<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Breeding Value and Allele Count Calculators</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.05</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">Breeding Value and Allele Count Calculators</div></div>
</div><!--header-->
<div class="contents">

<p>For calculations related to the loaded allele effects and internal additive breeding value model.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga3b43d07c08ed861ed0f22a69bb57b94a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a">gsc_split_by_bv</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const int top_n, const int lowIsBest)</td></tr>
<tr class="memdesc:ga3b43d07c08ed861ed0f22a69bb57b94a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group.  <a href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a">More...</a><br /></td></tr>
<tr class="separator:ga3b43d07c08ed861ed0f22a69bb57b94a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac94e98cc238a772aaa6151400df71629"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gac94e98cc238a772aaa6151400df71629">gsc_calculate_bvs</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gac94e98cc238a772aaa6151400df71629"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in a certain group.  <a href="group__calculators.html#gac94e98cc238a772aaa6151400df71629">More...</a><br /></td></tr>
<tr class="separator:gac94e98cc238a772aaa6151400df71629"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36">gsc_calculate_utility_bvs</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, const <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *effset)</td></tr>
<tr class="memdesc:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the fitness metric/breeding value for a set of genotypes.  <a href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36">More...</a><br /></td></tr>
<tr class="separator:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafbf562b4479f030957c8955ada159b0e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gafbf562b4479f030957c8955ada159b0e">gsc_calculate_allele_counts</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const char allele)</td></tr>
<tr class="memdesc:gafbf562b4479f030957c8955ada159b0e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears.  <a href="group__calculators.html#gafbf562b4479f030957c8955ada159b0e">More...</a><br /></td></tr>
<tr class="separator:gafbf562b4479f030957c8955ada159b0e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8f765775d463c59f8437da21dad6fab4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4">gsc_calculate_utility_allele_counts</a> (const size_t n_markers, const size_t n_genotypes, const char **const genotypes, const char allele, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:ga8f765775d463c59f8437da21dad6fab4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears.  <a href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4">More...</a><br /></td></tr>
<tr class="separator:ga8f765775d463c59f8437da21dad6fab4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga19bbe4c5c9d867e9c627bbda7aa9bd22"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga19bbe4c5c9d867e9c627bbda7aa9bd22">gsc_calculate_utility_allele_counts_pair</a> (const size_t n_markers, const size_t n_genotypes, const char **const genotypes, const char allele, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts, const char allele2, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:ga19bbe4c5c9d867e9c627bbda7aa9bd22"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear.  <a href="group__calculators.html#ga19bbe4c5c9d867e9c627bbda7aa9bd22">More...</a><br /></td></tr>
<tr class="separator:ga19bbe4c5c9d867e9c627bbda7aa9bd22"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3f2aedf883cd7cadb039feac6bb28ef1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga3f2aedf883cd7cadb039feac6bb28ef1">gsc_create_evenlength_blocks_each_chr</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapid, const int n)</td></tr>
<tr class="memdesc:ga3f2aedf883cd7cadb039feac6bb28ef1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct.  <a href="group__calculators.html#ga3f2aedf883cd7cadb039feac6bb28ef1">More...</a><br /></td></tr>
<tr class="separator:ga3f2aedf883cd7cadb039feac6bb28ef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaf6f32f41ae7ca27c1a48f8947919794a">gsc_load_blocks</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *block_file)</td></tr>
<tr class="memdesc:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks.  <a href="group__calculators.html#gaf6f32f41ae7ca27c1a48f8947919794a">More...</a><br /></td></tr>
<tr class="separator:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7b5ad446b4cd825f6dca63e11e362749"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga7b5ad446b4cd825f6dca63e11e362749">gsc_calculate_group_local_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const char *output_file, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga7b5ad446b4cd825f6dca63e11e362749"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga7b5ad446b4cd825f6dca63e11e362749">More...</a><br /></td></tr>
<tr class="separator:ga7b5ad446b4cd825f6dca63e11e362749"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gada411a339f47920e091fed1d586b1425"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gada411a339f47920e091fed1d586b1425">gsc_calculate_local_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const char *output_file)</td></tr>
<tr class="memdesc:gada411a339f47920e091fed1d586b1425"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#gada411a339f47920e091fed1d586b1425">More...</a><br /></td></tr>
<tr class="separator:gada411a339f47920e091fed1d586b1425"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae63afe0edeaf9eab31a0328fc72fc600"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae63afe0edeaf9eab31a0328fc72fc600">gsc_calculate_optimal_haplotype</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gae63afe0edeaf9eab31a0328fc72fc600"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>.  <a href="group__calculators.html#gae63afe0edeaf9eab31a0328fc72fc600">More...</a><br /></td></tr>
<tr class="separator:gae63afe0edeaf9eab31a0328fc72fc600"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga145430273fd296e49528a27489f0e520"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga145430273fd296e49528a27489f0e520">gsc_calculate_optimal_possible_haplotype</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga145430273fd296e49528a27489f0e520"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the highest-breeding-value haplotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#ga145430273fd296e49528a27489f0e520">More...</a><br /></td></tr>
<tr class="separator:ga145430273fd296e49528a27489f0e520"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaf950c24b4c08a7557ea74a771ebdc5d8">gsc_calculate_optimal_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values.  <a href="group__calculators.html#gaf950c24b4c08a7557ea74a771ebdc5d8">More...</a><br /></td></tr>
<tr class="separator:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga1f9eb57dc514f86c2923f5f91a140c4f">gsc_calculate_optimal_possible_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the breeding value of the highest breeding-value genotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#ga1f9eb57dc514f86c2923f5f91a140c4f">More...</a><br /></td></tr>
<tr class="separator:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae2af55ef05a2bb020ebbb2a9581ea60c">gsc_calculate_minimal_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values.  <a href="group__calculators.html#gae2af55ef05a2bb020ebbb2a9581ea60c">More...</a><br /></td></tr>
<tr class="separator:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p >For calculations related to the loaded allele effects and internal additive breeding value model. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="gafbf562b4479f030957c8955ada159b0e" name="gafbf562b4479f030957c8955ada159b0e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gafbf562b4479f030957c8955ada159b0e">&#9670;&nbsp;</a></span>gsc_calculate_allele_counts()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> gsc_calculate_allele_counts </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears. </p>
<p >Returns the results in a calculated DecimalMatrix. This structure should be deleted with <em>delete_dmatrix</em> once there is no more use for it.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gaa3ae83304655bb5e3889c6920addc0f0">calculate_allele_counts</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>simulation object containing the genotypes whose alleles will be counted </td></tr>
    <tr><td class="paramname">group</td><td>NO_GROUP to count alleles of all genotypes in the simulation, or a specific group identifier to count only the alleles of members of that group. </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a (num genotypes x num markers) <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a>, with each cell containing the number of incidences of the allele. The first axis of the DecimalMatrix corresponds to candidate genotypes, the second axis corresponds to genetic markers. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08798">8798</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gafbf562b4479f030957c8955ada159b0e_cgraph.png" border="0" usemap="#agroup__calculators_gafbf562b4479f030957c8955ada159b0e_cgraph" alt=""/></div>
<map name="agroup__calculators_gafbf562b4479f030957c8955ada159b0e_cgraph" id="agroup__calculators_gafbf562b4479f030957c8955ada159b0e_cgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="5,144,144,188"/>
<area shape="rect" href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="215,5,354,49"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="192,73,377,101"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="194,125,375,153"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="229,177,340,205"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="220,229,349,257"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="425,247,568,291"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="436,125,557,153"/>
</map>
</div>

</div>
</div>
<a id="gac94e98cc238a772aaa6151400df71629" name="gac94e98cc238a772aaa6151400df71629"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac94e98cc238a772aaa6151400df71629">&#9670;&nbsp;</a></span>gsc_calculate_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> gsc_calculate_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in a certain group. </p>
<p >To calculate the breeding value, the number of copies of each allele at each marker for each genotype are counted. The counts of each allele are multiplied by the additive marker effect of that allele at that marker. Then, each genotype's allele effects across all markers are summed up.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga97373dbd0bb7c3a93715623796c90176">calculate_bvs</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object in which the marker effects and genotypes to be used to calculate breeding values are stored. </td></tr>
    <tr><td class="paramname">group</td><td>NO_GROUP to calculate breeding values for all genotypes in the simulation, or a group number to calculate breeding values for all members of that group. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A (1 x number of genotypes) <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> containing the breeding value scores </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08707">8707</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gac94e98cc238a772aaa6151400df71629_cgraph.png" border="0" usemap="#agroup__calculators_gac94e98cc238a772aaa6151400df71629_cgraph" alt=""/></div>
<map name="agroup__calculators_gac94e98cc238a772aaa6151400df71629_cgraph" id="agroup__calculators_gac94e98cc238a772aaa6151400df71629_cgraph">
<area shape="rect" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="5,537,133,565"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="191,288,357,316"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="417,485,598,513"/>
<area shape="rect" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049" title="Create a bidirectional iterator." alt="" coords="181,537,367,565"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="181,589,367,617"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="190,641,358,669"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="415,109,600,153"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="434,177,581,221"/>
<area shape="rect" href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="438,245,577,289"/>
<area shape="rect" href="group__calculators.html#ga19bbe4c5c9d867e9c627bbda7aa9bd22" title="Calculates the number of times at each marker that two particular alleles appear." alt="" coords="438,313,577,357"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="441,381,575,409"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="452,433,563,461"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="443,5,572,33"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="648,24,792,68"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="659,485,781,513"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="427,537,588,581"/>
</map>
</div>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gac94e98cc238a772aaa6151400df71629_icgraph.png" border="0" usemap="#agroup__calculators_gac94e98cc238a772aaa6151400df71629_icgraph" alt=""/></div>
<map name="agroup__calculators_gac94e98cc238a772aaa6151400df71629_icgraph" id="agroup__calculators_gac94e98cc238a772aaa6151400df71629_icgraph">
<area shape="rect" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="382,32,509,60"/>
<area shape="rect" href="group__collgetters.html#ga25310a2e8c9fbfabca0c90a355f0d522" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="202,5,334,33"/>
<area shape="rect" href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="211,57,325,85"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,49,154,93"/>
</map>
</div>

</div>
</div>
<a id="ga7b5ad446b4cd825f6dca63e11e362749" name="ga7b5ad446b4cd825f6dca63e11e362749"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7b5ad446b4cd825f6dca63e11e362749">&#9670;&nbsp;</a></span>gsc_calculate_group_local_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_calculate_group_local_bvs </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file. </p>
<p >This gives block effects/local BVs for each haplotype of each individual in the group.</p>
<p >The output file will have format:</p>
<p >[genotype name]_1 [effect for first block, first allele] [effect for second block, second allele] ...</p>
<p >[genotype name]_2 [effect for first block, second allele] [effect for second block, second allele] ...</p>
<p >[genotype 2 name]_1 ...</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga2d9017a62e931839266507f9ed6bb4af">calculate_group_local_bvs</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">b</td><td>struct containing the blocks to use </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename of the file to which output block effects/local BVs will be saved. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to split the top individuals. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09185">9185</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga7b5ad446b4cd825f6dca63e11e362749_cgraph.png" border="0" usemap="#agroup__calculators_ga7b5ad446b4cd825f6dca63e11e362749_cgraph" alt=""/></div>
<map name="agroup__calculators_ga7b5ad446b4cd825f6dca63e11e362749_cgraph" id="agroup__calculators_ga7b5ad446b4cd825f6dca63e11e362749_cgraph">
<area shape="rect" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness..." alt="" coords="5,101,146,145"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="203,5,352,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="409,84,546,112"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="212,109,343,137"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="201,161,354,189"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="194,213,361,241"/>
</map>
</div>

</div>
</div>
<a id="gada411a339f47920e091fed1d586b1425" name="gada411a339f47920e091fed1d586b1425"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gada411a339f47920e091fed1d586b1425">&#9670;&nbsp;</a></span>gsc_calculate_local_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_calculate_local_bvs </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file. </p>
<p >This gives block effects for each haplotype of each individual currently saved to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>.</p>
<p >The output file will have format:</p>
<p >[genotype name]_1 [effect for first block, first allele] [effect for second block, second allele] ...</p>
<p >[genotype name]_2 [effect for first block, second allele] [effect for second block, second allele] ...</p>
<p >[genotype 2 name]_1 ...</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga5449c4624faa804b0ec3f12aa5df6893">calculate_local_bvs</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object to which individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">b</td><td>struct containing the blocks to use </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename of the file to which output block effects/local BV will be saved. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09298">9298</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gada411a339f47920e091fed1d586b1425_cgraph.png" border="0" usemap="#agroup__calculators_gada411a339f47920e091fed1d586b1425_cgraph" alt=""/></div>
<map name="agroup__calculators_gada411a339f47920e091fed1d586b1425_cgraph" id="agroup__calculators_gada411a339f47920e091fed1d586b1425_cgraph">
<area shape="rect" title="Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the f..." alt="" coords="5,5,167,33"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="215,5,382,33"/>
</map>
</div>

</div>
</div>
<a id="gae2af55ef05a2bb020ebbb2a9581ea60c" name="gae2af55ef05a2bb020ebbb2a9581ea60c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gae2af55ef05a2bb020ebbb2a9581ea60c">&#9670;&nbsp;</a></span>gsc_calculate_minimal_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double gsc_calculate_minimal_bv </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values. </p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must be initialised with marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gafaf7fbd54f72849ec14e3e1c28940bac">calculate_minimal_bv</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing markers and marker effects. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the fitness metric/breeding value of the worst genotype. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09638">9638</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gae2af55ef05a2bb020ebbb2a9581ea60c_cgraph.png" border="0" usemap="#agroup__calculators_gae2af55ef05a2bb020ebbb2a9581ea60c_cgraph" alt=""/></div>
<map name="agroup__calculators_gae2af55ef05a2bb020ebbb2a9581ea60c_cgraph" id="agroup__calculators_gae2af55ef05a2bb020ebbb2a9581ea60c_cgraph">
<area shape="rect" title="Takes a look at the currently&#45;loaded effect values and returns the lowest possible breeding value any..." alt="" coords="5,5,178,33"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="226,5,393,33"/>
</map>
</div>

</div>
</div>
<a id="gaf950c24b4c08a7557ea74a771ebdc5d8" name="gaf950c24b4c08a7557ea74a771ebdc5d8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaf950c24b4c08a7557ea74a771ebdc5d8">&#9670;&nbsp;</a></span>gsc_calculate_optimal_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double gsc_calculate_optimal_bv </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values. </p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must be initialised with marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga522229d6a4495b72de6eec76811ac96c">calculate_optimal_bv</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing markers and marker effects. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the fitness metric/breeding value of the best/ideal genotype. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09516">9516</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gaf950c24b4c08a7557ea74a771ebdc5d8_cgraph.png" border="0" usemap="#agroup__calculators_gaf950c24b4c08a7557ea74a771ebdc5d8_cgraph" alt=""/></div>
<map name="agroup__calculators_gaf950c24b4c08a7557ea74a771ebdc5d8_cgraph" id="agroup__calculators_gaf950c24b4c08a7557ea74a771ebdc5d8_cgraph">
<area shape="rect" title="Takes a look at the currently&#45;loaded effect values and returns the highest possible breeding value an..." alt="" coords="5,5,175,33"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="223,5,390,33"/>
</map>
</div>

</div>
</div>
<a id="gae63afe0edeaf9eab31a0328fc72fc600" name="gae63afe0edeaf9eab31a0328fc72fc600"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gae63afe0edeaf9eab31a0328fc72fc600">&#9670;&nbsp;</a></span>gsc_calculate_optimal_haplotype()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char * gsc_calculate_optimal_haplotype </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>. </p>
<p >Returns this as a null-terminated heap array of characters of length d-&gt;n_markers + one null byte.</p>
<p >The return value should be freed when usage is finished!</p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must be initialised with marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga75f9620f17bc4f791c7bd55fc3d12eed">calculate_optimal_haplotype</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing markers and marker effects. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap array filled with a null-terminated string containing the highest scoring allele at each marker. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09400">9400</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gae63afe0edeaf9eab31a0328fc72fc600_cgraph.png" border="0" usemap="#agroup__calculators_gae63afe0edeaf9eab31a0328fc72fc600_cgraph" alt=""/></div>
<map name="agroup__calculators_gae63afe0edeaf9eab31a0328fc72fc600_cgraph" id="agroup__calculators_gae63afe0edeaf9eab31a0328fc72fc600_cgraph">
<area shape="rect" title="Takes a look at the currently&#45;loaded effect values and creates a string containing the allele with th..." alt="" coords="5,24,156,68"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="204,5,371,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="227,57,348,85"/>
</map>
</div>

</div>
</div>
<a id="ga1f9eb57dc514f86c2923f5f91a140c4f" name="ga1f9eb57dc514f86c2923f5f91a140c4f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga1f9eb57dc514f86c2923f5f91a140c4f">&#9670;&nbsp;</a></span>gsc_calculate_optimal_possible_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double gsc_calculate_optimal_possible_bv </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the breeding value of the highest breeding-value genotype that can be created from the alleles present in a given group. </p>
<p >The highest value genotype is completely homozygous with the same alleles as the haplotype from <a class="el" href="group__calculators.html#ga145430273fd296e49528a27489f0e520" title="Calculates the highest-breeding-value haplotype that can be created from the alleles present in a giv...">gsc_calculate_optimal_possible_haplotype()</a>, because of the additive model of trait effects.</p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must be initialised with marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gac5506783193ab0351cee3056c2a72baf">calculate_optimal_possible_bv</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing markers and marker effects. </td></tr>
    <tr><td class="paramname">group</td><td>number to look at the available alleles in </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the fitness metric/breeding value of the best genotype </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09558">9558</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga1f9eb57dc514f86c2923f5f91a140c4f_cgraph.png" border="0" usemap="#agroup__calculators_ga1f9eb57dc514f86c2923f5f91a140c4f_cgraph" alt=""/></div>
<map name="agroup__calculators_ga1f9eb57dc514f86c2923f5f91a140c4f_cgraph" id="agroup__calculators_ga1f9eb57dc514f86c2923f5f91a140c4f_cgraph">
<area shape="rect" title="Calculates the breeding value of the highest breeding&#45;value genotype that can be created from the all..." alt="" coords="5,49,156,93"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="213,5,362,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="419,32,556,60"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="204,109,371,137"/>
</map>
</div>

</div>
</div>
<a id="ga145430273fd296e49528a27489f0e520" name="ga145430273fd296e49528a27489f0e520"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga145430273fd296e49528a27489f0e520">&#9670;&nbsp;</a></span>gsc_calculate_optimal_possible_haplotype()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char * gsc_calculate_optimal_possible_haplotype </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the highest-breeding-value haplotype that can be created from the alleles present in a given group. </p>
<p >The return value should be freed when usage is finished!</p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must be initialised with marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga85ed426b29c4d098fb9c503e7c9878f0">calculate_optimal_possible_haplotype</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing markers and marker effects. </td></tr>
    <tr><td class="paramname">group</td><td>group number to look at the available alleles in </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap array filled with a null-terminated string containing the highest scoring allele at each marker. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09441">9441</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga145430273fd296e49528a27489f0e520_cgraph.png" border="0" usemap="#agroup__calculators_ga145430273fd296e49528a27489f0e520_cgraph" alt=""/></div>
<map name="agroup__calculators_ga145430273fd296e49528a27489f0e520_cgraph" id="agroup__calculators_ga145430273fd296e49528a27489f0e520_cgraph">
<area shape="rect" title="Calculates the highest&#45;breeding&#45;value haplotype that can be created from the alleles present in a giv..." alt="" coords="5,76,156,120"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="213,5,362,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="419,32,556,60"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="204,109,371,137"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="227,161,348,189"/>
</map>
</div>

</div>
</div>
<a id="ga8f765775d463c59f8437da21dad6fab4" name="ga8f765775d463c59f8437da21dad6fab4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8f765775d463c59f8437da21dad6fab4">&#9670;&nbsp;</a></span>gsc_calculate_utility_allele_counts()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_calculate_utility_allele_counts </td>
          <td>(</td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>n_markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>n_genotypes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **const&#160;</td>
          <td class="paramname"><em>genotypes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>counts</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears. </p>
<p >Saves the result to a <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> provided by the calling context.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_markers</td><td>number of genetic markers in each genotype. </td></tr>
    <tr><td class="paramname">n_genotypes</td><td>number of genotypes to calculate counts for. (length of the following vector) </td></tr>
    <tr><td class="paramname">genotypes</td><td>pointers to genotype strings for which to count the alleles at particular markers. </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
    <tr><td class="paramname">counts</td><td>pointer to the <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> into which to put the number of <code>allele</code> occurences at each row/marker for each column/genotype in the group. The DecimalMatrix should have at least <em>n_genotypes</em> rows and <em>n_markers</em> columns, if not, no output will be saved. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08837">8837</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga8f765775d463c59f8437da21dad6fab4_icgraph.png" border="0" usemap="#agroup__calculators_ga8f765775d463c59f8437da21dad6fab4_icgraph" alt=""/></div>
<map name="agroup__calculators_ga8f765775d463c59f8437da21dad6fab4_icgraph" id="agroup__calculators_ga8f765775d463c59f8437da21dad6fab4_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="1292,51,1432,95"/>
<area shape="rect" href="group__calculators.html#gafbf562b4479f030957c8955ada159b0e" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="1093,20,1231,64"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="1079,88,1244,116"/>
<area shape="rect" href="group__calculators.html#gac94e98cc238a772aaa6151400df71629" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="898,57,1025,85"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="892,117,1031,145"/>
<area shape="rect" href="group__collgetters.html#ga25310a2e8c9fbfabca0c90a355f0d522" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="700,5,833,33"/>
<area shape="rect" href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="504,33,617,61"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,52,174,96"/>
<area shape="rect" href="sim-operations_8c.html#a7072c6a3834472e7fbd768c0275fdc54" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="689,109,844,153"/>
<area shape="rect" href="group__savers.html#ga6148ecde2587d6d37c313cfb88401771" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="716,177,817,205"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,133,641,177"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,37,400,65"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,89,431,117"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,141,429,169"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,193,429,237"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,261,431,289"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,313,398,341"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,193,192,237"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,261,195,289"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,313,190,357"/>
</map>
</div>

</div>
</div>
<a id="ga19bbe4c5c9d867e9c627bbda7aa9bd22" name="ga19bbe4c5c9d867e9c627bbda7aa9bd22"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga19bbe4c5c9d867e9c627bbda7aa9bd22">&#9670;&nbsp;</a></span>gsc_calculate_utility_allele_counts_pair()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_calculate_utility_allele_counts_pair </td>
          <td>(</td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>n_markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>n_genotypes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **const&#160;</td>
          <td class="paramname"><em>genotypes</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>counts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>counts2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that two particular alleles appear. </p>
<p >Saves the results for each allele to a <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> provided by the calling context.</p>
<p >This function is the same as <em>gsc_calculate_utility_allele_counts</em>, but for two alleles at a time, for the purpose of loop unrolling in breeding value calculations.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_markers</td><td>number of genetic markers in each genotype. </td></tr>
    <tr><td class="paramname">n_genotypes</td><td>number of genotypes to calculate counts for. (length of the following vector) </td></tr>
    <tr><td class="paramname">genotypes</td><td>pointers to genotype strings for which to count the alleles at particular markers. </td></tr>
    <tr><td class="paramname">allele</td><td>the first single-character allele to be counting. </td></tr>
    <tr><td class="paramname">counts</td><td>pointer to the <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> into which to put the number of <code>allele</code> occurences at each row/marker for each column/genotype in the group. The DecimalMatrix should have at least <em>n_genotypes</em> rows and <em>n_markers</em> columns, if not, no output will be saved. </td></tr>
    <tr><td class="paramname">allele2</td><td>the second single-character allele to be counting. </td></tr>
    <tr><td class="paramname">counts2</td><td>pointer to the <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> into which to put the number of <code>allele2</code> occurences at each row/marker for each column/genotype in the group. The DecimalMatrix should have at least <em>n_genotypes</em> rows and <em>n_markers</em> columns, if not, no output will be saved. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08881">8881</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga19bbe4c5c9d867e9c627bbda7aa9bd22_icgraph.png" border="0" usemap="#agroup__calculators_ga19bbe4c5c9d867e9c627bbda7aa9bd22_icgraph" alt=""/></div>
<map name="agroup__calculators_ga19bbe4c5c9d867e9c627bbda7aa9bd22_icgraph" id="agroup__calculators_ga19bbe4c5c9d867e9c627bbda7aa9bd22_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that two particular alleles appear." alt="" coords="1292,80,1432,124"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="1079,88,1244,116"/>
<area shape="rect" href="group__calculators.html#gac94e98cc238a772aaa6151400df71629" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="898,57,1025,85"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="892,117,1031,145"/>
<area shape="rect" href="group__collgetters.html#ga25310a2e8c9fbfabca0c90a355f0d522" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="700,5,833,33"/>
<area shape="rect" href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="504,33,617,61"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,52,174,96"/>
<area shape="rect" href="sim-operations_8c.html#a7072c6a3834472e7fbd768c0275fdc54" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="689,109,844,153"/>
<area shape="rect" href="group__savers.html#ga6148ecde2587d6d37c313cfb88401771" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="716,177,817,205"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,133,641,177"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,37,400,65"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,89,431,117"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,141,429,169"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,193,429,237"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,261,431,289"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,313,398,341"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,193,192,237"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,261,195,289"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,313,190,357"/>
</map>
</div>

</div>
</div>
<a id="ga5eb332e54b2f90f399a28a65cdb25e36" name="ga5eb332e54b2f90f399a28a65cdb25e36"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga5eb332e54b2f90f399a28a65cdb25e36">&#9670;&nbsp;</a></span>gsc_calculate_utility_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> gsc_calculate_utility_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>targets</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *&#160;</td>
          <td class="paramname"><em>effset</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculate the fitness metric/breeding value for a set of genotypes. </p>
<p >The breeding value is calculated as the sum, across all alleles, of the (genotypes x markers) count matrix for that allele multiplied by the (markers x 1) marker effects for that allele.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">targets</td><td>Iterator for the genotypes whose breeding values are to be calculated. </td></tr>
    <tr><td class="paramname">effset</td><td>Set of additive marker effects to use to calculate the breeding values. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>(1 x number of genotypes in iterator) <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a>, containing the breeding value scores </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08734">8734</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_cgraph.png" border="0" usemap="#agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_cgraph" alt=""/></div>
<map name="agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_cgraph" id="agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_cgraph">
<area shape="rect" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="5,277,171,305"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="219,5,404,49"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="238,73,385,117"/>
<area shape="rect" href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="242,141,381,185"/>
<area shape="rect" href="group__calculators.html#ga19bbe4c5c9d867e9c627bbda7aa9bd22" title="Calculates the number of times at each marker that two particular alleles appear." alt="" coords="242,209,381,253"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="244,277,379,305"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="221,329,402,357"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="256,381,367,409"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="247,433,376,461"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="452,451,595,495"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="463,329,584,357"/>
</map>
</div>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_icgraph.png" border="0" usemap="#agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_icgraph" alt=""/></div>
<map name="agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_icgraph" id="agroup__calculators_ga5eb332e54b2f90f399a28a65cdb25e36_icgraph">
<area shape="rect" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="1079,88,1244,116"/>
<area shape="rect" href="group__calculators.html#gac94e98cc238a772aaa6151400df71629" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="898,57,1025,85"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="892,117,1031,145"/>
<area shape="rect" href="group__collgetters.html#ga25310a2e8c9fbfabca0c90a355f0d522" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="700,5,833,33"/>
<area shape="rect" href="group__calculators.html#ga3b43d07c08ed861ed0f22a69bb57b94a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="504,33,617,61"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,52,174,96"/>
<area shape="rect" href="sim-operations_8c.html#a7072c6a3834472e7fbd768c0275fdc54" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="689,109,844,153"/>
<area shape="rect" href="group__savers.html#ga6148ecde2587d6d37c313cfb88401771" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="716,177,817,205"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,133,641,177"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,37,400,65"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,89,431,117"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,141,429,169"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,193,429,237"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,261,431,289"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,313,398,341"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,193,192,237"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,261,195,289"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,313,190,357"/>
</map>
</div>

</div>
</div>
<a id="ga3f2aedf883cd7cadb039feac6bb28ef1" name="ga3f2aedf883cd7cadb039feac6bb28ef1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3f2aedf883cd7cadb039feac6bb28ef1">&#9670;&nbsp;</a></span>gsc_create_evenlength_blocks_each_chr()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> gsc_create_evenlength_blocks_each_chr </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td>
          <td class="paramname"><em>mapid</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct. </p>
<p >Chromosomes where there are no markers tracked are excluded/no blocks are created in those chromosomes.</p>
<p >Chromosomes where only one marker is tracked put the marker in the first block, and have all other blocks empty.</p>
<p >Empty blocks have blocks.num_markers_in_block[index] equal to 0 and contain a null pointer at blocks.markers_in_block[index].</p>
<p >Remember to call the <a class="el" href="structgsc___marker_blocks.html" title="A struct used to store a set of blocks of markers.">gsc_MarkerBlocks</a> destructor <a class="el" href="group__deletors.html#ga8583ccabd176c47c4126f164321e0126" title="Delete a gsc_MarkerBlocks struct.">gsc_delete_markerblocks()</a> on the returned struct.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gad9411be8f1d526298252e8d46b66760d">create_evenlength_blocks_each_chr</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">mapid</td><td>ID of the recombination map whose markers and chromosomes will be divided into blocks, or NO_MAP to use the first-loaded/primary map by default. </td></tr>
    <tr><td class="paramname">mapid</td><td>recombination map to use to determine block lengths and marker allocations to blocks. If this value is NO_MAP, uses first/primary map. </td></tr>
    <tr><td class="paramname">n</td><td>number of blocks into which to split each chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct containing the markers identified as belonging to each block. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08936">8936</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga3f2aedf883cd7cadb039feac6bb28ef1_cgraph.png" border="0" usemap="#agroup__calculators_ga3f2aedf883cd7cadb039feac6bb28ef1_cgraph" alt=""/></div>
<map name="agroup__calculators_ga3f2aedf883cd7cadb039feac6bb28ef1_cgraph" id="agroup__calculators_ga3f2aedf883cd7cadb039feac6bb28ef1_cgraph">
<area shape="rect" title="Divide the genotype into blocks where each block contains all markers within a 1/n length section of ..." alt="" coords="5,24,159,68"/>
<area shape="rect" href="group__supporters.html#gaaa359f6aab15a4c27d344e1118e87457" title="Function to identify the lookup index of a recombination map identifier." alt="" coords="207,5,361,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="223,57,345,85"/>
</map>
</div>

</div>
</div>
<a id="gaf6f32f41ae7ca27c1a48f8947919794a" name="gaf6f32f41ae7ca27c1a48f8947919794a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaf6f32f41ae7ca27c1a48f8947919794a">&#9670;&nbsp;</a></span>gsc_load_blocks()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> gsc_load_blocks </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>block_file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks. </p>
<p >The block file is designed after the output from a call to the R SelectionTools package's <code>st.def.hblocks</code> function. It should have the format (tab-separated):</p>
<p >Chrom Pos Name Class Markers</p>
<p >[ignored] [ignored] [ignored] [ignored] [semicolon];[separated];[list] ;[of];[marker];[names];[belonging];[to];[this];[block];</p>
<p >...</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga9bf0f856deff9404c9963bbf02204fd0">load_blocks</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">block_file</td><td>string containing filename of the file with blocks </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct containing the markers identified as belonging to each block according to their definitions in the file. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l09089">9089</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-15" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-15-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-15-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-15-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_gaf6f32f41ae7ca27c1a48f8947919794a_cgraph.png" border="0" usemap="#agroup__calculators_gaf6f32f41ae7ca27c1a48f8947919794a_cgraph" alt=""/></div>
<map name="agroup__calculators_gaf6f32f41ae7ca27c1a48f8947919794a_cgraph" id="agroup__calculators_gaf6f32f41ae7ca27c1a48f8947919794a_cgraph">
<area shape="rect" title="Given a file containing definitions of blocks of markers, process that file and return a struct conta..." alt="" coords="5,65,124,93"/>
<area shape="rect" href="group__supporters.html#ga31fba32c2f869b50d97669e3e322d338" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="172,5,335,33"/>
<area shape="rect" href="group__supporters.html#ga6d0f2bc0f68056eed39469166167fefb" title="Linear search through a list of strings." alt="" coords="172,57,335,101"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="193,125,314,153"/>
</map>
</div>

</div>
</div>
<a id="ga3b43d07c08ed861ed0f22a69bb57b94a" name="ga3b43d07c08ed861ed0f22a69bb57b94a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3b43d07c08ed861ed0f22a69bb57b94a">&#9670;&nbsp;</a></span>gsc_split_by_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_split_by_bv </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>top_n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>lowIsBest</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group. </p>
<p >The new group number is returned.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to split the top individuals. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the set of marker effects to use to calculate BV. </td></tr>
    <tr><td class="paramname">top_n</td><td>The number of individuals to put in the new group. </td></tr>
    <tr><td class="paramname">lowIsBest</td><td>boolean, if TRUE the <code>top_n</code> with the lowest breeding value will be selected, if false the <code>top_n</code> with the highest breeding value are. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the newly-created split-off group </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08643">8643</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-16" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-16-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-16-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-16-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_cgraph.png" border="0" usemap="#agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_cgraph" alt=""/></div>
<map name="agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_cgraph" id="agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_cgraph">
<area shape="rect" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="5,651,119,679"/>
<area shape="rect" href="group__calculators.html#gac94e98cc238a772aaa6151400df71629" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="182,444,310,472"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="649,561,783,589"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="391,521,558,549"/>
<area shape="rect" href="group__collgetters.html#ga90490b08f70757c55e06b842abc5d1ce" title="Gets the indexes (0&#45;based, from the start of the linked list in the gsc_SimData) of each member of th..." alt="" coords="167,624,325,652"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="406,629,543,657"/>
<area shape="rect" href="sim-operations_8c.html#a3c42957f4439d2f2b00d9cc2fa54c6c7" title="Comparator function for qsort." alt="" coords="171,728,321,772"/>
<area shape="rect" href="sim-operations_8c.html#a4d1427e76e7fd129c6585493e80be049" title="Comparator function for qsort." alt="" coords="167,796,325,840"/>
<area shape="rect" href="group__groupmod.html#ga9a3bd44f93ffc4ee634a9c5d1aeb59a2" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="170,864,322,892"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="392,269,557,297"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="625,432,807,460"/>
<area shape="rect" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049" title="Create a bidirectional iterator." alt="" coords="1072,671,1258,699"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="382,367,567,395"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="623,160,809,204"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="642,228,790,272"/>
<area shape="rect" href="group__calculators.html#ga8f765775d463c59f8437da21dad6fab4" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="646,296,786,340"/>
<area shape="rect" href="group__calculators.html#ga19bbe4c5c9d867e9c627bbda7aa9bd22" title="Calculates the number of times at each marker that two particular alleles appear." alt="" coords="646,364,786,408"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="660,5,772,33"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="1101,109,1230,137"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="1315,76,1459,120"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="880,639,1001,667"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="1306,663,1468,707"/>
<area shape="rect" href="group__iterators.html#gaa26377edd1339407af2070924d3596de" title="Create a Random Access Iterator." alt="" coords="373,689,575,717"/>
<area shape="rect" href="group__deletors.html#ga3e88311f64b089f3d7a336f4efee3b05" title="Deletes a gsc_RandomAccessIterator object and frees its memory." alt="" coords="374,756,575,784"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="390,808,558,836"/>
<area shape="rect" href="group__iterators.html#ga5b9b804b6c3b0c165f20a765942fa810" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="654,785,778,813"/>
<area shape="rect" href="group__liteget.html#ga7c88af63ccbce0c1e9026339d38fe53a" title="Set the current group membership of a genotype." alt="" coords="421,912,527,940"/>
<area shape="rect" href="group__collgetters.html#gaf9692ce014cfe5439fd00e522a6bab9f" title="Identify group numbers that currently have members." alt="" coords="633,716,799,744"/>
<area shape="rect" href="group__collgetters.html#gaddb8ea59b22bbe0ec8c61da7f5ba8a98" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="861,691,1020,735"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="1112,727,1218,755"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="857,811,1024,839"/>
</map>
</div>
<div id="dynsection-17" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-17-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-17-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-17-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_icgraph.png" border="0" usemap="#agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_icgraph" alt=""/></div>
<map name="agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_icgraph" id="agroup__calculators_ga3b43d07c08ed861ed0f22a69bb57b94a_icgraph">
<area shape="rect" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="202,13,315,41"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,154,49"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
