<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Recombination Calculators</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">genomicSimulationC
   &#160;<span id="projectnumber">0.1-4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Recombination Calculators</div>  </div>
</div><!--header-->
<div class="contents">

<p>Experimental functions for retroactively calculating number of recombinations.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:gaad9cd4248231da3682ba2f08a40cdbe8"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gaad9cd4248231da3682ba2f08a40cdbe8">calculate_min_recombinations_fw1</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int certain)</td></tr>
<tr class="memdesc:gaad9cd4248231da3682ba2f08a40cdbe8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured.  <a href="group__recomb.html#gaad9cd4248231da3682ba2f08a40cdbe8">More...</a><br /></td></tr>
<tr class="separator:gaad9cd4248231da3682ba2f08a40cdbe8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3627a380ed86ebebe937c953bc3b2339"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga3627a380ed86ebebe937c953bc3b2339">calculate_min_recombinations_fwn</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int window_size, int certain)</td></tr>
<tr class="memdesc:ga3627a380ed86ebebe937c953bc3b2339"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it.  <a href="group__recomb.html#ga3627a380ed86ebebe937c953bc3b2339">More...</a><br /></td></tr>
<tr class="separator:ga3627a380ed86ebebe937c953bc3b2339"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa71cdbced90506820ef61bd842ecfa40"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gaa71cdbced90506820ef61bd842ecfa40">has_same_alleles</a> (char *p1, char *p2, int i)</td></tr>
<tr class="memdesc:gaa71cdbced90506820ef61bd842ecfa40"><td class="mdescLeft">&#160;</td><td class="mdescRight">Simple operator to determine if at marker i, two genotypes share at least one allele.  <a href="group__recomb.html#gaa71cdbced90506820ef61bd842ecfa40">More...</a><br /></td></tr>
<tr class="separator:gaa71cdbced90506820ef61bd842ecfa40"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab88223532880a22dfc966be4cdb4107b"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gab88223532880a22dfc966be4cdb4107b">has_same_alleles_window</a> (char *g1, char *g2, int start, int w)</td></tr>
<tr class="memdesc:gab88223532880a22dfc966be4cdb4107b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at least one allele.  <a href="group__recomb.html#gab88223532880a22dfc966be4cdb4107b">More...</a><br /></td></tr>
<tr class="separator:gab88223532880a22dfc966be4cdb4107b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">calculate_recombinations_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const char *output_file, int window_len, int certain)</td></tr>
<tr class="memdesc:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes.  <a href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">More...</a><br /></td></tr>
<tr class="separator:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Experimental functions for retroactively calculating number of recombinations. </p>
<p>This functionality is for interest only. It is not clear, or tidy, or checked against real data. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="gaad9cd4248231da3682ba2f08a40cdbe8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaad9cd4248231da3682ba2f08a40cdbe8">&#9670;&nbsp;</a></span>calculate_min_recombinations_fw1()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int* calculate_min_recombinations_fw1 </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>offspring</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured. </p>
<p>This function is a little lower-level (see the kinds of parameters required) and so a wrapper like calculate_recombinations_from_file is suggested for end users. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot...">calculate_recombinations_from_file()</a></dd></dl>
<p>The function reads start to end along each chromosome. At each marker, it checks if the alleles the offspring has could only have come from one parent/there is known parentage of that allele. If that is the case, it saves the provided id number of the source parent to the matching position in the result vector. If it is not the case, its behaviour depends on the <code>certain</code> parameter.</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct whose genetic map matches the provided genotypes. </td></tr>
    <tr><td class="paramname">parent1</td><td>a character vector containing one parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p1num</td><td>an integer that will be used to identify areas of the genome that come from the first parent in the returned vector. </td></tr>
    <tr><td class="paramname">parent2</td><td>a character vector containing the other parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p2num</td><td>an integer that will be used to identify areas of the genome that come from the second parent in the returned vector. </td></tr>
    <tr><td class="paramname">offspring</td><td>a character vector containing the alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> of the genotype whose likely recombinations we want to identify. </td></tr>
    <tr><td class="paramname">certain</td><td>a boolean. If TRUE, markers where the parent of origin cannot be identified will be set to 0, if FALSE, the value will be set to the id of the parent that provided the most recently identified allele in that chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap vector of length <code>d-&gt;n_markers</code> containing the id of the parent of origin at each marker in the <code>offspring</code> genotype. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_cgraph.png" border="0" usemap="#group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_cgraph" alt=""/></div>
<map name="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_cgraph" id="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_cgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="5,5,228,32"/>
<area shape="rect" href="group__recomb.html#gaa71cdbced90506820ef61bd842ecfa40" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="276,5,403,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_icgraph.png" border="0" usemap="#group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_icgraph" alt=""/></div>
<map name="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_icgraph" id="group__recomb_gaad9cd4248231da3682ba2f08a40cdbe8_icgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,13,445,39"/>
<area shape="rect" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="ga3627a380ed86ebebe937c953bc3b2339"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3627a380ed86ebebe937c953bc3b2339">&#9670;&nbsp;</a></span>calculate_min_recombinations_fwn()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int* calculate_min_recombinations_fwn </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>offspring</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>window_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it. </p>
<p>This function is a little lower-level (see the kinds of parameters required) and so a wrapper like calculate_recombinations_from_file is suggested for end users. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot...">calculate_recombinations_from_file()</a></dd></dl>
<p>The function reads start to end along each chromosome. At each marker, it checks if the alleles the offspring has in the window centered at that marker could have come from one parent but could not have come from the other/there is known parentage of that allele. If that is the case, it saves the provided id number of the source parent to the matching position in the result vector. If it is not the case, its behaviour depends on the <code>certain</code> parameter.</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<p>Behaviour when the window size is not an odd integer has not been tested.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct whose genetic map matches the provided genotypes. </td></tr>
    <tr><td class="paramname">parent1</td><td>a character vector containing one parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p1num</td><td>an integer that will be used to identify areas of the genome that come from the first parent in the returned vector. </td></tr>
    <tr><td class="paramname">parent2</td><td>a character vector containing the other parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p2num</td><td>an integer that will be used to identify areas of the genome that come from the second parent in the returned vector. </td></tr>
    <tr><td class="paramname">offspring</td><td>a character vector containing the alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> of the genotype whose likely recombinations we want to identify. </td></tr>
    <tr><td class="paramname">window_size</td><td>an odd integer representing the number of markers to check for known parentage around each marker </td></tr>
    <tr><td class="paramname">certain</td><td>a boolean. If TRUE, markers where the parent of origin cannot be identified will be set to 0, if FALSE, the value will be set to the id of the parent that provided the most recently identified allele in that chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap vector of length <code>d-&gt;n_markers</code> containing the id of the parent of origin at each marker in the <code>offspring</code> genotype. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_cgraph.png" border="0" usemap="#group__recomb_ga3627a380ed86ebebe937c953bc3b2339_cgraph" alt=""/></div>
<map name="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_cgraph" id="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_cgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="5,5,228,32"/>
<area shape="rect" href="group__recomb.html#gab88223532880a22dfc966be4cdb4107b" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="276,5,451,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_icgraph.png" border="0" usemap="#group__recomb_ga3627a380ed86ebebe937c953bc3b2339_icgraph" alt=""/></div>
<map name="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_icgraph" id="group__recomb_ga3627a380ed86ebebe937c953bc3b2339_icgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,13,445,39"/>
<area shape="rect" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="ga7813f256407ab2ea242df32ea9ad2c7e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7813f256407ab2ea242df32ea9ad2c7e">&#9670;&nbsp;</a></span>calculate_recombinations_from_file()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int calculate_recombinations_from_file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>input_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>window_len</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes. </p>
<p>The input file (which pairs up which targets and their parents the calculation should be carried out on) should have format:</p>
<p>[target name] [parent1name] [parent2name]</p>
<p>[target name] [parent1name] [parent2name]</p>
<p>...</p>
<p>The tab-separated output file produced by this function will have format: </p><pre class="fragment">[marker 1 name] [marker 2 name]...
</pre><p>[target name] [tab-separated recombination vector, containing the index at each marker of the parent the function guesses the target's alleles came from, or 0 if this is unknow]</p>
<p>...</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct containing the genotypes and map under consideration. </td></tr>
    <tr><td class="paramname">input_file</td><td>string containing the name of the file with the pairs of parents and offsprings of which to calculate recombinations </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename to which to save the results. </td></tr>
    <tr><td class="paramname">window_len</td><td>an odd integer representing the number of markers to check for known parentage around each marker </td></tr>
    <tr><td class="paramname">certain</td><td>TRUE to fill locations where parentage is unknown with 0, FALSE to fill locations where parentage is unknown with the most recent known parent </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 on success. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_ga7813f256407ab2ea242df32ea9ad2c7e_cgraph.png" border="0" usemap="#group__recomb_ga7813f256407ab2ea242df32ea9ad2c7e_cgraph" alt=""/></div>
<map name="group__recomb_ga7813f256407ab2ea242df32ea9ad2c7e_cgraph" id="group__recomb_ga7813f256407ab2ea242df32ea9ad2c7e_cgraph">
<area shape="rect" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,125,175,166"/>
<area shape="rect" href="group__recomb.html#gaad9cd4248231da3682ba2f08a40cdbe8" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,5,445,32"/>
<area shape="rect" href="group__recomb.html#ga3627a380ed86ebebe937c953bc3b2339" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,56,445,83"/>
<area shape="rect" href="group__supporters.html#ga392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="266,107,402,133"/>
<area shape="rect" href="group__getters.html#ga309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="265,157,403,184"/>
<area shape="rect" href="group__getters.html#ga67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="278,208,390,235"/>
<area shape="rect" href="group__getters.html#ga93779d37ca9081d07987fd83c27b84af" title="Search for a genotype with a particular name in a linked list of AlleleMatrix, and return its index." alt="" coords="267,259,401,285"/>
<area shape="rect" href="group__recomb.html#gaa71cdbced90506820ef61bd842ecfa40" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="517,5,644,32"/>
<area shape="rect" href="group__recomb.html#gab88223532880a22dfc966be4cdb4107b" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="493,56,668,83"/>
</map>
</div>

</div>
</div>
<a id="gaa71cdbced90506820ef61bd842ecfa40"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa71cdbced90506820ef61bd842ecfa40">&#9670;&nbsp;</a></span>has_same_alleles()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int has_same_alleles </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>p2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Simple operator to determine if at marker i, two genotypes share at least one allele. </p>
<p>Checks only 3 of four possible permutations because assumes there cannot be more than two alleles at a given marker.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">p1</td><td>pointer to a character array genotype of the type stored in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (2*n_markers long, representing the two alleles at a marker consecutively) for the first of the genotypes to compare. </td></tr>
    <tr><td class="paramname">p2</td><td>pointer to a character array genotype for the second of the genotypes to compare. </td></tr>
    <tr><td class="paramname">i</td><td>index of the marker at which to perform the check </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>boolean result of the check </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_gaa71cdbced90506820ef61bd842ecfa40_icgraph.png" border="0" usemap="#group__recomb_gaa71cdbced90506820ef61bd842ecfa40_icgraph" alt=""/></div>
<map name="group__recomb_gaa71cdbced90506820ef61bd842ecfa40_icgraph" id="group__recomb_gaa71cdbced90506820ef61bd842ecfa40_icgraph">
<area shape="rect" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="493,13,620,39"/>
<area shape="rect" href="group__recomb.html#gaad9cd4248231da3682ba2f08a40cdbe8" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,13,445,39"/>
<area shape="rect" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="gab88223532880a22dfc966be4cdb4107b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab88223532880a22dfc966be4cdb4107b">&#9670;&nbsp;</a></span>has_same_alleles_window()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int has_same_alleles_window </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>g1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>g2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>w</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at least one allele. </p>
<p>Checks only 3 of four possible permutations at each marker because assumes there cannot be more than two alleles at a given marker. For the return value to be true, there must be at least one match at every one of the markers in the window.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">g1</td><td>pointer to a character array genotype of the type stored in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (2*n_markers long, representing the two alleles at a marker consecutively) for the first of the genotypes to compare. </td></tr>
    <tr><td class="paramname">g2</td><td>pointer to a character array genotype for the second of the genotypes to compare. </td></tr>
    <tr><td class="paramname">start</td><td>index of the first marker in the window over which to perform the check </td></tr>
    <tr><td class="paramname">w</td><td>length of the window over which to perform the check </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>boolean result of the check </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__recomb_gab88223532880a22dfc966be4cdb4107b_icgraph.png" border="0" usemap="#group__recomb_gab88223532880a22dfc966be4cdb4107b_icgraph" alt=""/></div>
<map name="group__recomb_gab88223532880a22dfc966be4cdb4107b_icgraph" id="group__recomb_gab88223532880a22dfc966be4cdb4107b_icgraph">
<area shape="rect" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="493,13,668,39"/>
<area shape="rect" href="group__recomb.html#ga3627a380ed86ebebe937c953bc3b2339" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,13,445,39"/>
<area shape="rect" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
