<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Saving Functions</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.3</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">Saving Functions</div></div>
</div><!--header-->
<div class="contents">

<p>For saving persistent simulation results.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5c1144f08e16be0974f7a8f81b81c0d4">save_simdata</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga5c1144f08e16be0974f7a8f81b81c0d4">More...</a><br /></td></tr>
<tr class="separator:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga63e803fa3cdf0cbb209b2922ce2982fe">save_marker_blocks</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b)</td></tr>
<tr class="memdesc:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga63e803fa3cdf0cbb209b2922ce2982fe">More...</a><br /></td></tr>
<tr class="separator:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3e90f23745c5b9c68d85276e41fbe23f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga3e90f23745c5b9c68d85276e41fbe23f">save_allele_matrix</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char **markers)</td></tr>
<tr class="memdesc:ga3e90f23745c5b9c68d85276e41fbe23f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="group__savers.html#ga3e90f23745c5b9c68d85276e41fbe23f">More...</a><br /></td></tr>
<tr class="separator:ga3e90f23745c5b9c68d85276e41fbe23f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga8e14b8d8c1836e1402a4a5e7a74c274f">save_transposed_allele_matrix</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char **markers)</td></tr>
<tr class="memdesc:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="group__savers.html#ga8e14b8d8c1836e1402a4a5e7a74c274f">More...</a><br /></td></tr>
<tr class="separator:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaeb3b527db48fd67f0d42a39c681b2c65"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaeb3b527db48fd67f0d42a39c681b2c65">save_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, const int group_id)</td></tr>
<tr class="memdesc:gaeb3b527db48fd67f0d42a39c681b2c65"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#gaeb3b527db48fd67f0d42a39c681b2c65">More...</a><br /></td></tr>
<tr class="separator:gaeb3b527db48fd67f0d42a39c681b2c65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2790244044d202e6fb158d8f57cf6d6f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga2790244044d202e6fb158d8f57cf6d6f">save_transposed_group_alleles</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int group_id)</td></tr>
<tr class="memdesc:ga2790244044d202e6fb158d8f57cf6d6f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#ga2790244044d202e6fb158d8f57cf6d6f">More...</a><br /></td></tr>
<tr class="separator:ga2790244044d202e6fb158d8f57cf6d6f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga822c4500c4f5979c9eee8aa733434196"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga822c4500c4f5979c9eee8aa733434196">save_group_one_step_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int group)</td></tr>
<tr class="memdesc:ga822c4500c4f5979c9eee8aa733434196"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="group__savers.html#ga822c4500c4f5979c9eee8aa733434196">More...</a><br /></td></tr>
<tr class="separator:ga822c4500c4f5979c9eee8aa733434196"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5674dd2fbdcf205475c8036a09536f2e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5674dd2fbdcf205475c8036a09536f2e">save_one_step_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga5674dd2fbdcf205475c8036a09536f2e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga5674dd2fbdcf205475c8036a09536f2e">More...</a><br /></td></tr>
<tr class="separator:ga5674dd2fbdcf205475c8036a09536f2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7adbe41e3a1d04fbf2b32840e2d01f86"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga7adbe41e3a1d04fbf2b32840e2d01f86">save_group_full_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int group)</td></tr>
<tr class="memdesc:ga7adbe41e3a1d04fbf2b32840e2d01f86"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="group__savers.html#ga7adbe41e3a1d04fbf2b32840e2d01f86">More...</a><br /></td></tr>
<tr class="separator:ga7adbe41e3a1d04fbf2b32840e2d01f86"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga08c823d63ce4c648432cb740b5ab743b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga08c823d63ce4c648432cb740b5ab743b">save_full_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga08c823d63ce4c648432cb740b5ab743b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga08c823d63ce4c648432cb740b5ab743b">More...</a><br /></td></tr>
<tr class="separator:ga08c823d63ce4c648432cb740b5ab743b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf1a9b82d2978a1563daf31c8814a783"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783">save_AM_pedigree</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const <a class="el" href="struct_sim_data.html">SimData</a> *parents)</td></tr>
<tr class="memdesc:gabf1a9b82d2978a1563daf31c8814a783"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file.  <a href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783">More...</a><br /></td></tr>
<tr class="separator:gabf1a9b82d2978a1563daf31c8814a783"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaead4d69058f8fe7d59465100c801bbbd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaead4d69058f8fe7d59465100c801bbbd">save_parents_of</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const unsigned int p1, const unsigned int p2)</td></tr>
<tr class="memdesc:gaead4d69058f8fe7d59465100c801bbbd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="group__savers.html#gaead4d69058f8fe7d59465100c801bbbd">More...</a><br /></td></tr>
<tr class="separator:gaead4d69058f8fe7d59465100c801bbbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7cd7382db4574b2b0cffb316eab47a29"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga7cd7382db4574b2b0cffb316eab47a29">save_group_bvs</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int group)</td></tr>
<tr class="memdesc:ga7cd7382db4574b2b0cffb316eab47a29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in a group to a file.  <a href="group__savers.html#ga7cd7382db4574b2b0cffb316eab47a29">More...</a><br /></td></tr>
<tr class="separator:ga7cd7382db4574b2b0cffb316eab47a29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga378aa1f3cc71a3b3e5e17c9c103c4727"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga378aa1f3cc71a3b3e5e17c9c103c4727">save_manual_bvs</a> (FILE *f, const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *e, const unsigned int *ids, const char **names)</td></tr>
<tr class="memdesc:ga378aa1f3cc71a3b3e5e17c9c103c4727"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>.  <a href="group__savers.html#ga378aa1f3cc71a3b3e5e17c9c103c4727">More...</a><br /></td></tr>
<tr class="separator:ga378aa1f3cc71a3b3e5e17c9c103c4727"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa9a8f4e7b82dda77c7a688d83fa08e28"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaa9a8f4e7b82dda77c7a688d83fa08e28">save_bvs</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gaa9a8f4e7b82dda77c7a688d83fa08e28"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#gaa9a8f4e7b82dda77c7a688d83fa08e28">More...</a><br /></td></tr>
<tr class="separator:gaa9a8f4e7b82dda77c7a688d83fa08e28"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga38b0958216f509a3b0a10c4c49183f44"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga38b0958216f509a3b0a10c4c49183f44">save_count_matrix</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const char allele)</td></tr>
<tr class="memdesc:ga38b0958216f509a3b0a10c4c49183f44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga38b0958216f509a3b0a10c4c49183f44">More...</a><br /></td></tr>
<tr class="separator:ga38b0958216f509a3b0a10c4c49183f44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8bbfd2ed13f449e5934b91701cc94a9f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga8bbfd2ed13f449e5934b91701cc94a9f">save_count_matrix_of_group</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const char allele, const int group)</td></tr>
<tr class="memdesc:ga8bbfd2ed13f449e5934b91701cc94a9f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="group__savers.html#ga8bbfd2ed13f449e5934b91701cc94a9f">More...</a><br /></td></tr>
<tr class="separator:ga8bbfd2ed13f449e5934b91701cc94a9f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p >For saving persistent simulation results. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga3e90f23745c5b9c68d85276e41fbe23f" name="ga3e90f23745c5b9c68d85276e41fbe23f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3e90f23745c5b9c68d85276e41fbe23f">&#9670;&nbsp;</a></span>save_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file. </p>
<p >Uses the following format: </p><pre class="fragment">    [marker name]   [marker name]
</pre><p> [id]OR[name] [allele pairs for each marker]</p>
<p >[id]OR[name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. if this is null, the header row will be empty. </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga3e90f23745c5b9c68d85276e41fbe23f_icgraph.png" border="0" usemap="#agroup__savers_ga3e90f23745c5b9c68d85276e41fbe23f_icgraph" alt=""/></div>
<map name="agroup__savers_ga3e90f23745c5b9c68d85276e41fbe23f_icgraph" id="agroup__savers_ga3e90f23745c5b9c68d85276e41fbe23f_icgraph">
<area shape="rect" title="Prints all the genotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="439,161,566,189"/>
<area shape="rect" href="group__crossers.html#ga7f9865d8d1912cda7e8524d1ff761de6" title="Performs random crosses among members of a group." alt="" coords="218,25,388,53"/>
<area shape="rect" href="group__crossers.html#ga43f436342a1afc3b14343a8f6e9385f9" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="219,83,387,111"/>
<area shape="rect" href="sim-operations_8c.html#a42ded1341b4c5a03c2233e2a3875fe07" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="215,135,391,163"/>
<area shape="rect" href="group__crossers.html#ga5e1da39f8341127e758d32393e8654aa" title="Creates an identical copy of each member of a group." alt="" coords="254,187,352,215"/>
<area shape="rect" href="group__crossers.html#gab081b254e7abce63067b8c02fbac774f" title="Creates a doubled haploid from each member of a group." alt="" coords="223,239,383,267"/>
<area shape="rect" href="group__crossers.html#gaf0d8ffcde1a2098e7ae807a031bbe7b9" title="Selfs each member of a group for a certain number of generations." alt="" coords="256,291,350,319"/>
<area shape="rect" href="group__crossers.html#gaa41e75da9e6e4f8db5466e8929dec2ee" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,162,49"/>
<area shape="rect" href="group__crossers.html#ga9d79117014f247c078c6576c70d600fb" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,73,164,117"/>
<area shape="rect" href="group__crossers.html#ga7fafc033bf3aaa14c9d5affbe9421ca9" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,141,167,169"/>
<area shape="rect" href="group__crossers.html#gafb306165b0cf0f5cd752f3eaf4c60928" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,193,162,237"/>
</map>
</div>

</div>
</div>
<a id="gabf1a9b82d2978a1563daf31c8814a783" name="gabf1a9b82d2978a1563daf31c8814a783"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabf1a9b82d2978a1563daf31c8814a783">&#9670;&nbsp;</a></span>save_AM_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_AM_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>parents</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p >Note this does not follow through the linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> containing the genotypes to print </td></tr>
    <tr><td class="paramname">parents</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given genotypes. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gabf1a9b82d2978a1563daf31c8814a783_cgraph.png" border="0" usemap="#agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_cgraph" alt=""/></div>
<map name="agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_cgraph" id="agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="5,32,136,60"/>
<area shape="rect" href="group__savers.html#gaead4d69058f8fe7d59465100c801bbbd" title="Recursively save the parents of a particular id to a file." alt="" coords="184,32,298,60"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="352,5,464,33"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="346,57,469,85"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="517,57,691,85"/>
</map>
</div>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gabf1a9b82d2978a1563daf31c8814a783_icgraph.png" border="0" usemap="#agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_icgraph" alt=""/></div>
<map name="agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_icgraph" id="agroup__savers_gabf1a9b82d2978a1563daf31c8814a783_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="439,161,569,189"/>
<area shape="rect" href="group__crossers.html#ga7f9865d8d1912cda7e8524d1ff761de6" title="Performs random crosses among members of a group." alt="" coords="218,25,388,53"/>
<area shape="rect" href="group__crossers.html#ga43f436342a1afc3b14343a8f6e9385f9" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="219,83,387,111"/>
<area shape="rect" href="sim-operations_8c.html#a42ded1341b4c5a03c2233e2a3875fe07" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="215,135,391,163"/>
<area shape="rect" href="group__crossers.html#ga5e1da39f8341127e758d32393e8654aa" title="Creates an identical copy of each member of a group." alt="" coords="254,187,352,215"/>
<area shape="rect" href="group__crossers.html#gab081b254e7abce63067b8c02fbac774f" title="Creates a doubled haploid from each member of a group." alt="" coords="223,239,383,267"/>
<area shape="rect" href="group__crossers.html#gaf0d8ffcde1a2098e7ae807a031bbe7b9" title="Selfs each member of a group for a certain number of generations." alt="" coords="256,291,350,319"/>
<area shape="rect" href="group__crossers.html#gaa41e75da9e6e4f8db5466e8929dec2ee" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,162,49"/>
<area shape="rect" href="group__crossers.html#ga9d79117014f247c078c6576c70d600fb" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,73,164,117"/>
<area shape="rect" href="group__crossers.html#ga7fafc033bf3aaa14c9d5affbe9421ca9" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,141,167,169"/>
<area shape="rect" href="group__crossers.html#gafb306165b0cf0f5cd752f3eaf4c60928" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,193,162,237"/>
</map>
</div>

</div>
</div>
<a id="gaa9a8f4e7b82dda77c7a688d83fa08e28" name="gaa9a8f4e7b82dda77c7a688d83fa08e28"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa9a8f4e7b82dda77c7a688d83fa08e28">&#9670;&nbsp;</a></span>save_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name] [BV]</p>
<p >[id] [name] [BV]</p>
<p >...</p>
<p >The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gaa9a8f4e7b82dda77c7a688d83fa08e28_cgraph.png" border="0" usemap="#agroup__savers_gaa9a8f4e7b82dda77c7a688d83fa08e28_cgraph" alt=""/></div>
<map name="agroup__savers_gaa9a8f4e7b82dda77c7a688d83fa08e28_cgraph" id="agroup__savers_gaa9a8f4e7b82dda77c7a688d83fa08e28_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in the SimData to a file." alt="" coords="5,273,79,301"/>
<area shape="rect" href="group__calculators.html#ga75f4e1199f9ead752b64cd5fa3b5a1a7" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="127,217,226,245"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="306,329,412,357"/>
<area shape="rect" href="group__maths.html#gab1f0bafd56d3496348ea3ccddfe9d937" title="Multiply two sets of a DecimalMatrix and vector, and add both products to the first column of a provi..." alt="" coords="280,5,437,49"/>
<area shape="rect" href="group__maths.html#ga029101c5b02d8ab22e74ce3c103f6114" title="Multiply a DecimalMatrix to a vector, and add that product to the first column of a provided DecimalM..." alt="" coords="274,73,443,117"/>
<area shape="rect" href="group__calculators.html#gae883ac8145dd7e45ada82cc878d405d7" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="281,141,436,185"/>
<area shape="rect" href="group__calculators.html#ga11a8e25ed696aaf04d27b2276e345fe6" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="284,209,434,253"/>
<area shape="rect" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="282,277,435,305"/>
</map>
</div>

</div>
</div>
<a id="ga38b0958216f509a3b0a10c4c49183f44" name="ga38b0958216f509a3b0a10c4c49183f44"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga38b0958216f509a3b0a10c4c49183f44">&#9670;&nbsp;</a></span>save_count_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p >The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele count for each marker]</p>
<p >[marker name] [allele count for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga38b0958216f509a3b0a10c4c49183f44_cgraph.png" border="0" usemap="#agroup__savers_ga38b0958216f509a3b0a10c4c49183f44_cgraph" alt=""/></div>
<map name="agroup__savers_ga38b0958216f509a3b0a10c4c49183f44_cgraph" id="agroup__savers_ga38b0958216f509a3b0a10c4c49183f44_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="5,44,135,72"/>
<area shape="rect" href="group__calculators.html#gae9e20b885533620494187b26e7d44d48" title="Calculates the number of times at each marker that a particular allele appears for each genotype in t..." alt="" coords="183,5,318,49"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="197,73,304,101"/>
<area shape="rect" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="366,13,519,41"/>
</map>
</div>

</div>
</div>
<a id="ga8bbfd2ed13f449e5934b91701cc94a9f" name="ga8bbfd2ed13f449e5934b91701cc94a9f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8bbfd2ed13f449e5934b91701cc94a9f">&#9670;&nbsp;</a></span>save_count_matrix_of_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix_of_group </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in a group to a file. </p>
<p >The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele count for each marker]</p>
<p >[marker name] [allele count for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the allele count of. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga8bbfd2ed13f449e5934b91701cc94a9f_cgraph.png" border="0" usemap="#agroup__savers_ga8bbfd2ed13f449e5934b91701cc94a9f_cgraph" alt=""/></div>
<map name="agroup__savers_ga8bbfd2ed13f449e5934b91701cc94a9f_cgraph" id="agroup__savers_ga8bbfd2ed13f449e5934b91701cc94a9f_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="5,152,152,196"/>
<area shape="rect" href="group__calculators.html#ga3b7bb854c37c21eaa60ca155ac6e7d86" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="201,40,352,84"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="573,83,683,111"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="223,160,330,188"/>
<area shape="rect" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="200,212,353,240"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="401,212,525,240"/>
<area shape="rect" href="group__collgetters.html#gaf3f95ff569cca762e4de8be67c4d4097" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="403,5,523,33"/>
</map>
</div>

</div>
</div>
<a id="ga08c823d63ce4c648432cb740b5ab743b" name="ga08c823d63ce4c648432cb740b5ab743b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga08c823d63ce4c648432cb740b5ab743b">&#9670;&nbsp;</a></span>save_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing all genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga08c823d63ce4c648432cb740b5ab743b_cgraph.png" border="0" usemap="#agroup__savers_ga08c823d63ce4c648432cb740b5ab743b_cgraph" alt=""/></div>
<map name="agroup__savers_ga08c823d63ce4c648432cb740b5ab743b_cgraph" id="agroup__savers_ga08c823d63ce4c648432cb740b5ab743b_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in the SimData to a file." alt="" coords="5,32,132,60"/>
<area shape="rect" href="group__savers.html#gaead4d69058f8fe7d59465100c801bbbd" title="Recursively save the parents of a particular id to a file." alt="" coords="180,32,294,60"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="348,5,460,33"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="342,57,465,85"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="513,57,687,85"/>
</map>
</div>

</div>
</div>
<a id="gaeb3b527db48fd67f0d42a39c681b2c65" name="gaeb3b527db48fd67f0d42a39c681b2c65"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaeb3b527db48fd67f0d42a39c681b2c65">&#9670;&nbsp;</a></span>save_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">    [marker name]   [marker name]
</pre><p> [id]OR[name] [allele pairs for each marker]</p>
<p >[id]OR[name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gaeb3b527db48fd67f0d42a39c681b2c65_cgraph.png" border="0" usemap="#agroup__savers_gaeb3b527db48fd67f0d42a39c681b2c65_cgraph" alt=""/></div>
<map name="agroup__savers_gaeb3b527db48fd67f0d42a39c681b2c65_cgraph" id="agroup__savers_gaeb3b527db48fd67f0d42a39c681b2c65_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,84,136,112"/>
<area shape="rect" href="group__collgetters.html#gaf3f95ff569cca762e4de8be67c4d4097" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="186,5,306,33"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="356,84,465,112"/>
<area shape="rect" href="group__collgetters.html#ga7bb2cb27f43af0f6330c6fa9105d8dea" title="Gets the ids of each member of the group." alt="" coords="195,109,297,137"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="184,161,308,189"/>
</map>
</div>

</div>
</div>
<a id="ga7cd7382db4574b2b0cffb316eab47a29" name="ga7cd7382db4574b2b0cffb316eab47a29"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7cd7382db4574b2b0cffb316eab47a29">&#9670;&nbsp;</a></span>save_group_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[group member id] [group member name] [BV]</p>
<p >[group member id] [group member name] [BV]</p>
<p >...</p>
<p >The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the breeding values of. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga7cd7382db4574b2b0cffb316eab47a29_cgraph.png" border="0" usemap="#agroup__savers_ga7cd7382db4574b2b0cffb316eab47a29_cgraph" alt=""/></div>
<map name="agroup__savers_ga7cd7382db4574b2b0cffb316eab47a29_cgraph" id="agroup__savers_ga7cd7382db4574b2b0cffb316eab47a29_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in a group to a file." alt="" coords="5,438,118,466"/>
<area shape="rect" href="group__calculators.html#gabf3d2edad86e88ef0021cb608b009b87" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="166,229,304,257"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="758,186,867,214"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="394,386,500,414"/>
<area shape="rect" href="group__collgetters.html#ga7bb2cb27f43af0f6330c6fa9105d8dea" title="Gets the ids of each member of the group." alt="" coords="598,489,701,517"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="173,541,297,569"/>
<area shape="rect" href="group__maths.html#gab1f0bafd56d3496348ea3ccddfe9d937" title="Multiply two sets of a DecimalMatrix and vector, and add both products to the first column of a provi..." alt="" coords="368,250,525,294"/>
<area shape="rect" href="group__maths.html#ga029101c5b02d8ab22e74ce3c103f6114" title="Multiply a DecimalMatrix to a vector, and add that product to the first column of a provided DecimalM..." alt="" coords="362,318,531,362"/>
<area shape="rect" href="group__calculators.html#ga3b7bb854c37c21eaa60ca155ac6e7d86" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="371,62,522,106"/>
<area shape="rect" href="group__calculators.html#ga50e0c672eb7e42ef3c8d94861d2262af" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="352,130,541,174"/>
<area shape="rect" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="370,198,523,226"/>
<area shape="rect" href="group__collgetters.html#gaf3f95ff569cca762e4de8be67c4d4097" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="589,134,710,162"/>
</map>
</div>

</div>
</div>
<a id="ga7adbe41e3a1d04fbf2b32840e2d01f86" name="ga7adbe41e3a1d04fbf2b32840e2d01f86"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7adbe41e3a1d04fbf2b32840e2d01f86">&#9670;&nbsp;</a></span>save_group_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the pedigree of. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_cgraph.png" border="0" usemap="#agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_cgraph" alt=""/></div>
<map name="agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_cgraph" id="agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="5,109,171,137"/>
<area shape="rect" href="group__collgetters.html#ga7bb2cb27f43af0f6330c6fa9105d8dea" title="Gets the ids of each member of the group." alt="" coords="230,5,332,33"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="398,57,507,85"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="219,109,343,137"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="391,173,514,201"/>
<area shape="rect" href="group__savers.html#gaead4d69058f8fe7d59465100c801bbbd" title="Recursively save the parents of a particular id to a file." alt="" coords="224,237,338,265"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="562,173,736,201"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="397,237,509,265"/>
</map>
</div>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_icgraph.png" border="0" usemap="#agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_icgraph" alt=""/></div>
<map name="agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_icgraph" id="agroup__savers_ga7adbe41e3a1d04fbf2b32840e2d01f86_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="195,5,360,33"/>
<area shape="rect" href="group__collgetters.html#ga530d4ce0844f7709a6868d63f018ef28" title="Gets the full pedigree string (as per save_group_full_pedigree() ) of each member of the group." alt="" coords="5,5,147,33"/>
</map>
</div>

</div>
</div>
<a id="ga822c4500c4f5979c9eee8aa733434196" name="ga822c4500c4f5979c9eee8aa733434196"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga822c4500c4f5979c9eee8aa733434196">&#9670;&nbsp;</a></span>save_group_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[group member name] [parent 1 name] [parent 2 name]</p>
<p >[group member name] [parent 1 name] [parent 2 name]</p>
<p >...</p>
<p >The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the immediate parents of. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga822c4500c4f5979c9eee8aa733434196_cgraph.png" border="0" usemap="#agroup__savers_ga822c4500c4f5979c9eee8aa733434196_cgraph" alt=""/></div>
<map name="agroup__savers_ga822c4500c4f5979c9eee8aa733434196_cgraph" id="agroup__savers_ga822c4500c4f5979c9eee8aa733434196_cgraph">
<area shape="rect" title="Print the parents of each genotype in a group to a file." alt="" coords="5,101,152,145"/>
<area shape="rect" href="group__collgetters.html#ga7bb2cb27f43af0f6330c6fa9105d8dea" title="Gets the ids of each member of the group." alt="" coords="211,5,313,33"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="404,57,513,85"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="200,109,324,137"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="206,161,318,189"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="200,213,324,241"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="372,213,545,241"/>
</map>
</div>

</div>
</div>
<a id="ga378aa1f3cc71a3b3e5e17c9c103c4727" name="ga378aa1f3cc71a3b3e5e17c9c103c4727"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga378aa1f3cc71a3b3e5e17c9c103c4727">&#9670;&nbsp;</a></span>save_manual_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_manual_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int *&#160;</td>
          <td class="paramname"><em>ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>names</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name] [BV]</p>
<p >[id] [name] [BV]</p>
<p >...</p>
<p >The function assumes the array of ids, array of names, and columns of the <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> are ordered the same, so a single index produces the corresponding value from each. It is used by as-you-go savers within crossing functions. For general use consider <code>save_bvs</code> or <code>save_group_bvs</code> instead.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the BVs in the first row. </td></tr>
    <tr><td class="paramname">ids</td><td>array of ids to print alongside the BVs. </td></tr>
    <tr><td class="paramname">names</td><td>array of names to print alongside the BVs. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga378aa1f3cc71a3b3e5e17c9c103c4727_icgraph.png" border="0" usemap="#agroup__savers_ga378aa1f3cc71a3b3e5e17c9c103c4727_icgraph" alt=""/></div>
<map name="agroup__savers_ga378aa1f3cc71a3b3e5e17c9c103c4727_icgraph" id="agroup__savers_ga378aa1f3cc71a3b3e5e17c9c103c4727_icgraph">
<area shape="rect" title="Print the provided breeding values of each provided name and id to a file, with the same format as a ..." alt="" coords="439,161,561,189"/>
<area shape="rect" href="group__crossers.html#ga7f9865d8d1912cda7e8524d1ff761de6" title="Performs random crosses among members of a group." alt="" coords="218,25,388,53"/>
<area shape="rect" href="group__crossers.html#ga43f436342a1afc3b14343a8f6e9385f9" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="219,83,387,111"/>
<area shape="rect" href="sim-operations_8c.html#a42ded1341b4c5a03c2233e2a3875fe07" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="215,135,391,163"/>
<area shape="rect" href="group__crossers.html#ga5e1da39f8341127e758d32393e8654aa" title="Creates an identical copy of each member of a group." alt="" coords="254,187,352,215"/>
<area shape="rect" href="group__crossers.html#gab081b254e7abce63067b8c02fbac774f" title="Creates a doubled haploid from each member of a group." alt="" coords="223,239,383,267"/>
<area shape="rect" href="group__crossers.html#gaf0d8ffcde1a2098e7ae807a031bbe7b9" title="Selfs each member of a group for a certain number of generations." alt="" coords="256,291,350,319"/>
<area shape="rect" href="group__crossers.html#gaa41e75da9e6e4f8db5466e8929dec2ee" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,162,49"/>
<area shape="rect" href="group__crossers.html#ga9d79117014f247c078c6576c70d600fb" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,73,164,117"/>
<area shape="rect" href="group__crossers.html#ga7fafc033bf3aaa14c9d5affbe9421ca9" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,141,167,169"/>
<area shape="rect" href="group__crossers.html#gafb306165b0cf0f5cd752f3eaf4c60928" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,193,162,237"/>
</map>
</div>

</div>
</div>
<a id="ga63e803fa3cdf0cbb209b2922ce2982fe" name="ga63e803fa3cdf0cbb209b2922ce2982fe"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga63e803fa3cdf0cbb209b2922ce2982fe">&#9670;&nbsp;</a></span>save_marker_blocks()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_marker_blocks </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the markers contained in a set of blocks to a file. </p>
<p >Column separators are tabs.</p>
<p >The printing format is:</p>
<p >Chrom Pos Name Class Markers</p>
<p >0 0 b0 b m1;m2;m3;m4;</p>
<p >0 0 b0 b m7;m9;</p>
<p >...</p>
<p >where m1, m2, m3, m4 are the names of the markers in the first block and m7 and m9 are the names of the markers in the second block.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> whose data we print </td></tr>
    <tr><td class="paramname">b</td><td><a class="el" href="struct_marker_blocks.html" title="A struct used to store a set of blocks of markers.">MarkerBlocks</a> struct containing the groupings of markers to print. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga5674dd2fbdcf205475c8036a09536f2e" name="ga5674dd2fbdcf205475c8036a09536f2e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga5674dd2fbdcf205475c8036a09536f2e">&#9670;&nbsp;</a></span>save_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[genotype name] [parent 1 name] [parent 2 name]</p>
<p >[genotype name] [parent 1 name] [parent 2 name]</p>
<p >...</p>
<p >The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes and their pedigrees </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga5674dd2fbdcf205475c8036a09536f2e_cgraph.png" border="0" usemap="#agroup__savers_ga5674dd2fbdcf205475c8036a09536f2e_cgraph" alt=""/></div>
<map name="agroup__savers_ga5674dd2fbdcf205475c8036a09536f2e_cgraph" id="agroup__savers_ga5674dd2fbdcf205475c8036a09536f2e_cgraph">
<area shape="rect" title="Print the parents of each genotype in the SimData to a file." alt="" coords="5,32,169,60"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="222,5,335,33"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="217,57,340,85"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="388,57,561,85"/>
</map>
</div>

</div>
</div>
<a id="gaead4d69058f8fe7d59465100c801bbbd" name="gaead4d69058f8fe7d59465100c801bbbd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaead4d69058f8fe7d59465100c801bbbd">&#9670;&nbsp;</a></span>save_parents_of()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_parents_of </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Recursively save the parents of a particular id to a file. </p>
<p >It saves using the following format:</p>
<ul>
<li>no characters are saved if the parents of the id are unknown/0</li>
<li>if the id has one parent, repeated twice (it was produced by selfing), print "=(parentname)", where parentname is the name of the parent or its id if it does not have one, followed by whatever is printed by a call to this function on the parent's id.</li>
<li>if the id has two separate parents, print "=(parent1name,parent2name)", where parent1name and parent2name are the name2 of the two parents or their ids if they does not have names, each name immediately followed by whatever is printed by a call to this function on the corresponding parent's id.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given id. </td></tr>
    <tr><td class="paramname">p1</td><td>the session-unique id of the first parent to be saved. </td></tr>
    <tr><td class="paramname">p2</td><td>the session-unique id of the second parent to be saved. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gaead4d69058f8fe7d59465100c801bbbd_cgraph.png" border="0" usemap="#agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_cgraph" alt=""/></div>
<map name="agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_cgraph" id="agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_cgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="5,32,120,60"/>
<area shape="rect" href="group__search.html#ga0a9f180e2bff0952bd9e5af26c7ae37d" title="Returns the name of the genotype with a given id." alt="" coords="173,5,286,33"/>
<area shape="rect" href="group__search.html#gade1706129ba2510f1f1ed496b51efcb6" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="168,57,291,85"/>
<area shape="rect" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7" title="Returns the located index in an array of integers where the integer is target." alt="" coords="339,57,512,85"/>
</map>
</div>
<div id="dynsection-15" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-15-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-15-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-15-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gaead4d69058f8fe7d59465100c801bbbd_icgraph.png" border="0" usemap="#agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_icgraph" alt=""/></div>
<map name="agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_icgraph" id="agroup__savers_gaead4d69058f8fe7d59465100c801bbbd_icgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="652,252,767,280"/>
<area shape="rect" href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="457,187,587,215"/>
<area shape="rect" href="group__savers.html#ga08c823d63ce4c648432cb740b5ab743b" title="Print the full known pedigree of each genotype in the SimData to a file." alt="" coords="459,252,585,280"/>
<area shape="rect" href="group__savers.html#ga7adbe41e3a1d04fbf2b32840e2d01f86" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="439,329,604,357"/>
<area shape="rect" href="group__crossers.html#ga7f9865d8d1912cda7e8524d1ff761de6" title="Performs random crosses among members of a group." alt="" coords="218,19,388,47"/>
<area shape="rect" href="group__crossers.html#ga43f436342a1afc3b14343a8f6e9385f9" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="219,83,387,111"/>
<area shape="rect" href="sim-operations_8c.html#a42ded1341b4c5a03c2233e2a3875fe07" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="215,135,391,163"/>
<area shape="rect" href="group__crossers.html#ga5e1da39f8341127e758d32393e8654aa" title="Creates an identical copy of each member of a group." alt="" coords="254,187,352,215"/>
<area shape="rect" href="group__crossers.html#gab081b254e7abce63067b8c02fbac774f" title="Creates a doubled haploid from each member of a group." alt="" coords="223,239,383,267"/>
<area shape="rect" href="group__crossers.html#gaf0d8ffcde1a2098e7ae807a031bbe7b9" title="Selfs each member of a group for a certain number of generations." alt="" coords="256,291,350,319"/>
<area shape="rect" href="group__crossers.html#gaa41e75da9e6e4f8db5466e8929dec2ee" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,162,49"/>
<area shape="rect" href="group__crossers.html#ga9d79117014f247c078c6576c70d600fb" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,73,164,117"/>
<area shape="rect" href="group__crossers.html#ga7fafc033bf3aaa14c9d5affbe9421ca9" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,141,167,169"/>
<area shape="rect" href="group__crossers.html#gafb306165b0cf0f5cd752f3eaf4c60928" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,193,162,237"/>
<area shape="rect" href="group__collgetters.html#ga530d4ce0844f7709a6868d63f018ef28" title="Gets the full pedigree string (as per save_group_full_pedigree() ) of each member of the group." alt="" coords="232,343,374,371"/>
</map>
</div>

</div>
</div>
<a id="ga5c1144f08e16be0974f7a8f81b81c0d4" name="ga5c1144f08e16be0974f7a8f81b81c0d4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga5c1144f08e16be0974f7a8f81b81c0d4">&#9670;&nbsp;</a></span>save_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_simdata </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>m</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p >Column separators are tabs.</p>
<p >The printing format is:</p>
<p >name chr pos [allele to which these effects apply, 0+ columns]</p>
<p >[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p >[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p >...</p>
<p >If m-&gt;effects is NULL, m-&gt;ref_alleles is NULL, or m-&gt;genetic_map is NULL, then the relevant columns are omitted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> whose data we print </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga8e14b8d8c1836e1402a4a5e7a74c274f" name="ga8e14b8d8c1836e1402a4a5e7a74c274f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8e14b8d8c1836e1402a4a5e7a74c274f">&#9670;&nbsp;</a></span>save_transposed_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the gene data saved in the linked list starting with <code>m</code> to the file. </p>
<p >Uses the following format: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele pairs for each marker]</p>
<p >[marker name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga2790244044d202e6fb158d8f57cf6d6f" name="ga2790244044d202e6fb158d8f57cf6d6f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2790244044d202e6fb158d8f57cf6d6f">&#9670;&nbsp;</a></span>save_transposed_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">[id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele pairs for each marker]</p>
<p >[marker name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-16" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-16-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-16-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-16-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga2790244044d202e6fb158d8f57cf6d6f_cgraph.png" border="0" usemap="#agroup__savers_ga2790244044d202e6fb158d8f57cf6d6f_cgraph" alt=""/></div>
<map name="agroup__savers_ga2790244044d202e6fb158d8f57cf6d6f_cgraph" id="agroup__savers_ga2790244044d202e6fb158d8f57cf6d6f_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,76,163,120"/>
<area shape="rect" href="group__collgetters.html#gaf3f95ff569cca762e4de8be67c4d4097" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="213,5,333,33"/>
<area shape="rect" href="group__collgetters.html#ga706e640cd2cfb0f390f98c3b8c00a709" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="383,84,492,112"/>
<area shape="rect" href="group__collgetters.html#ga7bb2cb27f43af0f6330c6fa9105d8dea" title="Gets the ids of each member of the group." alt="" coords="222,109,324,137"/>
<area shape="rect" href="group__collgetters.html#gabb5a3786c56f6cfc7fa8a363bb1d2149" title="Gets a shallow copy of the names of each member of the group." alt="" coords="211,161,335,189"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
