<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Genotype Iterators</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.11</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">Genotype Iterators<div class="ingroups"><a class="el" href="group__getters.html">Data Access and Search Functions</a></div></div></div>
</div><!--header-->
<div class="contents">

<p>For iterating through the genotypes in the simulation.  
<a href="#details">More...</a></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Genotype Iterators:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators.png" border="0" usemap="#agroup____iterators" alt=""/></div>
<map name="agroup____iterators" id="agroup____iterators">
<area shape="rect" href="group__getters.html" title="For non&#45;persistent access to simulation results and contents." alt="" coords="5,5,172,48"/>
<area shape="rect" title="For iterating through the genotypes in the simulation." alt="" coords="220,13,348,40"/>
</map>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">An gsc_AlleleMatrix/gsc_AlleleMatrix index coordinate of a particular genotype in the simulation.  <a href="structgsc___geno_location.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A structure to iterate forwards and backwards through all genotypes in a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> or through only the members of a group.  <a href="structgsc___bidirectional_iterator.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A structure to iterate forwards through all positions in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> linked list in <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>.  <a href="structgsc___gappy_iterator.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially by storage order.  <a href="structgsc___empty_list_navigator.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A structure to search and cache indexes of all genotypes in a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> or of all the members of a group.  <a href="structgsc___random_access_iterator.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="define-members" name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:ga380f5e71802bb5191652598a3b1f80df"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga380f5e71802bb5191652598a3b1f80df">GSC_INVALID_GENO_LOCATION</a>&#160;&#160;&#160;(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>){.localAM=0,.localPos=<a class="el" href="sim-operations_8h.html#a69cfaeb60cca56d9073898585d3bfba5">GSC_NA_LOCALX</a>}</td></tr>
<tr class="memdesc:ga380f5e71802bb5191652598a3b1f80df"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constant representing a nonexistent location in the simulation.  <a href="group__iterators.html#ga380f5e71802bb5191652598a3b1f80df">More...</a><br /></td></tr>
<tr class="separator:ga380f5e71802bb5191652598a3b1f80df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabac400e75bbcf0fcfee5979feaae9fd6"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gabac400e75bbcf0fcfee5979feaae9fd6">GSC_IS_VALID_LOCATION</a>(g)&#160;&#160;&#160;(g.localAM != 0 &amp;&amp; g.localPos != <a class="el" href="sim-operations_8h.html#a69cfaeb60cca56d9073898585d3bfba5">GSC_NA_LOCALX</a>)</td></tr>
<tr class="memdesc:gabac400e75bbcf0fcfee5979feaae9fd6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check if a <a class="el" href="group__shortnames.html#ga86d2d8f2c416cdbd37c3f99e5d80c4e3">GenoLocation</a> is <a class="el" href="group__shortnames.html#ga7d0a1ac8a896751e36af006030708219">INVALID_GENO_LOCATION</a>.  <a href="group__iterators.html#gabac400e75bbcf0fcfee5979feaae9fd6">More...</a><br /></td></tr>
<tr class="separator:gabac400e75bbcf0fcfee5979feaae9fd6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga8e318c37d9100ca9195193b87ede8524"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga8e318c37d9100ca9195193b87ede8524">gsc_isValidLocation</a> (const <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> g)</td></tr>
<tr class="memdesc:ga8e318c37d9100ca9195193b87ede8524"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify whether a <a class="el" href="structgsc___geno_location.html" title="An gsc_AlleleMatrix/gsc_AlleleMatrix index coordinate of a particular genotype in the simulation.">gsc_GenoLocation</a> is INVALID_GENO_LOCATION.  <a href="group__iterators.html#ga8e318c37d9100ca9195193b87ede8524">More...</a><br /></td></tr>
<tr class="separator:ga8e318c37d9100ca9195193b87ede8524"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga05da8d96593f8404938cc24f659db049"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049">gsc_create_bidirectional_iter</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga05da8d96593f8404938cc24f659db049"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a bidirectional iterator.  <a href="group__iterators.html#ga05da8d96593f8404938cc24f659db049">More...</a><br /></td></tr>
<tr class="separator:ga05da8d96593f8404938cc24f659db049"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5cca400276ca83acceaab36d5f79aab1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1">gsc_create_bidirectional_iter_fromAM</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *am, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="separator:ga5cca400276ca83acceaab36d5f79aab1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa26377edd1339407af2070924d3596de"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa26377edd1339407af2070924d3596de">gsc_create_randomaccess_iter</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:gaa26377edd1339407af2070924d3596de"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a Random Access Iterator.  <a href="group__iterators.html#gaa26377edd1339407af2070924d3596de">More...</a><br /></td></tr>
<tr class="separator:gaa26377edd1339407af2070924d3596de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga16615a5663b4d3bd768d8380f820aa66"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66">gsc_get_nth_AlleleMatrix</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *listStart, const unsigned int n)</td></tr>
<tr class="memdesc:ga16615a5663b4d3bd768d8380f820aa66"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> by index in the linked list.  <a href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66">More...</a><br /></td></tr>
<tr class="separator:ga16615a5663b4d3bd768d8380f820aa66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd">gsc_set_bidirectional_iter_to_start</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the start of its sequence.  <a href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd">More...</a><br /></td></tr>
<tr class="separator:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga99820e199424a3ec8ba7e3c58802fba9"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9">gsc_set_bidirectional_iter_to_end</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga99820e199424a3ec8ba7e3c58802fba9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the end of its sequence.  <a href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9">More...</a><br /></td></tr>
<tr class="separator:ga99820e199424a3ec8ba7e3c58802fba9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga630ba7f8e864238e296cb9480dfbc728"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728">gsc_next_forwards</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga630ba7f8e864238e296cb9480dfbc728"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the next location from a bidirectional iterator.  <a href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728">More...</a><br /></td></tr>
<tr class="separator:ga630ba7f8e864238e296cb9480dfbc728"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaacc1f3131edc2c62e14442ca1b647871"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871">gsc_next_backwards</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gaacc1f3131edc2c62e14442ca1b647871"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the previous location from a bidirectional iterator.  <a href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871">More...</a><br /></td></tr>
<tr class="separator:gaacc1f3131edc2c62e14442ca1b647871"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaccc34724148be204eb4d8515000dfd5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5">gsc_next_get_nth</a> (<a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a> *it, const unsigned int n)</td></tr>
<tr class="memdesc:gaaccc34724148be204eb4d8515000dfd5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a location by index using a <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a>.  <a href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5">More...</a><br /></td></tr>
<tr class="separator:gaaccc34724148be204eb4d8515000dfd5"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p >For iterating through the genotypes in the simulation. </p>
<p >It is possible to iterate through one group or through every genotype in the simulation. </p>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a id="ga380f5e71802bb5191652598a3b1f80df" name="ga380f5e71802bb5191652598a3b1f80df"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga380f5e71802bb5191652598a3b1f80df">&#9670;&nbsp;</a></span>GSC_INVALID_GENO_LOCATION</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define GSC_INVALID_GENO_LOCATION&#160;&#160;&#160;(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>){.localAM=0,.localPos=<a class="el" href="sim-operations_8h.html#a69cfaeb60cca56d9073898585d3bfba5">GSC_NA_LOCALX</a>}</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Constant representing a nonexistent location in the simulation. </p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga7d0a1ac8a896751e36af006030708219">INVALID_GENO_LOCATION</a> </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8h_source.html#l01220">1220</a> of file <a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>.</p>

</div>
</div>
<a id="gabac400e75bbcf0fcfee5979feaae9fd6" name="gabac400e75bbcf0fcfee5979feaae9fd6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabac400e75bbcf0fcfee5979feaae9fd6">&#9670;&nbsp;</a></span>GSC_IS_VALID_LOCATION</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define GSC_IS_VALID_LOCATION</td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">g</td><td>)</td>
          <td>&#160;&#160;&#160;(g.localAM != 0 &amp;&amp; g.localPos != <a class="el" href="sim-operations_8h.html#a69cfaeb60cca56d9073898585d3bfba5">GSC_NA_LOCALX</a>)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Check if a <a class="el" href="group__shortnames.html#ga86d2d8f2c416cdbd37c3f99e5d80c4e3">GenoLocation</a> is <a class="el" href="group__shortnames.html#ga7d0a1ac8a896751e36af006030708219">INVALID_GENO_LOCATION</a>. </p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga74a123d16204f5ca9d85342f13d3ae9a">IS_VALID_LOCATION</a> </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8h_source.html#l01224">1224</a> of file <a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>.</p>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga05da8d96593f8404938cc24f659db049" name="ga05da8d96593f8404938cc24f659db049"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga05da8d96593f8404938cc24f659db049">&#9670;&nbsp;</a></span>gsc_create_bidirectional_iter()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> gsc_create_bidirectional_iter </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Create a bidirectional iterator. </p>
<p >A bidirectional iterator can be used to loop through members of a particular group or all genotypes in the simulation, forwards or backwards. The iterator is not initialised to any location at this point. The first call to a next* function will initialise it, or you can manually initialise using gsc_set_bidirectional_iter_to_start or gsc_set_bidirectional_iter_to_end. gsc_next_forwards will initialise it to the first in the group, or call gsc_next_backwards to initialise it to the last in the group.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gaf6e8c60198060f13a3ade462c56b9d77">create_bidirectional_iter</a></p>
<dl class="section warning"><dt>Warning</dt><dd>An initialised iterator is only valid if no genotypes have been added to the group and no genotypes in the simulation as a whole have been removed since its creation. Discard the old iterator and create a new one when the state of the simulation changes.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the genotypes to iterate through </td></tr>
    <tr><td class="paramname">group</td><td>the group number of the group to iterate through, or 0 to iterate through all genotypes. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>uninitialised <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> for the provided group. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01427">1427</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga05da8d96593f8404938cc24f659db049_cgraph.png" border="0" usemap="#agroup__iterators_ga05da8d96593f8404938cc24f659db049_cgraph" alt=""/></div>
<map name="agroup__iterators_ga05da8d96593f8404938cc24f659db049_cgraph" id="agroup__iterators_ga05da8d96593f8404938cc24f659db049_cgraph">
<area shape="rect" title="Create a bidirectional iterator." alt="" coords="5,13,192,41"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="240,5,401,49"/>
</map>
</div>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga05da8d96593f8404938cc24f659db049_icgraph.png" border="0" usemap="#agroup__iterators_ga05da8d96593f8404938cc24f659db049_icgraph" alt=""/></div>
<map name="agroup__iterators_ga05da8d96593f8404938cc24f659db049_icgraph" id="agroup__iterators_ga05da8d96593f8404938cc24f659db049_icgraph">
<area shape="rect" title="Create a bidirectional iterator." alt="" coords="1378,525,1565,553"/>
<area shape="rect" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="1173,5,1311,49"/>
<area shape="rect" href="group__calculators.html#ga30689b0f03641259091a72d065951e29" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="1178,73,1305,101"/>
<area shape="rect" href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe" title="Replace all occurences of a given allele with a different symbol representation." alt="" coords="1153,688,1330,716"/>
<area shape="rect" href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c" title="Combine a set of groups into one group." alt="" coords="1170,740,1314,768"/>
<area shape="rect" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="1163,517,1321,561"/>
<area shape="rect" href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b" title="Split by some allocator (generic function)" alt="" coords="497,724,633,768"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,176,400,204"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,263,431,291"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,617,398,645"/>
<area shape="rect" href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d" title="Allocates the genotypes with values of a label in a particular range to a new group." alt="" coords="501,228,629,272"/>
<area shape="rect" href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b" title="Allocates the genotypes with a particular value of a label to a new group." alt="" coords="501,296,629,340"/>
<area shape="rect" href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a" title="Flip a coin for each member of the group to decide if it should be moved to the new group." alt="" coords="498,364,631,408"/>
<area shape="rect" href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2" title="Split by some quality (generic function)" alt="" coords="958,637,1083,681"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="1163,792,1321,820"/>
<area shape="rect" href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="1191,844,1292,872"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="1171,896,1313,924"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="1173,948,1311,976"/>
<area shape="rect" href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="954,23,1087,51"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="280,107,394,135"/>
<area shape="rect" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83" title="Identify group numbers that currently have members." alt="" coords="938,525,1103,553"/>
<area shape="rect" href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2" title="Function to identify the next n sequential integers that do not identify a group that currently has m..." alt="" coords="698,533,888,561"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="709,429,877,457"/>
<area shape="rect" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5" title="Allocate each member of the group to one of n groups with custom probabilities for each group." alt="" coords="252,677,422,705"/>
<area shape="rect" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="258,729,416,757"/>
<area shape="rect" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="260,781,414,809"/>
<area shape="rect" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621" title="Allocate each member of the group to one of n groups with equal probability." alt="" coords="270,833,404,877"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="490,552,639,596"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="489,432,641,460"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="484,484,646,528"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="479,620,650,648"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="269,561,405,589"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="25,569,175,597"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,505,429,533"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,385,429,429"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,453,431,481"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,381,192,425"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,449,195,477"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,501,190,545"/>
<area shape="rect" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33" title="Split a group into families by their pedigrees." alt="" coords="717,637,869,665"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="720,689,866,733"/>
<area shape="rect" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1" title="Split a group into n one&#45;member groups." alt="" coords="709,757,877,785"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="936,948,1105,976"/>
</map>
</div>

</div>
</div>
<a id="ga5cca400276ca83acceaab36d5f79aab1" name="ga5cca400276ca83acceaab36d5f79aab1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga5cca400276ca83acceaab36d5f79aab1">&#9670;&nbsp;</a></span>gsc_create_bidirectional_iter_fromAM()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> gsc_create_bidirectional_iter_fromAM </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>am</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01432">1432</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga5cca400276ca83acceaab36d5f79aab1_icgraph.png" border="0" usemap="#agroup__iterators_ga5cca400276ca83acceaab36d5f79aab1_icgraph" alt=""/></div>
<map name="agroup__iterators_ga5cca400276ca83acceaab36d5f79aab1_icgraph" id="agroup__iterators_ga5cca400276ca83acceaab36d5f79aab1_icgraph">
<area shape="rect" title=" " alt="" coords="1613,108,1774,152"/>
<area shape="rect" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049" title="Create a bidirectional iterator." alt="" coords="1378,583,1565,611"/>
<area shape="rect" href="sim-operations_8c.html#a240267fbee56d8075f7b8fe10c81aaa4" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="1394,141,1549,185"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="1394,5,1549,49"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="1394,73,1549,117"/>
<area shape="rect" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="1173,335,1311,379"/>
<area shape="rect" href="group__calculators.html#ga30689b0f03641259091a72d065951e29" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="1178,713,1305,741"/>
<area shape="rect" href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe" title="Replace all occurences of a given allele with a different symbol representation." alt="" coords="1153,179,1330,207"/>
<area shape="rect" href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c" title="Combine a set of groups into one group." alt="" coords="1170,231,1314,259"/>
<area shape="rect" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="1163,925,1321,969"/>
<area shape="rect" href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b" title="Split by some allocator (generic function)" alt="" coords="497,1181,633,1225"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,401,400,429"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,453,431,481"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,532,398,560"/>
<area shape="rect" href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d" title="Allocates the genotypes with values of a label in a particular range to a new group." alt="" coords="501,760,629,804"/>
<area shape="rect" href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b" title="Allocates the genotypes with a particular value of a label to a new group." alt="" coords="501,832,629,876"/>
<area shape="rect" href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a" title="Flip a coin for each member of the group to decide if it should be moved to the new group." alt="" coords="498,1091,631,1135"/>
<area shape="rect" href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2" title="Split by some quality (generic function)" alt="" coords="958,925,1083,969"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="1163,283,1321,311"/>
<area shape="rect" href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="1191,505,1292,533"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="1171,557,1313,585"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="1173,661,1311,689"/>
<area shape="rect" href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="954,713,1087,741"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="280,785,394,813"/>
<area shape="rect" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83" title="Identify group numbers that currently have members." alt="" coords="938,1040,1103,1068"/>
<area shape="rect" href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2" title="Function to identify the next n sequential integers that do not identify a group that currently has m..." alt="" coords="698,1144,888,1172"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="709,1040,877,1068"/>
<area shape="rect" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5" title="Allocate each member of the group to one of n groups with custom probabilities for each group." alt="" coords="252,1095,422,1123"/>
<area shape="rect" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="258,1147,416,1175"/>
<area shape="rect" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="260,1199,414,1227"/>
<area shape="rect" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621" title="Allocate each member of the group to one of n groups with equal probability." alt="" coords="270,1251,404,1295"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="490,1023,639,1067"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="489,909,641,937"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="484,308,646,352"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="479,971,650,999"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="269,1031,405,1059"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="25,1031,175,1059"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,229,429,257"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,281,429,325"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,349,431,377"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,409,192,453"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,289,195,317"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,341,190,385"/>
<area shape="rect" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33" title="Split a group into families by their pedigrees." alt="" coords="717,988,869,1016"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="720,868,866,912"/>
<area shape="rect" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1" title="Split a group into n one&#45;member groups." alt="" coords="709,936,877,964"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="936,609,1105,637"/>
</map>
</div>

</div>
</div>
<a id="gaa26377edd1339407af2070924d3596de" name="gaa26377edd1339407af2070924d3596de"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa26377edd1339407af2070924d3596de">&#9670;&nbsp;</a></span>gsc_create_randomaccess_iter()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a> gsc_create_randomaccess_iter </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Create a Random Access Iterator. </p>
<p >A random access iterator and the function <a class="el" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator.">gsc_next_get_nth()</a> can be used to get any genotype in the simulation by index or any genotype in a group by within-group index.</p>
<p >The random access iterator gives you access to the (i)th group member in O(n) time (n being the number of genotypes in the simulation), but contains a persistent cache that allows subsequent O(1) access to group members.</p>
<p >The random access iterator created by this function is initialised. If the <code>groupSize</code> attribute of the returned iterator is set to 0, no members of that group could be found in the simulation, so the iterator will never return a valid position.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga6fe61bd6960ec26b020f0466cb6509d7">create_randomaccess_iter</a></p>
<dl class="section warning"><dt>Warning</dt><dd>An initialised iterator is only valid if no genotypes have been added to the group and no genotypes in the simulation as a whole have been removed since its creation. Discard the old iterator and create a new one when the state of the simulation changes.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the genotypes to iterate through </td></tr>
    <tr><td class="paramname">group</td><td>the group number of the group to iterate through, or 0 to iterate through all genotypes. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>initialised Random Access iterator for the provided group. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01474">1474</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaa26377edd1339407af2070924d3596de_cgraph.png" border="0" usemap="#agroup__iterators_gaa26377edd1339407af2070924d3596de_cgraph" alt=""/></div>
<map name="agroup__iterators_gaa26377edd1339407af2070924d3596de_cgraph" id="agroup__iterators_gaa26377edd1339407af2070924d3596de_cgraph">
<area shape="rect" title="Create a Random Access Iterator." alt="" coords="5,5,208,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="256,5,377,33"/>
</map>
</div>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaa26377edd1339407af2070924d3596de_icgraph.png" border="0" usemap="#agroup__iterators_gaa26377edd1339407af2070924d3596de_icgraph" alt=""/></div>
<map name="agroup__iterators_gaa26377edd1339407af2070924d3596de_icgraph" id="agroup__iterators_gaa26377edd1339407af2070924d3596de_icgraph">
<area shape="rect" title="Create a Random Access Iterator." alt="" coords="479,117,681,145"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="261,5,413,33"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,57,429,85"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,109,429,153"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,177,431,205"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="251,229,423,257"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="43,5,157,33"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,109,192,153"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,177,195,205"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,229,190,273"/>
</map>
</div>

</div>
</div>
<a id="ga16615a5663b4d3bd768d8380f820aa66" name="ga16615a5663b4d3bd768d8380f820aa66"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga16615a5663b4d3bd768d8380f820aa66">&#9670;&nbsp;</a></span>gsc_get_nth_AlleleMatrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> * gsc_get_nth_AlleleMatrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>listStart</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> by index in the linked list. </p>
<p >listStart is considered the 0th <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> (0-based indexing)</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">listStart</td><td>the 0th <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> in the linked list </td></tr>
    <tr><td class="paramname">n</td><td>the index of the desired <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> in the linked list </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>pointer to the nth <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, if it exists in the list, or NULL if the list is shorter than n </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01552">1552</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga16615a5663b4d3bd768d8380f820aa66_icgraph.png" border="0" usemap="#agroup__iterators_ga16615a5663b4d3bd768d8380f820aa66_icgraph" alt=""/></div>
<map name="agroup__iterators_ga16615a5663b4d3bd768d8380f820aa66_icgraph" id="agroup__iterators_ga16615a5663b4d3bd768d8380f820aa66_icgraph">
<area shape="rect" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="1091,392,1258,420"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="900,101,1043,145"/>
<area shape="rect" href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871" title="Get the previous location from a bidirectional iterator." alt="" coords="692,511,836,539"/>
<area shape="rect" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="909,365,1034,393"/>
<area shape="rect" href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9" title="Initialise a Bidirectional iterator to the end of its sequence." alt="" coords="900,469,1043,513"/>
<area shape="rect" href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5" title="Deletes all genotypes belonging to a particular group." alt="" coords="702,15,826,43"/>
<area shape="rect" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad" title="Push emptylist edited genotypes into the SimData." alt="" coords="687,67,841,111"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="683,135,845,179"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="460,5,609,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="230,13,365,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="471,73,597,101"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="440,125,628,153"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="442,229,626,257"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="442,289,626,333"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="440,373,628,401"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="473,177,595,205"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="205,305,390,349"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="203,373,392,401"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="208,425,387,469"/>
<area shape="rect" href="sim-operations_8c.html#a0fcef09a211d57fadb1dfa0b6d102901" title="parentChooser function parameter for gsc_make_random_crosses." alt="" coords="677,203,851,247"/>
<area shape="rect" href="sim-operations_8c.html#a2cee0d2e90de12744fd077b20a7db9bd" title="parentChooser function parameter for gsc_make_random_crosses_between." alt="" coords="676,323,852,367"/>
<area shape="rect" href="sim-operations_8c.html#a2c3d189234e1ac86e27a1ccaace3c363" title="parentChooser function parameter for gsc_make_targeted_crosses." alt="" coords="677,391,851,435"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="688,459,840,487"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="678,271,850,299"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="478,459,591,487"/>
</map>
</div>

</div>
</div>
<a id="ga8e318c37d9100ca9195193b87ede8524" name="ga8e318c37d9100ca9195193b87ede8524"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8e318c37d9100ca9195193b87ede8524">&#9670;&nbsp;</a></span>gsc_isValidLocation()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_isValidLocation </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>g</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Identify whether a <a class="el" href="structgsc___geno_location.html" title="An gsc_AlleleMatrix/gsc_AlleleMatrix index coordinate of a particular genotype in the simulation.">gsc_GenoLocation</a> is INVALID_GENO_LOCATION. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__shortnames.html#ga74a123d16204f5ca9d85342f13d3ae9a">IS_VALID_LOCATION</a> <a class="el" href="group__shortnames.html#ga74a123d16204f5ca9d85342f13d3ae9a">IS_VALID_LOCATION</a> has the same function. </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">g</td><td>location to check. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>FALSE if g has either of the attributes of INVALID_GENO_LOCATION, TRUE otherwise </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8h_source.html#l01233">1233</a> of file <a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>.</p>

</div>
</div>
<a id="gaacc1f3131edc2c62e14442ca1b647871" name="gaacc1f3131edc2c62e14442ca1b647871"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaacc1f3131edc2c62e14442ca1b647871">&#9670;&nbsp;</a></span>gsc_next_backwards()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_next_backwards </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the previous location from a bidirectional iterator. </p>
<p >Moves the pointer of a <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> backwards by one step, and returns the new location it points to. If the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> is not initialised, then initialises it to the very last element.</p>
<p >Slightly slower than gsc_next_forwards, because the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> linked list is not bidirectional. To find the preceding <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, it needs to count forwards from the beginning of the list to find the n-1th <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>.</p>
<p >Returns GSC_INVALID_GENO_LOCATION if the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> is corrupted or if it is at the beginning of the sequence. Test the return value of this function with <a class="el" href="group__iterators.html#gabac400e75bbcf0fcfee5979feaae9fd6">GSC_IS_VALID_LOCATION()</a>.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gaf09273de7be166868b7d8b0d9c5f2f2e">next_backwards</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">it</td><td>the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> to iterate backwards </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the location of the previous genotype in the sequence, or GSC_INVALID_GENO_LOCATION if the iterator is corrupted or the iterator's pointer is already at the first element. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01859">1859</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaacc1f3131edc2c62e14442ca1b647871_cgraph.png" border="0" usemap="#agroup__iterators_gaacc1f3131edc2c62e14442ca1b647871_cgraph" alt=""/></div>
<map name="agroup__iterators_gaacc1f3131edc2c62e14442ca1b647871_cgraph" id="agroup__iterators_gaacc1f3131edc2c62e14442ca1b647871_cgraph">
<area shape="rect" title="Get the previous location from a bidirectional iterator." alt="" coords="5,5,149,33"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="388,5,555,33"/>
<area shape="rect" href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9" title="Initialise a Bidirectional iterator to the end of its sequence." alt="" coords="197,27,340,71"/>
</map>
</div>

</div>
</div>
<a id="ga630ba7f8e864238e296cb9480dfbc728" name="ga630ba7f8e864238e296cb9480dfbc728"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga630ba7f8e864238e296cb9480dfbc728">&#9670;&nbsp;</a></span>gsc_next_forwards()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_next_forwards </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the next location from a bidirectional iterator. </p>
<p >Moves the pointer of a <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> forwards by one step, and returns the new location it points to. If the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> is not initialised, then initialises it to the very first element.</p>
<p >Returns GSC_INVALID_GENO_LOCATION if the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> is corrupted or if it is at the end of the sequence. Test the return value of this function with <a class="el" href="group__iterators.html#gabac400e75bbcf0fcfee5979feaae9fd6">GSC_IS_VALID_LOCATION()</a>.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga78e7dac2c4c329adbd048de304ebb498">next_forwards</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">it</td><td>the <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> to iterate forwards </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the location of the next genotype in the sequence, or GSC_INVALID_GENO_LOCATION if the iterator is corrupted or the iterator's pointer is already at the last element. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01745">1745</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga630ba7f8e864238e296cb9480dfbc728_cgraph.png" border="0" usemap="#agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_cgraph" alt=""/></div>
<map name="agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_cgraph" id="agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_cgraph">
<area shape="rect" title="Get the next location from a bidirectional iterator." alt="" coords="5,13,135,41"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="183,5,326,49"/>
</map>
</div>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga630ba7f8e864238e296cb9480dfbc728_icgraph.png" border="0" usemap="#agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_icgraph" alt=""/></div>
<map name="agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_icgraph" id="agroup__iterators_ga630ba7f8e864238e296cb9480dfbc728_icgraph">
<area shape="rect" title="Get the next location from a bidirectional iterator." alt="" coords="1190,681,1319,709"/>
<area shape="rect" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="966,75,1105,119"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="953,143,1118,171"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="721,139,860,167"/>
<area shape="rect" href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe" title="Replace all occurences of a given allele with a different symbol representation." alt="" coords="947,1127,1124,1155"/>
<area shape="rect" href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c" title="Combine a set of groups into one group." alt="" coords="963,1179,1108,1207"/>
<area shape="rect" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="956,741,1115,785"/>
<area shape="rect" href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b" title="Split by some allocator (generic function)" alt="" coords="260,1005,397,1049"/>
<area shape="rect" href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d" title="Allocates the genotypes with values of a label in a particular range to a new group." alt="" coords="264,436,393,480"/>
<area shape="rect" href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b" title="Allocates the genotypes with a particular value of a label to a new group." alt="" coords="264,504,393,548"/>
<area shape="rect" href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a" title="Flip a coin for each member of the group to decide if it should be moved to the new group." alt="" coords="262,796,395,840"/>
<area shape="rect" href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2" title="Split by some quality (generic function)" alt="" coords="728,801,853,845"/>
<area shape="rect" href="sim-operations_8c.html#a94b58634907af7e24ac22b543b8fcbe3" title="parentChooser function parameter for gsc_make_clones." alt="" coords="949,673,1122,717"/>
<area shape="rect" href="sim-operations_8c.html#aafe958ddf4ee8c990c521d6ebf60db71" title="parentChooser function parameter for gsc_self_n_times." alt="" coords="242,864,415,908"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="947,259,1124,287"/>
<area shape="rect" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361" title="Prints a matrix of genotype information to a file." alt="" coords="929,431,1142,459"/>
<area shape="rect" href="group__calculators.html#ga30689b0f03641259091a72d065951e29" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="727,27,854,55"/>
<area shape="rect" href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="492,27,624,55"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="43,123,156,151"/>
<area shape="rect" href="sim-operations_8c.html#a240267fbee56d8075f7b8fe10c81aaa4" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="481,131,635,175"/>
<area shape="rect" href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="507,79,608,107"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="247,368,410,412"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="36,504,163,532"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="5,604,194,632"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="7,428,192,456"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="7,308,192,352"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="5,376,194,404"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="38,752,161,780"/>
<area shape="rect" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83" title="Identify group numbers that currently have members." alt="" coords="708,749,873,777"/>
<area shape="rect" href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2" title="Function to identify the next n sequential integers that do not identify a group that currently has m..." alt="" coords="463,749,652,777"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="474,645,642,673"/>
<area shape="rect" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5" title="Allocate each member of the group to one of n groups with custom probabilities for each group." alt="" coords="14,1103,185,1131"/>
<area shape="rect" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="20,931,179,959"/>
<area shape="rect" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="22,983,177,1011"/>
<area shape="rect" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621" title="Allocate each member of the group to one of n groups with equal probability." alt="" coords="33,1035,166,1079"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="254,624,403,668"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="252,572,405,600"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="243,692,414,720"/>
<area shape="rect" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33" title="Split a group into families by their pedigrees." alt="" coords="482,801,634,829"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="485,853,630,897"/>
<area shape="rect" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1" title="Split a group into n one&#45;member groups." alt="" coords="474,921,641,949"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="713,191,868,235"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="721,259,860,287"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="473,251,642,279"/>
<area shape="rect" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file." alt="" coords="715,336,866,380"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="700,431,881,459"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="479,303,637,331"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="481,423,635,467"/>
<area shape="rect" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="478,355,638,399"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="487,491,629,519"/>
</map>
</div>

</div>
</div>
<a id="gaaccc34724148be204eb4d8515000dfd5" name="gaaccc34724148be204eb4d8515000dfd5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaaccc34724148be204eb4d8515000dfd5">&#9670;&nbsp;</a></span>gsc_next_get_nth()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_next_get_nth </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a location by index using a <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a>. </p>
<p >Gives the location of the provided global index (if <code>it-&gt;group == 0</code>) or the location of the provided group index (if <code>it-&gt;group</code> is not 0), by first searching the <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a>'s cache for it and if not, searching the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> for it and adding it and its predecessors to the cache.</p>
<p >Returns GSC_INVALID_GENO_LOCATION if the iterator is corrupted or the index is invalid. Check the return value with <a class="el" href="group__iterators.html#gabac400e75bbcf0fcfee5979feaae9fd6">GSC_IS_VALID_LOCATION()</a>.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga6c778f594f3b64af885c8cfa3ce3a784">next_get_nth</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">it</td><td>the <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a> to read and update the cache of. </td></tr>
    <tr><td class="paramname">n</td><td>If the iterator is iterating through all genotypes, the global index in the simulation of the genotype you want to access. If the iterator is iterating through the group, the within-group index of the genotype you want to access. In either case the first genotype is at index 0. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the location of the nth genotype/nth group member, or GSC_INVALID_GENO_LOCATION if the index is invalid. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01980">1980</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaaccc34724148be204eb4d8515000dfd5_cgraph.png" border="0" usemap="#agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_cgraph" alt=""/></div>
<map name="agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_cgraph" id="agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_cgraph">
<area shape="rect" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="5,57,130,85"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="208,5,315,33"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="178,57,345,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="201,109,322,137"/>
</map>
</div>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaaccc34724148be204eb4d8515000dfd5_icgraph.png" border="0" usemap="#agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_icgraph" alt=""/></div>
<map name="agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_icgraph" id="agroup__iterators_gaaccc34724148be204eb4d8515000dfd5_icgraph">
<area shape="rect" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="702,149,827,177"/>
<area shape="rect" href="sim-operations_8c.html#a0fcef09a211d57fadb1dfa0b6d102901" title="parentChooser function parameter for gsc_make_random_crosses." alt="" coords="480,5,653,49"/>
<area shape="rect" href="sim-operations_8c.html#a2cee0d2e90de12744fd077b20a7db9bd" title="parentChooser function parameter for gsc_make_random_crosses_between." alt="" coords="479,73,654,117"/>
<area shape="rect" href="sim-operations_8c.html#a2c3d189234e1ac86e27a1ccaace3c363" title="parentChooser function parameter for gsc_make_targeted_crosses." alt="" coords="480,141,653,185"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="491,209,643,237"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="481,261,652,289"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,13,429,41"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,73,429,117"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,149,431,177"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,81,192,125"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,149,195,177"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,201,190,245"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="280,209,394,237"/>
</map>
</div>

</div>
</div>
<a id="ga99820e199424a3ec8ba7e3c58802fba9" name="ga99820e199424a3ec8ba7e3c58802fba9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga99820e199424a3ec8ba7e3c58802fba9">&#9670;&nbsp;</a></span>gsc_set_bidirectional_iter_to_end()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_set_bidirectional_iter_to_end </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialise a Bidirectional iterator to the end of its sequence. </p>
<p >Can be used to reset a <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> so that it is pointing at the very last member of the group it is looping through.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga7137b74725f2990fbc42af04f9fd0669">set_bidirectional_iter_to_end</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">it</td><td>BidirectioanlIterator to initialise </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>location of the last member of the sequence. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01654">1654</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga99820e199424a3ec8ba7e3c58802fba9_cgraph.png" border="0" usemap="#agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_cgraph" alt=""/></div>
<map name="agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_cgraph" id="agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_cgraph">
<area shape="rect" title="Initialise a Bidirectional iterator to the end of its sequence." alt="" coords="5,5,149,49"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="197,13,364,41"/>
</map>
</div>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_ga99820e199424a3ec8ba7e3c58802fba9_icgraph.png" border="0" usemap="#agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_icgraph" alt=""/></div>
<map name="agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_icgraph" id="agroup__iterators_ga99820e199424a3ec8ba7e3c58802fba9_icgraph">
<area shape="rect" title="Initialise a Bidirectional iterator to the end of its sequence." alt="" coords="197,5,340,49"/>
<area shape="rect" href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871" title="Get the previous location from a bidirectional iterator." alt="" coords="5,13,149,41"/>
</map>
</div>

</div>
</div>
<a id="gaa4fa2aec3b7f49f9925fe82c8ff38dcd" name="gaa4fa2aec3b7f49f9925fe82c8ff38dcd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa4fa2aec3b7f49f9925fe82c8ff38dcd">&#9670;&nbsp;</a></span>gsc_set_bidirectional_iter_to_start()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_set_bidirectional_iter_to_start </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Initialise a Bidirectional iterator to the start of its sequence. </p>
<p >Can be used to reset a <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> so that it is pointing at the very first member of the group it is looping through.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga60fa7078ac84231afe8ec7cd9cf850ef">set_bidirectional_iter_to_start</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">it</td><td>BidirectioanlIterator to initialise </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>location of the first member of the sequence. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01576">1576</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__iterators_gaa4fa2aec3b7f49f9925fe82c8ff38dcd_icgraph.png" border="0" usemap="#agroup__iterators_gaa4fa2aec3b7f49f9925fe82c8ff38dcd_icgraph" alt=""/></div>
<map name="agroup__iterators_gaa4fa2aec3b7f49f9925fe82c8ff38dcd_icgraph" id="agroup__iterators_gaa4fa2aec3b7f49f9925fe82c8ff38dcd_icgraph">
<area shape="rect" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="1367,668,1511,712"/>
<area shape="rect" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="966,5,1105,49"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="953,676,1118,704"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="721,728,860,756"/>
<area shape="rect" href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe" title="Replace all occurences of a given allele with a different symbol representation." alt="" coords="947,73,1124,101"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="1190,447,1319,475"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="947,833,1124,861"/>
<area shape="rect" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361" title="Prints a matrix of genotype information to a file." alt="" coords="929,924,1142,952"/>
<area shape="rect" href="group__calculators.html#ga30689b0f03641259091a72d065951e29" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="727,676,854,704"/>
<area shape="rect" href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="492,676,624,704"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="43,423,156,451"/>
<area shape="rect" href="sim-operations_8c.html#a240267fbee56d8075f7b8fe10c81aaa4" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="481,780,635,824"/>
<area shape="rect" href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1" title="Prints breeding values of genotypes in the simulation to a file." alt="" coords="507,728,608,756"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="247,780,410,824"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="36,564,163,592"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="5,711,194,739"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="7,763,192,791"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="7,815,192,859"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="5,883,194,911"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="38,659,161,687"/>
<area shape="rect" href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c" title="Combine a set of groups into one group." alt="" coords="963,228,1108,256"/>
<area shape="rect" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="956,280,1115,324"/>
<area shape="rect" href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b" title="Split by some allocator (generic function)" alt="" coords="260,83,397,127"/>
<area shape="rect" href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d" title="Allocates the genotypes with values of a label in a particular range to a new group." alt="" coords="264,199,393,243"/>
<area shape="rect" href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b" title="Allocates the genotypes with a particular value of a label to a new group." alt="" coords="264,387,393,431"/>
<area shape="rect" href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a" title="Flip a coin for each member of the group to decide if it should be moved to the new group." alt="" coords="262,455,395,499"/>
<area shape="rect" href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2" title="Split by some quality (generic function)" alt="" coords="728,168,853,212"/>
<area shape="rect" href="sim-operations_8c.html#a94b58634907af7e24ac22b543b8fcbe3" title="parentChooser function parameter for gsc_make_clones." alt="" coords="949,524,1122,568"/>
<area shape="rect" href="sim-operations_8c.html#aafe958ddf4ee8c990c521d6ebf60db71" title="parentChooser function parameter for gsc_self_n_times." alt="" coords="242,651,415,695"/>
<area shape="rect" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83" title="Identify group numbers that currently have members." alt="" coords="708,288,873,316"/>
<area shape="rect" href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2" title="Function to identify the next n sequential integers that do not identify a group that currently has m..." alt="" coords="463,265,652,293"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="474,369,642,397"/>
<area shape="rect" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5" title="Allocate each member of the group to one of n groups with custom probabilities for each group." alt="" coords="14,13,185,41"/>
<area shape="rect" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="20,65,179,93"/>
<area shape="rect" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="22,117,177,145"/>
<area shape="rect" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621" title="Allocate each member of the group to one of n groups with equal probability." alt="" coords="33,169,166,213"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="254,267,403,311"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="252,335,405,363"/>
<area shape="rect" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33" title="Split a group into families by their pedigrees." alt="" coords="482,145,634,173"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="485,197,630,241"/>
<area shape="rect" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1" title="Split a group into n one&#45;member groups." alt="" coords="474,93,641,121"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="713,891,868,935"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="721,835,860,863"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="473,848,642,876"/>
<area shape="rect" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file." alt="" coords="715,1101,866,1145"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="700,989,881,1017"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="479,1140,637,1168"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="481,952,635,996"/>
<area shape="rect" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="478,1020,638,1064"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="487,1088,629,1116"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
