<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Saving Functions</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.02</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">Saving Functions</div></div>
</div><!--header-->
<div class="contents">

<p>For saving persistent simulation results.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga364d2ea13aadfa347d9c12223b4cc4d1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga364d2ea13aadfa347d9c12223b4cc4d1">gsc_save_markerblocks</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b)</td></tr>
<tr class="memdesc:ga364d2ea13aadfa347d9c12223b4cc4d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga364d2ea13aadfa347d9c12223b4cc4d1">More...</a><br /></td></tr>
<tr class="separator:ga364d2ea13aadfa347d9c12223b4cc4d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga12e6eb5a3dccf1c79fd1b68df69c2668"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga12e6eb5a3dccf1c79fd1b68df69c2668">gsc_save_names_header</a> (FILE *f, size_t n, const char **names)</td></tr>
<tr class="memdesc:ga12e6eb5a3dccf1c79fd1b68df69c2668"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints a list of names as a tab-separated row to a file.  <a href="group__savers.html#ga12e6eb5a3dccf1c79fd1b68df69c2668">More...</a><br /></td></tr>
<tr class="separator:ga12e6eb5a3dccf1c79fd1b68df69c2668"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7bca249d1a6132b5faff18069a0e37b5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga7bca249d1a6132b5faff18069a0e37b5">gsc_save_allele_matrix</a> (FILE *f, const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:ga7bca249d1a6132b5faff18069a0e37b5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="group__savers.html#ga7bca249d1a6132b5faff18069a0e37b5">More...</a><br /></td></tr>
<tr class="separator:ga7bca249d1a6132b5faff18069a0e37b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga25ecc575d779bed4f46626cd61fbe65d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga25ecc575d779bed4f46626cd61fbe65d">gsc_save_transposed_allele_matrix</a> (FILE *f, const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m, const char **markers)</td></tr>
<tr class="memdesc:ga25ecc575d779bed4f46626cd61fbe65d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="group__savers.html#ga25ecc575d779bed4f46626cd61fbe65d">More...</a><br /></td></tr>
<tr class="separator:ga25ecc575d779bed4f46626cd61fbe65d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab849eb8ece54052720d78fe248d09298"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gab849eb8ece54052720d78fe248d09298">gsc_save_group_genotypes</a> (FILE *f, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gab849eb8ece54052720d78fe248d09298"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#gab849eb8ece54052720d78fe248d09298">More...</a><br /></td></tr>
<tr class="separator:gab849eb8ece54052720d78fe248d09298"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac0853546f2c1334ec45bc43b693f8ce0"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gac0853546f2c1334ec45bc43b693f8ce0">gsc_save_transposed_group_genotypes</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gac0853546f2c1334ec45bc43b693f8ce0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#gac0853546f2c1334ec45bc43b693f8ce0">More...</a><br /></td></tr>
<tr class="separator:gac0853546f2c1334ec45bc43b693f8ce0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9678bbe9c8db5f7774188b9574c0c09c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga9678bbe9c8db5f7774188b9574c0c09c">gsc_save_count_matrix</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char allele)</td></tr>
<tr class="memdesc:ga9678bbe9c8db5f7774188b9574c0c09c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file.  <a href="group__savers.html#ga9678bbe9c8db5f7774188b9574c0c09c">More...</a><br /></td></tr>
<tr class="separator:ga9678bbe9c8db5f7774188b9574c0c09c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga47adc7657700b186067b2c26dabf5b13"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga47adc7657700b186067b2c26dabf5b13">gsc_save_group_count_matrix</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char allele, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga47adc7657700b186067b2c26dabf5b13"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="group__savers.html#ga47adc7657700b186067b2c26dabf5b13">More...</a><br /></td></tr>
<tr class="separator:ga47adc7657700b186067b2c26dabf5b13"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2bc846644aa32182f20feaa67ff3df8d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga2bc846644aa32182f20feaa67ff3df8d">gsc_save_one_step_pedigree</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga2bc846644aa32182f20feaa67ff3df8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file.  <a href="group__savers.html#ga2bc846644aa32182f20feaa67ff3df8d">More...</a><br /></td></tr>
<tr class="separator:ga2bc846644aa32182f20feaa67ff3df8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad10b44011b806daa35f8378eb90c5365"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gad10b44011b806daa35f8378eb90c5365">gsc_save_group_one_step_pedigree</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:gad10b44011b806daa35f8378eb90c5365"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="group__savers.html#gad10b44011b806daa35f8378eb90c5365">More...</a><br /></td></tr>
<tr class="separator:gad10b44011b806daa35f8378eb90c5365"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6a5bcd7945c34e0ec99ccc92311d947c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga6a5bcd7945c34e0ec99ccc92311d947c">gsc_save_full_pedigree</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga6a5bcd7945c34e0ec99ccc92311d947c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file.  <a href="group__savers.html#ga6a5bcd7945c34e0ec99ccc92311d947c">More...</a><br /></td></tr>
<tr class="separator:ga6a5bcd7945c34e0ec99ccc92311d947c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga34df718dd14346f72cc20928351cc4d2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga34df718dd14346f72cc20928351cc4d2">gsc_save_group_full_pedigree</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga34df718dd14346f72cc20928351cc4d2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="group__savers.html#ga34df718dd14346f72cc20928351cc4d2">More...</a><br /></td></tr>
<tr class="separator:ga34df718dd14346f72cc20928351cc4d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga78f467f0dd08c5501b804caca1e0c6b9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga78f467f0dd08c5501b804caca1e0c6b9">gsc_save_allelematrix_full_pedigree</a> (FILE *f, const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *parents)</td></tr>
<tr class="memdesc:ga78f467f0dd08c5501b804caca1e0c6b9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> to a file.  <a href="group__savers.html#ga78f467f0dd08c5501b804caca1e0c6b9">More...</a><br /></td></tr>
<tr class="separator:ga78f467f0dd08c5501b804caca1e0c6b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8e5d08dbcfc4cc77d1d1c6603b75a2ed"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed">gsc_save_parents_of</a> (FILE *f, const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> p1, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> p2)</td></tr>
<tr class="memdesc:ga8e5d08dbcfc4cc77d1d1c6603b75a2ed"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed">More...</a><br /></td></tr>
<tr class="separator:ga8e5d08dbcfc4cc77d1d1c6603b75a2ed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9081436408a37abb9d37e035b3cfaa87"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga9081436408a37abb9d37e035b3cfaa87">gsc_save_bvs</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga9081436408a37abb9d37e035b3cfaa87"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file.  <a href="group__savers.html#ga9081436408a37abb9d37e035b3cfaa87">More...</a><br /></td></tr>
<tr class="separator:ga9081436408a37abb9d37e035b3cfaa87"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3453d7fc7d552e7c254cf00788c45506"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga3453d7fc7d552e7c254cf00788c45506">gsc_save_group_bvs</a> (FILE *f, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga3453d7fc7d552e7c254cf00788c45506"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in a group to a file.  <a href="group__savers.html#ga3453d7fc7d552e7c254cf00788c45506">More...</a><br /></td></tr>
<tr class="separator:ga3453d7fc7d552e7c254cf00788c45506"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga707e7b98f0321e3f2460abb234e1bb88"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga707e7b98f0321e3f2460abb234e1bb88">gsc_save_manual_bvs</a> (FILE *f, const <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *e, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *ids, const char **names)</td></tr>
<tr class="memdesc:ga707e7b98f0321e3f2460abb234e1bb88"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>gsc_save_bvs</code> or <code>gsc_save_group_bvs</code>.  <a href="group__savers.html#ga707e7b98f0321e3f2460abb234e1bb88">More...</a><br /></td></tr>
<tr class="separator:ga707e7b98f0321e3f2460abb234e1bb88"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p >For saving persistent simulation results. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga7bca249d1a6132b5faff18069a0e37b5" name="ga7bca249d1a6132b5faff18069a0e37b5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7bca249d1a6132b5faff18069a0e37b5">&#9670;&nbsp;</a></span>gsc_save_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file. </p>
<p >Uses the following format:</p>
<p >[id]OR[name] [allele pairs for each marker]</p>
<p >[id]OR[name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga80f72e29ba4545cbe72a2cbb6cd023de">save_allele_matrix</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> whose data we print </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga7bca249d1a6132b5faff18069a0e37b5_icgraph.png" border="0" usemap="#agroup__savers_ga7bca249d1a6132b5faff18069a0e37b5_icgraph" alt=""/></div>
<map name="agroup__savers_ga7bca249d1a6132b5faff18069a0e37b5_icgraph" id="agroup__savers_ga7bca249d1a6132b5faff18069a0e37b5_icgraph">
<area shape="rect" title="Prints all the genotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="892,140,1047,168"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
</map>
</div>

</div>
</div>
<a id="ga78f467f0dd08c5501b804caca1e0c6b9" name="ga78f467f0dd08c5501b804caca1e0c6b9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga78f467f0dd08c5501b804caca1e0c6b9">&#9670;&nbsp;</a></span>gsc_save_allelematrix_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_allelematrix_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>parents</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a single <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p >Note this does not follow through the linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> containing the genotypes to print </td></tr>
    <tr><td class="paramname">parents</td><td>pointer to an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given genotypes. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga78f467f0dd08c5501b804caca1e0c6b9_cgraph.png" border="0" usemap="#agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_cgraph" alt=""/></div>
<map name="agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_cgraph" id="agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single gsc_AlleleMatrix to a file." alt="" coords="5,24,154,68"/>
<area shape="rect" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" title="Recursively save the parents of a particular id to a file." alt="" coords="202,32,344,60"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="398,5,538,33"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="392,57,543,85"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="591,24,741,68"/>
</map>
</div>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga78f467f0dd08c5501b804caca1e0c6b9_icgraph.png" border="0" usemap="#agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_icgraph" alt=""/></div>
<map name="agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_icgraph" id="agroup__savers_ga78f467f0dd08c5501b804caca1e0c6b9_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single gsc_AlleleMatrix to a file." alt="" coords="892,132,1040,176"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
</map>
</div>

</div>
</div>
<a id="ga9081436408a37abb9d37e035b3cfaa87" name="ga9081436408a37abb9d37e035b3cfaa87"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga9081436408a37abb9d37e035b3cfaa87">&#9670;&nbsp;</a></span>gsc_save_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name] [BV]</p>
<p >[id] [name] [BV]</p>
<p >...</p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must have loaded marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gaf3caad0e03e9bac0a06acdb0e6d80e8f">save_bvs</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga9081436408a37abb9d37e035b3cfaa87_cgraph.png" border="0" usemap="#agroup__savers_ga9081436408a37abb9d37e035b3cfaa87_cgraph" alt=""/></div>
<map name="agroup__savers_ga9081436408a37abb9d37e035b3cfaa87_cgraph" id="agroup__savers_ga9081436408a37abb9d37e035b3cfaa87_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in the gsc_SimData to a file." alt="" coords="5,299,107,327"/>
<area shape="rect" href="group__calculators.html#gaae6cddc5291e4ad120ac28a7d043c7fa" title="Calculates the fitness metric/breeding value for each genotype in the gsc_AlleleMatrix,..." alt="" coords="175,217,302,245"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="396,329,530,357"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="155,351,322,379"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="370,5,555,49"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="389,73,536,117"/>
<area shape="rect" href="group__calculators.html#gae37153681a5a2b78b3accbe7e01f65d8" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="393,141,533,185"/>
<area shape="rect" href="group__calculators.html#ga34b1d566ad458097f95823278f6bbd8e" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="393,209,533,253"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="372,277,553,305"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="603,277,725,305"/>
</map>
</div>

</div>
</div>
<a id="ga9678bbe9c8db5f7774188b9574c0c09c" name="ga9678bbe9c8db5f7774188b9574c0c09c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga9678bbe9c8db5f7774188b9574c0c09c">&#9670;&nbsp;</a></span>gsc_save_count_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_count_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file. </p>
<p >The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele count for each marker]</p>
<p >[marker name] [allele count for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga311d03837ccf30973a719e2d751d1f48">save_count_matrix</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga9678bbe9c8db5f7774188b9574c0c09c_cgraph.png" border="0" usemap="#agroup__savers_ga9678bbe9c8db5f7774188b9574c0c09c_cgraph" alt=""/></div>
<map name="agroup__savers_ga9678bbe9c8db5f7774188b9574c0c09c_cgraph" id="agroup__savers_ga9678bbe9c8db5f7774188b9574c0c09c_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in the gsc_SimData ..." alt="" coords="5,44,163,72"/>
<area shape="rect" href="group__calculators.html#gae2d46345f01bc74e45a4d07b2a531113" title="Calculates the number of times at each marker that a particular allele appears for each genotype in t..." alt="" coords="216,5,340,49"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="211,73,345,101"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="393,13,574,41"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="622,13,744,41"/>
</map>
</div>

</div>
</div>
<a id="ga6a5bcd7945c34e0ec99ccc92311d947c" name="ga6a5bcd7945c34e0ec99ccc92311d947c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga6a5bcd7945c34e0ec99ccc92311d947c">&#9670;&nbsp;</a></span>gsc_save_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga6e1ab8f7cd21180ca9d6942ac3f4c026">save_full_pedigree</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing all genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga6a5bcd7945c34e0ec99ccc92311d947c_cgraph.png" border="0" usemap="#agroup__savers_ga6a5bcd7945c34e0ec99ccc92311d947c_cgraph" alt=""/></div>
<map name="agroup__savers_ga6a5bcd7945c34e0ec99ccc92311d947c_cgraph" id="agroup__savers_ga6a5bcd7945c34e0ec99ccc92311d947c_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in the gsc_SimData to a file." alt="" coords="5,32,160,60"/>
<area shape="rect" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" title="Recursively save the parents of a particular id to a file." alt="" coords="208,32,350,60"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="404,5,544,33"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="398,57,549,85"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="597,24,747,68"/>
</map>
</div>

</div>
</div>
<a id="ga3453d7fc7d552e7c254cf00788c45506" name="ga3453d7fc7d552e7c254cf00788c45506"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3453d7fc7d552e7c254cf00788c45506">&#9670;&nbsp;</a></span>gsc_save_group_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_group_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[group member id] [group member name] [BV]</p>
<p >[group member id] [group member name] [BV]</p>
<p >...</p>
<p >The <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> must have loaded marker effects for this function to succeed.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga9de3a7edd3958006119e6e6ea30d9593">save_group_bvs</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the breeding values of. </td></tr>
    <tr><td class="paramname">effID</td><td>Identifier of the marker effect set to be used to calculate these breeding values </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga3453d7fc7d552e7c254cf00788c45506_cgraph.png" border="0" usemap="#agroup__savers_ga3453d7fc7d552e7c254cf00788c45506_cgraph" alt=""/></div>
<map name="agroup__savers_ga3453d7fc7d552e7c254cf00788c45506_cgraph" id="agroup__savers_ga3453d7fc7d552e7c254cf00788c45506_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in a group to a file." alt="" coords="5,254,146,282"/>
<area shape="rect" href="group__calculators.html#ga381bf7e42397f428a322dd8e118a4243" title="Calculates the fitness metric/breeding value for each genotype in the gsc_AlleleMatrix in a certain g..." alt="" coords="194,377,360,405"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="838,181,975,209"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="434,557,568,585"/>
<area shape="rect" href="group__supporters.html#gac15b90373af3e2828a786215e90963a8" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="417,609,584,637"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="650,27,781,55"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="201,79,353,107"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="408,233,593,277"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="427,369,574,413"/>
<area shape="rect" href="group__calculators.html#ga12c13aaf31fccff2f9ddd612035c426a" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="431,301,571,345"/>
<area shape="rect" href="group__calculators.html#ga160313a4276db1a2897ab0d161f5caa8" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="431,437,571,481"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="410,505,591,533"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="641,309,790,337"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="655,505,776,533"/>
</map>
</div>

</div>
</div>
<a id="ga47adc7657700b186067b2c26dabf5b13" name="ga47adc7657700b186067b2c26dabf5b13"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga47adc7657700b186067b2c26dabf5b13">&#9670;&nbsp;</a></span>gsc_save_group_count_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_group_count_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in a group to a file. </p>
<p >The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele count for each marker]</p>
<p >[marker name] [allele count for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga5c3ad1442e8ccb084fde3adc034c86ff">save_group_count_matrix</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the allele count of. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga47adc7657700b186067b2c26dabf5b13_cgraph.png" border="0" usemap="#agroup__savers_ga47adc7657700b186067b2c26dabf5b13_cgraph" alt=""/></div>
<map name="agroup__savers_ga47adc7657700b186067b2c26dabf5b13_cgraph" id="agroup__savers_ga47adc7657700b186067b2c26dabf5b13_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="5,179,159,223"/>
<area shape="rect" href="group__calculators.html#ga12c13aaf31fccff2f9ddd612035c426a" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="227,40,368,84"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="636,108,774,136"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="230,160,365,188"/>
<area shape="rect" href="group__maths.html#gae001e0609ac52df335530f9473cb57e1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="207,212,388,240"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="447,263,577,291"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="436,315,588,343"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="438,5,586,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="452,211,573,239"/>
</map>
</div>

</div>
</div>
<a id="ga34df718dd14346f72cc20928351cc4d2" name="ga34df718dd14346f72cc20928351cc4d2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga34df718dd14346f72cc20928351cc4d2">&#9670;&nbsp;</a></span>gsc_save_group_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_group_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p >[id] [name]=([parent pedigree])</p>
<p >...</p>
<p >Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p >The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga7eba288c46c5645de15761d784673c01">save_group_full_pedigree</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the pedigree of. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga34df718dd14346f72cc20928351cc4d2_cgraph.png" border="0" usemap="#agroup__savers_ga34df718dd14346f72cc20928351cc4d2_cgraph" alt=""/></div>
<map name="agroup__savers_ga34df718dd14346f72cc20928351cc4d2_cgraph" id="agroup__savers_ga34df718dd14346f72cc20928351cc4d2_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="5,101,143,145"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="212,5,343,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="419,84,556,112"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="201,109,354,137"/>
<area shape="rect" href="group__collgetters.html#ga3010fa26c1c6a31c61286e6c60c97dd5" title="Gets the ids of either the first or second parent of each member of the group." alt="" coords="191,161,364,189"/>
<area shape="rect" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" title="Recursively save the parents of a particular id to a file." alt="" coords="206,237,349,265"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="418,211,558,239"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="412,263,563,291"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="611,229,761,273"/>
</map>
</div>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga34df718dd14346f72cc20928351cc4d2_icgraph.png" border="0" usemap="#agroup__savers_ga34df718dd14346f72cc20928351cc4d2_icgraph" alt=""/></div>
<map name="agroup__savers_ga34df718dd14346f72cc20928351cc4d2_icgraph" id="agroup__savers_ga34df718dd14346f72cc20928351cc4d2_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="223,5,360,49"/>
<area shape="rect" href="group__collgetters.html#ga62d40b0b1cb6c5ded508b8a5f161d38a" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="5,13,175,41"/>
</map>
</div>

</div>
</div>
<a id="gab849eb8ece54052720d78fe248d09298" name="gab849eb8ece54052720d78fe248d09298"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab849eb8ece54052720d78fe248d09298">&#9670;&nbsp;</a></span>gsc_save_group_genotypes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_group_genotypes </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">    [marker name]   [marker name]
</pre><p> [id]OR[name] [allele pairs for each marker]</p>
<p >[id]OR[name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<p ><em>Has short name:</em> save_group_genotype</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gab849eb8ece54052720d78fe248d09298_cgraph.png" border="0" usemap="#agroup__savers_gab849eb8ece54052720d78fe248d09298_cgraph" alt=""/></div>
<map name="agroup__savers_gab849eb8ece54052720d78fe248d09298_cgraph" id="agroup__savers_gab849eb8ece54052720d78fe248d09298_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,84,187,112"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="237,5,385,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="435,84,572,112"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="246,109,376,137"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="235,161,387,189"/>
</map>
</div>

</div>
</div>
<a id="gad10b44011b806daa35f8378eb90c5365" name="gad10b44011b806daa35f8378eb90c5365"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad10b44011b806daa35f8378eb90c5365">&#9670;&nbsp;</a></span>gsc_save_group_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_group_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in a group to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[group member name] [parent 1 name] [parent 2 name]</p>
<p >[group member name] [parent 1 name] [parent 2 name]</p>
<p >...</p>
<p >The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga5ed0a70f29334b047cab8dfcdf63805f">save_group_one_step_pedigree</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the immediate parents of. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gad10b44011b806daa35f8378eb90c5365_cgraph.png" border="0" usemap="#agroup__savers_gad10b44011b806daa35f8378eb90c5365_cgraph" alt=""/></div>
<map name="agroup__savers_gad10b44011b806daa35f8378eb90c5365_cgraph" id="agroup__savers_gad10b44011b806daa35f8378eb90c5365_cgraph">
<area shape="rect" title="Print the parents of each genotype in a group to a file." alt="" coords="5,101,148,145"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="217,5,348,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="423,84,560,112"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="206,109,359,137"/>
<area shape="rect" href="group__collgetters.html#ga3010fa26c1c6a31c61286e6c60c97dd5" title="Gets the ids of either the first or second parent of each member of the group." alt="" coords="196,161,369,189"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="212,213,353,241"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="417,205,566,249"/>
</map>
</div>

</div>
</div>
<a id="ga707e7b98f0321e3f2460abb234e1bb88" name="ga707e7b98f0321e3f2460abb234e1bb88"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga707e7b98f0321e3f2460abb234e1bb88">&#9670;&nbsp;</a></span>gsc_save_manual_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_manual_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *&#160;</td>
          <td class="paramname"><em>ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>names</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>gsc_save_bvs</code> or <code>gsc_save_group_bvs</code>. </p>
<p >The following tab-separated format is used:</p>
<p >[id] [name] [BV]</p>
<p >[id] [name] [BV]</p>
<p >...</p>
<p >The function assumes the array of ids, array of names, and columns of the <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> are ordered the same, so a single index produces the corresponding value from each. It is used by as-you-go savers within crossing functions. For general use consider <code>gsc_save_bvs</code> or <code>gsc_save_group_bvs</code> instead.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> containing the BVs in the first row. </td></tr>
    <tr><td class="paramname">ids</td><td>array of ids to print alongside the BVs. </td></tr>
    <tr><td class="paramname">names</td><td>array of names to print alongside the BVs. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga707e7b98f0321e3f2460abb234e1bb88_icgraph.png" border="0" usemap="#agroup__savers_ga707e7b98f0321e3f2460abb234e1bb88_icgraph" alt=""/></div>
<map name="agroup__savers_ga707e7b98f0321e3f2460abb234e1bb88_icgraph" id="agroup__savers_ga707e7b98f0321e3f2460abb234e1bb88_icgraph">
<area shape="rect" title="Print the provided breeding values of each provided name and id to a file, with the same format as a ..." alt="" coords="892,140,1042,168"/>
<area shape="rect" href="sim-operations_8c.html#a7072c6a3834472e7fbd768c0275fdc54" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
</map>
</div>

</div>
</div>
<a id="ga364d2ea13aadfa347d9c12223b4cc4d1" name="ga364d2ea13aadfa347d9c12223b4cc4d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga364d2ea13aadfa347d9c12223b4cc4d1">&#9670;&nbsp;</a></span>gsc_save_markerblocks()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_markerblocks </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the markers contained in a set of blocks to a file. </p>
<p >Column separators are tabs.</p>
<p >The printing format is:</p>
<p >Chrom Pos Name Class Markers</p>
<p >0 0 b0 b m1;m2;m3;m4;</p>
<p >0 0 b0 b m7;m9;</p>
<p >...</p>
<p >where m1, m2, m3, m4 are the names of the markers in the first block and m7 and m9 are the names of the markers in the second block. Note that chromosome numbers and positions are not actually specified, just replaced with 0. Only the last column is meaningful.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gad985933c857f19424a26483b0ca567de">save_markerblocks</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> whose data we print </td></tr>
    <tr><td class="paramname">b</td><td><a class="el" href="structgsc___marker_blocks.html" title="A struct used to store a set of blocks of markers.">gsc_MarkerBlocks</a> struct containing the groupings of markers to print. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga12e6eb5a3dccf1c79fd1b68df69c2668" name="ga12e6eb5a3dccf1c79fd1b68df69c2668"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga12e6eb5a3dccf1c79fd1b68df69c2668">&#9670;&nbsp;</a></span>gsc_save_names_header()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_names_header </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>names</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints a list of names as a tab-separated row to a file. </p>
<p >Used, for example, for printing marker names as a header in a genotype output file.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#ga0937dcb150cfddfc0b14e4df5322b760">save_names_header</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">n</td><td>number of names to print </td></tr>
    <tr><td class="paramname">names</td><td>list of strings/names to print. Must contain at least [n] names. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga12e6eb5a3dccf1c79fd1b68df69c2668_icgraph.png" border="0" usemap="#agroup__savers_ga12e6eb5a3dccf1c79fd1b68df69c2668_icgraph" alt=""/></div>
<map name="agroup__savers_ga12e6eb5a3dccf1c79fd1b68df69c2668_icgraph" id="agroup__savers_ga12e6eb5a3dccf1c79fd1b68df69c2668_icgraph">
<area shape="rect" title="Prints a list of names as a tab&#45;separated row to a file." alt="" coords="898,140,1065,168"/>
<area shape="rect" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="689,132,850,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
</map>
</div>

</div>
</div>
<a id="ga2bc846644aa32182f20feaa67ff3df8d" name="ga2bc846644aa32182f20feaa67ff3df8d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2bc846644aa32182f20feaa67ff3df8d">&#9670;&nbsp;</a></span>gsc_save_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to a file. </p>
<p >The following tab-separated format is used:</p>
<p >[genotype name] [parent 1 name] [parent 2 name]</p>
<p >[genotype name] [parent 1 name] [parent 2 name]</p>
<p >...</p>
<p >The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> struct.</p>
<p >If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gac4a53ada7d4ee19e988f3d1865c46cdf">save_one_step_pedigree</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the genotypes and their pedigrees </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga2bc846644aa32182f20feaa67ff3df8d_cgraph.png" border="0" usemap="#agroup__savers_ga2bc846644aa32182f20feaa67ff3df8d_cgraph" alt=""/></div>
<map name="agroup__savers_ga2bc846644aa32182f20feaa67ff3df8d_cgraph" id="agroup__savers_ga2bc846644aa32182f20feaa67ff3df8d_cgraph">
<area shape="rect" title="Print the parents of each genotype in the gsc_SimData to a file." alt="" coords="5,13,197,41"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="245,13,385,41"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="433,5,582,49"/>
</map>
</div>

</div>
</div>
<a id="ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" name="ga8e5d08dbcfc4cc77d1d1c6603b75a2ed"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed">&#9670;&nbsp;</a></span>gsc_save_parents_of()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_parents_of </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>p2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Recursively save the parents of a particular id to a file. </p>
<p >It saves using the following format:</p>
<ul>
<li>no characters are saved if the parents of the id are unknown/0</li>
<li>if the id has one parent, repeated twice (it was produced by selfing), print "=(parentname)", where parentname is the name of the parent or its id if it does not have one, followed by whatever is printed by a call to this function on the parent's id.</li>
<li>if the id has two separate parents, print "=(parent1name,parent2name)", where parent1name and parent2name are the name2 of the two parents or their ids if they does not have names, each name immediately followed by whatever is printed by a call to this function on the corresponding parent's id.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given id. </td></tr>
    <tr><td class="paramname">p1</td><td>the session-unique id of the first parent to be saved. </td></tr>
    <tr><td class="paramname">p2</td><td>the session-unique id of the second parent to be saved. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-15" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-15-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-15-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-15-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_cgraph.png" border="0" usemap="#agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_cgraph" alt=""/></div>
<map name="agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_cgraph" id="agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_cgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="5,32,148,60"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="201,5,342,33"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="196,57,347,85"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="395,24,544,68"/>
</map>
</div>
<div id="dynsection-16" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-16-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-16-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-16-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_icgraph.png" border="0" usemap="#agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_icgraph" alt=""/></div>
<map name="agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_icgraph" id="agroup__savers_ga8e5d08dbcfc4cc77d1d1c6603b75a2ed_icgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="1109,200,1251,228"/>
<area shape="rect" href="group__savers.html#ga78f467f0dd08c5501b804caca1e0c6b9" title="Print the full known pedigree of each genotype in a single gsc_AlleleMatrix to a file." alt="" coords="910,132,1058,176"/>
<area shape="rect" href="group__savers.html#ga6a5bcd7945c34e0ec99ccc92311d947c" title="Print the full known pedigree of each genotype in the gsc_SimData to a file." alt="" coords="907,200,1061,228"/>
<area shape="rect" href="group__savers.html#ga34df718dd14346f72cc20928351cc4d2" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="915,252,1053,296"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="697,132,851,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
<area shape="rect" href="group__collgetters.html#ga62d40b0b1cb6c5ded508b8a5f161d38a" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="689,260,859,288"/>
</map>
</div>

</div>
</div>
<a id="ga25ecc575d779bed4f46626cd61fbe65d" name="ga25ecc575d779bed4f46626cd61fbe65d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga25ecc575d779bed4f46626cd61fbe65d">&#9670;&nbsp;</a></span>gsc_save_transposed_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_transposed_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the gene data saved in the linked list starting with <code>m</code> to the file. </p>
<p >Uses the following format: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele pairs for each marker]</p>
<p >[marker name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<p ><em>Has short name:</em> <a class="el" href="group__shortnames.html#gaa98ff50f311a1b8fa27f2c13ecfbd816">save_transposed_allele_matrix</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="gac0853546f2c1334ec45bc43b693f8ce0" name="gac0853546f2c1334ec45bc43b693f8ce0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac0853546f2c1334ec45bc43b693f8ce0">&#9670;&nbsp;</a></span>gsc_save_transposed_group_genotypes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_save_transposed_group_genotypes </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">[id]OR[name]    [id]OR[name] ...
</pre><p> [marker name] [allele pairs for each marker]</p>
<p >[marker name] [allele pairs for each marker]</p>
<p >...</p>
<p >ID will be printed if the genotype does not have a name saved</p>
<p ><em>Has short name:</em> save_transposed_group_genotype</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p >fprintf(f, "\n");</p>
<div id="dynsection-17" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-17-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-17-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-17-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__savers_gac0853546f2c1334ec45bc43b693f8ce0_cgraph.png" border="0" usemap="#agroup__savers_gac0853546f2c1334ec45bc43b693f8ce0_cgraph" alt=""/></div>
<map name="agroup__savers_gac0853546f2c1334ec45bc43b693f8ce0_cgraph" id="agroup__savers_gac0853546f2c1334ec45bc43b693f8ce0_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,76,152,120"/>
<area shape="rect" href="group__collgetters.html#gab256e999fe9540658a2a3aba11260cb5" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="202,5,350,33"/>
<area shape="rect" href="group__collgetters.html#gae9131a8c72149706f67f6e4a48add79c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="400,84,537,112"/>
<area shape="rect" href="group__collgetters.html#gae5fac6863433c073871766b80aa6d966" title="Gets the ids of each member of the group." alt="" coords="211,109,341,137"/>
<area shape="rect" href="group__collgetters.html#ga52a8eb90f5a6a1f9dabea81955d0bb79" title="Gets a shallow copy of the names of each member of the group." alt="" coords="200,161,352,189"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
