<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Saving Functions</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">genomicSimulationC
   &#160;<span id="projectnumber">0.1-4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Saving Functions</div>  </div>
</div><!--header-->
<div class="contents">

<p>For saving persistent simulation results.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga4d63cd249222ab25b3112106813f29b2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga4d63cd249222ab25b3112106813f29b2">save_simdata</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga4d63cd249222ab25b3112106813f29b2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga4d63cd249222ab25b3112106813f29b2">More...</a><br /></td></tr>
<tr class="separator:ga4d63cd249222ab25b3112106813f29b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga2d795d5594b914eac4e5f6fcee3afbda">save_marker_blocks</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b)</td></tr>
<tr class="memdesc:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga2d795d5594b914eac4e5f6fcee3afbda">More...</a><br /></td></tr>
<tr class="separator:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga42a237afba9dc8f3a7339cb97d929cb6">save_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="group__savers.html#ga42a237afba9dc8f3a7339cb97d929cb6">More...</a><br /></td></tr>
<tr class="separator:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaadf131f9f90546d6ab9355cbaa1cf6e1">save_transposed_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="group__savers.html#gaadf131f9f90546d6ab9355cbaa1cf6e1">More...</a><br /></td></tr>
<tr class="separator:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga9e43115d9ee4ac406c14161fc6ccb327">save_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#ga9e43115d9ee4ac406c14161fc6ccb327">More...</a><br /></td></tr>
<tr class="separator:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga0f53ee5421c91ca041a1b8d4603238b5">save_transposed_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#ga0f53ee5421c91ca041a1b8d4603238b5">More...</a><br /></td></tr>
<tr class="separator:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga979efe39128de97176da5460cab7dfa7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga979efe39128de97176da5460cab7dfa7">save_group_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga979efe39128de97176da5460cab7dfa7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="group__savers.html#ga979efe39128de97176da5460cab7dfa7">More...</a><br /></td></tr>
<tr class="separator:ga979efe39128de97176da5460cab7dfa7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga6b87b484064fd373aec0bfa4ce8dbef1">save_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga6b87b484064fd373aec0bfa4ce8dbef1">More...</a><br /></td></tr>
<tr class="separator:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6">save_group_full_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6">More...</a><br /></td></tr>
<tr class="separator:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga40dbed7f680c5c1181a8ed27e616417b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga40dbed7f680c5c1181a8ed27e616417b">save_full_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga40dbed7f680c5c1181a8ed27e616417b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga40dbed7f680c5c1181a8ed27e616417b">More...</a><br /></td></tr>
<tr class="separator:ga40dbed7f680c5c1181a8ed27e616417b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabf5bc0b61b008aeda217ea9a666ad9cb">save_AM_pedigree</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_sim_data.html">SimData</a> *parents)</td></tr>
<tr class="memdesc:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file.  <a href="group__savers.html#gabf5bc0b61b008aeda217ea9a666ad9cb">More...</a><br /></td></tr>
<tr class="separator:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabc1307a944deb18e0aef315783ab123e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabc1307a944deb18e0aef315783ab123e">save_parents_of</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, unsigned int p1, unsigned int p2)</td></tr>
<tr class="memdesc:gabc1307a944deb18e0aef315783ab123e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="group__savers.html#gabc1307a944deb18e0aef315783ab123e">More...</a><br /></td></tr>
<tr class="separator:gabc1307a944deb18e0aef315783ab123e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0da46978e8e546af5eb3b2635902a82b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b">save_group_bvs</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga0da46978e8e546af5eb3b2635902a82b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in a group to a file.  <a href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b">More...</a><br /></td></tr>
<tr class="separator:ga0da46978e8e546af5eb3b2635902a82b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacecc6d6f4179a14a12132e37b0176877"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gacecc6d6f4179a14a12132e37b0176877">save_manual_bvs</a> (FILE *f, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *e, unsigned int *ids, char **names)</td></tr>
<tr class="memdesc:gacecc6d6f4179a14a12132e37b0176877"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>.  <a href="group__savers.html#gacecc6d6f4179a14a12132e37b0176877">More...</a><br /></td></tr>
<tr class="separator:gacecc6d6f4179a14a12132e37b0176877"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6024d629560199d7c97d86bda15f55fc"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc">save_bvs</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga6024d629560199d7c97d86bda15f55fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc">More...</a><br /></td></tr>
<tr class="separator:ga6024d629560199d7c97d86bda15f55fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga490906a5bfdb22b90aba5cdf6282fd03">save_count_matrix</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele)</td></tr>
<tr class="memdesc:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga490906a5bfdb22b90aba5cdf6282fd03">More...</a><br /></td></tr>
<tr class="separator:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab732ec2d648a6dc01444ef74f780fc70"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gab732ec2d648a6dc01444ef74f780fc70">save_count_matrix_of_group</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele, int group)</td></tr>
<tr class="memdesc:gab732ec2d648a6dc01444ef74f780fc70"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="group__savers.html#gab732ec2d648a6dc01444ef74f780fc70">More...</a><br /></td></tr>
<tr class="separator:gab732ec2d648a6dc01444ef74f780fc70"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>For saving persistent simulation results. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga42a237afba9dc8f3a7339cb97d929cb6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga42a237afba9dc8f3a7339cb97d929cb6">&#9670;&nbsp;</a></span>save_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file. </p>
<p>Uses the following format: </p><pre class="fragment">    [marker name]   [marker name]
</pre><p>[id]OR[name] [allele pairs for each marker]</p>
<p>[id]OR[name] [allele pairs for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. if this is null, the header row will be empty. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga42a237afba9dc8f3a7339cb97d929cb6_icgraph.png" border="0" usemap="#group__savers_ga42a237afba9dc8f3a7339cb97d929cb6_icgraph" alt=""/></div>
<map name="group__savers_ga42a237afba9dc8f3a7339cb97d929cb6_icgraph" id="group__savers_ga42a237afba9dc8f3a7339cb97d929cb6_icgraph">
<area shape="rect" title="Prints all the genotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="441,143,569,170"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="gabf5bc0b61b008aeda217ea9a666ad9cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabf5bc0b61b008aeda217ea9a666ad9cb">&#9670;&nbsp;</a></span>save_AM_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_AM_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>parents</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p>[id] [name]=([parent pedigree])</p>
<p>...</p>
<p>Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p>The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p>Note this does not follow through the linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> containing the genotypes to print </td></tr>
    <tr><td class="paramname">parents</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given genotypes. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_cgraph.png" border="0" usemap="#group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_cgraph" alt=""/></div>
<map name="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_cgraph" id="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="5,31,136,57"/>
<area shape="rect" href="group__savers.html#gabc1307a944deb18e0aef315783ab123e" title="Recursively save the parents of a particular id to a file." alt="" coords="184,31,299,57"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="352,5,465,32"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="347,56,471,83"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="519,31,693,57"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_icgraph.png" border="0" usemap="#group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_icgraph" alt=""/></div>
<map name="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_icgraph" id="group__savers_gabf5bc0b61b008aeda217ea9a666ad9cb_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="441,143,572,170"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ga6024d629560199d7c97d86bda15f55fc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga6024d629560199d7c97d86bda15f55fc">&#9670;&nbsp;</a></span>save_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name] [BV]</p>
<p>[id] [name] [BV]</p>
<p>...</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga6024d629560199d7c97d86bda15f55fc_cgraph.png" border="0" usemap="#group__savers_ga6024d629560199d7c97d86bda15f55fc_cgraph" alt=""/></div>
<map name="group__savers_ga6024d629560199d7c97d86bda15f55fc_cgraph" id="group__savers_ga6024d629560199d7c97d86bda15f55fc_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in the SimData to a file." alt="" coords="5,263,80,290"/>
<area shape="rect" href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="128,209,228,235"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="308,318,415,345"/>
<area shape="rect" href="group__maths.html#gad2c9cda1aab9679e0bfe3c88959213c6" title="Multiply two sets of a DecimalMatrix and vector, and add both products to the first column of a provi..." alt="" coords="282,5,441,47"/>
<area shape="rect" href="group__maths.html#gabb8addf721442aecdd7dca3dfd955c3c" title="Multiply a DecimalMatrix to a vector, and add that product to the first column of a provided DecimalM..." alt="" coords="276,71,447,112"/>
<area shape="rect" href="group__calculators.html#ga32b287af2e2955edd565eedf3043f02c" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="283,136,439,177"/>
<area shape="rect" href="group__calculators.html#ga3e812683e26615b797eb4cd98b694d52" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="286,201,437,243"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="284,267,439,294"/>
</map>
</div>

</div>
</div>
<a id="ga490906a5bfdb22b90aba5cdf6282fd03"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga490906a5bfdb22b90aba5cdf6282fd03">&#9670;&nbsp;</a></span>save_count_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p>[marker name] [allele count for each marker]</p>
<p>[marker name] [allele count for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga490906a5bfdb22b90aba5cdf6282fd03_cgraph.png" border="0" usemap="#group__savers_ga490906a5bfdb22b90aba5cdf6282fd03_cgraph" alt=""/></div>
<map name="group__savers_ga490906a5bfdb22b90aba5cdf6282fd03_cgraph" id="group__savers_ga490906a5bfdb22b90aba5cdf6282fd03_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="5,42,136,69"/>
<area shape="rect" href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in t..." alt="" coords="184,5,320,47"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="199,71,305,98"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="368,13,523,39"/>
</map>
</div>

</div>
</div>
<a id="gab732ec2d648a6dc01444ef74f780fc70"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab732ec2d648a6dc01444ef74f780fc70">&#9670;&nbsp;</a></span>save_count_matrix_of_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix_of_group </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in a group to a file. </p>
<p>The following tab-separated format is used: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p>[marker name] [allele count for each marker]</p>
<p>[marker name] [allele count for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the allele count of. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gab732ec2d648a6dc01444ef74f780fc70_cgraph.png" border="0" usemap="#group__savers_gab732ec2d648a6dc01444ef74f780fc70_cgraph" alt=""/></div>
<map name="group__savers_gab732ec2d648a6dc01444ef74f780fc70_cgraph" id="group__savers_gab732ec2d648a6dc01444ef74f780fc70_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="5,98,152,139"/>
<area shape="rect" href="group__calculators.html#gad86ea5492e6c1cf918d6449a9eb530de" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="201,39,353,80"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="576,46,687,73"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="224,156,331,182"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="200,206,355,233"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="403,122,528,149"/>
<area shape="rect" href="group__getters.html#gac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="405,72,526,98"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="589,97,674,124"/>
</map>
</div>

</div>
</div>
<a id="ga40dbed7f680c5c1181a8ed27e616417b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga40dbed7f680c5c1181a8ed27e616417b">&#9670;&nbsp;</a></span>save_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p>[id] [name]=([parent pedigree])</p>
<p>...</p>
<p>Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p>The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing all genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga40dbed7f680c5c1181a8ed27e616417b_cgraph.png" border="0" usemap="#group__savers_ga40dbed7f680c5c1181a8ed27e616417b_cgraph" alt=""/></div>
<map name="group__savers_ga40dbed7f680c5c1181a8ed27e616417b_cgraph" id="group__savers_ga40dbed7f680c5c1181a8ed27e616417b_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in the SimData to a file." alt="" coords="5,31,132,57"/>
<area shape="rect" href="group__savers.html#gabc1307a944deb18e0aef315783ab123e" title="Recursively save the parents of a particular id to a file." alt="" coords="180,31,295,57"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="348,5,461,32"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="343,56,467,83"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="515,31,689,57"/>
</map>
</div>

</div>
</div>
<a id="ga9e43115d9ee4ac406c14161fc6ccb327"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga9e43115d9ee4ac406c14161fc6ccb327">&#9670;&nbsp;</a></span>save_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">    [marker name]   [marker name]
</pre><p>[id]OR[name] [allele pairs for each marker]</p>
<p>[id]OR[name] [allele pairs for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga9e43115d9ee4ac406c14161fc6ccb327_cgraph.png" border="0" usemap="#group__savers_ga9e43115d9ee4ac406c14161fc6ccb327_cgraph" alt=""/></div>
<map name="group__savers_ga9e43115d9ee4ac406c14161fc6ccb327_cgraph" id="group__savers_ga9e43115d9ee4ac406c14161fc6ccb327_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,50,136,77"/>
<area shape="rect" href="group__getters.html#gac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="186,25,307,51"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="357,25,468,51"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="195,75,298,102"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="184,126,309,153"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="370,75,455,102"/>
</map>
</div>

</div>
</div>
<a id="ga0da46978e8e546af5eb3b2635902a82b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0da46978e8e546af5eb3b2635902a82b">&#9670;&nbsp;</a></span>save_group_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the breeding value of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[group member id] [group member name] [BV]</p>
<p>[group member id] [group member name] [BV]</p>
<p>...</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the breeding values of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga0da46978e8e546af5eb3b2635902a82b_cgraph.png" border="0" usemap="#group__savers_ga0da46978e8e546af5eb3b2635902a82b_cgraph" alt=""/></div>
<map name="group__savers_ga0da46978e8e546af5eb3b2635902a82b_cgraph" id="group__savers_ga0da46978e8e546af5eb3b2635902a82b_cgraph">
<area shape="rect" title="Print the breeding value of each genotype in a group to a file." alt="" coords="5,406,119,432"/>
<area shape="rect" href="group__calculators.html#ga7f80905718ac87e9da1c30a2cec77ea6" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="167,252,305,279"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="765,146,876,172"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="395,406,502,432"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="603,298,706,324"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="592,247,717,274"/>
<area shape="rect" href="group__maths.html#gad2c9cda1aab9679e0bfe3c88959213c6" title="Multiply two sets of a DecimalMatrix and vector, and add both products to the first column of a provi..." alt="" coords="369,274,528,316"/>
<area shape="rect" href="group__maths.html#gabb8addf721442aecdd7dca3dfd955c3c" title="Multiply a DecimalMatrix to a vector, and add that product to the first column of a provided DecimalM..." alt="" coords="363,340,534,381"/>
<area shape="rect" href="group__calculators.html#gad86ea5492e6c1cf918d6449a9eb530de" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="373,92,525,133"/>
<area shape="rect" href="group__calculators.html#ga0b2b686b9dde986b549f15d72fa4c919" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="353,157,544,198"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="371,223,526,250"/>
<area shape="rect" href="group__getters.html#gac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="594,146,715,172"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="778,247,863,274"/>
</map>
</div>

</div>
</div>
<a id="ga75e854cbc72b31a5e2eb03c63c33ecc6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga75e854cbc72b31a5e2eb03c63c33ecc6">&#9670;&nbsp;</a></span>save_group_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p>[id] [name]=([parent pedigree])</p>
<p>...</p>
<p>Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p>The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the pedigree of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_cgraph.png" border="0" usemap="#group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_cgraph" alt=""/></div>
<map name="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_cgraph" id="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="5,76,172,103"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="231,76,334,103"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="400,26,511,52"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="220,26,345,52"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="393,151,517,178"/>
<area shape="rect" href="group__savers.html#gabc1307a944deb18e0aef315783ab123e" title="Recursively save the parents of a particular id to a file." alt="" coords="225,202,340,228"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="413,76,498,103"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="565,176,740,203"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="399,202,512,228"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_icgraph.png" border="0" usemap="#group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_icgraph" alt=""/></div>
<map name="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_icgraph" id="group__savers_ga75e854cbc72b31a5e2eb03c63c33ecc6_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="196,5,363,32"/>
<area shape="rect" href="group__getters.html#ga86e56272267b8d4980e5062d6c0fef7c" title="Gets the full pedigree string (as per save_group_full_pedigree() ) of each member of the group." alt="" coords="5,5,148,32"/>
</map>
</div>

</div>
</div>
<a id="ga979efe39128de97176da5460cab7dfa7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga979efe39128de97176da5460cab7dfa7">&#9670;&nbsp;</a></span>save_group_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[group member name] [parent 1 name] [parent 2 name]</p>
<p>[group member name] [parent 1 name] [parent 2 name]</p>
<p>...</p>
<p>The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the immediate parents of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga979efe39128de97176da5460cab7dfa7_cgraph.png" border="0" usemap="#group__savers_ga979efe39128de97176da5460cab7dfa7_cgraph" alt=""/></div>
<map name="group__savers_ga979efe39128de97176da5460cab7dfa7_cgraph" id="group__savers_ga979efe39128de97176da5460cab7dfa7_cgraph">
<area shape="rect" title="Print the parents of each genotype in a group to a file." alt="" coords="5,71,152,112"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="211,78,314,105"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="405,28,516,54"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="200,28,325,54"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="206,180,319,206"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="201,129,325,156"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="418,78,503,105"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="373,154,548,181"/>
</map>
</div>

</div>
</div>
<a id="gacecc6d6f4179a14a12132e37b0176877"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gacecc6d6f4179a14a12132e37b0176877">&#9670;&nbsp;</a></span>save_manual_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_manual_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>names</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name] [BV]</p>
<p>[id] [name] [BV]</p>
<p>...</p>
<p>The function assumes the array of ids, array of names, and columns of the <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> are ordered the same, so a single index produces the corresponding value from each. It is used by as-you-go savers within crossing functions. For general use consider <code>save_bvs</code> or <code>save_group_bvs</code> instead.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the BVs in the first row. </td></tr>
    <tr><td class="paramname">ids</td><td>array of ids to print alongside the BVs. </td></tr>
    <tr><td class="paramname">names</td><td>array of names to print alongside the BVs. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gacecc6d6f4179a14a12132e37b0176877_icgraph.png" border="0" usemap="#group__savers_gacecc6d6f4179a14a12132e37b0176877_icgraph" alt=""/></div>
<map name="group__savers_gacecc6d6f4179a14a12132e37b0176877_icgraph" id="group__savers_gacecc6d6f4179a14a12132e37b0176877_icgraph">
<area shape="rect" title="Print the provided breeding values of each provided name and id to a file, with the same format as a ..." alt="" coords="441,143,564,170"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ga2d795d5594b914eac4e5f6fcee3afbda"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2d795d5594b914eac4e5f6fcee3afbda">&#9670;&nbsp;</a></span>save_marker_blocks()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_marker_blocks </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the markers contained in a set of blocks to a file. </p>
<p>Column separators are tabs.</p>
<p>The printing format is:</p>
<p>Chrom Pos Name Class Markers</p>
<p>0 0 b0 b m1;m2;m3;m4;</p>
<p>0 0 b0 b m7;m9;</p>
<p>...</p>
<p>where m1, m2, m3, m4 are the names of the markers in the first block and m7 and m9 are the names of the markers in the second block.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> whose data we print </td></tr>
    <tr><td class="paramname">b</td><td><a class="el" href="struct_marker_blocks.html" title="A struct used to store a set of blocks of markers.">MarkerBlocks</a> struct containing the groupings of markers to print. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga6b87b484064fd373aec0bfa4ce8dbef1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga6b87b484064fd373aec0bfa4ce8dbef1">&#9670;&nbsp;</a></span>save_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[genotype name] [parent 1 name] [parent 2 name]</p>
<p>[genotype name] [parent 1 name] [parent 2 name]</p>
<p>...</p>
<p>The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes and their pedigrees </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga6b87b484064fd373aec0bfa4ce8dbef1_cgraph.png" border="0" usemap="#group__savers_ga6b87b484064fd373aec0bfa4ce8dbef1_cgraph" alt=""/></div>
<map name="group__savers_ga6b87b484064fd373aec0bfa4ce8dbef1_cgraph" id="group__savers_ga6b87b484064fd373aec0bfa4ce8dbef1_cgraph">
<area shape="rect" title="Print the parents of each genotype in the SimData to a file." alt="" coords="5,31,169,57"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="223,5,336,32"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="217,56,341,83"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="389,31,564,57"/>
</map>
</div>

</div>
</div>
<a id="gabc1307a944deb18e0aef315783ab123e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabc1307a944deb18e0aef315783ab123e">&#9670;&nbsp;</a></span>save_parents_of()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_parents_of </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Recursively save the parents of a particular id to a file. </p>
<p>It saves using the following format:</p>
<ul>
<li>no characters are saved if the parents of the id are unknown/0</li>
<li>if the id has one parent, repeated twice (it was produced by selfing), print "=(parentname)", where parentname is the name of the parent or its id if it does not have one, followed by whatever is printed by a call to this function on the parent's id.</li>
<li>if the id has two separate parents, print "=(parent1name,parent2name)", where parent1name and parent2name are the name2 of the two parents or their ids if they does not have names, each name immediately followed by whatever is printed by a call to this function on the corresponding parent's id.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given id. </td></tr>
    <tr><td class="paramname">p1</td><td>the session-unique id of the first parent to be saved. </td></tr>
    <tr><td class="paramname">p2</td><td>the session-unique id of the second parent to be saved. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gabc1307a944deb18e0aef315783ab123e_cgraph.png" border="0" usemap="#group__savers_gabc1307a944deb18e0aef315783ab123e_cgraph" alt=""/></div>
<map name="group__savers_gabc1307a944deb18e0aef315783ab123e_cgraph" id="group__savers_gabc1307a944deb18e0aef315783ab123e_cgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="5,31,120,57"/>
<area shape="rect" href="group__getters.html#ga6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="173,5,287,32"/>
<area shape="rect" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="168,56,292,83"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="340,31,515,57"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_gabc1307a944deb18e0aef315783ab123e_icgraph.png" border="0" usemap="#group__savers_gabc1307a944deb18e0aef315783ab123e_icgraph" alt=""/></div>
<map name="group__savers_gabc1307a944deb18e0aef315783ab123e_icgraph" id="group__savers_gabc1307a944deb18e0aef315783ab123e_icgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="656,209,771,235"/>
<area shape="rect" href="group__savers.html#gabf5bc0b61b008aeda217ea9a666ad9cb" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="459,158,590,185"/>
<area shape="rect" href="group__savers.html#ga40dbed7f680c5c1181a8ed27e616417b" title="Print the full known pedigree of each genotype in the SimData to a file." alt="" coords="461,209,588,235"/>
<area shape="rect" href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="441,259,608,286"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,19,390,46"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,209,352,235"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="group__getters.html#ga86e56272267b8d4980e5062d6c0fef7c" title="Gets the full pedigree string (as per save_group_full_pedigree() ) of each member of the group." alt="" coords="233,259,376,286"/>
</map>
</div>

</div>
</div>
<a id="ga4d63cd249222ab25b3112106813f29b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga4d63cd249222ab25b3112106813f29b2">&#9670;&nbsp;</a></span>save_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_simdata </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>m</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>Column separators are tabs.</p>
<p>The printing format is:</p>
<p>name chr pos [allele to which these effects apply, 0+ columns]</p>
<p>[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p>[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p>...</p>
<p>If m-&gt;effects is NULL, m-&gt;ref_alleles is NULL, or m-&gt;genetic_map is NULL, then the relevant columns are omitted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> whose data we print </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="gaadf131f9f90546d6ab9355cbaa1cf6e1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaadf131f9f90546d6ab9355cbaa1cf6e1">&#9670;&nbsp;</a></span>save_transposed_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the gene data saved in the linked list starting with <code>m</code> to the file. </p>
<p>Uses the following format: </p><pre class="fragment">    [id]OR[name]    [id]OR[name] ...
</pre><p>[marker name] [allele pairs for each marker]</p>
<p>[marker name] [allele pairs for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga0f53ee5421c91ca041a1b8d4603238b5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0f53ee5421c91ca041a1b8d4603238b5">&#9670;&nbsp;</a></span>save_transposed_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">[id]OR[name]    [id]OR[name] ...
</pre><p>[marker name] [allele pairs for each marker]</p>
<p>[marker name] [allele pairs for each marker]</p>
<p>...</p>
<p>ID will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__savers_ga0f53ee5421c91ca041a1b8d4603238b5_cgraph.png" border="0" usemap="#group__savers_ga0f53ee5421c91ca041a1b8d4603238b5_cgraph" alt=""/></div>
<map name="group__savers_ga0f53ee5421c91ca041a1b8d4603238b5_cgraph" id="group__savers_ga0f53ee5421c91ca041a1b8d4603238b5_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,42,164,84"/>
<area shape="rect" href="group__getters.html#gac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="214,24,335,51"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="385,24,496,51"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="223,75,326,102"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="212,126,337,152"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="398,75,483,102"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
