<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Breeding Value and Allele Count Calculators</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">genomicSimulationC
   &#160;<span id="projectnumber">0.1-4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Breeding Value and Allele Count Calculators</div>  </div>
</div><!--header-->
<div class="contents">

<p>For calculations related to the loaded allele effects and internal additive breeding value model.  
<a href="#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:gae34db60faf62b898d0572ea6c039b96a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">split_by_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, int top_n, int lowIsBest)</td></tr>
<tr class="memdesc:gae34db60faf62b898d0572ea6c039b96a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group.  <a href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">More...</a><br /></td></tr>
<tr class="separator:gae34db60faf62b898d0572ea6c039b96a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga955d1b1f461fcc6d5e8c33623164ad6e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga955d1b1f461fcc6d5e8c33623164ad6e">calculate_group_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga955d1b1f461fcc6d5e8c33623164ad6e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="group__calculators.html#ga955d1b1f461fcc6d5e8c33623164ad6e">More...</a><br /></td></tr>
<tr class="separator:ga955d1b1f461fcc6d5e8c33623164ad6e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5">calculate_bvs</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *e)</td></tr>
<tr class="memdesc:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5">More...</a><br /></td></tr>
<tr class="separator:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae7797aeac935227749fc6eae09a02a96"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae7797aeac935227749fc6eae09a02a96">calculate_count_matrix_of_allele_for_ids</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, unsigned int *for_ids, unsigned int n_ids, char allele)</td></tr>
<tr class="memdesc:gae7797aeac935227749fc6eae09a02a96"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype is a set of given genotype.  <a href="group__calculators.html#gae7797aeac935227749fc6eae09a02a96">More...</a><br /></td></tr>
<tr class="separator:gae7797aeac935227749fc6eae09a02a96"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga42f2398ccb53ffafce9b4a17e3222636"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636">calculate_full_count_matrix_of_allele</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char allele)</td></tr>
<tr class="memdesc:ga42f2398ccb53ffafce9b4a17e3222636"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.  <a href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636">More...</a><br /></td></tr>
<tr class="separator:ga42f2398ccb53ffafce9b4a17e3222636"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2f61debfc95036e86953f6111236026f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga2f61debfc95036e86953f6111236026f">create_n_blocks_by_chr</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n)</td></tr>
<tr class="memdesc:ga2f61debfc95036e86953f6111236026f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct.  <a href="group__calculators.html#ga2f61debfc95036e86953f6111236026f">More...</a><br /></td></tr>
<tr class="separator:ga2f61debfc95036e86953f6111236026f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga463082fc380a6815f07bd06401927bc0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga463082fc380a6815f07bd06401927bc0">read_block_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *block_file)</td></tr>
<tr class="memdesc:ga463082fc380a6815f07bd06401927bc0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks.  <a href="group__calculators.html#ga463082fc380a6815f07bd06401927bc0">More...</a><br /></td></tr>
<tr class="separator:ga463082fc380a6815f07bd06401927bc0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga8fdd3011c2ce5228c8c8d6486533369b">calculate_group_local_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const char *output_file, int group)</td></tr>
<tr class="memdesc:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga8fdd3011c2ce5228c8c8d6486533369b">More...</a><br /></td></tr>
<tr class="separator:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7c6e2c2f4447accd4a30031256c1610a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga7c6e2c2f4447accd4a30031256c1610a">calculate_local_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const char *output_file)</td></tr>
<tr class="memdesc:ga7c6e2c2f4447accd4a30031256c1610a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga7c6e2c2f4447accd4a30031256c1610a">More...</a><br /></td></tr>
<tr class="separator:ga7c6e2c2f4447accd4a30031256c1610a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadf9a424cc9474272059ff328bfadf755"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gadf9a424cc9474272059ff328bfadf755">calculate_optimal_alleles</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gadf9a424cc9474272059ff328bfadf755"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__calculators.html#gadf9a424cc9474272059ff328bfadf755">More...</a><br /></td></tr>
<tr class="separator:gadf9a424cc9474272059ff328bfadf755"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga67d98ae97e5352f32380c8b89b1e4047"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga67d98ae97e5352f32380c8b89b1e4047">calculate_optimum_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga67d98ae97e5352f32380c8b89b1e4047"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values.  <a href="group__calculators.html#ga67d98ae97e5352f32380c8b89b1e4047">More...</a><br /></td></tr>
<tr class="separator:ga67d98ae97e5352f32380c8b89b1e4047"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacef164c7781f3aa15928b482c3845418"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gacef164c7781f3aa15928b482c3845418">calculate_minimum_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gacef164c7781f3aa15928b482c3845418"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values.  <a href="group__calculators.html#gacef164c7781f3aa15928b482c3845418">More...</a><br /></td></tr>
<tr class="separator:gacef164c7781f3aa15928b482c3845418"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>For calculations related to the loaded allele effects and internal additive breeding value model. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="gab5b177cf40a7e28276c460d60a3ecbb5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab5b177cf40a7e28276c460d60a3ecbb5">&#9670;&nbsp;</a></span>calculate_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct. </p>
<p>The breeding value is calculated for each genotype by taking the sum of (number of copies of this allele at this marker times this allele's effect at this marker) for each marker for each different allele. To do this, the vector containing each allele's effect values is multiplied by a matrix containing the counts of that allele at each marker.</p>
<p>The function exits with error code 1 if no marker effect file is loaded.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object to which the genotypes belong. </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> that effect values have been loaded into. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the score for each individual in the group. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_cgraph.png" border="0" usemap="#group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_cgraph" alt=""/></div>
<map name="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_cgraph" id="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_cgraph">
<area shape="rect" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="5,144,105,171"/>
<area shape="rect" href="group__maths.html#gac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="169,5,276,32"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="340,85,495,112"/>
<area shape="rect" href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="155,107,291,149"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="169,173,276,200"/>
<area shape="rect" href="group__maths.html#ga1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="157,224,288,251"/>
<area shape="rect" href="group__maths.html#gabe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="153,275,292,301"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_icgraph.png" border="0" usemap="#group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_icgraph" alt=""/></div>
<map name="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_icgraph" id="group__calculators_gab5b177cf40a7e28276c460d60a3ecbb5_icgraph">
<area shape="rect" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="441,169,541,195"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc" title="Print the breeding value of each genotype in the SimData to a file." alt="" coords="267,219,342,246"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,270,352,297"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="gae7797aeac935227749fc6eae09a02a96"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gae7797aeac935227749fc6eae09a02a96">&#9670;&nbsp;</a></span>calculate_count_matrix_of_allele_for_ids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_count_matrix_of_allele_for_ids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>for_ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears for each genotype is a set of given genotype. </p>
<p>Returns the result as a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>. Useful for multiplying to effect matrix to calculate breeding values.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the start of a linked list that contains the alleles of each genotype in <code>for_ids</code>. </td></tr>
    <tr><td class="paramname">for_ids</td><td>poinder to an array of ids of different genotypes. </td></tr>
    <tr><td class="paramname">n_ids</td><td>number of genotypes to calculate the counts of/length of <code>for_ids</code> </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> countaining the number of <code>allele</code> occurences at each row/marker for each column/genotype in <code>for_ids</code>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gae7797aeac935227749fc6eae09a02a96_cgraph.png" border="0" usemap="#group__calculators_gae7797aeac935227749fc6eae09a02a96_cgraph" alt=""/></div>
<map name="group__calculators_gae7797aeac935227749fc6eae09a02a96_cgraph" id="group__calculators_gae7797aeac935227749fc6eae09a02a96_cgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="5,23,161,65"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="209,5,364,32"/>
<area shape="rect" href="group__getters.html#gaf49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="229,56,345,83"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="412,56,587,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gae7797aeac935227749fc6eae09a02a96_icgraph.png" border="0" usemap="#group__calculators_gae7797aeac935227749fc6eae09a02a96_icgraph" alt=""/></div>
<map name="group__calculators_gae7797aeac935227749fc6eae09a02a96_icgraph" id="group__calculators_gae7797aeac935227749fc6eae09a02a96_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="563,78,719,119"/>
<area shape="rect" href="group__calculators.html#ga955d1b1f461fcc6d5e8c33623164ad6e" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="372,56,511,83"/>
<area shape="rect" href="group__savers.html#gab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="368,107,515,149"/>
<area shape="rect" href="group__getters.html#gaacbc190e8ea144de14b586e5fe3ae865" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="211,5,316,32"/>
<area shape="rect" href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b" title="Print the breeding value of each genotype in a group to a file." alt="" coords="207,56,320,83"/>
<area shape="rect" href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="220,107,307,133"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,99,159,141"/>
</map>
</div>

</div>
</div>
<a id="ga42f2398ccb53ffafce9b4a17e3222636"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga42f2398ccb53ffafce9b4a17e3222636">&#9670;&nbsp;</a></span>calculate_full_count_matrix_of_allele()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_full_count_matrix_of_allele </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>. </p>
<p>Returns the result as a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>. Useful for multiplying to effect matrix to calculate breeding values.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that contains the genotypes to count alleles. </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> countaining the number of <code>allele</code> occurences at each row/marker for each column/genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_cgraph.png" border="0" usemap="#group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_cgraph" alt=""/></div>
<map name="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_cgraph" id="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_cgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="5,5,141,47"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="189,13,344,39"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_icgraph.png" border="0" usemap="#group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_icgraph" alt=""/></div>
<map name="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_icgraph" id="group__calculators_ga42f2398ccb53ffafce9b4a17e3222636_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="620,176,756,217"/>
<area shape="rect" href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="457,158,557,185"/>
<area shape="rect" href="group__savers.html#ga490906a5bfdb22b90aba5cdf6282fd03" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="441,209,572,235"/>
<area shape="rect" href="group__crossers.html#ga3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,34,390,61"/>
<area shape="rect" href="group__crossers.html#gabb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc" title="Print the breeding value of each genotype in the SimData to a file." alt="" coords="267,209,342,235"/>
<area shape="rect" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,259,352,286"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ga955d1b1f461fcc6d5e8c33623164ad6e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga955d1b1f461fcc6d5e8c33623164ad6e">&#9670;&nbsp;</a></span>calculate_group_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_group_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct. </p>
<p>The breeding value is calculated for each genotype by taking the sum of (number of copies of this allele at this marker times this allele's effect at this marker) for each marker for each different allele. To do this, the vector containing each allele's effect values is multiplied by a matrix containing the counts of that allele at each marker.</p>
<p>The function exits with error code 1 if no marker effect file is loaded.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">group</td><td>calculate breeding values for each genotype in the group with this group number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the score for each individual in the group. </dd></dl>
<p>counts = sum; // use counts as a temp</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_cgraph.png" border="0" usemap="#group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_cgraph" alt=""/></div>
<map name="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_cgraph" id="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_cgraph">
<area shape="rect" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="5,187,144,213"/>
<area shape="rect" href="group__maths.html#ga2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="215,5,325,32"/>
<area shape="rect" href="group__calculators.html#gae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="192,209,348,250"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="396,107,551,133"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="217,157,323,184"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="219,275,321,301"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="418,325,529,352"/>
<area shape="rect" href="group__maths.html#ga1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="205,107,335,133"/>
<area shape="rect" href="group__maths.html#gabe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="201,376,339,403"/>
<area shape="rect" href="group__getters.html#gaf49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="415,216,531,243"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="599,216,773,243"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="431,275,516,301"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_icgraph.png" border="0" usemap="#group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_icgraph" alt=""/></div>
<map name="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_icgraph" id="group__calculators_ga955d1b1f461fcc6d5e8c33623164ad6e_icgraph">
<area shape="rect" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="368,56,507,83"/>
<area shape="rect" href="group__getters.html#gaacbc190e8ea144de14b586e5fe3ae865" title="Gets the breeding values/breeding values/fitnesses of each member of the group." alt="" coords="211,5,316,32"/>
<area shape="rect" href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b" title="Print the breeding value of each genotype in a group to a file." alt="" coords="207,56,320,83"/>
<area shape="rect" href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="220,107,307,133"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,99,159,141"/>
</map>
</div>

</div>
</div>
<a id="ga8fdd3011c2ce5228c8c8d6486533369b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8fdd3011c2ce5228c8c8d6486533369b">&#9670;&nbsp;</a></span>calculate_group_local_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void calculate_group_local_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file. </p>
<p>This gives block effects/local BVs for each haplotype of each individual in the group.</p>
<p>The output file will have format:</p>
<p>[genotype name]_1 [effect for first block, first allele] [effect for second block, second allele] ...</p>
<p>[genotype name]_2 [effect for first block, second allele] [effect for second block, second allele] ...</p>
<p>[genotype 2 name]_1 ...</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">b</td><td>struct containing the blocks to use </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename of the file to which output block effects/local BVs will be saved. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to split the top individuals. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga8fdd3011c2ce5228c8c8d6486533369b_cgraph.png" border="0" usemap="#group__calculators_ga8fdd3011c2ce5228c8c8d6486533369b_cgraph" alt=""/></div>
<map name="group__calculators_ga8fdd3011c2ce5228c8c8d6486533369b_cgraph" id="group__calculators_ga8fdd3011c2ce5228c8c8d6486533369b_cgraph">
<area shape="rect" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness..." alt="" coords="5,22,179,49"/>
<area shape="rect" href="group__getters.html#gac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="229,73,350,99"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="400,22,511,49"/>
<area shape="rect" href="group__getters.html#ga3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="227,22,352,49"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="413,73,498,99"/>
</map>
</div>

</div>
</div>
<a id="ga7c6e2c2f4447accd4a30031256c1610a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7c6e2c2f4447accd4a30031256c1610a">&#9670;&nbsp;</a></span>calculate_local_bvs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void calculate_local_bvs </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file. </p>
<p>This gives block effects for each haplotype of each individual currently saved to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.</p>
<p>The output file will have format:</p>
<p>[genotype name]_1 [effect for first block, first allele] [effect for second block, second allele] ...</p>
<p>[genotype name]_2 [effect for first block, second allele] [effect for second block, second allele] ...</p>
<p>[genotype 2 name]_1 ...</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">b</td><td>struct containing the blocks to use </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename of the file to which output block effects/local BV will be saved. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="gacef164c7781f3aa15928b482c3845418"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gacef164c7781f3aa15928b482c3845418">&#9670;&nbsp;</a></span>calculate_minimum_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double calculate_minimum_bv </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values. </p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must be initialised with marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing markers and marker effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the fitness metric/breeding value of the worst genotype. </dd></dl>

</div>
</div>
<a id="gadf9a424cc9474272059ff328bfadf755"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gadf9a424cc9474272059ff328bfadf755">&#9670;&nbsp;</a></span>calculate_optimal_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* calculate_optimal_alleles </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </p>
<p>Returns this as a null-terminated heap array of characters of length d-&gt;n_markers + one null byte.</p>
<p>The return value should be freed when usage is finished!</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must be initialised with marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing markers and marker effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap array filled with a null-terminated string containing the highest scoring allele at each marker. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gadf9a424cc9474272059ff328bfadf755_cgraph.png" border="0" usemap="#group__calculators_gadf9a424cc9474272059ff328bfadf755_cgraph" alt=""/></div>
<map name="group__calculators_gadf9a424cc9474272059ff328bfadf755_cgraph" id="group__calculators_gadf9a424cc9474272059ff328bfadf755_cgraph">
<area shape="rect" title="Takes a look at the currently&#45;loaded effect values and creates a string containing the allele with th..." alt="" coords="5,5,172,32"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="220,5,305,32"/>
</map>
</div>

</div>
</div>
<a id="ga67d98ae97e5352f32380c8b89b1e4047"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga67d98ae97e5352f32380c8b89b1e4047">&#9670;&nbsp;</a></span>calculate_optimum_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double calculate_optimum_bv </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values. </p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must be initialised with marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing markers and marker effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the fitness metric/breeding value of the best/ideal genotype. </dd></dl>

</div>
</div>
<a id="ga2f61debfc95036e86953f6111236026f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2f61debfc95036e86953f6111236026f">&#9670;&nbsp;</a></span>create_n_blocks_by_chr()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> create_n_blocks_by_chr </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct. </p>
<p>Chromosomes where there are no markers tracked are excluded/no blocks are created in those chromosomes.</p>
<p>Chromosomes where only one marker is tracked put the marker in the first block, and have all other blocks empty.</p>
<p>Empty blocks have blocks.num_markers_in_block[index] equal to 0 and contain a null pointer at blocks.markers_in_block[index].</p>
<p>Remember to call the <a class="el" href="struct_marker_blocks.html" title="A struct used to store a set of blocks of markers.">MarkerBlocks</a> destructor <a class="el" href="group__deletors.html#gaea2a41cc103e8d11bedff564f93a4759" title="Deletes a MarkerBlocks object and frees its associated memory.">delete_markerblocks()</a> on the returned struct.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">n</td><td>number of blocks into which to split each chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct containing the markers identified as belonging to each block. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga2f61debfc95036e86953f6111236026f_cgraph.png" border="0" usemap="#group__calculators_ga2f61debfc95036e86953f6111236026f_cgraph" alt=""/></div>
<map name="group__calculators_ga2f61debfc95036e86953f6111236026f_cgraph" id="group__calculators_ga2f61debfc95036e86953f6111236026f_cgraph">
<area shape="rect" title="Divide the genotype into blocks where each block contains all markers within a 1/n length section of ..." alt="" coords="5,5,168,32"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="216,5,301,32"/>
</map>
</div>

</div>
</div>
<a id="ga463082fc380a6815f07bd06401927bc0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga463082fc380a6815f07bd06401927bc0">&#9670;&nbsp;</a></span>read_block_file()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> read_block_file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>block_file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks. </p>
<p>The block file is designed after the output from a call to the R SelectionTools package's <code>st.def.hblocks</code> function. It should have the format (tab-separated):</p>
<p>Chrom Pos Name Class Markers</p>
<p>[ignored] [ignored] [ignored] [ignored] [semicolon];[separated];[list] ;[of];[marker];[names];[belonging];[to];[this];[block];</p>
<p>...</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">block_file</td><td>string containing filename of the file with blocks </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct containing the markers identified as belonging to each block according to their definitions in the file. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_ga463082fc380a6815f07bd06401927bc0_cgraph.png" border="0" usemap="#group__calculators_ga463082fc380a6815f07bd06401927bc0_cgraph" alt=""/></div>
<map name="group__calculators_ga463082fc380a6815f07bd06401927bc0_cgraph" id="group__calculators_ga463082fc380a6815f07bd06401927bc0_cgraph">
<area shape="rect" title="Given a file containing definitions of blocks of markers, process that file and return a struct conta..." alt="" coords="5,64,113,91"/>
<area shape="rect" href="group__supporters.html#ga392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="161,5,297,32"/>
<area shape="rect" href="group__supporters.html#gaed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="161,57,297,98"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="187,123,272,149"/>
</map>
</div>

</div>
</div>
<a id="gae34db60faf62b898d0572ea6c039b96a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gae34db60faf62b898d0572ea6c039b96a">&#9670;&nbsp;</a></span>split_by_bv()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int split_by_bv </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>top_n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>lowIsBest</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group. </p>
<p>The new group number is returned.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to split the top individuals. </td></tr>
    <tr><td class="paramname">top_n</td><td>The number of individuals to put in the new group. </td></tr>
    <tr><td class="paramname">lowIsBest</td><td>boolean, if TRUE the <code>top_n</code> with the lowest breeding value will be selected, if false the <code>top_n</code> with the highest breeding value are. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the newly-created split-off group </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gae34db60faf62b898d0572ea6c039b96a_cgraph.png" border="0" usemap="#group__calculators_gae34db60faf62b898d0572ea6c039b96a_cgraph" alt=""/></div>
<map name="group__calculators_gae34db60faf62b898d0572ea6c039b96a_cgraph" id="group__calculators_gae34db60faf62b898d0572ea6c039b96a_cgraph">
<area shape="rect" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="5,233,92,260"/>
<area shape="rect" href="group__supporters.html#ga311381904a1ca73ea26dfede9fcddea2" title="Comparator function for qsort." alt="" coords="140,91,335,117"/>
<area shape="rect" href="group__supporters.html#ga2aae9a568cd0cd5313808e458a9fe388" title="Comparator function for qsort." alt="" coords="167,142,307,183"/>
<area shape="rect" href="group__calculators.html#ga955d1b1f461fcc6d5e8c33623164ad6e" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix in a certain group..." alt="" coords="168,259,307,285"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="407,427,514,453"/>
<area shape="rect" href="group__getters.html#ga95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="395,5,526,32"/>
<area shape="rect" href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="179,448,296,475"/>
<area shape="rect" href="group__maths.html#ga2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="405,208,516,235"/>
<area shape="rect" href="group__calculators.html#gae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="383,361,539,402"/>
<area shape="rect" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="587,309,741,336"/>
<area shape="rect" href="group__getters.html#ga961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="409,56,512,83"/>
<area shape="rect" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="609,56,719,83"/>
<area shape="rect" href="group__maths.html#ga1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="395,259,526,285"/>
<area shape="rect" href="group__maths.html#gabe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="391,157,530,184"/>
<area shape="rect" href="group__getters.html#gaf49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="606,368,722,395"/>
<area shape="rect" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="789,368,964,395"/>
<area shape="rect" href="group__supporters.html#gad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="621,5,707,32"/>
<area shape="rect" href="group__supporters.html#gae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="390,477,531,504"/>
<area shape="rect" href="group__getters.html#ga3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="595,477,733,504"/>
<area shape="rect" href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="791,477,962,504"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="group__calculators_gae34db60faf62b898d0572ea6c039b96a_icgraph.png" border="0" usemap="#group__calculators_gae34db60faf62b898d0572ea6c039b96a_icgraph" alt=""/></div>
<map name="group__calculators_gae34db60faf62b898d0572ea6c039b96a_icgraph" id="group__calculators_gae34db60faf62b898d0572ea6c039b96a_icgraph">
<area shape="rect" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="207,13,293,39"/>
<area shape="rect" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
