<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Data Searching Functions</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.02</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle"><div class="title">Data Searching Functions<div class="ingroups"><a class="el" href="group__getters.html">Data Access and Search Functions</a></div></div></div>
</div><!--header-->
<div class="contents">

<p>These functions search the simulation data for the genotype that matches a particular known piece of information, eg a name, id, global index, or set of parents.  
<a href="#details">More...</a></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Data Searching Functions:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search.png" border="0" usemap="#agroup____search" alt=""/></div>
<map name="agroup____search" id="agroup____search">
<area shape="rect" href="group__getters.html" title="For non&#45;persistent access to simulation results and contents." alt="" coords="5,5,172,48"/>
<area shape="rect" title="These functions search the simulation data for the genotype that matches a particular known piece of ..." alt="" coords="220,13,391,40"/>
</map>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga5c7cd898a40e515c3164c98821068978"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga5c7cd898a40e515c3164c98821068978">gsc_get_name_of_id</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> id)</td></tr>
<tr class="memdesc:ga5c7cd898a40e515c3164c98821068978"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the name of the genotype with a given id.  <a href="group__search.html#ga5c7cd898a40e515c3164c98821068978">More...</a><br /></td></tr>
<tr class="separator:ga5c7cd898a40e515c3164c98821068978"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad9d1d0b5575da681af7789354447612a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gad9d1d0b5575da681af7789354447612a">gsc_get_parents_of_id</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> id, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> output[static 2])</td></tr>
<tr class="memdesc:gad9d1d0b5575da681af7789354447612a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>.  <a href="group__search.html#gad9d1d0b5575da681af7789354447612a">More...</a><br /></td></tr>
<tr class="separator:gad9d1d0b5575da681af7789354447612a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2678c4659fdb74d5b5fb074ad40137a3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga2678c4659fdb74d5b5fb074ad40137a3">gsc_get_ids_of_names</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const int n_names, const char **names, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *output)</td></tr>
<tr class="memdesc:ga2678c4659fdb74d5b5fb074ad40137a3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for genotypes with certain names in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> and save the ids of those names.  <a href="group__search.html#ga2678c4659fdb74d5b5fb074ad40137a3">More...</a><br /></td></tr>
<tr class="separator:ga2678c4659fdb74d5b5fb074ad40137a3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaefd205844e67944ce7017049516db3bd"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gaefd205844e67944ce7017049516db3bd">gsc_get_index_of_child</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> parent1id, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> parent2id)</td></tr>
<tr class="memdesc:gaefd205844e67944ce7017049516db3bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its index.  <a href="group__search.html#gaefd205844e67944ce7017049516db3bd">More...</a><br /></td></tr>
<tr class="separator:gaefd205844e67944ce7017049516db3bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga292d9c26b8d027c931a84fd09a32518f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const char *name)</td></tr>
<tr class="memdesc:ga292d9c26b8d027c931a84fd09a32518f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with a particular name in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its index.  <a href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">More...</a><br /></td></tr>
<tr class="separator:ga292d9c26b8d027c931a84fd09a32518f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7cab70998890225a5cac0d4f08850a55"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga7cab70998890225a5cac0d4f08850a55">gsc_get_id_of_index</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const int index)</td></tr>
<tr class="memdesc:ga7cab70998890225a5cac0d4f08850a55"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the id of a genotype by its index.  <a href="group__search.html#ga7cab70998890225a5cac0d4f08850a55">More...</a><br /></td></tr>
<tr class="separator:ga7cab70998890225a5cac0d4f08850a55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6bf46b86dfd6f562de44b86f64ba6eda"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga6bf46b86dfd6f562de44b86f64ba6eda">gsc_get_genes_of_index</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const int index)</td></tr>
<tr class="memdesc:ga6bf46b86dfd6f562de44b86f64ba6eda"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of a genotype by its index.  <a href="group__search.html#ga6bf46b86dfd6f562de44b86f64ba6eda">More...</a><br /></td></tr>
<tr class="separator:ga6bf46b86dfd6f562de44b86f64ba6eda"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p >These functions search the simulation data for the genotype that matches a particular known piece of information, eg a name, id, global index, or set of parents. </p>
<p >Depending on the size of the simulation, they may not be fast. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga6bf46b86dfd6f562de44b86f64ba6eda" name="ga6bf46b86dfd6f562de44b86f64ba6eda"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga6bf46b86dfd6f562de44b86f64ba6eda">&#9670;&nbsp;</a></span>gsc_get_genes_of_index()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char * gsc_get_genes_of_index </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the alleles of a genotype by its index. </p>
<p >The index is assumed to be 0-based, starting at the first entry of <code>start</code> and continuing through the linked list to which <code>start</code> belongs. Exits if the linked list does not contain that index.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which to locate the id at a particular index. </td></tr>
    <tr><td class="paramname">index</td><td>the index of the target. It is assumed to start at 0 at the start of the matrix in <code>start</code> and be incremented up until the last entry in the matrix of the last entry in the linked list. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the alleles of the genotype that has idnex <code>index</code>, as a copy of the pointer to the heap memory where the genotype is saved (so <em>don't</em> free the pointer returned from this function). It points to a sequence of characters, ordered according to the markers in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> to which the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> belongs. </dd></dl>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga6bf46b86dfd6f562de44b86f64ba6eda_icgraph.png" border="0" usemap="#agroup__search_ga6bf46b86dfd6f562de44b86f64ba6eda_icgraph" alt=""/></div>
<map name="agroup__search_ga6bf46b86dfd6f562de44b86f64ba6eda_icgraph" id="agroup__search_ga6bf46b86dfd6f562de44b86f64ba6eda_icgraph">
<area shape="rect" title="Get the alleles of a genotype by its index." alt="" coords="250,13,414,41"/>
<area shape="rect" href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,202,49"/>
</map>
</div>

</div>
</div>
<a id="ga7cab70998890225a5cac0d4f08850a55" name="ga7cab70998890225a5cac0d4f08850a55"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7cab70998890225a5cac0d4f08850a55">&#9670;&nbsp;</a></span>gsc_get_id_of_index()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> gsc_get_id_of_index </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the id of a genotype by its index. </p>
<p >The index is assumed to be 0-based, starting at the first entry of <code>start</code> and continuing through the linked list to which <code>start</code> belongs. Exits if the linked list does not contain that index.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which to locate the id at a particular index. </td></tr>
    <tr><td class="paramname">index</td><td>the index of the target. It is assumed to start at 0 at the start of the matrix in <code>start</code> and be incremented up until the last entry in the matrix of the last entry in the linked list. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the lifetime-unique id of the genotype found at that index. </dd></dl>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga7cab70998890225a5cac0d4f08850a55_icgraph.png" border="0" usemap="#agroup__search_ga7cab70998890225a5cac0d4f08850a55_icgraph" alt=""/></div>
<map name="agroup__search_ga7cab70998890225a5cac0d4f08850a55_icgraph" id="agroup__search_ga7cab70998890225a5cac0d4f08850a55_icgraph">
<area shape="rect" title="Get the id of a genotype by its index." alt="" coords="250,13,389,41"/>
<area shape="rect" href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,202,49"/>
</map>
</div>

</div>
</div>
<a id="ga2678c4659fdb74d5b5fb074ad40137a3" name="ga2678c4659fdb74d5b5fb074ad40137a3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2678c4659fdb74d5b5fb074ad40137a3">&#9670;&nbsp;</a></span>gsc_get_ids_of_names()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void gsc_get_ids_of_names </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>n_names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for genotypes with certain names in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> and save the ids of those names. </p>
<p >Exits if any name cannot be found.</p>
<p >This function must check every name in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">n_names</td><td>the length of the array of names which are being sought. </td></tr>
    <tr><td class="paramname">names</td><td>an array of strings/names whose ids we want to find. It must have <br  />
 at least [n_names] entries. </td></tr>
    <tr><td class="paramname">output</td><td>pointer to an array with at least enough space to store [n_names] PedigreeIDs. The ID corresponding to each name in <code>names</code> is saved at the corresponding index in <code>output</code> when it is found by the function. </td></tr>
  </table>
  </dd>
</dl>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga2678c4659fdb74d5b5fb074ad40137a3_icgraph.png" border="0" usemap="#agroup__search_ga2678c4659fdb74d5b5fb074ad40137a3_icgraph" alt=""/></div>
<map name="agroup__search_ga2678c4659fdb74d5b5fb074ad40137a3_icgraph" id="agroup__search_ga2678c4659fdb74d5b5fb074ad40137a3_icgraph">
<area shape="rect" title="Search for genotypes with certain names in a linked list of gsc_AlleleMatrix and save the ids of thos..." alt="" coords="233,13,387,41"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,5,185,49"/>
</map>
</div>

</div>
</div>
<a id="gaefd205844e67944ce7017049516db3bd" name="gaefd205844e67944ce7017049516db3bd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaefd205844e67944ce7017049516db3bd">&#9670;&nbsp;</a></span>gsc_get_index_of_child()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int gsc_get_index_of_child </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>parent1id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>parent2id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its index. </p>
<p >Exits if such a child cannot be found.</p>
<p >This function must check every genotype in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which the child is assumed to be found. </td></tr>
    <tr><td class="paramname">parent1id</td><td>one of the parents of the genotype must have this id. </td></tr>
    <tr><td class="paramname">parent2id</td><td>the other parent of the genotype must have this id. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the index (0-based, starting at the start of <code>start</code>) of the first sequentially located genotype whose parents match the two parent ids provided </dd></dl>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_gaefd205844e67944ce7017049516db3bd_icgraph.png" border="0" usemap="#agroup__search_gaefd205844e67944ce7017049516db3bd_icgraph" alt=""/></div>
<map name="agroup__search_gaefd205844e67944ce7017049516db3bd_icgraph" id="agroup__search_gaefd205844e67944ce7017049516db3bd_icgraph">
<area shape="rect" title="Search for a genotype with parentage matching two given parent ids in a linked list of gsc_AlleleMatr..." alt="" coords="233,13,389,41"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,5,185,49"/>
</map>
</div>

</div>
</div>
<a id="ga292d9c26b8d027c931a84fd09a32518f" name="ga292d9c26b8d027c931a84fd09a32518f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga292d9c26b8d027c931a84fd09a32518f">&#9670;&nbsp;</a></span>gsc_get_index_of_name()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int gsc_get_index_of_name </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for a genotype with a particular name in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its index. </p>
<p >Exits if such a child cannot be found.</p>
<p >This function must check every genotype in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which the genotype is assumed to be found. </td></tr>
    <tr><td class="paramname">name</td><td>a string to match to the name of the target </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the index (0-based, starting at the start of <code>start</code>) of the first sequentially located genotype whose name is the same as the provided name. </dd></dl>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga292d9c26b8d027c931a84fd09a32518f_icgraph.png" border="0" usemap="#agroup__search_ga292d9c26b8d027c931a84fd09a32518f_icgraph" alt=""/></div>
<map name="agroup__search_ga292d9c26b8d027c931a84fd09a32518f_icgraph" id="agroup__search_ga292d9c26b8d027c931a84fd09a32518f_icgraph">
<area shape="rect" title="Search for a genotype with a particular name in a linked list of gsc_AlleleMatrix,..." alt="" coords="250,44,411,72"/>
<area shape="rect" href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,202,49"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="9,73,198,101"/>
</map>
</div>

</div>
</div>
<a id="ga5c7cd898a40e515c3164c98821068978" name="ga5c7cd898a40e515c3164c98821068978"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga5c7cd898a40e515c3164c98821068978">&#9670;&nbsp;</a></span>gsc_get_name_of_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char * gsc_get_name_of_id </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the name of the genotype with a given id. </p>
<p >The function uses a bisection search on the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> where it should be found. Searching by id is therefore fast. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs.">gsc_get_from_ordered_pedigree_list()</a></dd></dl>
<p>This function assumes that ids are never reshuffled in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>. This is true as long as gsc_condense_allele_matrix's process of moving genotypes while retaining their order is the only genotype-rearranging function in use. This function's algorithm will need to be revisited if different genotype-rearranging processes are implemented.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">id</td><td>the id of the genotype whose name is sought </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the name of the genotype that has id <code>id</code>, as a copy of the pointer to the heap memory where the name is saved (so <em>don't</em> free the pointer returned from this function). Returns NULL if the ID does not exist. Note the return value might also be NULL if the genotype of this ID has no name. </dd></dl>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga5c7cd898a40e515c3164c98821068978_cgraph.png" border="0" usemap="#agroup__search_ga5c7cd898a40e515c3164c98821068978_cgraph" alt=""/></div>
<map name="agroup__search_ga5c7cd898a40e515c3164c98821068978_cgraph" id="agroup__search_ga5c7cd898a40e515c3164c98821068978_cgraph">
<area shape="rect" title="Returns the name of the genotype with a given id." alt="" coords="5,13,146,41"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="194,5,343,49"/>
</map>
</div>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_ga5c7cd898a40e515c3164c98821068978_icgraph.png" border="0" usemap="#agroup__search_ga5c7cd898a40e515c3164c98821068978_icgraph" alt=""/></div>
<map name="agroup__search_ga5c7cd898a40e515c3164c98821068978_icgraph" id="agroup__search_ga5c7cd898a40e515c3164c98821068978_icgraph">
<area shape="rect" title="Returns the name of the genotype with a given id." alt="" coords="1348,112,1489,140"/>
<area shape="rect" href="group__collgetters.html#ga69155fb6c1a62f930f2a0a8b6c419cfb" title="Gets the names of either the first or second parent of each member of the group." alt="" coords="1130,5,1279,49"/>
<area shape="rect" href="group__savers.html#gad10b44011b806daa35f8378eb90c5365" title="Print the parents of each genotype in a group to a file." alt="" coords="1134,73,1276,117"/>
<area shape="rect" href="group__savers.html#ga2bc846644aa32182f20feaa67ff3df8d" title="Print the parents of each genotype in the gsc_SimData to a file." alt="" coords="1109,141,1300,169"/>
<area shape="rect" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" title="Recursively save the parents of a particular id to a file." alt="" coords="1134,217,1276,245"/>
<area shape="rect" href="group__savers.html#ga78f467f0dd08c5501b804caca1e0c6b9" title="Print the full known pedigree of each genotype in a single gsc_AlleleMatrix to a file." alt="" coords="910,149,1058,193"/>
<area shape="rect" href="group__savers.html#ga6a5bcd7945c34e0ec99ccc92311d947c" title="Print the full known pedigree of each genotype in the gsc_SimData to a file." alt="" coords="907,217,1061,245"/>
<area shape="rect" href="group__savers.html#ga34df718dd14346f72cc20928351cc4d2" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="915,269,1053,313"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="697,149,851,193"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,149,641,193"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,23,400,51"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,75,431,103"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,127,429,155"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,179,429,223"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,247,431,275"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,299,398,327"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,115,174,159"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,183,192,227"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,251,195,279"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,303,190,347"/>
<area shape="rect" href="group__collgetters.html#ga62d40b0b1cb6c5ded508b8a5f161d38a" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="689,277,859,305"/>
</map>
</div>

</div>
</div>
<a id="gad9d1d0b5575da681af7789354447612a" name="gad9d1d0b5575da681af7789354447612a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad9d1d0b5575da681af7789354447612a">&#9670;&nbsp;</a></span>gsc_get_parents_of_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int gsc_get_parents_of_id </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>output</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>. </p>
<p >The function uses a bisection search on the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> where it should be found. Searching by id is therefore fast. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs.">gsc_get_from_ordered_pedigree_list()</a></dd></dl>
<p>This function assumes that ids are never reshuffled in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>. This is true as long as gsc_condense_allele_matrix's process of moving genotypes while retaining their order is the only genotype-rearranging function in use. This function's algorithm will need to be revisited if different genotype-rearranging processes are implemented.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of gsc_AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">id</td><td>the id of the genotype whose parents are sought </td></tr>
    <tr><td class="paramname">output</td><td>An array which the calling function can access where this function will put its results. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 when the id is successfully identified and at least one parent's id is known, 1 if neither parent is known, and 2 if the ID passed in does not exist. The ids of both parents if at least one parent is known/nonzero are saved to the array <code>output</code>. </dd></dl>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_gad9d1d0b5575da681af7789354447612a_cgraph.png" border="0" usemap="#agroup__search_gad9d1d0b5575da681af7789354447612a_cgraph" alt=""/></div>
<map name="agroup__search_gad9d1d0b5575da681af7789354447612a_cgraph" id="agroup__search_gad9d1d0b5575da681af7789354447612a_cgraph">
<area shape="rect" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="5,13,157,41"/>
<area shape="rect" href="group__supporters.html#ga8b0703c50a67ddcbe49f6b8d4ba3162b" title="Binary search through a list of PedigreeIDs." alt="" coords="205,5,354,49"/>
</map>
</div>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="group__search_gad9d1d0b5575da681af7789354447612a_icgraph.png" border="0" usemap="#agroup__search_gad9d1d0b5575da681af7789354447612a_icgraph" alt=""/></div>
<map name="agroup__search_gad9d1d0b5575da681af7789354447612a_icgraph" id="agroup__search_gad9d1d0b5575da681af7789354447612a_icgraph">
<area shape="rect" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="1299,200,1451,228"/>
<area shape="rect" href="group__savers.html#ga8e5d08dbcfc4cc77d1d1c6603b75a2ed" title="Recursively save the parents of a particular id to a file." alt="" coords="1109,200,1251,228"/>
<area shape="rect" href="group__savers.html#ga78f467f0dd08c5501b804caca1e0c6b9" title="Print the full known pedigree of each genotype in a single gsc_AlleleMatrix to a file." alt="" coords="910,132,1058,176"/>
<area shape="rect" href="group__savers.html#ga6a5bcd7945c34e0ec99ccc92311d947c" title="Print the full known pedigree of each genotype in the gsc_SimData to a file." alt="" coords="907,200,1061,228"/>
<area shape="rect" href="group__savers.html#ga34df718dd14346f72cc20928351cc4d2" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="915,252,1053,296"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="697,132,851,176"/>
<area shape="rect" href="group__crossers.html#ga28b0fcf60b6a8f6e8ca470f7b5d0a4b7" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#ga15af204958dc69717334c46b19219a2d" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga03af5446006203244dd3080cf75af26f" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#ga652376b5a34d24f239e5fc0b167c067b" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#ga4450224009587f4cd4a972413423c96e" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gab19b3eb80b5fe460d17c93e0c27f7aaf" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="26,97,174,141"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,192,209"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,233,195,261"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,285,190,329"/>
<area shape="rect" href="group__collgetters.html#ga62d40b0b1cb6c5ded508b8a5f161d38a" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="689,260,859,288"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
