<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Guarantees</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Guarantees </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p >A list of understandings about the simulation tool that you can trust will hold. C function names are used, but the assumptions still hold in the R version:</p>
<ul>
<li>No function can change genetic marker indexes or order, or the genetic map. The genetic map is set and static from the point it is loaded. A new <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must be created to change it</li>
<li>No function except delete_group can change genotype indexes or invalidate iterators. delete_group deletes all genotypes in the named groups, and shuffles all genotypes after the now-deleted ones backwards (retaining order) to fill in gaps where now-deleted genotypes sat.</li>
<li>No function will change the relative order of genotypes (follows from above).</li>
<li>Progeny from the same cross will be given sequential indexes (and IDs, if IDs are allocated). That is, when <a class="el" href="struct_gen_options.html#a76fff8d868300904c232812b3331ef46" title="The number of offspring to produce from each cross.">GenOptions.family_size</a> &gt; 1 when calling a crossing function, that fullsib family are stored side-by-side to each other.</li>
<li>Genotypes loaded from an input file will be stored (and given IDs, and indexes) in the same order as they appeared in the input file.</li>
<li>The order of genetic markers may not match the order of genetic markers in the input file: they will be re-ordered to match the genetic map (eg sorted by lower chromosome number and earlier position).</li>
</ul>
<h1><a class="anchor" id="autotoc_md44"></a>
Valid values</h1>
<ul>
<li>Genotype indexes will be integer values greater than or equal to 0. These correspond to the position the genotype is currently stored at inside the simulation's data structures.</li>
<li>Genotype Pedigree IDs will be integer values greater than or equal to 1. 0 denotes no/unknown ID. This is used to track pedigree, and is unique to each genotype for the lifetime of the simulation.</li>
<li>Group IDs/group numbers will be integer values greater than or equal to 1. All genotypes will be a member of a valid group; 0 is an error code. This is used to group together genotypes and run commands on groups of genotypes.</li>
<li>Label IDs will be integer values greater than or equal to 1. Labels can be used to save information on each genotype (eg age, sex). </li>
</ul>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
