<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Migration Guide</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.11</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Migration Guide </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><a class="anchor" id="md_Migration"></a></p>
<p >Sometimes, function signatures may need to be changed at a version update. The changes needed for a script to be updated to run in a newer version of genomicSimulation can be found in this guide. Angled brackets represent function inputs.</p>
<p >This guide belongs to the C version of genomicSimulation. Function signature changes in the R version of genomicSimulation will be managed with <code>.Deprecated</code> warnings.</p>
<h1><a class="anchor" id="autotoc_md64"></a>
0.2.5 â†’ 0.2.6</h1>
<p >Function signatures for file output functions were simplified. So were function signatures of breeding value and allele count matrix calculation functions. One extra parameter was added to <code>load_data_files</code>.</p>
<table class="doxtable">
<tr>
<th>0.2.5</th><th>0.2.6 </th></tr>
<tr>
<td><code>load_data_files(&lt;d&gt;, &lt;data_file&gt;, &lt;map_file&gt;, &lt;effect_file&gt;)</code> </td><td><code>load_data_files(&lt;d&gt;, &lt;data_file&gt;, &lt;map_file&gt;, &lt;effect_file&gt;, DETECT_FILE_FORMAT)</code> </td></tr>
<tr>
<td><code>save_markerblocks(&lt;f&gt;, &lt;d&gt;, &lt;b&gt;)</code> </td><td><code>save_markerblocks(&lt;!fname&gt;, &lt;d&gt;, &lt;b&gt;, NO_MAP)</code><br  />
followed by running the command <code>sed 's/^/0\t0\tb0\tb\t/' &lt;!fname&gt; | sed '1i\Chrom\tPos\tName\tClass\tMarkers'</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>)<br  />
(It is suggested you use one of the simpler new output formats without meaningless columns instead.) </td></tr>
<tr>
<td><code>save_group_genotypes(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;)</code> </td><td><code>save_genotypes(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, GSC_FALSE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_transposed_group_genotypes(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;)</code> </td><td><code>save_genotypes(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, GSC_TRUE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_count_matrix(&lt;f&gt;, &lt;d&gt;, &lt;allele&gt;)</code> </td><td><code>save_allele_counts(&lt;!fname&gt;, &lt;d&gt;, NO_GROUP, &lt;allele&gt;, GSC_FALSE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_group_count_matrix(&lt;f&gt;, &lt;d&gt;, &lt;allele&gt;, &lt;group&gt;)</code> </td><td><code>save_allele_counts(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, &lt;allele&gt;, GSC_FALSE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_one_step_pedigree(&lt;f&gt;, &lt;d&gt;)</code> </td><td><code>save_pedigrees(&lt;!fname&gt;, &lt;d&gt;, NO_GROUP, GSC_FALSE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_group_one_step_pedigree(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;)</code> </td><td><code>save_pedigrees(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, GSC_FALSE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_full_pedigree(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;)</code> </td><td><code>save_pedigrees(&lt;!fname&gt;, &lt;d&gt;, NO_GROUP, GSC_TRUE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_group_full_pedigree(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;)</code> </td><td><p class="starttd"><code>save_pedigrees(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, GSC_TRUE)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>)</p>
<p class="endtd"></p>
</td></tr>
<tr>
<td><code>save_bvs(&lt;f&gt;, &lt;d&gt;, &lt;effID&gt;)</code> </td><td><code>save_bvs(&lt;!fname&gt;, &lt;d&gt;, NO_GROUP, &lt;effID&gt;)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>save_group_bvs(&lt;f&gt;, &lt;d&gt;, &lt;group&gt;, &lt;effID&gt;)</code> </td><td><code>save_bvs(&lt;!fname&gt;, &lt;d&gt;, &lt;group&gt;, &lt;effID&gt;)</code><br  />
(where <code>&lt;!fname&gt;</code> is the file name that would, in the old version, have been opened as file pointer <code>&lt;f&gt;</code>) </td></tr>
<tr>
<td><code>calculate_bvs(&lt;d&gt;-&gt;m, &lt;eff&gt;)</code><br  />
where <code>&lt;d&gt;</code> is a SimData object </td><td><code>calculate_bvs(&lt;d&gt;, NO_GROUP, &lt;eff&gt;)</code> </td></tr>
<tr>
<td><code>calculate_bvs(&lt;m&gt;, &lt;eff&gt;)</code><br  />
 if <code>&lt;m&gt;</code> does not belong to a SimData object </td><td><code>BidirectionalIterator it = gsc_create_bidirectional_iter_fromAM(&lt;m&gt;, NO_GROUP);</code><br  />
<code>gsc_calculate_utility_bvs(&amp;it, &lt;eff&gt;);</code><br  />
<code>gsc_delete_bidirectional_iter(&amp;it);</code> </td></tr>
<tr>
<td><code>calculate_group_bvs(&lt;d&gt;, &lt;group&gt;, &lt;eff&gt;)</code> </td><td><code>calculate_bvs(&lt;d&gt;, &lt;group&gt;, &lt;eff&gt;)</code> </td></tr>
<tr>
<td><code>calculate_count_matrix(&lt;d&gt;-&gt;m, &lt;allele&gt;, &lt;counts&gt;)</code><br  />
where <code>&lt;d&gt;</code> is a SimData object </td><td><code>gsc_delete_dmatrix(&amp;&lt;counts&gt;);</code> (if needed)<br  />
<code>&lt;counts&gt; = calculate_allele_counts(&lt;d&gt;, NO_GROUP, &lt;allele&gt;);</code> </td></tr>
<tr>
<td><code>calculate_group_count_matrix(&lt;d&gt;-&gt;m, &lt;group&gt;, &lt;allele&gt;, &amp;&lt;counts&gt;)</code><br  />
where <code>&lt;d&gt;</code> is a SimData object </td><td><code>gsc_delete_dmatrix(&amp;&lt;counts&gt;);</code> (if needed)<br  />
<code>&lt;counts&gt; = calculate_allele_counts(&lt;d&gt;, &lt;group&gt;, &lt;allele&gt;);</code> </td></tr>
<tr>
<td><code>calculate_full_count_matrix(&lt;d&gt;-&gt;m, &lt;allele&gt;)</code><br  />
where <code>&lt;d&gt;</code> is a SimData object </td><td><code>calculate_allele_counts(&lt;d&gt;, NO_GROUP, &lt;allele&gt;)</code> </td></tr>
<tr>
<td><code>calculate_full_count_matrix(&lt;m&gt;, &lt;allele&gt;)</code><br  />
if <code>&lt;m&gt;</code> does not belong to a SimData object </td><td><code>DecimalMatrix counts = gsc_generate_zero_dmatrix(&lt;m&gt;-&gt;n_genotypes, &lt;m&gt;-&gt;n_markers);</code><br  />
<code>gsc_calculate_utility_allele_counts(&lt;m&gt;-&gt;n_markers, &lt;m&gt;-&gt;n_genotypes, &lt;m&gt;-&gt;alleles, &lt;allele&gt;, &amp;counts);</code><br  />
The result is in <code>counts</code> </td></tr>
<tr>
<td><code>gsc_save_names_header(&lt;f&gt;, &lt;n&gt;, &lt;names&gt;)</code> </td><td><code>gsc_save_utility_genotypes(&lt;f&gt;, NULL, &lt;n&gt;, &lt;names&gt;, GSC_FALSE)</code> </td></tr>
<tr>
<td><code>gsc_save_allele_matrix(&lt;f&gt;, &lt;m&gt;)</code> </td><td><code>BidirectionalIterator it = gsc_create_bidirectional_iter_fromAM(&lt;m&gt;, NO_GROUP);</code><br  />
<code>gsc_save_utility_genotypes(&lt;f&gt;, &amp;it, 0, NULL, GSC_FALSE);</code><br  />
<code>gsc_delete_bidirectional_iter(&amp;it);</code> </td></tr>
<tr>
<td><code>gsc_save_transposed_allele_matrix(&lt;f&gt;, &lt;m&gt;, &lt;names&gt;)</code> </td><td><code>BidirectionalIterator it = gsc_create_bidirectional_iter_fromAM(&lt;m&gt;, NO_GROUP);</code><br  />
 <code>gsc_save_utility_genotypes(&lt;f&gt;, &amp;it, &lt;m&gt;-&gt;n_markers, &lt;names&gt;, GSC_TRUE);</code><br  />
<code>gsc_delete_bidirectional_iter(&amp;it);</code> </td></tr>
<tr>
<td><code>gsc_save_allelematrix_full_pedigree(&lt;f&gt;, &lt;m&gt;, &lt;parents&gt;)</code> </td><td><code>BidirectionalIterator it = gsc_create_bidirectional_iter_fromAM(&lt;m&gt;, NO_GROUP);</code><br  />
 <code>gsc_save_utility_pedigrees(&lt;f&gt;, &amp;it, GSC_TRUE, &lt;parents&gt;-&gt;m);</code><br  />
<code>gsc_delete_bidirectional_iter(&amp;it);</code> </td></tr>
<tr>
<td><code><a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> bvs = gsc_calculate_bvs(&lt;m&gt;, &lt;eff&gt;);</code><br  />
<code>gsc_save_manual_bvs(&lt;f&gt;, &amp;bvs, &lt;m&gt;-&gt;ids, (const char**) &lt;m&gt;-&gt;names);</code><br  />
<code>gsc_delete_dmatrix(&amp;bvs);</code> </td><td><code><a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> it = gsc_create_bidirectional_iter_fromAM(&lt;m&gt;, NO_GROUP);</code><br  />
<code>gsc_save_utility_bvs(&lt;f&gt;, &amp;it, &lt;eff&gt;);</code><br  />
<code>gsc_delete_bidirectional_iter(&amp;it);</code> </td></tr>
</table>
<p >The signedness and size of integer parameters may have changed as part of an effort to increase consistency in integer types. The changes will depend on the definitions given to type aliases <code>GSC_ID_T</code>, <code>GSC_GLOBALX_T</code>, <code>GSC_LOCALX_T</code>, <code>GSC_GENOLEN_T</code>, so a list exact type casts you need to add to upgrade a script from v0.2.5 to v0.2.6 cannot be made. You will need to add your own type casts as requested by the compiler.</p>
<h1><a class="anchor" id="autotoc_md65"></a>
0.2.4 â†’ 0.2.5</h1>
<p >See the file <code>names_stopgap.h</code> in the v0.2.5 release.</p>
<div class="fragment"><div class="line">#define GSC_DEPRECATED_VERBOSE</div>
<div class="line">#include &quot;names_stopgap.h&quot;</div>
</div><!-- fragment --> </div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
