<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Latest News</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.3</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Latest News </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p ><a class="anchor" id="md_NEWS"></a></p>
<h1><a class="anchor" id="autotoc_md3"></a>
genomicSimulation 0.2.3</h1>
<h1><a class="anchor" id="autotoc_md4"></a>
New Features</h1>
<ul>
<li>Add functions calculate_optimal_available_alleles and calculate_optimal_available_bv, to calculate the best combination of alleles and best possible breeding value score given the pool of alleles available in a particular group.</li>
<li>Added custom integer labels, eg. for tracking age:<ul>
<li>Create a new label: create_new_label (every genotype has a value for every label; multiple labels can exist at a time).</li>
<li>Change the values of a label for some or all genotypes: set_labels_to_const; increment_labels; set_labels_to_values</li>
<li>Group manipulation using the values of a label: split_by_label_value; split_by_label_range</li>
</ul>
</li>
<li>Added two iterators: <a class="el" href="struct_bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a SimData or through only the ...">BidirectionalIterator</a> and <a class="el" href="struct_random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a SimData or of all the members of a grou...">RandomAccessIterator</a>. They can iterate through every genotype in the simulation, or through just the members of a specific group. Iterators need to be reset or destroyed and re-created when genotypes are created or destroyed.<ul>
<li>Creating and initialising iterators: create_bidirectional_iter + set_bidirectional_iter_to_start or set_bidirectional_iter_to_end; create_randomaccess_iter</li>
<li>Iterating: next_forwards; next_backwards; next_get_nth</li>
<li>Deleting: delete_bidirectional_iter; delete_randomaccess_iter</li>
</ul>
</li>
<li>Iterators return <a class="el" href="struct_geno_location.html" title="An AlleleMatrix/AlleleMatrix index coordinate of a particular genotype in the simulation.">GenoLocation</a> values. A new family of get_ functions has been added to access the data of the genotype at a particular <a class="el" href="struct_geno_location.html" title="An AlleleMatrix/AlleleMatrix index coordinate of a particular genotype in the simulation.">GenoLocation</a>.<ul>
<li>get_name; get_alleles; get_first_parent; get_second_parent; get_id; get_group; get_label_value</li>
</ul>
</li>
</ul>
<h1><a class="anchor" id="autotoc_md5"></a>
Improvements</h1>
<ul>
<li>BREAKING CHANGE: Removed automatic re-seeding in get_chromosome_locations. Users should now set the random number generator seed in their own programs (see the example sim.c). This removes the unexpected behaviour where genomicSimulation overwrites your own random seeds.</li>
<li>BREAKING CHANGE: Collective data access functions (get_group_* family, get_existing_groups and get_existing_group_counts) now use parameter modification (save their results to an array passed as a parameter) rather than by returning a heap array.</li>
<li>BREAKING CHANGE: Changed function signature of cross_these_combinations to allow for more flexible usage: instead of one <code>combinations[2][n_crosses]</code> parameter, there are a pair of <code>firstParents[n_crosses]</code> and <code>secondParents[n_crosses]</code> parameters.</li>
<li>Added 'const' to all readonly parameters. Compilation should no longer produce warnings.</li>
<li>Swapped to using Mattias Gustavsson's "rnd" implementation of a Permuted Congruential Generator (<a href="https://github.com/mattiasgustavsson/libs">https://github.com/mattiasgustavsson/libs</a>) for random number generation, rather than the builtin and subpar "rand()".</li>
</ul>
<h1><a class="anchor" id="autotoc_md6"></a>
Bug Fixes</h1>
<ul>
<li>Up NAME_LENGTH (the buffer size for name loading) to 45, as a temporary fix.</li>
<li>Allow uint ordered search to ignore 0s mid-list, to solve a bug when not all genotypes are allocated IDs</li>
</ul>
<h1><a class="anchor" id="autotoc_md7"></a>
genomicSimulationC 0.2.2</h1>
<ul>
<li>Swapped out a custom randlim() in shuffle_up_to for the basic call</li>
<li>Removed a "negative ID" warning for get_group_parent_names when the group contains members that have no known parents.</li>
<li>Fixed a bug in get_n_new_group_nums which incorrectly identified some existing groups as empty.</li>
<li>Fixed some memory leaks in the tests</li>
<li>Fixed an inappropriate memory access bug in crossing functions when will_allocate_ids = FALSE</li>
<li>Fixed an inappropriate memory access bug in make_clones when trying to inherit the name of a parent with no name.</li>
</ul>
<h1><a class="anchor" id="autotoc_md8"></a>
genomicSimulationC 0.2.1</h1>
<ul>
<li>Added function make_clones (and its genotype-generating subfunction generate_clone) to make identical duplicates of members of a group.</li>
<li>Added function to randomly cross between two groups, with the option of a cap on the number of uses of each group member as a parent of a cross:<ul>
<li>cross_randomly_between</li>
</ul>
</li>
<li>Functions split_into_individuals and split_into_families now return the groups they create.</li>
<li>Added more functions for splitting groups:<ul>
<li>split_into_halfsib_families</li>
<li>split_evenly_into_two</li>
<li>split_evenly_into_n</li>
<li>split_by_specific_counts_into_n</li>
<li>split_randomly_into_two</li>
<li>split_randomly_into_n</li>
<li>split_by_probabilities_into_n</li>
</ul>
</li>
</ul>
<p >Minor changes:</p>
<ul>
<li>Risk of losing reference to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object in load_all_simdata fixed.</li>
<li>Fixed read from freed memory error in self_n_times, fixed some dereference null pointer risks in other crossing functions.</li>
<li>Improve parameter checking in crossing and saving functions.</li>
<li>Fixed get_group_[data] functions requesting a 0-length block of heap space when they are provided with a nonexistent group. Instead they now stop and return NULL</li>
<li>Fix segfault when trying to select more individuals by GEBV than exist in the group (e.g. asking for the best 5 members of a group of 2). Now, it just moves all group members to the new selected group, and doesn't worry about the missing requested remainder.</li>
<li>Added two new helper functions: get_n_new_group_nums and shuffle_up_to</li>
</ul>
<h1><a class="anchor" id="autotoc_md9"></a>
genomicSimulationC 0.2</h1>
<ul>
<li>State corresponding to <a href="https://github.com/vllrs/genomicSimulation/releases/tag/v0.2">genomicSimulation R package v0.2</a> release. </li>
</ul>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
