<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: sim-operations.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle"><div class="title">sim-operations.c File Reference</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><code>#include &quot;<a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>&quot;</code><br />
</div><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga08d9e7703e6779d2122baa4309fca0ab"><td class="memItemLeft" align="right" valign="top">void *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga08d9e7703e6779d2122baa4309fca0ab">get_malloc</a> (size_t size)</td></tr>
<tr class="memdesc:ga08d9e7703e6779d2122baa4309fca0ab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace calls to malloc direct with this function, which errors and exits with status 2 if memory allocation fails.  <a href="group__supporters.html#ga08d9e7703e6779d2122baa4309fca0ab">More...</a><br /></td></tr>
<tr class="separator:ga08d9e7703e6779d2122baa4309fca0ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa7eaf69609ea4cb99cf61c9ac4e8f9fd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gaa7eaf69609ea4cb99cf61c9ac4e8f9fd">create_empty_allelematrix</a> (int n_markers, int n_genotypes)</td></tr>
<tr class="memdesc:gaa7eaf69609ea4cb99cf61c9ac4e8f9fd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object of a given size.  <a href="group__loaders.html#gaa7eaf69609ea4cb99cf61c9ac4e8f9fd">More...</a><br /></td></tr>
<tr class="separator:gaa7eaf69609ea4cb99cf61c9ac4e8f9fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf7350bdc690b7cb5592892858d2830a9"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">create_empty_simdata</a> ()</td></tr>
<tr class="memdesc:gaf7350bdc690b7cb5592892858d2830a9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object on the heap.  <a href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">More...</a><br /></td></tr>
<tr class="separator:gaf7350bdc690b7cb5592892858d2830a9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gac052ad3cb1f8975c21b28eea560a6e0b">clear_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object on the heap.  <a href="group__loaders.html#gac052ad3cb1f8975c21b28eea560a6e0b">More...</a><br /></td></tr>
<tr class="separator:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3d910624bfa7ac1bfacf706c3357fc5d"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga3d910624bfa7ac1bfacf706c3357fc5d">randn</a> (void)</td></tr>
<tr class="memdesc:ga3d910624bfa7ac1bfacf706c3357fc5d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a double from the normal distribution N(0,1) based on rand() and the Box-Muller algorithm.  <a href="group__maths.html#ga3d910624bfa7ac1bfacf706c3357fc5d">More...</a><br /></td></tr>
<tr class="separator:ga3d910624bfa7ac1bfacf706c3357fc5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad4e0f7934e82157e3115f8df7aec01f0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gad4e0f7934e82157e3115f8df7aec01f0">randb</a> (void)</td></tr>
<tr class="memdesc:gad4e0f7934e82157e3115f8df7aec01f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly one of 0 or 1, by splitting the output of rand() in half.  <a href="group__maths.html#gad4e0f7934e82157e3115f8df7aec01f0">More...</a><br /></td></tr>
<tr class="separator:gad4e0f7934e82157e3115f8df7aec01f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7d9dbe538d6fb096af14c3163c98e800"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga7d9dbe538d6fb096af14c3163c98e800">randpoi</a> (double lambda)</td></tr>
<tr class="memdesc:ga7d9dbe538d6fb096af14c3163c98e800"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly a number from the Poisson distribution with parameter <code>lambda</code>, using the Knuth approach.  <a href="group__maths.html#ga7d9dbe538d6fb096af14c3163c98e800">More...</a><br /></td></tr>
<tr class="separator:ga7d9dbe538d6fb096af14c3163c98e800"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga89e066191f12e056cebb8d2cd74a2a29"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga89e066191f12e056cebb8d2cd74a2a29">randlim</a> (int limit)</td></tr>
<tr class="memdesc:ga89e066191f12e056cebb8d2cd74a2a29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate an integer randomly between 0 and limit inclusive.  <a href="group__maths.html#ga89e066191f12e056cebb8d2cd74a2a29">More...</a><br /></td></tr>
<tr class="separator:ga89e066191f12e056cebb8d2cd74a2a29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga076a7b5225f3bb1f0d7912715df393b3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga076a7b5225f3bb1f0d7912715df393b3">set_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_index, int to_index)</td></tr>
<tr class="memdesc:ga076a7b5225f3bb1f0d7912715df393b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate lifetime-unique ids to each genotype in the range of whole <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> indexes <code>from_index</code> to <code>to_index</code> inclusive.  <a href="group__supporters.html#ga076a7b5225f3bb1f0d7912715df393b3">More...</a><br /></td></tr>
<tr class="separator:ga076a7b5225f3bb1f0d7912715df393b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga392c23248f22e9faf085198dbb1d8711"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="struct_table_size.html">TableSize</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga392c23248f22e9faf085198dbb1d8711">get_file_dimensions</a> (const char *filename, char sep)</td></tr>
<tr class="memdesc:ga392c23248f22e9faf085198dbb1d8711"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens a table file and reads the number of columns and rows (including headers) separated by <code>sep</code> into a <a class="el" href="struct_table_size.html" title="A simple struct used for returning the dimensions of a matrix or table.">TableSize</a> struct that is returned.  <a href="group__supporters.html#ga392c23248f22e9faf085198dbb1d8711">More...</a><br /></td></tr>
<tr class="separator:ga392c23248f22e9faf085198dbb1d8711"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6943b0633c1744bacad62ab2bf68ebd"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd">get_from_ordered_uint_list</a> (unsigned int target, unsigned int *list, unsigned int list_len)</td></tr>
<tr class="memdesc:gaa6943b0633c1744bacad62ab2bf68ebd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the located index in an array of integers where the integer is <code>target</code>.  <a href="group__supporters.html#gaa6943b0633c1744bacad62ab2bf68ebd">More...</a><br /></td></tr>
<tr class="separator:gaa6943b0633c1744bacad62ab2bf68ebd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaed2ae969c3ce4ed1466a885dcc559472"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaed2ae969c3ce4ed1466a885dcc559472">get_from_unordered_str_list</a> (char *target, char **list, int list_len)</td></tr>
<tr class="memdesc:gaed2ae969c3ce4ed1466a885dcc559472"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the first located index in an array of strings where the string is the same as the string <code>target</code>.  <a href="group__supporters.html#gaed2ae969c3ce4ed1466a885dcc559472">More...</a><br /></td></tr>
<tr class="separator:gaed2ae969c3ce4ed1466a885dcc559472"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa471b9b9b77c8010dff8d318e9a968a7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa471b9b9b77c8010dff8d318e9a968a7">shuffle_up_to</a> (int *sequence, size_t total_n, size_t n_to_shuffle)</td></tr>
<tr class="memdesc:gaa471b9b9b77c8010dff8d318e9a968a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Produce a random ordering of the first n elements in an array of integers using a (partial) Fisher-Yates shuffle.  <a href="group__supporters.html#gaa471b9b9b77c8010dff8d318e9a968a7">More...</a><br /></td></tr>
<tr class="separator:gaa471b9b9b77c8010dff8d318e9a968a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab751e196194afd414e942f0900adba72"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gab751e196194afd414e942f0900adba72">set_names</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *a, char *prefix, int suffix, int from_index)</td></tr>
<tr class="memdesc:gab751e196194afd414e942f0900adba72"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills the designated section of the <code>.names</code> array in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> with the pattern <code>prefix</code>index.  <a href="group__supporters.html#gab751e196194afd414e942f0900adba72">More...</a><br /></td></tr>
<tr class="separator:gab751e196194afd414e942f0900adba72"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad7d864d43092823ac2ed8a6cbd595e35"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gad7d864d43092823ac2ed8a6cbd595e35">get_integer_digits</a> (int i)</td></tr>
<tr class="memdesc:gad7d864d43092823ac2ed8a6cbd595e35"><td class="mdescLeft">&#160;</td><td class="mdescRight">Count and return the number of digits in <code>i</code>.  <a href="group__supporters.html#gad7d864d43092823ac2ed8a6cbd595e35">More...</a><br /></td></tr>
<tr class="separator:gad7d864d43092823ac2ed8a6cbd595e35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga80cab739edbd12bbd2ad1f9b77b27548">_simdata_pos_compare</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga80cab739edbd12bbd2ad1f9b77b27548">More...</a><br /></td></tr>
<tr class="separator:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2aae9a568cd0cd5313808e458a9fe388"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga2aae9a568cd0cd5313808e458a9fe388">_descending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga2aae9a568cd0cd5313808e458a9fe388"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga2aae9a568cd0cd5313808e458a9fe388">More...</a><br /></td></tr>
<tr class="separator:ga2aae9a568cd0cd5313808e458a9fe388"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga311381904a1ca73ea26dfede9fcddea2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga311381904a1ca73ea26dfede9fcddea2">_ascending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga311381904a1ca73ea26dfede9fcddea2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga311381904a1ca73ea26dfede9fcddea2">More...</a><br /></td></tr>
<tr class="separator:ga311381904a1ca73ea26dfede9fcddea2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3744bc6f3fcb52952d520029055e8af8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga3744bc6f3fcb52952d520029055e8af8">_ascending_float_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:ga3744bc6f3fcb52952d520029055e8af8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga3744bc6f3fcb52952d520029055e8af8">More...</a><br /></td></tr>
<tr class="separator:ga3744bc6f3fcb52952d520029055e8af8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadae2e9f7e1061cb20426ba2333b7629b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b">_ascending_int_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:gadae2e9f7e1061cb20426ba2333b7629b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b">More...</a><br /></td></tr>
<tr class="separator:gadae2e9f7e1061cb20426ba2333b7629b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacda79ff4e2a3229d13c5393d006136db"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gacda79ff4e2a3229d13c5393d006136db">_ascending_int_dcomparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:gacda79ff4e2a3229d13c5393d006136db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#gacda79ff4e2a3229d13c5393d006136db">More...</a><br /></td></tr>
<tr class="separator:gacda79ff4e2a3229d13c5393d006136db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga887027349503be24b1a847c9b833aaac"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga887027349503be24b1a847c9b833aaac">condense_allele_matrix</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga887027349503be24b1a847c9b833aaac"><td class="mdescLeft">&#160;</td><td class="mdescRight">A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__supporters.html#ga887027349503be24b1a847c9b833aaac">More...</a><br /></td></tr>
<tr class="separator:ga887027349503be24b1a847c9b833aaac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6796dadab6ccc822dd82999961cfa9e5"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga6796dadab6ccc822dd82999961cfa9e5">get_name_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id)</td></tr>
<tr class="memdesc:ga6796dadab6ccc822dd82999961cfa9e5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the name of the genotype with a given id.  <a href="group__getters.html#ga6796dadab6ccc822dd82999961cfa9e5">More...</a><br /></td></tr>
<tr class="separator:ga6796dadab6ccc822dd82999961cfa9e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8f116c3594e0e084bda2fb301548f10a"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga8f116c3594e0e084bda2fb301548f10a">get_genes_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id)</td></tr>
<tr class="memdesc:ga8f116c3594e0e084bda2fb301548f10a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the alleles at each marker of the genotype with a given id.  <a href="group__getters.html#ga8f116c3594e0e084bda2fb301548f10a">More...</a><br /></td></tr>
<tr class="separator:ga8f116c3594e0e084bda2fb301548f10a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga229c9de94875ea41a23723d337f5f92f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f">get_parents_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id, unsigned int output[2])</td></tr>
<tr class="memdesc:ga229c9de94875ea41a23723d337f5f92f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>.  <a href="group__getters.html#ga229c9de94875ea41a23723d337f5f92f">More...</a><br /></td></tr>
<tr class="separator:ga229c9de94875ea41a23723d337f5f92f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga84037d781b8850f9d90425ad16a0d059"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga84037d781b8850f9d90425ad16a0d059">get_ids_of_names</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int n_names, char *names[n_names], unsigned int *output)</td></tr>
<tr class="memdesc:ga84037d781b8850f9d90425ad16a0d059"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for genotypes with certain names in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> and save the ids of those names.  <a href="group__getters.html#ga84037d781b8850f9d90425ad16a0d059">More...</a><br /></td></tr>
<tr class="separator:ga84037d781b8850f9d90425ad16a0d059"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaadec6c1a0fe783461a72ceaee3ccefad"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gaadec6c1a0fe783461a72ceaee3ccefad">get_id_of_child</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int parent1id, unsigned int parent2id)</td></tr>
<tr class="memdesc:gaadec6c1a0fe783461a72ceaee3ccefad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its id.  <a href="group__getters.html#gaadec6c1a0fe783461a72ceaee3ccefad">More...</a><br /></td></tr>
<tr class="separator:gaadec6c1a0fe783461a72ceaee3ccefad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga79a381534123851f35f8c5cf1ee35364"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga79a381534123851f35f8c5cf1ee35364">get_index_of_child</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int parent1id, unsigned int parent2id)</td></tr>
<tr class="memdesc:ga79a381534123851f35f8c5cf1ee35364"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="group__getters.html#ga79a381534123851f35f8c5cf1ee35364">More...</a><br /></td></tr>
<tr class="separator:ga79a381534123851f35f8c5cf1ee35364"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga93779d37ca9081d07987fd83c27b84af"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga93779d37ca9081d07987fd83c27b84af">get_index_of_name</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, char *name)</td></tr>
<tr class="memdesc:ga93779d37ca9081d07987fd83c27b84af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with a particular name in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="group__getters.html#ga93779d37ca9081d07987fd83c27b84af">More...</a><br /></td></tr>
<tr class="separator:ga93779d37ca9081d07987fd83c27b84af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga67f649fd36494324e609b06fcc960785"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga67f649fd36494324e609b06fcc960785">get_id_of_index</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int index)</td></tr>
<tr class="memdesc:ga67f649fd36494324e609b06fcc960785"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the id of a genotype by its index.  <a href="group__getters.html#ga67f649fd36494324e609b06fcc960785">More...</a><br /></td></tr>
<tr class="separator:ga67f649fd36494324e609b06fcc960785"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabb1b5976faef4289fa7d5338d43aafb8"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gabb1b5976faef4289fa7d5338d43aafb8">get_genes_of_index</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int index)</td></tr>
<tr class="memdesc:gabb1b5976faef4289fa7d5338d43aafb8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of a genotype by its index.  <a href="group__getters.html#gabb1b5976faef4289fa7d5338d43aafb8">More...</a><br /></td></tr>
<tr class="separator:gabb1b5976faef4289fa7d5338d43aafb8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gade6b01bfc6450638a3a341b76e28a97d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gade6b01bfc6450638a3a341b76e28a97d">combine_groups</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int list_len, int group_ids[list_len])</td></tr>
<tr class="memdesc:gade6b01bfc6450638a3a341b76e28a97d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Combine a set of groups into one group.  <a href="group__groupmod.html#gade6b01bfc6450638a3a341b76e28a97d">More...</a><br /></td></tr>
<tr class="separator:gade6b01bfc6450638a3a341b76e28a97d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf2eda46d016c384b166ce65605952b7c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c">split_from_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int indexes_to_split[n])</td></tr>
<tr class="memdesc:gaf2eda46d016c384b166ce65605952b7c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Take a list of indexes and allocate the genotypes at those indexes to a new group.  <a href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c">More...</a><br /></td></tr>
<tr class="separator:gaf2eda46d016c384b166ce65605952b7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4cad3df4dd91ba6d8f8be038e14aa182"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga4cad3df4dd91ba6d8f8be038e14aa182">split_into_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int *results)</td></tr>
<tr class="memdesc:ga4cad3df4dd91ba6d8f8be038e14aa182"><td class="mdescLeft">&#160;</td><td class="mdescRight">Give every individual in the group a new group number that does not belong to any other existing group (thereby allocating each genotype in the group to a new group of 1).  <a href="group__groupmod.html#ga4cad3df4dd91ba6d8f8be038e14aa182">More...</a><br /></td></tr>
<tr class="separator:ga4cad3df4dd91ba6d8f8be038e14aa182"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga62aec37e088d85c1df20a53b6d65daa4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga62aec37e088d85c1df20a53b6d65daa4">split_into_families</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int *results)</td></tr>
<tr class="memdesc:ga62aec37e088d85c1df20a53b6d65daa4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into a set of smaller groups, each of which contains the genotypes in the original group that share a particular pair of parents.  <a href="group__groupmod.html#ga62aec37e088d85c1df20a53b6d65daa4">More...</a><br /></td></tr>
<tr class="separator:ga62aec37e088d85c1df20a53b6d65daa4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga39c5cb49068858b095acea9385e017ec"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga39c5cb49068858b095acea9385e017ec">split_into_halfsib_families</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int parent, int *results)</td></tr>
<tr class="memdesc:ga39c5cb49068858b095acea9385e017ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into a set of smaller groups, each containing the genotypes from the original group that share one parent.  <a href="group__groupmod.html#ga39c5cb49068858b095acea9385e017ec">More...</a><br /></td></tr>
<tr class="separator:ga39c5cb49068858b095acea9385e017ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga64b60c64209718e14b77ff4f4a991fe0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga64b60c64209718e14b77ff4f4a991fe0">split_evenly_into_two</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ga64b60c64209718e14b77ff4f4a991fe0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into two groups of equal size (or size differing only by one, if the original group had an odd number of members) using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#ga64b60c64209718e14b77ff4f4a991fe0">More...</a><br /></td></tr>
<tr class="separator:ga64b60c64209718e14b77ff4f4a991fe0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae4916db977e1014b5a4b163edf37427b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gae4916db977e1014b5a4b163edf37427b">split_evenly_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int n, int *results)</td></tr>
<tr class="memdesc:gae4916db977e1014b5a4b163edf37427b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size.), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#gae4916db977e1014b5a4b163edf37427b">More...</a><br /></td></tr>
<tr class="separator:gae4916db977e1014b5a4b163edf37427b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8fe9bd972582622c399c8b6a7c81a6d1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga8fe9bd972582622c399c8b6a7c81a6d1">split_by_specific_counts_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int n, int *counts, int *results)</td></tr>
<tr class="memdesc:ga8fe9bd972582622c399c8b6a7c81a6d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#ga8fe9bd972582622c399c8b6a7c81a6d1">More...</a><br /></td></tr>
<tr class="separator:ga8fe9bd972582622c399c8b6a7c81a6d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacd7e433cf6af229f0d2932b04468e9ca"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gacd7e433cf6af229f0d2932b04468e9ca">split_randomly_into_two</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:gacd7e433cf6af229f0d2932b04468e9ca"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flip a coin for each member of the group to decide if it should be moved to the new group.  <a href="group__groupmod.html#gacd7e433cf6af229f0d2932b04468e9ca">More...</a><br /></td></tr>
<tr class="separator:gacd7e433cf6af229f0d2932b04468e9ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga38909617537e075fa1bea7cc15ca86df"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga38909617537e075fa1bea7cc15ca86df">split_randomly_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int n, int *results)</td></tr>
<tr class="memdesc:ga38909617537e075fa1bea7cc15ca86df"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with equal probability.  <a href="group__groupmod.html#ga38909617537e075fa1bea7cc15ca86df">More...</a><br /></td></tr>
<tr class="separator:ga38909617537e075fa1bea7cc15ca86df"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga75110a594a4cd6ddfc5f6905a83c422e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga75110a594a4cd6ddfc5f6905a83c422e">split_by_probabilities_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int n, double *probs, int *results)</td></tr>
<tr class="memdesc:ga75110a594a4cd6ddfc5f6905a83c422e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with custom probabilities for each group.  <a href="group__groupmod.html#ga75110a594a4cd6ddfc5f6905a83c422e">More...</a><br /></td></tr>
<tr class="separator:ga75110a594a4cd6ddfc5f6905a83c422e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9e5a5fb63d74d84617d91d4ae9acf1ee"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga9e5a5fb63d74d84617d91d4ae9acf1ee">get_existing_groups</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int *n_groups)</td></tr>
<tr class="memdesc:ga9e5a5fb63d74d84617d91d4ae9acf1ee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify every group number that currently has members.  <a href="group__getters.html#ga9e5a5fb63d74d84617d91d4ae9acf1ee">More...</a><br /></td></tr>
<tr class="separator:ga9e5a5fb63d74d84617d91d4ae9acf1ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga66bed294fc6cad3ca04026cbbdb62e02"><td class="memItemLeft" align="right" valign="top">int **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga66bed294fc6cad3ca04026cbbdb62e02">get_existing_group_counts</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int *n_groups)</td></tr>
<tr class="memdesc:ga66bed294fc6cad3ca04026cbbdb62e02"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify every group number that currently has members and the number of genotypes currently allocated to that group.  <a href="group__getters.html#ga66bed294fc6cad3ca04026cbbdb62e02">More...</a><br /></td></tr>
<tr class="separator:ga66bed294fc6cad3ca04026cbbdb62e02"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae974a74145fe876e303e14336d3c5d3a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gae974a74145fe876e303e14336d3c5d3a">get_new_group_num</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gae974a74145fe876e303e14336d3c5d3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that does not identify a group that currently has member(s).  <a href="group__supporters.html#gae974a74145fe876e303e14336d3c5d3a">More...</a><br /></td></tr>
<tr class="separator:gae974a74145fe876e303e14336d3c5d3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga71dbb7fd0d0317d24aa06b6bbb472367"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga71dbb7fd0d0317d24aa06b6bbb472367">get_n_new_group_nums</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int *result)</td></tr>
<tr class="memdesc:ga71dbb7fd0d0317d24aa06b6bbb472367"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next n sequential integers that do not identify a group that currently has member(s).  <a href="group__supporters.html#ga71dbb7fd0d0317d24aa06b6bbb472367">More...</a><br /></td></tr>
<tr class="separator:ga71dbb7fd0d0317d24aa06b6bbb472367"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad6d2b66296cd80f5bfe303e06de7935c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c">get_group_size</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:gad6d2b66296cd80f5bfe303e06de7935c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to count the number of genotypes that currently belong to the specified group.  <a href="group__getters.html#gad6d2b66296cd80f5bfe303e06de7935c">More...</a><br /></td></tr>
<tr class="separator:gad6d2b66296cd80f5bfe303e06de7935c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab6a654922ed92b3a42e3e36ecea50d07"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gab6a654922ed92b3a42e3e36ecea50d07">get_group_genes</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:gab6a654922ed92b3a42e3e36ecea50d07"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the genes/alleles of each member of the group.  <a href="group__getters.html#gab6a654922ed92b3a42e3e36ecea50d07">More...</a><br /></td></tr>
<tr class="separator:gab6a654922ed92b3a42e3e36ecea50d07"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac68da3207f10c75563ffd90f93924d68"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gac68da3207f10c75563ffd90f93924d68">get_group_names</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:gac68da3207f10c75563ffd90f93924d68"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the names of each member of the group.  <a href="group__getters.html#gac68da3207f10c75563ffd90f93924d68">More...</a><br /></td></tr>
<tr class="separator:gac68da3207f10c75563ffd90f93924d68"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac32d82b3a249f254656a3dd56fcb2a4d"><td class="memItemLeft" align="right" valign="top">unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gac32d82b3a249f254656a3dd56fcb2a4d">get_group_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:gac32d82b3a249f254656a3dd56fcb2a4d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of each member of the group.  <a href="group__getters.html#gac32d82b3a249f254656a3dd56fcb2a4d">More...</a><br /></td></tr>
<tr class="separator:gac32d82b3a249f254656a3dd56fcb2a4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga201ff5b2b875f2f6e12618ace0e9c35b"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga201ff5b2b875f2f6e12618ace0e9c35b">get_group_indexes</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:ga201ff5b2b875f2f6e12618ace0e9c35b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the indexes (0-based, from the start of the linked list in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>) of each member of the group.  <a href="group__getters.html#ga201ff5b2b875f2f6e12618ace0e9c35b">More...</a><br /></td></tr>
<tr class="separator:ga201ff5b2b875f2f6e12618ace0e9c35b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5588ab0e42298444cc56d81cbd07790c"><td class="memItemLeft" align="right" valign="top">double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#ga5588ab0e42298444cc56d81cbd07790c">get_group_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:ga5588ab0e42298444cc56d81cbd07790c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the breeding values/breeding values/fitnesses of each member of the group.  <a href="group__getters.html#ga5588ab0e42298444cc56d81cbd07790c">More...</a><br /></td></tr>
<tr class="separator:ga5588ab0e42298444cc56d81cbd07790c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab7e7e04745c611fcfbacfd04d63b80e0"><td class="memItemLeft" align="right" valign="top">unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gab7e7e04745c611fcfbacfd04d63b80e0">get_group_parent_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size, int parent)</td></tr>
<tr class="memdesc:gab7e7e04745c611fcfbacfd04d63b80e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of either the first or second parent of each member of the group.  <a href="group__getters.html#gab7e7e04745c611fcfbacfd04d63b80e0">More...</a><br /></td></tr>
<tr class="separator:gab7e7e04745c611fcfbacfd04d63b80e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac358bb2fe5955a4c1988c5c2c9b0a7ac"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gac358bb2fe5955a4c1988c5c2c9b0a7ac">get_group_parent_names</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size, int parent)</td></tr>
<tr class="memdesc:gac358bb2fe5955a4c1988c5c2c9b0a7ac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the names of either the first or second parent of each member of the group.  <a href="group__getters.html#gac358bb2fe5955a4c1988c5c2c9b0a7ac">More...</a><br /></td></tr>
<tr class="separator:gac358bb2fe5955a4c1988c5c2c9b0a7ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa52503593f21762088c32210c9365537"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__getters.html#gaa52503593f21762088c32210c9365537">get_group_pedigrees</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:gaa52503593f21762088c32210c9365537"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the full pedigree string (as per <a class="el" href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file.">save_group_full_pedigree()</a> ) of each member of the group.  <a href="group__getters.html#gaa52503593f21762088c32210c9365537">More...</a><br /></td></tr>
<tr class="separator:gaa52503593f21762088c32210c9365537"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1">generate_zero_dmatrix</a> (int r, int c)</td></tr>
<tr class="memdesc:gabf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a matrix of c columns, r rows with all 0.  <a href="group__maths.html#gabf3f578b0aa59d5b68ccd0738dd6a9d1">More...</a><br /></td></tr>
<tr class="separator:gabf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabb8addf721442aecdd7dca3dfd955c3c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gabb8addf721442aecdd7dca3dfd955c3c">add_matrixvector_product_to_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *result, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *a, double *b)</td></tr>
<tr class="memdesc:gabb8addf721442aecdd7dca3dfd955c3c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> to a vector, and add that product to the first column of a provided <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__maths.html#gabb8addf721442aecdd7dca3dfd955c3c">More...</a><br /></td></tr>
<tr class="separator:gabb8addf721442aecdd7dca3dfd955c3c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad2c9cda1aab9679e0bfe3c88959213c6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gad2c9cda1aab9679e0bfe3c88959213c6">add_doublematrixvector_product_to_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *result, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *amat, double *avec, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *bmat, double *bvec)</td></tr>
<tr class="memdesc:gad2c9cda1aab9679e0bfe3c88959213c6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply two sets of a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> and vector, and add both products to the first column of a provided <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__maths.html#gad2c9cda1aab9679e0bfe3c88959213c6">More...</a><br /></td></tr>
<tr class="separator:gad2c9cda1aab9679e0bfe3c88959213c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a">delete_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *m)</td></tr>
<tr class="memdesc:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> and frees its memory.  <a href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a">More...</a><br /></td></tr>
<tr class="separator:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafecdc653eee11af9dbee476eca981a0b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gafecdc653eee11af9dbee476eca981a0b">delete_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:gafecdc653eee11af9dbee476eca981a0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears memory of all genotypes and associated information belonging to a particular group.  <a href="group__deletors.html#gafecdc653eee11af9dbee476eca981a0b">More...</a><br /></td></tr>
<tr class="separator:gafecdc653eee11af9dbee476eca981a0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaaf7e12b92ec0f93bb3bc2ec67155bc94">delete_genmap</a> (<a class="el" href="struct_genetic_map.html">GeneticMap</a> *m)</td></tr>
<tr class="memdesc:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_genetic_map.html" title="A type that stores the genetic map for a set of markers.">GeneticMap</a> object and frees its memory.  <a href="group__deletors.html#gaaf7e12b92ec0f93bb3bc2ec67155bc94">More...</a><br /></td></tr>
<tr class="separator:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gadd4a9c17e8c3e59fc00eee25635b110f">delete_allele_matrix</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes the full <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object and frees its memory.  <a href="group__deletors.html#gadd4a9c17e8c3e59fc00eee25635b110f">More...</a><br /></td></tr>
<tr class="separator:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga961f4b0757752fee2bd08adde7ece8ad"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga961f4b0757752fee2bd08adde7ece8ad">delete_effect_matrix</a> (<a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *m)</td></tr>
<tr class="memdesc:ga961f4b0757752fee2bd08adde7ece8ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes an <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> object and frees its memory.  <a href="group__deletors.html#ga961f4b0757752fee2bd08adde7ece8ad">More...</a><br /></td></tr>
<tr class="separator:ga961f4b0757752fee2bd08adde7ece8ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga6fc4a687a2186e9e9dcf3229d9141041">delete_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object and frees its memory.  <a href="group__deletors.html#ga6fc4a687a2186e9e9dcf3229d9141041">More...</a><br /></td></tr>
<tr class="separator:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaea2a41cc103e8d11bedff564f93a4759"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaea2a41cc103e8d11bedff564f93a4759">delete_markerblocks</a> (<a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> *b)</td></tr>
<tr class="memdesc:gaea2a41cc103e8d11bedff564f93a4759"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_marker_blocks.html" title="A struct used to store a set of blocks of markers.">MarkerBlocks</a> object and frees its associated memory.  <a href="group__deletors.html#gaea2a41cc103e8d11bedff564f93a4759">More...</a><br /></td></tr>
<tr class="separator:gaea2a41cc103e8d11bedff564f93a4759"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga156666e21151a9f267e31598574366a6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga156666e21151a9f267e31598574366a6">load_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga156666e21151a9f267e31598574366a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="group__loaders.html#ga156666e21151a9f267e31598574366a6">More...</a><br /></td></tr>
<tr class="separator:ga156666e21151a9f267e31598574366a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad5863641a67355c2c1523e1a23b6f581"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gad5863641a67355c2c1523e1a23b6f581">load_transposed_encoded_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:gad5863641a67355c2c1523e1a23b6f581"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="group__loaders.html#gad5863641a67355c2c1523e1a23b6f581">More...</a><br /></td></tr>
<tr class="separator:gad5863641a67355c2c1523e1a23b6f581"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8bd85cb47ec8e0c4049f937ef40ae102"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga8bd85cb47ec8e0c4049f937ef40ae102">load_more_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga8bd85cb47ec8e0c4049f937ef40ae102"><td class="mdescLeft">&#160;</td><td class="mdescRight">Appends genotype data from a file to an existing <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__loaders.html#ga8bd85cb47ec8e0c4049f937ef40ae102">More...</a><br /></td></tr>
<tr class="separator:ga8bd85cb47ec8e0c4049f937ef40ae102"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga8edb09b9b9eb27f84ac59cc4342ea855">load_genmap_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with data from a genetic map.  <a href="group__loaders.html#ga8edb09b9b9eb27f84ac59cc4342ea855">More...</a><br /></td></tr>
<tr class="separator:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadebbfa88bf139ddca501ab158f3746c5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gadebbfa88bf139ddca501ab158f3746c5">get_sorted_markers</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int actual_n_markers)</td></tr>
<tr class="memdesc:gadebbfa88bf139ddca501ab158f3746c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object, and sorts its markers, the rows of its parent gen <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (because they are ordered by the markers), and its genetic map so that the markers are ordered by chromosome number then position.  <a href="group__supporters.html#gadebbfa88bf139ddca501ab158f3746c5">More...</a><br /></td></tr>
<tr class="separator:gadebbfa88bf139ddca501ab158f3746c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5e049945be5700a6d5a5654b6b332842"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga5e049945be5700a6d5a5654b6b332842">get_chromosome_locations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga5e049945be5700a6d5a5654b6b332842"><td class="mdescLeft">&#160;</td><td class="mdescRight">Updates the chr_ends, n_chr and chr_lengths fields in <a class="el" href="struct_sim_data.html#ae1ae6d192937ca2edaba69edf8308711" title="A GeneticMap.">SimData.map</a>.  <a href="group__supporters.html#ga5e049945be5700a6d5a5654b6b332842">More...</a><br /></td></tr>
<tr class="separator:ga5e049945be5700a6d5a5654b6b332842"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9d6b231bd6ddde82af66a5d276a2c739"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga9d6b231bd6ddde82af66a5d276a2c739">load_effects_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga9d6b231bd6ddde82af66a5d276a2c739"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with effect values.  <a href="group__loaders.html#ga9d6b231bd6ddde82af66a5d276a2c739">More...</a><br /></td></tr>
<tr class="separator:ga9d6b231bd6ddde82af66a5d276a2c739"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga54cd625669413c2ef171060e9844d814"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">load_all_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *data_file, const char *map_file, const char *effect_file)</td></tr>
<tr class="memdesc:ga54cd625669413c2ef171060e9844d814"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination from scratch with marker allele data, a genetic map, and effect values.  <a href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">More...</a><br /></td></tr>
<tr class="separator:ga54cd625669413c2ef171060e9844d814"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gabefce26ab485e6886d7bbfcaa5692f3a">calculate_min_recombinations_fw1</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int certain)</td></tr>
<tr class="memdesc:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured.  <a href="group__recomb.html#gabefce26ab485e6886d7bbfcaa5692f3a">More...</a><br /></td></tr>
<tr class="separator:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9c9c066183d70aa46d4ed4d224116580"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga9c9c066183d70aa46d4ed4d224116580">calculate_min_recombinations_fwn</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int window_size, int certain)</td></tr>
<tr class="memdesc:ga9c9c066183d70aa46d4ed4d224116580"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it.  <a href="group__recomb.html#ga9c9c066183d70aa46d4ed4d224116580">More...</a><br /></td></tr>
<tr class="separator:ga9c9c066183d70aa46d4ed4d224116580"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">calculate_recombinations_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const char *output_file, int window_len, int certain)</td></tr>
<tr class="memdesc:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes.  <a href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">More...</a><br /></td></tr>
<tr class="separator:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae406f4a0266b7480e49899693f1b00da"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#gae406f4a0266b7480e49899693f1b00da">generate_gamete</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent_genome, char *output)</td></tr>
<tr class="memdesc:gae406f4a0266b7480e49899693f1b00da"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alleles of a given parent.  <a href="group__meiosis.html#gae406f4a0266b7480e49899693f1b00da">More...</a><br /></td></tr>
<tr class="separator:gae406f4a0266b7480e49899693f1b00da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa18b10bcd660dee5213868cc3244f826"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#gaa18b10bcd660dee5213868cc3244f826">generate_cross</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1_genome, char *parent2_genome, char *output)</td></tr>
<tr class="memdesc:gaa18b10bcd660dee5213868cc3244f826"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of crossing two genotypes.  <a href="group__meiosis.html#gaa18b10bcd660dee5213868cc3244f826">More...</a><br /></td></tr>
<tr class="separator:gaa18b10bcd660dee5213868cc3244f826"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac1f46b3de0afb4674ae6d34311e7aea7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#gac1f46b3de0afb4674ae6d34311e7aea7">generate_doubled_haploid</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent_genome, char *output)</td></tr>
<tr class="memdesc:gac1f46b3de0afb4674ae6d34311e7aea7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent.  <a href="group__meiosis.html#gac1f46b3de0afb4674ae6d34311e7aea7">More...</a><br /></td></tr>
<tr class="separator:gac1f46b3de0afb4674ae6d34311e7aea7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6cf5e2546b2e75585ca2ed4387f25cd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#gaa6cf5e2546b2e75585ca2ed4387f25cd">generate_clone</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent_genome, char *output)</td></tr>
<tr class="memdesc:gaa6cf5e2546b2e75585ca2ed4387f25cd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an identical copy of a given genotype.  <a href="group__meiosis.html#gaa6cf5e2546b2e75585ca2ed4387f25cd">More...</a><br /></td></tr>
<tr class="separator:gaa6cf5e2546b2e75585ca2ed4387f25cd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae66231fb141a76432d1211e2cf0baf25"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gae66231fb141a76432d1211e2cf0baf25">cross_random_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_group, int n_crosses, int cap, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gae66231fb141a76432d1211e2cf0baf25"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses among members of a group.  <a href="group__crossers.html#gae66231fb141a76432d1211e2cf0baf25">More...</a><br /></td></tr>
<tr class="separator:gae66231fb141a76432d1211e2cf0baf25"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3be2cb18d1c722b3a766a7a5663a8f0c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group1, int group2, int n_crosses, int cap1, int cap2, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga3be2cb18d1c722b3a766a7a5663a8f0c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses where the first parent comes from one group and the second from another.  <a href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">More...</a><br /></td></tr>
<tr class="separator:ga3be2cb18d1c722b3a766a7a5663a8f0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabb793988799d792dfe4007efdf988068"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gabb793988799d792dfe4007efdf988068">cross_these_combinations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n_combinations, int combinations[2][n_combinations], <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gabb793988799d792dfe4007efdf988068"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs the crosses of pairs of parents whose ids are provided in an array.  <a href="group__crossers.html#gabb793988799d792dfe4007efdf988068">More...</a><br /></td></tr>
<tr class="separator:gabb793988799d792dfe4007efdf988068"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadf5beb682ac15fa5a0176e480adc9d91"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91">self_n_times</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gadf5beb682ac15fa5a0176e480adc9d91"><td class="mdescLeft">&#160;</td><td class="mdescRight">Selfs each member of a group for a certain number of generations.  <a href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91">More...</a><br /></td></tr>
<tr class="separator:gadf5beb682ac15fa5a0176e480adc9d91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf078d6bc1b5d4234a9b960e6132480aa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa">make_doubled_haploids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gaf078d6bc1b5d4234a9b960e6132480aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates a doubled haploid from each member of a group.  <a href="group__crossers.html#gaf078d6bc1b5d4234a9b960e6132480aa">More...</a><br /></td></tr>
<tr class="separator:gaf078d6bc1b5d4234a9b960e6132480aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa42e48ab06f9051e5ed6879409cbe2f4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa42e48ab06f9051e5ed6879409cbe2f4">make_clones</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, int inherit_names, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gaa42e48ab06f9051e5ed6879409cbe2f4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates an identical copy of each member of a group.  <a href="group__crossers.html#gaa42e48ab06f9051e5ed6879409cbe2f4">More...</a><br /></td></tr>
<tr class="separator:gaa42e48ab06f9051e5ed6879409cbe2f4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3f1b06b7aa35e6a05918e16d1deabeee"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee">make_all_unidirectional_crosses</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga3f1b06b7aa35e6a05918e16d1deabeee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between all pairs of parents in the group <code>from_group</code> and allocates the resulting offspring to a new group.  <a href="group__crossers.html#ga3f1b06b7aa35e6a05918e16d1deabeee">More...</a><br /></td></tr>
<tr class="separator:ga3f1b06b7aa35e6a05918e16d1deabeee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga68aca99a41527b019316c8243e6fc7a7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7">make_n_crosses_from_top_m_percent</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int m, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga68aca99a41527b019316c8243e6fc7a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Find the top m percent of a group and perform random crosses between those top individuals.  <a href="group__crossers.html#ga68aca99a41527b019316c8243e6fc7a7">More...</a><br /></td></tr>
<tr class="separator:ga68aca99a41527b019316c8243e6fc7a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3b67d38c13c7a8bef2d0fc91c0365177"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177">make_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga3b67d38c13c7a8bef2d0fc91c0365177"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between pairs of parents identified by name in a file and allocate the resulting offspring to a new group.  <a href="group__crossers.html#ga3b67d38c13c7a8bef2d0fc91c0365177">More...</a><br /></td></tr>
<tr class="separator:ga3b67d38c13c7a8bef2d0fc91c0365177"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa7c90b4015e4abce750176c6d891cad2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2">make_double_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gaa7c90b4015e4abce750176c6d891cad2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between previously-generated offspring of pairs of parents identified by name in a file.  <a href="group__crossers.html#gaa7c90b4015e4abce750176c6d891cad2">More...</a><br /></td></tr>
<tr class="separator:gaa7c90b4015e4abce750176c6d891cad2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae34db60faf62b898d0572ea6c039b96a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">split_by_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, int top_n, int lowIsBest)</td></tr>
<tr class="memdesc:gae34db60faf62b898d0572ea6c039b96a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group.  <a href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">More...</a><br /></td></tr>
<tr class="separator:gae34db60faf62b898d0572ea6c039b96a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7f80905718ac87e9da1c30a2cec77ea6"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga7f80905718ac87e9da1c30a2cec77ea6">calculate_group_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, unsigned int group)</td></tr>
<tr class="memdesc:ga7f80905718ac87e9da1c30a2cec77ea6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__calculators.html#ga7f80905718ac87e9da1c30a2cec77ea6">More...</a><br /></td></tr>
<tr class="separator:ga7f80905718ac87e9da1c30a2cec77ea6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5">calculate_bvs</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *e)</td></tr>
<tr class="memdesc:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="group__calculators.html#gab5b177cf40a7e28276c460d60a3ecbb5">More...</a><br /></td></tr>
<tr class="separator:gab5b177cf40a7e28276c460d60a3ecbb5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad86ea5492e6c1cf918d6449a9eb530de"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gad86ea5492e6c1cf918d6449a9eb530de">calculate_group_count_matrix_of_allele</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, unsigned int group, char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:gad86ea5492e6c1cf918d6449a9eb530de"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a group.  <a href="group__calculators.html#gad86ea5492e6c1cf918d6449a9eb530de">More...</a><br /></td></tr>
<tr class="separator:gad86ea5492e6c1cf918d6449a9eb530de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0b2b686b9dde986b549f15d72fa4c919"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga0b2b686b9dde986b549f15d72fa4c919">calculate_group_doublecount_matrix_of_allele</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, unsigned int group, char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts, char allele2, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:ga0b2b686b9dde986b549f15d72fa4c919"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear for each genotype in a group.  <a href="group__calculators.html#ga0b2b686b9dde986b549f15d72fa4c919">More...</a><br /></td></tr>
<tr class="separator:ga0b2b686b9dde986b549f15d72fa4c919"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga32b287af2e2955edd565eedf3043f02c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga32b287af2e2955edd565eedf3043f02c">calculate_count_matrix_of_allele</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:ga32b287af2e2955edd565eedf3043f02c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (does not go through the linked list).  <a href="group__calculators.html#ga32b287af2e2955edd565eedf3043f02c">More...</a><br /></td></tr>
<tr class="separator:ga32b287af2e2955edd565eedf3043f02c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3e812683e26615b797eb4cd98b694d52"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga3e812683e26615b797eb4cd98b694d52">calculate_doublecount_matrix_of_allele</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts, char allele2, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:ga3e812683e26615b797eb4cd98b694d52"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear for each genotype in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (does not go through the linked list).  <a href="group__calculators.html#ga3e812683e26615b797eb4cd98b694d52">More...</a><br /></td></tr>
<tr class="separator:ga3e812683e26615b797eb4cd98b694d52"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga42f2398ccb53ffafce9b4a17e3222636"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636">calculate_full_count_matrix_of_allele</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char allele)</td></tr>
<tr class="memdesc:ga42f2398ccb53ffafce9b4a17e3222636"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in the linked list starting at the given <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.  <a href="group__calculators.html#ga42f2398ccb53ffafce9b4a17e3222636">More...</a><br /></td></tr>
<tr class="separator:ga42f2398ccb53ffafce9b4a17e3222636"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2f61debfc95036e86953f6111236026f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga2f61debfc95036e86953f6111236026f">create_n_blocks_by_chr</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n)</td></tr>
<tr class="memdesc:ga2f61debfc95036e86953f6111236026f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct.  <a href="group__calculators.html#ga2f61debfc95036e86953f6111236026f">More...</a><br /></td></tr>
<tr class="separator:ga2f61debfc95036e86953f6111236026f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga463082fc380a6815f07bd06401927bc0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga463082fc380a6815f07bd06401927bc0">read_block_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *block_file)</td></tr>
<tr class="memdesc:ga463082fc380a6815f07bd06401927bc0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks.  <a href="group__calculators.html#ga463082fc380a6815f07bd06401927bc0">More...</a><br /></td></tr>
<tr class="separator:ga463082fc380a6815f07bd06401927bc0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga8fdd3011c2ce5228c8c8d6486533369b">calculate_group_local_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const char *output_file, int group)</td></tr>
<tr class="memdesc:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga8fdd3011c2ce5228c8c8d6486533369b">More...</a><br /></td></tr>
<tr class="separator:ga8fdd3011c2ce5228c8c8d6486533369b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7c6e2c2f4447accd4a30031256c1610a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga7c6e2c2f4447accd4a30031256c1610a">calculate_local_bvs</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const char *output_file)</td></tr>
<tr class="memdesc:ga7c6e2c2f4447accd4a30031256c1610a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga7c6e2c2f4447accd4a30031256c1610a">More...</a><br /></td></tr>
<tr class="separator:ga7c6e2c2f4447accd4a30031256c1610a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab0550fd0210bf851b8b2b2ccbd7363cf"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gab0550fd0210bf851b8b2b2ccbd7363cf">calculate_optimal_alleles</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gab0550fd0210bf851b8b2b2ccbd7363cf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__calculators.html#gab0550fd0210bf851b8b2b2ccbd7363cf">More...</a><br /></td></tr>
<tr class="separator:gab0550fd0210bf851b8b2b2ccbd7363cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga67d98ae97e5352f32380c8b89b1e4047"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga67d98ae97e5352f32380c8b89b1e4047">calculate_optimum_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga67d98ae97e5352f32380c8b89b1e4047"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values.  <a href="group__calculators.html#ga67d98ae97e5352f32380c8b89b1e4047">More...</a><br /></td></tr>
<tr class="separator:ga67d98ae97e5352f32380c8b89b1e4047"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacef164c7781f3aa15928b482c3845418"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gacef164c7781f3aa15928b482c3845418">calculate_minimum_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gacef164c7781f3aa15928b482c3845418"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values.  <a href="group__calculators.html#gacef164c7781f3aa15928b482c3845418">More...</a><br /></td></tr>
<tr class="separator:gacef164c7781f3aa15928b482c3845418"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4d63cd249222ab25b3112106813f29b2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga4d63cd249222ab25b3112106813f29b2">save_simdata</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga4d63cd249222ab25b3112106813f29b2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga4d63cd249222ab25b3112106813f29b2">More...</a><br /></td></tr>
<tr class="separator:ga4d63cd249222ab25b3112106813f29b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga2d795d5594b914eac4e5f6fcee3afbda">save_marker_blocks</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b)</td></tr>
<tr class="memdesc:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga2d795d5594b914eac4e5f6fcee3afbda">More...</a><br /></td></tr>
<tr class="separator:ga2d795d5594b914eac4e5f6fcee3afbda"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga42a237afba9dc8f3a7339cb97d929cb6">save_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="group__savers.html#ga42a237afba9dc8f3a7339cb97d929cb6">More...</a><br /></td></tr>
<tr class="separator:ga42a237afba9dc8f3a7339cb97d929cb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaadf131f9f90546d6ab9355cbaa1cf6e1">save_transposed_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="group__savers.html#gaadf131f9f90546d6ab9355cbaa1cf6e1">More...</a><br /></td></tr>
<tr class="separator:gaadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga9e43115d9ee4ac406c14161fc6ccb327">save_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#ga9e43115d9ee4ac406c14161fc6ccb327">More...</a><br /></td></tr>
<tr class="separator:ga9e43115d9ee4ac406c14161fc6ccb327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga0f53ee5421c91ca041a1b8d4603238b5">save_transposed_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#ga0f53ee5421c91ca041a1b8d4603238b5">More...</a><br /></td></tr>
<tr class="separator:ga0f53ee5421c91ca041a1b8d4603238b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga979efe39128de97176da5460cab7dfa7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga979efe39128de97176da5460cab7dfa7">save_group_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga979efe39128de97176da5460cab7dfa7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="group__savers.html#ga979efe39128de97176da5460cab7dfa7">More...</a><br /></td></tr>
<tr class="separator:ga979efe39128de97176da5460cab7dfa7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga6b87b484064fd373aec0bfa4ce8dbef1">save_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga6b87b484064fd373aec0bfa4ce8dbef1">More...</a><br /></td></tr>
<tr class="separator:ga6b87b484064fd373aec0bfa4ce8dbef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6">save_group_full_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="group__savers.html#ga75e854cbc72b31a5e2eb03c63c33ecc6">More...</a><br /></td></tr>
<tr class="separator:ga75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga40dbed7f680c5c1181a8ed27e616417b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga40dbed7f680c5c1181a8ed27e616417b">save_full_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga40dbed7f680c5c1181a8ed27e616417b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga40dbed7f680c5c1181a8ed27e616417b">More...</a><br /></td></tr>
<tr class="separator:ga40dbed7f680c5c1181a8ed27e616417b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabf5bc0b61b008aeda217ea9a666ad9cb">save_AM_pedigree</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_sim_data.html">SimData</a> *parents)</td></tr>
<tr class="memdesc:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file.  <a href="group__savers.html#gabf5bc0b61b008aeda217ea9a666ad9cb">More...</a><br /></td></tr>
<tr class="separator:gabf5bc0b61b008aeda217ea9a666ad9cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabc1307a944deb18e0aef315783ab123e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabc1307a944deb18e0aef315783ab123e">save_parents_of</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, unsigned int p1, unsigned int p2)</td></tr>
<tr class="memdesc:gabc1307a944deb18e0aef315783ab123e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="group__savers.html#gabc1307a944deb18e0aef315783ab123e">More...</a><br /></td></tr>
<tr class="separator:gabc1307a944deb18e0aef315783ab123e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0da46978e8e546af5eb3b2635902a82b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b">save_group_bvs</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ga0da46978e8e546af5eb3b2635902a82b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in a group to a file.  <a href="group__savers.html#ga0da46978e8e546af5eb3b2635902a82b">More...</a><br /></td></tr>
<tr class="separator:ga0da46978e8e546af5eb3b2635902a82b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6024d629560199d7c97d86bda15f55fc"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc">save_bvs</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga6024d629560199d7c97d86bda15f55fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga6024d629560199d7c97d86bda15f55fc">More...</a><br /></td></tr>
<tr class="separator:ga6024d629560199d7c97d86bda15f55fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacecc6d6f4179a14a12132e37b0176877"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gacecc6d6f4179a14a12132e37b0176877">save_manual_bvs</a> (FILE *f, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *e, unsigned int *ids, char **names)</td></tr>
<tr class="memdesc:gacecc6d6f4179a14a12132e37b0176877"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>.  <a href="group__savers.html#gacecc6d6f4179a14a12132e37b0176877">More...</a><br /></td></tr>
<tr class="separator:gacecc6d6f4179a14a12132e37b0176877"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga490906a5bfdb22b90aba5cdf6282fd03">save_count_matrix</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele)</td></tr>
<tr class="memdesc:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga490906a5bfdb22b90aba5cdf6282fd03">More...</a><br /></td></tr>
<tr class="separator:ga490906a5bfdb22b90aba5cdf6282fd03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab732ec2d648a6dc01444ef74f780fc70"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gab732ec2d648a6dc01444ef74f780fc70">save_count_matrix_of_group</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele, int group)</td></tr>
<tr class="memdesc:gab732ec2d648a6dc01444ef74f780fc70"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="group__savers.html#gab732ec2d648a6dc01444ef74f780fc70">More...</a><br /></td></tr>
<tr class="separator:gab732ec2d648a6dc01444ef74f780fc70"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="var-members" name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:a5c6b92aede8f29f53476aa78d4e4133a"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a></td></tr>
<tr class="memdesc:a5c6b92aede8f29f53476aa78d4e4133a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Options parameter to run <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> functions in their bare-bones form.  <a href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">More...</a><br /></td></tr>
<tr class="separator:a5c6b92aede8f29f53476aa78d4e4133a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6db8fbe9028cc259012cbdf075c222b2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a6db8fbe9028cc259012cbdf075c222b2">RAND_HALFPOINT</a> = (RAND_MAX/2)</td></tr>
<tr class="separator:a6db8fbe9028cc259012cbdf075c222b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Variable Documentation</h2>
<a id="a5c6b92aede8f29f53476aa78d4e4133a" name="a5c6b92aede8f29f53476aa78d4e4133a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5c6b92aede8f29f53476aa78d4e4133a">&#9670;&nbsp;</a></span>BASIC_OPT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const <a class="el" href="struct_gen_options.html">GenOptions</a> BASIC_OPT</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">= {</div>
<div class="line">    .will_name_offspring = <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">    .offspring_name_prefix = NULL,</div>
<div class="line">    .family_size = 1,</div>
<div class="line">    .will_track_pedigree = <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">    .will_allocate_ids = <a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">    .filename_prefix = NULL,</div>
<div class="line">    .will_save_pedigree_to_file = <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">    .will_save_bvs_to_file = <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">    .will_save_alleles_to_file = <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">    .will_save_to_simdata = <a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a></div>
<div class="line">}</div>
<div class="ttc" id="asim-operations_8h_html_aa8cecfc5c5c054d2875c03e77b7be15d"><div class="ttname"><a href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a></div><div class="ttdeci">#define TRUE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:13</div></div>
<div class="ttc" id="asim-operations_8h_html_aa93f0eb578d23995850d61f7d61c55c1"><div class="ttname"><a href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a></div><div class="ttdeci">#define FALSE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:14</div></div>
</div><!-- fragment -->
<p>Options parameter to run <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> functions in their bare-bones form. </p>

</div>
</div>
<a id="a6db8fbe9028cc259012cbdf075c222b2" name="a6db8fbe9028cc259012cbdf075c222b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6db8fbe9028cc259012cbdf075c222b2">&#9670;&nbsp;</a></span>RAND_HALFPOINT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RAND_HALFPOINT = (RAND_MAX/2)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
