<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: sim-operations.h File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">genomicSimulationC
   &#160;<span id="projectnumber">0.1-3</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">sim-operations.h File Reference</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><code>#include &lt;string.h&gt;</code><br />
<code>#include &lt;limits.h&gt;</code><br />
<code>#include &lt;stdlib.h&gt;</code><br />
<code>#include &lt;stdio.h&gt;</code><br />
<code>#include &lt;time.h&gt;</code><br />
<code>#include &lt;math.h&gt;</code><br />
</div><div class="textblock"><div class="dynheader">
Include dependency graph for sim-operations.h:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h__incl.png" border="0" usemap="#sim-operations_8h" alt=""/></div>
<map name="sim-operations_8h" id="sim-operations_8h">
<area shape="rect" title=" " alt="" coords="199,5,317,32"/>
<area shape="rect" title=" " alt="" coords="5,80,71,107"/>
<area shape="rect" title=" " alt="" coords="95,80,159,107"/>
<area shape="rect" title=" " alt="" coords="183,80,247,107"/>
<area shape="rect" title=" " alt="" coords="271,80,333,107"/>
<area shape="rect" title=" " alt="" coords="357,80,415,107"/>
<area shape="rect" title=" " alt="" coords="440,80,503,107"/>
</map>
</div>
</div><div class="textblock"><div class="dynheader">
This graph shows which files directly or indirectly include this file:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h__dep__incl.png" border="0" usemap="#sim-operations_8hdep" alt=""/></div>
<map name="sim-operations_8hdep" id="sim-operations_8hdep">
<area shape="rect" title=" " alt="" coords="5,5,124,32"/>
<area shape="rect" href="sim-operations_8c.html" title=" " alt="" coords="5,80,124,107"/>
</map>
</div>
</div>
<p><a href="sim-operations_8h_source.html">Go to the source code of this file.</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_marker_position.html">MarkerPosition</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A struct representing a single marker location.  <a href="struct_marker_position.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_table_size.html">TableSize</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A simple struct used for returning the dimensions of a matrix or table.  <a href="struct_table_size.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A heap matrix that contains floating point numbers.  <a href="struct_decimal_matrix.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_gen_options.html">GenOptions</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A type that contains choices of settings for <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> functions that create a new AlleleMatrix/generation.  <a href="struct_gen_options.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_genetic_map.html">GeneticMap</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A type that stores the genetic map for a set of markers.  <a href="struct_genetic_map.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_new_cross.html">NewCross</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A type used for containing the basic genotype and parentage of a new cross.  <a href="struct_new_cross.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_effect_matrix.html">EffectMatrix</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A type that stores a matrix of effect values and their names.  <a href="struct_effect_matrix.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="struct_sim_data.html">SimData</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Composite type that is used to run crossing simulations.  <a href="struct_sim_data.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:a598a3330b3c21701223ee0ca14316eca"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a598a3330b3c21701223ee0ca14316eca">PI</a>&#160;&#160;&#160;3.1415926535897932384626433832795028841971693993751</td></tr>
<tr class="separator:a598a3330b3c21701223ee0ca14316eca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a156b862ebf6d213f5da19b9e3ccb779e"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a156b862ebf6d213f5da19b9e3ccb779e">TOL</a>&#160;&#160;&#160;0.00001</td></tr>
<tr class="separator:a156b862ebf6d213f5da19b9e3ccb779e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa8cecfc5c5c054d2875c03e77b7be15d"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>&#160;&#160;&#160;1</td></tr>
<tr class="separator:aa8cecfc5c5c054d2875c03e77b7be15d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa93f0eb578d23995850d61f7d61c55c1"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>&#160;&#160;&#160;0</td></tr>
<tr class="separator:aa93f0eb578d23995850d61f7d61c55c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:a210a64911b3ec2ed32894e3969365e7f"><td class="memItemLeft" align="right" valign="top">typedef struct <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a210a64911b3ec2ed32894e3969365e7f">AlleleMatrix</a></td></tr>
<tr class="memdesc:a210a64911b3ec2ed32894e3969365e7f"><td class="mdescLeft">&#160;</td><td class="mdescRight">A linked list entry that stores a matrix of alleles for a set of SNP markers and subjects.  <a href="sim-operations_8h.html#a210a64911b3ec2ed32894e3969365e7f">More...</a><br /></td></tr>
<tr class="separator:a210a64911b3ec2ed32894e3969365e7f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:a3d910624bfa7ac1bfacf706c3357fc5d"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3d910624bfa7ac1bfacf706c3357fc5d">randn</a> (void)</td></tr>
<tr class="memdesc:a3d910624bfa7ac1bfacf706c3357fc5d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a double from the normal distribution N(0,1) based on rand() and the Box-Muller algorithm.  <a href="sim-operations_8h.html#a3d910624bfa7ac1bfacf706c3357fc5d">More...</a><br /></td></tr>
<tr class="separator:a3d910624bfa7ac1bfacf706c3357fc5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4e0f7934e82157e3115f8df7aec01f0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad4e0f7934e82157e3115f8df7aec01f0">randb</a> (void)</td></tr>
<tr class="memdesc:ad4e0f7934e82157e3115f8df7aec01f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly one of 0 or 1, by splitting the output of rand() in half.  <a href="sim-operations_8h.html#ad4e0f7934e82157e3115f8df7aec01f0">More...</a><br /></td></tr>
<tr class="separator:ad4e0f7934e82157e3115f8df7aec01f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7d9dbe538d6fb096af14c3163c98e800"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a7d9dbe538d6fb096af14c3163c98e800">randpoi</a> (double lambda)</td></tr>
<tr class="memdesc:a7d9dbe538d6fb096af14c3163c98e800"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly a number from the Poisson distribution with parameter <code>lambda</code>, using the Knuth approach.  <a href="sim-operations_8h.html#a7d9dbe538d6fb096af14c3163c98e800">More...</a><br /></td></tr>
<tr class="separator:a7d9dbe538d6fb096af14c3163c98e800"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89e066191f12e056cebb8d2cd74a2a29"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a89e066191f12e056cebb8d2cd74a2a29">randlim</a> (int limit)</td></tr>
<tr class="memdesc:a89e066191f12e056cebb8d2cd74a2a29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generate an integer randomly between 0 and limit inclusive.  <a href="sim-operations_8h.html#a89e066191f12e056cebb8d2cd74a2a29">More...</a><br /></td></tr>
<tr class="separator:a89e066191f12e056cebb8d2cd74a2a29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aac694dd044a29fd93443368a7b9e8d8a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aac694dd044a29fd93443368a7b9e8d8a">create_empty_allelematrix</a> (int n_markers, int n_subjects)</td></tr>
<tr class="memdesc:aac694dd044a29fd93443368a7b9e8d8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object of a given size.  <a href="sim-operations_8h.html#aac694dd044a29fd93443368a7b9e8d8a">More...</a><br /></td></tr>
<tr class="separator:aac694dd044a29fd93443368a7b9e8d8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a392c23248f22e9faf085198dbb1d8711"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="struct_table_size.html">TableSize</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a392c23248f22e9faf085198dbb1d8711">get_file_dimensions</a> (const char *filename, char sep)</td></tr>
<tr class="memdesc:a392c23248f22e9faf085198dbb1d8711"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens a table file and reads the number of columns and rows (including headers) separated by <code>sep</code> into a <a class="el" href="struct_table_size.html" title="A simple struct used for returning the dimensions of a matrix or table.">TableSize</a> struct that is returned.  <a href="sim-operations_8h.html#a392c23248f22e9faf085198dbb1d8711">More...</a><br /></td></tr>
<tr class="separator:a392c23248f22e9faf085198dbb1d8711"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa6943b0633c1744bacad62ab2bf68ebd"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa6943b0633c1744bacad62ab2bf68ebd">get_from_ordered_uint_list</a> (unsigned int target, unsigned int *list, unsigned int list_len)</td></tr>
<tr class="memdesc:aa6943b0633c1744bacad62ab2bf68ebd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the located index in an array of integers where the integer is <code>target</code>.  <a href="sim-operations_8h.html#aa6943b0633c1744bacad62ab2bf68ebd">More...</a><br /></td></tr>
<tr class="separator:aa6943b0633c1744bacad62ab2bf68ebd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aed2ae969c3ce4ed1466a885dcc559472"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aed2ae969c3ce4ed1466a885dcc559472">get_from_unordered_str_list</a> (char *target, char **list, int list_len)</td></tr>
<tr class="memdesc:aed2ae969c3ce4ed1466a885dcc559472"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the first located index in an array of strings where the string is the same as the string <code>target</code>.  <a href="sim-operations_8h.html#aed2ae969c3ce4ed1466a885dcc559472">More...</a><br /></td></tr>
<tr class="separator:aed2ae969c3ce4ed1466a885dcc559472"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6aa75e72bb29725ffe32cd003c547480"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a6aa75e72bb29725ffe32cd003c547480">get_name_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id)</td></tr>
<tr class="memdesc:a6aa75e72bb29725ffe32cd003c547480"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the name of the genotype with a given id.  <a href="sim-operations_8h.html#a6aa75e72bb29725ffe32cd003c547480">More...</a><br /></td></tr>
<tr class="separator:a6aa75e72bb29725ffe32cd003c547480"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af49ac14d9ff4483bce6424f823d30ca0"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#af49ac14d9ff4483bce6424f823d30ca0">get_genes_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id)</td></tr>
<tr class="memdesc:af49ac14d9ff4483bce6424f823d30ca0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the alleles at each marker of the genotype with a given id.  <a href="sim-operations_8h.html#af49ac14d9ff4483bce6424f823d30ca0">More...</a><br /></td></tr>
<tr class="separator:af49ac14d9ff4483bce6424f823d30ca0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a229c9de94875ea41a23723d337f5f92f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a229c9de94875ea41a23723d337f5f92f">get_parents_of_id</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int id, unsigned int output[2])</td></tr>
<tr class="memdesc:a229c9de94875ea41a23723d337f5f92f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>.  <a href="sim-operations_8h.html#a229c9de94875ea41a23723d337f5f92f">More...</a><br /></td></tr>
<tr class="separator:a229c9de94875ea41a23723d337f5f92f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84037d781b8850f9d90425ad16a0d059"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a84037d781b8850f9d90425ad16a0d059">get_ids_of_names</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int n_names, char *names[n_names], unsigned int *output)</td></tr>
<tr class="memdesc:a84037d781b8850f9d90425ad16a0d059"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for genotypes with certain names in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> and save the ids of those names.  <a href="sim-operations_8h.html#a84037d781b8850f9d90425ad16a0d059">More...</a><br /></td></tr>
<tr class="separator:a84037d781b8850f9d90425ad16a0d059"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aadec6c1a0fe783461a72ceaee3ccefad"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aadec6c1a0fe783461a72ceaee3ccefad">get_id_of_child</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int parent1id, unsigned int parent2id)</td></tr>
<tr class="memdesc:aadec6c1a0fe783461a72ceaee3ccefad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its id.  <a href="sim-operations_8h.html#aadec6c1a0fe783461a72ceaee3ccefad">More...</a><br /></td></tr>
<tr class="separator:aadec6c1a0fe783461a72ceaee3ccefad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79a381534123851f35f8c5cf1ee35364"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a79a381534123851f35f8c5cf1ee35364">get_index_of_child</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, unsigned int parent1id, unsigned int parent2id)</td></tr>
<tr class="memdesc:a79a381534123851f35f8c5cf1ee35364"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="sim-operations_8h.html#a79a381534123851f35f8c5cf1ee35364">More...</a><br /></td></tr>
<tr class="separator:a79a381534123851f35f8c5cf1ee35364"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a93779d37ca9081d07987fd83c27b84af"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a93779d37ca9081d07987fd83c27b84af">get_index_of_name</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, char *name)</td></tr>
<tr class="memdesc:a93779d37ca9081d07987fd83c27b84af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with a particular name in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="sim-operations_8h.html#a93779d37ca9081d07987fd83c27b84af">More...</a><br /></td></tr>
<tr class="separator:a93779d37ca9081d07987fd83c27b84af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a67f649fd36494324e609b06fcc960785"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a67f649fd36494324e609b06fcc960785">get_id_of_index</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int index)</td></tr>
<tr class="memdesc:a67f649fd36494324e609b06fcc960785"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the id of a genotype by its index.  <a href="sim-operations_8h.html#a67f649fd36494324e609b06fcc960785">More...</a><br /></td></tr>
<tr class="separator:a67f649fd36494324e609b06fcc960785"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a309dbf26800ee65e982df2b347980906"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a309dbf26800ee65e982df2b347980906">get_genes_of_index</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, int index)</td></tr>
<tr class="memdesc:a309dbf26800ee65e982df2b347980906"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of a genotype by its index.  <a href="sim-operations_8h.html#a309dbf26800ee65e982df2b347980906">More...</a><br /></td></tr>
<tr class="separator:a309dbf26800ee65e982df2b347980906"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad6d2b66296cd80f5bfe303e06de7935c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad6d2b66296cd80f5bfe303e06de7935c">get_group_size</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ad6d2b66296cd80f5bfe303e06de7935c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to count the number of genotypes that currently belong to the specified group.  <a href="sim-operations_8h.html#ad6d2b66296cd80f5bfe303e06de7935c">More...</a><br /></td></tr>
<tr class="separator:ad6d2b66296cd80f5bfe303e06de7935c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac159a31261e14b325bd623899eefa796"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ac159a31261e14b325bd623899eefa796">get_group_genes</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:ac159a31261e14b325bd623899eefa796"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the genes/alleles of each member of the group.  <a href="sim-operations_8h.html#ac159a31261e14b325bd623899eefa796">More...</a><br /></td></tr>
<tr class="separator:ac159a31261e14b325bd623899eefa796"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3e00f06acc41eb012e99ae46d7a18ccc"><td class="memItemLeft" align="right" valign="top">char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3e00f06acc41eb012e99ae46d7a18ccc">get_group_names</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:a3e00f06acc41eb012e99ae46d7a18ccc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the names of each member of the group.  <a href="sim-operations_8h.html#a3e00f06acc41eb012e99ae46d7a18ccc">More...</a><br /></td></tr>
<tr class="separator:a3e00f06acc41eb012e99ae46d7a18ccc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a961042e468630f31a3607dac42f8fa34"><td class="memItemLeft" align="right" valign="top">unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a961042e468630f31a3607dac42f8fa34">get_group_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:a961042e468630f31a3607dac42f8fa34"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of each member of the group.  <a href="sim-operations_8h.html#a961042e468630f31a3607dac42f8fa34">More...</a><br /></td></tr>
<tr class="separator:a961042e468630f31a3607dac42f8fa34"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95aad1344c24e6072e6a50dcc67b60a8"><td class="memItemLeft" align="right" valign="top">unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a95aad1344c24e6072e6a50dcc67b60a8">get_group_indexes</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id, int group_size)</td></tr>
<tr class="memdesc:a95aad1344c24e6072e6a50dcc67b60a8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the indexes (0-based, from the start of the linked list in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>) of each member of the group.  <a href="sim-operations_8h.html#a95aad1344c24e6072e6a50dcc67b60a8">More...</a><br /></td></tr>
<tr class="separator:a95aad1344c24e6072e6a50dcc67b60a8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab4a5ef571ebeb5e546461398eecfb0ff"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ab4a5ef571ebeb5e546461398eecfb0ff">set_subject_names</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *a, char *prefix, int suffix, int from_index)</td></tr>
<tr class="memdesc:ab4a5ef571ebeb5e546461398eecfb0ff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills the designated section of the <code>.subject_names</code> array in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> with the pattern <code>prefix</code>index.  <a href="sim-operations_8h.html#ab4a5ef571ebeb5e546461398eecfb0ff">More...</a><br /></td></tr>
<tr class="separator:ab4a5ef571ebeb5e546461398eecfb0ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac5f92e5e4bd9d414392a2f022fdc5596"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ac5f92e5e4bd9d414392a2f022fdc5596">set_subject_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_index, int to_index)</td></tr>
<tr class="memdesc:ac5f92e5e4bd9d414392a2f022fdc5596"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate lifetime-unique ids to each genotype in the range of whole <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> indexes <code>from_index</code> to <code>to_index</code> inclusive.  <a href="sim-operations_8h.html#ac5f92e5e4bd9d414392a2f022fdc5596">More...</a><br /></td></tr>
<tr class="separator:ac5f92e5e4bd9d414392a2f022fdc5596"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad7d864d43092823ac2ed8a6cbd595e35"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad7d864d43092823ac2ed8a6cbd595e35">get_integer_digits</a> (int i)</td></tr>
<tr class="memdesc:ad7d864d43092823ac2ed8a6cbd595e35"><td class="mdescLeft">&#160;</td><td class="mdescRight">Count and return the number of digits in <code>i</code>.  <a href="sim-operations_8h.html#ad7d864d43092823ac2ed8a6cbd595e35">More...</a><br /></td></tr>
<tr class="separator:ad7d864d43092823ac2ed8a6cbd595e35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae974a74145fe876e303e14336d3c5d3a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ae974a74145fe876e303e14336d3c5d3a">get_new_group_num</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ae974a74145fe876e303e14336d3c5d3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that does not identify a group that currently has member(s).  <a href="sim-operations_8h.html#ae974a74145fe876e303e14336d3c5d3a">More...</a><br /></td></tr>
<tr class="separator:ae974a74145fe876e303e14336d3c5d3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace12db201ad55ffc72c26a76ee8fbfc5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ace12db201ad55ffc72c26a76ee8fbfc5">set_group_list</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int by_n, int new_group)</td></tr>
<tr class="separator:ace12db201ad55ffc72c26a76ee8fbfc5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a887027349503be24b1a847c9b833aaac"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a887027349503be24b1a847c9b833aaac">condense_allele_matrix</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:a887027349503be24b1a847c9b833aaac"><td class="mdescLeft">&#160;</td><td class="mdescRight">A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="sim-operations_8h.html#a887027349503be24b1a847c9b833aaac">More...</a><br /></td></tr>
<tr class="separator:a887027349503be24b1a847c9b833aaac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad1a5b9f3e850da13931ee02fbb170d41"><td class="memItemLeft" align="right" valign="top">void *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad1a5b9f3e850da13931ee02fbb170d41">get_malloc</a> (size_t size)</td></tr>
<tr class="memdesc:ad1a5b9f3e850da13931ee02fbb170d41"><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace calls to malloc direct with this function, which errors and exits with status 2 if memory allocation fails.  <a href="sim-operations_8h.html#ad1a5b9f3e850da13931ee02fbb170d41">More...</a><br /></td></tr>
<tr class="separator:ad1a5b9f3e850da13931ee02fbb170d41"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a80cab739edbd12bbd2ad1f9b77b27548"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a80cab739edbd12bbd2ad1f9b77b27548">_simdata_pos_compare</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a80cab739edbd12bbd2ad1f9b77b27548"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8h.html#a80cab739edbd12bbd2ad1f9b77b27548">More...</a><br /></td></tr>
<tr class="separator:a80cab739edbd12bbd2ad1f9b77b27548"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2aae9a568cd0cd5313808e458a9fe388"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a2aae9a568cd0cd5313808e458a9fe388">_descending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a2aae9a568cd0cd5313808e458a9fe388"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8h.html#a2aae9a568cd0cd5313808e458a9fe388">More...</a><br /></td></tr>
<tr class="separator:a2aae9a568cd0cd5313808e458a9fe388"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a311381904a1ca73ea26dfede9fcddea2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a311381904a1ca73ea26dfede9fcddea2">_ascending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a311381904a1ca73ea26dfede9fcddea2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8h.html#a311381904a1ca73ea26dfede9fcddea2">More...</a><br /></td></tr>
<tr class="separator:a311381904a1ca73ea26dfede9fcddea2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3744bc6f3fcb52952d520029055e8af8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3744bc6f3fcb52952d520029055e8af8">_ascending_float_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:a3744bc6f3fcb52952d520029055e8af8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8h.html#a3744bc6f3fcb52952d520029055e8af8">More...</a><br /></td></tr>
<tr class="separator:a3744bc6f3fcb52952d520029055e8af8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ade6b01bfc6450638a3a341b76e28a97d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ade6b01bfc6450638a3a341b76e28a97d">combine_groups</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int list_len, int group_ids[list_len])</td></tr>
<tr class="memdesc:ade6b01bfc6450638a3a341b76e28a97d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Combine a set of groups into one group.  <a href="sim-operations_8h.html#ade6b01bfc6450638a3a341b76e28a97d">More...</a><br /></td></tr>
<tr class="separator:ade6b01bfc6450638a3a341b76e28a97d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5a61d77e05e6bb0d8f4235d21b8b1f3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad5a61d77e05e6bb0d8f4235d21b8b1f3">split_into_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:ad5a61d77e05e6bb0d8f4235d21b8b1f3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Give every individual in the group a new group number that does not belong to any other existing group (thereby allocating each genotype in the group to a new group of 1).  <a href="sim-operations_8h.html#ad5a61d77e05e6bb0d8f4235d21b8b1f3">More...</a><br /></td></tr>
<tr class="separator:ad5a61d77e05e6bb0d8f4235d21b8b1f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3305ddcc29138d2d92689f39a7885ca0"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3305ddcc29138d2d92689f39a7885ca0">get_existing_groups</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int *n_groups)</td></tr>
<tr class="memdesc:a3305ddcc29138d2d92689f39a7885ca0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify every group number that currently has members.  <a href="sim-operations_8h.html#a3305ddcc29138d2d92689f39a7885ca0">More...</a><br /></td></tr>
<tr class="separator:a3305ddcc29138d2d92689f39a7885ca0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af2eda46d016c384b166ce65605952b7c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#af2eda46d016c384b166ce65605952b7c">split_from_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int indexes_to_split[n])</td></tr>
<tr class="memdesc:af2eda46d016c384b166ce65605952b7c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Take a list of indexes and allocate the genotypes at those indexes to a new group.  <a href="sim-operations_8h.html#af2eda46d016c384b166ce65605952b7c">More...</a><br /></td></tr>
<tr class="separator:af2eda46d016c384b166ce65605952b7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afecdc653eee11af9dbee476eca981a0b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#afecdc653eee11af9dbee476eca981a0b">delete_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:afecdc653eee11af9dbee476eca981a0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears memory of all genotypes and associated information belonging to a particular group.  <a href="sim-operations_8h.html#afecdc653eee11af9dbee476eca981a0b">More...</a><br /></td></tr>
<tr class="separator:afecdc653eee11af9dbee476eca981a0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aaf7e12b92ec0f93bb3bc2ec67155bc94">delete_genmap</a> (<a class="el" href="struct_genetic_map.html">GeneticMap</a> *m)</td></tr>
<tr class="memdesc:aaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_genetic_map.html" title="A type that stores the genetic map for a set of markers.">GeneticMap</a> object and frees its memory.  <a href="sim-operations_8h.html#aaf7e12b92ec0f93bb3bc2ec67155bc94">More...</a><br /></td></tr>
<tr class="separator:aaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:add4a9c17e8c3e59fc00eee25635b110f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#add4a9c17e8c3e59fc00eee25635b110f">delete_allele_matrix</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:add4a9c17e8c3e59fc00eee25635b110f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes the full <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object and frees its memory.  <a href="sim-operations_8h.html#add4a9c17e8c3e59fc00eee25635b110f">More...</a><br /></td></tr>
<tr class="separator:add4a9c17e8c3e59fc00eee25635b110f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a961f4b0757752fee2bd08adde7ece8ad"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a961f4b0757752fee2bd08adde7ece8ad">delete_effect_matrix</a> (<a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *m)</td></tr>
<tr class="memdesc:a961f4b0757752fee2bd08adde7ece8ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes an <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> object and frees its memory.  <a href="sim-operations_8h.html#a961f4b0757752fee2bd08adde7ece8ad">More...</a><br /></td></tr>
<tr class="separator:a961f4b0757752fee2bd08adde7ece8ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6fc4a687a2186e9e9dcf3229d9141041"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041">delete_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:a6fc4a687a2186e9e9dcf3229d9141041"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object and frees its memory.  <a href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041">More...</a><br /></td></tr>
<tr class="separator:a6fc4a687a2186e9e9dcf3229d9141041"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#abf3f578b0aa59d5b68ccd0738dd6a9d1">generate_zero_dmatrix</a> (int r, int c)</td></tr>
<tr class="memdesc:abf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a matrix of c columns, r rows with all 0.  <a href="sim-operations_8h.html#abf3f578b0aa59d5b68ccd0738dd6a9d1">More...</a><br /></td></tr>
<tr class="separator:abf3f578b0aa59d5b68ccd0738dd6a9d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abe8da784853a382f703129997ed08996"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#abe8da784853a382f703129997ed08996">subset_dmatrix_row</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *m, int row_index)</td></tr>
<tr class="memdesc:abe8da784853a382f703129997ed08996"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> composed only of row <code>row_index</code> from <code>m</code>  <a href="sim-operations_8h.html#abe8da784853a382f703129997ed08996">More...</a><br /></td></tr>
<tr class="separator:abe8da784853a382f703129997ed08996"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac711a312bddd30d394adda822a7c7679"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ac711a312bddd30d394adda822a7c7679">add_dmatrices</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *a, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *b)</td></tr>
<tr class="memdesc:ac711a312bddd30d394adda822a7c7679"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add the two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="sim-operations_8h.html#ac711a312bddd30d394adda822a7c7679">More...</a><br /></td></tr>
<tr class="separator:ac711a312bddd30d394adda822a7c7679"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2c58c3fef64a20bca3191f343b947933"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a2c58c3fef64a20bca3191f343b947933">add_to_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *a, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *b)</td></tr>
<tr class="memdesc:a2c58c3fef64a20bca3191f343b947933"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add the two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="sim-operations_8h.html#a2c58c3fef64a20bca3191f343b947933">More...</a><br /></td></tr>
<tr class="separator:a2c58c3fef64a20bca3191f343b947933"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1219ce0fef428e6b5317a034280497b0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a1219ce0fef428e6b5317a034280497b0">multiply_dmatrices</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *a, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *b)</td></tr>
<tr class="memdesc:a1219ce0fef428e6b5317a034280497b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform matrix multiplication on two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="sim-operations_8h.html#a1219ce0fef428e6b5317a034280497b0">More...</a><br /></td></tr>
<tr class="separator:a1219ce0fef428e6b5317a034280497b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf6542d59034f88f9e1ec48e5f053c4a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aaf6542d59034f88f9e1ec48e5f053c4a">delete_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *m)</td></tr>
<tr class="memdesc:aaf6542d59034f88f9e1ec48e5f053c4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> and frees its memory.  <a href="sim-operations_8h.html#aaf6542d59034f88f9e1ec48e5f053c4a">More...</a><br /></td></tr>
<tr class="separator:aaf6542d59034f88f9e1ec48e5f053c4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a156666e21151a9f267e31598574366a6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6">load_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:a156666e21151a9f267e31598574366a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6">More...</a><br /></td></tr>
<tr class="separator:a156666e21151a9f267e31598574366a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8bd85cb47ec8e0c4049f937ef40ae102"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102">load_more_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:a8bd85cb47ec8e0c4049f937ef40ae102"><td class="mdescLeft">&#160;</td><td class="mdescRight">Appends genotype data from a file to an existing <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102">More...</a><br /></td></tr>
<tr class="separator:a8bd85cb47ec8e0c4049f937ef40ae102"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeba5195a9b36f7f4dab457bfd84a65b2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aeba5195a9b36f7f4dab457bfd84a65b2">load_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="separator:aeba5195a9b36f7f4dab457bfd84a65b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af4f7f33b6ca4f50bcc60504a6ee75ee2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#af4f7f33b6ca4f50bcc60504a6ee75ee2">load_more_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="separator:af4f7f33b6ca4f50bcc60504a6ee75ee2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5863641a67355c2c1523e1a23b6f581"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581">load_transposed_encoded_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ad5863641a67355c2c1523e1a23b6f581"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581">More...</a><br /></td></tr>
<tr class="separator:ad5863641a67355c2c1523e1a23b6f581"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8edb09b9b9eb27f84ac59cc4342ea855"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855">load_genmap_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:a8edb09b9b9eb27f84ac59cc4342ea855"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with data from a genetic map.  <a href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855">More...</a><br /></td></tr>
<tr class="separator:a8edb09b9b9eb27f84ac59cc4342ea855"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adebbfa88bf139ddca501ab158f3746c5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5">get_sorted_markers</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int actual_n_markers)</td></tr>
<tr class="memdesc:adebbfa88bf139ddca501ab158f3746c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object, and sorts its markers, the rows of its parent gen <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (because they are ordered by the markers), and its genetic map so that the markers are ordered by chromosome number then position.  <a href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5">More...</a><br /></td></tr>
<tr class="separator:adebbfa88bf139ddca501ab158f3746c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5e049945be5700a6d5a5654b6b332842"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a5e049945be5700a6d5a5654b6b332842">get_chromosome_locations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:a5e049945be5700a6d5a5654b6b332842"><td class="mdescLeft">&#160;</td><td class="mdescRight">Updates the chr_ends, n_chr and chr_lengths fields in <a class="el" href="struct_sim_data.html#ae1ae6d192937ca2edaba69edf8308711">SimData.map</a>.  <a href="sim-operations_8h.html#a5e049945be5700a6d5a5654b6b332842">More...</a><br /></td></tr>
<tr class="separator:a5e049945be5700a6d5a5654b6b332842"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9d6b231bd6ddde82af66a5d276a2c739"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a9d6b231bd6ddde82af66a5d276a2c739">load_effects_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:a9d6b231bd6ddde82af66a5d276a2c739"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with effect values.  <a href="sim-operations_8h.html#a9d6b231bd6ddde82af66a5d276a2c739">More...</a><br /></td></tr>
<tr class="separator:a9d6b231bd6ddde82af66a5d276a2c739"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54cd625669413c2ef171060e9844d814"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814">load_all_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *data_file, const char *map_file, const char *effect_file)</td></tr>
<tr class="memdesc:a54cd625669413c2ef171060e9844d814"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination from scratch with marker allele data, a genetic map, and effect values.  <a href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814">More...</a><br /></td></tr>
<tr class="separator:a54cd625669413c2ef171060e9844d814"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad9cd4248231da3682ba2f08a40cdbe8"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aad9cd4248231da3682ba2f08a40cdbe8">calculate_min_recombinations_fw1</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int certain)</td></tr>
<tr class="memdesc:aad9cd4248231da3682ba2f08a40cdbe8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured.  <a href="sim-operations_8h.html#aad9cd4248231da3682ba2f08a40cdbe8">More...</a><br /></td></tr>
<tr class="separator:aad9cd4248231da3682ba2f08a40cdbe8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3627a380ed86ebebe937c953bc3b2339"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3627a380ed86ebebe937c953bc3b2339">calculate_min_recombinations_fwn</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int window_size, int certain)</td></tr>
<tr class="memdesc:a3627a380ed86ebebe937c953bc3b2339"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it.  <a href="sim-operations_8h.html#a3627a380ed86ebebe937c953bc3b2339">More...</a><br /></td></tr>
<tr class="separator:a3627a380ed86ebebe937c953bc3b2339"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa71cdbced90506820ef61bd842ecfa40"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa71cdbced90506820ef61bd842ecfa40">has_same_alleles</a> (char *p1, char *p2, int i)</td></tr>
<tr class="memdesc:aa71cdbced90506820ef61bd842ecfa40"><td class="mdescLeft">&#160;</td><td class="mdescRight">Simple operator to determine if at marker i, two genotypes share at least one allele.  <a href="sim-operations_8h.html#aa71cdbced90506820ef61bd842ecfa40">More...</a><br /></td></tr>
<tr class="separator:aa71cdbced90506820ef61bd842ecfa40"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab88223532880a22dfc966be4cdb4107b"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ab88223532880a22dfc966be4cdb4107b">has_same_alleles_window</a> (char *g1, char *g2, int start, int w)</td></tr>
<tr class="memdesc:ab88223532880a22dfc966be4cdb4107b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at least one allele.  <a href="sim-operations_8h.html#ab88223532880a22dfc966be4cdb4107b">More...</a><br /></td></tr>
<tr class="separator:ab88223532880a22dfc966be4cdb4107b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7813f256407ab2ea242df32ea9ad2c7e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e">calculate_recombinations_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const char *output_file, int window_len, int certain)</td></tr>
<tr class="memdesc:a7813f256407ab2ea242df32ea9ad2c7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes.  <a href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e">More...</a><br /></td></tr>
<tr class="separator:a7813f256407ab2ea242df32ea9ad2c7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae406f4a0266b7480e49899693f1b00da"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da">generate_gamete</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ae406f4a0266b7480e49899693f1b00da"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alleles of a given parent.  <a href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da">More...</a><br /></td></tr>
<tr class="separator:ae406f4a0266b7480e49899693f1b00da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa18b10bcd660dee5213868cc3244f826"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826">generate_cross</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1_genome, char *parent2_genome, char *output)</td></tr>
<tr class="memdesc:aa18b10bcd660dee5213868cc3244f826"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of crossing two subjects.  <a href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826">More...</a><br /></td></tr>
<tr class="separator:aa18b10bcd660dee5213868cc3244f826"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac1f46b3de0afb4674ae6d34311e7aea7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7">generate_doubled_haploid</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ac1f46b3de0afb4674ae6d34311e7aea7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent.  <a href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7">More...</a><br /></td></tr>
<tr class="separator:ac1f46b3de0afb4674ae6d34311e7aea7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ea77c76c329625646419a10683669f9"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9">cross_random_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_group, int n_crosses, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:a3ea77c76c329625646419a10683669f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses among members of a group.  <a href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9">More...</a><br /></td></tr>
<tr class="separator:a3ea77c76c329625646419a10683669f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb793988799d792dfe4007efdf988068"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068">cross_these_combinations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n_combinations, int combinations[2][n_combinations], <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:abb793988799d792dfe4007efdf988068"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs the crosses of pairs of parents whose ids are provided in an array.  <a href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068">More...</a><br /></td></tr>
<tr class="separator:abb793988799d792dfe4007efdf988068"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf5beb682ac15fa5a0176e480adc9d91"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91">self_n_times</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:adf5beb682ac15fa5a0176e480adc9d91"><td class="mdescLeft">&#160;</td><td class="mdescRight">Selfs each member of a group for a certain number of generations.  <a href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91">More...</a><br /></td></tr>
<tr class="separator:adf5beb682ac15fa5a0176e480adc9d91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af078d6bc1b5d4234a9b960e6132480aa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa">make_doubled_haploids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:af078d6bc1b5d4234a9b960e6132480aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates a doubled haploid from each member of a group.  <a href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa">More...</a><br /></td></tr>
<tr class="separator:af078d6bc1b5d4234a9b960e6132480aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f1b06b7aa35e6a05918e16d1deabeee"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee">make_all_unidirectional_crosses</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int from_group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:a3f1b06b7aa35e6a05918e16d1deabeee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between all pairs of parents in the group <code>from_group</code> and allocates the resulting offspring to a new group.  <a href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee">More...</a><br /></td></tr>
<tr class="separator:a3f1b06b7aa35e6a05918e16d1deabeee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a68aca99a41527b019316c8243e6fc7a7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7">make_n_crosses_from_top_m_percent</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int n, int m, int group, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:a68aca99a41527b019316c8243e6fc7a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Find the top m percent of a group and perform random crosses between those top individuals.  <a href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7">More...</a><br /></td></tr>
<tr class="separator:a68aca99a41527b019316c8243e6fc7a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3b67d38c13c7a8bef2d0fc91c0365177"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177">make_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:a3b67d38c13c7a8bef2d0fc91c0365177"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between pairs of parents identified by name in a file and allocate the resulting offspring to a new group.  <a href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177">More...</a><br /></td></tr>
<tr class="separator:a3b67d38c13c7a8bef2d0fc91c0365177"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7c90b4015e4abce750176c6d891cad2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2">make_double_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:aa7c90b4015e4abce750176c6d891cad2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between previously-generated offspring of pairs of parents identified by name in a file.  <a href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2">More...</a><br /></td></tr>
<tr class="separator:aa7c90b4015e4abce750176c6d891cad2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa91ffff83cf204fff8baef3790738b7a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a">split_group_by_fitness</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group, int top_n, int lowIsBest)</td></tr>
<tr class="memdesc:aa91ffff83cf204fff8baef3790738b7a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the top <code>top_n</code> fitness/GEBV individuals in the group and puts them in a new group.  <a href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a">More...</a><br /></td></tr>
<tr class="separator:aa91ffff83cf204fff8baef3790738b7a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac297afb7cf2c56f6f3f436090d7962dd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd">calculate_fitness_metric_of_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:ac297afb7cf2c56f6f3f436090d7962dd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/GEBV for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd">More...</a><br /></td></tr>
<tr class="separator:ac297afb7cf2c56f6f3f436090d7962dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a238032359089625f3fd3b4447cdbbdff"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff">calculate_fitness_metric</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *e)</td></tr>
<tr class="memdesc:a238032359089625f3fd3b4447cdbbdff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/GEBV for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff">More...</a><br /></td></tr>
<tr class="separator:a238032359089625f3fd3b4447cdbbdff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae7797aeac935227749fc6eae09a02a96"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ae7797aeac935227749fc6eae09a02a96">calculate_count_matrix_of_allele_for_ids</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, unsigned int *for_ids, unsigned int n_ids, char allele)</td></tr>
<tr class="memdesc:ae7797aeac935227749fc6eae09a02a96"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype is a set of given genotype.  <a href="sim-operations_8h.html#ae7797aeac935227749fc6eae09a02a96">More...</a><br /></td></tr>
<tr class="separator:ae7797aeac935227749fc6eae09a02a96"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42f2398ccb53ffafce9b4a17e3222636"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a42f2398ccb53ffafce9b4a17e3222636">calculate_full_count_matrix_of_allele</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char allele)</td></tr>
<tr class="memdesc:a42f2398ccb53ffafce9b4a17e3222636"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.  <a href="sim-operations_8h.html#a42f2398ccb53ffafce9b4a17e3222636">More...</a><br /></td></tr>
<tr class="separator:a42f2398ccb53ffafce9b4a17e3222636"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4f233b73d31da864cdf803b52231e555"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a4f233b73d31da864cdf803b52231e555">calculate_group_block_effects</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *block_file, char *output_file, int group)</td></tr>
<tr class="separator:a4f233b73d31da864cdf803b52231e555"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d63cd249222ab25b3112106813f29b2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a4d63cd249222ab25b3112106813f29b2">save_simdata</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:a4d63cd249222ab25b3112106813f29b2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="sim-operations_8h.html#a4d63cd249222ab25b3112106813f29b2">More...</a><br /></td></tr>
<tr class="separator:a4d63cd249222ab25b3112106813f29b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42a237afba9dc8f3a7339cb97d929cb6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a42a237afba9dc8f3a7339cb97d929cb6">save_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:a42a237afba9dc8f3a7339cb97d929cb6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the geneotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="sim-operations_8h.html#a42a237afba9dc8f3a7339cb97d929cb6">More...</a><br /></td></tr>
<tr class="separator:a42a237afba9dc8f3a7339cb97d929cb6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aadf131f9f90546d6ab9355cbaa1cf6e1">save_transposed_allele_matrix</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, char **markers)</td></tr>
<tr class="memdesc:aadf131f9f90546d6ab9355cbaa1cf6e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="sim-operations_8h.html#aadf131f9f90546d6ab9355cbaa1cf6e1">More...</a><br /></td></tr>
<tr class="separator:aadf131f9f90546d6ab9355cbaa1cf6e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9e43115d9ee4ac406c14161fc6ccb327"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327">save_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:a9e43115d9ee4ac406c14161fc6ccb327"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327">More...</a><br /></td></tr>
<tr class="separator:a9e43115d9ee4ac406c14161fc6ccb327"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f53ee5421c91ca041a1b8d4603238b5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5">save_transposed_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group_id)</td></tr>
<tr class="memdesc:a0f53ee5421c91ca041a1b8d4603238b5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5">More...</a><br /></td></tr>
<tr class="separator:a0f53ee5421c91ca041a1b8d4603238b5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a979efe39128de97176da5460cab7dfa7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7">save_group_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:a979efe39128de97176da5460cab7dfa7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7">More...</a><br /></td></tr>
<tr class="separator:a979efe39128de97176da5460cab7dfa7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6b87b484064fd373aec0bfa4ce8dbef1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a6b87b484064fd373aec0bfa4ce8dbef1">save_one_step_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:a6b87b484064fd373aec0bfa4ce8dbef1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="sim-operations_8h.html#a6b87b484064fd373aec0bfa4ce8dbef1">More...</a><br /></td></tr>
<tr class="separator:a6b87b484064fd373aec0bfa4ce8dbef1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6">save_group_full_pedigree</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:a75e854cbc72b31a5e2eb03c63c33ecc6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6">More...</a><br /></td></tr>
<tr class="separator:a75e854cbc72b31a5e2eb03c63c33ecc6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad2fb503b61c909787f62b2a64385557"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557">save_full_pedigree</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *parents)</td></tr>
<tr class="memdesc:aad2fb503b61c909787f62b2a64385557"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file.  <a href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557">More...</a><br /></td></tr>
<tr class="separator:aad2fb503b61c909787f62b2a64385557"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4a41a75435588bd183d4d4d91082e601"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a4a41a75435588bd183d4d4d91082e601">save_parents_of</a> (FILE *f, <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, unsigned int id)</td></tr>
<tr class="memdesc:a4a41a75435588bd183d4d4d91082e601"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="sim-operations_8h.html#a4a41a75435588bd183d4d4d91082e601">More...</a><br /></td></tr>
<tr class="separator:a4a41a75435588bd183d4d4d91082e601"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d37c413cb4c979b4eafa1d0b656edd1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1">save_group_fitness</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, int group)</td></tr>
<tr class="memdesc:a4d37c413cb4c979b4eafa1d0b656edd1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the GEBV of each genotype in a group to a file.  <a href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1">More...</a><br /></td></tr>
<tr class="separator:a4d37c413cb4c979b4eafa1d0b656edd1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f5320627e7155d6e69e4c3db6208be4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a8f5320627e7155d6e69e4c3db6208be4">save_fitness</a> (FILE *f, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *e, unsigned int *ids, char **names)</td></tr>
<tr class="memdesc:a8f5320627e7155d6e69e4c3db6208be4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a set of pre-calculated GEBVs with provided names and ids to a file, with the same format as a regular call to <code>save_all_fitness</code> or <code>save_group_fitness</code>.  <a href="sim-operations_8h.html#a8f5320627e7155d6e69e4c3db6208be4">More...</a><br /></td></tr>
<tr class="separator:a8f5320627e7155d6e69e4c3db6208be4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd53d955f89343aec74bda4195aaa750"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750">save_all_fitness</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:abd53d955f89343aec74bda4195aaa750"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the GEBV of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750">More...</a><br /></td></tr>
<tr class="separator:abd53d955f89343aec74bda4195aaa750"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a490906a5bfdb22b90aba5cdf6282fd03"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a490906a5bfdb22b90aba5cdf6282fd03">save_count_matrix</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele)</td></tr>
<tr class="memdesc:a490906a5bfdb22b90aba5cdf6282fd03"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="sim-operations_8h.html#a490906a5bfdb22b90aba5cdf6282fd03">More...</a><br /></td></tr>
<tr class="separator:a490906a5bfdb22b90aba5cdf6282fd03"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab732ec2d648a6dc01444ef74f780fc70"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70">save_count_matrix_of_group</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, char allele, int group)</td></tr>
<tr class="memdesc:ab732ec2d648a6dc01444ef74f780fc70"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70">More...</a><br /></td></tr>
<tr class="separator:ab732ec2d648a6dc01444ef74f780fc70"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab569f6043f8d810aee8d8c5ba380abbd"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#ab569f6043f8d810aee8d8c5ba380abbd">calculate_ideal_genotype</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ab569f6043f8d810aee8d8c5ba380abbd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="sim-operations_8h.html#ab569f6043f8d810aee8d8c5ba380abbd">More...</a><br /></td></tr>
<tr class="separator:ab569f6043f8d810aee8d8c5ba380abbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:a74936c18ca56d0cfa05dc9d786288c17"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="struct_sim_data.html">SimData</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a74936c18ca56d0cfa05dc9d786288c17">EMPTY_SIMDATA</a></td></tr>
<tr class="separator:a74936c18ca56d0cfa05dc9d786288c17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c6b92aede8f29f53476aa78d4e4133a"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a></td></tr>
<tr class="separator:a5c6b92aede8f29f53476aa78d4e4133a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6db8fbe9028cc259012cbdf075c222b2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8h.html#a6db8fbe9028cc259012cbdf075c222b2">RAND_HALFPOINT</a></td></tr>
<tr class="separator:a6db8fbe9028cc259012cbdf075c222b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a id="aa93f0eb578d23995850d61f7d61c55c1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa93f0eb578d23995850d61f7d61c55c1">&#9670;&nbsp;</a></span>FALSE</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define FALSE&#160;&#160;&#160;0</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a598a3330b3c21701223ee0ca14316eca"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a598a3330b3c21701223ee0ca14316eca">&#9670;&nbsp;</a></span>PI</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define PI&#160;&#160;&#160;3.1415926535897932384626433832795028841971693993751</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a156b862ebf6d213f5da19b9e3ccb779e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a156b862ebf6d213f5da19b9e3ccb779e">&#9670;&nbsp;</a></span>TOL</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define TOL&#160;&#160;&#160;0.00001</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="aa8cecfc5c5c054d2875c03e77b7be15d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa8cecfc5c5c054d2875c03e77b7be15d">&#9670;&nbsp;</a></span>TRUE</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define TRUE&#160;&#160;&#160;1</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Typedef Documentation</h2>
<a id="a210a64911b3ec2ed32894e3969365e7f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a210a64911b3ec2ed32894e3969365e7f">&#9670;&nbsp;</a></span>AlleleMatrix</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef struct <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A linked list entry that stores a matrix of alleles for a set of SNP markers and subjects. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alleles</td><td>a matrix of SNP markers by subjects containing pairs of alleles eg TT, TA. Use <code>alleles[subject index][marker index * 2]</code> to get the first allele and <code>alleles[subject index][marker index * 2 + 1]</code> to get the second. In usual use the char** alleles will point to an 1000-ptr-long piece of heap memory, and if you need to add more than 100 subjects you will add another <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to the linked list </td></tr>
    <tr><td class="paramname">subject_names</td><td>array of strings containing the names of the lines/subjects whose data is stored in <code>alleles</code>. NULL if they do not have names. </td></tr>
    <tr><td class="paramname">ids</td><td>unique ids for each subject in the matrix. If you use a creator function, these will be generated for you. </td></tr>
    <tr><td class="paramname">n_subjects</td><td>the number of subjects currently loaded in <code>alleles</code> </td></tr>
    <tr><td class="paramname">n_markers</td><td>the number of markers in <code>alleles</code> </td></tr>
    <tr><td class="paramname">pedigrees</td><td>two lists of integer IDs of the parents of this subject if tracked, or 0 if we don't know/care about this pedigree. </td></tr>
    <tr><td class="paramname">next</td><td>pointer to the next <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in the linked list, or NULL if this entry is the last. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="a311381904a1ca73ea26dfede9fcddea2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a311381904a1ca73ea26dfede9fcddea2">&#9670;&nbsp;</a></span>_ascending_double_comparer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int _ascending_double_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p>Used to compare an array of doubles* to sort them in ascending order of the doubles they point to. </p><dl class="section see"><dt>See also</dt><dd>get_top_n_fitness_subject_indexes()</dd></dl>
<p>Sorts lower numbers before higher numbers. If they are equal, their order after comparison is undefined. </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a311381904a1ca73ea26dfede9fcddea2_icgraph.png" border="0" usemap="#sim-operations_8h_a311381904a1ca73ea26dfede9fcddea2_icgraph" alt=""/></div>
<map name="sim-operations_8h_a311381904a1ca73ea26dfede9fcddea2_icgraph" id="sim-operations_8h_a311381904a1ca73ea26dfede9fcddea2_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="405,13,600,39"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,13,357,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="a3744bc6f3fcb52952d520029055e8af8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3744bc6f3fcb52952d520029055e8af8">&#9670;&nbsp;</a></span>_ascending_float_comparer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int _ascending_float_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p>Used to compare an array of floats to sort them in ascending order. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle...">generate_gamete()</a></dd></dl>
<p>Sorts lower numbers before higher numbers. If floats are equal, their order after comparison is undefined. </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3744bc6f3fcb52952d520029055e8af8_icgraph.png" border="0" usemap="#sim-operations_8h_a3744bc6f3fcb52952d520029055e8af8_icgraph" alt=""/></div>
<map name="sim-operations_8h_a3744bc6f3fcb52952d520029055e8af8_icgraph" id="sim-operations_8h_a3744bc6f3fcb52952d520029055e8af8_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="661,209,841,235"/>
<area shape="rect" href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="472,133,583,159"/>
<area shape="rect" href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="441,209,613,235"/>
<area shape="rect" href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="466,259,589,286"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,50,390,77"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,158,352,185"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,209,385,235"/>
</map>
</div>

</div>
</div>
<a id="a2aae9a568cd0cd5313808e458a9fe388"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2aae9a568cd0cd5313808e458a9fe388">&#9670;&nbsp;</a></span>_descending_double_comparer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int _descending_double_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p>Used to compare an array of doubles* to sort them in descending order of the doubles they point to. </p><dl class="section see"><dt>See also</dt><dd>get_top_n_fitness_subject_indexes()</dd></dl>
<p>Sorts higher numbers before lower numbers. If they are equal, their order after comparison is undefined. </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a2aae9a568cd0cd5313808e458a9fe388_icgraph.png" border="0" usemap="#sim-operations_8h_a2aae9a568cd0cd5313808e458a9fe388_icgraph" alt=""/></div>
<map name="sim-operations_8h_a2aae9a568cd0cd5313808e458a9fe388_icgraph" id="sim-operations_8h_a2aae9a568cd0cd5313808e458a9fe388_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="405,5,545,47"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,13,357,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="a80cab739edbd12bbd2ad1f9b77b27548"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a80cab739edbd12bbd2ad1f9b77b27548">&#9670;&nbsp;</a></span>_simdata_pos_compare()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int _simdata_pos_compare </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p>Used to compare an array of MarkerPosition** </p><dl class="section see"><dt>See also</dt><dd>sort_markers()</dd></dl>
<p>Use this on an array of pointers to values in SimData.map.positions.</p>
<p>Sorts lower chromosome numbers before higher chromosome numbers. Within chromosomes, sorts lower positions before higher ones.</p>
<p>If chromosome number is 0, the <a class="el" href="struct_marker_position.html" title="A struct representing a single marker location.">MarkerPosition</a> is considered uninitialised. All unitialised values are moved to the end. The ordering of uninitialised values amongst themselves is undefined.</p>
<p>Reference: <a href="https://stackoverflow.com/questions/32948281/c-sort-two-arrays-the-same-way">https://stackoverflow.com/questions/32948281/c-sort-two-arrays-the-same-way</a> </p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a80cab739edbd12bbd2ad1f9b77b27548_icgraph.png" border="0" usemap="#sim-operations_8h_a80cab739edbd12bbd2ad1f9b77b27548_icgraph" alt=""/></div>
<map name="sim-operations_8h_a80cab739edbd12bbd2ad1f9b77b27548_icgraph" id="sim-operations_8h_a80cab739edbd12bbd2ad1f9b77b27548_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="576,5,736,32"/>
<area shape="rect" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="391,5,528,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,31,343,57"/>
</map>
</div>

</div>
</div>
<a id="ac711a312bddd30d394adda822a7c7679"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac711a312bddd30d394adda822a7c7679">&#9670;&nbsp;</a></span>add_dmatrices()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> add_dmatrices </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add the two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </p>
<p>The result is returned as a new <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>.</p>
<p>If the two matrices do not have the same dimensions, the matrix returned has it will throw an error.</p>
<p>The ordering of the parameters does not matter. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>pointer to the first <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Add this to b. </td></tr>
    <tr><td class="paramname">b</td><td>pointer to the second <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Add this to a. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a new <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the result of *a + *b. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_cgraph.png" border="0" usemap="#sim-operations_8h_ac711a312bddd30d394adda822a7c7679_cgraph" alt=""/></div>
<map name="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_cgraph" id="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_cgraph">
<area shape="rect" title="Add the two DecimalMatrix." alt="" coords="5,5,112,32"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="160,5,315,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_icgraph.png" border="0" usemap="#sim-operations_8h_ac711a312bddd30d394adda822a7c7679_icgraph" alt=""/></div>
<map name="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_icgraph" id="sim-operations_8h_ac711a312bddd30d394adda822a7c7679_icgraph">
<area shape="rect" title="Add the two DecimalMatrix." alt="" coords="652,169,759,195"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,169,604,195"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,219,361,246"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,270,352,297"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="a2c58c3fef64a20bca3191f343b947933"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2c58c3fef64a20bca3191f343b947933">&#9670;&nbsp;</a></span>add_to_dmatrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void add_to_dmatrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add the two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </p>
<p>The result is saved in the first <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>.</p>
<p>If the two matrices do not have the same dimensions, the matrix returned has it will throw an error.</p>
<p>The ordering of the parameters does not matter.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>pointer to the first <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Add this to b. </td></tr>
    <tr><td class="paramname">b</td><td>pointer to the second <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Add this to a. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a new <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the result of *a + *b. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a2c58c3fef64a20bca3191f343b947933_icgraph.png" border="0" usemap="#sim-operations_8h_a2c58c3fef64a20bca3191f343b947933_icgraph" alt=""/></div>
<map name="sim-operations_8h_a2c58c3fef64a20bca3191f343b947933_icgraph" id="sim-operations_8h_a2c58c3fef64a20bca3191f343b947933_icgraph">
<area shape="rect" title="Add the two DecimalMatrix." alt="" coords="616,31,727,57"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="405,23,568,65"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="216,5,348,32"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,56,357,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,49,159,90"/>
</map>
</div>

</div>
</div>
<a id="ae7797aeac935227749fc6eae09a02a96"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae7797aeac935227749fc6eae09a02a96">&#9670;&nbsp;</a></span>calculate_count_matrix_of_allele_for_ids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_count_matrix_of_allele_for_ids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>for_ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears for each genotype is a set of given genotype. </p>
<p>Returns the result as a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Useful for multiplying to effect matrix to calculate GEBVs.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the start of a linked list that contains the alleles of each genotype in <code>for_ids</code>. </td></tr>
    <tr><td class="paramname">for_ids</td><td>poinder to an array of ids of different genotypes. </td></tr>
    <tr><td class="paramname">n_ids</td><td>number of genotypes to calculate the counts of/length of <code>for_ids</code> </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> countaining the number of <code>allele</code> occurences at each row/marker for each column/genotype in <code>for_ids</code>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_cgraph.png" border="0" usemap="#sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_cgraph" alt=""/></div>
<map name="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_cgraph" id="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_cgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="5,23,161,65"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="209,5,364,32"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="229,56,345,83"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="412,56,587,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_icgraph.png" border="0" usemap="#sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_icgraph" alt=""/></div>
<map name="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_icgraph" id="sim-operations_8h_ae7797aeac935227749fc6eae09a02a96_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="616,57,772,98"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="405,23,568,65"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="413,89,560,130"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="216,5,348,32"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,56,357,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,49,159,90"/>
</map>
</div>

</div>
</div>
<a id="a238032359089625f3fd3b4447cdbbdff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a238032359089625f3fd3b4447cdbbdff">&#9670;&nbsp;</a></span>calculate_fitness_metric()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_fitness_metric </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the fitness metric/GEBV for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> struct. </p>
<p>The GEBV is calculated for each genotype by taking the sum of (number of copies of this allele at this marker times this allele's effect at this marker) for each marker for each different allele. To do this, the vector containing each allele's effect values is multiplied by a matrix containing the counts of that allele at each marker.</p>
<p>The function exits with error code 1 if no marker effect file is loaded.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object to which the genotypes belong. </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> that effect values have been loaded into. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the score for each individual in the group. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_cgraph.png" border="0" usemap="#sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_cgraph" alt=""/></div>
<map name="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_cgraph" id="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_cgraph">
<area shape="rect" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="5,144,168,171"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="232,5,339,32"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="403,85,557,112"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="217,107,353,149"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="232,173,339,200"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="220,224,351,251"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="216,275,355,301"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_icgraph.png" border="0" usemap="#sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_icgraph" alt=""/></div>
<map name="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_icgraph" id="sim-operations_8h_a238032359089625f3fd3b4447cdbbdff_icgraph">
<area shape="rect" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,169,604,195"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,219,361,246"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,270,352,297"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ac297afb7cf2c56f6f3f436090d7962dd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac297afb7cf2c56f6f3f436090d7962dd">&#9670;&nbsp;</a></span>calculate_fitness_metric_of_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_fitness_metric_of_group </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the fitness metric/GEBV for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> struct. </p>
<p>The GEBV is calculated for each genotype by taking the sum of (number of copies of this allele at this marker times this allele's effect at this marker) for each marker for each different allele. To do this, the vector containing each allele's effect values is multiplied by a matrix containing the counts of that allele at each marker.</p>
<p>The function exits with error code 1 if no marker effect file is loaded.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">group</td><td>calculate GEBVs for each genotype in the group with this group number. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the score for each individual in the group. </dd></dl>
<p>counts = sum; // use counts as a temp</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_cgraph.png" border="0" usemap="#sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_cgraph" alt=""/></div>
<map name="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_cgraph" id="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_cgraph">
<area shape="rect" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="5,179,168,221"/>
<area shape="rect" href="sim-operations_8c.html#a2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="239,5,349,32"/>
<area shape="rect" href="sim-operations_8c.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="216,209,372,250"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="420,107,575,133"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="241,157,347,184"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="243,275,345,301"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="442,325,553,352"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="229,107,359,133"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="225,376,363,403"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="439,216,555,243"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="623,216,797,243"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="455,275,540,301"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_icgraph.png" border="0" usemap="#sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_icgraph" alt=""/></div>
<map name="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_icgraph" id="sim-operations_8h_ac297afb7cf2c56f6f3f436090d7962dd_icgraph">
<area shape="rect" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="405,23,568,65"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="216,5,348,32"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,56,357,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,49,159,90"/>
</map>
</div>

</div>
</div>
<a id="a42f2398ccb53ffafce9b4a17e3222636"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42f2398ccb53ffafce9b4a17e3222636">&#9670;&nbsp;</a></span>calculate_full_count_matrix_of_allele()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> calculate_full_count_matrix_of_allele </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>. </p>
<p>Returns the result as a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. Useful for multiplying to effect matrix to calculate GEBVs.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that contains the genotypes to count alleles. </td></tr>
    <tr><td class="paramname">allele</td><td>the single-character allele to be counting. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> countaining the number of <code>allele</code> occurences at each row/marker for each column/genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_cgraph.png" border="0" usemap="#sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_cgraph" alt=""/></div>
<map name="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_cgraph" id="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_cgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="5,5,141,47"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="189,13,344,39"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_icgraph.png" border="0" usemap="#sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_icgraph" alt=""/></div>
<map name="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_icgraph" id="sim-operations_8h_a42f2398ccb53ffafce9b4a17e3222636_icgraph">
<area shape="rect" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="652,176,788,217"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,158,604,185"/>
<area shape="rect" href="sim-operations_8h.html#a490906a5bfdb22b90aba5cdf6282fd03" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="457,209,588,235"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,34,390,61"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,209,361,235"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,259,352,286"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="a4f233b73d31da864cdf803b52231e555"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4f233b73d31da864cdf803b52231e555">&#9670;&nbsp;</a></span>calculate_group_block_effects()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void calculate_group_block_effects </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>block_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>output_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ab569f6043f8d810aee8d8c5ba380abbd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab569f6043f8d810aee8d8c5ba380abbd">&#9670;&nbsp;</a></span>calculate_ideal_genotype()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* calculate_ideal_genotype </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </p>
<p>Returns this as a null-terminated heap array of characters of length d-&gt;n_markers + one null byte.</p>
<p>The return value should be freed when usage is finished!</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must be initialised with marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing markers and marker effects. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap array filled with a null-terminated string containing the highest scoring allele at each marker. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ab569f6043f8d810aee8d8c5ba380abbd_cgraph.png" border="0" usemap="#sim-operations_8h_ab569f6043f8d810aee8d8c5ba380abbd_cgraph" alt=""/></div>
<map name="sim-operations_8h_ab569f6043f8d810aee8d8c5ba380abbd_cgraph" id="sim-operations_8h_ab569f6043f8d810aee8d8c5ba380abbd_cgraph">
<area shape="rect" title="Takes a look at the currently&#45;loaded effect values and creates a string containing the allele with th..." alt="" coords="5,5,173,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="221,5,307,32"/>
</map>
</div>

</div>
</div>
<a id="aad9cd4248231da3682ba2f08a40cdbe8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad9cd4248231da3682ba2f08a40cdbe8">&#9670;&nbsp;</a></span>calculate_min_recombinations_fw1()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int* calculate_min_recombinations_fw1 </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>offspring</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured. </p>
<p>This function is a little lower-level (see the kinds of parameters required) and so a wrapper like calculate_recombinations_from_file is suggested for end users. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot...">calculate_recombinations_from_file()</a></dd></dl>
<p>The function reads start to end along each chromosome. At each marker, it checks if the alleles the offspring has could only have come from one parent/there is known parentage of that allele. If that is the case, it saves the provided id number of the source parent to the matching position in the result vector. If it is not the case, its behaviour depends on the <code>certain</code> parameter.</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct whose genetic map matches the provided genotypes. </td></tr>
    <tr><td class="paramname">parent1</td><td>a character vector containing one parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p1num</td><td>an integer that will be used to identify areas of the genome that come from the first parent in the returned vector. </td></tr>
    <tr><td class="paramname">parent2</td><td>a character vector containing the other parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p2num</td><td>an integer that will be used to identify areas of the genome that come from the second parent in the returned vector. </td></tr>
    <tr><td class="paramname">offspring</td><td>a character vector containing the alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> of the genotype whose likely recombinations we want to identify. </td></tr>
    <tr><td class="paramname">certain</td><td>a boolean. If TRUE, markers where the parent of origin cannot be identified will be set to 0, if FALSE, the value will be set to the id of the parent that provided the most recently identified allele in that chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap vector of length <code>d-&gt;n_markers</code> containing the id of the parent of origin at each marker in the <code>offspring</code> genotype. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_cgraph.png" border="0" usemap="#sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_cgraph" alt=""/></div>
<map name="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_cgraph" id="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_cgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="5,5,228,32"/>
<area shape="rect" href="sim-operations_8h.html#aa71cdbced90506820ef61bd842ecfa40" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="276,5,403,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_icgraph.png" border="0" usemap="#sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_icgraph" alt=""/></div>
<map name="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_icgraph" id="sim-operations_8h_aad9cd4248231da3682ba2f08a40cdbe8_icgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,13,445,39"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="a3627a380ed86ebebe937c953bc3b2339"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3627a380ed86ebebe937c953bc3b2339">&#9670;&nbsp;</a></span>calculate_min_recombinations_fwn()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int* calculate_min_recombinations_fwn </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p1num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>p2num</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>offspring</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>window_size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it. </p>
<p>This function is a little lower-level (see the kinds of parameters required) and so a wrapper like calculate_recombinations_from_file is suggested for end users. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot...">calculate_recombinations_from_file()</a></dd></dl>
<p>The function reads start to end along each chromosome. At each marker, it checks if the alleles the offspring has in the window centered at that marker could have come from one parent but could not have come from the other/there is known parentage of that allele. If that is the case, it saves the provided id number of the source parent to the matching position in the result vector. If it is not the case, its behaviour depends on the <code>certain</code> parameter.</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<p>Behaviour when the window size is not an odd integer has not been tested.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct whose genetic map matches the provided genotypes. </td></tr>
    <tr><td class="paramname">parent1</td><td>a character vector containing one parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p1num</td><td>an integer that will be used to identify areas of the genome that come from the first parent in the returned vector. </td></tr>
    <tr><td class="paramname">parent2</td><td>a character vector containing the other parent's alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </td></tr>
    <tr><td class="paramname">p2num</td><td>an integer that will be used to identify areas of the genome that come from the second parent in the returned vector. </td></tr>
    <tr><td class="paramname">offspring</td><td>a character vector containing the alleles at each marker in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> of the genotype whose likely recombinations we want to identify. </td></tr>
    <tr><td class="paramname">window_size</td><td>an odd integer representing the number of markers to check for known parentage around each marker </td></tr>
    <tr><td class="paramname">certain</td><td>a boolean. If TRUE, markers where the parent of origin cannot be identified will be set to 0, if FALSE, the value will be set to the id of the parent that provided the most recently identified allele in that chromosome. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap vector of length <code>d-&gt;n_markers</code> containing the id of the parent of origin at each marker in the <code>offspring</code> genotype. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_cgraph.png" border="0" usemap="#sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_cgraph" id="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_cgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="5,5,228,32"/>
<area shape="rect" href="sim-operations_8h.html#ab88223532880a22dfc966be4cdb4107b" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="276,5,451,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_icgraph.png" border="0" usemap="#sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_icgraph" alt=""/></div>
<map name="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_icgraph" id="sim-operations_8h_a3627a380ed86ebebe937c953bc3b2339_icgraph">
<area shape="rect" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,13,445,39"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="a7813f256407ab2ea242df32ea9ad2c7e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7813f256407ab2ea242df32ea9ad2c7e">&#9670;&nbsp;</a></span>calculate_recombinations_from_file()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int calculate_recombinations_from_file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>input_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>output_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>window_len</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>certain</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes. </p>
<p>The input file (which pairs up which targets and their parents the calculation should be carried out on) should have format:</p>
<p>[target name] [parent1name] [parent2name]</p>
<p>[target name] [parent1name] [parent2name]</p>
<p>...</p>
<p>The tab-separated output file produced by this function will have format: </p><pre class="fragment">[marker 1 name] [marker 2 name]...
</pre><p>[target name] [tab-separated recombination vector, containing the index at each marker of the parent the function guesses the target's alleles came from, or 0 if this is unknow]</p>
<p>...</p>
<p>Parents do not have to be directly identified as parents by the pedigree functionality of this library. A sample usage is performing a cross then multiple generations of selfing, then comparing the final inbred to the original two lines of the cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct containing the genotypes and map under consideration. </td></tr>
    <tr><td class="paramname">input_file</td><td>string containing the name of the file with the pairs of parents and offsprings of which to calculate recombinations </td></tr>
    <tr><td class="paramname">output_file</td><td>string containing the filename to which to save the results. </td></tr>
    <tr><td class="paramname">window_len</td><td>an odd integer representing the number of markers to check for known parentage around each marker </td></tr>
    <tr><td class="paramname">certain</td><td>TRUE to fill locations where parentage is unknown with 0, FALSE to fill locations where parentage is unknown with the most recent known parent </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 on success. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a7813f256407ab2ea242df32ea9ad2c7e_cgraph.png" border="0" usemap="#sim-operations_8h_a7813f256407ab2ea242df32ea9ad2c7e_cgraph" alt=""/></div>
<map name="sim-operations_8h_a7813f256407ab2ea242df32ea9ad2c7e_cgraph" id="sim-operations_8h_a7813f256407ab2ea242df32ea9ad2c7e_cgraph">
<area shape="rect" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,125,175,166"/>
<area shape="rect" href="sim-operations_8c.html#aad9cd4248231da3682ba2f08a40cdbe8" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,5,445,32"/>
<area shape="rect" href="sim-operations_8c.html#a3627a380ed86ebebe937c953bc3b2339" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,56,445,83"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="266,107,402,133"/>
<area shape="rect" href="sim-operations_8c.html#a309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="265,157,403,184"/>
<area shape="rect" href="sim-operations_8c.html#a67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="278,208,390,235"/>
<area shape="rect" href="sim-operations_8c.html#a93779d37ca9081d07987fd83c27b84af" title="Search for a genotype with a particular name in a linked list of AlleleMatrix, and return its index." alt="" coords="267,259,401,285"/>
<area shape="rect" href="sim-operations_8h.html#aa71cdbced90506820ef61bd842ecfa40" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="517,5,644,32"/>
<area shape="rect" href="sim-operations_8h.html#ab88223532880a22dfc966be4cdb4107b" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="493,56,668,83"/>
</map>
</div>

</div>
</div>
<a id="ade6b01bfc6450638a3a341b76e28a97d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ade6b01bfc6450638a3a341b76e28a97d">&#9670;&nbsp;</a></span>combine_groups()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int combine_groups </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>list_len</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_ids</em>[list_len]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Combine a set of groups into one group. </p>
<p>The function does so by setting the group membership of every genotype belonging to one of the groups to the same group number.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">list_len</td><td>the number of groups to be combined </td></tr>
    <tr><td class="paramname">group_ids</td><td>an array of group numbers containing the groups that are to be combined. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the new combined group. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ade6b01bfc6450638a3a341b76e28a97d_icgraph.png" border="0" usemap="#sim-operations_8h_ade6b01bfc6450638a3a341b76e28a97d_icgraph" alt=""/></div>
<map name="sim-operations_8h_ade6b01bfc6450638a3a341b76e28a97d_icgraph" id="sim-operations_8h_ade6b01bfc6450638a3a341b76e28a97d_icgraph">
<area shape="rect" title="Combine a set of groups into one group." alt="" coords="207,13,324,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="a887027349503be24b1a847c9b833aaac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a887027349503be24b1a847c9b833aaac">&#9670;&nbsp;</a></span>condense_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void condense_allele_matrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </p>
<p>Not intended to be called by an end user - functions which require it should be calling it already.</p>
<p>Ideally, we want all <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> structs in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>'s linked list to have no gaps. That is, if there are more than 1000 genotypes, the first AM should be full/contain 1000 genotypes, and so forth, and the AM at the end should have its n genotypes having indexes &lt; n (so all at the start of the AM with no gaps in-between).</p>
<p>This function achieves the cleanup by using two pointers: a filler out the front that identifies a genotype that needs to be shifted back/that occurs after a gap, and a filler that identifies each gap and copies the genotype at the filler back into it.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to operate. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a887027349503be24b1a847c9b833aaac_cgraph.png" border="0" usemap="#sim-operations_8h_a887027349503be24b1a847c9b833aaac_cgraph" alt=""/></div>
<map name="sim-operations_8h_a887027349503be24b1a847c9b833aaac_cgraph" id="sim-operations_8h_a887027349503be24b1a847c9b833aaac_cgraph">
<area shape="rect" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="5,5,164,32"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="212,5,349,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a887027349503be24b1a847c9b833aaac_icgraph.png" border="0" usemap="#sim-operations_8h_a887027349503be24b1a847c9b833aaac_icgraph" alt=""/></div>
<map name="sim-operations_8h_a887027349503be24b1a847c9b833aaac_icgraph" id="sim-operations_8h_a887027349503be24b1a847c9b833aaac_icgraph">
<area shape="rect" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="441,198,600,225"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#afecdc653eee11af9dbee476eca981a0b" title="Clears memory of all genotypes and associated information belonging to a particular group." alt="" coords="257,169,353,195"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="228,220,381,261"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,286,385,313"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,337,352,363"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="aac694dd044a29fd93443368a7b9e8d8a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aac694dd044a29fd93443368a7b9e8d8a">&#9670;&nbsp;</a></span>create_empty_allelematrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>* create_empty_allelematrix </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_subjects</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Creator for an empty <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object of a given size. </p>
<p>Includes memory allocation for <code>n_subjects</code> worth of <code>.alleles</code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_markers</td><td>number of rows/markers to create </td></tr>
    <tr><td class="paramname">n_subjects</td><td>number of individuals to create. This includes filling the first n_subjects entries of .alleles with heap char* of length n_markers, so that the alleles for these can be added without further memory allocation. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>pointer to the empty created <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_cgraph.png" border="0" usemap="#sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_cgraph" alt=""/></div>
<map name="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_cgraph" id="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_cgraph">
<area shape="rect" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="5,5,180,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="228,5,313,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_icgraph.png" border="0" usemap="#sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_icgraph" alt=""/></div>
<map name="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_icgraph" id="sim-operations_8h_aac694dd044a29fd93443368a7b9e8d8a_icgraph">
<area shape="rect" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="441,231,616,258"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,50,390,77"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="228,159,381,200"/>
<area shape="rect" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data." alt="" coords="218,224,391,265"/>
<area shape="rect" href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data." alt="" coords="225,289,384,331"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,355,385,382"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,406,352,433"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="27,297,146,323"/>
</map>
</div>

</div>
</div>
<a id="a3ea77c76c329625646419a10683669f9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3ea77c76c329625646419a10683669f9">&#9670;&nbsp;</a></span>cross_random_individuals()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int cross_random_individuals </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>from_group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_crosses</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs random crosses among members of a group. </p>
<p>If the group does not have at least two members, the simulation exits. Selfing/crossing an individual with itself is not permitted. The resulting genotypes are allocated to a new group.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross. The family_size parameter in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> allows you to repeat each particular randomly-chosen cross a certain number of times.</p>
<p>Parents are drawn uniformly from the group when picking which crosses to make.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that contains the genetic map and genotypes of the parent group. </td></tr>
    <tr><td class="paramname">from_group</td><td>group number from which to draw the parents. </td></tr>
    <tr><td class="paramname">n_crosses</td><td>number of random pairs of parents to cross. </td></tr>
    <tr><td class="paramname">g</td><td>options for the genotypes created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3ea77c76c329625646419a10683669f9_cgraph.png" border="0" usemap="#sim-operations_8h_a3ea77c76c329625646419a10683669f9_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3ea77c76c329625646419a10683669f9_cgraph" id="sim-operations_8h_a3ea77c76c329625646419a10683669f9_cgraph">
<area shape="rect" title="Performs random crosses among members of a group." alt="" coords="5,578,176,605"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="230,173,393,199"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="483,274,590,301"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="232,325,391,351"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="468,325,605,351"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="224,477,399,503"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="256,426,367,453"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="251,578,372,605"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="481,578,592,605"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="260,527,363,554"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="241,730,382,757"/>
<area shape="rect" href="sim-operations_8c.html#a89e066191f12e056cebb8d2cd74a2a29" title="Generate an integer randomly between 0 and limit inclusive." alt="" coords="279,781,343,807"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="247,831,375,858"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="265,882,358,909"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="248,933,375,959"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="244,679,379,706"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="483,223,590,250"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="683,122,837,149"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="469,5,605,47"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="471,71,602,98"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="467,122,606,149"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="494,527,579,554"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="447,375,627,402"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="509,426,564,453"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="505,477,569,503"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="467,730,606,757"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="675,730,845,757"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="703,919,817,946"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="479,970,594,997"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="893,945,1068,971"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="698,970,822,997"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="475,679,599,706"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3ea77c76c329625646419a10683669f9_icgraph.png" border="0" usemap="#sim-operations_8h_a3ea77c76c329625646419a10683669f9_icgraph" alt=""/></div>
<map name="sim-operations_8h_a3ea77c76c329625646419a10683669f9_icgraph" id="sim-operations_8h_a3ea77c76c329625646419a10683669f9_icgraph">
<area shape="rect" title="Performs random crosses among members of a group." alt="" coords="207,13,377,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="abb793988799d792dfe4007efdf988068"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abb793988799d792dfe4007efdf988068">&#9670;&nbsp;</a></span>cross_these_combinations()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int cross_these_combinations </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_combinations</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>combinations</em>[2][n_combinations], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs the crosses of pairs of parents whose ids are provided in an array. </p>
<p>The resulting genotypes are allocated to a new group.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross. The family_size parameter in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> allows you to repeat each particular cross a certain number of times.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that includes genetic map data and allele data needed to simulate crossing. </td></tr>
    <tr><td class="paramname">combinations</td><td>a 2D array of IDs, with the first dimension being parent 1 or 2, and the second being the IDs of those parents for each combination to cross. </td></tr>
    <tr><td class="paramname">n_combinations</td><td>the number of pairs of ids to cross/the length of <code>combinations</code> </td></tr>
    <tr><td class="paramname">g</td><td>options for the genotypes created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_abb793988799d792dfe4007efdf988068_cgraph.png" border="0" usemap="#sim-operations_8h_abb793988799d792dfe4007efdf988068_cgraph" alt=""/></div>
<map name="sim-operations_8h_abb793988799d792dfe4007efdf988068_cgraph" id="sim-operations_8h_abb793988799d792dfe4007efdf988068_cgraph">
<area shape="rect" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="5,505,183,532"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="237,151,399,177"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="490,252,597,279"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="239,353,397,380"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="475,303,612,329"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="231,455,405,481"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="263,404,373,431"/>
<area shape="rect" href="sim-operations_8c.html#a309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="249,556,387,583"/>
<area shape="rect" href="sim-operations_8c.html#a67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="262,607,374,633"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="247,657,389,684"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="254,708,382,735"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="271,759,365,785"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="255,809,381,836"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="251,505,385,532"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="490,33,597,60"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="689,63,844,89"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="475,85,611,126"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="478,151,609,177"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="474,201,613,228"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="501,505,586,532"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="453,353,633,380"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="516,404,571,431"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="511,455,575,481"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="474,657,613,684"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="681,657,852,684"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="710,835,823,861"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="486,784,601,811"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="900,809,1075,836"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="705,784,829,811"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="481,556,605,583"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_abb793988799d792dfe4007efdf988068_icgraph.png" border="0" usemap="#sim-operations_8h_abb793988799d792dfe4007efdf988068_icgraph" alt=""/></div>
<map name="sim-operations_8h_abb793988799d792dfe4007efdf988068_icgraph" id="sim-operations_8h_abb793988799d792dfe4007efdf988068_icgraph">
<area shape="rect" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,71,393,98"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,5,165,47"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,71,168,98"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,123,163,164"/>
</map>
</div>

</div>
</div>
<a id="add4a9c17e8c3e59fc00eee25635b110f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#add4a9c17e8c3e59fc00eee25635b110f">&#9670;&nbsp;</a></span>delete_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_allele_matrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deletes the full <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object and frees its memory. </p>
<p>m will now refer to an empty matrix, with pointers set to null and dimensions set to 0.</p>
<p>freeing its memory includes freeing the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, which was allocated on the heap by </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size.">create_empty_allelematrix()</a></dd></dl>
<p>all subsequent matrixes in the linked list are also deleted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the matrix whose data is to be cleared and memory freed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_add4a9c17e8c3e59fc00eee25635b110f_icgraph.png" border="0" usemap="#sim-operations_8h_add4a9c17e8c3e59fc00eee25635b110f_icgraph" alt=""/></div>
<map name="sim-operations_8h_add4a9c17e8c3e59fc00eee25635b110f_icgraph" id="sim-operations_8h_add4a9c17e8c3e59fc00eee25635b110f_icgraph">
<area shape="rect" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="648,250,785,277"/>
<area shape="rect" href="sim-operations_8h.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="441,225,600,251"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,57,390,83"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,275,385,302"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,326,352,353"/>
<area shape="rect" href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041" title="Deletes a SimData object and frees its memory." alt="" coords="465,377,576,403"/>
<area shape="rect" href="sim-operations_8h.html#afecdc653eee11af9dbee476eca981a0b" title="Clears memory of all genotypes and associated information belonging to a particular group." alt="" coords="257,225,353,251"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="228,159,381,200"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="245,377,364,403"/>
</map>
</div>

</div>
</div>
<a id="aaf6542d59034f88f9e1ec48e5f053c4a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaf6542d59034f88f9e1ec48e5f053c4a">&#9670;&nbsp;</a></span>delete_dmatrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_dmatrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deletes a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> and frees its memory. </p>
<p>m will now refer to an empty matrix, with every pointer set to null and dimensions set to 0.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the matrix whose data is to be cleared and memory freed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aaf6542d59034f88f9e1ec48e5f053c4a_icgraph.png" border="0" usemap="#sim-operations_8h_aaf6542d59034f88f9e1ec48e5f053c4a_icgraph" alt=""/></div>
<map name="sim-operations_8h_aaf6542d59034f88f9e1ec48e5f053c4a_icgraph" id="sim-operations_8h_aaf6542d59034f88f9e1ec48e5f053c4a_icgraph">
<area shape="rect" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="652,346,759,373"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,270,604,297"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,169,390,195"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,219,393,246"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,270,385,297"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,321,361,347"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,371,352,398"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="441,52,604,93"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="239,30,371,57"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="229,107,380,134"/>
<area shape="rect" href="sim-operations_8h.html#a961f4b0757752fee2bd08adde7ece8ad" title="Deletes an EffectMatrix object and frees its memory." alt="" coords="453,523,592,550"/>
<area shape="rect" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="454,473,591,499"/>
<area shape="rect" href="sim-operations_8h.html#a9d6b231bd6ddde82af66a5d276a2c739" title="Populates a SimData combination with effect values." alt="" coords="441,574,604,601"/>
<area shape="rect" href="sim-operations_8h.html#a490906a5bfdb22b90aba5cdf6282fd03" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="457,625,588,651"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="449,676,596,717"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,100,163,141"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,165,165,207"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,231,168,258"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,283,163,324"/>
<area shape="rect" href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041" title="Deletes a SimData object and frees its memory." alt="" coords="249,523,360,550"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="27,498,146,525"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="219,473,390,499"/>
</map>
</div>

</div>
</div>
<a id="a961f4b0757752fee2bd08adde7ece8ad"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a961f4b0757752fee2bd08adde7ece8ad">&#9670;&nbsp;</a></span>delete_effect_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_effect_matrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deletes an <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> object and frees its memory. </p>
<p>m will now refer to an empty matrix, with every pointer set to null and dimensions set to 0.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the matrix whose data is to be cleared and memory freed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_cgraph.png" border="0" usemap="#sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_cgraph" alt=""/></div>
<map name="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_cgraph" id="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_cgraph">
<area shape="rect" title="Deletes an EffectMatrix object and frees its memory." alt="" coords="5,5,144,32"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="192,5,299,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_icgraph.png" border="0" usemap="#sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_icgraph" alt=""/></div>
<map name="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_icgraph" id="sim-operations_8h_a961f4b0757752fee2bd08adde7ece8ad_icgraph">
<area shape="rect" title="Deletes an EffectMatrix object and frees its memory." alt="" coords="331,5,469,32"/>
<area shape="rect" href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041" title="Deletes a SimData object and frees its memory." alt="" coords="172,5,283,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
</map>
</div>

</div>
</div>
<a id="aaf7e12b92ec0f93bb3bc2ec67155bc94"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaf7e12b92ec0f93bb3bc2ec67155bc94">&#9670;&nbsp;</a></span>delete_genmap()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_genmap </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_genetic_map.html">GeneticMap</a> *&#160;</td>
          <td class="paramname"><em>m</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deletes a <a class="el" href="struct_genetic_map.html" title="A type that stores the genetic map for a set of markers.">GeneticMap</a> object and frees its memory. </p>
<p>m will now refer to an empty matrix, with every pointer set to null and dimensions set to 0.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the matrix whose data is to be cleared and memory freed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aaf7e12b92ec0f93bb3bc2ec67155bc94_icgraph.png" border="0" usemap="#sim-operations_8h_aaf7e12b92ec0f93bb3bc2ec67155bc94_icgraph" alt=""/></div>
<map name="sim-operations_8h_aaf7e12b92ec0f93bb3bc2ec67155bc94_icgraph" id="sim-operations_8h_aaf7e12b92ec0f93bb3bc2ec67155bc94_icgraph">
<area shape="rect" title="Deletes a GeneticMap object and frees its memory." alt="" coords="576,56,687,83"/>
<area shape="rect" href="sim-operations_8h.html#a6fc4a687a2186e9e9dcf3229d9141041" title="Deletes a SimData object and frees its memory." alt="" coords="202,5,313,32"/>
<area shape="rect" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="391,56,528,83"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,107,343,133"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,56,124,83"/>
</map>
</div>

</div>
</div>
<a id="afecdc653eee11af9dbee476eca981a0b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afecdc653eee11af9dbee476eca981a0b">&#9670;&nbsp;</a></span>delete_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_group </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Clears memory of all genotypes and associated information belonging to a particular group. </p>
<p>Uses a call to <a class="el" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S...">condense_allele_matrix()</a> to ensure that the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> remains valid after deletion.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the subset of data to be cleared </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_afecdc653eee11af9dbee476eca981a0b_cgraph.png" border="0" usemap="#sim-operations_8h_afecdc653eee11af9dbee476eca981a0b_cgraph" alt=""/></div>
<map name="sim-operations_8h_afecdc653eee11af9dbee476eca981a0b_cgraph" id="sim-operations_8h_afecdc653eee11af9dbee476eca981a0b_cgraph">
<area shape="rect" title="Clears memory of all genotypes and associated information belonging to a particular group." alt="" coords="5,5,101,32"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="149,5,308,32"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="356,5,493,32"/>
</map>
</div>

</div>
</div>
<a id="a6fc4a687a2186e9e9dcf3229d9141041"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6fc4a687a2186e9e9dcf3229d9141041">&#9670;&nbsp;</a></span>delete_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void delete_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>m</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Deletes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object and frees its memory. </p>
<p>m will now refer to an empty matrix, with every pointer set to null and dimensions set to 0.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the matrix whose data is to be cleared and memory freed. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_cgraph.png" border="0" usemap="#sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_cgraph" alt=""/></div>
<map name="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_cgraph" id="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_cgraph">
<area shape="rect" title="Deletes a SimData object and frees its memory." alt="" coords="5,56,116,83"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="165,5,302,32"/>
<area shape="rect" href="sim-operations_8c.html#a961f4b0757752fee2bd08adde7ece8ad" title="Deletes an EffectMatrix object and frees its memory." alt="" coords="164,56,303,83"/>
<area shape="rect" href="sim-operations_8c.html#aaf7e12b92ec0f93bb3bc2ec67155bc94" title="Deletes a GeneticMap object and frees its memory." alt="" coords="178,107,289,133"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="351,56,457,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_icgraph.png" border="0" usemap="#sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_icgraph" alt=""/></div>
<map name="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_icgraph" id="sim-operations_8h_a6fc4a687a2186e9e9dcf3229d9141041_icgraph">
<area shape="rect" title="Deletes a SimData object and frees its memory." alt="" coords="172,5,283,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
</map>
</div>

</div>
</div>
<a id="aa18b10bcd660dee5213868cc3244f826"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa18b10bcd660dee5213868cc3244f826">&#9670;&nbsp;</a></span>generate_cross()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void generate_cross </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent1_genome</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent2_genome</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the alleles of the outcome of crossing two subjects. </p>
<p>Gametes are generated at the same time but are independent.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle...">generate_gamete()</a>, on which this is based.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that includes genetic map data needed to simulate meiosis and the value of n_markers </td></tr>
    <tr><td class="paramname">parent1_genome</td><td>a 2x(n_markers) array of characters containing the alleles of the first parent </td></tr>
    <tr><td class="paramname">parent2_genome</td><td>a 2x(n_markers) array of characters containing the alleles of the second parent. </td></tr>
    <tr><td class="paramname">output</td><td>a 2x(n_marker) array of chars which will be overwritten with the offspring genome. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_cgraph.png" border="0" usemap="#sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_cgraph" alt=""/></div>
<map name="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_cgraph" id="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_cgraph">
<area shape="rect" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="5,81,116,108"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="164,5,344,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="211,56,297,83"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="227,107,281,133"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="222,157,286,184"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_icgraph.png" border="0" usemap="#sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_icgraph" alt=""/></div>
<map name="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_icgraph" id="sim-operations_8h_aa18b10bcd660dee5213868cc3244f826_icgraph">
<area shape="rect" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="441,118,552,145"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,169,352,195"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ac1f46b3de0afb4674ae6d34311e7aea7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac1f46b3de0afb4674ae6d34311e7aea7">&#9670;&nbsp;</a></span>generate_doubled_haploid()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void generate_doubled_haploid </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent_genome</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent. </p>
<p>One gamete is generated, then doubled. The output will be perfectly homozygous.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle...">generate_gamete()</a>, on which this is based.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that includes genetic map data needed to simulate meiosis and the value of n_markers </td></tr>
    <tr><td class="paramname">parent_genome</td><td>a 2x(n_markers) array of characters containing the alleles of the first parent </td></tr>
    <tr><td class="paramname">output</td><td>a 2x(n_marker) array of chars which will be overwritten with the offspring genome. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_cgraph.png" border="0" usemap="#sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_cgraph" alt=""/></div>
<map name="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_cgraph" id="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_cgraph">
<area shape="rect" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="5,81,177,108"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="225,5,405,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="273,56,358,83"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="288,107,343,133"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="283,157,347,184"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_icgraph.png" border="0" usemap="#sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_icgraph" alt=""/></div>
<map name="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_icgraph" id="sim-operations_8h_ac1f46b3de0afb4674ae6d34311e7aea7_icgraph">
<area shape="rect" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="215,5,387,32"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="5,5,167,32"/>
</map>
</div>

</div>
</div>
<a id="ae406f4a0266b7480e49899693f1b00da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae406f4a0266b7480e49899693f1b00da">&#9670;&nbsp;</a></span>generate_gamete()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void generate_gamete </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>parent_genome</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alleles of a given parent. </p>
<p>It generates the number of crossover events in each chromosome by drawing from a Poisson distribution with parameter corresponding to the length of the chromosome (in Morgans.).</p>
<p>It generates the positions of those crossover events from a uniform distribution.</p>
<p>It picks a random one of the gametes generated by picking randomly which column of the parent's alleles to start with. When crossover events occur it starts reading the other column.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object containing map positions for the markers that make up the rows of <code>parent_table</code>. sort_markers() and locate_chromosomes() should have been called previously. </td></tr>
    <tr><td class="paramname">parent_genome</td><td>the char* containing the parent's genome as a character string made up of sequential pairs of alleles for each marker in d-&gt;markers. </td></tr>
    <tr><td class="paramname">output</td><td>the char* to which to save the gamete. It saves the alleles every second character, starting at 0, so that calling generate_gamete(..., offspring_genome) &amp; generate_gamete(..., offspring_genome + 1) can be used to generate both halves of its genome. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ae406f4a0266b7480e49899693f1b00da_cgraph.png" border="0" usemap="#sim-operations_8h_ae406f4a0266b7480e49899693f1b00da_cgraph" alt=""/></div>
<map name="sim-operations_8h_ae406f4a0266b7480e49899693f1b00da_cgraph" id="sim-operations_8h_ae406f4a0266b7480e49899693f1b00da_cgraph">
<area shape="rect" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="5,81,128,108"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="176,5,356,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="223,56,309,83"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="239,107,293,133"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="234,157,298,184"/>
</map>
</div>

</div>
</div>
<a id="abf3f578b0aa59d5b68ccd0738dd6a9d1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abf3f578b0aa59d5b68ccd0738dd6a9d1">&#9670;&nbsp;</a></span>generate_zero_dmatrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> generate_zero_dmatrix </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>r</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>c</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Generates a matrix of c columns, r rows with all 0. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">r</td><td>the number of rows/first index for the new matrix. </td></tr>
    <tr><td class="paramname">c</td><td>the number of columns/second index for the new matrix. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> with r rows, c cols, and a matrix of the correct size, with all values zeroed. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_abf3f578b0aa59d5b68ccd0738dd6a9d1_icgraph.png" border="0" usemap="#sim-operations_8h_abf3f578b0aa59d5b68ccd0738dd6a9d1_icgraph" alt=""/></div>
<map name="sim-operations_8h_abf3f578b0aa59d5b68ccd0738dd6a9d1_icgraph" id="sim-operations_8h_abf3f578b0aa59d5b68ccd0738dd6a9d1_icgraph">
<area shape="rect" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="864,228,1019,255"/>
<area shape="rect" href="sim-operations_8h.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="685,60,791,87"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="445,115,608,141"/>
<area shape="rect" href="sim-operations_8h.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="660,330,816,371"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="445,265,608,306"/>
<area shape="rect" href="sim-operations_8h.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="670,162,806,203"/>
<area shape="rect" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="669,396,807,423"/>
<area shape="rect" href="sim-operations_8h.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="673,228,803,255"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,115,390,141"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,64,393,91"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,165,385,192"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,216,361,243"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,13,352,40"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,249,163,290"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,122,165,163"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,5,168,32"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,57,163,98"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="453,330,600,371"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="239,267,371,293"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="229,317,380,344"/>
<area shape="rect" href="sim-operations_8h.html#a490906a5bfdb22b90aba5cdf6282fd03" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="461,169,592,196"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="245,421,364,448"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="441,447,612,473"/>
</map>
</div>

</div>
</div>
<a id="a5e049945be5700a6d5a5654b6b332842"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5e049945be5700a6d5a5654b6b332842">&#9670;&nbsp;</a></span>get_chromosome_locations()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void get_chromosome_locations </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Updates the chr_ends, n_chr and chr_lengths fields in <a class="el" href="struct_sim_data.html#ae1ae6d192937ca2edaba69edf8308711">SimData.map</a>. </p>
<p>This should only be run on a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> that has already been ordered. </p><dl class="section see"><dt>See also</dt><dd>sort_markers()</dd></dl>
<p>The function loops over all <a class="el" href="struct_marker_position.html" title="A struct representing a single marker location.">MarkerPosition</a> in SimData.map.positions twice, for flexible and minimal memory usage rather than maximum speed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object for which the fields under <code>map</code> need to be initialised or updated. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_cgraph.png" border="0" usemap="#sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_cgraph" alt=""/></div>
<map name="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_cgraph" id="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_cgraph">
<area shape="rect" title="Updates the chr_ends, n_chr and chr_lengths fields in SimData.map." alt="" coords="5,5,187,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="235,5,320,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_icgraph.png" border="0" usemap="#sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_icgraph" alt=""/></div>
<map name="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_icgraph" id="sim-operations_8h_a5e049945be5700a6d5a5654b6b332842_icgraph">
<area shape="rect" title="Updates the chr_ends, n_chr and chr_lengths fields in SimData.map." alt="" coords="391,5,572,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,31,343,57"/>
</map>
</div>

</div>
</div>
<a id="a3305ddcc29138d2d92689f39a7885ca0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3305ddcc29138d2d92689f39a7885ca0">&#9670;&nbsp;</a></span>get_existing_groups()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int* get_existing_groups </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>n_groups</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Identify every group number that currently has members. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">n_groups</td><td>a pointer to a location that can be accessed by the calling function. The number of groups/length of the returned array will be saved here. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a heap array of length [value saved at n_groups after this function has run] containing the group numbers that appear in the group memberships in <code>d</code>, sorted in ascending order. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_cgraph.png" border="0" usemap="#sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_cgraph" id="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_cgraph">
<area shape="rect" title="Identify every group number that currently has members." alt="" coords="5,5,144,32"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="192,5,363,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_icgraph.png" border="0" usemap="#sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_icgraph" alt=""/></div>
<map name="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_icgraph" id="sim-operations_8h_a3305ddcc29138d2d92689f39a7885ca0_icgraph">
<area shape="rect" title="Identify every group number that currently has members." alt="" coords="840,177,979,204"/>
<area shape="rect" href="sim-operations_8h.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="651,152,792,179"/>
<area shape="rect" href="sim-operations_8h.html#ad5a61d77e05e6bb0d8f4235d21b8b1f3" title="Give every individual in the group a new group number that does not belong to any other existing grou..." alt="" coords="651,203,791,229"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="207,5,377,32"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="425,64,603,91"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="437,115,591,157"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="433,181,595,208"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="467,232,561,259"/>
<area shape="rect" href="sim-operations_8h.html#af2eda46d016c384b166ce65605952b7c" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="455,283,573,309"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,137,159,178"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="213,57,371,98"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="211,123,373,149"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="216,174,368,215"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="217,261,367,288"/>
</map>
</div>

</div>
</div>
<a id="a392c23248f22e9faf085198dbb1d8711"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a392c23248f22e9faf085198dbb1d8711">&#9670;&nbsp;</a></span>get_file_dimensions()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">struct <a class="el" href="struct_table_size.html">TableSize</a> get_file_dimensions </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>sep</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Opens a table file and reads the number of columns and rows (including headers) separated by <code>sep</code> into a <a class="el" href="struct_table_size.html" title="A simple struct used for returning the dimensions of a matrix or table.">TableSize</a> struct that is returned. </p>
<p>If the file fails to open, the simulation exits.</p>
<p>Rows must be either empty or have same number of columns as the first. Empty rows are not counted.</p>
<p>If a row with an invalid number of columns is found, the number of columns in the return value is set to 0. number of rows in the return value in this case is arbitrary.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>the path/name to the table file whose dimensions we want </td></tr>
    <tr><td class="paramname">sep</td><td>the character that separates columns in the file eg tab </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><a class="el" href="struct_table_size.html" title="A simple struct used for returning the dimensions of a matrix or table.">TableSize</a> struct with .num_columns and .num_rows filled. These counts include header rows/columns and exclude blank rows. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a392c23248f22e9faf085198dbb1d8711_icgraph.png" border="0" usemap="#sim-operations_8h_a392c23248f22e9faf085198dbb1d8711_icgraph" alt=""/></div>
<map name="sim-operations_8h_a392c23248f22e9faf085198dbb1d8711_icgraph" id="sim-operations_8h_a392c23248f22e9faf085198dbb1d8711_icgraph">
<area shape="rect" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="393,209,529,235"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="183,5,334,47"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="174,71,343,112"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="182,136,335,177"/>
<area shape="rect" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data." alt="" coords="172,201,345,243"/>
<area shape="rect" href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data." alt="" coords="179,267,338,308"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="177,333,340,359"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="183,384,335,425"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,274,124,301"/>
</map>
</div>

</div>
</div>
<a id="aa6943b0633c1744bacad62ab2bf68ebd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa6943b0633c1744bacad62ab2bf68ebd">&#9670;&nbsp;</a></span>get_from_ordered_uint_list()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_from_ordered_uint_list </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>target</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>list_len</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the located index in an array of integers where the integer is <code>target</code>. </p>
<p>Returns -1 if no match was found. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar...">get_from_unordered_str_list()</a></dd></dl>
<p>The list is assumed to be sorted in ascending order.</p>
<p>It uses the bisection method to search the list.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">target</td><td>the integer to be located </td></tr>
    <tr><td class="paramname">list</td><td>the array of integers to search </td></tr>
    <tr><td class="paramname">list_len</td><td>length of the array of integers to search </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Index in <code>list</code> where we find the same integer as <code>target</code>, or -1 if no match is found. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa6943b0633c1744bacad62ab2bf68ebd_icgraph.png" border="0" usemap="#sim-operations_8h_aa6943b0633c1744bacad62ab2bf68ebd_icgraph" alt=""/></div>
<map name="sim-operations_8h_aa6943b0633c1744bacad62ab2bf68ebd_icgraph" id="sim-operations_8h_aa6943b0633c1744bacad62ab2bf68ebd_icgraph">
<area shape="rect" title="Returns the located index in an array of integers where the integer is target." alt="" coords="1040,240,1215,267"/>
<area shape="rect" href="sim-operations_8h.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="872,147,988,173"/>
<area shape="rect" href="sim-operations_8h.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="873,240,987,267"/>
<area shape="rect" href="sim-operations_8h.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="868,292,992,319"/>
<area shape="rect" href="sim-operations_8h.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="660,97,816,138"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="443,38,606,79"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="451,103,598,145"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="239,5,371,32"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="229,56,380,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,89,163,130"/>
<area shape="rect" href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="461,175,588,201"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="665,289,811,330"/>
<area shape="rect" href="sim-operations_8h.html#a6b87b484064fd373aec0bfa4ce8dbef1" title="Print the parents of each genotype in the SimData to a file." alt="" coords="656,355,820,381"/>
<area shape="rect" href="sim-operations_8h.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="681,237,795,264"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,208,390,235"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,259,393,285"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,107,385,133"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,157,352,184"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,193,165,234"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,259,168,285"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,310,163,351"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="441,237,608,264"/>
</map>
</div>

</div>
</div>
<a id="aed2ae969c3ce4ed1466a885dcc559472"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aed2ae969c3ce4ed1466a885dcc559472">&#9670;&nbsp;</a></span>get_from_unordered_str_list()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_from_unordered_str_list </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>target</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>list_len</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the first located index in an array of strings where the string is the same as the string <code>target</code>. </p>
<p>Returns -1 if no match was found. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target.">get_from_ordered_uint_list()</a></dd></dl>
<p>The list of strings is not assumed to be sorted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">target</td><td>the string to be located </td></tr>
    <tr><td class="paramname">list</td><td>the array of strings to search </td></tr>
    <tr><td class="paramname">list_len</td><td>length of the array of strings to search </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Index in <code>list</code> where we find the same string as <code>target</code>, or -1 if no match is found. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aed2ae969c3ce4ed1466a885dcc559472_icgraph.png" border="0" usemap="#sim-operations_8h_aed2ae969c3ce4ed1466a885dcc559472_icgraph" alt=""/></div>
<map name="sim-operations_8h_aed2ae969c3ce4ed1466a885dcc559472_icgraph" id="sim-operations_8h_aed2ae969c3ce4ed1466a885dcc559472_icgraph">
<area shape="rect" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="391,89,527,131"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="182,5,333,47"/>
<area shape="rect" href="sim-operations_8h.html#a9d6b231bd6ddde82af66a5d276a2c739" title="Populates a SimData combination with effect values." alt="" coords="176,71,339,98"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,122,343,149"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="181,173,334,215"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,97,124,123"/>
</map>
</div>

</div>
</div>
<a id="af49ac14d9ff4483bce6424f823d30ca0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af49ac14d9ff4483bce6424f823d30ca0">&#9670;&nbsp;</a></span>get_genes_of_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* get_genes_of_id </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the alleles at each marker of the genotype with a given id. </p>
<p>The function uses a bisection search on the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> where it should be found. Searching by id is therefore fast. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target.">get_from_ordered_uint_list()</a></dd></dl>
<p>This function assumes that ids are never reshuffled in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. As of 2021.04.12 this condition is always true, but this function will need to be revisited if this is no longer the case.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">id</td><td>the id of the genotype whose alleles are sought </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the alleles of the genotype that has id <code>id</code>, as a copy of the pointer to the heap memory where the genotype is saved (so <em>don't</em> free the pointer returned from this function). It points to a sequence of characters, ordered according to the markers in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to which the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> belongs. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_cgraph.png" border="0" usemap="#sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_cgraph" alt=""/></div>
<map name="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_cgraph" id="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_cgraph">
<area shape="rect" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="5,5,121,32"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="169,5,344,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_icgraph.png" border="0" usemap="#sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_icgraph" alt=""/></div>
<map name="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_icgraph" id="sim-operations_8h_af49ac14d9ff4483bce6424f823d30ca0_icgraph">
<area shape="rect" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="820,64,936,91"/>
<area shape="rect" href="sim-operations_8h.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="616,57,772,98"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="405,23,568,65"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="413,89,560,130"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="216,5,348,32"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,56,357,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,49,159,90"/>
</map>
</div>

</div>
</div>
<a id="a309dbf26800ee65e982df2b347980906"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a309dbf26800ee65e982df2b347980906">&#9670;&nbsp;</a></span>get_genes_of_index()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* get_genes_of_index </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the alleles of a genotype by its index. </p>
<p>The index is assumed to be 0-based, starting at the first entry of <code>start</code> and continuing through the linked list to which <code>start</code> belongs. Exits if the linked list does not contain that index.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which to locate the id at a particular index. </td></tr>
    <tr><td class="paramname">index</td><td>the index of the target. It is assumed to start at 0 at the start of the matrix in <code>start</code> and be incremented up until the last entry in the matrix of the last entry in the linked list. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the alleles of the genotype that has idnex <code>index</code>, as a copy of the pointer to the heap memory where the genotype is saved (so <em>don't</em> free the pointer returned from this function). It points to a sequence of characters, ordered according to the markers in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to which the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> belongs. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a309dbf26800ee65e982df2b347980906_icgraph.png" border="0" usemap="#sim-operations_8h_a309dbf26800ee65e982df2b347980906_icgraph" alt=""/></div>
<map name="sim-operations_8h_a309dbf26800ee65e982df2b347980906_icgraph" id="sim-operations_8h_a309dbf26800ee65e982df2b347980906_icgraph">
<area shape="rect" title="Get the alleles of a genotype by its index." alt="" coords="441,42,579,69"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="220,5,389,47"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,71,393,98"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,5,165,47"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,71,168,98"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,123,163,164"/>
</map>
</div>

</div>
</div>
<a id="ac159a31261e14b325bd623899eefa796"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac159a31261e14b325bd623899eefa796">&#9670;&nbsp;</a></span>get_group_genes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char** get_group_genes </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets a shallow copy of the genes/alleles of each member of the group. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group.">get_group_size()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group.">get_group_names()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group.">get_group_ids()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0-based, from the start of the linked list in the SimData) of each member of the gr...">get_group_indexes()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group we want data from </td></tr>
    <tr><td class="paramname">group_size</td><td>if group_size has already been calculated, pass it too, otherwise put in -1. This enables fewer calls to get_group_size when using multiple group-data-getter functions. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a vector containing pointers to the genes of each member of the group. the vector itself is on the heap and should be freed, but its contents are only shallow copies so should not be freed. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac159a31261e14b325bd623899eefa796_cgraph.png" border="0" usemap="#sim-operations_8h_ac159a31261e14b325bd623899eefa796_cgraph" alt=""/></div>
<map name="sim-operations_8h_ac159a31261e14b325bd623899eefa796_cgraph" id="sim-operations_8h_ac159a31261e14b325bd623899eefa796_cgraph">
<area shape="rect" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="5,31,127,57"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="175,5,285,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="187,56,273,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac159a31261e14b325bd623899eefa796_icgraph.png" border="0" usemap="#sim-operations_8h_ac159a31261e14b325bd623899eefa796_icgraph" alt=""/></div>
<map name="sim-operations_8h_ac159a31261e14b325bd623899eefa796_icgraph" id="sim-operations_8h_ac159a31261e14b325bd623899eefa796_icgraph">
<area shape="rect" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="425,147,547,174"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="217,5,367,47"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="207,71,377,98"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="211,122,373,149"/>
<area shape="rect" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="227,173,357,199"/>
<area shape="rect" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="213,224,371,265"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="245,290,339,317"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,64,159,105"/>
</map>
</div>

</div>
</div>
<a id="a961042e468630f31a3607dac42f8fa34"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a961042e468630f31a3607dac42f8fa34">&#9670;&nbsp;</a></span>get_group_ids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int* get_group_ids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the ids of each member of the group. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group.">get_group_size()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group.">get_group_genes()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group.">get_group_names()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0-based, from the start of the linked list in the SimData) of each member of the gr...">get_group_indexes()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group you want data from </td></tr>
    <tr><td class="paramname">group_size</td><td>if group_size has already been calculated, pass it too, otherwise put in -1. This enables fewer calls to get_group_size when using multiple group-data-getter functions. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a vector containing the ids of each memmber of the group. the vector itself is on the heap and should be freed. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_cgraph.png" border="0" usemap="#sim-operations_8h_a961042e468630f31a3607dac42f8fa34_cgraph" alt=""/></div>
<map name="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_cgraph" id="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_cgraph">
<area shape="rect" title="Gets the ids of each member of the group." alt="" coords="5,31,108,57"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="156,5,267,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="169,56,254,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_icgraph.png" border="0" usemap="#sim-operations_8h_a961042e468630f31a3607dac42f8fa34_icgraph" alt=""/></div>
<map name="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_icgraph" id="sim-operations_8h_a961042e468630f31a3607dac42f8fa34_icgraph">
<area shape="rect" title="Gets the ids of each member of the group." alt="" coords="640,263,743,289"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="427,58,590,99"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="226,12,358,39"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="207,120,377,147"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="428,175,589,201"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="435,226,582,267"/>
<area shape="rect" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="443,292,574,319"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="425,343,592,369"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="435,394,582,435"/>
<area shape="rect" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="429,459,588,501"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="461,525,556,552"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="217,68,367,95"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,87,159,129"/>
</map>
</div>

</div>
</div>
<a id="a95aad1344c24e6072e6a50dcc67b60a8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a95aad1344c24e6072e6a50dcc67b60a8">&#9670;&nbsp;</a></span>get_group_indexes()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int* get_group_indexes </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets the indexes (0-based, from the start of the linked list in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>) of each member of the group. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group.">get_group_size()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group.">get_group_genes()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group.">get_group_names()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group.">get_group_ids()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group you want data from </td></tr>
    <tr><td class="paramname">group_size</td><td>if group_size has already been calculated, pass it too, otherwise put in -1. This enables fewer calls to get_group_size when using multiple group-data-getter functions. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a vector containing the indexes of each member of the group. the vector itself is on the heap and should be freed. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_cgraph.png" border="0" usemap="#sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_cgraph" alt=""/></div>
<map name="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_cgraph" id="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_cgraph">
<area shape="rect" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="5,31,136,57"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="184,5,295,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="197,56,282,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_icgraph.png" border="0" usemap="#sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_icgraph" alt=""/></div>
<map name="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_icgraph" id="sim-operations_8h_a95aad1344c24e6072e6a50dcc67b60a8_icgraph">
<area shape="rect" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="412,42,543,69"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="207,5,364,47"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="210,71,361,98"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,64,159,105"/>
</map>
</div>

</div>
</div>
<a id="a3e00f06acc41eb012e99ae46d7a18ccc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3e00f06acc41eb012e99ae46d7a18ccc">&#9670;&nbsp;</a></span>get_group_names()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char** get_group_names </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets a shallow copy of the names of each member of the group. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group.">get_group_size()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group.">get_group_genes()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group.">get_group_ids()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0-based, from the start of the linked list in the SimData) of each member of the gr...">get_group_indexes()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group you want data from </td></tr>
    <tr><td class="paramname">group_size</td><td>if group_size has already been calculated, pass it too, otherwise put in -1. This enables fewer calls to get_group_size when using multiple group-data-getter functions. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a vector containing pointers to the names of each member of the group. the vector itself is on the heap and should be freed, but its contents are only shallow copies so should not be freed. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_cgraph.png" border="0" usemap="#sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_cgraph" id="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_cgraph">
<area shape="rect" title="Gets a shallow copy of the names of each member of the group." alt="" coords="5,31,131,57"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="179,5,289,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="191,56,277,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_icgraph.png" border="0" usemap="#sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_icgraph" alt=""/></div>
<map name="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_icgraph" id="sim-operations_8h_a3e00f06acc41eb012e99ae46d7a18ccc_icgraph">
<area shape="rect" title="Gets a shallow copy of the names of each member of the group." alt="" coords="220,187,345,214"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="13,5,164,47"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="15,71,162,112"/>
<area shape="rect" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="23,137,154,163"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="23,187,155,214"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="5,238,172,265"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="15,289,162,331"/>
<area shape="rect" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="9,355,168,396"/>
</map>
</div>

</div>
</div>
<a id="ad6d2b66296cd80f5bfe303e06de7935c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad6d2b66296cd80f5bfe303e06de7935c">&#9670;&nbsp;</a></span>get_group_size()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_group_size </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to count the number of genotypes that currently belong to the specified group. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group.">get_group_genes()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group.">get_group_names()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group.">get_group_ids()</a> </dd>
<dd>
<a class="el" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0-based, from the start of the linked list in the SimData) of each member of the gr...">get_group_indexes()</a></dd></dl>
<p>This goes through and checks every genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>, because there is currently no centralised tracking of groups.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group whose members need to be counted. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the number of genotypes currently belonging to this group. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad6d2b66296cd80f5bfe303e06de7935c_icgraph.png" border="0" usemap="#sim-operations_8h_ad6d2b66296cd80f5bfe303e06de7935c_icgraph" alt=""/></div>
<map name="sim-operations_8h_ad6d2b66296cd80f5bfe303e06de7935c_icgraph" id="sim-operations_8h_ad6d2b66296cd80f5bfe303e06de7935c_icgraph">
<area shape="rect" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="803,455,913,481"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="409,173,572,214"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="216,297,348,324"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,111,159,153"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="415,743,566,785"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="405,692,576,719"/>
<area shape="rect" href="sim-operations_8h.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="629,708,750,735"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="410,860,571,887"/>
<area shape="rect" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="425,575,556,601"/>
<area shape="rect" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="411,626,570,667"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="443,809,538,836"/>
<area shape="rect" href="sim-operations_8h.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="638,505,741,532"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="417,458,564,499"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="407,407,574,433"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="417,290,564,331"/>
<area shape="rect" href="sim-operations_8h.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="624,87,755,113"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="412,107,569,149"/>
<area shape="rect" href="sim-operations_8h.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="627,455,752,481"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,119,357,145"/>
</map>
</div>

</div>
</div>
<a id="aadec6c1a0fe783461a72ceaee3ccefad"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aadec6c1a0fe783461a72ceaee3ccefad">&#9670;&nbsp;</a></span>get_id_of_child()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int get_id_of_child </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>parent1id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>parent2id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its id. </p>
<p>Exits if such a child cannot be found.</p>
<p>This function must check every genotype in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the child is assumed to be found. </td></tr>
    <tr><td class="paramname">parent1id</td><td>one of the parents of the genotype must have this id. </td></tr>
    <tr><td class="paramname">parent2id</td><td>the other parent of the genotype must have this id. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the id of the first sequentially located genotype whose parents match the two parent ids provided </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aadec6c1a0fe783461a72ceaee3ccefad_icgraph.png" border="0" usemap="#sim-operations_8h_aadec6c1a0fe783461a72ceaee3ccefad_icgraph" alt=""/></div>
<map name="sim-operations_8h_aadec6c1a0fe783461a72ceaee3ccefad_icgraph" id="sim-operations_8h_aadec6c1a0fe783461a72ceaee3ccefad_icgraph">
<area shape="rect" title="Search for a genotype with parentage matching two given parent ids in a linked list of AlleleMatrix,..." alt="" coords="205,13,313,39"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,5,157,47"/>
</map>
</div>

</div>
</div>
<a id="a67f649fd36494324e609b06fcc960785"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a67f649fd36494324e609b06fcc960785">&#9670;&nbsp;</a></span>get_id_of_index()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int get_id_of_index </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the id of a genotype by its index. </p>
<p>The index is assumed to be 0-based, starting at the first entry of <code>start</code> and continuing through the linked list to which <code>start</code> belongs. Exits if the linked list does not contain that index.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which to locate the id at a particular index. </td></tr>
    <tr><td class="paramname">index</td><td>the index of the target. It is assumed to start at 0 at the start of the matrix in <code>start</code> and be incremented up until the last entry in the matrix of the last entry in the linked list. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the lifetime-unique id of the genotype found at that index. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a67f649fd36494324e609b06fcc960785_icgraph.png" border="0" usemap="#sim-operations_8h_a67f649fd36494324e609b06fcc960785_icgraph" alt=""/></div>
<map name="sim-operations_8h_a67f649fd36494324e609b06fcc960785_icgraph" id="sim-operations_8h_a67f649fd36494324e609b06fcc960785_icgraph">
<area shape="rect" title="Get the id of a genotype by its index." alt="" coords="441,42,553,69"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="220,5,389,47"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,71,393,98"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,5,165,47"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,71,168,98"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,123,163,164"/>
</map>
</div>

</div>
</div>
<a id="a84037d781b8850f9d90425ad16a0d059"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a84037d781b8850f9d90425ad16a0d059">&#9670;&nbsp;</a></span>get_ids_of_names()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void get_ids_of_names </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n_names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>names</em>[n_names], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>output</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for genotypes with certain names in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> and save the ids of those names. </p>
<p>Exits if any name cannot be found.</p>
<p>This function must check every name in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">n_names</td><td>the length of the array of names which are being sought. </td></tr>
    <tr><td class="paramname">names</td><td>an array of names whose ids are being sought. </td></tr>
    <tr><td class="paramname">output</td><td>pointer to an array of length at least <code>n_names</code> which can be accessed by the calling function. The ids of each name are saved to corresponding indexes in the array this pointer points to. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a84037d781b8850f9d90425ad16a0d059_icgraph.png" border="0" usemap="#sim-operations_8h_a84037d781b8850f9d90425ad16a0d059_icgraph" alt=""/></div>
<map name="sim-operations_8h_a84037d781b8850f9d90425ad16a0d059_icgraph" id="sim-operations_8h_a84037d781b8850f9d90425ad16a0d059_icgraph">
<area shape="rect" title="Search for genotypes with certain names in a linked list of AlleleMatrix and save the ids of those na..." alt="" coords="205,13,332,39"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,5,157,47"/>
</map>
</div>

</div>
</div>
<a id="a79a381534123851f35f8c5cf1ee35364"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a79a381534123851f35f8c5cf1ee35364">&#9670;&nbsp;</a></span>get_index_of_child()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_index_of_child </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>parent1id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>parent2id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index. </p>
<p>Exits if such a child cannot be found.</p>
<p>This function must check every genotype in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the child is assumed to be found. </td></tr>
    <tr><td class="paramname">parent1id</td><td>one of the parents of the genotype must have this id. </td></tr>
    <tr><td class="paramname">parent2id</td><td>the other parent of the genotype must have this id. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the index (0-based, starting at the start of <code>start</code>) of the first sequentially located genotype whose parents match the two parent ids provided </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a79a381534123851f35f8c5cf1ee35364_icgraph.png" border="0" usemap="#sim-operations_8h_a79a381534123851f35f8c5cf1ee35364_icgraph" alt=""/></div>
<map name="sim-operations_8h_a79a381534123851f35f8c5cf1ee35364_icgraph" id="sim-operations_8h_a79a381534123851f35f8c5cf1ee35364_icgraph">
<area shape="rect" title="Search for a genotype with parentage matching two given parent ids in a linked list of AlleleMatrix,..." alt="" coords="205,13,335,39"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,5,157,47"/>
</map>
</div>

</div>
</div>
<a id="a93779d37ca9081d07987fd83c27b84af"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a93779d37ca9081d07987fd83c27b84af">&#9670;&nbsp;</a></span>get_index_of_name()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_index_of_name </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Search for a genotype with a particular name in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index. </p>
<p>Exits if such a child cannot be found.</p>
<p>This function must check every genotype in the linked list for matches, so will be relatively slow.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the genotype is assumed to be found. </td></tr>
    <tr><td class="paramname">name</td><td>a string to match to the subject_name of the target </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the index (0-based, starting at the start of <code>start</code>) of the first sequentially located genotype whose name is the same as the provided name. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a93779d37ca9081d07987fd83c27b84af_icgraph.png" border="0" usemap="#sim-operations_8h_a93779d37ca9081d07987fd83c27b84af_icgraph" alt=""/></div>
<map name="sim-operations_8h_a93779d37ca9081d07987fd83c27b84af_icgraph" id="sim-operations_8h_a93779d37ca9081d07987fd83c27b84af_icgraph">
<area shape="rect" title="Search for a genotype with a particular name in a linked list of AlleleMatrix, and return its index." alt="" coords="223,42,357,69"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="9,71,171,98"/>
</map>
</div>

</div>
</div>
<a id="ad7d864d43092823ac2ed8a6cbd595e35"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad7d864d43092823ac2ed8a6cbd595e35">&#9670;&nbsp;</a></span>get_integer_digits()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_integer_digits </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Count and return the number of digits in <code>i</code>. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">i</td><td>the integer whose digits are to be counted. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the number of digits to print <code>i</code> </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad7d864d43092823ac2ed8a6cbd595e35_icgraph.png" border="0" usemap="#sim-operations_8h_ad7d864d43092823ac2ed8a6cbd595e35_icgraph" alt=""/></div>
<map name="sim-operations_8h_ad7d864d43092823ac2ed8a6cbd595e35_icgraph" id="sim-operations_8h_ad7d864d43092823ac2ed8a6cbd595e35_icgraph">
<area shape="rect" title="Count and return the number of digits in i." alt="" coords="624,143,748,170"/>
<area shape="rect" href="sim-operations_8h.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="441,143,576,170"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="ad1a5b9f3e850da13931ee02fbb170d41"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad1a5b9f3e850da13931ee02fbb170d41">&#9670;&nbsp;</a></span>get_malloc()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void* get_malloc </td>
          <td>(</td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>size</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Replace calls to malloc direct with this function, which errors and exits with status 2 if memory allocation fails. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size</td><td>size of space to be allocated. Treat like the parameter of a regular malloc call. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>pointer to the allocated space. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad1a5b9f3e850da13931ee02fbb170d41_icgraph.png" border="0" usemap="#sim-operations_8h_ad1a5b9f3e850da13931ee02fbb170d41_icgraph" alt=""/></div>
<map name="sim-operations_8h_ad1a5b9f3e850da13931ee02fbb170d41_icgraph" id="sim-operations_8h_ad1a5b9f3e850da13931ee02fbb170d41_icgraph">
<area shape="rect" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="872,411,957,437"/>
<area shape="rect" href="sim-operations_8h.html#ab569f6043f8d810aee8d8c5ba380abbd" title="Takes a look at the currently&#45;loaded effect values and creates a string containing the allele with th..." alt="" coords="649,5,817,32"/>
<area shape="rect" href="sim-operations_8h.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="646,259,821,285"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="429,66,583,107"/>
<area shape="rect" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data." alt="" coords="419,131,593,173"/>
<area shape="rect" href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data." alt="" coords="427,197,585,238"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="459,364,553,391"/>
<area shape="rect" href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="678,309,789,336"/>
<area shape="rect" href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="647,461,819,488"/>
<area shape="rect" href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="672,1151,795,1177"/>
<area shape="rect" href="sim-operations_8h.html#a5e049945be5700a6d5a5654b6b332842" title="Updates the chr_ends, n_chr and chr_lengths fields in SimData.map." alt="" coords="643,1049,824,1076"/>
<area shape="rect" href="sim-operations_8h.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="673,512,794,539"/>
<area shape="rect" href="sim-operations_8h.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="682,597,785,624"/>
<area shape="rect" href="sim-operations_8h.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="441,15,571,41"/>
<area shape="rect" href="sim-operations_8h.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="671,711,796,737"/>
<area shape="rect" href="sim-operations_8h.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="665,1100,802,1127"/>
<area shape="rect" href="sim-operations_8h.html#a9d6b231bd6ddde82af66a5d276a2c739" title="Populates a SimData combination with effect values." alt="" coords="652,911,815,937"/>
<area shape="rect" href="sim-operations_8h.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="666,360,801,387"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="421,313,591,340"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="417,263,595,289"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="425,415,587,441"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,255,159,297"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="209,45,367,86"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="207,263,369,289"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="212,146,364,187"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="229,1049,347,1076"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="421,1100,591,1127"/>
<area shape="rect" href="sim-operations_8c.html#a8de70c24458ffb2f3dea5590f7f8cdff" title="Given a set of blocks of markers in a file, for each genotype in a group, calculate the local GEBV fo..." alt="" coords="431,466,581,507"/>
<area shape="rect" href="sim-operations_8h.html#a9e43115d9ee4ac406c14161fc6ccb327" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="441,597,571,624"/>
<area shape="rect" href="sim-operations_8h.html#a0f53ee5421c91ca041a1b8d4603238b5" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="427,531,585,573"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="425,649,587,690"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="222,715,354,741"/>
<area shape="rect" href="sim-operations_8h.html#ab732ec2d648a6dc01444ef74f780fc70" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="433,831,579,873"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="423,897,589,924"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="433,766,579,807"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="213,212,363,239"/>
</map>
</div>

</div>
</div>
<a id="a6aa75e72bb29725ffe32cd003c547480"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6aa75e72bb29725ffe32cd003c547480">&#9670;&nbsp;</a></span>get_name_of_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">char* get_name_of_id </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the name of the genotype with a given id. </p>
<p>The function uses a bisection search on the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> where it should be found. Searching by id is therefore fast. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target.">get_from_ordered_uint_list()</a></dd></dl>
<p>This function assumes that ids are never reshuffled in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. As of 2021.04.12 this condition is always true, but this function will need to be revisited if this is no longer the case.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">id</td><td>the id of the genotype whose name is sought </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the name of the genotype that has id <code>id</code>, as a copy of the pointer to the heap memory where the name is saved (so <em>don't</em> free the pointer returned from this function) </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_cgraph.png" border="0" usemap="#sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_cgraph" alt=""/></div>
<map name="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_cgraph" id="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_cgraph">
<area shape="rect" title="Returns the name of the genotype with a given id." alt="" coords="5,5,119,32"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="167,5,341,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_icgraph.png" border="0" usemap="#sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_icgraph" alt=""/></div>
<map name="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_icgraph" id="sim-operations_8h_a6aa75e72bb29725ffe32cd003c547480_icgraph">
<area shape="rect" title="Returns the name of the genotype with a given id." alt="" coords="868,150,981,177"/>
<area shape="rect" href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="461,133,588,159"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="665,113,811,155"/>
<area shape="rect" href="sim-operations_8h.html#a6b87b484064fd373aec0bfa4ce8dbef1" title="Print the parents of each genotype in the SimData to a file." alt="" coords="656,179,820,206"/>
<area shape="rect" href="sim-operations_8h.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="681,254,795,281"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,50,390,77"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,209,352,235"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="441,254,608,281"/>
</map>
</div>

</div>
</div>
<a id="ae974a74145fe876e303e14336d3c5d3a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae974a74145fe876e303e14336d3c5d3a">&#9670;&nbsp;</a></span>get_new_group_num()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_new_group_num </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to identify the next sequential integer that does not identify a group that currently has member(s). </p>
<p>This calls <a class="el" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members.">get_existing_groups()</a> every time, so functions that do repeated group creation, like split_into_individuals, are recommended to use their own implementation rather than calling this function repeatedly for best speed/memory usage.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the next sequential currently-unused group number. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_cgraph.png" border="0" usemap="#sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_cgraph" alt=""/></div>
<map name="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_cgraph" id="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_cgraph">
<area shape="rect" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="5,5,147,32"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="195,5,333,32"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="381,5,552,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_icgraph.png" border="0" usemap="#sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_icgraph" alt=""/></div>
<map name="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_icgraph" id="sim-operations_8h_ae974a74145fe876e303e14336d3c5d3a_icgraph">
<area shape="rect" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="651,152,792,179"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="207,5,377,32"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="425,64,603,91"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="437,115,591,157"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="433,181,595,208"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="467,232,561,259"/>
<area shape="rect" href="sim-operations_8h.html#af2eda46d016c384b166ce65605952b7c" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="455,283,573,309"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,137,159,178"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="213,57,371,98"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="211,123,373,149"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="216,174,368,215"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="217,261,367,288"/>
</map>
</div>

</div>
</div>
<a id="a229c9de94875ea41a23723d337f5f92f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a229c9de94875ea41a23723d337f5f92f">&#9670;&nbsp;</a></span>get_parents_of_id()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int get_parents_of_id </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>output</em>[2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>. </p>
<p>The function uses a bisection search on the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> where it should be found. Searching by id is therefore fast. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target.">get_from_ordered_uint_list()</a></dd></dl>
<p>This function assumes that ids are never reshuffled in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. As of 2021.04.12 this condition is always true, but this function will need to be revisited if this is no longer the case.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">start</td><td>Pointer to the first of a linked list of AlleleMatrixes in which the genotype with the provided id is assumed to be found. </td></tr>
    <tr><td class="paramname">id</td><td>the id of the genotype whose parents are sought </td></tr>
    <tr><td class="paramname">output</td><td>An array which the calling function can access where this function will put its results. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>0 when the id is successfully identified and at least one parent's id is known, 1 otherwise. The ids of both parents if at least one parent is known/nonzero are saved to the array <code>output</code>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_cgraph.png" border="0" usemap="#sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_cgraph" alt=""/></div>
<map name="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_cgraph" id="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_cgraph">
<area shape="rect" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="5,5,129,32"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="177,5,352,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_icgraph.png" border="0" usemap="#sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_icgraph" alt=""/></div>
<map name="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_icgraph" id="sim-operations_8h_a229c9de94875ea41a23723d337f5f92f_icgraph">
<area shape="rect" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="868,83,992,110"/>
<area shape="rect" href="sim-operations_8h.html#a979efe39128de97176da5460cab7dfa7" title="Print the parents of each genotype in a group to a file." alt="" coords="665,17,811,59"/>
<area shape="rect" href="sim-operations_8h.html#a6b87b484064fd373aec0bfa4ce8dbef1" title="Print the parents of each genotype in the SimData to a file." alt="" coords="656,83,820,110"/>
<area shape="rect" href="sim-operations_8h.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="681,158,795,185"/>
<area shape="rect" href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="461,133,588,159"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="441,183,608,210"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,34,390,61"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,209,352,235"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="adebbfa88bf139ddca501ab158f3746c5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adebbfa88bf139ddca501ab158f3746c5">&#9670;&nbsp;</a></span>get_sorted_markers()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void get_sorted_markers </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>actual_n_markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object, and sorts its markers, the rows of its parent gen <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (because they are ordered by the markers), and its genetic map so that the markers are ordered by chromosome number then position. </p>
<p>Markers that do not have a position in d-&gt;map.positions are deleted from all those three lists. This is done by malloc-ing new memory, copying data over in the new sorted order, and freeing the old memory.</p>
<p>Note: only the starting generation in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> list is reordered, and all additional <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> objects are deleted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to have its markers, genetic map, and allele matrix sorted. The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> pointed to by d will be modified by this function. </td></tr>
    <tr><td class="paramname">actual_n_markers</td><td>If previously calculated, include the number of markers in <code>d-&gt;markers</code> that have a position loaded into <code>d-&gt;map.positions</code>. If this has not been calculated yet, make the value -1 and this function will calculate it. The value is calculated as the number of positions in <code>d-&gt;map.positions</code> that have a chromosome number of 0. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_cgraph.png" border="0" usemap="#sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_cgraph" alt=""/></div>
<map name="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_cgraph" id="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_cgraph">
<area shape="rect" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="5,107,143,133"/>
<area shape="rect" href="sim-operations_8c.html#a80cab739edbd12bbd2ad1f9b77b27548" title="Comparator function for qsort." alt="" coords="191,5,351,32"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="217,56,324,83"/>
<area shape="rect" href="sim-operations_8c.html#aaf7e12b92ec0f93bb3bc2ec67155bc94" title="Deletes a GeneticMap object and frees its memory." alt="" coords="215,107,326,133"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="193,157,348,184"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="228,208,313,235"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_icgraph.png" border="0" usemap="#sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_icgraph" alt=""/></div>
<map name="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_icgraph" id="sim-operations_8h_adebbfa88bf139ddca501ab158f3746c5_icgraph">
<area shape="rect" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="391,5,528,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
<area shape="rect" href="sim-operations_8h.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,31,343,57"/>
</map>
</div>

</div>
</div>
<a id="aa71cdbced90506820ef61bd842ecfa40"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa71cdbced90506820ef61bd842ecfa40">&#9670;&nbsp;</a></span>has_same_alleles()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int has_same_alleles </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>p2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>i</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Simple operator to determine if at marker i, two genotypes share at least one allele. </p>
<p>Checks only 3 of four possible permutations because assumes there cannot be more than two alleles at a given marker.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">p1</td><td>pointer to a character array genotype of the type stored in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (2*n_markers long, representing the two alleles at a marker consecutively) for the first of the genotypes to compare. </td></tr>
    <tr><td class="paramname">p2</td><td>pointer to a character array genotype for the second of the genotypes to compare. </td></tr>
    <tr><td class="paramname">i</td><td>index of the marker at which to perform the check </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>boolean result of the check </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa71cdbced90506820ef61bd842ecfa40_icgraph.png" border="0" usemap="#sim-operations_8h_aa71cdbced90506820ef61bd842ecfa40_icgraph" alt=""/></div>
<map name="sim-operations_8h_aa71cdbced90506820ef61bd842ecfa40_icgraph" id="sim-operations_8h_aa71cdbced90506820ef61bd842ecfa40_icgraph">
<area shape="rect" title="Simple operator to determine if at marker i, two genotypes share at least one allele." alt="" coords="493,13,620,39"/>
<area shape="rect" href="sim-operations_8h.html#aad9cd4248231da3682ba2f08a40cdbe8" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="223,13,445,39"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="ab88223532880a22dfc966be4cdb4107b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab88223532880a22dfc966be4cdb4107b">&#9670;&nbsp;</a></span>has_same_alleles_window()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int has_same_alleles_window </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>g1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>g2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>start</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>w</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at least one allele. </p>
<p>Checks only 3 of four possible permutations at each marker because assumes there cannot be more than two alleles at a given marker. For the return value to be true, there must be at least one match at every one of the markers in the window.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">g1</td><td>pointer to a character array genotype of the type stored in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (2*n_markers long, representing the two alleles at a marker consecutively) for the first of the genotypes to compare. </td></tr>
    <tr><td class="paramname">g2</td><td>pointer to a character array genotype for the second of the genotypes to compare. </td></tr>
    <tr><td class="paramname">start</td><td>index of the first marker in the window over which to perform the check </td></tr>
    <tr><td class="paramname">w</td><td>length of the window over which to perform the check </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>boolean result of the check </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ab88223532880a22dfc966be4cdb4107b_icgraph.png" border="0" usemap="#sim-operations_8h_ab88223532880a22dfc966be4cdb4107b_icgraph" alt=""/></div>
<map name="sim-operations_8h_ab88223532880a22dfc966be4cdb4107b_icgraph" id="sim-operations_8h_ab88223532880a22dfc966be4cdb4107b_icgraph">
<area shape="rect" title="Simple operator to determine if at markers with indexes i to i+w inclusive, two genotypes share at le..." alt="" coords="493,13,668,39"/>
<area shape="rect" href="sim-operations_8h.html#a3627a380ed86ebebe937c953bc3b2339" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="223,13,445,39"/>
<area shape="rect" href="sim-operations_8h.html#a7813f256407ab2ea242df32ea9ad2c7e" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="5,5,175,47"/>
</map>
</div>

</div>
</div>
<a id="a54cd625669413c2ef171060e9844d814"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a54cd625669413c2ef171060e9844d814">&#9670;&nbsp;</a></span>load_all_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_all_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>data_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>map_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>effect_file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination from scratch with marker allele data, a genetic map, and effect values. </p>
<p>Note: this function shows the order that files need to be loaded into <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>, i.e. Allele data first (so we know what markers we care about), then genetic map data (so we know what markers we can use in simulation, and rearrange them to be ordered), then effects data (to be saved in the correct order according to newly ordered markers).</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated </td></tr>
    <tr><td class="paramname">data_file</td><td>string containing name/path of file containing SNP marker allele data. </td></tr>
    <tr><td class="paramname">map_file</td><td>string name/path of file containing genetic map data. </td></tr>
    <tr><td class="paramname">effect_file</td><td>string name/path of file containing effect values. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a54cd625669413c2ef171060e9844d814_cgraph.png" border="0" usemap="#sim-operations_8h_a54cd625669413c2ef171060e9844d814_cgraph" alt=""/></div>
<map name="sim-operations_8h_a54cd625669413c2ef171060e9844d814_cgraph" id="sim-operations_8h_a54cd625669413c2ef171060e9844d814_cgraph">
<area shape="rect" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,300,124,327"/>
<area shape="rect" href="sim-operations_8c.html#a6fc4a687a2186e9e9dcf3229d9141041" title="Deletes a SimData object and frees its memory." alt="" coords="202,107,313,133"/>
<area shape="rect" href="sim-operations_8c.html#a5e049945be5700a6d5a5654b6b332842" title="Updates the chr_ends, n_chr and chr_lengths fields in SimData.map." alt="" coords="391,427,572,453"/>
<area shape="rect" href="sim-operations_8c.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="413,325,550,352"/>
<area shape="rect" href="sim-operations_8c.html#a9d6b231bd6ddde82af66a5d276a2c739" title="Populates a SimData combination with effect values." alt="" coords="176,220,339,247"/>
<area shape="rect" href="sim-operations_8c.html#a8edb09b9b9eb27f84ac59cc4342ea855" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,275,343,301"/>
<area shape="rect" href="sim-operations_8c.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data." alt="" coords="178,521,337,562"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="413,5,550,32"/>
<area shape="rect" href="sim-operations_8c.html#a961f4b0757752fee2bd08adde7ece8ad" title="Deletes an EffectMatrix object and frees its memory." alt="" coords="412,56,551,83"/>
<area shape="rect" href="sim-operations_8c.html#aaf7e12b92ec0f93bb3bc2ec67155bc94" title="Deletes a GeneticMap object and frees its memory." alt="" coords="645,275,755,301"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="647,107,753,133"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="657,427,743,453"/>
<area shape="rect" href="sim-operations_8c.html#a80cab739edbd12bbd2ad1f9b77b27548" title="Comparator function for qsort." alt="" coords="620,376,780,403"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="623,325,777,352"/>
<area shape="rect" href="sim-operations_8c.html#aed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="413,209,549,250"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="394,477,569,504"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="413,629,549,656"/>
<area shape="rect" href="sim-operations_8c.html#ac5f92e5e4bd9d414392a2f022fdc5596" title="Allocate lifetime&#45;unique ids to each genotype in the range of whole SimData indexes from_index to to_..." alt="" coords="425,579,538,605"/>
</map>
</div>

</div>
</div>
<a id="a9d6b231bd6ddde82af66a5d276a2c739"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9d6b231bd6ddde82af66a5d276a2c739">&#9670;&nbsp;</a></span>load_effects_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void load_effects_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with effect values. </p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must already have its allele data and map data loaded (so that it has an ordered <code>markers</code> list and no markers that will not be used for simulation.</p>
<p>It loads in the file as rows of effects for each allele that appears in the allele data out of 'A', 'C', 'G', 'T'.</p>
<p>The file should have format:</p>
<p>[marker] [allele] [effect]</p>
<p>[marker] [allele] [effect]</p>
<p>...</p>
<p>The function assumes the maximum line length is 99 characters. It also assumes that the array ref_alleles is the same length as m's marker_names vector.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated. </td></tr>
    <tr><td class="paramname">filename</td><td>string name/path of file containing effect values. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_cgraph.png" border="0" usemap="#sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_cgraph" alt=""/></div>
<map name="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_cgraph" id="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_cgraph">
<area shape="rect" title="Populates a SimData combination with effect values." alt="" coords="5,64,168,91"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="231,5,337,32"/>
<area shape="rect" href="sim-operations_8c.html#aed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="216,57,352,98"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="241,123,327,149"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_icgraph.png" border="0" usemap="#sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_icgraph" alt=""/></div>
<map name="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_icgraph" id="sim-operations_8h_a9d6b231bd6ddde82af66a5d276a2c739_icgraph">
<area shape="rect" title="Populates a SimData combination with effect values." alt="" coords="172,5,335,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
</map>
</div>

</div>
</div>
<a id="aeba5195a9b36f7f4dab457bfd84a65b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aeba5195a9b36f7f4dab457bfd84a65b2">&#9670;&nbsp;</a></span>load_genes_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_genes_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a8edb09b9b9eb27f84ac59cc4342ea855"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8edb09b9b9eb27f84ac59cc4342ea855">&#9670;&nbsp;</a></span>load_genmap_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void load_genmap_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with data from a genetic map. </p>
<p>Map positions must be in cM.</p>
<p>Note: this function should be called second when populating a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object, after populating it with marker allele data. This is because this function loads the genmap data corresponding to markers used in the allele file, then deletes markers in <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> that do not have positions for ease of simulation.</p>
<p>The file's format should be:</p>
<p>marker chr pos</p>
<p>[marker name] [chr] [pos]</p>
<p>[marker name] [chr] [pos]</p>
<p>...</p>
<p>The function assumes the maximum line length is 99 characters. It also assumes that there is only one mapping per marker in the file.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated </td></tr>
    <tr><td class="paramname">filename</td><td>string name/path of file containing genetic map data. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_cgraph.png" border="0" usemap="#sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_cgraph" alt=""/></div>
<map name="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_cgraph" id="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_cgraph">
<area shape="rect" title="Populates a SimData combination with data from a genetic map." alt="" coords="5,81,176,108"/>
<area shape="rect" href="sim-operations_8c.html#aaf7e12b92ec0f93bb3bc2ec67155bc94" title="Deletes a GeneticMap object and frees its memory." alt="" coords="478,5,589,32"/>
<area shape="rect" href="sim-operations_8c.html#a5e049945be5700a6d5a5654b6b332842" title="Updates the chr_ends, n_chr and chr_lengths fields in SimData.map." alt="" coords="224,56,405,83"/>
<area shape="rect" href="sim-operations_8c.html#aed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="247,158,383,199"/>
<area shape="rect" href="sim-operations_8c.html#adebbfa88bf139ddca501ab158f3746c5" title="Takes a SimData object, and sorts its markers, the rows of its parent gen AlleleMatrix (because they ..." alt="" coords="246,107,383,133"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="491,56,576,83"/>
<area shape="rect" href="sim-operations_8c.html#a80cab739edbd12bbd2ad1f9b77b27548" title="Comparator function for qsort." alt="" coords="453,107,613,133"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="480,157,587,184"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="456,208,611,235"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_icgraph.png" border="0" usemap="#sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_icgraph" alt=""/></div>
<map name="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_icgraph" id="sim-operations_8h_a8edb09b9b9eb27f84ac59cc4342ea855_icgraph">
<area shape="rect" title="Populates a SimData combination with data from a genetic map." alt="" coords="172,5,343,32"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,5,124,32"/>
</map>
</div>

</div>
</div>
<a id="af4f7f33b6ca4f50bcc60504a6ee75ee2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af4f7f33b6ca4f50bcc60504a6ee75ee2">&#9670;&nbsp;</a></span>load_more_genes_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_more_genes_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a8bd85cb47ec8e0c4049f937ef40ae102"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8bd85cb47ec8e0c4049f937ef40ae102">&#9670;&nbsp;</a></span>load_more_transposed_genes_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_more_transposed_genes_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Appends genotype data from a file to an existing <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data.">load_transposed_genes_to_simdata()</a></dd></dl>
<p>Given a file with the following format:</p>
<p>name [line] [line] [line] ... [line]</p>
<p>[marker] [SNP pair] [SNP pair] [SNP pair] ... [SNP pair]</p>
<p>[marker] [SNP pair] [SNP pair] [SNP pair] ... [SNP pair]</p>
<p>...</p>
<p>Where [line] is a code for a line, [marker] is a code for a marker, and [SNP pair] is eg TT, TA.</p>
<p>If a given marker does not exist in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>'s set of markers, it is ignored. for the purposes of loading. No markers can be added to a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> after the creation step.</p>
<p>An output message stating the number of genotypes and number of markers loaded is printed to stdout.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated </td></tr>
    <tr><td class="paramname">filename</td><td>string containing name/path of file containing SNP marker allele data. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the loaded genotypes. All genotypes are loaded into the same group. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a8bd85cb47ec8e0c4049f937ef40ae102_cgraph.png" border="0" usemap="#sim-operations_8h_a8bd85cb47ec8e0c4049f937ef40ae102_cgraph" alt=""/></div>
<map name="sim-operations_8h_a8bd85cb47ec8e0c4049f937ef40ae102_cgraph" id="sim-operations_8h_a8bd85cb47ec8e0c4049f937ef40ae102_cgraph">
<area shape="rect" title="Appends genotype data from a file to an existing SimData." alt="" coords="5,150,159,191"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="215,5,373,32"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="207,56,381,83"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="456,81,541,108"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="226,157,362,184"/>
<area shape="rect" href="sim-operations_8c.html#aed2ae969c3ce4ed1466a885dcc559472" title="Returns the first located index in an array of strings where the string is the same as the string tar..." alt="" coords="226,209,362,250"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="223,275,365,301"/>
<area shape="rect" href="sim-operations_8c.html#ac5f92e5e4bd9d414392a2f022fdc5596" title="Allocate lifetime&#45;unique ids to each genotype in the range of whole SimData indexes from_index to to_..." alt="" coords="237,325,351,352"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="430,5,567,32"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="429,275,568,301"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="616,275,787,301"/>
</map>
</div>

</div>
</div>
<a id="ad5863641a67355c2c1523e1a23b6f581"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad5863641a67355c2c1523e1a23b6f581">&#9670;&nbsp;</a></span>load_transposed_encoded_genes_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_transposed_encoded_genes_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data.">load_transposed_genes_to_simdata()</a> Assumes it is starting from a clean/empty <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.</dd></dl>
<p>Given a file with the following format:</p>
<p>name [line] [line] [line] ... [line]</p>
<p>[marker] [encoded] [encoded] [encoded] ... [encoded]</p>
<p>[marker] [encoded] [encoded] [encoded] ... [encoded]</p>
<p>...</p>
<p>Where [line] is a code for a line, [marker] is a code for a marker, and [encoded] is the standard IUPAC encoding for a particular pair. Because this simulation tracks phase, and this encoding does not, the phase at heterozygous markers is chosen randomly.</p>
<p>Code =&gt; Alleles key: A =&gt; AA ; C =&gt; CC ; G =&gt; GG ; T =&gt; TT ; R =&gt; AG ; Y =&gt; CT ; S =&gt; CG ; W =&gt; AT ; K =&gt; GT ; M =&gt; AC</p>
<p>Note: this function should be called first when populating a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object - it clears everything in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. This is because all the data in <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> is based on what markers exist in the loaded marker allele file.</p>
<p>An output message stating the number of genotypes and number of markers loaded is printed to stdout.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated </td></tr>
    <tr><td class="paramname">filename</td><td>string containing name/path of file containing SNP marker allele data. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the loaded genotypes. All genotypes are loaded into the same group. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad5863641a67355c2c1523e1a23b6f581_cgraph.png" border="0" usemap="#sim-operations_8h_ad5863641a67355c2c1523e1a23b6f581_cgraph" alt=""/></div>
<map name="sim-operations_8h_ad5863641a67355c2c1523e1a23b6f581_cgraph" id="sim-operations_8h_ad5863641a67355c2c1523e1a23b6f581_cgraph">
<area shape="rect" title="Populates a SimData combination with marker allele data." alt="" coords="5,99,179,141"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="227,5,401,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="449,31,535,57"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="246,107,382,133"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="287,157,341,184"/>
<area shape="rect" href="sim-operations_8c.html#ac5f92e5e4bd9d414392a2f022fdc5596" title="Allocate lifetime&#45;unique ids to each genotype in the range of whole SimData indexes from_index to to_..." alt="" coords="257,208,371,235"/>
</map>
</div>

</div>
</div>
<a id="a156666e21151a9f267e31598574366a6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a156666e21151a9f267e31598574366a6">&#9670;&nbsp;</a></span>load_transposed_genes_to_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int load_transposed_genes_to_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data.">load_transposed_encoded_genes_to_simdata()</a> Assumes it is starting from a clean/empty <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.</dd></dl>
<p>Given a file with the following format:</p>
<p>name [line] [line] [line] ... [line]</p>
<p>[marker] [SNP pair] [SNP pair] [SNP pair] ... [SNP pair]</p>
<p>[marker] [SNP pair] [SNP pair] [SNP pair] ... [SNP pair]</p>
<p>...</p>
<p>Where [line] is a code for a line, [marker] is a code for a marker, and [SNP pair] is eg TT, TA.</p>
<p>Note: this function should be called first when populating a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object - it clears everything in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>. This is because all the data in <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> is based on what markers exist in the loaded marker allele file.</p>
<p>An output message stating the number of genotypes and number of markers loaded is printed to stdout.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to be populated </td></tr>
    <tr><td class="paramname">filename</td><td>string containing name/path of file containing SNP marker allele data. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the loaded genotypes. All genotypes are loaded into the same group. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a156666e21151a9f267e31598574366a6_cgraph.png" border="0" usemap="#sim-operations_8h_a156666e21151a9f267e31598574366a6_cgraph" alt=""/></div>
<map name="sim-operations_8h_a156666e21151a9f267e31598574366a6_cgraph" id="sim-operations_8h_a156666e21151a9f267e31598574366a6_cgraph">
<area shape="rect" title="Populates a SimData combination with marker allele data." alt="" coords="5,74,164,115"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="212,5,387,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="435,31,520,57"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="231,107,367,133"/>
<area shape="rect" href="sim-operations_8c.html#ac5f92e5e4bd9d414392a2f022fdc5596" title="Allocate lifetime&#45;unique ids to each genotype in the range of whole SimData indexes from_index to to_..." alt="" coords="243,157,356,184"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a156666e21151a9f267e31598574366a6_icgraph.png" border="0" usemap="#sim-operations_8h_a156666e21151a9f267e31598574366a6_icgraph" alt=""/></div>
<map name="sim-operations_8h_a156666e21151a9f267e31598574366a6_icgraph" id="sim-operations_8h_a156666e21151a9f267e31598574366a6_icgraph">
<area shape="rect" title="Populates a SimData combination with marker allele data." alt="" coords="172,5,331,47"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,13,124,39"/>
</map>
</div>

</div>
</div>
<a id="a3f1b06b7aa35e6a05918e16d1deabeee"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3f1b06b7aa35e6a05918e16d1deabeee">&#9670;&nbsp;</a></span>make_all_unidirectional_crosses()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_all_unidirectional_crosses </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>from_group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Perform crosses between all pairs of parents in the group <code>from_group</code> and allocates the resulting offspring to a new group. </p>
<p>If the group has n members, there will be n * (n-1) / 2 offspring produced.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object containing or markers and parent alleles </td></tr>
    <tr><td class="paramname">from_group</td><td>group number from which to do all these crosses. </td></tr>
    <tr><td class="paramname">g</td><td>options for the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3f1b06b7aa35e6a05918e16d1deabeee_cgraph.png" border="0" usemap="#sim-operations_8h_a3f1b06b7aa35e6a05918e16d1deabeee_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3f1b06b7aa35e6a05918e16d1deabeee_cgraph" id="sim-operations_8h_a3f1b06b7aa35e6a05918e16d1deabeee_cgraph">
<area shape="rect" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="5,759,163,801"/>
<area shape="rect" href="sim-operations_8c.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="211,476,388,503"/>
<area shape="rect" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="234,767,365,793"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="468,856,579,883"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="442,299,605,325"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="695,451,802,477"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="444,451,603,477"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="680,501,817,528"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="436,603,611,629"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="468,805,579,832"/>
<area shape="rect" href="sim-operations_8c.html#a309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="455,653,592,680"/>
<area shape="rect" href="sim-operations_8c.html#a67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="467,704,579,731"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="453,139,594,165"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="459,88,587,115"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="477,349,570,376"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="460,189,587,216"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="456,552,591,579"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="695,333,802,360"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="895,308,1049,335"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="681,385,817,426"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="683,232,814,259"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="679,181,818,208"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="706,653,791,680"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="659,805,839,832"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="721,856,776,883"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="717,755,781,781"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="679,5,818,32"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="887,5,1057,32"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="915,131,1029,157"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="691,80,806,107"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="1105,105,1280,132"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="910,80,1034,107"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="687,552,811,579"/>
</map>
</div>

</div>
</div>
<a id="a3b67d38c13c7a8bef2d0fc91c0365177"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3b67d38c13c7a8bef2d0fc91c0365177">&#9670;&nbsp;</a></span>make_crosses_from_file()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_crosses_from_file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>input_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Perform crosses between pairs of parents identified by name in a file and allocate the resulting offspring to a new group. </p>
<p>The input file should have format:</p>
<p>[parent1 name] [parent2 name]</p>
<p>[parent1 name] [parent2 name]</p>
<p>...</p>
<p>where each row represents a separate cross to carry out.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object containing or markers and parent alleles </td></tr>
    <tr><td class="paramname">input_file</td><td>file instructing which crosses to perform </td></tr>
    <tr><td class="paramname">g</td><td>options for the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a3b67d38c13c7a8bef2d0fc91c0365177_cgraph.png" border="0" usemap="#sim-operations_8h_a3b67d38c13c7a8bef2d0fc91c0365177_cgraph" alt=""/></div>
<map name="sim-operations_8h_a3b67d38c13c7a8bef2d0fc91c0365177_cgraph" id="sim-operations_8h_a3b67d38c13c7a8bef2d0fc91c0365177_cgraph">
<area shape="rect" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,556,168,583"/>
<area shape="rect" href="sim-operations_8c.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,505,393,532"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="237,556,373,583"/>
<area shape="rect" href="sim-operations_8c.html#a93779d37ca9081d07987fd83c27b84af" title="Search for a genotype with a particular name in a linked list of AlleleMatrix, and return its index." alt="" coords="237,607,372,633"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="447,151,610,177"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="701,252,807,279"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="449,353,608,380"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="685,303,823,329"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="441,455,616,481"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="473,404,584,431"/>
<area shape="rect" href="sim-operations_8c.html#a309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="460,556,597,583"/>
<area shape="rect" href="sim-operations_8c.html#a67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="473,607,585,633"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="458,657,599,684"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="465,708,593,735"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="482,759,575,785"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="465,809,592,836"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="461,505,596,532"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="701,33,807,60"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="900,63,1055,89"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="686,85,822,126"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="689,151,819,177"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="685,201,823,228"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="711,505,797,532"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="664,353,844,380"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="727,404,781,431"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="722,455,786,481"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="685,657,823,684"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="892,657,1063,684"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="921,835,1034,861"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="697,784,811,811"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="1111,809,1285,836"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="915,784,1039,811"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="692,556,816,583"/>
</map>
</div>

</div>
</div>
<a id="aa7c90b4015e4abce750176c6d891cad2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa7c90b4015e4abce750176c6d891cad2">&#9670;&nbsp;</a></span>make_double_crosses_from_file()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_double_crosses_from_file </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>input_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Perform crosses between previously-generated offspring of pairs of parents identified by name in a file. </p>
<p>The resulting offspring are allocated to a new group.</p>
<p>The input file should have format:</p>
<p>[gparent1 name] [gparent2 name] [gparent3 name] [gparent4 name]</p>
<p>[gparent1 name] [gparent2 name] [gparent3 name] [gparent4 name]</p>
<p>...</p>
<p>where each row represents a separate cross to carry out.</p>
<p>Results of the cross [gparent1 name] x [gparent2 name] and [gparent3 name] x [gparent4 name] must have already been generated with pedigree tracking options turned on. Messages will be printed to stderr if such offspring cannot be found.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object containing or markers and parent alleles </td></tr>
    <tr><td class="paramname">input_file</td><td>file instructing which crosses to perform </td></tr>
    <tr><td class="paramname">g</td><td>options for the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa7c90b4015e4abce750176c6d891cad2_cgraph.png" border="0" usemap="#sim-operations_8h_aa7c90b4015e4abce750176c6d891cad2_cgraph" alt=""/></div>
<map name="sim-operations_8h_aa7c90b4015e4abce750176c6d891cad2_cgraph" id="sim-operations_8h_aa7c90b4015e4abce750176c6d891cad2_cgraph">
<area shape="rect" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="5,599,157,641"/>
<area shape="rect" href="sim-operations_8c.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="205,505,383,532"/>
<area shape="rect" href="sim-operations_8c.html#a392c23248f22e9faf085198dbb1d8711" title="Opens a table file and reads the number of columns and rows (including headers) separated by sep into..." alt="" coords="226,556,362,583"/>
<area shape="rect" href="sim-operations_8c.html#aadec6c1a0fe783461a72ceaee3ccefad" title="Search for a genotype with parentage matching two given parent ids in a linked list of AlleleMatrix,..." alt="" coords="240,607,348,633"/>
<area shape="rect" href="sim-operations_8c.html#a84037d781b8850f9d90425ad16a0d059" title="Search for genotypes with certain names in a linked list of AlleleMatrix and save the ids of those na..." alt="" coords="231,657,357,684"/>
<area shape="rect" href="sim-operations_8c.html#a79a381534123851f35f8c5cf1ee35364" title="Search for a genotype with parentage matching two given parent ids in a linked list of AlleleMatrix,..." alt="" coords="229,708,359,735"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="437,151,599,177"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="690,252,797,279"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="439,353,597,380"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="675,303,812,329"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="431,455,605,481"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="463,404,573,431"/>
<area shape="rect" href="sim-operations_8c.html#a309dbf26800ee65e982df2b347980906" title="Get the alleles of a genotype by its index." alt="" coords="449,556,587,583"/>
<area shape="rect" href="sim-operations_8c.html#a67f649fd36494324e609b06fcc960785" title="Get the id of a genotype by its index." alt="" coords="462,607,574,633"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="447,657,589,684"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="454,708,582,735"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="471,759,565,785"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="455,809,581,836"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="451,505,585,532"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="690,33,797,60"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="889,63,1044,89"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="675,85,811,126"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="678,151,809,177"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="674,201,813,228"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="701,505,786,532"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="653,353,833,380"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="716,404,771,431"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="711,455,775,481"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="674,657,813,684"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="881,657,1052,684"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="910,835,1023,861"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="686,784,801,811"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="1100,809,1275,836"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="905,784,1029,811"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="681,556,805,583"/>
</map>
</div>

</div>
</div>
<a id="af078d6bc1b5d4234a9b960e6132480aa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af078d6bc1b5d4234a9b960e6132480aa">&#9670;&nbsp;</a></span>make_doubled_haploids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_doubled_haploids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Creates a doubled haploid from each member of a group. </p>
<p>The resulting genotypes are allocated to a new group.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this operation. The family_size parameter in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> allows you to generate multiple doubled haploid offspring from each member of the group. These multiple offspring all originate from independent processes of generating a gamete then doubling its alleles.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that contains the genetic map and genotypes of the parent group. </td></tr>
    <tr><td class="paramname">group</td><td>the genotypes on which to perform the operation. </td></tr>
    <tr><td class="paramname">g</td><td>options for the genotypes created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_af078d6bc1b5d4234a9b960e6132480aa_cgraph.png" border="0" usemap="#sim-operations_8h_af078d6bc1b5d4234a9b960e6132480aa_cgraph" alt=""/></div>
<map name="sim-operations_8h_af078d6bc1b5d4234a9b960e6132480aa_cgraph" id="sim-operations_8h_af078d6bc1b5d4234a9b960e6132480aa_cgraph">
<area shape="rect" title="Creates a doubled haploid from each member of a group." alt="" coords="5,553,167,579"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="221,173,383,199"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="474,274,581,301"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="223,375,381,402"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="459,325,596,351"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="215,477,389,503"/>
<area shape="rect" href="sim-operations_8c.html#ac1f46b3de0afb4674ae6d34311e7aea7" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="216,426,388,453"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="241,527,363,554"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="472,578,583,605"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="251,578,353,605"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="231,730,373,757"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="238,781,366,807"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="255,831,349,858"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="239,882,365,909"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="235,629,369,655"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="474,223,581,250"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="673,122,828,149"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="459,5,595,47"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="462,71,593,98"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="458,122,597,149"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="485,527,570,554"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="437,375,617,402"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="500,426,555,453"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="495,477,559,503"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="458,730,597,757"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="665,730,836,757"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="694,869,807,895"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="470,919,585,946"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="884,894,1059,921"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="689,919,813,946"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="465,629,589,655"/>
</map>
</div>

</div>
</div>
<a id="a68aca99a41527b019316c8243e6fc7a7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a68aca99a41527b019316c8243e6fc7a7">&#9670;&nbsp;</a></span>make_n_crosses_from_top_m_percent()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int make_n_crosses_from_top_m_percent </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Find the top m percent of a group and perform random crosses between those top individuals. </p>
<p>The resulting offspring are allocated to a new group.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object containing or markers and parent alleles </td></tr>
    <tr><td class="paramname">n</td><td>number of random crosses to make. </td></tr>
    <tr><td class="paramname">m</td><td>percentage (as a decimal, i.e. 0.2 for 20percent). Take the best [this portion] of the group for performing the random crosses. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to identify top parents and do crosses </td></tr>
    <tr><td class="paramname">g</td><td>options for the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a68aca99a41527b019316c8243e6fc7a7_cgraph.png" border="0" usemap="#sim-operations_8h_a68aca99a41527b019316c8243e6fc7a7_cgraph" alt=""/></div>
<map name="sim-operations_8h_a68aca99a41527b019316c8243e6fc7a7_cgraph" id="sim-operations_8h_a68aca99a41527b019316c8243e6fc7a7_cgraph">
<area shape="rect" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,637,159,679"/>
<area shape="rect" href="sim-operations_8c.html#ade6b01bfc6450638a3a341b76e28a97d" title="Combine a set of groups into one group." alt="" coords="223,569,341,595"/>
<area shape="rect" href="sim-operations_8c.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="417,391,588,418"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="905,586,1016,613"/>
<area shape="rect" href="sim-operations_8c.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,1147,357,1174"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="654,771,817,798"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="907,1025,1014,1051"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="656,62,815,89"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="892,11,1029,38"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="648,417,823,443"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="680,113,791,139"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="675,467,796,494"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="684,518,787,545"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="665,1257,806,1283"/>
<area shape="rect" href="sim-operations_8c.html#a89e066191f12e056cebb8d2cd74a2a29" title="Generate an integer randomly between 0 and limit inclusive." alt="" coords="703,265,767,291"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="671,315,799,342"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="689,366,782,393"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="672,163,799,190"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="668,214,803,241"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="907,721,1014,747"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="1107,797,1261,823"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="893,655,1029,696"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="895,822,1026,849"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="891,873,1030,899"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="918,442,1003,469"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="871,62,1051,89"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="933,113,988,139"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="929,163,993,190"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="891,1257,1030,1283"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="1099,1257,1269,1283"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="1127,289,1241,315"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="903,238,1018,265"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="1317,289,1492,315"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="1122,238,1246,265"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="899,339,1023,366"/>
<area shape="rect" href="sim-operations_8c.html#a311381904a1ca73ea26dfede9fcddea2" title="Comparator function for qsort." alt="" coords="405,1198,600,1225"/>
<area shape="rect" href="sim-operations_8c.html#a2aae9a568cd0cd5313808e458a9fe388" title="Comparator function for qsort." alt="" coords="433,1249,573,1291"/>
<area shape="rect" href="sim-operations_8c.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="421,967,584,1008"/>
<area shape="rect" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="670,569,801,595"/>
<area shape="rect" href="sim-operations_8c.html#af2eda46d016c384b166ce65605952b7c" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="444,1315,561,1342"/>
<area shape="rect" href="sim-operations_8c.html#a2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="680,1025,791,1051"/>
<area shape="rect" href="sim-operations_8c.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="883,1076,1039,1117"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="1126,978,1242,1005"/>
</map>
</div>

</div>
</div>
<a id="a1219ce0fef428e6b5317a034280497b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1219ce0fef428e6b5317a034280497b0">&#9670;&nbsp;</a></span>multiply_dmatrices()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> multiply_dmatrices </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>b</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Perform matrix multiplication on two <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </p>
<p>Result is returned as a new matrix.</p>
<p>If the two matrices do not have dimensions that allow them to be multiplied, it will throw an error.</p>
<p>The order of parameters a and b matters.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>pointer to the first <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </td></tr>
    <tr><td class="paramname">b</td><td>pointer to the second <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the result of the matrix multiplication (*a) x (*b), as a new <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_cgraph.png" border="0" usemap="#sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_cgraph" alt=""/></div>
<map name="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_cgraph" id="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_cgraph">
<area shape="rect" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="5,5,136,32"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="184,5,339,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_icgraph.png" border="0" usemap="#sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_icgraph" alt=""/></div>
<map name="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_icgraph" id="sim-operations_8h_a1219ce0fef428e6b5317a034280497b0_icgraph">
<area shape="rect" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="652,208,783,235"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,132,604,159"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="441,251,604,293"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,157,390,184"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,133"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,208,385,235"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,5,361,32"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,56,352,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,234,163,275"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,99,165,141"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,165,168,192"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,34,163,75"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="239,309,371,336"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="229,259,380,285"/>
</map>
</div>

</div>
</div>
<a id="ad4e0f7934e82157e3115f8df7aec01f0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad4e0f7934e82157e3115f8df7aec01f0">&#9670;&nbsp;</a></span>randb()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int randb </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Generates randomly one of 0 or 1, by splitting the output of rand() in half. </p>
<dl class="section return"><dt>Returns</dt><dd>0 or 1 at random. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad4e0f7934e82157e3115f8df7aec01f0_icgraph.png" border="0" usemap="#sim-operations_8h_ad4e0f7934e82157e3115f8df7aec01f0_icgraph" alt=""/></div>
<map name="sim-operations_8h_ad4e0f7934e82157e3115f8df7aec01f0_icgraph" id="sim-operations_8h_ad4e0f7934e82157e3115f8df7aec01f0_icgraph">
<area shape="rect" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="663,234,717,261"/>
<area shape="rect" href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="473,133,583,159"/>
<area shape="rect" href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="442,209,614,235"/>
<area shape="rect" href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="467,259,589,286"/>
<area shape="rect" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data." alt="" coords="441,311,615,352"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,50,390,77"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,158,352,185"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,209,385,235"/>
</map>
</div>

</div>
</div>
<a id="a89e066191f12e056cebb8d2cd74a2a29"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a89e066191f12e056cebb8d2cd74a2a29">&#9670;&nbsp;</a></span>randlim()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int randlim </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>limit</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Generate an integer randomly between 0 and limit inclusive. </p>
<p>Written by Jerry Coffin <a href="https://stackoverflow.com/questions/2999075/generate-a-random-number-within-range/2999130#2999130">https://stackoverflow.com/questions/2999075/generate-a-random-number-within-range/2999130#2999130</a></p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">limit</td><td>the upper limit of the range to be generated </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a random integer in the range [0,limit]. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a89e066191f12e056cebb8d2cd74a2a29_icgraph.png" border="0" usemap="#sim-operations_8h_a89e066191f12e056cebb8d2cd74a2a29_icgraph" alt=""/></div>
<map name="sim-operations_8h_a89e066191f12e056cebb8d2cd74a2a29_icgraph" id="sim-operations_8h_a89e066191f12e056cebb8d2cd74a2a29_icgraph">
<area shape="rect" title="Generate an integer randomly between 0 and limit inclusive." alt="" coords="425,13,489,39"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="207,13,377,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="a3d910624bfa7ac1bfacf706c3357fc5d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3d910624bfa7ac1bfacf706c3357fc5d">&#9670;&nbsp;</a></span>randn()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double randn </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Generates a double from the normal distribution N(0,1) based on rand() and the Box-Muller algorithm. </p>
<dl class="section return"><dt>Returns</dt><dd>a double drawn from the N(0,1) random distribution. </dd></dl>

</div>
</div>
<a id="a7d9dbe538d6fb096af14c3163c98e800"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7d9dbe538d6fb096af14c3163c98e800">&#9670;&nbsp;</a></span>randpoi()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int randpoi </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>lambda</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Generates randomly a number from the Poisson distribution with parameter <code>lambda</code>, using the Knuth approach. </p>
<p>This approach is known to run slow for larger values of lambda.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">lambda</td><td>the parameter of the Poisson distribution. Corresponds to both expected value and variance of the distribution. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a random integer generated from the Poisson distribution with parameter lambda. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a7d9dbe538d6fb096af14c3163c98e800_icgraph.png" border="0" usemap="#sim-operations_8h_a7d9dbe538d6fb096af14c3163c98e800_icgraph" alt=""/></div>
<map name="sim-operations_8h_a7d9dbe538d6fb096af14c3163c98e800_icgraph" id="sim-operations_8h_a7d9dbe538d6fb096af14c3163c98e800_icgraph">
<area shape="rect" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="661,209,725,235"/>
<area shape="rect" href="sim-operations_8h.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="472,133,583,159"/>
<area shape="rect" href="sim-operations_8h.html#ac1f46b3de0afb4674ae6d34311e7aea7" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="441,209,613,235"/>
<area shape="rect" href="sim-operations_8h.html#ae406f4a0266b7480e49899693f1b00da" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="466,259,589,286"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,50,390,77"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,158,352,185"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,209,385,235"/>
</map>
</div>

</div>
</div>
<a id="abd53d955f89343aec74bda4195aaa750"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abd53d955f89343aec74bda4195aaa750">&#9670;&nbsp;</a></span>save_all_fitness()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_all_fitness </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the GEBV of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name] [GEBV]</p>
<p>[id] [name] [GEBV]</p>
<p>...</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_abd53d955f89343aec74bda4195aaa750_cgraph.png" border="0" usemap="#sim-operations_8h_abd53d955f89343aec74bda4195aaa750_cgraph" alt=""/></div>
<map name="sim-operations_8h_abd53d955f89343aec74bda4195aaa750_cgraph" id="sim-operations_8h_abd53d955f89343aec74bda4195aaa750_cgraph">
<area shape="rect" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="5,224,119,251"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="167,173,329,200"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="393,275,500,301"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="393,5,500,32"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="564,85,719,112"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="379,107,515,149"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="381,173,512,200"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="377,224,516,251"/>
</map>
</div>

</div>
</div>
<a id="a42a237afba9dc8f3a7339cb97d929cb6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42a237afba9dc8f3a7339cb97d929cb6">&#9670;&nbsp;</a></span>save_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the geneotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file. </p>
<p>Uses the following format: </p><pre class="fragment">    [marker name]   [marker name]
</pre><p>[subject id]OR[subject name] [allele pairs for each marker]</p>
<p>[subject id]OR[subject name] [allele pairs for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. if this is null, the header row will be empty. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a42a237afba9dc8f3a7339cb97d929cb6_icgraph.png" border="0" usemap="#sim-operations_8h_a42a237afba9dc8f3a7339cb97d929cb6_icgraph" alt=""/></div>
<map name="sim-operations_8h_a42a237afba9dc8f3a7339cb97d929cb6_icgraph" id="sim-operations_8h_a42a237afba9dc8f3a7339cb97d929cb6_icgraph">
<area shape="rect" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="441,143,569,170"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="a490906a5bfdb22b90aba5cdf6282fd03"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a490906a5bfdb22b90aba5cdf6282fd03">&#9670;&nbsp;</a></span>save_count_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used: </p><pre class="fragment">    [subject id]OR[subject name]    [subject id]OR[subject name] ...
</pre><p>[marker name] [allele count for each marker]</p>
<p>[marker name] [allele count for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a490906a5bfdb22b90aba5cdf6282fd03_cgraph.png" border="0" usemap="#sim-operations_8h_a490906a5bfdb22b90aba5cdf6282fd03_cgraph" alt=""/></div>
<map name="sim-operations_8h_a490906a5bfdb22b90aba5cdf6282fd03_cgraph" id="sim-operations_8h_a490906a5bfdb22b90aba5cdf6282fd03_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in the SimData to a..." alt="" coords="5,42,136,69"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="184,5,320,47"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="199,71,305,98"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="368,13,523,39"/>
</map>
</div>

</div>
</div>
<a id="ab732ec2d648a6dc01444ef74f780fc70"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab732ec2d648a6dc01444ef74f780fc70">&#9670;&nbsp;</a></span>save_count_matrix_of_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_count_matrix_of_group </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>allele</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the number of copies of a particular allele at each marker of each genotype in a group to a file. </p>
<p>The following tab-separated format is used: </p><pre class="fragment">    [subject id]OR[subject name]    [subject id]OR[subject name] ...
</pre><p>[marker name] [allele count for each marker]</p>
<p>[marker name] [allele count for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the allele count of. </td></tr>
    <tr><td class="paramname">allele</td><td>the allele character to count </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ab732ec2d648a6dc01444ef74f780fc70_cgraph.png" border="0" usemap="#sim-operations_8h_ab732ec2d648a6dc01444ef74f780fc70_cgraph" alt=""/></div>
<map name="sim-operations_8h_ab732ec2d648a6dc01444ef74f780fc70_cgraph" id="sim-operations_8h_ab732ec2d648a6dc01444ef74f780fc70_cgraph">
<area shape="rect" title="Print the number of copies of a particular allele at each marker of each genotype in a group to a fil..." alt="" coords="5,133,152,174"/>
<area shape="rect" href="sim-operations_8c.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="200,23,356,65"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="225,89,331,116"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="227,241,329,268"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="426,191,537,217"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="215,191,341,217"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="404,5,559,32"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="423,56,539,83"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="607,56,781,83"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="439,241,524,268"/>
</map>
</div>

</div>
</div>
<a id="a8f5320627e7155d6e69e4c3db6208be4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8f5320627e7155d6e69e4c3db6208be4">&#9670;&nbsp;</a></span>save_fitness()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_fitness </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>e</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>ids</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>names</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print a set of pre-calculated GEBVs with provided names and ids to a file, with the same format as a regular call to <code>save_all_fitness</code> or <code>save_group_fitness</code>. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name] [GEBV]</p>
<p>[id] [name] [GEBV]</p>
<p>...</p>
<p>The function assumes the array of ids, array of names, and columns of the <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> are ordered the same, so a single index produces the corresponding value from each.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">e</td><td>pointer to the <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> containing the GEBVs in the first row. </td></tr>
    <tr><td class="paramname">ids</td><td>array of ids to print alongside the GEBVs. </td></tr>
    <tr><td class="paramname">names</td><td>array of names to print alongside the GEBVs. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a8f5320627e7155d6e69e4c3db6208be4_icgraph.png" border="0" usemap="#sim-operations_8h_a8f5320627e7155d6e69e4c3db6208be4_icgraph" alt=""/></div>
<map name="sim-operations_8h_a8f5320627e7155d6e69e4c3db6208be4_icgraph" id="sim-operations_8h_a8f5320627e7155d6e69e4c3db6208be4_icgraph">
<area shape="rect" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="441,143,535,170"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="aad2fb503b61c909787f62b2a64385557"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aad2fb503b61c909787f62b2a64385557">&#9670;&nbsp;</a></span>save_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>parents</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p>[id] [name]=([parent pedigree])</p>
<p>...</p>
<p>Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p>The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<p>Note this does not follow through the linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> containing the genotypes to print </td></tr>
    <tr><td class="paramname">parents</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given genotypes. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_cgraph.png" border="0" usemap="#sim-operations_8h_aad2fb503b61c909787f62b2a64385557_cgraph" alt=""/></div>
<map name="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_cgraph" id="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="5,19,132,45"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="348,5,461,32"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="180,56,295,83"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="515,31,689,57"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="343,56,467,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_icgraph.png" border="0" usemap="#sim-operations_8h_aad2fb503b61c909787f62b2a64385557_icgraph" alt=""/></div>
<map name="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_icgraph" id="sim-operations_8h_aad2fb503b61c909787f62b2a64385557_icgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="441,143,568,170"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="a9e43115d9ee4ac406c14161fc6ccb327"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9e43115d9ee4ac406c14161fc6ccb327">&#9670;&nbsp;</a></span>save_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">    [marker name]   [marker name]
</pre><p>[subject id]OR[subject name] [allele pairs for each marker]</p>
<p>[subject id]OR[subject name] [allele pairs for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a9e43115d9ee4ac406c14161fc6ccb327_cgraph.png" border="0" usemap="#sim-operations_8h_a9e43115d9ee4ac406c14161fc6ccb327_cgraph" alt=""/></div>
<map name="sim-operations_8h_a9e43115d9ee4ac406c14161fc6ccb327_cgraph" id="sim-operations_8h_a9e43115d9ee4ac406c14161fc6ccb327_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,50,136,77"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="186,25,307,51"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="357,25,468,51"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="195,75,298,102"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="184,126,309,153"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="370,75,455,102"/>
</map>
</div>

</div>
</div>
<a id="a4d37c413cb4c979b4eafa1d0b656edd1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4d37c413cb4c979b4eafa1d0b656edd1">&#9670;&nbsp;</a></span>save_group_fitness()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_fitness </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the GEBV of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[group member id] [group member name] [GEBV]</p>
<p>[group member id] [group member name] [GEBV]</p>
<p>...</p>
<p>The <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> must have loaded marker effects for this function to succeed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the GEBVs of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a4d37c413cb4c979b4eafa1d0b656edd1_cgraph.png" border="0" usemap="#sim-operations_8h_a4d37c413cb4c979b4eafa1d0b656edd1_cgraph" alt=""/></div>
<map name="sim-operations_8h_a4d37c413cb4c979b4eafa1d0b656edd1_cgraph" id="sim-operations_8h_a4d37c413cb4c979b4eafa1d0b656edd1_cgraph">
<area shape="rect" title="Print the GEBV of each genotype in a group to a file." alt="" coords="5,376,137,403"/>
<area shape="rect" href="sim-operations_8c.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="185,209,348,250"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="421,5,527,32"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="423,427,525,453"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="622,401,733,428"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="411,477,537,504"/>
<area shape="rect" href="sim-operations_8c.html#a2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="419,56,529,83"/>
<area shape="rect" href="sim-operations_8c.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="396,209,552,250"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="600,157,755,184"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="409,157,539,184"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="405,325,543,352"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="619,216,735,243"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="803,216,977,243"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="635,464,720,491"/>
</map>
</div>

</div>
</div>
<a id="a75e854cbc72b31a5e2eb03c63c33ecc6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a75e854cbc72b31a5e2eb03c63c33ecc6">&#9670;&nbsp;</a></span>save_group_full_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_full_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the full known pedigree of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[id] [name]=([parent 1 pedigree],[parent 2 pedigree])</p>
<p>[id] [name]=([parent pedigree])</p>
<p>...</p>
<p>Note that this pedigree is recursively constructed, so if a genotype's parents are known, [parent pedigree] is replaced with a string of format name=([parent1 pedigree],[parent2 pedigree]) alike. If the two parents of a genotype are the same individual (i.e. it was produced by selfing, the comma and second parent pedigree are ommitted, as in the second line sample in the format above.</p>
<p>The parents of a genotype are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0, the individual is printed without brackets or parent pedigrees and recursion stops here.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the pedigree of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a75e854cbc72b31a5e2eb03c63c33ecc6_cgraph.png" border="0" usemap="#sim-operations_8h_a75e854cbc72b31a5e2eb03c63c33ecc6_cgraph" alt=""/></div>
<map name="sim-operations_8h_a75e854cbc72b31a5e2eb03c63c33ecc6_cgraph" id="sim-operations_8h_a75e854cbc72b31a5e2eb03c63c33ecc6_cgraph">
<area shape="rect" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="5,49,172,76"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="231,24,334,51"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="400,24,511,51"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="220,75,345,101"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="225,149,340,176"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="413,75,498,101"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="399,137,512,164"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="393,188,517,215"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="565,163,740,189"/>
</map>
</div>

</div>
</div>
<a id="a979efe39128de97176da5460cab7dfa7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a979efe39128de97176da5460cab7dfa7">&#9670;&nbsp;</a></span>save_group_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_group_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in a group to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[group member name] [parent 1 name] [parent 2 name]</p>
<p>[group member name] [parent 1 name] [parent 2 name]</p>
<p>...</p>
<p>The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id of an individual is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the group members. </td></tr>
    <tr><td class="paramname">group</td><td>group number of the group of individuals to print the immediate parents of. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a979efe39128de97176da5460cab7dfa7_cgraph.png" border="0" usemap="#sim-operations_8h_a979efe39128de97176da5460cab7dfa7_cgraph" alt=""/></div>
<map name="sim-operations_8h_a979efe39128de97176da5460cab7dfa7_cgraph" id="sim-operations_8h_a979efe39128de97176da5460cab7dfa7_cgraph">
<area shape="rect" title="Print the parents of each genotype in a group to a file." alt="" coords="5,71,152,112"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="211,78,314,105"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="405,28,516,54"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="200,28,325,54"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="206,180,319,206"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="201,129,325,156"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="418,78,503,105"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="373,154,548,181"/>
</map>
</div>

</div>
</div>
<a id="a6b87b484064fd373aec0bfa4ce8dbef1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6b87b484064fd373aec0bfa4ce8dbef1">&#9670;&nbsp;</a></span>save_one_step_pedigree()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_one_step_pedigree </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>The following tab-separated format is used:</p>
<p>[genotype name] [parent 1 name] [parent 2 name]</p>
<p>[genotype name] [parent 1 name] [parent 2 name]</p>
<p>...</p>
<p>The parents are identified by the two ids saved in the genotype's pedigrees field in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> struct.</p>
<p>If a group member or parent has no name, the name will be replaced in the output file with its session-unique id. If the parent id is 0 (which means the parent is unknown) no name or id is printed for that parent.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes and their pedigrees </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a6b87b484064fd373aec0bfa4ce8dbef1_cgraph.png" border="0" usemap="#sim-operations_8h_a6b87b484064fd373aec0bfa4ce8dbef1_cgraph" alt=""/></div>
<map name="sim-operations_8h_a6b87b484064fd373aec0bfa4ce8dbef1_cgraph" id="sim-operations_8h_a6b87b484064fd373aec0bfa4ce8dbef1_cgraph">
<area shape="rect" title="Print the parents of each genotype in the SimData to a file." alt="" coords="5,31,169,57"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="223,5,336,32"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="217,56,341,83"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="389,31,564,57"/>
</map>
</div>

</div>
</div>
<a id="a4a41a75435588bd183d4d4d91082e601"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4a41a75435588bd183d4d4d91082e601">&#9670;&nbsp;</a></span>save_parents_of()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_parents_of </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Recursively save the parents of a particular id to a file. </p>
<p>It saves using the following format:</p>
<ul>
<li>no characters are saved if the parents of the id are unknown/0</li>
<li>if the id has one parent, repeated twice (it was produced by selfing), print "=(parentname)", where parentname is the name of the parent or its id if it does not have one, followed by whatever is printed by a call to this function on the parent's id.</li>
<li>if the id has two separate parents, print "=(parent1name,parent2name)", where parent1name and parent2name are the name2 of the two parents or their ids if they does not have names, each name immediately followed by whatever is printed by a call to this function on the corresponding parent's id.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> that heads the linked list containing the parents and other ancestry of the given id. </td></tr>
    <tr><td class="paramname">id</td><td>the session-unique id of the genotype whose parents we wish to recursively save. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_cgraph.png" border="0" usemap="#sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_cgraph" alt=""/></div>
<map name="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_cgraph" id="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_cgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="5,48,120,74"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="336,22,449,49"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="331,73,455,100"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="168,48,283,74"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="503,48,677,74"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_icgraph.png" border="0" usemap="#sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_icgraph" alt=""/></div>
<map name="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_icgraph" id="sim-operations_8h_a4a41a75435588bd183d4d4d91082e601_icgraph">
<area shape="rect" title="Recursively save the parents of a particular id to a file." alt="" coords="656,158,771,185"/>
<area shape="rect" href="sim-operations_8h.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="461,133,588,159"/>
<area shape="rect" href="sim-operations_8h.html#a75e854cbc72b31a5e2eb03c63c33ecc6" title="Print the full known pedigree of each genotype in a group to a file." alt="" coords="441,183,608,210"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,34,390,61"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,134"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,158,385,185"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,209,352,235"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="a4d63cd249222ab25b3112106813f29b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4d63cd249222ab25b3112106813f29b2">&#9670;&nbsp;</a></span>save_simdata()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_simdata </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>m</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file. </p>
<p>Column separators are tabs.</p>
<p>The printing format is:</p>
<p>name chr pos [allele to which these effects apply, 0+ columns] <br  />
</p>
<p>[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p>[marker name] [chr number] [chr pos] [effects, 0+ columns]</p>
<p>...</p>
<p>If m-&gt;effects is NULL, m-&gt;ref_alleles is NULL, or m-&gt;genetic_map is NULL, then the relevant columns are omitted.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> whose data we print </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="aadf131f9f90546d6ab9355cbaa1cf6e1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aadf131f9f90546d6ab9355cbaa1cf6e1">&#9670;&nbsp;</a></span>save_transposed_allele_matrix()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_allele_matrix </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>markers</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints all the gene data saved in the linked list starting with <code>m</code> to the file. </p>
<p>Uses the following format: </p><pre class="fragment">    [subject id]OR[subject name]    [subject id]OR[subject name] ...
</pre><p>[marker name] [allele pairs for each marker]</p>
<p>[marker name] [allele pairs for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose data we print </td></tr>
    <tr><td class="paramname">markers</td><td>array of strings that correspond to names of the markers. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a0f53ee5421c91ca041a1b8d4603238b5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0f53ee5421c91ca041a1b8d4603238b5">&#9670;&nbsp;</a></span>save_transposed_group_alleles()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void save_transposed_group_alleles </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Prints the genotypes of each individual in a given group to a file, with the following format. </p>
<pre class="fragment">[subject id]OR[subject name]    [subject id]OR[subject name] ...
</pre><p>[marker name] [allele pairs for each marker]</p>
<p>[marker name] [allele pairs for each marker]</p>
<p>...</p>
<p>Subject id will be printed if the genotype does not have a name saved</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">f</td><td>file pointer opened for writing to put the output </td></tr>
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> containing the genotypes of the group and the marker names. </td></tr>
    <tr><td class="paramname">group_id</td><td>group number of the group of individuals whose genotypes to print. </td></tr>
  </table>
  </dd>
</dl>
<p>fprintf(f, "\n");</p>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_a0f53ee5421c91ca041a1b8d4603238b5_cgraph.png" border="0" usemap="#sim-operations_8h_a0f53ee5421c91ca041a1b8d4603238b5_cgraph" alt=""/></div>
<map name="sim-operations_8h_a0f53ee5421c91ca041a1b8d4603238b5_cgraph" id="sim-operations_8h_a0f53ee5421c91ca041a1b8d4603238b5_cgraph">
<area shape="rect" title="Prints the genotypes of each individual in a given group to a file, with the following format." alt="" coords="5,42,164,84"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="214,24,335,51"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="385,24,496,51"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="223,75,326,102"/>
<area shape="rect" href="sim-operations_8c.html#a3e00f06acc41eb012e99ae46d7a18ccc" title="Gets a shallow copy of the names of each member of the group." alt="" coords="212,126,337,152"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="398,75,483,102"/>
</map>
</div>

</div>
</div>
<a id="adf5beb682ac15fa5a0176e480adc9d91"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf5beb682ac15fa5a0176e480adc9d91">&#9670;&nbsp;</a></span>self_n_times()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int self_n_times </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Selfs each member of a group for a certain number of generations. </p>
<p>The resulting genotypes are allocated to a new group.</p>
<p>Only the genotype after all n generations is saved. Intermediate steps will be lost.</p>
<p>Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this operation. The family_size parameter in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> allows you to generate multiple selfed offspring from each member of the group. These multiple selfed offspring all originate from independent processes of selfing the parent with itself then selfing that intermediate offspring with itself for the remaining n-1 generations.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that contains the genetic map and genotypes of the parent group. </td></tr>
    <tr><td class="paramname">n</td><td>number of generations of selfing simulation to carry out. </td></tr>
    <tr><td class="paramname">group</td><td>the genotypes on which to perform these n generations of selfing. </td></tr>
    <tr><td class="paramname">g</td><td>options for the genotypes created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_adf5beb682ac15fa5a0176e480adc9d91_cgraph.png" border="0" usemap="#sim-operations_8h_adf5beb682ac15fa5a0176e480adc9d91_cgraph" alt=""/></div>
<map name="sim-operations_8h_adf5beb682ac15fa5a0176e480adc9d91_cgraph" id="sim-operations_8h_adf5beb682ac15fa5a0176e480adc9d91_cgraph">
<area shape="rect" title="Selfs each member of a group for a certain number of generations." alt="" coords="5,578,100,605"/>
<area shape="rect" href="sim-operations_8c.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="154,173,317,199"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="407,274,514,301"/>
<area shape="rect" href="sim-operations_8c.html#a887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="156,325,315,351"/>
<area shape="rect" href="sim-operations_8c.html#add4a9c17e8c3e59fc00eee25635b110f" title="Deletes the full AlleleMatrix object and frees its memory." alt="" coords="392,325,529,351"/>
<area shape="rect" href="sim-operations_8c.html#aac694dd044a29fd93443368a7b9e8d8a" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="148,477,323,503"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="418,553,503,579"/>
<area shape="rect" href="sim-operations_8c.html#aa18b10bcd660dee5213868cc3244f826" title="Get the alleles of the outcome of crossing two subjects." alt="" coords="180,426,291,453"/>
<area shape="rect" href="sim-operations_8c.html#ac159a31261e14b325bd623899eefa796" title="Gets a shallow copy of the genes/alleles of each member of the group." alt="" coords="175,629,296,655"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="405,629,516,655"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="184,578,287,605"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="165,781,306,807"/>
<area shape="rect" href="sim-operations_8c.html#a42a237afba9dc8f3a7339cb97d929cb6" title="Prints all the geneotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="171,831,299,858"/>
<area shape="rect" href="sim-operations_8c.html#a8f5320627e7155d6e69e4c3db6208be4" title="Print a set of pre&#45;calculated GEBVs with provided names and ids to a file, with the same format as a ..." alt="" coords="189,882,282,909"/>
<area shape="rect" href="sim-operations_8c.html#aad2fb503b61c909787f62b2a64385557" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="172,933,299,959"/>
<area shape="rect" href="sim-operations_8c.html#ab4a5ef571ebeb5e546461398eecfb0ff" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="168,730,303,757"/>
<area shape="rect" href="sim-operations_8c.html#ac711a312bddd30d394adda822a7c7679" title="Add the two DecimalMatrix." alt="" coords="407,223,514,250"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="607,122,761,149"/>
<area shape="rect" href="sim-operations_8c.html#a42f2398ccb53ffafce9b4a17e3222636" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="393,5,529,47"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="395,71,526,98"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="391,122,530,149"/>
<area shape="rect" href="sim-operations_8c.html#a3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="371,375,551,402"/>
<area shape="rect" href="sim-operations_8c.html#ad4e0f7934e82157e3115f8df7aec01f0" title="Generates randomly one of 0 or 1, by splitting the output of rand() in half." alt="" coords="433,426,488,453"/>
<area shape="rect" href="sim-operations_8c.html#a7d9dbe538d6fb096af14c3163c98e800" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="429,477,493,503"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="391,781,530,807"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="599,781,769,807"/>
<area shape="rect" href="sim-operations_8c.html#a6aa75e72bb29725ffe32cd003c547480" title="Returns the name of the genotype with a given id." alt="" coords="627,919,741,946"/>
<area shape="rect" href="sim-operations_8c.html#a4a41a75435588bd183d4d4d91082e601" title="Recursively save the parents of a particular id to a file." alt="" coords="403,970,518,997"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="817,945,992,971"/>
<area shape="rect" href="sim-operations_8c.html#a229c9de94875ea41a23723d337f5f92f" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="622,970,746,997"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="399,730,523,757"/>
</map>
</div>

</div>
</div>
<a id="ace12db201ad55ffc72c26a76ee8fbfc5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ace12db201ad55ffc72c26a76ee8fbfc5">&#9670;&nbsp;</a></span>set_group_list()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void set_group_list </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>by_n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>new_group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ac5f92e5e4bd9d414392a2f022fdc5596"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac5f92e5e4bd9d414392a2f022fdc5596">&#9670;&nbsp;</a></span>set_subject_ids()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void set_subject_ids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>from_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>to_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Allocate lifetime-unique ids to each genotype in the range of whole <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> indexes <code>from_index</code> to <code>to_index</code> inclusive. </p>
<p>Not intended to be called by an end user.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform actions </td></tr>
    <tr><td class="paramname">from_index</td><td>the starting 0-based index of the range to allocate ids </td></tr>
    <tr><td class="paramname">to_index</td><td>the last 0-based index in the range to allocate ids </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ac5f92e5e4bd9d414392a2f022fdc5596_icgraph.png" border="0" usemap="#sim-operations_8h_ac5f92e5e4bd9d414392a2f022fdc5596_icgraph" alt=""/></div>
<map name="sim-operations_8h_ac5f92e5e4bd9d414392a2f022fdc5596_icgraph" id="sim-operations_8h_ac5f92e5e4bd9d414392a2f022fdc5596_icgraph">
<area shape="rect" title="Allocate lifetime&#45;unique ids to each genotype in the range of whole SimData indexes from_index to to_..." alt="" coords="393,78,507,105"/>
<area shape="rect" href="sim-operations_8h.html#a8bd85cb47ec8e0c4049f937ef40ae102" title="Appends genotype data from a file to an existing SimData." alt="" coords="182,5,335,47"/>
<area shape="rect" href="sim-operations_8h.html#ad5863641a67355c2c1523e1a23b6f581" title="Populates a SimData combination with marker allele data." alt="" coords="172,71,345,112"/>
<area shape="rect" href="sim-operations_8h.html#a156666e21151a9f267e31598574366a6" title="Populates a SimData combination with marker allele data." alt="" coords="179,136,338,177"/>
<area shape="rect" href="sim-operations_8h.html#a54cd625669413c2ef171060e9844d814" title="Populates a SimData combination from scratch with marker allele data, a genetic map,..." alt="" coords="5,143,124,170"/>
</map>
</div>

</div>
</div>
<a id="ab4a5ef571ebeb5e546461398eecfb0ff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab4a5ef571ebeb5e546461398eecfb0ff">&#9670;&nbsp;</a></span>set_subject_names()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void set_subject_names </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>prefix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>suffix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>from_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Fills the designated section of the <code>.subject_names</code> array in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> with the pattern <code>prefix</code>index. </p>
<p>This function is not intended to be called by an end user.</p>
<p>In future this function could be expanded to allow for different naming formats.</p>
<p>The index is padded with zeros depending on the size of <code>a-&gt;n_subjects</code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>pointer to the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> whose <code>.subject_names</code> to modify </td></tr>
    <tr><td class="paramname">prefix</td><td>the prefix to add to the suffix to make the new subject name </td></tr>
    <tr><td class="paramname">suffix</td><td>suffixes start at this value and increment for each additional subject_name </td></tr>
    <tr><td class="paramname">from_index</td><td>the new names are added to this index and all those following it in this <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>. </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_cgraph.png" border="0" usemap="#sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_cgraph" alt=""/></div>
<map name="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_cgraph" id="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_cgraph">
<area shape="rect" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="5,31,140,57"/>
<area shape="rect" href="sim-operations_8c.html#ad7d864d43092823ac2ed8a6cbd595e35" title="Count and return the number of digits in i." alt="" coords="188,5,312,32"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="207,56,293,83"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_icgraph.png" border="0" usemap="#sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_icgraph" alt=""/></div>
<map name="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_icgraph" id="sim-operations_8h_ab4a5ef571ebeb5e546461398eecfb0ff_icgraph">
<area shape="rect" title="Fills the designated section of the .subject_names array in an AlleleMatrix with the pattern prefixin..." alt="" coords="441,143,576,170"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,31,390,58"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,118,393,145"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,169,385,195"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,219,352,246"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,5,163,47"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,71,165,112"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,137,168,163"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,188,163,229"/>
</map>
</div>

</div>
</div>
<a id="af2eda46d016c384b166ce65605952b7c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af2eda46d016c384b166ce65605952b7c">&#9670;&nbsp;</a></span>split_from_group()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int split_from_group </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>indexes_to_split</em>[n]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Take a list of indexes and allocate the genotypes at those indexes to a new group. </p>
<p>Does not check if all the indexes are valid/if all indexes have successfully had their groups changed.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">n</td><td>the number of indexes provided </td></tr>
    <tr><td class="paramname">indexes_to_split</td><td>an array containing the indexes (0-based, starting at the first entry at <code>d-&gt;m</code>) of the genotypes to allocate to the new group. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the new group to which the provided indexes have been allocated. </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_cgraph.png" border="0" usemap="#sim-operations_8h_af2eda46d016c384b166ce65605952b7c_cgraph" alt=""/></div>
<map name="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_cgraph" id="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_cgraph">
<area shape="rect" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="5,5,123,32"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="171,5,312,32"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="360,5,499,32"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="547,5,717,32"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_icgraph.png" border="0" usemap="#sim-operations_8h_af2eda46d016c384b166ce65605952b7c_icgraph" alt=""/></div>
<map name="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_icgraph" id="sim-operations_8h_af2eda46d016c384b166ce65605952b7c_icgraph">
<area shape="rect" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="405,13,523,39"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,13,357,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="aa91ffff83cf204fff8baef3790738b7a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa91ffff83cf204fff8baef3790738b7a">&#9670;&nbsp;</a></span>split_group_by_fitness()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int split_group_by_fitness </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>top_n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>lowIsBest</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Takes the top <code>top_n</code> fitness/GEBV individuals in the group and puts them in a new group. </p>
<p>The new group number is returned.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object to which the groups and individuals belong. It must have a marker effect file loaded to successfully run this function. </td></tr>
    <tr><td class="paramname">group</td><td>group number from which to split the top individuals. </td></tr>
    <tr><td class="paramname">top_n</td><td>The number of individuals to put in the new group. </td></tr>
    <tr><td class="paramname">lowIsBest</td><td>boolean, if TRUE the <code>top_n</code> with the lowest fitness/GEBV score will be selected, if false the <code>top_n</code> with the highest fitness/GEBV score are. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the newly-created split-off group </dd></dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_cgraph.png" border="0" usemap="#sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_cgraph" alt=""/></div>
<map name="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_cgraph" id="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_cgraph">
<area shape="rect" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="5,229,156,256"/>
<area shape="rect" href="sim-operations_8c.html#a311381904a1ca73ea26dfede9fcddea2" title="Comparator function for qsort." alt="" coords="204,83,399,109"/>
<area shape="rect" href="sim-operations_8c.html#a2aae9a568cd0cd5313808e458a9fe388" title="Comparator function for qsort." alt="" coords="231,134,371,175"/>
<area shape="rect" href="sim-operations_8c.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="220,251,383,293"/>
<area shape="rect" href="sim-operations_8c.html#aaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="471,427,578,453"/>
<area shape="rect" href="sim-operations_8c.html#a95aad1344c24e6072e6a50dcc67b60a8" title="Gets the indexes (0&#45;based, from the start of the linked list in the SimData) of each member of the gr..." alt="" coords="459,5,590,32"/>
<area shape="rect" href="sim-operations_8c.html#af2eda46d016c384b166ce65605952b7c" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="243,449,360,476"/>
<area shape="rect" href="sim-operations_8c.html#a2c58c3fef64a20bca3191f343b947933" title="Add the two DecimalMatrix." alt="" coords="469,208,580,235"/>
<area shape="rect" href="sim-operations_8c.html#ae7797aeac935227749fc6eae09a02a96" title="Calculates the number of times at each marker that a particular allele appears for each genotype is a..." alt="" coords="447,361,603,402"/>
<area shape="rect" href="sim-operations_8c.html#abf3f578b0aa59d5b68ccd0738dd6a9d1" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="651,309,805,336"/>
<area shape="rect" href="sim-operations_8c.html#a961042e468630f31a3607dac42f8fa34" title="Gets the ids of each member of the group." alt="" coords="473,56,576,83"/>
<area shape="rect" href="sim-operations_8c.html#ad6d2b66296cd80f5bfe303e06de7935c" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="673,56,783,83"/>
<area shape="rect" href="sim-operations_8c.html#a1219ce0fef428e6b5317a034280497b0" title="Perform matrix multiplication on two DecimalMatrix." alt="" coords="459,259,590,285"/>
<area shape="rect" href="sim-operations_8c.html#abe8da784853a382f703129997ed08996" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="455,157,594,184"/>
<area shape="rect" href="sim-operations_8c.html#af49ac14d9ff4483bce6424f823d30ca0" title="Returns the alleles at each marker of the genotype with a given id." alt="" coords="670,368,786,395"/>
<area shape="rect" href="sim-operations_8c.html#aa6943b0633c1744bacad62ab2bf68ebd" title="Returns the located index in an array of integers where the integer is target." alt="" coords="853,368,1028,395"/>
<area shape="rect" href="sim-operations_8c.html#ad1a5b9f3e850da13931ee02fbb170d41" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="685,5,771,32"/>
<area shape="rect" href="sim-operations_8c.html#ae974a74145fe876e303e14336d3c5d3a" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="454,477,595,504"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="659,477,797,504"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="855,477,1026,504"/>
</map>
</div>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_icgraph.png" border="0" usemap="#sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_icgraph" alt=""/></div>
<map name="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_icgraph" id="sim-operations_8h_aa91ffff83cf204fff8baef3790738b7a_icgraph">
<area shape="rect" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="207,13,357,39"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="5,5,159,47"/>
</map>
</div>

</div>
</div>
<a id="ad5a61d77e05e6bb0d8f4235d21b8b1f3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad5a61d77e05e6bb0d8f4235d21b8b1f3">&#9670;&nbsp;</a></span>split_into_individuals()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void split_into_individuals </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>group_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Give every individual in the group a new group number that does not belong to any other existing group (thereby allocating each genotype in the group to a new group of 1). </p>
<p>Note: this does not return the group numbers of all the newly created groups-of-one.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> struct on which to perform the operation </td></tr>
    <tr><td class="paramname">group_id</td><td>the group number of the group to be split </td></tr>
  </table>
  </dd>
</dl>
<div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_ad5a61d77e05e6bb0d8f4235d21b8b1f3_cgraph.png" border="0" usemap="#sim-operations_8h_ad5a61d77e05e6bb0d8f4235d21b8b1f3_cgraph" alt=""/></div>
<map name="sim-operations_8h_ad5a61d77e05e6bb0d8f4235d21b8b1f3_cgraph" id="sim-operations_8h_ad5a61d77e05e6bb0d8f4235d21b8b1f3_cgraph">
<area shape="rect" title="Give every individual in the group a new group number that does not belong to any other existing grou..." alt="" coords="5,5,145,32"/>
<area shape="rect" href="sim-operations_8c.html#a3305ddcc29138d2d92689f39a7885ca0" title="Identify every group number that currently has members." alt="" coords="193,5,332,32"/>
<area shape="rect" href="sim-operations_8c.html#adae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="380,5,551,32"/>
</map>
</div>

</div>
</div>
<a id="abe8da784853a382f703129997ed08996"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abe8da784853a382f703129997ed08996">&#9670;&nbsp;</a></span>subset_dmatrix_row()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> subset_dmatrix_row </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>row_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get a <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> composed only of row <code>row_index</code> from <code>m</code> </p>
<p>Checks that <code>row_index</code> is a valid index. it will throw an error if the index is not valid</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">m</td><td>pointer to the <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a> whose subset we are taking. </td></tr>
    <tr><td class="paramname">row_index</td><td>index in m of the row/first index from m to take as a subset </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the row of <code>m</code> at <code>row_index</code> as a new independent <a class="el" href="struct_decimal_matrix.html" title="A heap matrix that contains floating point numbers.">DecimalMatrix</a>. </dd></dl>
<div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="sim-operations_8h_abe8da784853a382f703129997ed08996_icgraph.png" border="0" usemap="#sim-operations_8h_abe8da784853a382f703129997ed08996_icgraph" alt=""/></div>
<map name="sim-operations_8h_abe8da784853a382f703129997ed08996_icgraph" id="sim-operations_8h_abe8da784853a382f703129997ed08996_icgraph">
<area shape="rect" title="Get a DecimalMatrix composed only of row row_index from m" alt="" coords="652,208,791,235"/>
<area shape="rect" href="sim-operations_8h.html#a238032359089625f3fd3b4447cdbbdff" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix, and returns the results in ..." alt="" coords="441,132,604,159"/>
<area shape="rect" href="sim-operations_8h.html#ac297afb7cf2c56f6f3f436090d7962dd" title="Calculates the fitness metric/GEBV for each genotype in the AlleleMatrix in a certain group,..." alt="" coords="441,251,604,293"/>
<area shape="rect" href="sim-operations_8h.html#a3ea77c76c329625646419a10683669f9" title="Performs random crosses among members of a group." alt="" coords="219,157,390,184"/>
<area shape="rect" href="sim-operations_8h.html#abb793988799d792dfe4007efdf988068" title="Performs the crosses of pairs of parents whose ids are provided in an array." alt="" coords="216,107,393,133"/>
<area shape="rect" href="sim-operations_8h.html#af078d6bc1b5d4234a9b960e6132480aa" title="Creates a doubled haploid from each member of a group." alt="" coords="224,208,385,235"/>
<area shape="rect" href="sim-operations_8h.html#abd53d955f89343aec74bda4195aaa750" title="Print the GEBV of each genotype in the SimData to a file." alt="" coords="248,5,361,32"/>
<area shape="rect" href="sim-operations_8h.html#adf5beb682ac15fa5a0176e480adc9d91" title="Selfs each member of a group for a certain number of generations." alt="" coords="257,56,352,83"/>
<area shape="rect" href="sim-operations_8h.html#a68aca99a41527b019316c8243e6fc7a7" title="Find the top m percent of a group and perform random crosses between those top individuals." alt="" coords="10,234,163,275"/>
<area shape="rect" href="sim-operations_8h.html#a3f1b06b7aa35e6a05918e16d1deabeee" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,99,165,141"/>
<area shape="rect" href="sim-operations_8h.html#a3b67d38c13c7a8bef2d0fc91c0365177" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,165,168,192"/>
<area shape="rect" href="sim-operations_8h.html#aa7c90b4015e4abce750176c6d891cad2" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="11,34,163,75"/>
<area shape="rect" href="sim-operations_8h.html#a4d37c413cb4c979b4eafa1d0b656edd1" title="Print the GEBV of each genotype in a group to a file." alt="" coords="239,309,371,336"/>
<area shape="rect" href="sim-operations_8h.html#aa91ffff83cf204fff8baef3790738b7a" title="Takes the top top_n fitness/GEBV individuals in the group and puts them in a new group." alt="" coords="229,259,380,285"/>
</map>
</div>

</div>
</div>
<h2 class="groupheader">Variable Documentation</h2>
<a id="a5c6b92aede8f29f53476aa78d4e4133a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5c6b92aede8f29f53476aa78d4e4133a">&#9670;&nbsp;</a></span>BASIC_OPT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const <a class="el" href="struct_gen_options.html">GenOptions</a> BASIC_OPT</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a74936c18ca56d0cfa05dc9d786288c17"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a74936c18ca56d0cfa05dc9d786288c17">&#9670;&nbsp;</a></span>EMPTY_SIMDATA</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const <a class="el" href="struct_sim_data.html">SimData</a> EMPTY_SIMDATA</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a6db8fbe9028cc259012cbdf075c222b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6db8fbe9028cc259012cbdf075c222b2">&#9670;&nbsp;</a></span>RAND_HALFPOINT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int RAND_HALFPOINT</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
