<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: sim-operations.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.12</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle"><div class="title">sim-operations.c File Reference</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><code>#include &quot;<a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>&quot;</code><br />
</div>
<p><a href="sim-operations_8c_source.html">Go to the source code of this file.</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">union &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="define-members" name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:ad84149faf0b815bcee066748ef31eb60"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ad84149faf0b815bcee066748ef31eb60">SIM_OPERATIONS</a></td></tr>
<tr class="separator:ad84149faf0b815bcee066748ef31eb60"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:aaaf3ebc75feaf901829232712e55677b"><td class="memItemLeft" align="right" valign="top">static void *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b">gsc_malloc_wrap</a> (const size_t size, char exitonfail)</td></tr>
<tr class="memdesc:aaaf3ebc75feaf901829232712e55677b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace calls to malloc direct with this function.  <a href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b">More...</a><br /></td></tr>
<tr class="separator:aaaf3ebc75feaf901829232712e55677b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga104fcea7a8f69cb80c136ab45c4eab76"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76">gsc_create_empty_allelematrix</a> (const unsigned int n_markers, const unsigned int n_labels, const int *labelDefaults, const unsigned int n_genotypes)</td></tr>
<tr class="memdesc:ga104fcea7a8f69cb80c136ab45c4eab76"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> object of a given size.  <a href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76">More...</a><br /></td></tr>
<tr class="separator:ga104fcea7a8f69cb80c136ab45c4eab76"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab37984b371a35d44dbce9d954bbf2366"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gab37984b371a35d44dbce9d954bbf2366">gsc_create_empty_simdata</a> (unsigned int RNGseed)</td></tr>
<tr class="memdesc:gab37984b371a35d44dbce9d954bbf2366"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object on the heap.  <a href="group__loaders.html#gab37984b371a35d44dbce9d954bbf2366">More...</a><br /></td></tr>
<tr class="separator:gab37984b371a35d44dbce9d954bbf2366"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga982b122b76326694fd0f26babeb027e7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga982b122b76326694fd0f26babeb027e7">gsc_clear_simdata</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga982b122b76326694fd0f26babeb027e7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object on the heap.  <a href="group__loaders.html#ga982b122b76326694fd0f26babeb027e7">More...</a><br /></td></tr>
<tr class="separator:ga982b122b76326694fd0f26babeb027e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga897c7abb536f30f0cbc1baf2abcd2067"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga897c7abb536f30f0cbc1baf2abcd2067">gsc_randpoi</a> (rnd_pcg_t *rng, double lambda)</td></tr>
<tr class="memdesc:ga897c7abb536f30f0cbc1baf2abcd2067"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly a number from the Poisson distribution with parameter <code>lambda</code>, using the Knuth approach.  <a href="group__maths.html#ga897c7abb536f30f0cbc1baf2abcd2067">More...</a><br /></td></tr>
<tr class="separator:ga897c7abb536f30f0cbc1baf2abcd2067"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga31fba32c2f869b50d97669e3e322d338"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="structgsc___table_size.html">gsc_TableSize</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga31fba32c2f869b50d97669e3e322d338">gsc_get_file_dimensions</a> (const char *filename, const char sep)</td></tr>
<tr class="memdesc:ga31fba32c2f869b50d97669e3e322d338"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens a table file and reads the number of columns and rows (including headers) separated by <code>sep</code> into a <a class="el" href="structgsc___table_size.html">gsc_TableSize</a> struct that is returned.  <a href="group__supporters.html#ga31fba32c2f869b50d97669e3e322d338">More...</a><br /></td></tr>
<tr class="separator:ga31fba32c2f869b50d97669e3e322d338"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga094a7c1d7f83050cce4d00fdcc89d6b3"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga094a7c1d7f83050cce4d00fdcc89d6b3">gsc_get_from_ordered_pedigree_list</a> (const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> target, const unsigned int listLen, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *list)</td></tr>
<tr class="memdesc:ga094a7c1d7f83050cce4d00fdcc89d6b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary search through list of unsigned integers.  <a href="group__supporters.html#ga094a7c1d7f83050cce4d00fdcc89d6b3">More...</a><br /></td></tr>
<tr class="separator:ga094a7c1d7f83050cce4d00fdcc89d6b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga60214f1e80d453c7a1fe6fdae4b6c36d"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga60214f1e80d453c7a1fe6fdae4b6c36d">gsc_get_from_unordered_str_list</a> (const char *target, const size_t listLen, const char **list)</td></tr>
<tr class="memdesc:ga60214f1e80d453c7a1fe6fdae4b6c36d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Linear search through a list of strings.  <a href="group__supporters.html#ga60214f1e80d453c7a1fe6fdae4b6c36d">More...</a><br /></td></tr>
<tr class="separator:ga60214f1e80d453c7a1fe6fdae4b6c36d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5230d5bd9791204d2dded825ee280074"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga5230d5bd9791204d2dded825ee280074">gsc_get_from_ordered_str_list</a> (const char *target, const size_t listLen, const char **list)</td></tr>
<tr class="memdesc:ga5230d5bd9791204d2dded825ee280074"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary search through a list of strings.  <a href="group__supporters.html#ga5230d5bd9791204d2dded825ee280074">More...</a><br /></td></tr>
<tr class="separator:ga5230d5bd9791204d2dded825ee280074"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4929f1ad6756e3c710ad0781baca0f42"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga4929f1ad6756e3c710ad0781baca0f42">gsc_shuffle_up_to</a> (rnd_pcg_t *rng, void *sequence, const size_t item_size, const size_t total_n, const size_t n_to_shuffle)</td></tr>
<tr class="memdesc:ga4929f1ad6756e3c710ad0781baca0f42"><td class="mdescLeft">&#160;</td><td class="mdescRight">Produce a random ordering of the first n elements in an array using a (partial) Fisher-Yates shuffle.  <a href="group__supporters.html#ga4929f1ad6756e3c710ad0781baca0f42">More...</a><br /></td></tr>
<tr class="separator:ga4929f1ad6756e3c710ad0781baca0f42"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a368b4ddb966aef02950fe327d46f051b"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a368b4ddb966aef02950fe327d46f051b">gsc_set_names</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *a, const char *prefix, const int suffix, const unsigned int from_index)</td></tr>
<tr class="memdesc:a368b4ddb966aef02950fe327d46f051b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills the designated section of the <code>.names</code> array in an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> with the pattern "`prefix`index".  <a href="sim-operations_8c.html#a368b4ddb966aef02950fe327d46f051b">More...</a><br /></td></tr>
<tr class="separator:a368b4ddb966aef02950fe327d46f051b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab82b3920fa4760dad88332ee430e7cf2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gab82b3920fa4760dad88332ee430e7cf2">gsc_create_new_label</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const int setTo)</td></tr>
<tr class="memdesc:gab82b3920fa4760dad88332ee430e7cf2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialises a new custom label.  <a href="group__supporters.html#gab82b3920fa4760dad88332ee430e7cf2">More...</a><br /></td></tr>
<tr class="separator:gab82b3920fa4760dad88332ee430e7cf2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga51d67d9f3619f81972f94b3666f0fe85"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga51d67d9f3619f81972f94b3666f0fe85">gsc_change_label_default</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const int newDefault)</td></tr>
<tr class="memdesc:ga51d67d9f3619f81972f94b3666f0fe85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the default value of a custom label.  <a href="group__supporters.html#ga51d67d9f3619f81972f94b3666f0fe85">More...</a><br /></td></tr>
<tr class="separator:ga51d67d9f3619f81972f94b3666f0fe85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga22e1de8931fdd488bd7e34a9ab416811"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga22e1de8931fdd488bd7e34a9ab416811">gsc_change_label_to</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> whichGroup, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const int setTo)</td></tr>
<tr class="memdesc:ga22e1de8931fdd488bd7e34a9ab416811"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the values of a custom label.  <a href="group__supporters.html#ga22e1de8931fdd488bd7e34a9ab416811">More...</a><br /></td></tr>
<tr class="separator:ga22e1de8931fdd488bd7e34a9ab416811"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf12b89150af1cdedbd0d77c5099c852d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaf12b89150af1cdedbd0d77c5099c852d">gsc_change_label_by_amount</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> whichGroup, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const int byValue)</td></tr>
<tr class="memdesc:gaf12b89150af1cdedbd0d77c5099c852d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Increment the values of a custom label.  <a href="group__supporters.html#gaf12b89150af1cdedbd0d77c5099c852d">More...</a><br /></td></tr>
<tr class="separator:gaf12b89150af1cdedbd0d77c5099c852d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga96c7ab2a732cb2f0e562eafe04670c75"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga96c7ab2a732cb2f0e562eafe04670c75">gsc_change_label_to_values</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> whichGroup, const unsigned int startIndex, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const size_t n_values, const int *values)</td></tr>
<tr class="memdesc:ga96c7ab2a732cb2f0e562eafe04670c75"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy a vector of integers into a custom label.  <a href="group__supporters.html#ga96c7ab2a732cb2f0e562eafe04670c75">More...</a><br /></td></tr>
<tr class="separator:ga96c7ab2a732cb2f0e562eafe04670c75"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac926561083ce538188ff2cccc9580d42"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gac926561083ce538188ff2cccc9580d42">gsc_change_names_to_values</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> whichGroup, const unsigned int startIndex, const size_t n_values, const char **values)</td></tr>
<tr class="memdesc:gac926561083ce538188ff2cccc9580d42"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy a vector of strings into the genotype name field.  <a href="group__supporters.html#gac926561083ce538188ff2cccc9580d42">More...</a><br /></td></tr>
<tr class="separator:gac926561083ce538188ff2cccc9580d42"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae4f31bd6abde28d757b5d7680dbcfcfe"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe">gsc_change_allele_symbol</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *which_marker, const char from, const char to)</td></tr>
<tr class="memdesc:gae4f31bd6abde28d757b5d7680dbcfcfe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace all occurences of a given allele with a different symbol representation.  <a href="group__supporters.html#gae4f31bd6abde28d757b5d7680dbcfcfe">More...</a><br /></td></tr>
<tr class="separator:gae4f31bd6abde28d757b5d7680dbcfcfe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga34ea6f4aea77b6bd95506ca071226544"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga34ea6f4aea77b6bd95506ca071226544">gsc_get_integer_digits</a> (const int i)</td></tr>
<tr class="memdesc:ga34ea6f4aea77b6bd95506ca071226544"><td class="mdescLeft">&#160;</td><td class="mdescRight">Count and return the number of digits in <code>i</code>.  <a href="group__supporters.html#ga34ea6f4aea77b6bd95506ca071226544">More...</a><br /></td></tr>
<tr class="separator:ga34ea6f4aea77b6bd95506ca071226544"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d1427e76e7fd129c6585493e80be049"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a4d1427e76e7fd129c6585493e80be049">gsc_helper_descending_pdouble_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a4d1427e76e7fd129c6585493e80be049"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#a4d1427e76e7fd129c6585493e80be049">More...</a><br /></td></tr>
<tr class="separator:a4d1427e76e7fd129c6585493e80be049"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a50de1f9769153df70263edc86de41fa5"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a50de1f9769153df70263edc86de41fa5">gsc_helper_ascending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a50de1f9769153df70263edc86de41fa5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#a50de1f9769153df70263edc86de41fa5">More...</a><br /></td></tr>
<tr class="separator:a50de1f9769153df70263edc86de41fa5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c42957f4439d2f2b00d9cc2fa54c6c7"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a3c42957f4439d2f2b00d9cc2fa54c6c7">gsc_helper_ascending_pdouble_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a3c42957f4439d2f2b00d9cc2fa54c6c7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#a3c42957f4439d2f2b00d9cc2fa54c6c7">More...</a><br /></td></tr>
<tr class="separator:a3c42957f4439d2f2b00d9cc2fa54c6c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acf2e8ece7fb3470e814da69e8f162813"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#acf2e8ece7fb3470e814da69e8f162813">gsc_helper_indirect_alphabetical_str_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:acf2e8ece7fb3470e814da69e8f162813"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#acf2e8ece7fb3470e814da69e8f162813">More...</a><br /></td></tr>
<tr class="separator:acf2e8ece7fb3470e814da69e8f162813"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaa406c38a85c3953b1cda622ff535d12"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aaa406c38a85c3953b1cda622ff535d12">gsc_helper_mapfileunit_ascending_chr_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:aaa406c38a85c3953b1cda622ff535d12"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#aaa406c38a85c3953b1cda622ff535d12">More...</a><br /></td></tr>
<tr class="separator:aaa406c38a85c3953b1cda622ff535d12"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6899282e8507284be584200d05e786cf"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a6899282e8507284be584200d05e786cf">gsc_helper_mapfileunit_ascending_d_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:a6899282e8507284be584200d05e786cf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#a6899282e8507284be584200d05e786cf">More...</a><br /></td></tr>
<tr class="separator:a6899282e8507284be584200d05e786cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae46ff71b819d28408e8b55dd4281a4f3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gae46ff71b819d28408e8b55dd4281a4f3">gsc_move_genotype</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> from, <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> to, int *label_defaults)</td></tr>
<tr class="memdesc:gae46ff71b819d28408e8b55dd4281a4f3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Move all details of the genotype at one <a class="el" href="structgsc___geno_location.html" title="An gsc_AlleleMatrix/gsc_AlleleMatrix index coordinate of a particular genotype in the simulation.">gsc_GenoLocation</a> to another <a class="el" href="structgsc___geno_location.html" title="An gsc_AlleleMatrix/gsc_AlleleMatrix index coordinate of a particular genotype in the simulation.">gsc_GenoLocation</a>.  <a href="group__deletors.html#gae46ff71b819d28408e8b55dd4281a4f3">More...</a><br /></td></tr>
<tr class="separator:gae46ff71b819d28408e8b55dd4281a4f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d94248d1163c9c40e2910e578ce4708"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708">gsc_nextgappy_valid_pos</a> (struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *it)</td></tr>
<tr class="memdesc:a6d94248d1163c9c40e2910e578ce4708"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next valid position, if the cursor is not already a valid position.  <a href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708">More...</a><br /></td></tr>
<tr class="separator:a6d94248d1163c9c40e2910e578ce4708"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf656d2e43014ca8dd4f5d39adeb68f6"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#adf656d2e43014ca8dd4f5d39adeb68f6">gsc_nextgappy_get_gap</a> (struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *it)</td></tr>
<tr class="memdesc:adf656d2e43014ca8dd4f5d39adeb68f6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next empty position, if the cursor is not already an empty position.  <a href="sim-operations_8c.html#adf656d2e43014ca8dd4f5d39adeb68f6">More...</a><br /></td></tr>
<tr class="separator:adf656d2e43014ca8dd4f5d39adeb68f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5486daced98d9bab2126b9d7af11356"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#af5486daced98d9bab2126b9d7af11356">gsc_nextgappy_get_nongap</a> (struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *it)</td></tr>
<tr class="memdesc:af5486daced98d9bab2126b9d7af11356"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next filled position, if the cursor is not already a filled position.  <a href="sim-operations_8c.html#af5486daced98d9bab2126b9d7af11356">More...</a><br /></td></tr>
<tr class="separator:af5486daced98d9bab2126b9d7af11356"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafa617e95d5ed52b4e25434a20006809c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c">gsc_condense_allele_matrix</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:gafa617e95d5ed52b4e25434a20006809c"><td class="mdescLeft">&#160;</td><td class="mdescRight">A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>.  <a href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c">More...</a><br /></td></tr>
<tr class="separator:gafa617e95d5ed52b4e25434a20006809c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga05da8d96593f8404938cc24f659db049"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049">gsc_create_bidirectional_iter</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga05da8d96593f8404938cc24f659db049"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a bidirectional iterator.  <a href="group__iterators.html#ga05da8d96593f8404938cc24f659db049">More...</a><br /></td></tr>
<tr class="separator:ga05da8d96593f8404938cc24f659db049"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5cca400276ca83acceaab36d5f79aab1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1">gsc_create_bidirectional_iter_fromAM</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *am, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="separator:ga5cca400276ca83acceaab36d5f79aab1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa26377edd1339407af2070924d3596de"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa26377edd1339407af2070924d3596de">gsc_create_randomaccess_iter</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:gaa26377edd1339407af2070924d3596de"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a Random Access Iterator.  <a href="group__iterators.html#gaa26377edd1339407af2070924d3596de">More...</a><br /></td></tr>
<tr class="separator:gaa26377edd1339407af2070924d3596de"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga16615a5663b4d3bd768d8380f820aa66"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66">gsc_get_nth_AlleleMatrix</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *listStart, const unsigned int n)</td></tr>
<tr class="memdesc:ga16615a5663b4d3bd768d8380f820aa66"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> by index in the linked list.  <a href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66">More...</a><br /></td></tr>
<tr class="separator:ga16615a5663b4d3bd768d8380f820aa66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd">gsc_set_bidirectional_iter_to_start</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the start of its sequence.  <a href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd">More...</a><br /></td></tr>
<tr class="separator:gaa4fa2aec3b7f49f9925fe82c8ff38dcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga99820e199424a3ec8ba7e3c58802fba9"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9">gsc_set_bidirectional_iter_to_end</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga99820e199424a3ec8ba7e3c58802fba9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the end of its sequence.  <a href="group__iterators.html#ga99820e199424a3ec8ba7e3c58802fba9">More...</a><br /></td></tr>
<tr class="separator:ga99820e199424a3ec8ba7e3c58802fba9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga630ba7f8e864238e296cb9480dfbc728"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728">gsc_next_forwards</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga630ba7f8e864238e296cb9480dfbc728"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the next location from a bidirectional iterator.  <a href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728">More...</a><br /></td></tr>
<tr class="separator:ga630ba7f8e864238e296cb9480dfbc728"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaacc1f3131edc2c62e14442ca1b647871"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871">gsc_next_backwards</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gaacc1f3131edc2c62e14442ca1b647871"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the previous location from a bidirectional iterator.  <a href="group__iterators.html#gaacc1f3131edc2c62e14442ca1b647871">More...</a><br /></td></tr>
<tr class="separator:gaacc1f3131edc2c62e14442ca1b647871"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaccc34724148be204eb4d8515000dfd5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5">gsc_next_get_nth</a> (<a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a> *it, const unsigned int n)</td></tr>
<tr class="memdesc:gaaccc34724148be204eb4d8515000dfd5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a location by index using a <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a>.  <a href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5">More...</a><br /></td></tr>
<tr class="separator:gaaccc34724148be204eb4d8515000dfd5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5c7cd898a40e515c3164c98821068978"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga5c7cd898a40e515c3164c98821068978">gsc_get_name_of_id</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> id)</td></tr>
<tr class="memdesc:ga5c7cd898a40e515c3164c98821068978"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the name of the genotype with a given id.  <a href="group__search.html#ga5c7cd898a40e515c3164c98821068978">More...</a><br /></td></tr>
<tr class="separator:ga5c7cd898a40e515c3164c98821068978"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad9d1d0b5575da681af7789354447612a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gad9d1d0b5575da681af7789354447612a">gsc_get_parents_of_id</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> id, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> output[static 2])</td></tr>
<tr class="memdesc:gad9d1d0b5575da681af7789354447612a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>.  <a href="group__search.html#gad9d1d0b5575da681af7789354447612a">More...</a><br /></td></tr>
<tr class="separator:gad9d1d0b5575da681af7789354447612a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7dd3ac972d163ae0c8792a5989387b29"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga7dd3ac972d163ae0c8792a5989387b29">gsc_get_ids_of_names</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const size_t n_names, const char **names, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *output)</td></tr>
<tr class="memdesc:ga7dd3ac972d163ae0c8792a5989387b29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for genotypes with certain names in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> and save the ids of those names.  <a href="group__search.html#ga7dd3ac972d163ae0c8792a5989387b29">More...</a><br /></td></tr>
<tr class="separator:ga7dd3ac972d163ae0c8792a5989387b29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6663c53f59634d0ea5f865accf2173c"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gaa6663c53f59634d0ea5f865accf2173c">gsc_get_index_of_child</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> parent1id, const <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> parent2id)</td></tr>
<tr class="memdesc:gaa6663c53f59634d0ea5f865accf2173c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its index.  <a href="group__search.html#gaa6663c53f59634d0ea5f865accf2173c">More...</a><br /></td></tr>
<tr class="separator:gaa6663c53f59634d0ea5f865accf2173c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad6c28080cf3e1f85fd78229aba53bf78"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const char *name)</td></tr>
<tr class="memdesc:gad6c28080cf3e1f85fd78229aba53bf78"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with a particular name in a linked list of <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>, and return its global index in the list.  <a href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">More...</a><br /></td></tr>
<tr class="separator:gad6c28080cf3e1f85fd78229aba53bf78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad771fc548df03810adee69ab36cfce9d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gad771fc548df03810adee69ab36cfce9d">gsc_get_id_of_index</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const unsigned int index)</td></tr>
<tr class="memdesc:gad771fc548df03810adee69ab36cfce9d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the id of a genotype by its index.  <a href="group__search.html#gad771fc548df03810adee69ab36cfce9d">More...</a><br /></td></tr>
<tr class="separator:gad771fc548df03810adee69ab36cfce9d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga560870b592f5936255e9bd2a55e7eb56"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga560870b592f5936255e9bd2a55e7eb56">gsc_get_genes_of_index</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *start, const unsigned int index)</td></tr>
<tr class="memdesc:ga560870b592f5936255e9bd2a55e7eb56"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of a genotype by its index.  <a href="group__search.html#ga560870b592f5936255e9bd2a55e7eb56">More...</a><br /></td></tr>
<tr class="separator:ga560870b592f5936255e9bd2a55e7eb56"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9655fe60b19077ecf4d8a0b0e229ea6c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c">gsc_combine_groups</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const size_t list_len, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *grouplist)</td></tr>
<tr class="memdesc:ga9655fe60b19077ecf4d8a0b0e229ea6c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Combine a set of groups into one group.  <a href="group__groupmod.html#ga9655fe60b19077ecf4d8a0b0e229ea6c">More...</a><br /></td></tr>
<tr class="separator:ga9655fe60b19077ecf4d8a0b0e229ea6c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad7acd8960fc510d4aebb587ad35878f7"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7">gsc_make_group_from</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const size_t index_list_len, const unsigned int *genotype_indexes)</td></tr>
<tr class="memdesc:gad7acd8960fc510d4aebb587ad35878f7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Take a list of indexes and allocate the genotypes at those indexes to a new group.  <a href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7">More...</a><br /></td></tr>
<tr class="separator:gad7acd8960fc510d4aebb587ad35878f7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab2b56b3f3c1754e6f8a5440aaf2ba38b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b">gsc_split_by_label_value</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const int valueToSplit)</td></tr>
<tr class="memdesc:gab2b56b3f3c1754e6f8a5440aaf2ba38b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocates the genotypes with a particular value of a label to a new group.  <a href="group__groupmod.html#gab2b56b3f3c1754e6f8a5440aaf2ba38b">More...</a><br /></td></tr>
<tr class="separator:gab2b56b3f3c1754e6f8a5440aaf2ba38b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5290a4afbf8a34b3149f476cc7a9a97d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d">gsc_split_by_label_range</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> whichLabel, const int valueLowBound, const int valueHighBound)</td></tr>
<tr class="memdesc:ga5290a4afbf8a34b3149f476cc7a9a97d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocates the genotypes with values of a label in a particular range to a new group.  <a href="group__groupmod.html#ga5290a4afbf8a34b3149f476cc7a9a97d">More...</a><br /></td></tr>
<tr class="separator:ga5290a4afbf8a34b3149f476cc7a9a97d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga25286575497913b2613f200f6ff439f2"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2">gsc_scaffold_split_by_somequality</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, void *somequality_data, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>(*somequality_tester)(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>, void *, size_t, size_t, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *), size_t maxentries_results, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga25286575497913b2613f200f6ff439f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split by some quality (generic function)  <a href="group__groupmod.html#ga25286575497913b2613f200f6ff439f2">More...</a><br /></td></tr>
<tr class="separator:ga25286575497913b2613f200f6ff439f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a986ca18e011ef3906d609b47c9a84c54"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a986ca18e011ef3906d609b47c9a84c54">gsc_helper_split_by_quality_halfsibtemplate</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, void *datastore, size_t maxgroups, size_t groupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>(*getparent)(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>))</td></tr>
<tr class="separator:a986ca18e011ef3906d609b47c9a84c54"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7732d896b52fde5a51e8ee337d7310ec"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a7732d896b52fde5a51e8ee337d7310ec">gsc_helper_split_by_quality_halfsib1</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, void *datastore, size_t maxgroups, size_t groupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="separator:a7732d896b52fde5a51e8ee337d7310ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f8a1c3546a8c12067e790c2f2cc9782"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a0f8a1c3546a8c12067e790c2f2cc9782">gsc_helper_split_by_quality_halfsib2</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, void *datastore, size_t maxgroups, size_t groupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="separator:a0f8a1c3546a8c12067e790c2f2cc9782"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga556be18433d7185e212112028e22a425"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga556be18433d7185e212112028e22a425">gsc_split_into_halfsib_families</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const int parent, size_t maxentries_results, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga556be18433d7185e212112028e22a425"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into families of half-siblings by shared first or second parent.  <a href="group__groupmod.html#ga556be18433d7185e212112028e22a425">More...</a><br /></td></tr>
<tr class="separator:ga556be18433d7185e212112028e22a425"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb9ba063ab6784b4676c7fdafa7a53a4"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#afb9ba063ab6784b4676c7fdafa7a53a4">gsc_helper_split_by_quality_family</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, void *datastore, size_t maxgroups, size_t groupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="separator:afb9ba063ab6784b4676c7fdafa7a53a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf863cc5728f3a83f0002a9e43ada3a33"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33">gsc_split_into_families</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, size_t maxentries_results, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:gaf863cc5728f3a83f0002a9e43ada3a33"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into families by their pedigrees.  <a href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33">More...</a><br /></td></tr>
<tr class="separator:gaf863cc5728f3a83f0002a9e43ada3a33"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a69a260388571f8c32ad879dfdc8a3c8f"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a69a260388571f8c32ad879dfdc8a3c8f">gsc_helper_split_by_quality_individuate</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, void *datastore, size_t maxgroups, size_t groupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="separator:a69a260388571f8c32ad879dfdc8a3c8f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga532a3797b8529c797d39e2ad9069ecf1"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1">gsc_split_into_individuals</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, size_t maxentries_results, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga532a3797b8529c797d39e2ad9069ecf1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n one-member groups.  <a href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1">More...</a><br /></td></tr>
<tr class="separator:ga532a3797b8529c797d39e2ad9069ecf1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga47292b150273a395fe282f86c0f78e0b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b">gsc_split_evenly_into_two</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:ga47292b150273a395fe282f86c0f78e0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into two groups of equal size (or size differing only by one, if the original group had an odd number of members) using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b">More...</a><br /></td></tr>
<tr class="separator:ga47292b150273a395fe282f86c0f78e0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga908be31b28b0f6ee07eb64703feaa81b"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b">gsc_scaffold_split_by_someallocation</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, void *someallocator_data, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>(*someallocator)(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *, void *, size_t, size_t *, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *), size_t n_outgroups, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *outgroups)</td></tr>
<tr class="memdesc:ga908be31b28b0f6ee07eb64703feaa81b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split by some allocator (generic function)  <a href="group__groupmod.html#ga908be31b28b0f6ee07eb64703feaa81b">More...</a><br /></td></tr>
<tr class="separator:ga908be31b28b0f6ee07eb64703feaa81b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a040d495e43abdd2c52d3720f71d52588"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a040d495e43abdd2c52d3720f71d52588">gsc_helper_split_by_allocator_knowncounts</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, void *datastore, size_t n_outgroups, size_t *subgroupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *outgroups)</td></tr>
<tr class="separator:a040d495e43abdd2c52d3720f71d52588"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga68012ed8f95ac60ae4875dbf605f5f95"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95">gsc_split_evenly_into_n</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const size_t n, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga68012ed8f95ac60ae4875dbf605f5f95"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size.), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95">More...</a><br /></td></tr>
<tr class="separator:ga68012ed8f95ac60ae4875dbf605f5f95"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga02433970bb02f719e2de61c44f573ff4"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4">gsc_split_into_buckets</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const size_t n, const unsigned int *counts, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga02433970bb02f719e2de61c44f573ff4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4">More...</a><br /></td></tr>
<tr class="separator:ga02433970bb02f719e2de61c44f573ff4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad3a5afbd400e53b51c16a2ab278b221a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a">gsc_split_randomly_into_two</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gad3a5afbd400e53b51c16a2ab278b221a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flip a coin for each member of the group to decide if it should be moved to the new group.  <a href="group__groupmod.html#gad3a5afbd400e53b51c16a2ab278b221a">More...</a><br /></td></tr>
<tr class="separator:gad3a5afbd400e53b51c16a2ab278b221a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9baab4183af973c0ef583a9ed3306c97"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a9baab4183af973c0ef583a9ed3306c97">gsc_helper_split_by_allocator_equalprob</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, void *datastore, size_t n_outgroups, size_t *subgroupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *outgroups)</td></tr>
<tr class="separator:a9baab4183af973c0ef583a9ed3306c97"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae44e90e4bb377a0a2d882d3ee0932621"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621">gsc_split_randomly_into_n</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const size_t n, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:gae44e90e4bb377a0a2d882d3ee0932621"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with equal probability.  <a href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621">More...</a><br /></td></tr>
<tr class="separator:gae44e90e4bb377a0a2d882d3ee0932621"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af7acb36091cc63f5b7c2c7cc36d5911c"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#af7acb36091cc63f5b7c2c7cc36d5911c">gsc_helper_split_by_allocator_unequalprob</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, void *datastore, size_t n_outgroups, size_t *subgroupsfound, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *outgroups)</td></tr>
<tr class="separator:af7acb36091cc63f5b7c2c7cc36d5911c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0433ee84c0760569489e2f1ed31dc7c5"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5">gsc_split_by_probabilities</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const size_t n, const double *probs, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga0433ee84c0760569489e2f1ed31dc7c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with custom probabilities for each group.  <a href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5">More...</a><br /></td></tr>
<tr class="separator:ga0433ee84c0760569489e2f1ed31dc7c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf76f5cdc38c881c9ee8eee32a1770e83"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83">gsc_get_existing_groups</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *output)</td></tr>
<tr class="memdesc:gaf76f5cdc38c881c9ee8eee32a1770e83"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify group numbers that currently have members.  <a href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83">More...</a><br /></td></tr>
<tr class="separator:gaf76f5cdc38c881c9ee8eee32a1770e83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf344362b92201c509c3efd9fba21f6a9"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9">gsc_get_existing_group_counts</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *out_groups, unsigned int *out_sizes)</td></tr>
<tr class="memdesc:gaf344362b92201c509c3efd9fba21f6a9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify group numbers that currently have members, and how many members they have.  <a href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9">More...</a><br /></td></tr>
<tr class="separator:gaf344362b92201c509c3efd9fba21f6a9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga069acddecab22dc14ca9416378f5eacb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga069acddecab22dc14ca9416378f5eacb">gsc_get_next_free_group_num</a> (const size_t n_existing_groups, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *existing_groups, size_t *cursor, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> previous)</td></tr>
<tr class="memdesc:ga069acddecab22dc14ca9416378f5eacb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Iterator to get the next currently-free group number.  <a href="group__supporters.html#ga069acddecab22dc14ca9416378f5eacb">More...</a><br /></td></tr>
<tr class="separator:ga069acddecab22dc14ca9416378f5eacb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga22609353c4951dcf3ed54537ecb6eeda"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda">gsc_get_new_group_num</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga22609353c4951dcf3ed54537ecb6eeda"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that does not identify a group that currently has member(s).  <a href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda">More...</a><br /></td></tr>
<tr class="separator:ga22609353c4951dcf3ed54537ecb6eeda"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7f90f55f1daced69a12f333fd4828ce2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2">gsc_get_n_new_group_nums</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const size_t n, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *result)</td></tr>
<tr class="memdesc:ga7f90f55f1daced69a12f333fd4828ce2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next n sequential integers that do not identify a group that currently has member(s).  <a href="group__supporters.html#ga7f90f55f1daced69a12f333fd4828ce2">More...</a><br /></td></tr>
<tr class="separator:ga7f90f55f1daced69a12f333fd4828ce2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6bcf8ef9adbfc05d8cf695ec978badf"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa6bcf8ef9adbfc05d8cf695ec978badf">gsc_get_new_label_id</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:gaa6bcf8ef9adbfc05d8cf695ec978badf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that is not already allocated to a label in the simulation.  <a href="group__supporters.html#gaa6bcf8ef9adbfc05d8cf695ec978badf">More...</a><br /></td></tr>
<tr class="separator:gaa6bcf8ef9adbfc05d8cf695ec978badf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7881c537673d2d2d9dc3bb285676bfd4"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga7881c537673d2d2d9dc3bb285676bfd4">gsc_get_new_eff_set_id</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga7881c537673d2d2d9dc3bb285676bfd4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that is not already allocated to a marker effect set ID in the simulation.  <a href="group__supporters.html#ga7881c537673d2d2d9dc3bb285676bfd4">More...</a><br /></td></tr>
<tr class="separator:ga7881c537673d2d2d9dc3bb285676bfd4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga48ab10327d06674174428ca2db0b32ed"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga48ab10327d06674174428ca2db0b32ed">gsc_get_new_map_id</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d)</td></tr>
<tr class="memdesc:ga48ab10327d06674174428ca2db0b32ed"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that is not already allocated to a map ID in the simulation.  <a href="group__supporters.html#ga48ab10327d06674174428ca2db0b32ed">More...</a><br /></td></tr>
<tr class="separator:ga48ab10327d06674174428ca2db0b32ed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga47353034dddec12b2276cd86c2374e67"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga47353034dddec12b2276cd86c2374e67">gsc_get_index_of_label</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> label)</td></tr>
<tr class="memdesc:ga47353034dddec12b2276cd86c2374e67"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the label lookup index of a label identifier.  <a href="group__supporters.html#ga47353034dddec12b2276cd86c2374e67">More...</a><br /></td></tr>
<tr class="separator:ga47353034dddec12b2276cd86c2374e67"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaccd39d5505c6a05a95fb8469662c9a0"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaaccd39d5505c6a05a95fb8469662c9a0">gsc_get_index_of_eff_set</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> eff_set_id)</td></tr>
<tr class="memdesc:gaaccd39d5505c6a05a95fb8469662c9a0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the lookup index of a marker effect set identifier.  <a href="group__supporters.html#gaaccd39d5505c6a05a95fb8469662c9a0">More...</a><br /></td></tr>
<tr class="separator:gaaccd39d5505c6a05a95fb8469662c9a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3a07c81d30aa7738de1672420031b3da"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga3a07c81d30aa7738de1672420031b3da">gsc_get_index_of_map</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map)</td></tr>
<tr class="memdesc:ga3a07c81d30aa7738de1672420031b3da"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the lookup index of a recombination map identifier.  <a href="group__supporters.html#ga3a07c81d30aa7738de1672420031b3da">More...</a><br /></td></tr>
<tr class="separator:ga3a07c81d30aa7738de1672420031b3da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafd291a4e0e38f9cf8963c3cd972fe8a7"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gafd291a4e0e38f9cf8963c3cd972fe8a7">gsc_get_group_size</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gafd291a4e0e38f9cf8963c3cd972fe8a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to count the number of genotypes that currently belong to the specified group.  <a href="group__collgetters.html#gafd291a4e0e38f9cf8963c3cd972fe8a7">More...</a><br /></td></tr>
<tr class="separator:gafd291a4e0e38f9cf8963c3cd972fe8a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2b7c79c5606acd2acd0bf4025c56970a"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga2b7c79c5606acd2acd0bf4025c56970a">gsc_get_group_genes</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, char **output)</td></tr>
<tr class="memdesc:ga2b7c79c5606acd2acd0bf4025c56970a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the genes/alleles of each member of the group.  <a href="group__collgetters.html#ga2b7c79c5606acd2acd0bf4025c56970a">More...</a><br /></td></tr>
<tr class="separator:ga2b7c79c5606acd2acd0bf4025c56970a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac338e0b139b4f50a55e91fa6aaf87b25"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gac338e0b139b4f50a55e91fa6aaf87b25">gsc_get_group_names</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, char **output)</td></tr>
<tr class="memdesc:gac338e0b139b4f50a55e91fa6aaf87b25"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the names of each member of the group.  <a href="group__collgetters.html#gac338e0b139b4f50a55e91fa6aaf87b25">More...</a><br /></td></tr>
<tr class="separator:gac338e0b139b4f50a55e91fa6aaf87b25"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae7e7c996acde461041a47f3800c77f46"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gae7e7c996acde461041a47f3800c77f46">gsc_get_group_ids</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *output)</td></tr>
<tr class="memdesc:gae7e7c996acde461041a47f3800c77f46"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of each member of the group.  <a href="group__collgetters.html#gae7e7c996acde461041a47f3800c77f46">More...</a><br /></td></tr>
<tr class="separator:gae7e7c996acde461041a47f3800c77f46"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad05c43cfd7de8c7077a1aa24403a8f73"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gad05c43cfd7de8c7077a1aa24403a8f73">gsc_get_group_indexes</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, unsigned int *output)</td></tr>
<tr class="memdesc:gad05c43cfd7de8c7077a1aa24403a8f73"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the 0-based global indexes of each member of the group.  <a href="group__collgetters.html#gad05c43cfd7de8c7077a1aa24403a8f73">More...</a><br /></td></tr>
<tr class="separator:gad05c43cfd7de8c7077a1aa24403a8f73"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaeb456f1672fa37415f985e07baae49a6"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6">gsc_get_group_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, unsigned int group_size, double *output)</td></tr>
<tr class="memdesc:gaeb456f1672fa37415f985e07baae49a6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the breeding values/breeding values/fitnesses of each member of the group.  <a href="group__collgetters.html#gaeb456f1672fa37415f985e07baae49a6">More...</a><br /></td></tr>
<tr class="separator:gaeb456f1672fa37415f985e07baae49a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5814616d34a8e7c2446641286c0745a0"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga5814616d34a8e7c2446641286c0745a0">gsc_get_group_parent_ids</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, const int whichParent, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> *output)</td></tr>
<tr class="memdesc:ga5814616d34a8e7c2446641286c0745a0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of either the first or second parent of each member of the group.  <a href="group__collgetters.html#ga5814616d34a8e7c2446641286c0745a0">More...</a><br /></td></tr>
<tr class="separator:ga5814616d34a8e7c2446641286c0745a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7eeb668312d559b35df34dfe8727f814"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga7eeb668312d559b35df34dfe8727f814">gsc_get_group_parent_names</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, const int whichParent, char **output)</td></tr>
<tr class="memdesc:ga7eeb668312d559b35df34dfe8727f814"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the names of either the first or second parent of each member of the group.  <a href="group__collgetters.html#ga7eeb668312d559b35df34dfe8727f814">More...</a><br /></td></tr>
<tr class="separator:ga7eeb668312d559b35df34dfe8727f814"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9e31fa104eccb5ed54832b21b7c5d73e"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e">gsc_get_group_pedigrees</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id, unsigned int group_size, char **output)</td></tr>
<tr class="memdesc:ga9e31fa104eccb5ed54832b21b7c5d73e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group.  <a href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e">More...</a><br /></td></tr>
<tr class="separator:ga9e31fa104eccb5ed54832b21b7c5d73e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9c1e4641b679df1d6ec3d885ef6d1fd2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga9c1e4641b679df1d6ec3d885ef6d1fd2">gsc_generate_zero_dmatrix</a> (const size_t r, const size_t c)</td></tr>
<tr class="memdesc:ga9c1e4641b679df1d6ec3d885ef6d1fd2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a matrix of c columns, r rows with all 0.  <a href="group__maths.html#ga9c1e4641b679df1d6ec3d885ef6d1fd2">More...</a><br /></td></tr>
<tr class="separator:ga9c1e4641b679df1d6ec3d885ef6d1fd2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6fdcedf004c9d08e5135d30deb5b12fa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa">gsc_add_matrixvector_product_to_dmatrix</a> (<a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *result, const <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *a, const double *b)</td></tr>
<tr class="memdesc:ga6fdcedf004c9d08e5135d30deb5b12fa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply a <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> to a vector, and add that product to the first column of a provided <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a>.  <a href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa">More...</a><br /></td></tr>
<tr class="separator:ga6fdcedf004c9d08e5135d30deb5b12fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadaec2ba547549392530e91e50a3c985d"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d">gsc_add_doublematrixvector_product_to_dmatrix</a> (<a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *result, const <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *amat, const double *avec, const <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *bmat, const double *bvec)</td></tr>
<tr class="memdesc:gadaec2ba547549392530e91e50a3c985d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply two sets of a <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> and vector, and add both products to the first column of a provided <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a>.  <a href="group__maths.html#gadaec2ba547549392530e91e50a3c985d">More...</a><br /></td></tr>
<tr class="separator:gadaec2ba547549392530e91e50a3c985d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga99bbd9355f031efae628b27e6ae72f6a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a">gsc_delete_dmatrix</a> (<a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *m)</td></tr>
<tr class="memdesc:ga99bbd9355f031efae628b27e6ae72f6a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="structgsc___decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">gsc_DecimalMatrix</a> and frees its memory.  <a href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a">More...</a><br /></td></tr>
<tr class="separator:ga99bbd9355f031efae628b27e6ae72f6a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaddc395516c92fe17e017a3a48e5fc3e5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5">gsc_delete_group</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gaddc395516c92fe17e017a3a48e5fc3e5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes all genotypes belonging to a particular group.  <a href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5">More...</a><br /></td></tr>
<tr class="separator:gaddc395516c92fe17e017a3a48e5fc3e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga27c2247d05e610fc83a83a5bae3d8d94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga27c2247d05e610fc83a83a5bae3d8d94">gsc_delete_eff_set</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga27c2247d05e610fc83a83a5bae3d8d94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a particular set of marker effects from memory.  <a href="group__deletors.html#ga27c2247d05e610fc83a83a5bae3d8d94">More...</a><br /></td></tr>
<tr class="separator:ga27c2247d05e610fc83a83a5bae3d8d94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga060559de8b20b9a1dbbd73297a6c4079"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga060559de8b20b9a1dbbd73297a6c4079">gsc_delete_label</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___label_i_d.html">gsc_LabelID</a> which_label)</td></tr>
<tr class="memdesc:ga060559de8b20b9a1dbbd73297a6c4079"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears memory of this label from the simulation and all its genotypes.  <a href="group__deletors.html#ga060559de8b20b9a1dbbd73297a6c4079">More...</a><br /></td></tr>
<tr class="separator:ga060559de8b20b9a1dbbd73297a6c4079"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf28e802f4bf8149f58a49c6262b02f4c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaf28e802f4bf8149f58a49c6262b02f4c">gsc_delete_genome</a> (<a class="el" href="structgsc___known_genome.html">gsc_KnownGenome</a> *g)</td></tr>
<tr class="memdesc:gaf28e802f4bf8149f58a49c6262b02f4c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes and clears the memory of a <a class="el" href="structgsc___known_genome.html" title="A type that stores the genome structure used in simulation.">gsc_KnownGenome</a> object and its children.  <a href="group__deletors.html#gaf28e802f4bf8149f58a49c6262b02f4c">More...</a><br /></td></tr>
<tr class="separator:gaf28e802f4bf8149f58a49c6262b02f4c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6075a65332dee0dc0032f29cad35fd74"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga6075a65332dee0dc0032f29cad35fd74">gsc_delete_recombination_map</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> which_map)</td></tr>
<tr class="memdesc:ga6075a65332dee0dc0032f29cad35fd74"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a particular recombination map from memory.  <a href="group__deletors.html#ga6075a65332dee0dc0032f29cad35fd74">More...</a><br /></td></tr>
<tr class="separator:ga6075a65332dee0dc0032f29cad35fd74"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6d16e1e88561612bb4efa22b80a5bed"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaa6d16e1e88561612bb4efa22b80a5bed">gsc_delete_recombination_map_nointegrity</a> (<a class="el" href="structgsc___recombination_map.html">gsc_RecombinationMap</a> *m)</td></tr>
<tr class="memdesc:gaa6d16e1e88561612bb4efa22b80a5bed"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes and clears the memory of a <a class="el" href="structgsc___recombination_map.html" title="A type that stores linkage groups and crossover probabilities for simulating meiosis.">gsc_RecombinationMap</a> struct.  <a href="group__deletors.html#gaa6d16e1e88561612bb4efa22b80a5bed">More...</a><br /></td></tr>
<tr class="separator:gaa6d16e1e88561612bb4efa22b80a5bed"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9c647cacb393998930c560962c1f7f6d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga9c647cacb393998930c560962c1f7f6d">gsc_delete_allele_matrix</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:ga9c647cacb393998930c560962c1f7f6d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Delete the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> linked list from m onwards and frees its memory.  <a href="group__deletors.html#ga9c647cacb393998930c560962c1f7f6d">More...</a><br /></td></tr>
<tr class="separator:ga9c647cacb393998930c560962c1f7f6d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga41c27e8d966e9adde4a844c3c022629b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga41c27e8d966e9adde4a844c3c022629b">gsc_delete_effect_matrix</a> (<a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *m)</td></tr>
<tr class="memdesc:ga41c27e8d966e9adde4a844c3c022629b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes an <a class="el" href="structgsc___effect_matrix.html" title="A type that stores a matrix of effect values and their names.">gsc_EffectMatrix</a> object and frees its memory.  <a href="group__deletors.html#ga41c27e8d966e9adde4a844c3c022629b">More...</a><br /></td></tr>
<tr class="separator:ga41c27e8d966e9adde4a844c3c022629b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1f4bef6ea148aaa0a1fa30681b29fe50"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga1f4bef6ea148aaa0a1fa30681b29fe50">gsc_delete_simdata</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *m)</td></tr>
<tr class="memdesc:ga1f4bef6ea148aaa0a1fa30681b29fe50"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object and frees its memory.  <a href="group__deletors.html#ga1f4bef6ea148aaa0a1fa30681b29fe50">More...</a><br /></td></tr>
<tr class="separator:ga1f4bef6ea148aaa0a1fa30681b29fe50"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8583ccabd176c47c4126f164321e0126"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga8583ccabd176c47c4126f164321e0126">gsc_delete_markerblocks</a> (<a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> *b)</td></tr>
<tr class="memdesc:ga8583ccabd176c47c4126f164321e0126"><td class="mdescLeft">&#160;</td><td class="mdescRight">Delete a <a class="el" href="structgsc___marker_blocks.html" title="A struct used to store a set of blocks of markers.">gsc_MarkerBlocks</a> struct.  <a href="group__deletors.html#ga8583ccabd176c47c4126f164321e0126">More...</a><br /></td></tr>
<tr class="separator:ga8583ccabd176c47c4126f164321e0126"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga83d81991036aca26e2ec5b03f8b33b8a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a">gsc_delete_bidirectional_iter</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga83d81991036aca26e2ec5b03f8b33b8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="structgsc___bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a gsc_SimData or through only ...">gsc_BidirectionalIterator</a> object.  <a href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a">More...</a><br /></td></tr>
<tr class="separator:ga83d81991036aca26e2ec5b03f8b33b8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3e88311f64b089f3d7a336f4efee3b05"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga3e88311f64b089f3d7a336f4efee3b05">gsc_delete_randomaccess_iter</a> (<a class="el" href="structgsc___random_access_iterator.html">gsc_RandomAccessIterator</a> *it)</td></tr>
<tr class="memdesc:ga3e88311f64b089f3d7a336f4efee3b05"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="structgsc___random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a gsc_SimData or of all the members of a ...">gsc_RandomAccessIterator</a> object and frees its memory.  <a href="group__deletors.html#ga3e88311f64b089f3d7a336f4efee3b05">More...</a><br /></td></tr>
<tr class="separator:ga3e88311f64b089f3d7a336f4efee3b05"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1a3a39309d4ce6447581e73f7483fbfb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga1a3a39309d4ce6447581e73f7483fbfb">gsc_tablefilereader_create</a> (const char *filename)</td></tr>
<tr class="memdesc:ga1a3a39309d4ce6447581e73f7483fbfb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Open a file for reading with <a class="el" href="structgsc___table_file_reader.html" title="Stream reader for files of some tabular format.">gsc_TableFileReader</a>.  <a href="group__structs.html#ga1a3a39309d4ce6447581e73f7483fbfb">More...</a><br /></td></tr>
<tr class="separator:ga1a3a39309d4ce6447581e73f7483fbfb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga85b05dcdc5ec484d918ed2d3330c3ae8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga85b05dcdc5ec484d918ed2d3330c3ae8">gsc_tablefilereader_close</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tbl)</td></tr>
<tr class="memdesc:ga85b05dcdc5ec484d918ed2d3330c3ae8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Close a <a class="el" href="structgsc___table_file_reader.html" title="Stream reader for files of some tabular format.">gsc_TableFileReader</a>'s file pointer.  <a href="group__structs.html#ga85b05dcdc5ec484d918ed2d3330c3ae8">More...</a><br /></td></tr>
<tr class="separator:ga85b05dcdc5ec484d918ed2d3330c3ae8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac2206b9a58abb4b533d11dec4bd9f322"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322">gsc_helper_tablefilereader_refill_buffer</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tbl)</td></tr>
<tr class="memdesc:gac2206b9a58abb4b533d11dec4bd9f322"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read another buffer's worth of characters from a <a class="el" href="structgsc___table_file_reader.html" title="Stream reader for files of some tabular format.">gsc_TableFileReader</a>'s file.  <a href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322">More...</a><br /></td></tr>
<tr class="separator:gac2206b9a58abb4b533d11dec4bd9f322"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9861b51e8a5c74f843aac748c932c3a0"><td class="memItemLeft" align="right" valign="top">enum <a class="el" href="group__structs.html#gaa092055f7433a18bae99970ac6d2b79f">gsc_TableFileCurrentStatus</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0">gsc_helper_tablefilereader_classify_char</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tbl)</td></tr>
<tr class="memdesc:ga9861b51e8a5c74f843aac748c932c3a0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Classify the character under the cursor of a TableFileReader as cell contents or otherwise.  <a href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0">More...</a><br /></td></tr>
<tr class="separator:ga9861b51e8a5c74f843aac748c932c3a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga85446dd929940e823745067d0a1d1d5d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d">gsc_tablefilecell_deep_copy</a> (<a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *c)</td></tr>
<tr class="memdesc:ga85446dd929940e823745067d0a1d1d5d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate memory to store a deep copy of a <a class="el" href="structgsc___table_file_cell.html" title="Represent a cell read by a gsc_TableFileReader.">gsc_TableFileCell</a>, if previously only a shallow copy.  <a href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d">More...</a><br /></td></tr>
<tr class="separator:ga85446dd929940e823745067d0a1d1d5d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9b389be1c0eada385d1015e445886c05"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05">gsc_tablefilereader_get_next_cell</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tbl)</td></tr>
<tr class="memdesc:ga9b389be1c0eada385d1015e445886c05"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read forwards in TableFileReader and return the next cell's contents, as well as how many column gaps and newlines preceeded it.  <a href="group__structs.html#ga9b389be1c0eada385d1015e445886c05">More...</a><br /></td></tr>
<tr class="separator:ga9b389be1c0eada385d1015e445886c05"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad57230a4c812ba8a6deed26100beae7b"><td class="memItemLeft" align="right" valign="top">_Bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gad57230a4c812ba8a6deed26100beae7b">gsc_get_index_of_genetic_marker</a> (const char *target, <a class="el" href="structgsc___known_genome.html">gsc_KnownGenome</a> g, unsigned int *out)</td></tr>
<tr class="memdesc:gad57230a4c812ba8a6deed26100beae7b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return whether or not a marker name is present in the tracked markers, and at what index.  <a href="group__supporters.html#gad57230a4c812ba8a6deed26100beae7b">More...</a><br /></td></tr>
<tr class="separator:gad57230a4c812ba8a6deed26100beae7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9022b0b2e12395904757c59db6010d5b"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a9022b0b2e12395904757c59db6010d5b">gsc_helper_tablefilereader_get_next_cell_wqueue</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tf, <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> **queue, size_t *queuesize)</td></tr>
<tr class="memdesc:a9022b0b2e12395904757c59db6010d5b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cells from a TableFileReader.  <a href="sim-operations_8c.html#a9022b0b2e12395904757c59db6010d5b">More...</a><br /></td></tr>
<tr class="separator:a9022b0b2e12395904757c59db6010d5b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11ff196fb220dfd8336587f016767b66"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a11ff196fb220dfd8336587f016767b66">gsc_helper_parse_3cell_header</a> (<a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *tf, const char **canonical_titles, int *col_order, <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *unprocessedqueue, size_t *queuesize)</td></tr>
<tr class="memdesc:a11ff196fb220dfd8336587f016767b66"><td class="mdescLeft">&#160;</td><td class="mdescRight">Header row reading and processing for map and effect set files.  <a href="sim-operations_8c.html#a11ff196fb220dfd8336587f016767b66">More...</a><br /></td></tr>
<tr class="separator:a11ff196fb220dfd8336587f016767b66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3c2207b2bd170d752a0a26a9fa2aa19"><td class="memItemLeft" align="right" valign="top">static size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ad3c2207b2bd170d752a0a26a9fa2aa19">gsc_helper_parse_mapfile</a> (const char *filename, struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> **out)</td></tr>
<tr class="memdesc:ad3c2207b2bd170d752a0a26a9fa2aa19"><td class="mdescLeft">&#160;</td><td class="mdescRight">Extract the contents of a genetic map file.  <a href="sim-operations_8c.html#ad3c2207b2bd170d752a0a26a9fa2aa19">More...</a><br /></td></tr>
<tr class="separator:ad3c2207b2bd170d752a0a26a9fa2aa19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acf770138dc0f9f33bc715156c83ccfd0"><td class="memItemLeft" align="right" valign="top">static unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#acf770138dc0f9f33bc715156c83ccfd0">gsc_helper_str_markerlist_leftjoin</a> (<a class="el" href="structgsc___known_genome.html">gsc_KnownGenome</a> g, unsigned int n_markers_in_list, struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> **markerlist)</td></tr>
<tr class="memdesc:acf770138dc0f9f33bc715156c83ccfd0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Discard markers whose names are not present in a <a class="el" href="structgsc___known_genome.html" title="A type that stores the genome structure used in simulation.">gsc_KnownGenome</a>.  <a href="sim-operations_8c.html#acf770138dc0f9f33bc715156c83ccfd0">More...</a><br /></td></tr>
<tr class="separator:acf770138dc0f9f33bc715156c83ccfd0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a337459fb31ef07b88a77068780ab746e"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a337459fb31ef07b88a77068780ab746e">gsc_helper_insert_recombmap_into_simdata</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___recombination_map.html">gsc_RecombinationMap</a> map)</td></tr>
<tr class="memdesc:a337459fb31ef07b88a77068780ab746e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Save a RecombinationMap to the SimData and allocate it a mapID.  <a href="sim-operations_8c.html#a337459fb31ef07b88a77068780ab746e">More...</a><br /></td></tr>
<tr class="separator:a337459fb31ef07b88a77068780ab746e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3717c5ed02cf2a321b7424dc0e668e19"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a3717c5ed02cf2a321b7424dc0e668e19">gsc_helper_insert_eff_set_into_simdata</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> effset)</td></tr>
<tr class="memdesc:a3717c5ed02cf2a321b7424dc0e668e19"><td class="mdescLeft">&#160;</td><td class="mdescRight">Save an EffectMatrix to the SimData and allocate it an EffectID.  <a href="sim-operations_8c.html#a3717c5ed02cf2a321b7424dc0e668e19">More...</a><br /></td></tr>
<tr class="separator:a3717c5ed02cf2a321b7424dc0e668e19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac1d97be6bdb21bbb271fc44616461fd8"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ac1d97be6bdb21bbb271fc44616461fd8">gsc_helper_sort_markerlist</a> (unsigned int n_markers, struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> *markerlist)</td></tr>
<tr class="memdesc:ac1d97be6bdb21bbb271fc44616461fd8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Sort markerlist by chromosome name, and by position within each chromosome.  <a href="sim-operations_8c.html#ac1d97be6bdb21bbb271fc44616461fd8">More...</a><br /></td></tr>
<tr class="separator:ac1d97be6bdb21bbb271fc44616461fd8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gada5c2aaacd8b8a145493a0db5c017068"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gada5c2aaacd8b8a145493a0db5c017068">gsc_create_recombmap_from_markerlist</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, unsigned int n_markers, struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> *markerlist)</td></tr>
<tr class="memdesc:gada5c2aaacd8b8a145493a0db5c017068"><td class="mdescLeft">&#160;</td><td class="mdescRight">Parse a list of markers/chrs/positions into a <a class="el" href="structgsc___recombination_map.html" title="A type that stores linkage groups and crossover probabilities for simulating meiosis.">gsc_RecombinationMap</a> and save to SimData.  <a href="group__loaders.html#gada5c2aaacd8b8a145493a0db5c017068">More...</a><br /></td></tr>
<tr class="separator:gada5c2aaacd8b8a145493a0db5c017068"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5cbe0b821180b257b21d9551ec797f20"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga5cbe0b821180b257b21d9551ec797f20">gsc_create_uniformspaced_recombmap</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, unsigned int n_markers, char **markernames, double expected_n_recombinations)</td></tr>
<tr class="memdesc:ga5cbe0b821180b257b21d9551ec797f20"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a uniformly-spaced <a class="el" href="structgsc___recombination_map.html" title="A type that stores linkage groups and crossover probabilities for simulating meiosis.">gsc_RecombinationMap</a> from a list of marker names and save to SimData.  <a href="group__loaders.html#ga5cbe0b821180b257b21d9551ec797f20">More...</a><br /></td></tr>
<tr class="separator:ga5cbe0b821180b257b21d9551ec797f20"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac180eb5f1d27907a69314e68cc0d37af"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af">gsc_load_mapfile</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:gac180eb5f1d27907a69314e68cc0d37af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Load a genetic map to a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object.  <a href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af">More...</a><br /></td></tr>
<tr class="separator:gac180eb5f1d27907a69314e68cc0d37af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga176416bff8f5e258dbe1619637c8b831"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831">gsc_load_effectfile</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga176416bff8f5e258dbe1619637c8b831"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> combination with effect values.  <a href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831">More...</a><br /></td></tr>
<tr class="separator:ga176416bff8f5e258dbe1619637c8b831"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaaec31480065a89defa3e02a0f902cb8"><td class="memItemLeft" align="right" valign="top">static enum <a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741">gsc_GenotypeFileCellStyle</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aaaec31480065a89defa3e02a0f902cb8">gsc_helper_genotype_matrix_identify_cell_style</a> (<a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> c)</td></tr>
<tr class="memdesc:aaaec31480065a89defa3e02a0f902cb8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify what formatting a genotype matrix is representing alleles as.  <a href="sim-operations_8c.html#aaaec31480065a89defa3e02a0f902cb8">More...</a><br /></td></tr>
<tr class="separator:aaaec31480065a89defa3e02a0f902cb8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c6136857e4f65bff718565b48e18afa"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a7c6136857e4f65bff718565b48e18afa">gsc_helper_genotypecell_to_allelematrix</a> (<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, unsigned int markerix, enum <a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741">gsc_GenotypeFileCellStyle</a> style, char *cell, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *forrng)</td></tr>
<tr class="memdesc:a7c6136857e4f65bff718565b48e18afa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Parse a string and save it as the alleles of a genotype at a particular location and genetic marker.  <a href="sim-operations_8c.html#a7c6136857e4f65bff718565b48e18afa">More...</a><br /></td></tr>
<tr class="separator:a7c6136857e4f65bff718565b48e18afa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a14bf1934790250afa4e708fe902c44a0"><td class="memItemLeft" align="right" valign="top">static struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a14bf1934790250afa4e708fe902c44a0">gsc_create_emptylistnavigator</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> allocation_group)</td></tr>
<tr class="memdesc:a14bf1934790250afa4e708fe902c44a0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a new <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a>, including an empty AlleleMatrix suitable for inserting into the provided SimData once the emptylist is finalised.  <a href="sim-operations_8c.html#a14bf1934790250afa4e708fe902c44a0">More...</a><br /></td></tr>
<tr class="separator:a14bf1934790250afa4e708fe902c44a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a52e6f87fe92eca15a0f4da957c5f8e19"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a52e6f87fe92eca15a0f4da957c5f8e19">gsc_emptylistnavigator_get_first</a> (struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *it)</td></tr>
<tr class="memdesc:a52e6f87fe92eca15a0f4da957c5f8e19"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reset the cursor of a <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a> to the first genotype.  <a href="sim-operations_8c.html#a52e6f87fe92eca15a0f4da957c5f8e19">More...</a><br /></td></tr>
<tr class="separator:a52e6f87fe92eca15a0f4da957c5f8e19"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a67c2447b0d771997a444efe026cf0600"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a67c2447b0d771997a444efe026cf0600">gsc_emptylistnavigator_get_next</a> (struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *it)</td></tr>
<tr class="memdesc:a67c2447b0d771997a444efe026cf0600"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the next sequential genotype in an <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a>.  <a href="sim-operations_8c.html#a67c2447b0d771997a444efe026cf0600">More...</a><br /></td></tr>
<tr class="separator:a67c2447b0d771997a444efe026cf0600"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a622965bf813b7d185c0effe64aca6bad"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad">gsc_emptylistnavigator_finaliselist</a> (struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *it)</td></tr>
<tr class="memdesc:a622965bf813b7d185c0effe64aca6bad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Push emptylist edited genotypes into the SimData.  <a href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad">More...</a><br /></td></tr>
<tr class="separator:a622965bf813b7d185c0effe64aca6bad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a39561a6563fa3b016f8833252acc5e5a"><td class="memItemLeft" align="right" valign="top">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a39561a6563fa3b016f8833252acc5e5a">gsc_helper_genotypefile_matrix_detect_orientation</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *cellqueue, const size_t firstrowlen, const size_t queuelen, struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> format, const char *filenameforlog)</td></tr>
<tr class="memdesc:a39561a6563fa3b016f8833252acc5e5a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determine whether a genotype matrix is row- or column-oriented.  <a href="sim-operations_8c.html#a39561a6563fa3b016f8833252acc5e5a">More...</a><br /></td></tr>
<tr class="separator:a39561a6563fa3b016f8833252acc5e5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac22c96ec7ddd91b676805d17de44689c"><td class="memItemLeft" align="right" valign="top">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ac22c96ec7ddd91b676805d17de44689c">gsc_helper_genotypefile_matrix_detect_cellstyle</a> (const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *cellqueue, const size_t firstrowlen, const size_t queuelen, struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> format, const char *filenameforlog)</td></tr>
<tr class="memdesc:ac22c96ec7ddd91b676805d17de44689c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determine the style in which alleles are stored in a genotype matrix.  <a href="sim-operations_8c.html#ac22c96ec7ddd91b676805d17de44689c">More...</a><br /></td></tr>
<tr class="separator:ac22c96ec7ddd91b676805d17de44689c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a410d2515f415529d8efe2abd9ab62048"><td class="memItemLeft" align="right" valign="top">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a410d2515f415529d8efe2abd9ab62048">gsc_helper_genotypefile_matrix_detect_header</a> (const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *cellqueue, const size_t firstrowlen, const size_t queuelen, struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> format, const char *filenameforlog)</td></tr>
<tr class="memdesc:a410d2515f415529d8efe2abd9ab62048"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determine whether a genotype matrix has a header row or not.  <a href="sim-operations_8c.html#a410d2515f415529d8efe2abd9ab62048">More...</a><br /></td></tr>
<tr class="separator:a410d2515f415529d8efe2abd9ab62048"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae903fb523337566ac34b578f02435c2c"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ae903fb523337566ac34b578f02435c2c">gsc_helper_genotypefile_matrix_detect_cornercell_presence</a> (const size_t ncellsfirstrow, const size_t ncellssecondrow, const _Bool secondrowheaderisempty)</td></tr>
<tr class="memdesc:ae903fb523337566ac34b578f02435c2c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Determine whether a genotype matrix has a corner cell or not.  <a href="sim-operations_8c.html#ae903fb523337566ac34b578f02435c2c">More...</a><br /></td></tr>
<tr class="separator:ae903fb523337566ac34b578f02435c2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac16e627ca627fc79dd175d17b3ca96af"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___file_format_spec.html">gsc_FileFormatSpec</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#gac16e627ca627fc79dd175d17b3ca96af">gsc_define_matrix_format_details</a> (const <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a> has_header, const <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a> markers_as_rows, const enum <a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741">gsc_GenotypeFileCellStyle</a> cell_style)</td></tr>
<tr class="memdesc:gac16e627ca627fc79dd175d17b3ca96af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Give genomicSimulation hints on the format of a genotype matrix file to be loaded.  <a href="group__structs.html#gac16e627ca627fc79dd175d17b3ca96af">More...</a><br /></td></tr>
<tr class="separator:gac16e627ca627fc79dd175d17b3ca96af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca78dbac4d7154e95b6ddca575937d77"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77">gsc_load_genotypefile_matrix</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *filename, const <a class="el" href="structgsc___file_format_spec.html">gsc_FileFormatSpec</a> format)</td></tr>
<tr class="memdesc:aca78dbac4d7154e95b6ddca575937d77"><td class="mdescLeft">&#160;</td><td class="mdescRight">Loads a genotype file, with or without existing genome model in the SimData.  <a href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77">More...</a><br /></td></tr>
<tr class="separator:aca78dbac4d7154e95b6ddca575937d77"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga50888a3ffcd2853e07bb4c1b6f0d067a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a">gsc_load_genotypefile</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *filename, const <a class="el" href="structgsc___file_format_spec.html">gsc_FileFormatSpec</a> format)</td></tr>
<tr class="memdesc:ga50888a3ffcd2853e07bb4c1b6f0d067a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Load a set of genotypes to a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object.  <a href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a">More...</a><br /></td></tr>
<tr class="separator:ga50888a3ffcd2853e07bb4c1b6f0d067a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga394371951741d8c2db18a85069d6b61e"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="structgsc___multi_i_d_set.html">gsc_MultiIDSet</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e">gsc_load_data_files</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *genotype_file, const char *map_file, const char *effect_file, const <a class="el" href="structgsc___file_format_spec.html">gsc_FileFormatSpec</a> format)</td></tr>
<tr class="memdesc:ga394371951741d8c2db18a85069d6b61e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a> object with marker allele data, a genetic map, and (optionally) marker effect values.  <a href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e">More...</a><br /></td></tr>
<tr class="separator:ga394371951741d8c2db18a85069d6b61e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaecfe54dacdba616a25212ff12a32398c"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gaecfe54dacdba616a25212ff12a32398c">gsc_calculate_min_recombinations_fw1</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapid, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int certain)</td></tr>
<tr class="memdesc:gaecfe54dacdba616a25212ff12a32398c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured.  <a href="group__recomb.html#gaecfe54dacdba616a25212ff12a32398c">More...</a><br /></td></tr>
<tr class="separator:gaecfe54dacdba616a25212ff12a32398c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf955aa7d648a243e559ac23040e35fc5"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gaf955aa7d648a243e559ac23040e35fc5">gsc_calculate_min_recombinations_fwn</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapid, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int window_size, int certain)</td></tr>
<tr class="memdesc:gaf955aa7d648a243e559ac23040e35fc5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it.  <a href="group__recomb.html#gaf955aa7d648a243e559ac23040e35fc5">More...</a><br /></td></tr>
<tr class="separator:gaf955aa7d648a243e559ac23040e35fc5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga17bf2aa42201490431e56ba42ace85fc"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc">gsc_calculate_recombinations_from_file</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *input_file, const char *output_file, int window_len, int certain)</td></tr>
<tr class="memdesc:ga17bf2aa42201490431e56ba42ace85fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes.  <a href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc">More...</a><br /></td></tr>
<tr class="separator:ga17bf2aa42201490431e56ba42ace85fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga923d9402fffe6eacac7971b703343073"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga923d9402fffe6eacac7971b703343073">gsc_generate_gamete</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *parent_genome, char *output, const unsigned int map_index)</td></tr>
<tr class="memdesc:ga923d9402fffe6eacac7971b703343073"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alleles of a given parent.  <a href="group__meiosis.html#ga923d9402fffe6eacac7971b703343073">More...</a><br /></td></tr>
<tr class="separator:ga923d9402fffe6eacac7971b703343073"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6a07e899a26f8aff2b241e3a6f7d8665"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga6a07e899a26f8aff2b241e3a6f7d8665">gsc_generate_doubled_haploid</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *parent_genome, char *output, const unsigned int map_index)</td></tr>
<tr class="memdesc:ga6a07e899a26f8aff2b241e3a6f7d8665"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent.  <a href="group__meiosis.html#ga6a07e899a26f8aff2b241e3a6f7d8665">More...</a><br /></td></tr>
<tr class="separator:ga6a07e899a26f8aff2b241e3a6f7d8665"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3793a9acc76baad86c7f3c31afd71999"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga3793a9acc76baad86c7f3c31afd71999">gsc_generate_clone</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ga3793a9acc76baad86c7f3c31afd71999"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an identical copy of a given genotype.  <a href="group__meiosis.html#ga3793a9acc76baad86c7f3c31afd71999">More...</a><br /></td></tr>
<tr class="separator:ga3793a9acc76baad86c7f3c31afd71999"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2013669171acf8fd19e0ca77c75a3753"><td class="memItemLeft" align="right" valign="top">static FILE *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a2013669171acf8fd19e0ca77c75a3753">gsc_helper_genoptions_save_pedigrees_setup</a> (const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:a2013669171acf8fd19e0ca77c75a3753"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens file for writing save-as-you-go pedigrees in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>.  <a href="sim-operations_8c.html#a2013669171acf8fd19e0ca77c75a3753">More...</a><br /></td></tr>
<tr class="separator:a2013669171acf8fd19e0ca77c75a3753"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abada6cc420c61761415131a3ebfb890e"><td class="memItemLeft" align="right" valign="top">static FILE *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#abada6cc420c61761415131a3ebfb890e">gsc_helper_genoptions_save_bvs_setup</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g, unsigned int *effIndexp)</td></tr>
<tr class="memdesc:abada6cc420c61761415131a3ebfb890e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens file for writing save-as-you-go breeding values in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>.  <a href="sim-operations_8c.html#abada6cc420c61761415131a3ebfb890e">More...</a><br /></td></tr>
<tr class="separator:abada6cc420c61761415131a3ebfb890e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a311c08610c1a020fb05be94297d6f28f"><td class="memItemLeft" align="right" valign="top">static FILE *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f">gsc_helper_genoptions_save_genotypes_setup</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:a311c08610c1a020fb05be94297d6f28f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens file for writing save-as-you-go genotypes in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>.  <a href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f">More...</a><br /></td></tr>
<tr class="separator:a311c08610c1a020fb05be94297d6f28f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa66b6b0ec2f03012e39df20c6f35b532"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532">gsc_helper_genoptions_save_pedigrees</a> (FILE *fp, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *tosave)</td></tr>
<tr class="memdesc:aa66b6b0ec2f03012e39df20c6f35b532"><td class="mdescLeft">&#160;</td><td class="mdescRight">save-as-you-go (pedigrees)  <a href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532">More...</a><br /></td></tr>
<tr class="separator:aa66b6b0ec2f03012e39df20c6f35b532"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a240267fbee56d8075f7b8fe10c81aaa4"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a240267fbee56d8075f7b8fe10c81aaa4">gsc_helper_genoptions_save_bvs</a> (FILE *fe, <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *effMatrices, unsigned int effIndex, <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *tosave)</td></tr>
<tr class="memdesc:a240267fbee56d8075f7b8fe10c81aaa4"><td class="mdescLeft">&#160;</td><td class="mdescRight">save-as-you-go (breeding values)  <a href="sim-operations_8c.html#a240267fbee56d8075f7b8fe10c81aaa4">More...</a><br /></td></tr>
<tr class="separator:a240267fbee56d8075f7b8fe10c81aaa4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0fe1d1ace12db253637f9df278cf1639"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639">gsc_helper_genoptions_save_genotypes</a> (FILE *fg, <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *tosave)</td></tr>
<tr class="memdesc:a0fe1d1ace12db253637f9df278cf1639"><td class="mdescLeft">&#160;</td><td class="mdescRight">save-as-you-go (genotypes/alleles)  <a href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639">More...</a><br /></td></tr>
<tr class="separator:a0fe1d1ace12db253637f9df278cf1639"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab77b79416783c53492b3c2b0e3f5c7ea"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ab77b79416783c53492b3c2b0e3f5c7ea">gsc_helper_genoptions_give_names_and_ids</a> (<a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *am, <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:ab77b79416783c53492b3c2b0e3f5c7ea"><td class="mdescLeft">&#160;</td><td class="mdescRight">Apply <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a> naming scheme and <a class="el" href="structgsc___pedigree_i_d.html" title="A type representing a program-lifetime-unique identifier for a genotype, to be used in tracking pedig...">gsc_PedigreeID</a> allocation to a single <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>.  <a href="sim-operations_8c.html#ab77b79416783c53492b3c2b0e3f5c7ea">More...</a><br /></td></tr>
<tr class="separator:ab77b79416783c53492b3c2b0e3f5c7ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa5a1b4a1a80f8baf5b5730d7b7db0b77"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77">gsc_scaffold_make_new_genotypes</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g, void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, int(*parentChooser)(void *, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *, unsigned int *, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>[static 2]), void(*offspringGenerator)(<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>[static 2], <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>))</td></tr>
<tr class="memdesc:gaa5a1b4a1a80f8baf5b5730d7b7db0b77"><td class="mdescLeft">&#160;</td><td class="mdescRight">Make new genotypes (generic function)  <a href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77">More...</a><br /></td></tr>
<tr class="separator:gaa5a1b4a1a80f8baf5b5730d7b7db0b77"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0fcef09a211d57fadb1dfa0b6d102901"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a0fcef09a211d57fadb1dfa0b6d102901">gsc_helper_parentchooser_cross_randomly</a> (void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, unsigned int *counter, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2])</td></tr>
<tr class="memdesc:a0fcef09a211d57fadb1dfa0b6d102901"><td class="mdescLeft">&#160;</td><td class="mdescRight">parentChooser function parameter for gsc_make_random_crosses.  <a href="sim-operations_8c.html#a0fcef09a211d57fadb1dfa0b6d102901">More...</a><br /></td></tr>
<tr class="separator:a0fcef09a211d57fadb1dfa0b6d102901"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa16bd09b22305a169b4ca8e54432f966"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aa16bd09b22305a169b4ca8e54432f966">gsc_helper_make_offspring_cross</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2], <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> putHere)</td></tr>
<tr class="memdesc:aa16bd09b22305a169b4ca8e54432f966"><td class="mdescLeft">&#160;</td><td class="mdescRight">offspringGenerator function parameter for all crossing functions  <a href="sim-operations_8c.html#aa16bd09b22305a169b4ca8e54432f966">More...</a><br /></td></tr>
<tr class="separator:aa16bd09b22305a169b4ca8e54432f966"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae9f10c91e8c7f77b536ef19e2222182c"><td class="memItemLeft" align="right" valign="top">static unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ae9f10c91e8c7f77b536ef19e2222182c">gsc_helper_random_cross_checks</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> from_group, const unsigned int n_crosses, const unsigned int cap)</td></tr>
<tr class="memdesc:ae9f10c91e8c7f77b536ef19e2222182c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check input parameters of random crossing functions.  <a href="sim-operations_8c.html#ae9f10c91e8c7f77b536ef19e2222182c">More...</a><br /></td></tr>
<tr class="separator:ae9f10c91e8c7f77b536ef19e2222182c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4bfb4f901f7a921e6a3a2f1d8a9ac249"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249">gsc_make_random_crosses</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> from_group, const unsigned int n_crosses, const unsigned int cap, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> which_map, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:ga4bfb4f901f7a921e6a3a2f1d8a9ac249"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses among members of a group.  <a href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249">More...</a><br /></td></tr>
<tr class="separator:ga4bfb4f901f7a921e6a3a2f1d8a9ac249"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafd6535362ff9898f871cd4e1098d2ee6"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gafd6535362ff9898f871cd4e1098d2ee6">gsc_randomdraw_replacementrules</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, unsigned int max, unsigned int cap, unsigned int *member_uses, unsigned int noCollision)</td></tr>
<tr class="memdesc:gafd6535362ff9898f871cd4e1098d2ee6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Randomly pick a number in a range, optionally with a cap on how many times a number can be picked, and optionally required to be different to the last pick.  <a href="group__supporters.html#gafd6535362ff9898f871cd4e1098d2ee6">More...</a><br /></td></tr>
<tr class="separator:gafd6535362ff9898f871cd4e1098d2ee6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2cee0d2e90de12744fd077b20a7db9bd"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a2cee0d2e90de12744fd077b20a7db9bd">gsc_helper_parentchooser_cross_randomly_between</a> (void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, unsigned int *counter, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2])</td></tr>
<tr class="memdesc:a2cee0d2e90de12744fd077b20a7db9bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">parentChooser function parameter for gsc_make_random_crosses_between.  <a href="sim-operations_8c.html#a2cee0d2e90de12744fd077b20a7db9bd">More...</a><br /></td></tr>
<tr class="separator:a2cee0d2e90de12744fd077b20a7db9bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa43222283d6fb3745f5fab73b36af22d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d">gsc_make_random_crosses_between</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group1, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group2, const unsigned int n_crosses, const unsigned int cap1, const unsigned int cap2, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map1, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map2, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:gaa43222283d6fb3745f5fab73b36af22d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses where the first parent comes from one group and the second from another.  <a href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d">More...</a><br /></td></tr>
<tr class="separator:gaa43222283d6fb3745f5fab73b36af22d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2c3d189234e1ac86e27a1ccaace3c363"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a2c3d189234e1ac86e27a1ccaace3c363">gsc_helper_parentchooser_cross_targeted</a> (void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, unsigned int *counter, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2])</td></tr>
<tr class="memdesc:a2c3d189234e1ac86e27a1ccaace3c363"><td class="mdescLeft">&#160;</td><td class="mdescRight">parentChooser function parameter for gsc_make_targeted_crosses.  <a href="sim-operations_8c.html#a2c3d189234e1ac86e27a1ccaace3c363">More...</a><br /></td></tr>
<tr class="separator:a2c3d189234e1ac86e27a1ccaace3c363"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafe3072701f2dea2a505dd8fc9e8bc936"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936">gsc_make_targeted_crosses</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const size_t n_combinations, const unsigned int *firstParents, const unsigned int *secondParents, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map1, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map2, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:gafe3072701f2dea2a505dd8fc9e8bc936"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs the crosses of pairs of parents whose indexes are provided in an array.  <a href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936">More...</a><br /></td></tr>
<tr class="separator:gafe3072701f2dea2a505dd8fc9e8bc936"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aafe958ddf4ee8c990c521d6ebf60db71"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aafe958ddf4ee8c990c521d6ebf60db71">gsc_helper_parentchooser_selfing</a> (void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, unsigned int *counter, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2])</td></tr>
<tr class="memdesc:aafe958ddf4ee8c990c521d6ebf60db71"><td class="mdescLeft">&#160;</td><td class="mdescRight">parentChooser function parameter for gsc_self_n_times.  <a href="sim-operations_8c.html#aafe958ddf4ee8c990c521d6ebf60db71">More...</a><br /></td></tr>
<tr class="separator:aafe958ddf4ee8c990c521d6ebf60db71"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42606fdc7adae265c1663ef40d45226b"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a42606fdc7adae265c1663ef40d45226b">gsc_helper_make_offspring_self_n_times</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2], <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> putHere)</td></tr>
<tr class="memdesc:a42606fdc7adae265c1663ef40d45226b"><td class="mdescLeft">&#160;</td><td class="mdescRight">offspringGenerator function parameter for gsc_self_n_times.  <a href="sim-operations_8c.html#a42606fdc7adae265c1663ef40d45226b">More...</a><br /></td></tr>
<tr class="separator:a42606fdc7adae265c1663ef40d45226b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab625453074798be6ba8677a5d4986edf"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf">gsc_self_n_times</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const unsigned int n, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> which_map, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:gab625453074798be6ba8677a5d4986edf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Selfs each member of a group for a certain number of generations.  <a href="group__crossers.html#gab625453074798be6ba8677a5d4986edf">More...</a><br /></td></tr>
<tr class="separator:gab625453074798be6ba8677a5d4986edf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a226a60f8e133bcbb389863cd4dbd1898"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a226a60f8e133bcbb389863cd4dbd1898">gsc_helper_make_offspring_doubled_haploids</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2], <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> putHere)</td></tr>
<tr class="memdesc:a226a60f8e133bcbb389863cd4dbd1898"><td class="mdescLeft">&#160;</td><td class="mdescRight">offspringGenerator function parameter for gsc_make_doubled_haploids.  <a href="sim-operations_8c.html#a226a60f8e133bcbb389863cd4dbd1898">More...</a><br /></td></tr>
<tr class="separator:a226a60f8e133bcbb389863cd4dbd1898"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6c230330ca9a9db9459d4477512a21af"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af">gsc_make_doubled_haploids</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> which_map, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:ga6c230330ca9a9db9459d4477512a21af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates a doubled haploid from each member of a group.  <a href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af">More...</a><br /></td></tr>
<tr class="separator:ga6c230330ca9a9db9459d4477512a21af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a94b58634907af7e24ac22b543b8fcbe3"><td class="memItemLeft" align="right" valign="top">static int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a94b58634907af7e24ac22b543b8fcbe3">gsc_helper_parentchooser_cloning</a> (void *parentIterator, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, unsigned int *counter, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2])</td></tr>
<tr class="memdesc:a94b58634907af7e24ac22b543b8fcbe3"><td class="mdescLeft">&#160;</td><td class="mdescRight">parentChooser function parameter for gsc_make_clones.  <a href="sim-operations_8c.html#a94b58634907af7e24ac22b543b8fcbe3">More...</a><br /></td></tr>
<tr class="separator:a94b58634907af7e24ac22b543b8fcbe3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0152d52dcce6514ebba0fa08810ee8a"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ab0152d52dcce6514ebba0fa08810ee8a">gsc_helper_make_offspring_clones</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *datastore, <a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a> parents[static 2], <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> putHere)</td></tr>
<tr class="memdesc:ab0152d52dcce6514ebba0fa08810ee8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">offspringGenerator function parameter for gsc_make_clones.  <a href="sim-operations_8c.html#ab0152d52dcce6514ebba0fa08810ee8a">More...</a><br /></td></tr>
<tr class="separator:ab0152d52dcce6514ebba0fa08810ee8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa3907b6b8ad3de641916b2df83e43a01"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01">gsc_make_clones</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const _Bool inherit_names, <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:gaa3907b6b8ad3de641916b2df83e43a01"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates an identical copy of each member of a group.  <a href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01">More...</a><br /></td></tr>
<tr class="separator:gaa3907b6b8ad3de641916b2df83e43a01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa70fd79312587d8e9fc166c64612cbce"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce">gsc_make_all_unidirectional_crosses</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> from_group, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapID, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:gaa70fd79312587d8e9fc166c64612cbce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between all pairs of parents in the group <code>from_group</code> and allocates the resulting offspring to a new group.  <a href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce">More...</a><br /></td></tr>
<tr class="separator:gaa70fd79312587d8e9fc166c64612cbce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab19b3eb80b5fe460d17c93e0c27f7aaf"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ab19b3eb80b5fe460d17c93e0c27f7aaf">gsc_make_n_crosses_from_top_m_percent</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const int n, const int m, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapID, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="separator:ab19b3eb80b5fe460d17c93e0c27f7aaf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5c0be901d03d0e5dd3d74bc19a4bea70"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70">gsc_make_crosses_from_file</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *input_file, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map1, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map2, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:ga5c0be901d03d0e5dd3d74bc19a4bea70"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between pairs of parents identified by name in a file and allocate the resulting offspring to a new group.  <a href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70">More...</a><br /></td></tr>
<tr class="separator:ga5c0be901d03d0e5dd3d74bc19a4bea70"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2b8100d2d30cf1ada0db8403663076ff"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff">gsc_make_double_crosses_from_file</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *input_file, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map1, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> map2, const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a> g)</td></tr>
<tr class="memdesc:ga2b8100d2d30cf1ada0db8403663076ff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between previously-generated offspring of pairs of parents identified by name in a file.  <a href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff">More...</a><br /></td></tr>
<tr class="separator:ga2b8100d2d30cf1ada0db8403663076ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga70ea35197fe4dc0eedf17e1c7bb993cb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb">gsc_split_by_bv</a> (<a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const unsigned int top_n, const _Bool lowIsBest)</td></tr>
<tr class="memdesc:ga70ea35197fe4dc0eedf17e1c7bb993cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group.  <a href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb">More...</a><br /></td></tr>
<tr class="separator:ga70ea35197fe4dc0eedf17e1c7bb993cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga30689b0f03641259091a72d065951e29"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga30689b0f03641259091a72d065951e29">gsc_calculate_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga30689b0f03641259091a72d065951e29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in a certain group.  <a href="group__calculators.html#ga30689b0f03641259091a72d065951e29">More...</a><br /></td></tr>
<tr class="separator:ga30689b0f03641259091a72d065951e29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36">gsc_calculate_utility_bvs</a> (<a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, const <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *effset)</td></tr>
<tr class="memdesc:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the fitness metric/breeding value for a set of genotypes.  <a href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36">More...</a><br /></td></tr>
<tr class="separator:ga5eb332e54b2f90f399a28a65cdb25e36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2eaf7918fe1253fdf95b9ab81a192c96"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96">gsc_calculate_allele_counts</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const char allele)</td></tr>
<tr class="memdesc:ga2eaf7918fe1253fdf95b9ab81a192c96"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears.  <a href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96">More...</a><br /></td></tr>
<tr class="separator:ga2eaf7918fe1253fdf95b9ab81a192c96"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga82f4e25fa7f7a6f614c9101fa78843ea"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga82f4e25fa7f7a6f614c9101fa78843ea">gsc_calculate_utility_allele_counts</a> (const unsigned int n_markers, const unsigned int n_genotypes, const char **const genotypes, const char allele, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:ga82f4e25fa7f7a6f614c9101fa78843ea"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears.  <a href="group__calculators.html#ga82f4e25fa7f7a6f614c9101fa78843ea">More...</a><br /></td></tr>
<tr class="separator:ga82f4e25fa7f7a6f614c9101fa78843ea"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga38627a2841854474e89769ebd5e10902"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga38627a2841854474e89769ebd5e10902">gsc_calculate_utility_allele_counts_pair</a> (const unsigned int n_markers, const unsigned int n_genotypes, const char **const genotypes, const char allele, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts, const char allele2, <a class="el" href="structgsc___decimal_matrix.html">gsc_DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:ga38627a2841854474e89769ebd5e10902"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear.  <a href="group__calculators.html#ga38627a2841854474e89769ebd5e10902">More...</a><br /></td></tr>
<tr class="separator:ga38627a2841854474e89769ebd5e10902"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga44e3a2d5e4041091deec1954b4f62a86"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga44e3a2d5e4041091deec1954b4f62a86">gsc_create_evenlength_blocks_each_chr</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> mapid, const unsigned int n)</td></tr>
<tr class="memdesc:ga44e3a2d5e4041091deec1954b4f62a86"><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct.  <a href="group__calculators.html#ga44e3a2d5e4041091deec1954b4f62a86">More...</a><br /></td></tr>
<tr class="separator:ga44e3a2d5e4041091deec1954b4f62a86"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaf6f32f41ae7ca27c1a48f8947919794a">gsc_load_blocks</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const char *block_file)</td></tr>
<tr class="memdesc:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks.  <a href="group__calculators.html#gaf6f32f41ae7ca27c1a48f8947919794a">More...</a><br /></td></tr>
<tr class="separator:gaf6f32f41ae7ca27c1a48f8947919794a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7b5ad446b4cd825f6dca63e11e362749"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga7b5ad446b4cd825f6dca63e11e362749">gsc_calculate_group_local_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const char *output_file, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group)</td></tr>
<tr class="memdesc:ga7b5ad446b4cd825f6dca63e11e362749"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga7b5ad446b4cd825f6dca63e11e362749">More...</a><br /></td></tr>
<tr class="separator:ga7b5ad446b4cd825f6dca63e11e362749"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gada411a339f47920e091fed1d586b1425"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gada411a339f47920e091fed1d586b1425">gsc_calculate_local_bvs</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID, const char *output_file)</td></tr>
<tr class="memdesc:gada411a339f47920e091fed1d586b1425"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#gada411a339f47920e091fed1d586b1425">More...</a><br /></td></tr>
<tr class="separator:gada411a339f47920e091fed1d586b1425"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae63afe0edeaf9eab31a0328fc72fc600"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae63afe0edeaf9eab31a0328fc72fc600">gsc_calculate_optimal_haplotype</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gae63afe0edeaf9eab31a0328fc72fc600"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="structgsc___sim_data.html" title="Composite type that is used to run crossing simulations.">gsc_SimData</a>.  <a href="group__calculators.html#gae63afe0edeaf9eab31a0328fc72fc600">More...</a><br /></td></tr>
<tr class="separator:gae63afe0edeaf9eab31a0328fc72fc600"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga145430273fd296e49528a27489f0e520"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga145430273fd296e49528a27489f0e520">gsc_calculate_optimal_possible_haplotype</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga145430273fd296e49528a27489f0e520"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the highest-breeding-value haplotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#ga145430273fd296e49528a27489f0e520">More...</a><br /></td></tr>
<tr class="separator:ga145430273fd296e49528a27489f0e520"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaf950c24b4c08a7557ea74a771ebdc5d8">gsc_calculate_optimal_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values.  <a href="group__calculators.html#gaf950c24b4c08a7557ea74a771ebdc5d8">More...</a><br /></td></tr>
<tr class="separator:gaf950c24b4c08a7557ea74a771ebdc5d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga1f9eb57dc514f86c2923f5f91a140c4f">gsc_calculate_optimal_possible_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> group, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the breeding value of the highest breeding-value genotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#ga1f9eb57dc514f86c2923f5f91a140c4f">More...</a><br /></td></tr>
<tr class="separator:ga1f9eb57dc514f86c2923f5f91a140c4f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae2af55ef05a2bb020ebbb2a9581ea60c">gsc_calculate_minimal_bv</a> (const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values.  <a href="group__calculators.html#gae2af55ef05a2bb020ebbb2a9581ea60c">More...</a><br /></td></tr>
<tr class="separator:gae2af55ef05a2bb020ebbb2a9581ea60c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga26462f6816264c01ce29e3e56efb2dc3"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga26462f6816264c01ce29e3e56efb2dc3">gsc_save_markerblocks</a> (const char *fname, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> labelMapID)</td></tr>
<tr class="memdesc:ga26462f6816264c01ce29e3e56efb2dc3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga26462f6816264c01ce29e3e56efb2dc3">More...</a><br /></td></tr>
<tr class="separator:ga26462f6816264c01ce29e3e56efb2dc3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga82a426c2dd33d2cf1e4da04d4e987445"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445">gsc_save_genotypes</a> (const char *fname, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> groupID, const _Bool markers_as_rows)</td></tr>
<tr class="memdesc:ga82a426c2dd33d2cf1e4da04d4e987445"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints genotypes from the simulation to a file.  <a href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445">More...</a><br /></td></tr>
<tr class="separator:ga82a426c2dd33d2cf1e4da04d4e987445"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7636ac81969c8d41c8da27977bb19150"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150">gsc_save_allele_counts</a> (const char *fname, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> groupID, const char allele, const _Bool markers_as_rows)</td></tr>
<tr class="memdesc:ga7636ac81969c8d41c8da27977bb19150"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints allele counts of genotypes from the simulation to a file.  <a href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150">More...</a><br /></td></tr>
<tr class="separator:ga7636ac81969c8d41c8da27977bb19150"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadce5deef59fa66f9208c2ed4d53a4768"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768">gsc_save_pedigrees</a> (const char *fname, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> groupID, const _Bool full_pedigree)</td></tr>
<tr class="memdesc:gadce5deef59fa66f9208c2ed4d53a4768"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints pedigrees of genotypes in the simulation to a file.  <a href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768">More...</a><br /></td></tr>
<tr class="separator:gadce5deef59fa66f9208c2ed4d53a4768"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga51cba5fcfb6d036614262c57b88a1ee1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1">gsc_save_bvs</a> (const char *fname, const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *d, const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> groupID, const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> effID)</td></tr>
<tr class="memdesc:ga51cba5fcfb6d036614262c57b88a1ee1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints breeding values of genotypes in the simulation to a file.  <a href="group__savers.html#ga51cba5fcfb6d036614262c57b88a1ee1">More...</a><br /></td></tr>
<tr class="separator:ga51cba5fcfb6d036614262c57b88a1ee1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c0584559c66e7c6da1452f9a804958d"><td class="memItemLeft" align="right" valign="top">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a7c0584559c66e7c6da1452f9a804958d">gsc_helper_is_marker_in_chr</a> (const unsigned int markerix, const <a class="el" href="structgsc___linkage_group.html">gsc_LinkageGroup</a> chr, double *pos)</td></tr>
<tr class="memdesc:a7c0584559c66e7c6da1452f9a804958d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check if a marker index is found in a particular LinkageGroup, and provide its distance along the chromosome/linkage group if so.  <a href="sim-operations_8c.html#a7c0584559c66e7c6da1452f9a804958d">More...</a><br /></td></tr>
<tr class="separator:a7c0584559c66e7c6da1452f9a804958d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga193bc97c400e318504cdc9bd03684095"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga193bc97c400e318504cdc9bd03684095">gsc_save_utility_markerblocks</a> (FILE *f, const <a class="el" href="structgsc___marker_blocks.html">gsc_MarkerBlocks</a> b, const unsigned int n_markers, char **const marker_names, const <a class="el" href="structgsc___recombination_map.html">gsc_RecombinationMap</a> *map)</td></tr>
<tr class="memdesc:ga193bc97c400e318504cdc9bd03684095"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga193bc97c400e318504cdc9bd03684095">More...</a><br /></td></tr>
<tr class="separator:ga193bc97c400e318504cdc9bd03684095"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3bbb853ca13c593a940d86882dda2361"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361">gsc_scaffold_save_genotype_info</a> (FILE *f, <a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, unsigned int n_markers, char **const marker_names, const _Bool markers_as_rows, void(*bodycell_printer)(FILE *, <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>, unsigned int, void *), void *bodycell_printer_data)</td></tr>
<tr class="memdesc:a3bbb853ca13c593a940d86882dda2361"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints a matrix of genotype information to a file.  <a href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361">More...</a><br /></td></tr>
<tr class="separator:a3bbb853ca13c593a940d86882dda2361"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a223d292e567ad088b1582c24b9d894ab"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a223d292e567ad088b1582c24b9d894ab">gsc_helper_output_genotypematrix_cell</a> (FILE *f, <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, unsigned int markerix, void *<a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8fa507afa49e3544bae608492cfeb3aca43">GSC_NA</a>)</td></tr>
<tr class="memdesc:a223d292e567ad088b1582c24b9d894ab"><td class="mdescLeft">&#160;</td><td class="mdescRight">Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of each genotype.  <a href="sim-operations_8c.html#a223d292e567ad088b1582c24b9d894ab">More...</a><br /></td></tr>
<tr class="separator:a223d292e567ad088b1582c24b9d894ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aca57b78a1df91391b9be67b3dcd46016"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#aca57b78a1df91391b9be67b3dcd46016">gsc_helper_output_countmatrix_cell</a> (FILE *f, <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> loc, unsigned int markerix, void *data)</td></tr>
<tr class="memdesc:aca57b78a1df91391b9be67b3dcd46016"><td class="mdescLeft">&#160;</td><td class="mdescRight">Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the allele counts of a particular allele for each genetic marker.  <a href="sim-operations_8c.html#aca57b78a1df91391b9be67b3dcd46016">More...</a><br /></td></tr>
<tr class="separator:aca57b78a1df91391b9be67b3dcd46016"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad5a9b943294b42f094bd347f990967cb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb">gsc_save_utility_genotypes</a> (FILE *f, <a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, unsigned int n_markers, char **const marker_names, const _Bool markers_as_rows)</td></tr>
<tr class="memdesc:gad5a9b943294b42f094bd347f990967cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints simulated genotypes to a file.  <a href="group__savers.html#gad5a9b943294b42f094bd347f990967cb">More...</a><br /></td></tr>
<tr class="separator:gad5a9b943294b42f094bd347f990967cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga37bba3dd2808aad1e2e8df7581d74561"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561">gsc_save_utility_allele_counts</a> (FILE *f, <a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, unsigned int n_markers, char **const marker_names, const _Bool markers_as_rows, const char allele)</td></tr>
<tr class="memdesc:ga37bba3dd2808aad1e2e8df7581d74561"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints allele counts of simulated genotypes to a file.  <a href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561">More...</a><br /></td></tr>
<tr class="separator:ga37bba3dd2808aad1e2e8df7581d74561"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8102058ad778cb4a23bd645d6786244d"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a8102058ad778cb4a23bd645d6786244d">gsc_scaffold_save_ancestry_of</a> (const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *m, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> p1, <a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a> p2, void(*strprinter)(char *, size_t, void *), void(*intprinter)(long unsigned int, void *), void *printer_data)</td></tr>
<tr class="memdesc:a8102058ad778cb4a23bd645d6786244d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identifies and saves (recursively) the pedigree of a pair of parents.  <a href="sim-operations_8c.html#a8102058ad778cb4a23bd645d6786244d">More...</a><br /></td></tr>
<tr class="separator:a8102058ad778cb4a23bd645d6786244d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5808811b84717ecb459de3bc7565494"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#af5808811b84717ecb459de3bc7565494">gsc_helper_ancestry_strprinter_file</a> (char *str, size_t strlen, void *data)</td></tr>
<tr class="memdesc:af5808811b84717ecb459de3bc7565494"><td class="mdescLeft">&#160;</td><td class="mdescRight">Kernel for scaffold functions that require printing a string to a file (as opposed to saving the string in a different place)  <a href="sim-operations_8c.html#af5808811b84717ecb459de3bc7565494">More...</a><br /></td></tr>
<tr class="separator:af5808811b84717ecb459de3bc7565494"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a52ddaa114ab4d9e277295dee39c60c48"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a52ddaa114ab4d9e277295dee39c60c48">gsc_helper_ancestry_intprinter_file</a> (long unsigned int i, void *data)</td></tr>
<tr class="memdesc:a52ddaa114ab4d9e277295dee39c60c48"><td class="mdescLeft">&#160;</td><td class="mdescRight">Kernel for scaffold functions that require printing an integer to a file (as opposed to saving the integer in a different place)  <a href="sim-operations_8c.html#a52ddaa114ab4d9e277295dee39c60c48">More...</a><br /></td></tr>
<tr class="separator:a52ddaa114ab4d9e277295dee39c60c48"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga23f6de68db1db7fd293a1e3207eeef6c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c">gsc_save_utility_pedigrees</a> (FILE *f, <a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, const _Bool full_pedigree, const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *parent_pedigree_store)</td></tr>
<tr class="memdesc:ga23f6de68db1db7fd293a1e3207eeef6c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints pedigrees to a file.  <a href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c">More...</a><br /></td></tr>
<tr class="separator:ga23f6de68db1db7fd293a1e3207eeef6c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga710981b758de78c6dd9f4e5ab61971fc"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc">gsc_save_utility_bvs</a> (FILE *f, <a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *targets, const <a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *eff)</td></tr>
<tr class="memdesc:ga710981b758de78c6dd9f4e5ab61971fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate and print breeding values to a file.  <a href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc">More...</a><br /></td></tr>
<tr class="separator:ga710981b758de78c6dd9f4e5ab61971fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="var-members" name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:gabec7841eda2ede124314f2f49e224ca0"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#gabec7841eda2ede124314f2f49e224ca0">GSC_BASIC_OPT</a></td></tr>
<tr class="memdesc:gabec7841eda2ede124314f2f49e224ca0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default parameter values for GenOptions, to help with quick scripts and prototypes.  <a href="group__structs.html#gabec7841eda2ede124314f2f49e224ca0">More...</a><br /></td></tr>
<tr class="separator:gabec7841eda2ede124314f2f49e224ca0"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a id="ad84149faf0b815bcee066748ef31eb60" name="ad84149faf0b815bcee066748ef31eb60"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad84149faf0b815bcee066748ef31eb60">&#9670;&nbsp;</a></span>SIM_OPERATIONS</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define SIM_OPERATIONS</td>
        </tr>
      </table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l00002">2</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="a14bf1934790250afa4e708fe902c44a0" name="a14bf1934790250afa4e708fe902c44a0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a14bf1934790250afa4e708fe902c44a0">&#9670;&nbsp;</a></span>gsc_create_emptylistnavigator()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> gsc_create_emptylistnavigator </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>allocation_group</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Create a new <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a>, including an empty AlleleMatrix suitable for inserting into the provided SimData once the emptylist is finalised. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06356">6356</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a14bf1934790250afa4e708fe902c44a0_cgraph.png" border="0" usemap="#asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_cgraph" alt=""/></div>
<map name="asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_cgraph" id="asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_cgraph">
<area shape="rect" title="Create a new gsc_EmptyListNavigator, including an empty AlleleMatrix suitable for inserting into the ..." alt="" coords="5,5,202,33"/>
<area shape="rect" href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76" title="Creator for an empty gsc_AlleleMatrix object of a given size." alt="" coords="250,5,451,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="499,5,620,33"/>
</map>
</div>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a14bf1934790250afa4e708fe902c44a0_icgraph.png" border="0" usemap="#asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_icgraph" alt=""/></div>
<map name="asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_icgraph" id="asim-operations_8c_a14bf1934790250afa4e708fe902c44a0_icgraph">
<area shape="rect" title="Create a new gsc_EmptyListNavigator, including an empty AlleleMatrix suitable for inserting into the ..." alt="" coords="583,13,780,41"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a622965bf813b7d185c0effe64aca6bad" name="a622965bf813b7d185c0effe64aca6bad"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a622965bf813b7d185c0effe64aca6bad">&#9670;&nbsp;</a></span>gsc_emptylistnavigator_finaliselist()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_emptylistnavigator_finaliselist </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Push emptylist edited genotypes into the SimData. </p>
<p >This is the expected final step in use of <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a> </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06431">6431</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a622965bf813b7d185c0effe64aca6bad_cgraph.png" border="0" usemap="#asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_cgraph" alt=""/></div>
<map name="asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_cgraph" id="asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_cgraph">
<area shape="rect" title="Push emptylist edited genotypes into the SimData." alt="" coords="5,153,159,197"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="207,153,349,197"/>
<area shape="rect" href="group__deletors.html#ga9c647cacb393998930c560962c1f7f6d" title="Delete the gsc_AlleleMatrix linked list from m onwards and frees its memory." alt="" coords="407,5,571,33"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="433,57,544,85"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="436,109,542,137"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="405,161,572,189"/>
<area shape="rect" href="group__deletors.html#gae46ff71b819d28408e8b55dd4281a4f3" title="Move all details of the genotype at one gsc_GenoLocation to another gsc_GenoLocation." alt="" coords="419,213,558,241"/>
<area shape="rect" href="sim-operations_8c.html#adf656d2e43014ca8dd4f5d39adeb68f6" title="Sets the current cursor position in a gsc_GappyIterator to the next empty position,..." alt="" coords="408,265,570,293"/>
<area shape="rect" href="sim-operations_8c.html#af5486daced98d9bab2126b9d7af11356" title="Sets the current cursor position in a gsc_GappyIterator to the next filled position,..." alt="" coords="397,317,580,345"/>
<area shape="rect" href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708" title="Sets the current cursor position in a gsc_GappyIterator to the next valid position,..." alt="" coords="628,292,798,320"/>
</map>
</div>
<div id="dynsection-3" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-3-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-3-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-3-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a622965bf813b7d185c0effe64aca6bad_icgraph.png" border="0" usemap="#asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_icgraph" alt=""/></div>
<map name="asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_icgraph" id="asim-operations_8c_a622965bf813b7d185c0effe64aca6bad_icgraph">
<area shape="rect" title="Push emptylist edited genotypes into the SimData." alt="" coords="583,5,737,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a52e6f87fe92eca15a0f4da957c5f8e19" name="a52e6f87fe92eca15a0f4da957c5f8e19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a52e6f87fe92eca15a0f4da957c5f8e19">&#9670;&nbsp;</a></span>gsc_emptylistnavigator_get_first()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_emptylistnavigator_get_first </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Reset the cursor of a <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a> to the first genotype. </p>
<dl class="section return"><dt>Returns</dt><dd>GenoLocation for the first sequential genotype in the emptylist </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06373">6373</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-4" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-4-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-4-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-4-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_cgraph.png" border="0" usemap="#asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_cgraph" alt=""/></div>
<map name="asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_cgraph" id="asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_cgraph">
<area shape="rect" title="Reset the cursor of a gsc_EmptyListNavigator to the first genotype." alt="" coords="5,5,159,49"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="207,13,328,41"/>
</map>
</div>
<div id="dynsection-5" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-5-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-5-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-5-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_icgraph.png" border="0" usemap="#asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_icgraph" alt=""/></div>
<map name="asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_icgraph" id="asim-operations_8c_a52e6f87fe92eca15a0f4da957c5f8e19_icgraph">
<area shape="rect" title="Reset the cursor of a gsc_EmptyListNavigator to the first genotype." alt="" coords="583,5,737,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a67c2447b0d771997a444efe026cf0600" name="a67c2447b0d771997a444efe026cf0600"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a67c2447b0d771997a444efe026cf0600">&#9670;&nbsp;</a></span>gsc_emptylistnavigator_get_next()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_emptylistnavigator_get_next </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___empty_list_navigator.html">gsc_EmptyListNavigator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Get the next sequential genotype in an <a class="el" href="structgsc___empty_list_navigator.html" title="A structure to hold an initially empty AlleleMatrix list whose genotypes can be accessed sequentially...">gsc_EmptyListNavigator</a>. </p>
<dl class="section return"><dt>Returns</dt><dd>GenoLocation for the next sequential genotype in the emptylist </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06392">6392</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-6" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-6-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-6-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-6-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a67c2447b0d771997a444efe026cf0600_cgraph.png" border="0" usemap="#asim-operations_8c_a67c2447b0d771997a444efe026cf0600_cgraph" alt=""/></div>
<map name="asim-operations_8c_a67c2447b0d771997a444efe026cf0600_cgraph" id="asim-operations_8c_a67c2447b0d771997a444efe026cf0600_cgraph">
<area shape="rect" title="Get the next sequential genotype in an gsc_EmptyListNavigator." alt="" coords="5,24,159,68"/>
<area shape="rect" href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76" title="Creator for an empty gsc_AlleleMatrix object of a given size." alt="" coords="207,5,408,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="456,32,577,60"/>
</map>
</div>
<div id="dynsection-7" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-7-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-7-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-7-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a67c2447b0d771997a444efe026cf0600_icgraph.png" border="0" usemap="#asim-operations_8c_a67c2447b0d771997a444efe026cf0600_icgraph" alt=""/></div>
<map name="asim-operations_8c_a67c2447b0d771997a444efe026cf0600_icgraph" id="asim-operations_8c_a67c2447b0d771997a444efe026cf0600_icgraph">
<area shape="rect" title="Get the next sequential genotype in an gsc_EmptyListNavigator." alt="" coords="583,5,737,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a52ddaa114ab4d9e277295dee39c60c48" name="a52ddaa114ab4d9e277295dee39c60c48"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a52ddaa114ab4d9e277295dee39c60c48">&#9670;&nbsp;</a></span>gsc_helper_ancestry_intprinter_file()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_ancestry_intprinter_file </td>
          <td>(</td>
          <td class="paramtype">long unsigned int&#160;</td>
          <td class="paramname"><em>i</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Kernel for scaffold functions that require printing an integer to a file (as opposed to saving the integer in a different place) </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10822">10822</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-8" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-8-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-8-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-8-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a52ddaa114ab4d9e277295dee39c60c48_icgraph.png" border="0" usemap="#asim-operations_8c_a52ddaa114ab4d9e277295dee39c60c48_icgraph" alt=""/></div>
<map name="asim-operations_8c_a52ddaa114ab4d9e277295dee39c60c48_icgraph" id="asim-operations_8c_a52ddaa114ab4d9e277295dee39c60c48_icgraph">
<area shape="rect" title="Kernel for scaffold functions that require printing an integer to a file (as opposed to saving the in..." alt="" coords="1123,163,1264,207"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="899,171,1075,199"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="696,132,851,176"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="704,200,843,228"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="483,132,645,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="479,200,648,228"/>
</map>
</div>

</div>
</div>
<a id="af5808811b84717ecb459de3bc7565494" name="af5808811b84717ecb459de3bc7565494"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5808811b84717ecb459de3bc7565494">&#9670;&nbsp;</a></span>gsc_helper_ancestry_strprinter_file()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_ancestry_strprinter_file </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>str</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>strlen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Kernel for scaffold functions that require printing a string to a file (as opposed to saving the string in a different place) </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10815">10815</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-9" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-9-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-9-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-9-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_af5808811b84717ecb459de3bc7565494_icgraph.png" border="0" usemap="#asim-operations_8c_af5808811b84717ecb459de3bc7565494_icgraph" alt=""/></div>
<map name="asim-operations_8c_af5808811b84717ecb459de3bc7565494_icgraph" id="asim-operations_8c_af5808811b84717ecb459de3bc7565494_icgraph">
<area shape="rect" title="Kernel for scaffold functions that require printing a string to a file (as opposed to saving the stri..." alt="" coords="1123,163,1264,207"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="899,171,1075,199"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="696,132,851,176"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="704,200,843,228"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="483,132,645,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="479,200,648,228"/>
</map>
</div>

</div>
</div>
<a id="a50de1f9769153df70263edc86de41fa5" name="a50de1f9769153df70263edc86de41fa5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a50de1f9769153df70263edc86de41fa5">&#9670;&nbsp;</a></span>gsc_helper_ascending_double_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_ascending_double_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of doubles to sort in ascending order </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n...">gsc_split_by_bv()</a></dd></dl>
<p>Sorts lower numbers before higher numbers. If they are equal, their order after comparison is undefined. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01147">1147</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-10" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-10-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-10-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-10-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a50de1f9769153df70263edc86de41fa5_icgraph.png" border="0" usemap="#asim-operations_8c_a50de1f9769153df70263edc86de41fa5_icgraph" alt=""/></div>
<map name="asim-operations_8c_a50de1f9769153df70263edc86de41fa5_icgraph" id="asim-operations_8c_a50de1f9769153df70263edc86de41fa5_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="905,80,1055,124"/>
<area shape="rect" href="group__meiosis.html#ga6a07e899a26f8aff2b241e3a6f7d8665" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="705,45,857,89"/>
<area shape="rect" href="group__meiosis.html#ga923d9402fffe6eacac7971b703343073" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="706,117,856,145"/>
<area shape="rect" href="sim-operations_8c.html#a226a60f8e133bcbb389863cd4dbd1898" title="offspringGenerator function parameter for gsc_make_doubled_haploids." alt="" coords="479,20,657,64"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,5,431,33"/>
<area shape="rect" href="sim-operations_8c.html#aa16bd09b22305a169b4ca8e54432f966" title="offspringGenerator function parameter for all crossing functions" alt="" coords="479,109,657,153"/>
<area shape="rect" href="sim-operations_8c.html#a42606fdc7adae265c1663ef40d45226b" title="offspringGenerator function parameter for gsc_self_n_times." alt="" coords="479,199,657,243"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,57,429,85"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,109,429,153"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,177,431,205"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,109,192,153"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,177,195,205"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,229,190,273"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,229,398,257"/>
</map>
</div>

</div>
</div>
<a id="a3c42957f4439d2f2b00d9cc2fa54c6c7" name="a3c42957f4439d2f2b00d9cc2fa54c6c7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3c42957f4439d2f2b00d9cc2fa54c6c7">&#9670;&nbsp;</a></span>gsc_helper_ascending_pdouble_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_ascending_pdouble_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of doubles* to sort them in ascending order of the doubles they point to. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n...">gsc_split_by_bv()</a></dd></dl>
<p>Sorts lower numbers before higher numbers. If they are equal, their order after comparison is undefined. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01164">1164</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-11" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-11-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-11-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-11-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a3c42957f4439d2f2b00d9cc2fa54c6c7_icgraph.png" border="0" usemap="#asim-operations_8c_a3c42957f4439d2f2b00d9cc2fa54c6c7_icgraph" alt=""/></div>
<map name="asim-operations_8c_a3c42957f4439d2f2b00d9cc2fa54c6c7_icgraph" id="asim-operations_8c_a3c42957f4439d2f2b00d9cc2fa54c6c7_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="167,5,317,49"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="5,13,119,41"/>
</map>
</div>

</div>
</div>
<a id="a4d1427e76e7fd129c6585493e80be049" name="a4d1427e76e7fd129c6585493e80be049"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4d1427e76e7fd129c6585493e80be049">&#9670;&nbsp;</a></span>gsc_helper_descending_pdouble_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_descending_pdouble_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of doubles* to sort them in descending order of the doubles they point to. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n...">gsc_split_by_bv()</a></dd></dl>
<p>Sorts higher numbers before lower numbers. If they are equal, their order after comparison is undefined. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01131">1131</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-12" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-12-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-12-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-12-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a4d1427e76e7fd129c6585493e80be049_icgraph.png" border="0" usemap="#asim-operations_8c_a4d1427e76e7fd129c6585493e80be049_icgraph" alt=""/></div>
<map name="asim-operations_8c_a4d1427e76e7fd129c6585493e80be049_icgraph" id="asim-operations_8c_a4d1427e76e7fd129c6585493e80be049_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="167,5,324,49"/>
<area shape="rect" href="group__calculators.html#ga70ea35197fe4dc0eedf17e1c7bb993cb" title="Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n..." alt="" coords="5,13,119,41"/>
</map>
</div>

</div>
</div>
<a id="ab77b79416783c53492b3c2b0e3f5c7ea" name="ab77b79416783c53492b3c2b0e3f5c7ea"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab77b79416783c53492b3c2b0e3f5c7ea">&#9670;&nbsp;</a></span>gsc_helper_genoptions_give_names_and_ids()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_genoptions_give_names_and_ids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>am</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Apply <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a> naming scheme and <a class="el" href="structgsc___pedigree_i_d.html" title="A type representing a program-lifetime-unique identifier for a genotype, to be used in tracking pedig...">gsc_PedigreeID</a> allocation to a single <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07834">7834</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-13" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-13-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-13-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-13-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_cgraph.png" border="0" usemap="#asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_cgraph" alt=""/></div>
<map name="asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_cgraph" id="asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_cgraph">
<area shape="rect" title="Apply gsc_GenOptions naming scheme and gsc_PedigreeID allocation to a single gsc_AlleleMatrix." alt="" coords="5,24,160,68"/>
<area shape="rect" href="sim-operations_8c.html#a368b4ddb966aef02950fe327d46f051b" title="Fills the designated section of the .names array in an gsc_AlleleMatrix with the pattern &quot;`prefix`ind..." alt="" coords="208,32,321,60"/>
<area shape="rect" href="group__supporters.html#ga34ea6f4aea77b6bd95506ca071226544" title="Count and return the number of digits in i." alt="" coords="369,5,520,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="384,57,505,85"/>
</map>
</div>
<div id="dynsection-14" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-14-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-14-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-14-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_icgraph.png" border="0" usemap="#asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_icgraph" alt=""/></div>
<map name="asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_icgraph" id="asim-operations_8c_ab77b79416783c53492b3c2b0e3f5c7ea_icgraph">
<area shape="rect" title="Apply gsc_GenOptions naming scheme and gsc_PedigreeID allocation to a single gsc_AlleleMatrix." alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a240267fbee56d8075f7b8fe10c81aaa4" name="a240267fbee56d8075f7b8fe10c81aaa4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a240267fbee56d8075f7b8fe10c81aaa4">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_bvs()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_genoptions_save_bvs </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>fe</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a> *&#160;</td>
          <td class="paramname"><em>effMatrices</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>effIndex</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>tosave</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>save-as-you-go (breeding values) </p>
<p >Calculates and saves breeding values of genotypes in an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> to a file. Performs null-checks on file pointer and <em>effIndex</em>, and fails silently if checks do. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07808">7808</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-15" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-15-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-15-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-15-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_cgraph.png" border="0" usemap="#asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_cgraph" alt=""/></div>
<map name="asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_cgraph" id="asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_cgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="5,452,160,496"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="220,392,381,436"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="208,460,393,488"/>
<area shape="rect" href="group__savers.html#ga710981b758de78c6dd9f4e5ab61971fc" title="Calculate and print breeding values to a file." alt="" coords="231,512,370,540"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="441,287,606,315"/>
<area shape="rect" href="group__deletors.html#ga99bbd9355f031efae628b27e6ae72f6a" title="Deletes a gsc_DecimalMatrix and frees its memory." alt="" coords="680,433,814,461"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="682,485,812,513"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="888,477,1031,521"/>
<area shape="rect" href="group__liteget.html#ga0ce82f67cb5e35a1c73bef54138202f6" title="Get the persistent id of a genotype." alt="" coords="482,589,566,617"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="471,641,577,669"/>
<area shape="rect" href="group__maths.html#gadaec2ba547549392530e91e50a3c985d" title="Multiply two sets of a gsc_DecimalMatrix and vector, and add both products to the first column of a p..." alt="" coords="654,57,840,101"/>
<area shape="rect" href="group__maths.html#ga6fdcedf004c9d08e5135d30deb5b12fa" title="Multiply a gsc_DecimalMatrix to a vector, and add that product to the first column of a provided gsc_..." alt="" coords="673,177,821,221"/>
<area shape="rect" href="group__calculators.html#ga82f4e25fa7f7a6f614c9101fa78843ea" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="677,245,817,289"/>
<area shape="rect" href="group__calculators.html#ga38627a2841854474e89769ebd5e10902" title="Calculates the number of times at each marker that two particular alleles appear." alt="" coords="677,313,817,357"/>
<area shape="rect" href="group__maths.html#ga9c1e4641b679df1d6ec3d885ef6d1fd2" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="656,125,838,153"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="691,5,803,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="899,125,1020,153"/>
</map>
</div>
<div id="dynsection-16" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-16-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-16-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-16-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_icgraph.png" border="0" usemap="#asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_icgraph" alt=""/></div>
<map name="asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_icgraph" id="asim-operations_8c_a240267fbee56d8075f7b8fe10c81aaa4_icgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (breeding values)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="abada6cc420c61761415131a3ebfb890e" name="abada6cc420c61761415131a3ebfb890e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abada6cc420c61761415131a3ebfb890e">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_bvs_setup()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static FILE * gsc_helper_genoptions_save_bvs_setup </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>effIndexp</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Opens file for writing save-as-you-go breeding values in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">effIndexp</td><td>a location to save the index of the chosen effect set. It will be set to GSC_NA_IDX if the effect set ID in <em>g</em> was invalid. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07748">7748</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-17" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-17-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-17-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-17-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_abada6cc420c61761415131a3ebfb890e_cgraph.png" border="0" usemap="#asim-operations_8c_abada6cc420c61761415131a3ebfb890e_cgraph" alt=""/></div>
<map name="asim-operations_8c_abada6cc420c61761415131a3ebfb890e_cgraph" id="asim-operations_8c_abada6cc420c61761415131a3ebfb890e_cgraph">
<area shape="rect" title="Opens file for writing save&#45;as&#45;you&#45;go breeding values in accordance with gsc_GenOptions." alt="" coords="5,5,160,49"/>
<area shape="rect" href="group__supporters.html#gaaccd39d5505c6a05a95fb8469662c9a0" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="208,13,375,41"/>
</map>
</div>
<div id="dynsection-18" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-18-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-18-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-18-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_abada6cc420c61761415131a3ebfb890e_icgraph.png" border="0" usemap="#asim-operations_8c_abada6cc420c61761415131a3ebfb890e_icgraph" alt=""/></div>
<map name="asim-operations_8c_abada6cc420c61761415131a3ebfb890e_icgraph" id="asim-operations_8c_abada6cc420c61761415131a3ebfb890e_icgraph">
<area shape="rect" title="Opens file for writing save&#45;as&#45;you&#45;go breeding values in accordance with gsc_GenOptions." alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a0fe1d1ace12db253637f9df278cf1639" name="a0fe1d1ace12db253637f9df278cf1639"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0fe1d1ace12db253637f9df278cf1639">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_genotypes()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_genoptions_save_genotypes </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>fg</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>tosave</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>save-as-you-go (genotypes/alleles) </p>
<p >Saves genotypes in an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> to a file. Performs null-checks on file pointer and fails silently if checks do. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07823">7823</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-19" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-19-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-19-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-19-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_cgraph.png" border="0" usemap="#asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_cgraph" alt=""/></div>
<map name="asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_cgraph" id="asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_cgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="5,65,160,109"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="220,5,381,49"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="208,73,393,101"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="210,125,391,153"/>
<area shape="rect" href="sim-operations_8c.html#a223d292e567ad088b1582c24b9d894ab" title="Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of eac..." alt="" coords="441,60,691,88"/>
<area shape="rect" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361" title="Prints a matrix of genotype information to a file." alt="" coords="460,151,673,179"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="748,20,860,48"/>
<area shape="rect" href="group__liteget.html#ga0ce82f67cb5e35a1c73bef54138202f6" title="Get the persistent id of a genotype." alt="" coords="762,72,846,100"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="751,124,857,152"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="743,176,865,204"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="739,228,869,256"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="917,247,1060,291"/>
</map>
</div>
<div id="dynsection-20" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-20-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-20-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-20-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_icgraph.png" border="0" usemap="#asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_icgraph" alt=""/></div>
<map name="asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_icgraph" id="asim-operations_8c_a0fe1d1ace12db253637f9df278cf1639_icgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a311c08610c1a020fb05be94297d6f28f" name="a311c08610c1a020fb05be94297d6f28f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a311c08610c1a020fb05be94297d6f28f">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_genotypes_setup()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static FILE * gsc_helper_genoptions_save_genotypes_setup </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Opens file for writing save-as-you-go genotypes in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07769">7769</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-21" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-21-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-21-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-21-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a311c08610c1a020fb05be94297d6f28f_cgraph.png" border="0" usemap="#asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_cgraph" alt=""/></div>
<map name="asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_cgraph" id="asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_cgraph">
<area shape="rect" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="5,103,166,147"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="214,111,394,139"/>
<area shape="rect" href="sim-operations_8c.html#a223d292e567ad088b1582c24b9d894ab" title="Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of eac..." alt="" coords="442,45,692,73"/>
<area shape="rect" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361" title="Prints a matrix of genotype information to a file." alt="" coords="461,136,674,164"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="749,5,861,33"/>
<area shape="rect" href="group__liteget.html#ga0ce82f67cb5e35a1c73bef54138202f6" title="Get the persistent id of a genotype." alt="" coords="763,57,847,85"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="752,109,858,137"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="744,161,866,189"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="740,213,870,241"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="918,232,1061,276"/>
</map>
</div>
<div id="dynsection-22" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-22-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-22-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-22-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a311c08610c1a020fb05be94297d6f28f_icgraph.png" border="0" usemap="#asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_icgraph" alt=""/></div>
<map name="asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_icgraph" id="asim-operations_8c_a311c08610c1a020fb05be94297d6f28f_icgraph">
<area shape="rect" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="689,132,850,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="aa66b6b0ec2f03012e39df20c6f35b532" name="aa66b6b0ec2f03012e39df20c6f35b532"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa66b6b0ec2f03012e39df20c6f35b532">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_pedigrees()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_genoptions_save_pedigrees </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>fp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>tosave</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>save-as-you-go (pedigrees) </p>
<p >Saves pedigrees in <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> <em>tosave</em> to a file. Performs null-checks on file pointer <em>fp</em>, and fails silently if checks do. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07793">7793</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-23" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-23-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-23-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-23-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_cgraph.png" border="0" usemap="#asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_cgraph" alt=""/></div>
<map name="asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_cgraph" id="asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_cgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="5,208,160,252"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="220,148,381,192"/>
<area shape="rect" href="group__deletors.html#ga83d81991036aca26e2ec5b03f8b33b8a" title="Deletes a gsc_BidirectionalIterator object." alt="" coords="208,216,393,244"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="212,268,389,296"/>
<area shape="rect" href="group__liteget.html#ga847569827ccdd09ff54808c09aba742c" title="Get the first/left parent of a genotype." alt="" coords="472,5,610,33"/>
<area shape="rect" href="group__liteget.html#ga0ce82f67cb5e35a1c73bef54138202f6" title="Get the persistent id of a genotype." alt="" coords="499,57,583,85"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="488,109,594,137"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="695,237,835,265"/>
<area shape="rect" href="group__liteget.html#gab71ce9343a19dd2fc34da1af89fee00a" title="Get the second/right parent of a genotype." alt="" coords="462,237,621,265"/>
<area shape="rect" href="sim-operations_8c.html#a52ddaa114ab4d9e277295dee39c60c48" title="Kernel for scaffold functions that require printing an integer to a file (as opposed to saving the in..." alt="" coords="471,289,612,333"/>
<area shape="rect" href="sim-operations_8c.html#af5808811b84717ecb459de3bc7565494" title="Kernel for scaffold functions that require printing a string to a file (as opposed to saving the stri..." alt="" coords="471,357,612,401"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="477,477,606,505"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="693,495,837,539"/>
<area shape="rect" href="sim-operations_8c.html#a8102058ad778cb4a23bd645d6786244d" title="Identifies and saves (recursively) the pedigree of a pair of parents." alt="" coords="441,185,641,213"/>
<area shape="rect" href="group__supporters.html#ga094a7c1d7f83050cce4d00fdcc89d6b3" title="Binary search through list of unsigned integers." alt="" coords="889,204,1038,248"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="689,185,841,213"/>
</map>
</div>
<div id="dynsection-24" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-24-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-24-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-24-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_icgraph.png" border="0" usemap="#asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_icgraph" alt=""/></div>
<map name="asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_icgraph" id="asim-operations_8c_aa66b6b0ec2f03012e39df20c6f35b532_icgraph">
<area shape="rect" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a2013669171acf8fd19e0ca77c75a3753" name="a2013669171acf8fd19e0ca77c75a3753"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2013669171acf8fd19e0ca77c75a3753">&#9670;&nbsp;</a></span>gsc_helper_genoptions_save_pedigrees_setup()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static FILE * gsc_helper_genoptions_save_pedigrees_setup </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Opens file for writing save-as-you-go pedigrees in accordance with <a class="el" href="structgsc___gen_options.html" title="A type that contains choices of settings for gsc_SimData functions that create a new gsc_AlleleMatrix...">gsc_GenOptions</a>. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l07728">7728</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-25" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-25-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-25-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-25-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a2013669171acf8fd19e0ca77c75a3753_icgraph.png" border="0" usemap="#asim-operations_8c_a2013669171acf8fd19e0ca77c75a3753_icgraph" alt=""/></div>
<map name="asim-operations_8c_a2013669171acf8fd19e0ca77c75a3753_icgraph" id="asim-operations_8c_a2013669171acf8fd19e0ca77c75a3753_icgraph">
<area shape="rect" title="Opens file for writing save&#45;as&#45;you&#45;go pedigrees in accordance with gsc_GenOptions." alt="" coords="689,132,846,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="aaaec31480065a89defa3e02a0f902cb8" name="aaaec31480065a89defa3e02a0f902cb8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaaec31480065a89defa3e02a0f902cb8">&#9670;&nbsp;</a></span>gsc_helper_genotype_matrix_identify_cell_style()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static enum <a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741">gsc_GenotypeFileCellStyle</a> gsc_helper_genotype_matrix_identify_cell_style </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a>&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Identify what formatting a genotype matrix is representing alleles as. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741" title="Represent possible representations of alleles at a marker in a genotype file.">gsc_GenotypeFileCellStyle</a></dd></dl>
<p>For IUPAC encoding: <a href="https://genome.ucsc.edu/goldenPath/help/iupac.html">https://genome.ucsc.edu/goldenPath/help/iupac.html</a> </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06205">6205</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-26" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-26-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-26-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-26-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aaaec31480065a89defa3e02a0f902cb8_icgraph.png" border="0" usemap="#asim-operations_8c_aaaec31480065a89defa3e02a0f902cb8_icgraph" alt=""/></div>
<map name="asim-operations_8c_aaaec31480065a89defa3e02a0f902cb8_icgraph" id="asim-operations_8c_aaaec31480065a89defa3e02a0f902cb8_icgraph">
<area shape="rect" title="Identify what formatting a genotype matrix is representing alleles as." alt="" coords="794,40,968,84"/>
<area shape="rect" href="sim-operations_8c.html#ac22c96ec7ddd91b676805d17de44689c" title="Determine the style in which alleles are stored in a genotype matrix." alt="" coords="583,5,746,49"/>
<area shape="rect" href="sim-operations_8c.html#a410d2515f415529d8efe2abd9ab62048" title="Determine whether a genotype matrix has a header row or not." alt="" coords="584,73,745,117"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,40,535,84"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,48,338,76"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,48,155,76"/>
</map>
</div>

</div>
</div>
<a id="a7c6136857e4f65bff718565b48e18afa" name="a7c6136857e4f65bff718565b48e18afa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7c6136857e4f65bff718565b48e18afa">&#9670;&nbsp;</a></span>gsc_helper_genotypecell_to_allelematrix()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_genotypecell_to_allelematrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>markerix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">enum <a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741">gsc_GenotypeFileCellStyle</a>&#160;</td>
          <td class="paramname"><em>style</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>cell</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>forrng</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Parse a string and save it as the alleles of a genotype at a particular location and genetic marker. </p>
<p >If the style in which the allele string is encoded does not denote phase, linkage phase is randomly chosen. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06264">6264</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-27" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-27-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-27-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-27-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a7c6136857e4f65bff718565b48e18afa_icgraph.png" border="0" usemap="#asim-operations_8c_a7c6136857e4f65bff718565b48e18afa_icgraph" alt=""/></div>
<map name="asim-operations_8c_a7c6136857e4f65bff718565b48e18afa_icgraph" id="asim-operations_8c_a7c6136857e4f65bff718565b48e18afa_icgraph">
<area shape="rect" title="Parse a string and save it as the alleles of a genotype at a particular location and genetic marker." alt="" coords="583,5,748,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="ac22c96ec7ddd91b676805d17de44689c" name="ac22c96ec7ddd91b676805d17de44689c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac22c96ec7ddd91b676805d17de44689c">&#9670;&nbsp;</a></span>gsc_helper_genotypefile_matrix_detect_cellstyle()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> gsc_helper_genotypefile_matrix_detect_cellstyle </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *&#160;</td>
          <td class="paramname"><em>cellqueue</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>firstrowlen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>queuelen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td>
          <td class="paramname"><em>format</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filenameforlog</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Determine the style in which alleles are stored in a genotype matrix. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cellqueue</td><td>array of cells read from the matrix. It should contain at least one body cell, if the matrix has any boy cells: that is, at least two cells if <em>format.has_header</em> is false, or an entire first row plus two cells if <em>format.has_header</em> is unknown or true. </td></tr>
    <tr><td class="paramname">firstrowlen</td><td>number of cells in the first row of the matrix. </td></tr>
    <tr><td class="paramname">queuelen</td><td>number of cells saved sequentially in <em>cellqueue</em>. It is assumed that queuelen &gt;= firstrowlen. </td></tr>
    <tr><td class="paramname">format</td><td>established details about the formatting of the genotype matrix </td></tr>
    <tr><td class="paramname">filenameforlog</td><td>The name of the matrix file or other identifier for this matrix that should be used when printing out logging information. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a format specification for the genotype matrix, with details about the style of allele encoding used in the body of the matrix added if those were not already supplied. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06552">6552</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-28" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-28-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-28-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-28-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_cgraph.png" border="0" usemap="#asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_cgraph" alt=""/></div>
<map name="asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_cgraph" id="asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_cgraph">
<area shape="rect" title="Determine the style in which alleles are stored in a genotype matrix." alt="" coords="5,5,168,49"/>
<area shape="rect" href="sim-operations_8c.html#aaaec31480065a89defa3e02a0f902cb8" title="Identify what formatting a genotype matrix is representing alleles as." alt="" coords="216,5,390,49"/>
</map>
</div>
<div id="dynsection-29" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-29-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-29-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-29-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_icgraph.png" border="0" usemap="#asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_icgraph" alt=""/></div>
<map name="asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_icgraph" id="asim-operations_8c_ac22c96ec7ddd91b676805d17de44689c_icgraph">
<area shape="rect" title="Determine the style in which alleles are stored in a genotype matrix." alt="" coords="583,5,746,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="ae903fb523337566ac34b578f02435c2c" name="ae903fb523337566ac34b578f02435c2c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae903fb523337566ac34b578f02435c2c">&#9670;&nbsp;</a></span>gsc_helper_genotypefile_matrix_detect_cornercell_presence()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a> gsc_helper_genotypefile_matrix_detect_cornercell_presence </td>
          <td>(</td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>ncellsfirstrow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>ncellssecondrow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const _Bool&#160;</td>
          <td class="paramname"><em>secondrowheaderisempty</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Determine whether a genotype matrix has a corner cell or not. </p>
<p >A corner cell is a non-blank value in the first-row/first-column of the matrix, if the matrix has both a header row and header column.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">ncellsfirstrow</td><td>number of cells in the first row of the matrix. </td></tr>
    <tr><td class="paramname">ncellssecondrow</td><td>number of cells in the second row of the matrix. </td></tr>
    <tr><td class="paramname">secondrowheaderisempty</td><td>truthy if the value in the first column of the second row of the matrix is blank, falsy if there is a non-blank value in the first cell of that second row of the matrix. (note: genomicSimulation's TableFileReader takes any blank space (including spaces at the beginning of a row) as cell dividers. That interface is the assumed input for calculating these parameters). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>GSC_TRUE if there is a corner cell, GSC_FALSE if there is not a corner cell, and GSC_NA if this was unable to be determined. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06694">6694</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-30" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-30-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-30-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-30-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ae903fb523337566ac34b578f02435c2c_icgraph.png" border="0" usemap="#asim-operations_8c_ae903fb523337566ac34b578f02435c2c_icgraph" alt=""/></div>
<map name="asim-operations_8c_ae903fb523337566ac34b578f02435c2c_icgraph" id="asim-operations_8c_ae903fb523337566ac34b578f02435c2c_icgraph">
<area shape="rect" title="Determine whether a genotype matrix has a corner cell or not." alt="" coords="583,5,754,65"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,13,535,57"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,21,338,49"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,21,155,49"/>
</map>
</div>

</div>
</div>
<a id="a410d2515f415529d8efe2abd9ab62048" name="a410d2515f415529d8efe2abd9ab62048"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a410d2515f415529d8efe2abd9ab62048">&#9670;&nbsp;</a></span>gsc_helper_genotypefile_matrix_detect_header()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> gsc_helper_genotypefile_matrix_detect_header </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *&#160;</td>
          <td class="paramname"><em>cellqueue</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>firstrowlen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>queuelen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td>
          <td class="paramname"><em>format</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filenameforlog</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Determine whether a genotype matrix has a header row or not. </p>
<p >It is assumed that, during the detection process, </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="sim-operations_8c.html#a39561a6563fa3b016f8833252acc5e5a" title="Determine whether a genotype matrix is row- or column-oriented.">gsc_helper_genotypefile_matrix_detect_orientation</a> and </dd>
<dd>
<a class="el" href="sim-operations_8c.html#ac22c96ec7ddd91b676805d17de44689c" title="Determine the style in which alleles are stored in a genotype matrix.">gsc_helper_genotypefile_matrix_detect_cellstyle</a> will be called before this function. Doing so in the opposite order risks that the detection functions assume a file has an invalid configuration even though there is another potential configuration that would be valid to load.</dd></dl>
<p>This function does not guarantee that the returned format has a known (true or false) has_header field. has_header may still be unknown.</p>
<p >The decision as to whether the genotype matrix has a header row is based on:</p><ul>
<li>If the file is a single column, then it has no header row.</li>
<li>If any cell in the first row does not match <em>format.cell_style</em>, then the first row is assumed to be a header row.</li>
<li>If all cells in the first row match <em>format.cell_style</em>, then there is assumed to be no header row.</li>
</ul>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cellqueue</td><td>array of cells read from the matrix. It should contain at least two cells, and ideally the whole of the first row of the matrix. </td></tr>
    <tr><td class="paramname">firstrowlen</td><td>number of cells in the first row of the matrix. </td></tr>
    <tr><td class="paramname">queuelen</td><td>number of cells saved sequentially in <em>cellqueue</em>. It is assumed that queuelen &gt;= firstrowlen. </td></tr>
    <tr><td class="paramname">format</td><td>established details about the formatting of the genotype matrix </td></tr>
    <tr><td class="paramname">filenameforlog</td><td>The name of the matrix file or other identifier for this matrix that should be used when printing out logging information. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a format specification for the genotype matrix, with the field defining whether there is a header row in the matrix added, if that were not already supplied. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06632">6632</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-31" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-31-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-31-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-31-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a410d2515f415529d8efe2abd9ab62048_cgraph.png" border="0" usemap="#asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_cgraph" alt=""/></div>
<map name="asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_cgraph" id="asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_cgraph">
<area shape="rect" title="Determine whether a genotype matrix has a header row or not." alt="" coords="5,5,166,49"/>
<area shape="rect" href="sim-operations_8c.html#aaaec31480065a89defa3e02a0f902cb8" title="Identify what formatting a genotype matrix is representing alleles as." alt="" coords="214,5,388,49"/>
</map>
</div>
<div id="dynsection-32" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-32-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-32-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-32-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a410d2515f415529d8efe2abd9ab62048_icgraph.png" border="0" usemap="#asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_icgraph" alt=""/></div>
<map name="asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_icgraph" id="asim-operations_8c_a410d2515f415529d8efe2abd9ab62048_icgraph">
<area shape="rect" title="Determine whether a genotype matrix has a header row or not." alt="" coords="583,5,744,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a39561a6563fa3b016f8833252acc5e5a" name="a39561a6563fa3b016f8833252acc5e5a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a39561a6563fa3b016f8833252acc5e5a">&#9670;&nbsp;</a></span>gsc_helper_genotypefile_matrix_detect_orientation()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a> gsc_helper_genotypefile_matrix_detect_orientation </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *&#160;</td>
          <td class="paramname"><em>cellqueue</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>firstrowlen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>queuelen</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___genotype_file___matrix_format.html">gsc_GenotypeFile_MatrixFormat</a>&#160;</td>
          <td class="paramname"><em>format</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filenameforlog</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Determine whether a genotype matrix is row- or column-oriented. </p>
<p >The decision of whether a genotype matrix is storing markers as columns or rows is based on the following factors:</p><ul>
<li>If no names from the column headers of the file cannot be matched to marker names from the SimData (either because no markers are tracked by the SimData, because there is no column header in this matrix, or simply because none of the names match names tracked by the SimData), it is assumed that rows in the genotype matrix represent markers.</li>
<li>If any name from the column header of the genotype file can be matched to the name of a genetic marker tracked by the SimData, then it is assumed that columns in the genotype matrix represent markers.</li>
</ul>
<p >genomicSimulation requires marker names for file loading, so if markers are represented by columns in the matrix, the format specifier returned will also specify that the matrix does have a header row.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>SimData object the genotype file is assumed to be loaded to. Required because determining matrix orientation requires checking marker names against the markers tracked in the SimData. </td></tr>
    <tr><td class="paramname">cellqueue</td><td>array of cells read from the matrix. Requires at least the first row of the matrix, preferably the first row of the matrix plus the first two cells (header and first body cell) of the second row of the matrix. </td></tr>
    <tr><td class="paramname">firstrowlen</td><td>number of cells in the first row of the matrix. These cells can be accessed at indices 0-{firstrowlen-1} in <em>cellqueue</em>. </td></tr>
    <tr><td class="paramname">queuelen</td><td>number of cells saved sequentially in <em>cellqueue</em>. It is assumed that queuelen &gt;= firstrowlen. </td></tr>
    <tr><td class="paramname">format</td><td>established details about the formatting of the genotype matrix </td></tr>
    <tr><td class="paramname">filenameforlog</td><td>The name of the matrix file or other identifier for this matrix that should be used when printing out logging information. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a format specification for the genotype matrix, with details about the orientation of the matrix added if those were not already supplied. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06471">6471</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-33" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-33-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-33-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-33-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_cgraph.png" border="0" usemap="#asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_cgraph" alt=""/></div>
<map name="asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_cgraph" id="asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_cgraph">
<area shape="rect" title="Determine whether a genotype matrix is row&#45; or column&#45;oriented." alt="" coords="5,5,180,49"/>
<area shape="rect" href="group__supporters.html#gad57230a4c812ba8a6deed26100beae7b" title="Return whether or not a marker name is present in the tracked markers, and at what index." alt="" coords="228,5,399,49"/>
</map>
</div>
<div id="dynsection-34" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-34-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-34-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-34-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_icgraph.png" border="0" usemap="#asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_icgraph" alt=""/></div>
<map name="asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_icgraph" id="asim-operations_8c_a39561a6563fa3b016f8833252acc5e5a_icgraph">
<area shape="rect" title="Determine whether a genotype matrix is row&#45; or column&#45;oriented." alt="" coords="583,5,758,49"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="acf2e8ece7fb3470e814da69e8f162813" name="acf2e8ece7fb3470e814da69e8f162813"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acf2e8ece7fb3470e814da69e8f162813">&#9670;&nbsp;</a></span>gsc_helper_indirect_alphabetical_str_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_indirect_alphabetical_str_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of pointers to strings to sort them in alphabetical order of the strings. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01177">1177</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-35" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-35-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-35-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-35-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_acf2e8ece7fb3470e814da69e8f162813_icgraph.png" border="0" usemap="#asim-operations_8c_acf2e8ece7fb3470e814da69e8f162813_icgraph" alt=""/></div>
<map name="asim-operations_8c_acf2e8ece7fb3470e814da69e8f162813_icgraph" id="asim-operations_8c_acf2e8ece7fb3470e814da69e8f162813_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="583,36,764,80"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="400,73,521,101"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,44,338,72"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,44,155,72"/>
</map>
</div>

</div>
</div>
<a id="a3717c5ed02cf2a321b7424dc0e668e19" name="a3717c5ed02cf2a321b7424dc0e668e19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3717c5ed02cf2a321b7424dc0e668e19">&#9670;&nbsp;</a></span>gsc_helper_insert_eff_set_into_simdata()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a> gsc_helper_insert_eff_set_into_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___effect_matrix.html">gsc_EffectMatrix</a>&#160;</td>
          <td class="paramname"><em>effset</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Save an EffectMatrix to the SimData and allocate it an EffectID. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05625">5625</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-36" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-36-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-36-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-36-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_cgraph.png" border="0" usemap="#asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_cgraph" alt=""/></div>
<map name="asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_cgraph" id="asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_cgraph">
<area shape="rect" title="Save an EffectMatrix to the SimData and allocate it an EffectID." alt="" coords="5,24,149,68"/>
<area shape="rect" href="group__supporters.html#ga7881c537673d2d2d9dc3bb285676bfd4" title="Function to identify the next sequential integer that is not already allocated to a marker effect set..." alt="" coords="197,5,356,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="216,57,337,85"/>
</map>
</div>
<div id="dynsection-37" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-37-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-37-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-37-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_icgraph.png" border="0" usemap="#asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_icgraph" alt=""/></div>
<map name="asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_icgraph" id="asim-operations_8c_a3717c5ed02cf2a321b7424dc0e668e19_icgraph">
<area shape="rect" title="Save an EffectMatrix to the SimData and allocate it an EffectID." alt="" coords="561,5,705,49"/>
<area shape="rect" href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831" title="Populates a gsc_SimData combination with effect values." alt="" coords="386,13,513,41"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a337459fb31ef07b88a77068780ab746e" name="a337459fb31ef07b88a77068780ab746e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a337459fb31ef07b88a77068780ab746e">&#9670;&nbsp;</a></span>gsc_helper_insert_recombmap_into_simdata()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a> gsc_helper_insert_recombmap_into_simdata </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___recombination_map.html">gsc_RecombinationMap</a>&#160;</td>
          <td class="paramname"><em>map</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Save a RecombinationMap to the SimData and allocate it a mapID. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05597">5597</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-38" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-38-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-38-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-38-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a337459fb31ef07b88a77068780ab746e_cgraph.png" border="0" usemap="#asim-operations_8c_a337459fb31ef07b88a77068780ab746e_cgraph" alt=""/></div>
<map name="asim-operations_8c_a337459fb31ef07b88a77068780ab746e_cgraph" id="asim-operations_8c_a337459fb31ef07b88a77068780ab746e_cgraph">
<area shape="rect" title="Save a RecombinationMap to the SimData and allocate it a mapID." alt="" coords="5,24,204,68"/>
<area shape="rect" href="group__supporters.html#ga48ab10327d06674174428ca2db0b32ed" title="Function to identify the next sequential integer that is not already allocated to a map ID in the sim..." alt="" coords="252,5,398,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="264,57,386,85"/>
</map>
</div>
<div id="dynsection-39" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-39-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-39-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-39-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a337459fb31ef07b88a77068780ab746e_icgraph.png" border="0" usemap="#asim-operations_8c_a337459fb31ef07b88a77068780ab746e_icgraph" alt=""/></div>
<map name="asim-operations_8c_a337459fb31ef07b88a77068780ab746e_icgraph" id="asim-operations_8c_a337459fb31ef07b88a77068780ab746e_icgraph">
<area shape="rect" title="Save a RecombinationMap to the SimData and allocate it a mapID." alt="" coords="808,40,1006,84"/>
<area shape="rect" href="group__loaders.html#gada5c2aaacd8b8a145493a0db5c017068" title="Parse a list of markers/chrs/positions into a gsc_RecombinationMap and save to SimData." alt="" coords="591,5,752,49"/>
<area shape="rect" href="group__loaders.html#ga5cbe0b821180b257b21d9551ec797f20" title="Create a uniformly&#45;spaced gsc_RecombinationMap from a list of marker names and save to SimData." alt="" coords="583,73,760,117"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="400,16,521,44"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,48,338,76"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,48,155,76"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,69,535,113"/>
</map>
</div>

</div>
</div>
<a id="a7c0584559c66e7c6da1452f9a804958d" name="a7c0584559c66e7c6da1452f9a804958d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7c0584559c66e7c6da1452f9a804958d">&#9670;&nbsp;</a></span>gsc_helper_is_marker_in_chr()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a> gsc_helper_is_marker_in_chr </td>
          <td>(</td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>markerix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___linkage_group.html">gsc_LinkageGroup</a>&#160;</td>
          <td class="paramname"><em>chr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>pos</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Check if a marker index is found in a particular LinkageGroup, and provide its distance along the chromosome/linkage group if so. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10319">10319</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-40" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-40-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-40-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-40-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a7c0584559c66e7c6da1452f9a804958d_icgraph.png" border="0" usemap="#asim-operations_8c_a7c0584559c66e7c6da1452f9a804958d_icgraph" alt=""/></div>
<map name="asim-operations_8c_a7c0584559c66e7c6da1452f9a804958d_icgraph" id="asim-operations_8c_a7c0584559c66e7c6da1452f9a804958d_icgraph">
<area shape="rect" title="Check if a marker index is found in a particular LinkageGroup, and provide its distance along the chr..." alt="" coords="460,5,608,49"/>
<area shape="rect" href="group__savers.html#ga193bc97c400e318504cdc9bd03684095" title="Prints the markers contained in a set of blocks to a file." alt="" coords="214,13,412,41"/>
<area shape="rect" href="group__savers.html#ga26462f6816264c01ce29e3e56efb2dc3" title="Prints the markers contained in a set of blocks to a file." alt="" coords="5,13,166,41"/>
</map>
</div>

</div>
</div>
<a id="ab0152d52dcce6514ebba0fa08810ee8a" name="ab0152d52dcce6514ebba0fa08810ee8a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab0152d52dcce6514ebba0fa08810ee8a">&#9670;&nbsp;</a></span>gsc_helper_make_offspring_clones()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_make_offspring_clones </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>putHere</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>offspringGenerator function parameter for gsc_make_clones. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group.">gsc_make_clones</a></dd></dl>
<p>Generates new alleles by making a clone of its parent. Only looks at the first parent and does not check it is valid. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08752">8752</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-41" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-41-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-41-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-41-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_cgraph.png" border="0" usemap="#asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_cgraph" alt=""/></div>
<map name="asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_cgraph" id="asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_cgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_make_clones." alt="" coords="5,76,184,120"/>
<area shape="rect" href="group__meiosis.html#ga3793a9acc76baad86c7f3c31afd71999" title="Get an identical copy of a given genotype." alt="" coords="232,5,369,33"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="245,57,356,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="240,109,361,137"/>
<area shape="rect" href="group__liteget.html#gadff19e04cab018dd64ae232e0adcb3cc" title="Set the name of a genotype." alt="" coords="247,161,354,189"/>
</map>
</div>
<div id="dynsection-42" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-42-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-42-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-42-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_icgraph.png" border="0" usemap="#asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_icgraph" alt=""/></div>
<map name="asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_icgraph" id="asim-operations_8c_ab0152d52dcce6514ebba0fa08810ee8a_icgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_make_clones." alt="" coords="180,5,358,49"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="5,13,132,41"/>
</map>
</div>

</div>
</div>
<a id="aa16bd09b22305a169b4ca8e54432f966" name="aa16bd09b22305a169b4ca8e54432f966"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa16bd09b22305a169b4ca8e54432f966">&#9670;&nbsp;</a></span>gsc_helper_make_offspring_cross()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_make_offspring_cross </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>putHere</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>offspringGenerator function parameter for all crossing functions </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group.">gsc_make_random_crosses</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another.">gsc_make_random_crosses_between</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array.">gsc_make_targeted_crosses</a></dd></dl>
<p>Generates new alleles from two separate parents. Does not check they are valid parents.</p>
<p ><em>datastore</em> parameter goes unused </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08081">8081</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-43" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-43-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-43-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-43-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_cgraph.png" border="0" usemap="#asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_cgraph" alt=""/></div>
<map name="asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_cgraph" id="asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_cgraph">
<area shape="rect" title="offspringGenerator function parameter for all crossing functions" alt="" coords="5,61,184,105"/>
<area shape="rect" href="group__meiosis.html#ga923d9402fffe6eacac7971b703343073" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="232,43,381,71"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="251,95,362,123"/>
<area shape="rect" href="sim-operations_8c.html#a50de1f9769153df70263edc86de41fa5" title="Comparator function for qsort." alt="" coords="429,5,579,49"/>
<area shape="rect" href="group__maths.html#ga897c7abb536f30f0cbc1baf2abcd2067" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="459,73,550,101"/>
</map>
</div>
<div id="dynsection-44" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-44-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-44-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-44-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_icgraph.png" border="0" usemap="#asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_icgraph" alt=""/></div>
<map name="asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_icgraph" id="asim-operations_8c_aa16bd09b22305a169b4ca8e54432f966_icgraph">
<area shape="rect" title="offspringGenerator function parameter for all crossing functions" alt="" coords="479,57,657,101"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,5,429,33"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,57,429,101"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,125,431,153"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,57,192,101"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,125,195,153"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,177,190,221"/>
</map>
</div>

</div>
</div>
<a id="a226a60f8e133bcbb389863cd4dbd1898" name="a226a60f8e133bcbb389863cd4dbd1898"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a226a60f8e133bcbb389863cd4dbd1898">&#9670;&nbsp;</a></span>gsc_helper_make_offspring_doubled_haploids()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_make_offspring_doubled_haploids </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>putHere</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>offspringGenerator function parameter for gsc_make_doubled_haploids. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group.">gsc_make_doubled_haploids</a></dd></dl>
<p>Generates new alleles by making a doubled haploid. Only looks at the first parent and does not check it is valid. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08653">8653</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-45" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-45-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-45-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-45-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_cgraph.png" border="0" usemap="#asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_cgraph" alt=""/></div>
<map name="asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_cgraph" id="asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_cgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_make_doubled_haploids." alt="" coords="5,65,184,109"/>
<area shape="rect" href="group__meiosis.html#ga6a07e899a26f8aff2b241e3a6f7d8665" title="Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent." alt="" coords="232,35,383,79"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="252,103,363,131"/>
<area shape="rect" href="sim-operations_8c.html#a50de1f9769153df70263edc86de41fa5" title="Comparator function for qsort." alt="" coords="431,5,581,49"/>
<area shape="rect" href="group__maths.html#ga897c7abb536f30f0cbc1baf2abcd2067" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="461,73,552,101"/>
</map>
</div>
<div id="dynsection-46" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-46-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-46-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-46-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_icgraph.png" border="0" usemap="#asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_icgraph" alt=""/></div>
<map name="asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_icgraph" id="asim-operations_8c_a226a60f8e133bcbb389863cd4dbd1898_icgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_make_doubled_haploids." alt="" coords="242,5,420,49"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="5,13,194,41"/>
</map>
</div>

</div>
</div>
<a id="a42606fdc7adae265c1663ef40d45226b" name="a42606fdc7adae265c1663ef40d45226b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42606fdc7adae265c1663ef40d45226b">&#9670;&nbsp;</a></span>gsc_helper_make_offspring_self_n_times()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_make_offspring_self_n_times </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>putHere</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>offspringGenerator function parameter for gsc_self_n_times. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations.">gsc_self_n_times</a></dd></dl>
<p>Datastore should be a pointer to an unsigned int containing the number of generations to self for.</p>
<p >Generates new alleles by selfing from one parent. Only looks at the first parent and does not check it is valid. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08555">8555</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-47" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-47-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-47-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-47-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a42606fdc7adae265c1663ef40d45226b_cgraph.png" border="0" usemap="#asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_cgraph" alt=""/></div>
<map name="asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_cgraph" id="asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_cgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_self_n_times." alt="" coords="5,61,184,105"/>
<area shape="rect" href="group__meiosis.html#ga923d9402fffe6eacac7971b703343073" title="Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alle..." alt="" coords="232,43,381,71"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="251,95,362,123"/>
<area shape="rect" href="sim-operations_8c.html#a50de1f9769153df70263edc86de41fa5" title="Comparator function for qsort." alt="" coords="429,5,579,49"/>
<area shape="rect" href="group__maths.html#ga897c7abb536f30f0cbc1baf2abcd2067" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="459,73,550,101"/>
</map>
</div>
<div id="dynsection-48" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-48-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-48-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-48-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a42606fdc7adae265c1663ef40d45226b_icgraph.png" border="0" usemap="#asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_icgraph" alt=""/></div>
<map name="asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_icgraph" id="asim-operations_8c_a42606fdc7adae265c1663ef40d45226b_icgraph">
<area shape="rect" title="offspringGenerator function parameter for gsc_self_n_times." alt="" coords="176,5,354,49"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="5,13,128,41"/>
</map>
</div>

</div>
</div>
<a id="aaa406c38a85c3953b1cda622ff535d12" name="aaa406c38a85c3953b1cda622ff535d12"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaa406c38a85c3953b1cda622ff535d12">&#9670;&nbsp;</a></span>gsc_helper_mapfileunit_ascending_chr_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_mapfileunit_ascending_chr_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of gsc_MapFileUnits to sort them in ascending order of chromosome number. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01186">1186</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-49" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-49-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-49-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-49-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aaa406c38a85c3953b1cda622ff535d12_icgraph.png" border="0" usemap="#asim-operations_8c_aaa406c38a85c3953b1cda622ff535d12_icgraph" alt=""/></div>
<map name="asim-operations_8c_aaa406c38a85c3953b1cda622ff535d12_icgraph" id="asim-operations_8c_aaa406c38a85c3953b1cda622ff535d12_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="781,5,954,49"/>
<area shape="rect" href="sim-operations_8c.html#ac1d97be6bdb21bbb271fc44616461fd8" title="Sort markerlist by chromosome name, and by position within each chromosome." alt="" coords="555,13,733,41"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="386,13,507,41"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a6899282e8507284be584200d05e786cf" name="a6899282e8507284be584200d05e786cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6899282e8507284be584200d05e786cf">&#9670;&nbsp;</a></span>gsc_helper_mapfileunit_ascending_d_comparer()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_mapfileunit_ascending_d_comparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>p1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare a pair of gsc_MapFileUnits to sort them in ascending order of position. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01196">1196</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-50" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-50-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-50-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-50-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a6899282e8507284be584200d05e786cf_icgraph.png" border="0" usemap="#asim-operations_8c_a6899282e8507284be584200d05e786cf_icgraph" alt=""/></div>
<map name="asim-operations_8c_a6899282e8507284be584200d05e786cf_icgraph" id="asim-operations_8c_a6899282e8507284be584200d05e786cf_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="781,5,943,49"/>
<area shape="rect" href="sim-operations_8c.html#ac1d97be6bdb21bbb271fc44616461fd8" title="Sort markerlist by chromosome name, and by position within each chromosome." alt="" coords="555,13,733,41"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="386,13,507,41"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="aca57b78a1df91391b9be67b3dcd46016" name="aca57b78a1df91391b9be67b3dcd46016"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aca57b78a1df91391b9be67b3dcd46016">&#9670;&nbsp;</a></span>gsc_helper_output_countmatrix_cell()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_output_countmatrix_cell </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>markerix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the allele counts of a particular allele for each genetic marker. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10614">10614</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-51" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-51-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-51-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-51-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aca57b78a1df91391b9be67b3dcd46016_icgraph.png" border="0" usemap="#asim-operations_8c_aca57b78a1df91391b9be67b3dcd46016_icgraph" alt=""/></div>
<map name="asim-operations_8c_aca57b78a1df91391b9be67b3dcd46016_icgraph" id="asim-operations_8c_aca57b78a1df91391b9be67b3dcd46016_icgraph">
<area shape="rect" title="Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the allele counts of a particula..." alt="" coords="410,13,639,41"/>
<area shape="rect" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file." alt="" coords="212,5,362,49"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="5,13,164,41"/>
</map>
</div>

</div>
</div>
<a id="a223d292e567ad088b1582c24b9d894ab" name="a223d292e567ad088b1582c24b9d894ab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a223d292e567ad088b1582c24b9d894ab">&#9670;&nbsp;</a></span>gsc_helper_output_genotypematrix_cell()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_output_genotypematrix_cell </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>markerix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>GSC_NA</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of each genotype. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10603">10603</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-52" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-52-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-52-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-52-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a223d292e567ad088b1582c24b9d894ab_cgraph.png" border="0" usemap="#asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_cgraph" alt=""/></div>
<map name="asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_cgraph" id="asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_cgraph">
<area shape="rect" title="Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of eac..." alt="" coords="5,5,256,33"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="304,5,415,33"/>
</map>
</div>
<div id="dynsection-53" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-53-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-53-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-53-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a223d292e567ad088b1582c24b9d894ab_icgraph.png" border="0" usemap="#asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_icgraph" alt=""/></div>
<map name="asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_icgraph" id="asim-operations_8c_a223d292e567ad088b1582c24b9d894ab_icgraph">
<area shape="rect" title="Kernel for gsc_scaffold_save_genotype_info, when the goal is to save the (phased) allele pairs of eac..." alt="" coords="1126,173,1376,201"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="898,173,1078,201"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="692,97,847,141"/>
<area shape="rect" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="689,165,850,209"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="698,233,841,261"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a94b58634907af7e24ac22b543b8fcbe3" name="a94b58634907af7e24ac22b543b8fcbe3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a94b58634907af7e24ac22b543b8fcbe3">&#9670;&nbsp;</a></span>gsc_helper_parentchooser_cloning()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_parentchooser_cloning </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>parentIterator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>counter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>parentChooser function parameter for gsc_make_clones. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group.">gsc_make_clones</a></dd></dl>
<p>Very similar to <a class="el" href="sim-operations_8c.html#aafe958ddf4ee8c990c521d6ebf60db71">gsc_helper_parentchooser_selfing</a>, but also accesses the datastore to see if it needs to inherit the name from its parent.</p>
<p >Locates the next group member. Guarantees <em>parents</em> contains two valid gsc_GenoLocations (which are the same) at the time it returns a truthy value. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08725">8725</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-54" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-54-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-54-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-54-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_cgraph.png" border="0" usemap="#asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_cgraph" alt=""/></div>
<map name="asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_cgraph" id="asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_cgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_clones." alt="" coords="5,24,179,68"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="238,5,345,33"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="227,57,356,85"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="404,49,547,93"/>
</map>
</div>
<div id="dynsection-55" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-55-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-55-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-55-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_icgraph.png" border="0" usemap="#asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_icgraph" alt=""/></div>
<map name="asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_icgraph" id="asim-operations_8c_a94b58634907af7e24ac22b543b8fcbe3_icgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_clones." alt="" coords="180,5,353,49"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="5,13,132,41"/>
</map>
</div>

</div>
</div>
<a id="a0fcef09a211d57fadb1dfa0b6d102901" name="a0fcef09a211d57fadb1dfa0b6d102901"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0fcef09a211d57fadb1dfa0b6d102901">&#9670;&nbsp;</a></span>gsc_helper_parentchooser_cross_randomly()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_parentchooser_cross_randomly </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>parentIterator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>counter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>parentChooser function parameter for gsc_make_random_crosses. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group.">gsc_make_random_crosses</a></dd></dl>
<p>Chooses parents randomly without overruning parent usage caps and without permitting selfing. Guarantees <em>parents</em> contains two valid gsc_GenoLocations at the time it returns a truthy value. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08029">8029</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-56" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-56-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-56-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-56-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_cgraph.png" border="0" usemap="#asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_cgraph" alt=""/></div>
<map name="asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_cgraph" id="asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_cgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_random_crosses." alt="" coords="5,76,179,120"/>
<area shape="rect" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="277,57,402,85"/>
<area shape="rect" href="group__supporters.html#gafd6535362ff9898f871cd4e1098d2ee6" title="Randomly pick a number in a range, optionally with a cap on how many times a number can be picked,..." alt="" coords="227,109,452,137"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="531,5,637,33"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="500,57,667,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="523,109,644,137"/>
</map>
</div>
<div id="dynsection-57" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-57-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-57-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-57-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_icgraph.png" border="0" usemap="#asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_icgraph" alt=""/></div>
<map name="asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_icgraph" id="asim-operations_8c_a0fcef09a211d57fadb1dfa0b6d102901_icgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_random_crosses." alt="" coords="238,5,411,49"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="5,13,190,41"/>
</map>
</div>

</div>
</div>
<a id="a2cee0d2e90de12744fd077b20a7db9bd" name="a2cee0d2e90de12744fd077b20a7db9bd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2cee0d2e90de12744fd077b20a7db9bd">&#9670;&nbsp;</a></span>gsc_helper_parentchooser_cross_randomly_between()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_parentchooser_cross_randomly_between </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>parentIterator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>counter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>parentChooser function parameter for gsc_make_random_crosses_between. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another.">gsc_make_random_crosses_between</a></dd></dl>
<p>Chooses parents randomly without overruning parent usage caps and without permitting selfing. First and second parent come from different groups and may have different usage caps. Guarantees <em>parents</em> contains two valid gsc_GenoLocations at the time it returns a truthy value. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08259">8259</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-58" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-58-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-58-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-58-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_cgraph.png" border="0" usemap="#asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_cgraph" alt=""/></div>
<map name="asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_cgraph" id="asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_cgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_random_crosses_between." alt="" coords="5,76,181,120"/>
<area shape="rect" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="279,57,404,85"/>
<area shape="rect" href="group__supporters.html#gafd6535362ff9898f871cd4e1098d2ee6" title="Randomly pick a number in a range, optionally with a cap on how many times a number can be picked,..." alt="" coords="229,109,454,137"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="533,5,639,33"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="502,57,669,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="525,109,646,137"/>
</map>
</div>
<div id="dynsection-59" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-59-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-59-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-59-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_icgraph.png" border="0" usemap="#asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_icgraph" alt=""/></div>
<map name="asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_icgraph" id="asim-operations_8c_a2cee0d2e90de12744fd077b20a7db9bd_icgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_random_crosses_between." alt="" coords="238,5,413,49"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="5,5,190,49"/>
</map>
</div>

</div>
</div>
<a id="a2c3d189234e1ac86e27a1ccaace3c363" name="a2c3d189234e1ac86e27a1ccaace3c363"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2c3d189234e1ac86e27a1ccaace3c363">&#9670;&nbsp;</a></span>gsc_helper_parentchooser_cross_targeted()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_parentchooser_cross_targeted </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>parentIterator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>counter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>parentChooser function parameter for gsc_make_targeted_crosses. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array.">gsc_make_targeted_crosses</a></dd></dl>
<p>Locates the next pair of parents in the provided sequence of combinations. Guarantees <em>parents</em> contains two valid gsc_GenoLocations at the time it returns a truthy value. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08416">8416</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-60" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-60-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-60-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-60-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_cgraph.png" border="0" usemap="#asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_cgraph" alt=""/></div>
<map name="asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_cgraph" id="asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_cgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_targeted_crosses." alt="" coords="5,49,179,93"/>
<area shape="rect" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="227,57,351,85"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="430,5,536,33"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="399,57,566,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="422,109,543,137"/>
</map>
</div>
<div id="dynsection-61" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-61-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-61-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-61-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_icgraph.png" border="0" usemap="#asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_icgraph" alt=""/></div>
<map name="asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_icgraph" id="asim-operations_8c_a2c3d189234e1ac86e27a1ccaace3c363_icgraph">
<area shape="rect" title="parentChooser function parameter for gsc_make_targeted_crosses." alt="" coords="479,65,652,109"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,73,431,101"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,5,192,49"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,73,195,101"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,125,190,169"/>
</map>
</div>

</div>
</div>
<a id="aafe958ddf4ee8c990c521d6ebf60db71" name="aafe958ddf4ee8c990c521d6ebf60db71"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aafe958ddf4ee8c990c521d6ebf60db71">&#9670;&nbsp;</a></span>gsc_helper_parentchooser_selfing()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static int gsc_helper_parentchooser_selfing </td>
          <td>(</td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>parentIterator</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">union <a class="el" href="uniongsc__datastore__make__genotypes.html">gsc_datastore_make_genotypes</a> *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int *&#160;</td>
          <td class="paramname"><em>counter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___parent_choice.html">gsc_ParentChoice</a>&#160;</td>
          <td class="paramname"><em>parents</em>[static 2]&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>parentChooser function parameter for gsc_self_n_times. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)">gsc_scaffold_make_new_genotypes</a> </dd>
<dd>
<a class="el" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations.">gsc_self_n_times</a> </dd>
<dd>
<a class="el" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group.">gsc_make_doubled_haploids</a></dd></dl>
<p>Locates the next group member. Guarantees <em>parents</em> contains two valid gsc_GenoLocations (which are the same) at the time it returns a truthy value. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08531">8531</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-62" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-62-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-62-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-62-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_cgraph.png" border="0" usemap="#asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_cgraph" alt=""/></div>
<map name="asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_cgraph" id="asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_cgraph">
<area shape="rect" title="parentChooser function parameter for gsc_self_n_times." alt="" coords="5,5,179,49"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="227,13,356,41"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="404,5,547,49"/>
</map>
</div>
<div id="dynsection-63" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-63-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-63-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-63-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_icgraph.png" border="0" usemap="#asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_icgraph" alt=""/></div>
<map name="asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_icgraph" id="asim-operations_8c_aafe958ddf4ee8c990c521d6ebf60db71_icgraph">
<area shape="rect" title="parentChooser function parameter for gsc_self_n_times." alt="" coords="242,24,415,68"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="5,5,194,33"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="38,57,161,85"/>
</map>
</div>

</div>
</div>
<a id="a11ff196fb220dfd8336587f016767b66" name="a11ff196fb220dfd8336587f016767b66"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a11ff196fb220dfd8336587f016767b66">&#9670;&nbsp;</a></span>gsc_helper_parse_3cell_header()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8f">GSC_LOGICVAL</a> gsc_helper_parse_3cell_header </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *&#160;</td>
          <td class="paramname"><em>tf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>canonical_titles</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int *&#160;</td>
          <td class="paramname"><em>col_order</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> *&#160;</td>
          <td class="paramname"><em>unprocessedqueue</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>queuesize</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Header row reading and processing for map and effect set files. </p>
<p >Task 1: identifies if there is a header (in which case the returned queue will not contain the header cells) or there is no header (in which case the returned queue will contain the first row of cells).</p>
<p >The first three cells are identified as a header row if there is a newline after the third cell (and none between the first three cells), and the entries in those three header cells correspond to the strings in the <em>canonical_titles</em> parameter, in any order. If the first row does match the requirements of a header row, the ordering of those <em>canonical_titles</em> in the header row will be saved to <em>col_order</em>.</p>
<p >The function returns the number of cells read but not processed. These cells read but not processed (i.e. not any identified header cells) will be saved to <em>unprocessedqueue</em>. At maximum 4 cells will be read, so <em>unprocessedqueue</em> must be a pointer with a location for space for at least 4 TableFileCells.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">canonical_titles</td><td>vector of three column titles, which the three potential header cells will be matched against </td></tr>
    <tr><td class="paramname">col_order</td><td>if the first row is determined to be a header row, entry i in this vector will be filled with the column number (1, 2, or 3) that corresponds to the i-th title in <em>canonical_titles</em> </td></tr>
    <tr><td class="paramname">unprocessedqueue</td><td>vector of length at least 4 which will be overwritten with the cells that were read but determined to not be part of the header. </td></tr>
    <tr><td class="paramname">queuesize</td><td>space to save number of entries in the queue <em>unprocessedqueue</em> after execution of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>GSC_TRUE (1) if the cell has a valid header, GSC_FALSE (0) if the cell has no header and GSC_NA (-1) if the cell does not have 3 columns or there is some other formatting error. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05341">5341</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-64" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-64-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-64-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-64-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a11ff196fb220dfd8336587f016767b66_cgraph.png" border="0" usemap="#asim-operations_8c_a11ff196fb220dfd8336587f016767b66_cgraph" alt=""/></div>
<map name="asim-operations_8c_a11ff196fb220dfd8336587f016767b66_cgraph" id="asim-operations_8c_a11ff196fb220dfd8336587f016767b66_cgraph">
<area shape="rect" title="Header row reading and processing for map and effect set files." alt="" coords="5,28,163,72"/>
<area shape="rect" href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d" title="Allocate memory to store a deep copy of a gsc_TableFileCell, if previously only a shallow copy." alt="" coords="211,5,394,33"/>
<area shape="rect" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05" title="Read forwards in TableFileReader and return the next cell&#39;s contents, as well as how many column gaps..." alt="" coords="238,57,367,101"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="467,5,588,33"/>
<area shape="rect" href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0" title="Classify the character under the cursor of a TableFileReader as cell contents or otherwise." alt="" coords="442,57,613,101"/>
<area shape="rect" href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322" title="Read another buffer&#39;s worth of characters from a gsc_TableFileReader&#39;s file." alt="" coords="442,125,613,169"/>
</map>
</div>
<div id="dynsection-65" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-65-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-65-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-65-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a11ff196fb220dfd8336587f016767b66_icgraph.png" border="0" usemap="#asim-operations_8c_a11ff196fb220dfd8336587f016767b66_icgraph" alt=""/></div>
<map name="asim-operations_8c_a11ff196fb220dfd8336587f016767b66_icgraph" id="asim-operations_8c_a11ff196fb220dfd8336587f016767b66_icgraph">
<area shape="rect" title="Header row reading and processing for map and effect set files." alt="" coords="781,24,938,68"/>
<area shape="rect" href="sim-operations_8c.html#ad3c2207b2bd170d752a0a26a9fa2aa19" title="Extract the contents of a genetic map file." alt="" coords="561,5,733,33"/>
<area shape="rect" href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831" title="Populates a gsc_SimData combination with effect values." alt="" coords="386,57,513,85"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="389,5,510,33"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,32,338,60"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,32,155,60"/>
</map>
</div>

</div>
</div>
<a id="ad3c2207b2bd170d752a0a26a9fa2aa19" name="ad3c2207b2bd170d752a0a26a9fa2aa19"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad3c2207b2bd170d752a0a26a9fa2aa19">&#9670;&nbsp;</a></span>gsc_helper_parse_mapfile()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static size_t gsc_helper_parse_mapfile </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> **&#160;</td>
          <td class="paramname"><em>out</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Extract the contents of a genetic map file. </p>
<p >By default, the file's columns are assumed to be in the order:</p><ol type="1">
<li>marker name</li>
<li>chromosome/linkage group (name or number)</li>
<li>position in cM along that group Extra columns are not permitted. If a header row is included, its fields should be "marker", "chr" and "pos". These names can be reordered if the file presents the columns in a different order to the default order shown above.</li>
</ol>
<p >Marker names may contain any characters except tabs, spaces, commas, and newlines. Chromosome names must be integers (eg 1) or alphanumeric strings (eg 1A). Positions should be able to be parsed as a floating point number by the compiling computer's C standard library (usually allows integers, decimals, and scientific notation, but not fractions).</p>
<p >Columns may be space-separated or tab-separated. Column separators need not be consistent throughout the file, and may be repeated. That is, any consecutive sequence of spaces and tabs is interpreted as one column separator. Any single or consecutive pair of '<br  />
' and '\r' characters are inferred as representing a newline. Like with column separators, there is no requirement for consistency across the file in which character or pair of characters represents the end of a row.</p>
<p >In the current implementation, it attempts to load the entire file into memory at a time, so the system must have enough memory to hold the file if it is to run this function.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>name of the file to parse </td></tr>
    <tr><td class="paramname">out</td><td>location to save a heap-allocated vector containing the contents of each line of the file. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>number of lines successfully read from the file </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05431">5431</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-66" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-66-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-66-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-66-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_cgraph.png" border="0" usemap="#asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_cgraph" alt=""/></div>
<map name="asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_cgraph" id="asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_cgraph">
<area shape="rect" title="Extract the contents of a genetic map file." alt="" coords="5,113,177,141"/>
<area shape="rect" href="sim-operations_8c.html#a11ff196fb220dfd8336587f016767b66" title="Header row reading and processing for map and effect set files." alt="" coords="232,37,389,81"/>
<area shape="rect" href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d" title="Allocate memory to store a deep copy of a gsc_TableFileCell, if previously only a shallow copy." alt="" coords="444,15,627,43"/>
<area shape="rect" href="sim-operations_8c.html#a9022b0b2e12395904757c59db6010d5b" title="Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cell..." alt="" coords="225,105,396,149"/>
<area shape="rect" href="group__structs.html#ga85b05dcdc5ec484d918ed2d3330c3ae8" title="Close a gsc_TableFileReader&#39;s file pointer." alt="" coords="246,173,375,217"/>
<area shape="rect" href="group__structs.html#ga1a3a39309d4ce6447581e73f7483fbfb" title="Open a file for reading with gsc_TableFileReader." alt="" coords="246,241,375,285"/>
<area shape="rect" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05" title="Read forwards in TableFileReader and return the next cell&#39;s contents, as well as how many column gaps..." alt="" coords="471,67,600,111"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="700,15,822,43"/>
<area shape="rect" href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0" title="Classify the character under the cursor of a TableFileReader as cell contents or otherwise." alt="" coords="675,67,847,111"/>
<area shape="rect" href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322" title="Read another buffer&#39;s worth of characters from a gsc_TableFileReader&#39;s file." alt="" coords="675,135,847,179"/>
</map>
</div>
<div id="dynsection-67" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-67-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-67-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-67-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_icgraph.png" border="0" usemap="#asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_icgraph" alt=""/></div>
<map name="asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_icgraph" id="asim-operations_8c_ad3c2207b2bd170d752a0a26a9fa2aa19_icgraph">
<area shape="rect" title="Extract the contents of a genetic map file." alt="" coords="555,5,727,33"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="386,5,507,33"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,5,338,33"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,5,155,33"/>
</map>
</div>

</div>
</div>
<a id="ae9f10c91e8c7f77b536ef19e2222182c" name="ae9f10c91e8c7f77b536ef19e2222182c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae9f10c91e8c7f77b536ef19e2222182c">&#9670;&nbsp;</a></span>gsc_helper_random_cross_checks()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static unsigned int gsc_helper_random_cross_checks </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>from_group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>n_crosses</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>cap</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Check input parameters of random crossing functions. </p>
<p >(helper function)</p>
<dl class="section return"><dt>Returns</dt><dd>0 if the parameters fail these checks, number of members of <em>from_group</em> otherwise. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08095">8095</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-68" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-68-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-68-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-68-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_cgraph.png" border="0" usemap="#asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_cgraph" alt=""/></div>
<map name="asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_cgraph" id="asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_cgraph">
<area shape="rect" title="Check input parameters of random crossing functions." alt="" coords="5,5,179,49"/>
<area shape="rect" href="group__collgetters.html#gafd291a4e0e38f9cf8963c3cd972fe8a7" title="Function to count the number of genotypes that currently belong to the specified group." alt="" coords="227,13,364,41"/>
</map>
</div>
<div id="dynsection-69" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-69-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-69-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-69-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_icgraph.png" border="0" usemap="#asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_icgraph" alt=""/></div>
<map name="asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_icgraph" id="asim-operations_8c_ae9f10c91e8c7f77b536ef19e2222182c_icgraph">
<area shape="rect" title="Check input parameters of random crossing functions." alt="" coords="238,28,411,72"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="5,5,190,33"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="5,57,190,101"/>
</map>
</div>

</div>
</div>
<a id="ac1d97be6bdb21bbb271fc44616461fd8" name="ac1d97be6bdb21bbb271fc44616461fd8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac1d97be6bdb21bbb271fc44616461fd8">&#9670;&nbsp;</a></span>gsc_helper_sort_markerlist()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_helper_sort_markerlist </td>
          <td>(</td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> *&#160;</td>
          <td class="paramname"><em>markerlist</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Sort markerlist by chromosome name, and by position within each chromosome. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">n_markers</td><td>length of <em>markerlist</em> </td></tr>
    <tr><td class="paramname">markerlist</td><td>list of MapFileUnits to sort Could return number of unique chromosomes found in the markerlist, but we aren't worried about that level of performance right now. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05657">5657</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-70" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-70-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-70-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-70-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_cgraph.png" border="0" usemap="#asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_cgraph" alt=""/></div>
<map name="asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_cgraph" id="asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_cgraph">
<area shape="rect" title="Sort markerlist by chromosome name, and by position within each chromosome." alt="" coords="5,48,183,76"/>
<area shape="rect" href="sim-operations_8c.html#aaa406c38a85c3953b1cda622ff535d12" title="Comparator function for qsort." alt="" coords="231,5,404,49"/>
<area shape="rect" href="sim-operations_8c.html#a6899282e8507284be584200d05e786cf" title="Comparator function for qsort." alt="" coords="236,73,399,117"/>
</map>
</div>
<div id="dynsection-71" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-71-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-71-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-71-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_icgraph.png" border="0" usemap="#asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_icgraph" alt=""/></div>
<map name="asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_icgraph" id="asim-operations_8c_ac1d97be6bdb21bbb271fc44616461fd8_icgraph">
<area shape="rect" title="Sort markerlist by chromosome name, and by position within each chromosome." alt="" coords="555,5,733,33"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="386,5,507,33"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,5,338,33"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,5,155,33"/>
</map>
</div>

</div>
</div>
<a id="a9baab4183af973c0ef583a9ed3306c97" name="a9baab4183af973c0ef583a9ed3306c97"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9baab4183af973c0ef583a9ed3306c97">&#9670;&nbsp;</a></span>gsc_helper_split_by_allocator_equalprob()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_allocator_equalprob </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n_outgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>subgroupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>outgroups</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l03414">3414</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-72" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-72-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-72-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-72-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a9baab4183af973c0ef583a9ed3306c97_icgraph.png" border="0" usemap="#asim-operations_8c_a9baab4183af973c0ef583a9ed3306c97_icgraph" alt=""/></div>
<map name="asim-operations_8c_a9baab4183af973c0ef583a9ed3306c97_icgraph" id="asim-operations_8c_a9baab4183af973c0ef583a9ed3306c97_icgraph">
<area shape="rect" title=" " alt="" coords="187,5,327,49"/>
<area shape="rect" href="group__groupmod.html#gae44e90e4bb377a0a2d882d3ee0932621" title="Allocate each member of the group to one of n groups with equal probability." alt="" coords="5,5,139,49"/>
</map>
</div>

</div>
</div>
<a id="a040d495e43abdd2c52d3720f71d52588" name="a040d495e43abdd2c52d3720f71d52588"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a040d495e43abdd2c52d3720f71d52588">&#9670;&nbsp;</a></span>gsc_helper_split_by_allocator_knowncounts()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_allocator_knowncounts </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n_outgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>subgroupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>outgroups</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l03207">3207</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-73" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-73-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-73-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-73-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a040d495e43abdd2c52d3720f71d52588_icgraph.png" border="0" usemap="#asim-operations_8c_a040d495e43abdd2c52d3720f71d52588_icgraph" alt=""/></div>
<map name="asim-operations_8c_a040d495e43abdd2c52d3720f71d52588_icgraph" id="asim-operations_8c_a040d495e43abdd2c52d3720f71d52588_icgraph">
<area shape="rect" title=" " alt="" coords="212,24,372,68"/>
<area shape="rect" href="group__groupmod.html#ga68012ed8f95ac60ae4875dbf605f5f95" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="5,5,164,33"/>
<area shape="rect" href="group__groupmod.html#ga02433970bb02f719e2de61c44f573ff4" title="Split a group into n groups of equal size (or size differing only by one, if n does not perfectly div..." alt="" coords="7,57,162,85"/>
</map>
</div>

</div>
</div>
<a id="af7acb36091cc63f5b7c2c7cc36d5911c" name="af7acb36091cc63f5b7c2c7cc36d5911c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af7acb36091cc63f5b7c2c7cc36d5911c">&#9670;&nbsp;</a></span>gsc_helper_split_by_allocator_unequalprob()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_allocator_unequalprob </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n_outgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>subgroupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>outgroups</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l03479">3479</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-74" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-74-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-74-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-74-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_af7acb36091cc63f5b7c2c7cc36d5911c_icgraph.png" border="0" usemap="#asim-operations_8c_af7acb36091cc63f5b7c2c7cc36d5911c_icgraph" alt=""/></div>
<map name="asim-operations_8c_af7acb36091cc63f5b7c2c7cc36d5911c_icgraph" id="asim-operations_8c_af7acb36091cc63f5b7c2c7cc36d5911c_icgraph">
<area shape="rect" title=" " alt="" coords="224,5,378,49"/>
<area shape="rect" href="group__groupmod.html#ga0433ee84c0760569489e2f1ed31dc7c5" title="Allocate each member of the group to one of n groups with custom probabilities for each group." alt="" coords="5,13,176,41"/>
</map>
</div>

</div>
</div>
<a id="afb9ba063ab6784b4676c7fdafa7a53a4" name="afb9ba063ab6784b4676c7fdafa7a53a4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afb9ba063ab6784b4676c7fdafa7a53a4">&#9670;&nbsp;</a></span>gsc_helper_split_by_quality_family()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_quality_family </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>maxgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>groupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>results</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l02964">2964</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-75" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-75-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-75-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-75-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_cgraph.png" border="0" usemap="#asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_cgraph" alt=""/></div>
<map name="asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_cgraph" id="asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_cgraph">
<area shape="rect" title=" " alt="" coords="5,24,142,68"/>
<area shape="rect" href="group__liteget.html#ga847569827ccdd09ff54808c09aba742c" title="Get the first/left parent of a genotype." alt="" coords="200,5,339,33"/>
<area shape="rect" href="group__liteget.html#gab71ce9343a19dd2fc34da1af89fee00a" title="Get the second/right parent of a genotype." alt="" coords="190,57,349,85"/>
</map>
</div>
<div id="dynsection-76" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-76-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-76-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-76-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_icgraph.png" border="0" usemap="#asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_icgraph" alt=""/></div>
<map name="asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_icgraph" id="asim-operations_8c_afb9ba063ab6784b4676c7fdafa7a53a4_icgraph">
<area shape="rect" title=" " alt="" coords="206,5,342,49"/>
<area shape="rect" href="group__groupmod.html#gaf863cc5728f3a83f0002a9e43ada3a33" title="Split a group into families by their pedigrees." alt="" coords="5,13,158,41"/>
</map>
</div>

</div>
</div>
<a id="a7732d896b52fde5a51e8ee337d7310ec" name="a7732d896b52fde5a51e8ee337d7310ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7732d896b52fde5a51e8ee337d7310ec">&#9670;&nbsp;</a></span>gsc_helper_split_by_quality_halfsib1()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_quality_halfsib1 </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>maxgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>groupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>results</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l02883">2883</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-77" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-77-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-77-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-77-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_cgraph.png" border="0" usemap="#asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_cgraph" alt=""/></div>
<map name="asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_cgraph" id="asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_cgraph">
<area shape="rect" title=" " alt="" coords="5,28,142,72"/>
<area shape="rect" href="group__liteget.html#ga847569827ccdd09ff54808c09aba742c" title="Get the first/left parent of a genotype." alt="" coords="201,5,339,33"/>
<area shape="rect" href="sim-operations_8c.html#a986ca18e011ef3906d609b47c9a84c54" title=" " alt="" coords="190,57,350,101"/>
</map>
</div>
<div id="dynsection-78" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-78-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-78-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-78-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_icgraph.png" border="0" usemap="#asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_icgraph" alt=""/></div>
<map name="asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_icgraph" id="asim-operations_8c_a7732d896b52fde5a51e8ee337d7310ec_icgraph">
<area shape="rect" title=" " alt="" coords="199,5,335,49"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="5,5,151,49"/>
</map>
</div>

</div>
</div>
<a id="a0f8a1c3546a8c12067e790c2f2cc9782" name="a0f8a1c3546a8c12067e790c2f2cc9782"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0f8a1c3546a8c12067e790c2f2cc9782">&#9670;&nbsp;</a></span>gsc_helper_split_by_quality_halfsib2()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_quality_halfsib2 </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>maxgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>groupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>results</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l02892">2892</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-79" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-79-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-79-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-79-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_cgraph.png" border="0" usemap="#asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_cgraph" alt=""/></div>
<map name="asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_cgraph" id="asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_cgraph">
<area shape="rect" title=" " alt="" coords="5,28,142,72"/>
<area shape="rect" href="group__liteget.html#gab71ce9343a19dd2fc34da1af89fee00a" title="Get the second/right parent of a genotype." alt="" coords="190,5,350,33"/>
<area shape="rect" href="sim-operations_8c.html#a986ca18e011ef3906d609b47c9a84c54" title=" " alt="" coords="190,57,350,101"/>
</map>
</div>
<div id="dynsection-80" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-80-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-80-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-80-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_icgraph.png" border="0" usemap="#asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_icgraph" alt=""/></div>
<map name="asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_icgraph" id="asim-operations_8c_a0f8a1c3546a8c12067e790c2f2cc9782_icgraph">
<area shape="rect" title=" " alt="" coords="199,5,335,49"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="5,5,151,49"/>
</map>
</div>

</div>
</div>
<a id="a986ca18e011ef3906d609b47c9a84c54" name="a986ca18e011ef3906d609b47c9a84c54"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a986ca18e011ef3906d609b47c9a84c54">&#9670;&nbsp;</a></span>gsc_helper_split_by_quality_halfsibtemplate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_quality_halfsibtemplate </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>maxgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>groupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>results</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>(*)(<a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>)&#160;</td>
          <td class="paramname"><em>getparent</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l02859">2859</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-81" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-81-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-81-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-81-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a986ca18e011ef3906d609b47c9a84c54_icgraph.png" border="0" usemap="#asim-operations_8c_a986ca18e011ef3906d609b47c9a84c54_icgraph" alt=""/></div>
<map name="asim-operations_8c_a986ca18e011ef3906d609b47c9a84c54_icgraph" id="asim-operations_8c_a986ca18e011ef3906d609b47c9a84c54_icgraph">
<area shape="rect" title=" " alt="" coords="383,40,543,84"/>
<area shape="rect" href="sim-operations_8c.html#a7732d896b52fde5a51e8ee337d7310ec" title=" " alt="" coords="199,5,335,49"/>
<area shape="rect" href="sim-operations_8c.html#a0f8a1c3546a8c12067e790c2f2cc9782" title=" " alt="" coords="199,73,335,117"/>
<area shape="rect" href="group__groupmod.html#ga556be18433d7185e212112028e22a425" title="Split a group into families of half&#45;siblings by shared first or second parent." alt="" coords="5,40,151,84"/>
</map>
</div>

</div>
</div>
<a id="a69a260388571f8c32ad879dfdc8a3c8f" name="a69a260388571f8c32ad879dfdc8a3c8f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a69a260388571f8c32ad879dfdc8a3c8f">&#9670;&nbsp;</a></span>gsc_helper_split_by_quality_individuate()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_helper_split_by_quality_individuate </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>&#160;</td>
          <td class="paramname"><em>loc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>datastore</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>maxgroups</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>groupsfound</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> *&#160;</td>
          <td class="paramname"><em>results</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l03041">3041</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-82" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-82-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-82-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-82-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a69a260388571f8c32ad879dfdc8a3c8f_icgraph.png" border="0" usemap="#asim-operations_8c_a69a260388571f8c32ad879dfdc8a3c8f_icgraph" alt=""/></div>
<map name="asim-operations_8c_a69a260388571f8c32ad879dfdc8a3c8f_icgraph" id="asim-operations_8c_a69a260388571f8c32ad879dfdc8a3c8f_icgraph">
<area shape="rect" title=" " alt="" coords="221,5,357,49"/>
<area shape="rect" href="group__groupmod.html#ga532a3797b8529c797d39e2ad9069ecf1" title="Split a group into n one&#45;member groups." alt="" coords="5,13,173,41"/>
</map>
</div>

</div>
</div>
<a id="acf770138dc0f9f33bc715156c83ccfd0" name="acf770138dc0f9f33bc715156c83ccfd0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acf770138dc0f9f33bc715156c83ccfd0">&#9670;&nbsp;</a></span>gsc_helper_str_markerlist_leftjoin()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static unsigned int gsc_helper_str_markerlist_leftjoin </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___known_genome.html">gsc_KnownGenome</a>&#160;</td>
          <td class="paramname"><em>g</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_markers_in_list</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">struct <a class="el" href="structgsc___mapfile_unit.html">gsc_MapfileUnit</a> **&#160;</td>
          <td class="paramname"><em>markerlist</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Discard markers whose names are not present in a <a class="el" href="structgsc___known_genome.html" title="A type that stores the genome structure used in simulation.">gsc_KnownGenome</a>. </p>
<p >All markers in the genome <em>g</em> are retained in the list. All markers not in that genome are freed and deleted from the <em>markerlist</em>. The new length of the marker list after deletions is returned.</p>
<p >Markers without names to join on are discarded.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">g</td><td>genome object containing the list of markers to compare markers in <em>markerlist</em> against </td></tr>
    <tr><td class="paramname">n_markers_in_list</td><td>length of <em>markerlist</em> </td></tr>
    <tr><td class="paramname">markerlist</td><td>Pointer to a list of name/chromosome/position entries for various markers. Only the names will be read, but for each entry, if its name does not exist in the list of markers to keep, the entry will be deleted and the remainder of the marker list shuffled backwards. Markers without names are automatically discarded. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>new number of markers in markerlist, either <em>n_markers_in_list</em> or smaller. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05566">5566</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-83" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-83-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-83-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-83-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_cgraph.png" border="0" usemap="#asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_cgraph" alt=""/></div>
<map name="asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_cgraph" id="asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_cgraph">
<area shape="rect" title="Discard markers whose names are not present in a gsc_KnownGenome." alt="" coords="5,5,176,49"/>
<area shape="rect" href="group__supporters.html#gad57230a4c812ba8a6deed26100beae7b" title="Return whether or not a marker name is present in the tracked markers, and at what index." alt="" coords="224,5,395,49"/>
</map>
</div>
<div id="dynsection-84" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-84-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-84-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-84-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_icgraph.png" border="0" usemap="#asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_icgraph" alt=""/></div>
<map name="asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_icgraph" id="asim-operations_8c_acf770138dc0f9f33bc715156c83ccfd0_icgraph">
<area shape="rect" title="Discard markers whose names are not present in a gsc_KnownGenome." alt="" coords="555,5,726,49"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="386,13,507,41"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="a9022b0b2e12395904757c59db6010d5b" name="a9022b0b2e12395904757c59db6010d5b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9022b0b2e12395904757c59db6010d5b">&#9670;&nbsp;</a></span>gsc_helper_tablefilereader_get_next_cell_wqueue()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> gsc_helper_tablefilereader_get_next_cell_wqueue </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___table_file_reader.html">gsc_TableFileReader</a> *&#160;</td>
          <td class="paramname"><em>tf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___table_file_cell.html">gsc_TableFileCell</a> **&#160;</td>
          <td class="paramname"><em>queue</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>queuesize</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cells from a TableFileReader. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">queue</td><td>pointer to queue containing <em>queuesize</em> TableFileCells. The pointer will be updated to point one cell forwards (so that it still points to the 'start' of the queue) if the item returned comes from the queue. </td></tr>
    <tr><td class="paramname">queuesize</td><td>number of entries remaining in the queue. Will be updated to be one fewer if the item returned comes from the queue </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l05296">5296</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-85" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-85-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-85-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-85-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a9022b0b2e12395904757c59db6010d5b_cgraph.png" border="0" usemap="#asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_cgraph" alt=""/></div>
<map name="asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_cgraph" id="asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_cgraph">
<area shape="rect" title="Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cell..." alt="" coords="5,73,177,117"/>
<area shape="rect" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05" title="Read forwards in TableFileReader and return the next cell&#39;s contents, as well as how many column gaps..." alt="" coords="225,73,354,117"/>
<area shape="rect" href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0" title="Classify the character under the cursor of a TableFileReader as cell contents or otherwise." alt="" coords="402,5,573,49"/>
<area shape="rect" href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322" title="Read another buffer&#39;s worth of characters from a gsc_TableFileReader&#39;s file." alt="" coords="402,73,573,117"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="427,141,548,169"/>
</map>
</div>
<div id="dynsection-86" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-86-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-86-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-86-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a9022b0b2e12395904757c59db6010d5b_icgraph.png" border="0" usemap="#asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_icgraph" alt=""/></div>
<map name="asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_icgraph" id="asim-operations_8c_a9022b0b2e12395904757c59db6010d5b_icgraph">
<area shape="rect" title="Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cell..." alt="" coords="803,49,974,93"/>
<area shape="rect" href="sim-operations_8c.html#ad3c2207b2bd170d752a0a26a9fa2aa19" title="Extract the contents of a genetic map file." alt="" coords="583,5,755,33"/>
<area shape="rect" href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831" title="Populates a gsc_SimData combination with effect values." alt="" coords="605,57,733,85"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,109,535,153"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="400,5,521,33"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,57,338,85"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,57,155,85"/>
</map>
</div>

</div>
</div>
<a id="aca78dbac4d7154e95b6ddca575937d77" name="aca78dbac4d7154e95b6ddca575937d77"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aca78dbac4d7154e95b6ddca575937d77">&#9670;&nbsp;</a></span>gsc_load_genotypefile_matrix()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_load_genotypefile_matrix </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___file_format_spec.html">gsc_FileFormatSpec</a>&#160;</td>
          <td class="paramname"><em>format</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Loads a genotype file, with or without existing genome model in the SimData. </p>
<p >First, it reads the first row to determine if it is a header or not. This, and early information from the second row, can be used to determine:</p><ul>
<li>the number of columns in the file</li>
<li>whether markers are stored as rows and parent lines as columns, or vice-versa</li>
<li>how the pair of alleles for each SNP is encoded in the body of the matrix <dl class="section see"><dt>See also</dt><dd><a class="el" href="group__structs.html#ga169da3c2b32a6f9afd92e2e0801ad741" title="Represent possible representations of alleles at a marker in a genotype file.">gsc_GenotypeFileCellStyle</a></dd></dl>
Rows with more entries than the first will have their excess entries ignored. Rows with fewer entries than the first will also provide a warning that there is an incorrect number of cells in that row, and will simply have default null '\0' alleles in that marker/parent line.</li>
</ul>
<dl class="section return"><dt>Returns</dt><dd>group number for the set of genotypes loaded from the file </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l06762">6762</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-87" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-87-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-87-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-87-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_cgraph.png" border="0" usemap="#asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_cgraph" alt=""/></div>
<map name="asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_cgraph" id="asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_cgraph">
<area shape="rect" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="5,618,155,662"/>
<area shape="rect" href="sim-operations_8c.html#a14bf1934790250afa4e708fe902c44a0" title="Create a new gsc_EmptyListNavigator, including an empty AlleleMatrix suitable for inserting into the ..." alt="" coords="203,97,399,125"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="727,230,849,258"/>
<area shape="rect" href="group__loaders.html#ga5cbe0b821180b257b21d9551ec797f20" title="Create a uniformly&#45;spaced gsc_RecombinationMap from a list of marker names and save to SimData." alt="" coords="213,319,389,363"/>
<area shape="rect" href="group__supporters.html#gad57230a4c812ba8a6deed26100beae7b" title="Return whether or not a marker name is present in the tracked markers, and at what index." alt="" coords="462,509,633,553"/>
<area shape="rect" href="group__deletors.html#ga9c647cacb393998930c560962c1f7f6d" title="Delete the gsc_AlleleMatrix linked list from m onwards and frees its memory." alt="" coords="706,558,870,586"/>
<area shape="rect" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad" title="Push emptylist edited genotypes into the SimData." alt="" coords="224,678,378,722"/>
<area shape="rect" href="sim-operations_8c.html#a52e6f87fe92eca15a0f4da957c5f8e19" title="Reset the cursor of a gsc_EmptyListNavigator to the first genotype." alt="" coords="471,253,624,297"/>
<area shape="rect" href="sim-operations_8c.html#a67c2447b0d771997a444efe026cf0600" title="Get the next sequential genotype in an gsc_EmptyListNavigator." alt="" coords="224,149,378,193"/>
<area shape="rect" href="group__supporters.html#ga22609353c4951dcf3ed54537ecb6eeda" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="217,746,385,774"/>
<area shape="rect" href="sim-operations_8c.html#a7c6136857e4f65bff718565b48e18afa" title="Parse a string and save it as the alleles of a genotype at a particular location and genetic marker." alt="" coords="219,798,383,842"/>
<area shape="rect" href="sim-operations_8c.html#ac22c96ec7ddd91b676805d17de44689c" title="Determine the style in which alleles are stored in a genotype matrix." alt="" coords="220,934,382,978"/>
<area shape="rect" href="sim-operations_8c.html#ae903fb523337566ac34b578f02435c2c" title="Determine whether a genotype matrix has a corner cell or not." alt="" coords="216,1002,386,1062"/>
<area shape="rect" href="sim-operations_8c.html#a410d2515f415529d8efe2abd9ab62048" title="Determine whether a genotype matrix has a header row or not." alt="" coords="221,866,381,910"/>
<area shape="rect" href="sim-operations_8c.html#a39561a6563fa3b016f8833252acc5e5a" title="Determine whether a genotype matrix is row&#45; or column&#45;oriented." alt="" coords="214,507,388,551"/>
<area shape="rect" href="sim-operations_8c.html#acf2e8ece7fb3470e814da69e8f162813" title="Comparator function for qsort." alt="" coords="211,1086,391,1130"/>
<area shape="rect" href="sim-operations_8c.html#a9022b0b2e12395904757c59db6010d5b" title="Return the next cell from a queue of cells until the queue is exhausted, and thereafter read new cell..." alt="" coords="215,439,387,483"/>
<area shape="rect" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05" title="Read forwards in TableFileReader and return the next cell&#39;s contents, as well as how many column gaps..." alt="" coords="483,441,612,485"/>
<area shape="rect" href="group__liteget.html#gadff19e04cab018dd64ae232e0adcb3cc" title="Set the name of a genotype." alt="" coords="248,1154,354,1182"/>
<area shape="rect" href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d" title="Allocate memory to store a deep copy of a gsc_TableFileCell, if previously only a shallow copy." alt="" coords="456,201,639,229"/>
<area shape="rect" href="group__structs.html#ga85b05dcdc5ec484d918ed2d3330c3ae8" title="Close a gsc_TableFileReader&#39;s file pointer." alt="" coords="236,1206,366,1250"/>
<area shape="rect" href="group__structs.html#ga1a3a39309d4ce6447581e73f7483fbfb" title="Open a file for reading with gsc_TableFileReader." alt="" coords="236,1274,366,1318"/>
<area shape="rect" href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76" title="Creator for an empty gsc_AlleleMatrix object of a given size." alt="" coords="447,97,648,125"/>
<area shape="rect" href="sim-operations_8c.html#a337459fb31ef07b88a77068780ab746e" title="Save a RecombinationMap to the SimData and allocate it a mapID." alt="" coords="449,373,647,417"/>
<area shape="rect" href="group__supporters.html#ga48ab10327d06674174428ca2db0b32ed" title="Function to identify the next sequential integer that is not already allocated to a map ID in the sim..." alt="" coords="715,370,861,398"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="477,687,619,731"/>
<area shape="rect" href="group__liteget.html#ga5e1e89782604c9ab47f196e1da233362" title="Get the alleles of a genotype." alt="" coords="732,610,844,638"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="735,662,841,690"/>
<area shape="rect" href="group__iterators.html#ga16615a5663b4d3bd768d8380f820aa66" title="Get an gsc_AlleleMatrix by index in the linked list." alt="" coords="704,714,872,742"/>
<area shape="rect" href="group__deletors.html#gae46ff71b819d28408e8b55dd4281a4f3" title="Move all details of the genotype at one gsc_GenoLocation to another gsc_GenoLocation." alt="" coords="718,766,858,794"/>
<area shape="rect" href="sim-operations_8c.html#adf656d2e43014ca8dd4f5d39adeb68f6" title="Sets the current cursor position in a gsc_GappyIterator to the next empty position,..." alt="" coords="707,818,869,846"/>
<area shape="rect" href="sim-operations_8c.html#af5486daced98d9bab2126b9d7af11356" title="Sets the current cursor position in a gsc_GappyIterator to the next filled position,..." alt="" coords="696,870,880,898"/>
<area shape="rect" href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708" title="Sets the current cursor position in a gsc_GappyIterator to the next valid position,..." alt="" coords="936,822,1106,850"/>
<area shape="rect" href="group__collgetters.html#gaf76f5cdc38c881c9ee8eee32a1770e83" title="Identify group numbers that currently have members." alt="" coords="465,789,630,817"/>
<area shape="rect" href="group__collgetters.html#gaf344362b92201c509c3efd9fba21f6a9" title="Identify group numbers that currently have members, and how many members they have." alt="" coords="709,922,867,966"/>
<area shape="rect" href="group__iterators.html#ga05da8d96593f8404938cc24f659db049" title="Create a bidirectional iterator." alt="" coords="928,878,1114,906"/>
<area shape="rect" href="group__liteget.html#gae5d12aca851dd2358b63dc60b8ac6e5f" title="Get the current group membership of a genotype." alt="" coords="968,930,1074,958"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="956,982,1086,1010"/>
<area shape="rect" href="group__iterators.html#ga5cca400276ca83acceaab36d5f79aab1" title=" " alt="" coords="1162,870,1323,914"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="1171,974,1314,1018"/>
<area shape="rect" href="sim-operations_8c.html#aaaec31480065a89defa3e02a0f902cb8" title="Identify what formatting a genotype matrix is representing alleles as." alt="" coords="461,899,635,943"/>
<area shape="rect" href="group__structs.html#ga9861b51e8a5c74f843aac748c932c3a0" title="Classify the character under the cursor of a TableFileReader as cell contents or otherwise." alt="" coords="702,422,874,466"/>
<area shape="rect" href="group__structs.html#gac2206b9a58abb4b533d11dec4bd9f322" title="Read another buffer&#39;s worth of characters from a gsc_TableFileReader&#39;s file." alt="" coords="702,490,874,534"/>
</map>
</div>
<div id="dynsection-88" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-88-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-88-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-88-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_icgraph.png" border="0" usemap="#asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_icgraph" alt=""/></div>
<map name="asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_icgraph" id="asim-operations_8c_aca78dbac4d7154e95b6ddca575937d77_icgraph">
<area shape="rect" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="386,5,535,49"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="203,13,338,41"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,13,155,41"/>
</map>
</div>

</div>
</div>
<a id="ab19b3eb80b5fe460d17c93e0c27f7aaf" name="ab19b3eb80b5fe460d17c93e0c27f7aaf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab19b3eb80b5fe460d17c93e0c27f7aaf">&#9670;&nbsp;</a></span>gsc_make_n_crosses_from_top_m_percent()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structgsc___group_num.html">gsc_GroupNum</a> gsc_make_n_crosses_from_top_m_percent </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___sim_data.html">gsc_SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___group_num.html">gsc_GroupNum</a>&#160;</td>
          <td class="paramname"><em>group</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___map_i_d.html">gsc_MapID</a>&#160;</td>
          <td class="paramname"><em>mapID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___effect_i_d.html">gsc_EffectID</a>&#160;</td>
          <td class="paramname"><em>effID</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structgsc___gen_options.html">gsc_GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l08869">8869</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>

</div>
</div>
<a id="aaaf3ebc75feaf901829232712e55677b" name="aaaf3ebc75feaf901829232712e55677b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aaaf3ebc75feaf901829232712e55677b">&#9670;&nbsp;</a></span>gsc_malloc_wrap()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void * gsc_malloc_wrap </td>
          <td>(</td>
          <td class="paramtype">const size_t&#160;</td>
          <td class="paramname"><em>size</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char&#160;</td>
          <td class="paramname"><em>exitonfail</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Replace calls to malloc direct with this function. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">size</td><td>size of space to be allocated. Treat like the parameter of a regular malloc call. </td></tr>
    <tr><td class="paramname">exitonfail</td><td>if truthy, if memory allocation fails it will terminate execution. if falsy, on memory allocation failure it will only print an error message and return NULL (useful, for example, if the calling function wants to add its own error message, or return partial results even if it fails here). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>pointer to the allocated space. </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l00034">34</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-89" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-89-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-89-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-89-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_aaaf3ebc75feaf901829232712e55677b_icgraph.png" border="0" usemap="#asim-operations_8c_aaaf3ebc75feaf901829232712e55677b_icgraph" alt=""/></div>
<map name="asim-operations_8c_aaaf3ebc75feaf901829232712e55677b_icgraph" id="asim-operations_8c_aaaf3ebc75feaf901829232712e55677b_icgraph">
<area shape="rect" title="Replace calls to malloc direct with this function." alt="" coords="968,935,1089,963"/>
<area shape="rect" href="group__recomb.html#gaecfe54dacdba616a25212ff12a32398c" title="Identify markers in the genotype of offspring where recombination from its parents occured." alt="" coords="669,5,920,33"/>
<area shape="rect" href="group__recomb.html#gaf955aa7d648a243e559ac23040e35fc5" title="Identify markers in the genotype of offspring where recombination from its parents occured,..." alt="" coords="669,57,920,85"/>
<area shape="rect" href="group__calculators.html#gae63afe0edeaf9eab31a0328fc72fc600" title="Takes a look at the currently&#45;loaded effect values and creates a string containing the allele with th..." alt="" coords="719,109,870,153"/>
<area shape="rect" href="group__calculators.html#ga145430273fd296e49528a27489f0e520" title="Calculates the highest&#45;breeding&#45;value haplotype that can be created from the alleles present in a giv..." alt="" coords="719,177,870,221"/>
<area shape="rect" href="group__supporters.html#gac926561083ce538188ff2cccc9580d42" title="Copy a vector of strings into the genotype name field." alt="" coords="717,245,872,289"/>
<area shape="rect" href="group__loaders.html#ga104fcea7a8f69cb80c136ab45c4eab76" title="Creator for an empty gsc_AlleleMatrix object of a given size." alt="" coords="694,636,895,664"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="208,567,358,611"/>
<area shape="rect" href="sim-operations_8c.html#a67c2447b0d771997a444efe026cf0600" title="Get the next sequential genotype in an gsc_EmptyListNavigator." alt="" coords="447,567,600,611"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="191,1383,375,1411"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="191,1088,375,1132"/>
<area shape="rect" href="group__loaders.html#gab37984b371a35d44dbce9d954bbf2366" title="Creator for an empty gsc_SimData object on the heap." alt="" coords="704,1339,885,1367"/>
<area shape="rect" href="group__calculators.html#ga44e3a2d5e4041091deec1954b4f62a86" title="Divide the genotype into blocks where each block contains all markers within a 1/n length section of ..." alt="" coords="718,1391,871,1435"/>
<area shape="rect" href="group__supporters.html#gab82b3920fa4760dad88332ee430e7cf2" title="Initialises a new custom label." alt="" coords="720,1459,869,1487"/>
<area shape="rect" href="group__iterators.html#gaa26377edd1339407af2070924d3596de" title="Create a Random Access Iterator." alt="" coords="693,1131,896,1159"/>
<area shape="rect" href="group__loaders.html#gada5c2aaacd8b8a145493a0db5c017068" title="Parse a list of markers/chrs/positions into a gsc_RecombinationMap and save to SimData." alt="" coords="443,823,603,867"/>
<area shape="rect" href="group__loaders.html#gac180eb5f1d27907a69314e68cc0d37af" title="Load a genetic map to a gsc_SimData object." alt="" coords="222,831,344,859"/>
<area shape="rect" href="group__loaders.html#ga5cbe0b821180b257b21d9551ec797f20" title="Create a uniformly&#45;spaced gsc_RecombinationMap from a list of marker names and save to SimData." alt="" coords="435,755,611,799"/>
<area shape="rect" href="group__deletors.html#ga27c2247d05e610fc83a83a5bae3d8d94" title="Deletes a particular set of marker effects from memory." alt="" coords="730,1511,859,1539"/>
<area shape="rect" href="group__deletors.html#ga060559de8b20b9a1dbbd73297a6c4079" title="Clears memory of this label from the simulation and all its genotypes." alt="" coords="735,1563,854,1591"/>
<area shape="rect" href="group__deletors.html#ga6075a65332dee0dc0032f29cad35fd74" title="Deletes a particular recombination map from memory." alt="" coords="692,1615,897,1643"/>
<area shape="rect" href="sim-operations_8c.html#a52e6f87fe92eca15a0f4da957c5f8e19" title="Reset the cursor of a gsc_EmptyListNavigator to the first genotype." alt="" coords="447,687,600,731"/>
<area shape="rect" href="group__maths.html#ga9c1e4641b679df1d6ec3d885ef6d1fd2" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="704,1667,885,1695"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="710,1719,879,1747"/>
<area shape="rect" href="sim-operations_8c.html#a3717c5ed02cf2a321b7424dc0e668e19" title="Save an EffectMatrix to the SimData and allocate it an EffectID." alt="" coords="723,313,866,357"/>
<area shape="rect" href="group__loaders.html#ga176416bff8f5e258dbe1619637c8b831" title="Populates a gsc_SimData combination with effect values." alt="" coords="460,343,587,371"/>
<area shape="rect" href="sim-operations_8c.html#a337459fb31ef07b88a77068780ab746e" title="Save a RecombinationMap to the SimData and allocate it a mapID." alt="" coords="695,791,894,835"/>
<area shape="rect" href="sim-operations_8c.html#ab0152d52dcce6514ebba0fa08810ee8a" title="offspringGenerator function parameter for gsc_make_clones." alt="" coords="434,943,612,987"/>
<area shape="rect" href="group__calculators.html#gaf6f32f41ae7ca27c1a48f8947919794a" title="Given a file containing definitions of blocks of markers, process that file and return a struct conta..." alt="" coords="735,1771,854,1799"/>
<area shape="rect" href="group__iterators.html#gaaccc34724148be204eb4d8515000dfd5" title="Get a location by index using a gsc_RandomAccessIterator." alt="" coords="732,1235,857,1263"/>
<area shape="rect" href="sim-operations_8c.html#a3bbb853ca13c593a940d86882dda2361" title="Prints a matrix of genotype information to a file." alt="" coords="688,1823,901,1851"/>
<area shape="rect" href="sim-operations_8c.html#a368b4ddb966aef02950fe327d46f051b" title="Fills the designated section of the .names array in an gsc_AlleleMatrix with the pattern &quot;`prefix`ind..." alt="" coords="738,1875,851,1903"/>
<area shape="rect" href="group__supporters.html#ga4929f1ad6756e3c710ad0781baca0f42" title="Produce a random ordering of the first n elements in an array using a (partial) Fisher&#45;Yates shuffle." alt="" coords="732,1183,857,1211"/>
<area shape="rect" href="group__structs.html#ga85446dd929940e823745067d0a1d1d5d" title="Allocate memory to store a deep copy of a gsc_TableFileCell, if previously only a shallow copy." alt="" coords="432,395,615,423"/>
<area shape="rect" href="group__structs.html#ga9b389be1c0eada385d1015e445886c05" title="Read forwards in TableFileReader and return the next cell&#39;s contents, as well as how many column gaps..." alt="" coords="459,447,588,491"/>
<area shape="rect" href="group__recomb.html#ga17bf2aa42201490431e56ba42ace85fc" title="Provides guesses as to the location of recombination events that led to the creation of certain genot..." alt="" coords="425,24,621,68"/>
<area shape="rect" href="sim-operations_8c.html#a14bf1934790250afa4e708fe902c44a0" title="Create a new gsc_EmptyListNavigator, including an empty AlleleMatrix suitable for inserting into the ..." alt="" coords="425,635,621,663"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="442,1063,604,1107"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="5,575,141,603"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="220,993,346,1021"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="189,1181,377,1209"/>
<area shape="rect" href="group__groupmod.html#gad7acd8960fc510d4aebb587ad35878f7" title="Take a list of indexes and allocate the genotypes at those indexes to a new group." alt="" coords="447,1284,599,1312"/>
<area shape="rect" href="group__groupmod.html#ga47292b150273a395fe282f86c0f78e0b" title="Split a group into two groups of equal size (or size differing only by one, if the original group had..." alt="" coords="438,1336,609,1364"/>
<area shape="rect" href="group__calculators.html#ga2eaf7918fe1253fdf95b9ab81a192c96" title="Calculates the number of times at each marker that a particular allele appears." alt="" coords="454,1644,592,1688"/>
<area shape="rect" href="group__calculators.html#ga30689b0f03641259091a72d065951e29" title="Calculate the fitness metric/breeding value for every genotype in the simulation or every genotype in..." alt="" coords="219,1737,347,1765"/>
<area shape="rect" href="group__calculators.html#ga5eb332e54b2f90f399a28a65cdb25e36" title="Calculate the fitness metric/breeding value for a set of genotypes." alt="" coords="441,1764,606,1792"/>
<area shape="rect" href="sim-operations_8c.html#a0fcef09a211d57fadb1dfa0b6d102901" title="parentChooser function parameter for gsc_make_random_crosses." alt="" coords="437,1524,610,1568"/>
<area shape="rect" href="sim-operations_8c.html#a2cee0d2e90de12744fd077b20a7db9bd" title="parentChooser function parameter for gsc_make_random_crosses_between." alt="" coords="436,1388,611,1432"/>
<area shape="rect" href="sim-operations_8c.html#a2c3d189234e1ac86e27a1ccaace3c363" title="parentChooser function parameter for gsc_make_targeted_crosses." alt="" coords="437,1456,610,1500"/>
<area shape="rect" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file." alt="" coords="448,1816,598,1860"/>
</map>
</div>

</div>
</div>
<a id="adf656d2e43014ca8dd4f5d39adeb68f6" name="adf656d2e43014ca8dd4f5d39adeb68f6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adf656d2e43014ca8dd4f5d39adeb68f6">&#9670;&nbsp;</a></span>gsc_nextgappy_get_gap()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_nextgappy_get_gap </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next empty position, if the cursor is not already an empty position. </p>
<p >Empty positions do not contain genotypes. This function does not change the cursor position if the curstor is already on a gap. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01291">1291</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-90" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-90-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-90-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-90-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_cgraph.png" border="0" usemap="#asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_cgraph" alt=""/></div>
<map name="asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_cgraph" id="asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_cgraph">
<area shape="rect" title="Sets the current cursor position in a gsc_GappyIterator to the next empty position,..." alt="" coords="5,5,168,33"/>
<area shape="rect" href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708" title="Sets the current cursor position in a gsc_GappyIterator to the next valid position,..." alt="" coords="216,5,385,33"/>
</map>
</div>
<div id="dynsection-91" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-91-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-91-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-91-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_icgraph.png" border="0" usemap="#asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_icgraph" alt=""/></div>
<map name="asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_icgraph" id="asim-operations_8c_adf656d2e43014ca8dd4f5d39adeb68f6_icgraph">
<area shape="rect" title="Sets the current cursor position in a gsc_GappyIterator to the next empty position,..." alt="" coords="1077,65,1239,93"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="887,57,1029,101"/>
<area shape="rect" href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5" title="Deletes all genotypes belonging to a particular group." alt="" coords="696,5,819,33"/>
<area shape="rect" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad" title="Push emptylist edited genotypes into the SimData." alt="" coords="681,57,834,101"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="676,173,839,217"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="460,9,609,53"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="230,17,365,45"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,17,155,45"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="471,77,597,105"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="440,129,628,157"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="442,181,626,209"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="442,233,626,277"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="440,301,628,329"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="473,353,595,381"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="205,233,390,277"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="203,301,392,329"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="208,353,387,397"/>
</map>
</div>

</div>
</div>
<a id="af5486daced98d9bab2126b9d7af11356" name="af5486daced98d9bab2126b9d7af11356"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5486daced98d9bab2126b9d7af11356">&#9670;&nbsp;</a></span>gsc_nextgappy_get_nongap()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_nextgappy_get_nongap </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next filled position, if the cursor is not already a filled position. </p>
<p >Non-gap positions do contain genotypes. This function does not change the cursor position if the curstor is already on a non-gap. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01320">1320</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-92" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-92-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-92-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-92-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_af5486daced98d9bab2126b9d7af11356_cgraph.png" border="0" usemap="#asim-operations_8c_af5486daced98d9bab2126b9d7af11356_cgraph" alt=""/></div>
<map name="asim-operations_8c_af5486daced98d9bab2126b9d7af11356_cgraph" id="asim-operations_8c_af5486daced98d9bab2126b9d7af11356_cgraph">
<area shape="rect" title="Sets the current cursor position in a gsc_GappyIterator to the next filled position,..." alt="" coords="5,5,189,33"/>
<area shape="rect" href="sim-operations_8c.html#a6d94248d1163c9c40e2910e578ce4708" title="Sets the current cursor position in a gsc_GappyIterator to the next valid position,..." alt="" coords="237,5,406,33"/>
</map>
</div>
<div id="dynsection-93" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-93-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-93-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-93-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_af5486daced98d9bab2126b9d7af11356_icgraph.png" border="0" usemap="#asim-operations_8c_af5486daced98d9bab2126b9d7af11356_icgraph" alt=""/></div>
<map name="asim-operations_8c_af5486daced98d9bab2126b9d7af11356_icgraph" id="asim-operations_8c_af5486daced98d9bab2126b9d7af11356_icgraph">
<area shape="rect" title="Sets the current cursor position in a gsc_GappyIterator to the next filled position,..." alt="" coords="1077,65,1260,93"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="887,57,1029,101"/>
<area shape="rect" href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5" title="Deletes all genotypes belonging to a particular group." alt="" coords="696,5,819,33"/>
<area shape="rect" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad" title="Push emptylist edited genotypes into the SimData." alt="" coords="681,57,834,101"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="676,173,839,217"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="460,9,609,53"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="230,17,365,45"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,17,155,45"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="471,77,597,105"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="440,129,628,157"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="442,181,626,209"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="442,233,626,277"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="440,301,628,329"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="473,353,595,381"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="205,233,390,277"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="203,301,392,329"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="208,353,387,397"/>
</map>
</div>

</div>
</div>
<a id="a6d94248d1163c9c40e2910e578ce4708" name="a6d94248d1163c9c40e2910e578ce4708"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6d94248d1163c9c40e2910e578ce4708">&#9670;&nbsp;</a></span>gsc_nextgappy_valid_pos()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a> gsc_nextgappy_valid_pos </td>
          <td>(</td>
          <td class="paramtype">struct <a class="el" href="structgsc___gappy_iterator.html">gsc_GappyIterator</a> *&#160;</td>
          <td class="paramname"><em>it</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Sets the current cursor position in a <a class="el" href="structgsc___gappy_iterator.html" title="A structure to iterate forwards through all positions in the gsc_AlleleMatrix linked list in gsc_SimD...">gsc_GappyIterator</a> to the next valid position, if the cursor is not already a valid position. </p>
<p >Valid positions in the linked list of AlleleMatrices may contain genotypes or not. </p>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l01271">1271</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-94" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-94-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-94-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-94-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a6d94248d1163c9c40e2910e578ce4708_icgraph.png" border="0" usemap="#asim-operations_8c_a6d94248d1163c9c40e2910e578ce4708_icgraph" alt=""/></div>
<map name="asim-operations_8c_a6d94248d1163c9c40e2910e578ce4708_icgraph" id="asim-operations_8c_a6d94248d1163c9c40e2910e578ce4708_icgraph">
<area shape="rect" title="Sets the current cursor position in a gsc_GappyIterator to the next valid position,..." alt="" coords="1308,65,1478,93"/>
<area shape="rect" href="sim-operations_8c.html#adf656d2e43014ca8dd4f5d39adeb68f6" title="Sets the current cursor position in a gsc_GappyIterator to the next empty position,..." alt="" coords="1088,39,1250,67"/>
<area shape="rect" href="sim-operations_8c.html#af5486daced98d9bab2126b9d7af11356" title="Sets the current cursor position in a gsc_GappyIterator to the next filled position,..." alt="" coords="1077,91,1260,119"/>
<area shape="rect" href="group__supporters.html#gafa617e95d5ed52b4e25434a20006809c" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the g..." alt="" coords="887,57,1029,101"/>
<area shape="rect" href="group__deletors.html#gaddc395516c92fe17e017a3a48e5fc3e5" title="Deletes all genotypes belonging to a particular group." alt="" coords="696,5,819,33"/>
<area shape="rect" href="sim-operations_8c.html#a622965bf813b7d185c0effe64aca6bad" title="Push emptylist edited genotypes into the SimData." alt="" coords="681,57,834,101"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="676,173,839,217"/>
<area shape="rect" href="sim-operations_8c.html#aca78dbac4d7154e95b6ddca575937d77" title="Loads a genotype file, with or without existing genome model in the SimData." alt="" coords="460,9,609,53"/>
<area shape="rect" href="group__loaders.html#ga394371951741d8c2db18a85069d6b61e" title="Populates a gsc_SimData object with marker allele data, a genetic map, and (optionally) marker effect..." alt="" coords="230,17,365,45"/>
<area shape="rect" href="group__loaders.html#ga50888a3ffcd2853e07bb4c1b6f0d067a" title="Load a set of genotypes to a gsc_SimData object." alt="" coords="5,17,155,45"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="471,77,597,105"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="440,129,628,157"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="442,181,626,209"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="442,233,626,277"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="440,301,628,329"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="473,353,595,381"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="205,233,390,277"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="203,301,392,329"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="208,353,387,397"/>
</map>
</div>

</div>
</div>
<a id="a8102058ad778cb4a23bd645d6786244d" name="a8102058ad778cb4a23bd645d6786244d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8102058ad778cb4a23bd645d6786244d">&#9670;&nbsp;</a></span>gsc_scaffold_save_ancestry_of()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_scaffold_save_ancestry_of </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>p1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___pedigree_i_d.html">gsc_PedigreeID</a>&#160;</td>
          <td class="paramname"><em>p2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void(*)(char *, size_t, void *)&#160;</td>
          <td class="paramname"><em>strprinter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void(*)(long unsigned int, void *)&#160;</td>
          <td class="paramname"><em>intprinter</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>printer_data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Identifies and saves (recursively) the pedigree of a pair of parents. </p>
<p >This is a generic scaffold for the following functions: </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file.">gsc_save_utility_pedigrees</a> with full_pedigree = true and in future, for a function to save these pedigrees to in-memory strings </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10748">10748</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-95" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-95-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-95-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-95-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a8102058ad778cb4a23bd645d6786244d_cgraph.png" border="0" usemap="#asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_cgraph" alt=""/></div>
<map name="asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_cgraph" id="asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_cgraph">
<area shape="rect" title="Identifies and saves (recursively) the pedigree of a pair of parents." alt="" coords="5,32,206,60"/>
<area shape="rect" href="group__search.html#ga5c7cd898a40e515c3164c98821068978" title="Returns the name of the genotype with a given id." alt="" coords="259,5,400,33"/>
<area shape="rect" href="group__search.html#gad9d1d0b5575da681af7789354447612a" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="254,57,405,85"/>
<area shape="rect" href="group__supporters.html#ga094a7c1d7f83050cce4d00fdcc89d6b3" title="Binary search through list of unsigned integers." alt="" coords="453,24,602,68"/>
</map>
</div>
<div id="dynsection-96" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-96-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-96-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-96-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a8102058ad778cb4a23bd645d6786244d_icgraph.png" border="0" usemap="#asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_icgraph" alt=""/></div>
<map name="asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_icgraph" id="asim-operations_8c_a8102058ad778cb4a23bd645d6786244d_icgraph">
<area shape="rect" title="Identifies and saves (recursively) the pedigree of a pair of parents." alt="" coords="1123,171,1323,199"/>
<area shape="rect" href="group__savers.html#ga23f6de68db1db7fd293a1e3207eeef6c" title="Prints pedigrees to a file." alt="" coords="899,171,1075,199"/>
<area shape="rect" href="sim-operations_8c.html#aa66b6b0ec2f03012e39df20c6f35b532" title="save&#45;as&#45;you&#45;go (pedigrees)" alt="" coords="696,132,851,176"/>
<area shape="rect" href="group__savers.html#gadce5deef59fa66f9208c2ed4d53a4768" title="Prints pedigrees of genotypes in the simulation to a file." alt="" coords="704,200,843,228"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="483,132,645,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
<area shape="rect" href="group__collgetters.html#ga9e31fa104eccb5ed54832b21b7c5d73e" title="Gets the full pedigree string (as per gsc_save_group_full_pedigree() ) of each member of the group." alt="" coords="479,200,648,228"/>
</map>
</div>

</div>
</div>
<a id="a3bbb853ca13c593a940d86882dda2361" name="a3bbb853ca13c593a940d86882dda2361"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3bbb853ca13c593a940d86882dda2361">&#9670;&nbsp;</a></span>gsc_scaffold_save_genotype_info()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_scaffold_save_genotype_info </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>f</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structgsc___bidirectional_iterator.html">gsc_BidirectionalIterator</a> *&#160;</td>
          <td class="paramname"><em>targets</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>n_markers</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **const&#160;</td>
          <td class="paramname"><em>marker_names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const _Bool&#160;</td>
          <td class="paramname"><em>markers_as_rows</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void(*)(FILE *, <a class="el" href="structgsc___geno_location.html">gsc_GenoLocation</a>, unsigned int, void *)&#160;</td>
          <td class="paramname"><em>bodycell_printer</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>bodycell_printer_data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Prints a matrix of genotype information to a file. </p>
<p >This is a generic scaffold for the following functions: </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file.">gsc_save_utility_genotypes</a> </dd>
<dd>
<a class="el" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file.">gsc_save_utility_allele_counts</a> </dd></dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l10482">10482</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-97" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-97-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-97-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-97-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a3bbb853ca13c593a940d86882dda2361_cgraph.png" border="0" usemap="#asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_cgraph" alt=""/></div>
<map name="asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_cgraph" id="asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_cgraph">
<area shape="rect" title="Prints a matrix of genotype information to a file." alt="" coords="5,109,219,137"/>
<area shape="rect" href="group__liteget.html#ga0ce82f67cb5e35a1c73bef54138202f6" title="Get the persistent id of a genotype." alt="" coords="289,5,374,33"/>
<area shape="rect" href="group__liteget.html#ga22a270ae73aaa86d93212a4a40f999dc" title="Get the name of a genotype." alt="" coords="278,57,385,85"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="271,109,392,137"/>
<area shape="rect" href="group__iterators.html#ga630ba7f8e864238e296cb9480dfbc728" title="Get the next location from a bidirectional iterator." alt="" coords="267,161,396,189"/>
<area shape="rect" href="group__iterators.html#gaa4fa2aec3b7f49f9925fe82c8ff38dcd" title="Initialise a Bidirectional iterator to the start of its sequence." alt="" coords="444,180,587,224"/>
</map>
</div>
<div id="dynsection-98" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-98-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-98-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-98-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a3bbb853ca13c593a940d86882dda2361_icgraph.png" border="0" usemap="#asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_icgraph" alt=""/></div>
<map name="asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_icgraph" id="asim-operations_8c_a3bbb853ca13c593a940d86882dda2361_icgraph">
<area shape="rect" title="Prints a matrix of genotype information to a file." alt="" coords="1126,111,1339,139"/>
<area shape="rect" href="group__savers.html#ga37bba3dd2808aad1e2e8df7581d74561" title="Prints allele counts of simulated genotypes to a file." alt="" coords="913,55,1063,99"/>
<area shape="rect" href="group__savers.html#gad5a9b943294b42f094bd347f990967cb" title="Prints simulated genotypes to a file." alt="" coords="898,140,1078,168"/>
<area shape="rect" href="group__savers.html#ga7636ac81969c8d41c8da27977bb19150" title="Prints allele counts of genotypes from the simulation to a file." alt="" coords="690,45,849,73"/>
<area shape="rect" href="sim-operations_8c.html#a0fe1d1ace12db253637f9df278cf1639" title="save&#45;as&#45;you&#45;go (genotypes/alleles)" alt="" coords="692,97,847,141"/>
<area shape="rect" href="sim-operations_8c.html#a311c08610c1a020fb05be94297d6f28f" title="Opens file for writing save&#45;as&#45;you&#45;go genotypes in accordance with gsc_GenOptions." alt="" coords="689,165,850,209"/>
<area shape="rect" href="group__savers.html#ga82a426c2dd33d2cf1e4da04d4e987445" title="Prints genotypes from the simulation to a file." alt="" coords="698,233,841,261"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
<a id="a368b4ddb966aef02950fe327d46f051b" name="a368b4ddb966aef02950fe327d46f051b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a368b4ddb966aef02950fe327d46f051b">&#9670;&nbsp;</a></span>gsc_set_names()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void gsc_set_names </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> *&#160;</td>
          <td class="paramname"><em>a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>prefix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>suffix</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unsigned int&#160;</td>
          <td class="paramname"><em>from_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Fills the designated section of the <code>.names</code> array in an <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> with the pattern "`prefix`index". </p>
<p >In future this function could be expanded to allow for different naming formats.</p>
<p >The index is padded with zeros depending on the size of <code>a-&gt;n_genotypes</code>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">a</td><td>pointer to the <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a> whose <code>.names</code> to modify </td></tr>
    <tr><td class="paramname">prefix</td><td>the prefix to add to the suffix to make the new genotype name </td></tr>
    <tr><td class="paramname">suffix</td><td>suffixes start at this value and increment for each additional name </td></tr>
    <tr><td class="paramname">from_index</td><td>the new names are added to this index and all those following it in this <a class="el" href="structgsc___allele_matrix.html">gsc_AlleleMatrix</a>. </td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="sim-operations_8c_source.html#l00580">580</a> of file <a class="el" href="sim-operations_8c_source.html">sim-operations.c</a>.</p>
<div id="dynsection-99" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-99-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-99-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-99-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a368b4ddb966aef02950fe327d46f051b_cgraph.png" border="0" usemap="#asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_cgraph" alt=""/></div>
<map name="asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_cgraph" id="asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_cgraph">
<area shape="rect" title="Fills the designated section of the .names array in an gsc_AlleleMatrix with the pattern &quot;`prefix`ind..." alt="" coords="5,32,119,60"/>
<area shape="rect" href="group__supporters.html#ga34ea6f4aea77b6bd95506ca071226544" title="Count and return the number of digits in i." alt="" coords="167,5,318,33"/>
<area shape="rect" href="sim-operations_8c.html#aaaf3ebc75feaf901829232712e55677b" title="Replace calls to malloc direct with this function." alt="" coords="182,57,303,85"/>
</map>
</div>
<div id="dynsection-100" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-100-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-100-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-100-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a368b4ddb966aef02950fe327d46f051b_icgraph.png" border="0" usemap="#asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_icgraph" alt=""/></div>
<map name="asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_icgraph" id="asim-operations_8c_a368b4ddb966aef02950fe327d46f051b_icgraph">
<area shape="rect" title="Fills the designated section of the .names array in an gsc_AlleleMatrix with the pattern &quot;`prefix`ind..." alt="" coords="892,140,1005,168"/>
<area shape="rect" href="sim-operations_8c.html#ab77b79416783c53492b3c2b0e3f5c7ea" title="Apply gsc_GenOptions naming scheme and gsc_PedigreeID allocation to a single gsc_AlleleMatrix." alt="" coords="689,132,844,176"/>
<area shape="rect" href="group__crossers.html#gaa5a1b4a1a80f8baf5b5730d7b7db0b77" title="Make new genotypes (generic function)" alt="" coords="479,132,641,176"/>
<area shape="rect" href="group__crossers.html#gaa3907b6b8ad3de641916b2df83e43a01" title="Creates an identical copy of each member of a group." alt="" coords="274,5,400,33"/>
<area shape="rect" href="group__crossers.html#ga6c230330ca9a9db9459d4477512a21af" title="Creates a doubled haploid from each member of a group." alt="" coords="243,57,431,85"/>
<area shape="rect" href="group__crossers.html#ga4bfb4f901f7a921e6a3a2f1d8a9ac249" title="Performs random crosses among members of a group." alt="" coords="245,109,429,137"/>
<area shape="rect" href="group__crossers.html#gaa43222283d6fb3745f5fab73b36af22d" title="Performs random crosses where the first parent comes from one group and the second from another." alt="" coords="245,161,429,205"/>
<area shape="rect" href="group__crossers.html#gafe3072701f2dea2a505dd8fc9e8bc936" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="243,229,431,257"/>
<area shape="rect" href="group__crossers.html#gab625453074798be6ba8677a5d4986edf" title="Selfs each member of a group for a certain number of generations." alt="" coords="276,281,398,309"/>
<area shape="rect" href="group__crossers.html#gaa70fd79312587d8e9fc166c64612cbce" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,161,192,205"/>
<area shape="rect" href="group__crossers.html#ga5c0be901d03d0e5dd3d74bc19a4bea70" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,229,195,257"/>
<area shape="rect" href="group__crossers.html#ga2b8100d2d30cf1ada0db8403663076ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,281,190,325"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
