<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: sim-operations.c File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.4</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#define-members">Macros</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle"><div class="title">sim-operations.c File Reference</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><code>#include &quot;<a class="el" href="sim-operations_8h_source.html">sim-operations.h</a>&quot;</code><br />
</div><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="define-members" name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:ad84149faf0b815bcee066748ef31eb60"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#ad84149faf0b815bcee066748ef31eb60">SIM_OPERATIONS</a></td></tr>
<tr class="separator:ad84149faf0b815bcee066748ef31eb60"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga6072a152b624743eeafb30adaf9b7f17"><td class="memItemLeft" align="right" valign="top">void *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga6072a152b624743eeafb30adaf9b7f17">get_malloc</a> (const size_t size)</td></tr>
<tr class="memdesc:ga6072a152b624743eeafb30adaf9b7f17"><td class="mdescLeft">&#160;</td><td class="mdescRight">Replace calls to malloc direct with this function, which errors and exits with status 2 if memory allocation fails.  <a href="group__supporters.html#ga6072a152b624743eeafb30adaf9b7f17">More...</a><br /></td></tr>
<tr class="separator:ga6072a152b624743eeafb30adaf9b7f17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf0399b852faa346adeab52d57fa9701b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gaf0399b852faa346adeab52d57fa9701b">create_empty_allelematrix</a> (const int n_markers, const int n_labels, const int labelDefaults[n_labels], const int n_genotypes)</td></tr>
<tr class="memdesc:gaf0399b852faa346adeab52d57fa9701b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> object of a given size.  <a href="group__loaders.html#gaf0399b852faa346adeab52d57fa9701b">More...</a><br /></td></tr>
<tr class="separator:gaf0399b852faa346adeab52d57fa9701b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac60a0289c8c051dab7bd8383a70a6245"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gac60a0289c8c051dab7bd8383a70a6245">create_empty_simdata</a> (RND_U32 RNGseed)</td></tr>
<tr class="memdesc:gac60a0289c8c051dab7bd8383a70a6245"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creator for an empty <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object on the heap.  <a href="group__loaders.html#gac60a0289c8c051dab7bd8383a70a6245">More...</a><br /></td></tr>
<tr class="separator:gac60a0289c8c051dab7bd8383a70a6245"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#gac052ad3cb1f8975c21b28eea560a6e0b">clear_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object on the heap.  <a href="group__loaders.html#gac052ad3cb1f8975c21b28eea560a6e0b">More...</a><br /></td></tr>
<tr class="separator:gac052ad3cb1f8975c21b28eea560a6e0b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac0aed93b85c3ff464c1e29bd2093e066"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gac0aed93b85c3ff464c1e29bd2093e066">randpoi</a> (rnd_pcg_t *rng, double lambda)</td></tr>
<tr class="memdesc:gac0aed93b85c3ff464c1e29bd2093e066"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates randomly a number from the Poisson distribution with parameter <code>lambda</code>, using the Knuth approach.  <a href="group__maths.html#gac0aed93b85c3ff464c1e29bd2093e066">More...</a><br /></td></tr>
<tr class="separator:gac0aed93b85c3ff464c1e29bd2093e066"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga80c7861f7c1014e8c406ef279a54759f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga80c7861f7c1014e8c406ef279a54759f">set_ids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int from_index, const int to_index)</td></tr>
<tr class="memdesc:ga80c7861f7c1014e8c406ef279a54759f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate lifetime-unique ids to each genotype in the range of whole <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> indexes <code>from_index</code> to <code>to_index</code> inclusive.  <a href="group__supporters.html#ga80c7861f7c1014e8c406ef279a54759f">More...</a><br /></td></tr>
<tr class="separator:ga80c7861f7c1014e8c406ef279a54759f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga10dddc2d2028b8c1341837897b6eee4d"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="struct_table_size.html">TableSize</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga10dddc2d2028b8c1341837897b6eee4d">get_file_dimensions</a> (const char *filename, const char sep)</td></tr>
<tr class="memdesc:ga10dddc2d2028b8c1341837897b6eee4d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Opens a table file and reads the number of columns and rows (including headers) separated by <code>sep</code> into a <a class="el" href="struct_table_size.html" title="A simple struct used for returning the dimensions of a matrix or table.">TableSize</a> struct that is returned.  <a href="group__supporters.html#ga10dddc2d2028b8c1341837897b6eee4d">More...</a><br /></td></tr>
<tr class="separator:ga10dddc2d2028b8c1341837897b6eee4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga98f69379fe6a5cca7fb25ed1fe49e1c7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7">get_from_ordered_uint_list</a> (const unsigned int target, const unsigned int listLen, const unsigned int list[listLen])</td></tr>
<tr class="memdesc:ga98f69379fe6a5cca7fb25ed1fe49e1c7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the located index in an array of integers where the integer is <code>target</code>.  <a href="group__supporters.html#ga98f69379fe6a5cca7fb25ed1fe49e1c7">More...</a><br /></td></tr>
<tr class="separator:ga98f69379fe6a5cca7fb25ed1fe49e1c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4e3d1fd7ae67dd133c4dbdec9a84d455"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga4e3d1fd7ae67dd133c4dbdec9a84d455">get_from_ordered_pedigree_list</a> (const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> target, const unsigned int listLen, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> list[listLen])</td></tr>
<tr class="memdesc:ga4e3d1fd7ae67dd133c4dbdec9a84d455"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the located index in an array of PedigreeIDs where the <a class="el" href="struct_pedigree_i_d.html" title="A type representing a program-lifetime-unique identifier for a genotype, to be used in tracking pedig...">PedigreeID</a> is <code>target</code>.  <a href="group__supporters.html#ga4e3d1fd7ae67dd133c4dbdec9a84d455">More...</a><br /></td></tr>
<tr class="separator:ga4e3d1fd7ae67dd133c4dbdec9a84d455"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf78d759e6da39a08acff41421d0e7b2a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaf78d759e6da39a08acff41421d0e7b2a">get_from_unordered_str_list</a> (const char *target, const int listLen, const char *list[listLen])</td></tr>
<tr class="memdesc:gaf78d759e6da39a08acff41421d0e7b2a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the first located index in an array of strings where the string is the same as the string <code>target</code>.  <a href="group__supporters.html#gaf78d759e6da39a08acff41421d0e7b2a">More...</a><br /></td></tr>
<tr class="separator:gaf78d759e6da39a08acff41421d0e7b2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga575fe459225d6b565a199426f1b5c030"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga575fe459225d6b565a199426f1b5c030">shuffle_up_to</a> (rnd_pcg_t *rng, int *sequence, const size_t total_n, const size_t n_to_shuffle)</td></tr>
<tr class="memdesc:ga575fe459225d6b565a199426f1b5c030"><td class="mdescLeft">&#160;</td><td class="mdescRight">Produce a random ordering of the first n elements in an array of integers using a (partial) Fisher-Yates shuffle.  <a href="group__supporters.html#ga575fe459225d6b565a199426f1b5c030">More...</a><br /></td></tr>
<tr class="separator:ga575fe459225d6b565a199426f1b5c030"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafad0846098c4da7ca876f0a625f369e1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gafad0846098c4da7ca876f0a625f369e1">set_names</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *a, const char *prefix, const int suffix, const int from_index)</td></tr>
<tr class="memdesc:gafad0846098c4da7ca876f0a625f369e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills the designated section of the <code>.names</code> array in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> with the pattern <code>prefix</code>index.  <a href="group__supporters.html#gafad0846098c4da7ca876f0a625f369e1">More...</a><br /></td></tr>
<tr class="separator:gafad0846098c4da7ca876f0a625f369e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa27804410b00527dfd9c043f6193e187"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_label_i_d.html">LabelID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa27804410b00527dfd9c043f6193e187">create_new_label</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int setTo)</td></tr>
<tr class="memdesc:gaa27804410b00527dfd9c043f6193e187"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialises a new custom label.  <a href="group__supporters.html#gaa27804410b00527dfd9c043f6193e187">More...</a><br /></td></tr>
<tr class="separator:gaa27804410b00527dfd9c043f6193e187"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga194393d51e63eddbb8323b9a80726d4a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga194393d51e63eddbb8323b9a80726d4a">set_label_default</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int newDefault)</td></tr>
<tr class="memdesc:ga194393d51e63eddbb8323b9a80726d4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the default value of a custom label.  <a href="group__supporters.html#ga194393d51e63eddbb8323b9a80726d4a">More...</a><br /></td></tr>
<tr class="separator:ga194393d51e63eddbb8323b9a80726d4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa65399dd3afee549ae0827a10780c1e9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa65399dd3afee549ae0827a10780c1e9">set_labels_to_const</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> whichGroup, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int setTo)</td></tr>
<tr class="memdesc:gaa65399dd3afee549ae0827a10780c1e9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set the values of a custom label.  <a href="group__supporters.html#gaa65399dd3afee549ae0827a10780c1e9">More...</a><br /></td></tr>
<tr class="separator:gaa65399dd3afee549ae0827a10780c1e9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae1f6208bda5336d11cfba62a03cb0401"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gae1f6208bda5336d11cfba62a03cb0401">increment_labels</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> whichGroup, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int byValue)</td></tr>
<tr class="memdesc:gae1f6208bda5336d11cfba62a03cb0401"><td class="mdescLeft">&#160;</td><td class="mdescRight">Increment the values of a custom label.  <a href="group__supporters.html#gae1f6208bda5336d11cfba62a03cb0401">More...</a><br /></td></tr>
<tr class="separator:gae1f6208bda5336d11cfba62a03cb0401"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga968801e51e13db7e536c997cbc84cacc"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga968801e51e13db7e536c997cbc84cacc">set_labels_to_values</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> whichGroup, const int startIndex, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int n_values, const int values[n_values])</td></tr>
<tr class="memdesc:ga968801e51e13db7e536c997cbc84cacc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy a vector of integers into a custom label.  <a href="group__supporters.html#ga968801e51e13db7e536c997cbc84cacc">More...</a><br /></td></tr>
<tr class="separator:ga968801e51e13db7e536c997cbc84cacc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga45b1944818993a6830a732fb13b493cb"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga45b1944818993a6830a732fb13b493cb">set_names_to_values</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> whichGroup, const int startIndex, const int n_values, const char *values[n_values])</td></tr>
<tr class="memdesc:ga45b1944818993a6830a732fb13b493cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy a vector of strings into the genotype name field.  <a href="group__supporters.html#ga45b1944818993a6830a732fb13b493cb">More...</a><br /></td></tr>
<tr class="separator:ga45b1944818993a6830a732fb13b493cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4467dd870605266ac95eff75bacffb1f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga4467dd870605266ac95eff75bacffb1f">get_integer_digits</a> (const int i)</td></tr>
<tr class="memdesc:ga4467dd870605266ac95eff75bacffb1f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Count and return the number of digits in <code>i</code>.  <a href="group__supporters.html#ga4467dd870605266ac95eff75bacffb1f">More...</a><br /></td></tr>
<tr class="separator:ga4467dd870605266ac95eff75bacffb1f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga80cab739edbd12bbd2ad1f9b77b27548">_simdata_pos_compare</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga80cab739edbd12bbd2ad1f9b77b27548">More...</a><br /></td></tr>
<tr class="separator:ga80cab739edbd12bbd2ad1f9b77b27548"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2aae9a568cd0cd5313808e458a9fe388"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga2aae9a568cd0cd5313808e458a9fe388">_descending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga2aae9a568cd0cd5313808e458a9fe388"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga2aae9a568cd0cd5313808e458a9fe388">More...</a><br /></td></tr>
<tr class="separator:ga2aae9a568cd0cd5313808e458a9fe388"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga311381904a1ca73ea26dfede9fcddea2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga311381904a1ca73ea26dfede9fcddea2">_ascending_double_comparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:ga311381904a1ca73ea26dfede9fcddea2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga311381904a1ca73ea26dfede9fcddea2">More...</a><br /></td></tr>
<tr class="separator:ga311381904a1ca73ea26dfede9fcddea2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3744bc6f3fcb52952d520029055e8af8"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga3744bc6f3fcb52952d520029055e8af8">_ascending_float_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:ga3744bc6f3fcb52952d520029055e8af8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#ga3744bc6f3fcb52952d520029055e8af8">More...</a><br /></td></tr>
<tr class="separator:ga3744bc6f3fcb52952d520029055e8af8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadae2e9f7e1061cb20426ba2333b7629b"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b">_ascending_int_comparer</a> (const void *p0, const void *p1)</td></tr>
<tr class="memdesc:gadae2e9f7e1061cb20426ba2333b7629b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b">More...</a><br /></td></tr>
<tr class="separator:gadae2e9f7e1061cb20426ba2333b7629b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacda79ff4e2a3229d13c5393d006136db"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gacda79ff4e2a3229d13c5393d006136db">_ascending_int_dcomparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:gacda79ff4e2a3229d13c5393d006136db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="group__supporters.html#gacda79ff4e2a3229d13c5393d006136db">More...</a><br /></td></tr>
<tr class="separator:gacda79ff4e2a3229d13c5393d006136db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a60b92e4ec1dac6abca4fd6c09486c237"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a60b92e4ec1dac6abca4fd6c09486c237">_ascending_uint_dcomparer</a> (const void *pp0, const void *pp1)</td></tr>
<tr class="memdesc:a60b92e4ec1dac6abca4fd6c09486c237"><td class="mdescLeft">&#160;</td><td class="mdescRight">Comparator function for qsort.  <a href="sim-operations_8c.html#a60b92e4ec1dac6abca4fd6c09486c237">More...</a><br /></td></tr>
<tr class="separator:a60b92e4ec1dac6abca4fd6c09486c237"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga887027349503be24b1a847c9b833aaac"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga887027349503be24b1a847c9b833aaac">condense_allele_matrix</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga887027349503be24b1a847c9b833aaac"><td class="mdescLeft">&#160;</td><td class="mdescRight">A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__supporters.html#ga887027349503be24b1a847c9b833aaac">More...</a><br /></td></tr>
<tr class="separator:ga887027349503be24b1a847c9b833aaac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa3cc5e7ea10f92c655ba7e4b88468a35"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa3cc5e7ea10f92c655ba7e4b88468a35">create_bidirectional_iter</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:gaa3cc5e7ea10f92c655ba7e4b88468a35"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a bidirectional iterator.  <a href="group__iterators.html#gaa3cc5e7ea10f92c655ba7e4b88468a35">More...</a><br /></td></tr>
<tr class="separator:gaa3cc5e7ea10f92c655ba7e4b88468a35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga028299459e350a2bbc2ab3874a109a0e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_random_access_iterator.html">RandomAccessIterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga028299459e350a2bbc2ab3874a109a0e">create_randomaccess_iter</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:ga028299459e350a2bbc2ab3874a109a0e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Create a Random Access Iterator.  <a href="group__iterators.html#ga028299459e350a2bbc2ab3874a109a0e">More...</a><br /></td></tr>
<tr class="separator:ga028299459e350a2bbc2ab3874a109a0e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaab4c1841241403397fac8c6a2087a497"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaab4c1841241403397fac8c6a2087a497">get_nth_AlleleMatrix</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *listStart, const unsigned int n)</td></tr>
<tr class="memdesc:gaab4c1841241403397fac8c6a2087a497"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> by index in the linked list.  <a href="group__iterators.html#gaab4c1841241403397fac8c6a2087a497">More...</a><br /></td></tr>
<tr class="separator:gaab4c1841241403397fac8c6a2087a497"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7ff54a3cacc5c06c334cd9e8a2635204"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga7ff54a3cacc5c06c334cd9e8a2635204">validate_bidirectional_cache</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga7ff54a3cacc5c06c334cd9e8a2635204"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check nothing in the <a class="el" href="struct_bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a SimData or through only the ...">BidirectionalIterator</a> is obviously invalid.  <a href="group__iterators.html#ga7ff54a3cacc5c06c334cd9e8a2635204">More...</a><br /></td></tr>
<tr class="separator:ga7ff54a3cacc5c06c334cd9e8a2635204"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf1ee9717701b8580e118144c5671cb5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gabf1ee9717701b8580e118144c5671cb5">set_bidirectional_iter_to_start</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gabf1ee9717701b8580e118144c5671cb5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the start of its sequence.  <a href="group__iterators.html#gabf1ee9717701b8580e118144c5671cb5">More...</a><br /></td></tr>
<tr class="separator:gabf1ee9717701b8580e118144c5671cb5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa7e9bc9b0ea586369586d01237d30258"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gaa7e9bc9b0ea586369586d01237d30258">set_bidirectional_iter_to_end</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gaa7e9bc9b0ea586369586d01237d30258"><td class="mdescLeft">&#160;</td><td class="mdescRight">Initialise a Bidirectional iterator to the end of its sequence.  <a href="group__iterators.html#gaa7e9bc9b0ea586369586d01237d30258">More...</a><br /></td></tr>
<tr class="separator:gaa7e9bc9b0ea586369586d01237d30258"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadff0d81d4ef4fd1f5afcdc7885779389"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#gadff0d81d4ef4fd1f5afcdc7885779389">next_forwards</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gadff0d81d4ef4fd1f5afcdc7885779389"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the next location from a bidirectional iterator.  <a href="group__iterators.html#gadff0d81d4ef4fd1f5afcdc7885779389">More...</a><br /></td></tr>
<tr class="separator:gadff0d81d4ef4fd1f5afcdc7885779389"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9431cd1ae9e15331a5464ac1c1204a8d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga9431cd1ae9e15331a5464ac1c1204a8d">next_backwards</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:ga9431cd1ae9e15331a5464ac1c1204a8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the previous location from a bidirectional iterator.  <a href="group__iterators.html#ga9431cd1ae9e15331a5464ac1c1204a8d">More...</a><br /></td></tr>
<tr class="separator:ga9431cd1ae9e15331a5464ac1c1204a8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0bc5b9ebb0857a471e41630fda00403b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga0bc5b9ebb0857a471e41630fda00403b">next_get_nth</a> (<a class="el" href="struct_random_access_iterator.html">RandomAccessIterator</a> *it, const unsigned long int n)</td></tr>
<tr class="memdesc:ga0bc5b9ebb0857a471e41630fda00403b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get a location by index using a <a class="el" href="struct_random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a SimData or of all the members of a grou...">RandomAccessIterator</a>.  <a href="group__iterators.html#ga0bc5b9ebb0857a471e41630fda00403b">More...</a><br /></td></tr>
<tr class="separator:ga0bc5b9ebb0857a471e41630fda00403b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaea63100f4c2639907cce72c29a413e8"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gaaea63100f4c2639907cce72c29a413e8">get_name_of_id</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> id)</td></tr>
<tr class="memdesc:gaaea63100f4c2639907cce72c29a413e8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the name of the genotype with a given id.  <a href="group__search.html#gaaea63100f4c2639907cce72c29a413e8">More...</a><br /></td></tr>
<tr class="separator:gaaea63100f4c2639907cce72c29a413e8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga25a640c4ebdcdca0fade3ad3d74523e6"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga25a640c4ebdcdca0fade3ad3d74523e6">get_genes_of_id</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> id)</td></tr>
<tr class="memdesc:ga25a640c4ebdcdca0fade3ad3d74523e6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the alleles at each marker of the genotype with a given id.  <a href="group__search.html#ga25a640c4ebdcdca0fade3ad3d74523e6">More...</a><br /></td></tr>
<tr class="separator:ga25a640c4ebdcdca0fade3ad3d74523e6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa814762960a964cd0aaf82be0c18b8cb"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#gaa814762960a964cd0aaf82be0c18b8cb">get_parents_of_id</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> id, <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> output[2])</td></tr>
<tr class="memdesc:gaa814762960a964cd0aaf82be0c18b8cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Saves the ids of the parents of a genotype with a particular id to the output array <code>output</code>.  <a href="group__search.html#gaa814762960a964cd0aaf82be0c18b8cb">More...</a><br /></td></tr>
<tr class="separator:gaa814762960a964cd0aaf82be0c18b8cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga27f29c0ac93e571fd539e81ce6dacb4b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga27f29c0ac93e571fd539e81ce6dacb4b">get_ids_of_names</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const int n_names, const char *names[n_names], <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> *output)</td></tr>
<tr class="memdesc:ga27f29c0ac93e571fd539e81ce6dacb4b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for genotypes with certain names in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> and save the ids of those names.  <a href="group__search.html#ga27f29c0ac93e571fd539e81ce6dacb4b">More...</a><br /></td></tr>
<tr class="separator:ga27f29c0ac93e571fd539e81ce6dacb4b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7bef742844713ce42ef56eb73050d1bb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_pedigree_i_d.html">PedigreeID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga7bef742844713ce42ef56eb73050d1bb">get_id_of_child</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> parent1id, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> parent2id)</td></tr>
<tr class="memdesc:ga7bef742844713ce42ef56eb73050d1bb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its id.  <a href="group__search.html#ga7bef742844713ce42ef56eb73050d1bb">More...</a><br /></td></tr>
<tr class="separator:ga7bef742844713ce42ef56eb73050d1bb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6803a600c63f5789aa97275acf97365e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga6803a600c63f5789aa97275acf97365e">get_index_of_child</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> parent1id, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> parent2id)</td></tr>
<tr class="memdesc:ga6803a600c63f5789aa97275acf97365e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with parentage matching two given parent ids in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="group__search.html#ga6803a600c63f5789aa97275acf97365e">More...</a><br /></td></tr>
<tr class="separator:ga6803a600c63f5789aa97275acf97365e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga852d1f756b52ea86e30e63335dcc33e1"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga852d1f756b52ea86e30e63335dcc33e1">get_index_of_name</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const char *name)</td></tr>
<tr class="memdesc:ga852d1f756b52ea86e30e63335dcc33e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Search for a genotype with a particular name in a linked list of <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and return its index.  <a href="group__search.html#ga852d1f756b52ea86e30e63335dcc33e1">More...</a><br /></td></tr>
<tr class="separator:ga852d1f756b52ea86e30e63335dcc33e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8c2bc00d2b81a4c293a939f202250eb8"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_pedigree_i_d.html">PedigreeID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga8c2bc00d2b81a4c293a939f202250eb8">get_id_of_index</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const int index)</td></tr>
<tr class="memdesc:ga8c2bc00d2b81a4c293a939f202250eb8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the id of a genotype by its index.  <a href="group__search.html#ga8c2bc00d2b81a4c293a939f202250eb8">More...</a><br /></td></tr>
<tr class="separator:ga8c2bc00d2b81a4c293a939f202250eb8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga31eb9ab2cd17d7c107102922e2609aac"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__search.html#ga31eb9ab2cd17d7c107102922e2609aac">get_genes_of_index</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *start, const int index)</td></tr>
<tr class="memdesc:ga31eb9ab2cd17d7c107102922e2609aac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of a genotype by its index.  <a href="group__search.html#ga31eb9ab2cd17d7c107102922e2609aac">More...</a><br /></td></tr>
<tr class="separator:ga31eb9ab2cd17d7c107102922e2609aac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga54018cd747361983c4d3a90ea9abc05d"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga54018cd747361983c4d3a90ea9abc05d">combine_groups</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int list_len, const <a class="el" href="struct_group_num.html">GroupNum</a> group_ids[list_len])</td></tr>
<tr class="memdesc:ga54018cd747361983c4d3a90ea9abc05d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Combine a set of groups into one group.  <a href="group__groupmod.html#ga54018cd747361983c4d3a90ea9abc05d">More...</a><br /></td></tr>
<tr class="separator:ga54018cd747361983c4d3a90ea9abc05d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5cef765475767b55605d62171fd96334"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga5cef765475767b55605d62171fd96334">split_from_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int n, const int indexes_to_split[n])</td></tr>
<tr class="memdesc:ga5cef765475767b55605d62171fd96334"><td class="mdescLeft">&#160;</td><td class="mdescRight">Take a list of indexes and allocate the genotypes at those indexes to a new group.  <a href="group__groupmod.html#ga5cef765475767b55605d62171fd96334">More...</a><br /></td></tr>
<tr class="separator:ga5cef765475767b55605d62171fd96334"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga94e9e1a407c7effc6f886b0b13084184"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga94e9e1a407c7effc6f886b0b13084184">split_by_label_value</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int valueToSplit)</td></tr>
<tr class="memdesc:ga94e9e1a407c7effc6f886b0b13084184"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocates the genotypes with a particular value of a label to a new group.  <a href="group__groupmod.html#ga94e9e1a407c7effc6f886b0b13084184">More...</a><br /></td></tr>
<tr class="separator:ga94e9e1a407c7effc6f886b0b13084184"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6cc15e471588f5b7e00d358e6af232fd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga6cc15e471588f5b7e00d358e6af232fd">split_by_label_range</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel, const int valueLowBound, const int valueHighBound)</td></tr>
<tr class="memdesc:ga6cc15e471588f5b7e00d358e6af232fd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocates the genotypes with values of a label in a particular range to a new group.  <a href="group__groupmod.html#ga6cc15e471588f5b7e00d358e6af232fd">More...</a><br /></td></tr>
<tr class="separator:ga6cc15e471588f5b7e00d358e6af232fd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga252f2045d78cdbc78d79b4896464ea37"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga252f2045d78cdbc78d79b4896464ea37">split_into_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga252f2045d78cdbc78d79b4896464ea37"><td class="mdescLeft">&#160;</td><td class="mdescRight">Give every individual in the group a new group number that does not belong to any other existing group (thereby allocating each genotype in the group to a new group of 1).  <a href="group__groupmod.html#ga252f2045d78cdbc78d79b4896464ea37">More...</a><br /></td></tr>
<tr class="separator:ga252f2045d78cdbc78d79b4896464ea37"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad2f97fc0c88dd0838cc492246049dce7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gad2f97fc0c88dd0838cc492246049dce7">split_into_families</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:gad2f97fc0c88dd0838cc492246049dce7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into a set of smaller groups, each of which contains the genotypes in the original group that share a particular pair of parents.  <a href="group__groupmod.html#gad2f97fc0c88dd0838cc492246049dce7">More...</a><br /></td></tr>
<tr class="separator:gad2f97fc0c88dd0838cc492246049dce7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8178833ed59cb2ba13caf00aa33e8b85"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga8178833ed59cb2ba13caf00aa33e8b85">split_into_halfsib_families</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const int parent, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga8178833ed59cb2ba13caf00aa33e8b85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into a set of smaller groups, each containing the genotypes from the original group that share one parent.  <a href="group__groupmod.html#ga8178833ed59cb2ba13caf00aa33e8b85">More...</a><br /></td></tr>
<tr class="separator:ga8178833ed59cb2ba13caf00aa33e8b85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf21b446a646f846db7c4053f9e8bcb65"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gaf21b446a646f846db7c4053f9e8bcb65">split_evenly_into_two</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gaf21b446a646f846db7c4053f9e8bcb65"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into two groups of equal size (or size differing only by one, if the original group had an odd number of members) using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#gaf21b446a646f846db7c4053f9e8bcb65">More...</a><br /></td></tr>
<tr class="separator:gaf21b446a646f846db7c4053f9e8bcb65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac3eb8cc70c4df57ca456af66b25bdc66"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gac3eb8cc70c4df57ca456af66b25bdc66">split_evenly_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const int n, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:gac3eb8cc70c4df57ca456af66b25bdc66"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size.), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#gac3eb8cc70c4df57ca456af66b25bdc66">More...</a><br /></td></tr>
<tr class="separator:gac3eb8cc70c4df57ca456af66b25bdc66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf4934fda59a428b74bccc307d7f505b1"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gaf4934fda59a428b74bccc307d7f505b1">split_by_specific_counts_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const int n, const int *counts, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:gaf4934fda59a428b74bccc307d7f505b1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Split a group into n groups of equal size (or size differing only by one, if n does not perfectly divide the group size), using a random permutation of the group members to determine which goes where.  <a href="group__groupmod.html#gaf4934fda59a428b74bccc307d7f505b1">More...</a><br /></td></tr>
<tr class="separator:gaf4934fda59a428b74bccc307d7f505b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga20b19aaab1ae2bf54cb657284d2a6935"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga20b19aaab1ae2bf54cb657284d2a6935">split_randomly_into_two</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:ga20b19aaab1ae2bf54cb657284d2a6935"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flip a coin for each member of the group to decide if it should be moved to the new group.  <a href="group__groupmod.html#ga20b19aaab1ae2bf54cb657284d2a6935">More...</a><br /></td></tr>
<tr class="separator:ga20b19aaab1ae2bf54cb657284d2a6935"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga59d758c73d0b494035b12f644ca8b13d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#ga59d758c73d0b494035b12f644ca8b13d">split_randomly_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const int n, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:ga59d758c73d0b494035b12f644ca8b13d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with equal probability.  <a href="group__groupmod.html#ga59d758c73d0b494035b12f644ca8b13d">More...</a><br /></td></tr>
<tr class="separator:ga59d758c73d0b494035b12f644ca8b13d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaafec5b8128166ccc93da48dd7533b22e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__groupmod.html#gaafec5b8128166ccc93da48dd7533b22e">split_by_probabilities_into_n</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const int n, const double *probs, <a class="el" href="struct_group_num.html">GroupNum</a> *results)</td></tr>
<tr class="memdesc:gaafec5b8128166ccc93da48dd7533b22e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allocate each member of the group to one of n groups with custom probabilities for each group.  <a href="group__groupmod.html#gaafec5b8128166ccc93da48dd7533b22e">More...</a><br /></td></tr>
<tr class="separator:gaafec5b8128166ccc93da48dd7533b22e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacadb6fd4167febae0fcd3f653ee8a534"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gacadb6fd4167febae0fcd3f653ee8a534">get_existing_groups</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int n_groups, <a class="el" href="struct_group_num.html">GroupNum</a> *output)</td></tr>
<tr class="memdesc:gacadb6fd4167febae0fcd3f653ee8a534"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify group numbers that currently have members.  <a href="group__collgetters.html#gacadb6fd4167febae0fcd3f653ee8a534">More...</a><br /></td></tr>
<tr class="separator:gacadb6fd4167febae0fcd3f653ee8a534"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5544a69d0f6abf525423479fc77d4805"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga5544a69d0f6abf525423479fc77d4805">get_existing_group_counts</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int n_groups, <a class="el" href="struct_group_num.html">GroupNum</a> *output_groups, int *output_sizes)</td></tr>
<tr class="memdesc:ga5544a69d0f6abf525423479fc77d4805"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify group numbers that currently has members and the number of genotypes currently allocated to those groups.  <a href="group__collgetters.html#ga5544a69d0f6abf525423479fc77d4805">More...</a><br /></td></tr>
<tr class="separator:ga5544a69d0f6abf525423479fc77d4805"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf0bc58eab8b8e413c1b20140fab8d17f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaf0bc58eab8b8e413c1b20140fab8d17f">get_new_group_num</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:gaf0bc58eab8b8e413c1b20140fab8d17f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that does not identify a group that currently has member(s).  <a href="group__supporters.html#gaf0bc58eab8b8e413c1b20140fab8d17f">More...</a><br /></td></tr>
<tr class="separator:gaf0bc58eab8b8e413c1b20140fab8d17f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8db3a8d7b93a8441f1a95d7ff7f228aa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga8db3a8d7b93a8441f1a95d7ff7f228aa">get_index_of_label</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_label_i_d.html">LabelID</a> label)</td></tr>
<tr class="memdesc:ga8db3a8d7b93a8441f1a95d7ff7f228aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the label lookup index of a label identifier.  <a href="group__supporters.html#ga8db3a8d7b93a8441f1a95d7ff7f228aa">More...</a><br /></td></tr>
<tr class="separator:ga8db3a8d7b93a8441f1a95d7ff7f228aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa605100835fdec6b250fde0ad5a9cc8f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gaa605100835fdec6b250fde0ad5a9cc8f">get_index_of_eff_set</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_effect_i_d.html">EffectID</a> eff_set_id)</td></tr>
<tr class="memdesc:gaa605100835fdec6b250fde0ad5a9cc8f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the lookup index of a marker effect set identifier.  <a href="group__supporters.html#gaa605100835fdec6b250fde0ad5a9cc8f">More...</a><br /></td></tr>
<tr class="separator:gaa605100835fdec6b250fde0ad5a9cc8f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3dd65057ec4a330cf33fce01a5956d5f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_label_i_d.html">LabelID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga3dd65057ec4a330cf33fce01a5956d5f">get_new_label_id</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga3dd65057ec4a330cf33fce01a5956d5f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that is not already allocated to a label in the simulation.  <a href="group__supporters.html#ga3dd65057ec4a330cf33fce01a5956d5f">More...</a><br /></td></tr>
<tr class="separator:ga3dd65057ec4a330cf33fce01a5956d5f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga921f3d9c1ff27c103a436e4f36292a14"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_effect_i_d.html">EffectID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga921f3d9c1ff27c103a436e4f36292a14">get_new_eff_set_id</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga921f3d9c1ff27c103a436e4f36292a14"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next sequential integer that is not already allocated to a marker effect set ID in the simulation.  <a href="group__supporters.html#ga921f3d9c1ff27c103a436e4f36292a14">More...</a><br /></td></tr>
<tr class="separator:ga921f3d9c1ff27c103a436e4f36292a14"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5f4874bcb65118360f7bba3870efe5e2"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga5f4874bcb65118360f7bba3870efe5e2">get_n_new_group_nums</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int n, <a class="el" href="struct_group_num.html">GroupNum</a> *result)</td></tr>
<tr class="memdesc:ga5f4874bcb65118360f7bba3870efe5e2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to identify the next n sequential integers that do not identify a group that currently has member(s).  <a href="group__supporters.html#ga5f4874bcb65118360f7bba3870efe5e2">More...</a><br /></td></tr>
<tr class="separator:ga5f4874bcb65118360f7bba3870efe5e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6773f2b434edda042163890dea8f8dbd"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga6773f2b434edda042163890dea8f8dbd">get_group_size</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:ga6773f2b434edda042163890dea8f8dbd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to count the number of genotypes that currently belong to the specified group.  <a href="group__collgetters.html#ga6773f2b434edda042163890dea8f8dbd">More...</a><br /></td></tr>
<tr class="separator:ga6773f2b434edda042163890dea8f8dbd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2897db8c7fffee9b2a4bcd7f81e368f6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga2897db8c7fffee9b2a4bcd7f81e368f6">get_group_genes</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, char **output)</td></tr>
<tr class="memdesc:ga2897db8c7fffee9b2a4bcd7f81e368f6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the genes/alleles of each member of the group.  <a href="group__collgetters.html#ga2897db8c7fffee9b2a4bcd7f81e368f6">More...</a><br /></td></tr>
<tr class="separator:ga2897db8c7fffee9b2a4bcd7f81e368f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab535d4c5e7afdce3816ab40c19a46a78"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gab535d4c5e7afdce3816ab40c19a46a78">get_group_names</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, char **output)</td></tr>
<tr class="memdesc:gab535d4c5e7afdce3816ab40c19a46a78"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets a shallow copy of the names of each member of the group.  <a href="group__collgetters.html#gab535d4c5e7afdce3816ab40c19a46a78">More...</a><br /></td></tr>
<tr class="separator:gab535d4c5e7afdce3816ab40c19a46a78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacc8432b03b8b5a33e0adde825c02ee7a"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gacc8432b03b8b5a33e0adde825c02ee7a">get_group_ids</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> *output)</td></tr>
<tr class="memdesc:gacc8432b03b8b5a33e0adde825c02ee7a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of each member of the group.  <a href="group__collgetters.html#gacc8432b03b8b5a33e0adde825c02ee7a">More...</a><br /></td></tr>
<tr class="separator:gacc8432b03b8b5a33e0adde825c02ee7a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga635f69a6776949d41a7b35346a70fc51"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga635f69a6776949d41a7b35346a70fc51">get_group_indexes</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, int *output)</td></tr>
<tr class="memdesc:ga635f69a6776949d41a7b35346a70fc51"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the indexes (0-based, from the start of the linked list in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>) of each member of the group.  <a href="group__collgetters.html#ga635f69a6776949d41a7b35346a70fc51">More...</a><br /></td></tr>
<tr class="separator:ga635f69a6776949d41a7b35346a70fc51"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaabe380985e14b55021634b3c687004e3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gaabe380985e14b55021634b3c687004e3">get_group_bvs</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID, int group_size, double *output)</td></tr>
<tr class="memdesc:gaabe380985e14b55021634b3c687004e3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the breeding values/breeding values/fitnesses of each member of the group.  <a href="group__collgetters.html#gaabe380985e14b55021634b3c687004e3">More...</a><br /></td></tr>
<tr class="separator:gaabe380985e14b55021634b3c687004e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac6f09e3398d4bf55d071d3a77e2f60f9"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gac6f09e3398d4bf55d071d3a77e2f60f9">get_group_parent_ids</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, const int whichParent, <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> *output)</td></tr>
<tr class="memdesc:gac6f09e3398d4bf55d071d3a77e2f60f9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the ids of either the first or second parent of each member of the group.  <a href="group__collgetters.html#gac6f09e3398d4bf55d071d3a77e2f60f9">More...</a><br /></td></tr>
<tr class="separator:gac6f09e3398d4bf55d071d3a77e2f60f9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad9cd4b14b44c05037093599c72d3b074"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#gad9cd4b14b44c05037093599c72d3b074">get_group_parent_names</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, const int whichParent, char **output)</td></tr>
<tr class="memdesc:gad9cd4b14b44c05037093599c72d3b074"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the names of either the first or second parent of each member of the group.  <a href="group__collgetters.html#gad9cd4b14b44c05037093599c72d3b074">More...</a><br /></td></tr>
<tr class="separator:gad9cd4b14b44c05037093599c72d3b074"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga31b649f4b57e0fa7ed2490b4cc48c97e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__collgetters.html#ga31b649f4b57e0fa7ed2490b4cc48c97e">get_group_pedigrees</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id, int group_size, char **output)</td></tr>
<tr class="memdesc:ga31b649f4b57e0fa7ed2490b4cc48c97e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the full pedigree string (as per <a class="el" href="group__savers.html#ga821e9489b1668201f9bd38d9ac467673" title="Print the full known pedigree of each genotype in a group to a file.">save_group_full_pedigree()</a> ) of each member of the group.  <a href="group__collgetters.html#ga31b649f4b57e0fa7ed2490b4cc48c97e">More...</a><br /></td></tr>
<tr class="separator:ga31b649f4b57e0fa7ed2490b4cc48c97e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga26e7ca69aa8bd969e3a781331a0afe27"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27">generate_zero_dmatrix</a> (const int r, const int c)</td></tr>
<tr class="memdesc:ga26e7ca69aa8bd969e3a781331a0afe27"><td class="mdescLeft">&#160;</td><td class="mdescRight">Generates a matrix of c columns, r rows with all 0.  <a href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27">More...</a><br /></td></tr>
<tr class="separator:ga26e7ca69aa8bd969e3a781331a0afe27"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga029101c5b02d8ab22e74ce3c103f6114"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#ga029101c5b02d8ab22e74ce3c103f6114">add_matrixvector_product_to_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *result, const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *a, const double *b)</td></tr>
<tr class="memdesc:ga029101c5b02d8ab22e74ce3c103f6114"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> to a vector, and add that product to the first column of a provided <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__maths.html#ga029101c5b02d8ab22e74ce3c103f6114">More...</a><br /></td></tr>
<tr class="separator:ga029101c5b02d8ab22e74ce3c103f6114"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab1f0bafd56d3496348ea3ccddfe9d937"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__maths.html#gab1f0bafd56d3496348ea3ccddfe9d937">add_doublematrixvector_product_to_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *result, const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *amat, const double *avec, const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *bmat, const double *bvec)</td></tr>
<tr class="memdesc:gab1f0bafd56d3496348ea3ccddfe9d937"><td class="mdescLeft">&#160;</td><td class="mdescRight">Multiply two sets of a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> and vector, and add both products to the first column of a provided <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__maths.html#gab1f0bafd56d3496348ea3ccddfe9d937">More...</a><br /></td></tr>
<tr class="separator:gab1f0bafd56d3496348ea3ccddfe9d937"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a">delete_dmatrix</a> (<a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *m)</td></tr>
<tr class="memdesc:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> and frees its memory.  <a href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a">More...</a><br /></td></tr>
<tr class="separator:gaaf6542d59034f88f9e1ec48e5f053c4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga09859190fb4250475a866fa78a67be52"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga09859190fb4250475a866fa78a67be52">delete_group</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:ga09859190fb4250475a866fa78a67be52"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes all genotypes belonging to a particular group.  <a href="group__deletors.html#ga09859190fb4250475a866fa78a67be52">More...</a><br /></td></tr>
<tr class="separator:ga09859190fb4250475a866fa78a67be52"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga85e0f5ee7e16b9b029eb92ef30ae419a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga85e0f5ee7e16b9b029eb92ef30ae419a">delete_eff_set</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_effect_i_d.html">EffectID</a> whichID)</td></tr>
<tr class="memdesc:ga85e0f5ee7e16b9b029eb92ef30ae419a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a particular set of marker effects from memory.  <a href="group__deletors.html#ga85e0f5ee7e16b9b029eb92ef30ae419a">More...</a><br /></td></tr>
<tr class="separator:ga85e0f5ee7e16b9b029eb92ef30ae419a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaef409380cea0efeb551e6c3bcc455168"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaef409380cea0efeb551e6c3bcc455168">delete_label</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_label_i_d.html">LabelID</a> whichLabel)</td></tr>
<tr class="memdesc:gaef409380cea0efeb551e6c3bcc455168"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clears memory of this label from the simulation and all its genotypes.  <a href="group__deletors.html#gaef409380cea0efeb551e6c3bcc455168">More...</a><br /></td></tr>
<tr class="separator:gaef409380cea0efeb551e6c3bcc455168"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaaf7e12b92ec0f93bb3bc2ec67155bc94">delete_genmap</a> (<a class="el" href="struct_genetic_map.html">GeneticMap</a> *m)</td></tr>
<tr class="memdesc:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_genetic_map.html" title="A type that stores the genetic map for a set of markers.">GeneticMap</a> object and frees its memory.  <a href="group__deletors.html#gaaf7e12b92ec0f93bb3bc2ec67155bc94">More...</a><br /></td></tr>
<tr class="separator:gaaf7e12b92ec0f93bb3bc2ec67155bc94"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gadd4a9c17e8c3e59fc00eee25635b110f">delete_allele_matrix</a> (<a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Delete the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> linked list from m onwards and frees its memory.  <a href="group__deletors.html#gadd4a9c17e8c3e59fc00eee25635b110f">More...</a><br /></td></tr>
<tr class="separator:gadd4a9c17e8c3e59fc00eee25635b110f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga961f4b0757752fee2bd08adde7ece8ad"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga961f4b0757752fee2bd08adde7ece8ad">delete_effect_matrix</a> (<a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *m)</td></tr>
<tr class="memdesc:ga961f4b0757752fee2bd08adde7ece8ad"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes an <a class="el" href="struct_effect_matrix.html" title="A type that stores a matrix of effect values and their names.">EffectMatrix</a> object and frees its memory.  <a href="group__deletors.html#ga961f4b0757752fee2bd08adde7ece8ad">More...</a><br /></td></tr>
<tr class="separator:ga961f4b0757752fee2bd08adde7ece8ad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#ga6fc4a687a2186e9e9dcf3229d9141041">delete_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object and frees its memory.  <a href="group__deletors.html#ga6fc4a687a2186e9e9dcf3229d9141041">More...</a><br /></td></tr>
<tr class="separator:ga6fc4a687a2186e9e9dcf3229d9141041"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaea2a41cc103e8d11bedff564f93a4759"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gaea2a41cc103e8d11bedff564f93a4759">delete_markerblocks</a> (<a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> *b)</td></tr>
<tr class="memdesc:gaea2a41cc103e8d11bedff564f93a4759"><td class="mdescLeft">&#160;</td><td class="mdescRight">Delete a <a class="el" href="struct_marker_blocks.html" title="A struct used to store a set of blocks of markers.">MarkerBlocks</a> struct.  <a href="group__deletors.html#gaea2a41cc103e8d11bedff564f93a4759">More...</a><br /></td></tr>
<tr class="separator:gaea2a41cc103e8d11bedff564f93a4759"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac29aa376a4682a50ba040c9723745354"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gac29aa376a4682a50ba040c9723745354">delete_bidirectional_iter</a> (<a class="el" href="struct_bidirectional_iterator.html">BidirectionalIterator</a> *it)</td></tr>
<tr class="memdesc:gac29aa376a4682a50ba040c9723745354"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_bidirectional_iterator.html" title="A structure to iterate forwards and backwards through all genotypes in a SimData or through only the ...">BidirectionalIterator</a> object.  <a href="group__deletors.html#gac29aa376a4682a50ba040c9723745354">More...</a><br /></td></tr>
<tr class="separator:gac29aa376a4682a50ba040c9723745354"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabb7f5bb0df140f22d55c946df3a74f08"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__deletors.html#gabb7f5bb0df140f22d55c946df3a74f08">delete_randomaccess_iter</a> (<a class="el" href="struct_random_access_iterator.html">RandomAccessIterator</a> *it)</td></tr>
<tr class="memdesc:gabb7f5bb0df140f22d55c946df3a74f08"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deletes a <a class="el" href="struct_random_access_iterator.html" title="A structure to search and cache indexes of all genotypes in a SimData or of all the members of a grou...">RandomAccessIterator</a> object and frees its memory.  <a href="group__deletors.html#gabb7f5bb0df140f22d55c946df3a74f08">More...</a><br /></td></tr>
<tr class="separator:gabb7f5bb0df140f22d55c946df3a74f08"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8a94c6099bacfff7a56a417162393510"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga8a94c6099bacfff7a56a417162393510">load_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga8a94c6099bacfff7a56a417162393510"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="group__loaders.html#ga8a94c6099bacfff7a56a417162393510">More...</a><br /></td></tr>
<tr class="separator:ga8a94c6099bacfff7a56a417162393510"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga65488a38fc6d9f437175e58d6fc481d1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga65488a38fc6d9f437175e58d6fc481d1">load_transposed_encoded_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga65488a38fc6d9f437175e58d6fc481d1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with marker allele data.  <a href="group__loaders.html#ga65488a38fc6d9f437175e58d6fc481d1">More...</a><br /></td></tr>
<tr class="separator:ga65488a38fc6d9f437175e58d6fc481d1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga38ad77c70ea0719c4c59644127fd3368"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga38ad77c70ea0719c4c59644127fd3368">load_more_transposed_genes_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga38ad77c70ea0719c4c59644127fd3368"><td class="mdescLeft">&#160;</td><td class="mdescRight">Appends genotype data from a file to an existing <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__loaders.html#ga38ad77c70ea0719c4c59644127fd3368">More...</a><br /></td></tr>
<tr class="separator:ga38ad77c70ea0719c4c59644127fd3368"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga8edb09b9b9eb27f84ac59cc4342ea855">load_genmap_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with data from a genetic map.  <a href="group__loaders.html#ga8edb09b9b9eb27f84ac59cc4342ea855">More...</a><br /></td></tr>
<tr class="separator:ga8edb09b9b9eb27f84ac59cc4342ea855"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadebbfa88bf139ddca501ab158f3746c5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#gadebbfa88bf139ddca501ab158f3746c5">get_sorted_markers</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, int actual_n_markers)</td></tr>
<tr class="memdesc:gadebbfa88bf139ddca501ab158f3746c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object, and sorts its markers, the rows of its parent gen <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (because they are ordered by the markers), and its genetic map so that the markers are ordered by chromosome number then position.  <a href="group__supporters.html#gadebbfa88bf139ddca501ab158f3746c5">More...</a><br /></td></tr>
<tr class="separator:gadebbfa88bf139ddca501ab158f3746c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5e049945be5700a6d5a5654b6b332842"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__supporters.html#ga5e049945be5700a6d5a5654b6b332842">get_chromosome_locations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga5e049945be5700a6d5a5654b6b332842"><td class="mdescLeft">&#160;</td><td class="mdescRight">Updates the chr_ends, n_chr and chr_lengths fields in <a class="el" href="struct_sim_data.html#ae1ae6d192937ca2edaba69edf8308711" title="A GeneticMap.">SimData.map</a>.  <a href="group__supporters.html#ga5e049945be5700a6d5a5654b6b332842">More...</a><br /></td></tr>
<tr class="separator:ga5e049945be5700a6d5a5654b6b332842"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga29f27b356c8ce89107314097868532d3"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_effect_i_d.html">EffectID</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga29f27b356c8ce89107314097868532d3">load_effects_to_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *filename)</td></tr>
<tr class="memdesc:ga29f27b356c8ce89107314097868532d3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination with effect values.  <a href="group__loaders.html#ga29f27b356c8ce89107314097868532d3">More...</a><br /></td></tr>
<tr class="separator:ga29f27b356c8ce89107314097868532d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3d8878f2c2c5fafdae57ffc46a78dfbc"><td class="memItemLeft" align="right" valign="top">struct <a class="el" href="struct_group_and_effect_set.html">GroupAndEffectSet</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__loaders.html#ga3d8878f2c2c5fafdae57ffc46a78dfbc">load_all_simdata</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *data_file, const char *map_file, const char *effect_file)</td></tr>
<tr class="memdesc:ga3d8878f2c2c5fafdae57ffc46a78dfbc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Populates a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> combination from scratch with marker allele data, a genetic map, and effect values.  <a href="group__loaders.html#ga3d8878f2c2c5fafdae57ffc46a78dfbc">More...</a><br /></td></tr>
<tr class="separator:ga3d8878f2c2c5fafdae57ffc46a78dfbc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#gabefce26ab485e6886d7bbfcaa5692f3a">calculate_min_recombinations_fw1</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int certain)</td></tr>
<tr class="memdesc:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured.  <a href="group__recomb.html#gabefce26ab485e6886d7bbfcaa5692f3a">More...</a><br /></td></tr>
<tr class="separator:gabefce26ab485e6886d7bbfcaa5692f3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9c9c066183d70aa46d4ed4d224116580"><td class="memItemLeft" align="right" valign="top">int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga9c9c066183d70aa46d4ed4d224116580">calculate_min_recombinations_fwn</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, char *parent1, unsigned int p1num, char *parent2, unsigned int p2num, char *offspring, int window_size, int certain)</td></tr>
<tr class="memdesc:ga9c9c066183d70aa46d4ed4d224116580"><td class="mdescLeft">&#160;</td><td class="mdescRight">Identify markers in the genotype of <code>offspring</code> where recombination from its parents occured, as judged by the marker itself and a short window around it.  <a href="group__recomb.html#ga9c9c066183d70aa46d4ed4d224116580">More...</a><br /></td></tr>
<tr class="separator:ga9c9c066183d70aa46d4ed4d224116580"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">calculate_recombinations_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const char *output_file, int window_len, int certain)</td></tr>
<tr class="memdesc:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Provides guesses as to the location of recombination events that led to the creation of certain genotypes from certain other genotypes.  <a href="group__recomb.html#ga7813f256407ab2ea242df32ea9ad2c7e">More...</a><br /></td></tr>
<tr class="separator:ga7813f256407ab2ea242df32ea9ad2c7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga90f7b0635f358ca029d16b996e203cbf"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga90f7b0635f358ca029d16b996e203cbf">generate_gamete</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ga90f7b0635f358ca029d16b996e203cbf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fills a char* with the simulated result of meiosis (reduction and recombination) from the marker alleles of a given parent.  <a href="group__meiosis.html#ga90f7b0635f358ca029d16b996e203cbf">More...</a><br /></td></tr>
<tr class="separator:ga90f7b0635f358ca029d16b996e203cbf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafc34eb6a9bf8128a2619ffd322db8689"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#gafc34eb6a9bf8128a2619ffd322db8689">generate_cross</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *parent1_genome, const char *parent2_genome, char *output)</td></tr>
<tr class="memdesc:gafc34eb6a9bf8128a2619ffd322db8689"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of crossing two genotypes.  <a href="group__meiosis.html#gafc34eb6a9bf8128a2619ffd322db8689">More...</a><br /></td></tr>
<tr class="separator:gafc34eb6a9bf8128a2619ffd322db8689"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga795bfb971a6c406f5f06c14467fd2287"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga795bfb971a6c406f5f06c14467fd2287">generate_doubled_haploid</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ga795bfb971a6c406f5f06c14467fd2287"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the alleles of the outcome of producing a doubled haploid from a gamete from a given parent.  <a href="group__meiosis.html#ga795bfb971a6c406f5f06c14467fd2287">More...</a><br /></td></tr>
<tr class="separator:ga795bfb971a6c406f5f06c14467fd2287"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga020af863f40a1ce6be2d26cc36af1848"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__meiosis.html#ga020af863f40a1ce6be2d26cc36af1848">generate_clone</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *parent_genome, char *output)</td></tr>
<tr class="memdesc:ga020af863f40a1ce6be2d26cc36af1848"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get an identical copy of a given genotype.  <a href="group__meiosis.html#ga020af863f40a1ce6be2d26cc36af1848">More...</a><br /></td></tr>
<tr class="separator:ga020af863f40a1ce6be2d26cc36af1848"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga823b906485f3edf2771935f2ad4f12db"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga823b906485f3edf2771935f2ad4f12db">cross_random_individuals</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> from_group, const int n_crosses, const int cap, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga823b906485f3edf2771935f2ad4f12db"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses among members of a group.  <a href="group__crossers.html#ga823b906485f3edf2771935f2ad4f12db">More...</a><br /></td></tr>
<tr class="separator:ga823b906485f3edf2771935f2ad4f12db"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga21fd20fa9a626c2933b7ec114c3faca5"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga21fd20fa9a626c2933b7ec114c3faca5">cross_randomly_between</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group1, const <a class="el" href="struct_group_num.html">GroupNum</a> group2, const int n_crosses, const int cap1, const int cap2, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga21fd20fa9a626c2933b7ec114c3faca5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs random crosses where the first parent comes from one group and the second from another.  <a href="group__crossers.html#ga21fd20fa9a626c2933b7ec114c3faca5">More...</a><br /></td></tr>
<tr class="separator:ga21fd20fa9a626c2933b7ec114c3faca5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a877bbfdb03961d00c30f7e2fa55efe7c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="sim-operations_8c.html#a877bbfdb03961d00c30f7e2fa55efe7c">cross_these_combinations</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int n_combinations, const int firstParents[n_combinations], const int secondParents[n_combinations], const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:a877bbfdb03961d00c30f7e2fa55efe7c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Performs the crosses of pairs of parents whose indexes are provided in an array.  <a href="sim-operations_8c.html#a877bbfdb03961d00c30f7e2fa55efe7c">More...</a><br /></td></tr>
<tr class="separator:a877bbfdb03961d00c30f7e2fa55efe7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaafc0d183fdb9762dc0817e5bd3009178"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gaafc0d183fdb9762dc0817e5bd3009178">self_n_times</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int n, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gaafc0d183fdb9762dc0817e5bd3009178"><td class="mdescLeft">&#160;</td><td class="mdescRight">Selfs each member of a group for a certain number of generations.  <a href="group__crossers.html#gaafc0d183fdb9762dc0817e5bd3009178">More...</a><br /></td></tr>
<tr class="separator:gaafc0d183fdb9762dc0817e5bd3009178"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6a3afc7743003c66fe24e7a2872e26cc"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga6a3afc7743003c66fe24e7a2872e26cc">make_doubled_haploids</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga6a3afc7743003c66fe24e7a2872e26cc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates a doubled haploid from each member of a group.  <a href="group__crossers.html#ga6a3afc7743003c66fe24e7a2872e26cc">More...</a><br /></td></tr>
<tr class="separator:ga6a3afc7743003c66fe24e7a2872e26cc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4eb395614f3ec7ffecb3b779c16e6f17"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga4eb395614f3ec7ffecb3b779c16e6f17">make_clones</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const int inherit_names, <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga4eb395614f3ec7ffecb3b779c16e6f17"><td class="mdescLeft">&#160;</td><td class="mdescRight">Creates an identical copy of each member of a group.  <a href="group__crossers.html#ga4eb395614f3ec7ffecb3b779c16e6f17">More...</a><br /></td></tr>
<tr class="separator:ga4eb395614f3ec7ffecb3b779c16e6f17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab173426b15ba3e135e1eea25035afd85"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gab173426b15ba3e135e1eea25035afd85">make_all_unidirectional_crosses</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> from_group, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gab173426b15ba3e135e1eea25035afd85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between all pairs of parents in the group <code>from_group</code> and allocates the resulting offspring to a new group.  <a href="group__crossers.html#gab173426b15ba3e135e1eea25035afd85">More...</a><br /></td></tr>
<tr class="separator:gab173426b15ba3e135e1eea25035afd85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafdd1fa0e8dd54ba1f3089c05986686cb"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#gafdd1fa0e8dd54ba1f3089c05986686cb">make_n_crosses_from_top_m_percent</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const int n, const int m, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:gafdd1fa0e8dd54ba1f3089c05986686cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Find the top m percent of a group and perform random crosses between those top individuals.  <a href="group__crossers.html#gafdd1fa0e8dd54ba1f3089c05986686cb">More...</a><br /></td></tr>
<tr class="separator:gafdd1fa0e8dd54ba1f3089c05986686cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga18a22fc990289359e790fa207613c121"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga18a22fc990289359e790fa207613c121">make_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga18a22fc990289359e790fa207613c121"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between pairs of parents identified by name in a file and allocate the resulting offspring to a new group.  <a href="group__crossers.html#ga18a22fc990289359e790fa207613c121">More...</a><br /></td></tr>
<tr class="separator:ga18a22fc990289359e790fa207613c121"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0cc412ee0f607370ad3b34f4c39a15ff"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__crossers.html#ga0cc412ee0f607370ad3b34f4c39a15ff">make_double_crosses_from_file</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const char *input_file, const <a class="el" href="struct_gen_options.html">GenOptions</a> g)</td></tr>
<tr class="memdesc:ga0cc412ee0f607370ad3b34f4c39a15ff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform crosses between previously-generated offspring of pairs of parents identified by name in a file.  <a href="group__crossers.html#ga0cc412ee0f607370ad3b34f4c39a15ff">More...</a><br /></td></tr>
<tr class="separator:ga0cc412ee0f607370ad3b34f4c39a15ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga88e086f459bf255ded35aff26f976ab8"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_group_num.html">GroupNum</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga88e086f459bf255ded35aff26f976ab8">split_by_bv</a> (<a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID, const int top_n, const int lowIsBest)</td></tr>
<tr class="memdesc:ga88e086f459bf255ded35aff26f976ab8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes the <code>top_n</code> individuals in the group with the best breeding values/fitnesses and puts them in a new group.  <a href="group__calculators.html#ga88e086f459bf255ded35aff26f976ab8">More...</a><br /></td></tr>
<tr class="separator:ga88e086f459bf255ded35aff26f976ab8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafd3faffcf504432504b50b669055f46b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gafd3faffcf504432504b50b669055f46b">calculate_group_bvs</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gafd3faffcf504432504b50b669055f46b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> in a certain group, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a>.  <a href="group__calculators.html#gafd3faffcf504432504b50b669055f46b">More...</a><br /></td></tr>
<tr class="separator:gafd3faffcf504432504b50b669055f46b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga75f4e1199f9ead752b64cd5fa3b5a1a7"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga75f4e1199f9ead752b64cd5fa3b5a1a7">calculate_bvs</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const <a class="el" href="struct_effect_matrix.html">EffectMatrix</a> *e)</td></tr>
<tr class="memdesc:ga75f4e1199f9ead752b64cd5fa3b5a1a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the fitness metric/breeding value for each genotype in the <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>, and returns the results in a <a class="el" href="struct_decimal_matrix.html" title="A row-major heap matrix that contains floating point numbers.">DecimalMatrix</a> struct.  <a href="group__calculators.html#ga75f4e1199f9ead752b64cd5fa3b5a1a7">More...</a><br /></td></tr>
<tr class="separator:ga75f4e1199f9ead752b64cd5fa3b5a1a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gace12e242f730dd2b7747919b955a5035"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gace12e242f730dd2b7747919b955a5035">calculate_group_count_matrix_of_allele</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:gace12e242f730dd2b7747919b955a5035"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a group.  <a href="group__calculators.html#gace12e242f730dd2b7747919b955a5035">More...</a><br /></td></tr>
<tr class="separator:gace12e242f730dd2b7747919b955a5035"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gace545bfe1524e08d158793bd738cdf5f"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gace545bfe1524e08d158793bd738cdf5f">calculate_group_doublecount_matrix_of_allele</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts, const char allele2, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:gace545bfe1524e08d158793bd738cdf5f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear for each genotype in a group.  <a href="group__calculators.html#gace545bfe1524e08d158793bd738cdf5f">More...</a><br /></td></tr>
<tr class="separator:gace545bfe1524e08d158793bd738cdf5f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae883ac8145dd7e45ada82cc878d405d7"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae883ac8145dd7e45ada82cc878d405d7">calculate_count_matrix_of_allele</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts)</td></tr>
<tr class="memdesc:gae883ac8145dd7e45ada82cc878d405d7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in a particular <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (does not go through the linked list).  <a href="group__calculators.html#gae883ac8145dd7e45ada82cc878d405d7">More...</a><br /></td></tr>
<tr class="separator:gae883ac8145dd7e45ada82cc878d405d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga11a8e25ed696aaf04d27b2276e345fe6"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga11a8e25ed696aaf04d27b2276e345fe6">calculate_doublecount_matrix_of_allele</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char allele, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts, const char allele2, <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *counts2)</td></tr>
<tr class="memdesc:ga11a8e25ed696aaf04d27b2276e345fe6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that two particular alleles appear for each genotype in an <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> (does not go through the linked list).  <a href="group__calculators.html#ga11a8e25ed696aaf04d27b2276e345fe6">More...</a><br /></td></tr>
<tr class="separator:ga11a8e25ed696aaf04d27b2276e345fe6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae9e20b885533620494187b26e7d44d48"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gae9e20b885533620494187b26e7d44d48">calculate_full_count_matrix_of_allele</a> (const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char allele)</td></tr>
<tr class="memdesc:gae9e20b885533620494187b26e7d44d48"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the number of times at each marker that a particular allele appears for each genotype in the linked list starting at the given <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a>.  <a href="group__calculators.html#gae9e20b885533620494187b26e7d44d48">More...</a><br /></td></tr>
<tr class="separator:gae9e20b885533620494187b26e7d44d48"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa7adc67211c9168dc29e9a4c1219c8f0"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaa7adc67211c9168dc29e9a4c1219c8f0">create_n_blocks_by_chr</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const int n)</td></tr>
<tr class="memdesc:gaa7adc67211c9168dc29e9a4c1219c8f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Divide the genotype into blocks where each block contains all markers within a 1/n length section of each chromosome in the map, and return the resulting blocks in a struct.  <a href="group__calculators.html#gaa7adc67211c9168dc29e9a4c1219c8f0">More...</a><br /></td></tr>
<tr class="separator:gaa7adc67211c9168dc29e9a4c1219c8f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga43ce1b960fe8dce2376319c20ecc0561"><td class="memItemLeft" align="right" valign="top"><a class="el" href="struct_marker_blocks.html">MarkerBlocks</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga43ce1b960fe8dce2376319c20ecc0561">read_block_file</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const char *block_file)</td></tr>
<tr class="memdesc:ga43ce1b960fe8dce2376319c20ecc0561"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a file containing definitions of blocks of markers, process that file and return a struct containing the definitions of those blocks.  <a href="group__calculators.html#ga43ce1b960fe8dce2376319c20ecc0561">More...</a><br /></td></tr>
<tr class="separator:ga43ce1b960fe8dce2376319c20ecc0561"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2ec60edb767c3b11c4ef6cfa9d2df8ee"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga2ec60edb767c3b11c4ef6cfa9d2df8ee">calculate_group_local_bvs</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID, const char *output_file, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:ga2ec60edb767c3b11c4ef6cfa9d2df8ee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype in a group, calculate the local fitness metric/breeding value for the first allele at each marker in the block, and the local fitness metric/breeding value for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga2ec60edb767c3b11c4ef6cfa9d2df8ee">More...</a><br /></td></tr>
<tr class="separator:ga2ec60edb767c3b11c4ef6cfa9d2df8ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1fb48abb9d45fab973a4a071f5f52b7e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#ga1fb48abb9d45fab973a4a071f5f52b7e">calculate_local_bvs</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID, const char *output_file)</td></tr>
<tr class="memdesc:ga1fb48abb9d45fab973a4a071f5f52b7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Given a set of blocks of markers in a file, for each genotype saved, calculate the local BV for the first allele at each marker in the block, and the local BV for the second allele at each marker in the block, then save the result to a file.  <a href="group__calculators.html#ga1fb48abb9d45fab973a4a071f5f52b7e">More...</a><br /></td></tr>
<tr class="separator:ga1fb48abb9d45fab973a4a071f5f52b7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf4094eb1d8fd4baec5a35bb40d88bcce"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaf4094eb1d8fd4baec5a35bb40d88bcce">calculate_optimal_alleles</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gaf4094eb1d8fd4baec5a35bb40d88bcce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and creates a string containing the allele with the highest effect value for each marker as ordered in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a>.  <a href="group__calculators.html#gaf4094eb1d8fd4baec5a35bb40d88bcce">More...</a><br /></td></tr>
<tr class="separator:gaf4094eb1d8fd4baec5a35bb40d88bcce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaccf07bc53314ff8ec4918ab2bbf07368"><td class="memItemLeft" align="right" valign="top">char *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaccf07bc53314ff8ec4918ab2bbf07368">calculate_optimal_available_alleles</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gaccf07bc53314ff8ec4918ab2bbf07368"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the highest-breeding-value haplotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#gaccf07bc53314ff8ec4918ab2bbf07368">More...</a><br /></td></tr>
<tr class="separator:gaccf07bc53314ff8ec4918ab2bbf07368"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa5f6c9539772c5086be1bfb023274d65"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaa5f6c9539772c5086be1bfb023274d65">calculate_optimum_bv</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gaa5f6c9539772c5086be1bfb023274d65"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the highest possible breeding value any (diploid) genotype could have using those effect values.  <a href="group__calculators.html#gaa5f6c9539772c5086be1bfb023274d65">More...</a><br /></td></tr>
<tr class="separator:gaa5f6c9539772c5086be1bfb023274d65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaecfeb3d0da1b66583e4f7bc7f9646b53"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gaecfeb3d0da1b66583e4f7bc7f9646b53">calculate_optimal_available_bv</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gaecfeb3d0da1b66583e4f7bc7f9646b53"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculates the breeding value of the highest breeding-value genotype that can be created from the alleles present in a given group.  <a href="group__calculators.html#gaecfeb3d0da1b66583e4f7bc7f9646b53">More...</a><br /></td></tr>
<tr class="separator:gaecfeb3d0da1b66583e4f7bc7f9646b53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gafb63ee57aa293e07cdd0cb14bf255348"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__calculators.html#gafb63ee57aa293e07cdd0cb14bf255348">calculate_minimum_bv</a> (const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gafb63ee57aa293e07cdd0cb14bf255348"><td class="mdescLeft">&#160;</td><td class="mdescRight">Takes a look at the currently-loaded effect values and returns the lowest possible breeding value any (diploid) genotype could score using those effect values.  <a href="group__calculators.html#gafb63ee57aa293e07cdd0cb14bf255348">More...</a><br /></td></tr>
<tr class="separator:gafb63ee57aa293e07cdd0cb14bf255348"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5c1144f08e16be0974f7a8f81b81c0d4">save_simdata</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *m)</td></tr>
<tr class="memdesc:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the setup data (everything except the actual genotypes) stored inside a <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga5c1144f08e16be0974f7a8f81b81c0d4">More...</a><br /></td></tr>
<tr class="separator:ga5c1144f08e16be0974f7a8f81b81c0d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga63e803fa3cdf0cbb209b2922ce2982fe">save_marker_blocks</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_marker_blocks.html">MarkerBlocks</a> b)</td></tr>
<tr class="memdesc:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the markers contained in a set of blocks to a file.  <a href="group__savers.html#ga63e803fa3cdf0cbb209b2922ce2982fe">More...</a><br /></td></tr>
<tr class="separator:ga63e803fa3cdf0cbb209b2922ce2982fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5b6c6c66af73fea9f0545b826d899a70"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5b6c6c66af73fea9f0545b826d899a70">save_names_header</a> (FILE *f, size_t n, const char *names[n])</td></tr>
<tr class="memdesc:ga5b6c6c66af73fea9f0545b826d899a70"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints a list of names as a tab-separated row to a file.  <a href="group__savers.html#ga5b6c6c66af73fea9f0545b826d899a70">More...</a><br /></td></tr>
<tr class="separator:ga5b6c6c66af73fea9f0545b826d899a70"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf147e82611c7d5d3cf3c5b8b27c4fb07"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaf147e82611c7d5d3cf3c5b8b27c4fb07">save_allele_matrix</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m)</td></tr>
<tr class="memdesc:gaf147e82611c7d5d3cf3c5b8b27c4fb07"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the genotype data saved in the linked list of AlleleMatrices starting with <code>m</code> to a file.  <a href="group__savers.html#gaf147e82611c7d5d3cf3c5b8b27c4fb07">More...</a><br /></td></tr>
<tr class="separator:gaf147e82611c7d5d3cf3c5b8b27c4fb07"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga8e14b8d8c1836e1402a4a5e7a74c274f">save_transposed_allele_matrix</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const char **markers)</td></tr>
<tr class="memdesc:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints all the gene data saved in the linked list starting with <code>m</code> to the file.  <a href="group__savers.html#ga8e14b8d8c1836e1402a4a5e7a74c274f">More...</a><br /></td></tr>
<tr class="separator:ga8e14b8d8c1836e1402a4a5e7a74c274f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae65797e9f37cac9cff90abadd8c219c5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gae65797e9f37cac9cff90abadd8c219c5">save_group_alleles</a> (FILE *f, <a class="el" href="struct_sim_data.html">SimData</a> *d, <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gae65797e9f37cac9cff90abadd8c219c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#gae65797e9f37cac9cff90abadd8c219c5">More...</a><br /></td></tr>
<tr class="separator:gae65797e9f37cac9cff90abadd8c219c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf856013dc9108a075ed5c1145487285"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gaaf856013dc9108a075ed5c1145487285">save_transposed_group_alleles</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group_id)</td></tr>
<tr class="memdesc:gaaf856013dc9108a075ed5c1145487285"><td class="mdescLeft">&#160;</td><td class="mdescRight">Prints the genotypes of each individual in a given group to a file, with the following format.  <a href="group__savers.html#gaaf856013dc9108a075ed5c1145487285">More...</a><br /></td></tr>
<tr class="separator:gaaf856013dc9108a075ed5c1145487285"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8ebf516e290bc4fd0e171a6d621efc2b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga8ebf516e290bc4fd0e171a6d621efc2b">save_group_one_step_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:ga8ebf516e290bc4fd0e171a6d621efc2b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in a group to a file.  <a href="group__savers.html#ga8ebf516e290bc4fd0e171a6d621efc2b">More...</a><br /></td></tr>
<tr class="separator:ga8ebf516e290bc4fd0e171a6d621efc2b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5674dd2fbdcf205475c8036a09536f2e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5674dd2fbdcf205475c8036a09536f2e">save_one_step_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga5674dd2fbdcf205475c8036a09536f2e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the parents of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga5674dd2fbdcf205475c8036a09536f2e">More...</a><br /></td></tr>
<tr class="separator:ga5674dd2fbdcf205475c8036a09536f2e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga821e9489b1668201f9bd38d9ac467673"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga821e9489b1668201f9bd38d9ac467673">save_group_full_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:ga821e9489b1668201f9bd38d9ac467673"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a group to a file.  <a href="group__savers.html#ga821e9489b1668201f9bd38d9ac467673">More...</a><br /></td></tr>
<tr class="separator:ga821e9489b1668201f9bd38d9ac467673"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga08c823d63ce4c648432cb740b5ab743b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga08c823d63ce4c648432cb740b5ab743b">save_full_pedigree</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d)</td></tr>
<tr class="memdesc:ga08c823d63ce4c648432cb740b5ab743b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga08c823d63ce4c648432cb740b5ab743b">More...</a><br /></td></tr>
<tr class="separator:ga08c823d63ce4c648432cb740b5ab743b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf1a9b82d2978a1563daf31c8814a783"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783">save_AM_pedigree</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, const <a class="el" href="struct_sim_data.html">SimData</a> *parents)</td></tr>
<tr class="memdesc:gabf1a9b82d2978a1563daf31c8814a783"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the full known pedigree of each genotype in a single <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> to a file.  <a href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783">More...</a><br /></td></tr>
<tr class="separator:gabf1a9b82d2978a1563daf31c8814a783"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2ad03e903efdd56805722af554328592"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga2ad03e903efdd56805722af554328592">save_parents_of</a> (FILE *f, const <a class="el" href="struct_allele_matrix.html">AlleleMatrix</a> *m, <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> p1, <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> p2)</td></tr>
<tr class="memdesc:ga2ad03e903efdd56805722af554328592"><td class="mdescLeft">&#160;</td><td class="mdescRight">Recursively save the parents of a particular id to a file.  <a href="group__savers.html#ga2ad03e903efdd56805722af554328592">More...</a><br /></td></tr>
<tr class="separator:ga2ad03e903efdd56805722af554328592"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadb547ac914ca6f84033394402ffece7e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#gadb547ac914ca6f84033394402ffece7e">save_group_bvs</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_group_num.html">GroupNum</a> group, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:gadb547ac914ca6f84033394402ffece7e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in a group to a file.  <a href="group__savers.html#gadb547ac914ca6f84033394402ffece7e">More...</a><br /></td></tr>
<tr class="separator:gadb547ac914ca6f84033394402ffece7e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9ff67eb9572c42530b64b70f4a307488"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga9ff67eb9572c42530b64b70f4a307488">save_bvs</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const <a class="el" href="struct_effect_i_d.html">EffectID</a> effID)</td></tr>
<tr class="memdesc:ga9ff67eb9572c42530b64b70f4a307488"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the breeding value of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga9ff67eb9572c42530b64b70f4a307488">More...</a><br /></td></tr>
<tr class="separator:ga9ff67eb9572c42530b64b70f4a307488"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4950114dda9c59fa276b72b13823d772"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga4950114dda9c59fa276b72b13823d772">save_manual_bvs</a> (FILE *f, const <a class="el" href="struct_decimal_matrix.html">DecimalMatrix</a> *e, const <a class="el" href="struct_pedigree_i_d.html">PedigreeID</a> *ids, const char **names)</td></tr>
<tr class="memdesc:ga4950114dda9c59fa276b72b13823d772"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the provided breeding values of each provided name and id to a file, with the same format as a regular call to <code>save_bvs</code> or <code>save_group_bvs</code>.  <a href="group__savers.html#ga4950114dda9c59fa276b72b13823d772">More...</a><br /></td></tr>
<tr class="separator:ga4950114dda9c59fa276b72b13823d772"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga38b0958216f509a3b0a10c4c49183f44"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga38b0958216f509a3b0a10c4c49183f44">save_count_matrix</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const char allele)</td></tr>
<tr class="memdesc:ga38b0958216f509a3b0a10c4c49183f44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> to a file.  <a href="group__savers.html#ga38b0958216f509a3b0a10c4c49183f44">More...</a><br /></td></tr>
<tr class="separator:ga38b0958216f509a3b0a10c4c49183f44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga5c490bff6698ad221c9eeec1254623c5"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__savers.html#ga5c490bff6698ad221c9eeec1254623c5">save_count_matrix_of_group</a> (FILE *f, const <a class="el" href="struct_sim_data.html">SimData</a> *d, const char allele, const <a class="el" href="struct_group_num.html">GroupNum</a> group)</td></tr>
<tr class="memdesc:ga5c490bff6698ad221c9eeec1254623c5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the number of copies of a particular allele at each marker of each genotype in a group to a file.  <a href="group__savers.html#ga5c490bff6698ad221c9eeec1254623c5">More...</a><br /></td></tr>
<tr class="separator:ga5c490bff6698ad221c9eeec1254623c5"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="var-members" name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:ga5c6b92aede8f29f53476aa78d4e4133a"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__structs.html#ga5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a></td></tr>
<tr class="memdesc:ga5c6b92aede8f29f53476aa78d4e4133a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Options parameter to run <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> functions in their bare-bones form.  <a href="group__structs.html#ga5c6b92aede8f29f53476aa78d4e4133a">More...</a><br /></td></tr>
<tr class="separator:ga5c6b92aede8f29f53476aa78d4e4133a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga694896423aa0294aaf5314e368a2abd4"><td class="memItemLeft" align="right" valign="top">const <a class="el" href="struct_geno_location.html">GenoLocation</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__iterators.html#ga694896423aa0294aaf5314e368a2abd4">INVALID_GENO_LOCATION</a></td></tr>
<tr class="memdesc:ga694896423aa0294aaf5314e368a2abd4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Constant representing a nonexistent location in the simulation.  <a href="group__iterators.html#ga694896423aa0294aaf5314e368a2abd4">More...</a><br /></td></tr>
<tr class="separator:ga694896423aa0294aaf5314e368a2abd4"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a id="ad84149faf0b815bcee066748ef31eb60" name="ad84149faf0b815bcee066748ef31eb60"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad84149faf0b815bcee066748ef31eb60">&#9670;&nbsp;</a></span>SIM_OPERATIONS</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define SIM_OPERATIONS</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="a60b92e4ec1dac6abca4fd6c09486c237" name="a60b92e4ec1dac6abca4fd6c09486c237"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a60b92e4ec1dac6abca4fd6c09486c237">&#9670;&nbsp;</a></span>_ascending_uint_dcomparer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int _ascending_uint_dcomparer </td>
          <td>(</td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp0</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const void *&#160;</td>
          <td class="paramname"><em>pp1</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Comparator function for qsort. </p>
<p >Used to compare an array of unsigned integers* to sort the integers pointed to in ascending order. </p><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__collgetters.html#ga5544a69d0f6abf525423479fc77d4805" title="Identify group numbers that currently has members and the number of genotypes currently allocated to ...">get_existing_group_counts()</a></dd></dl>
<p>Sorts lower numbers before higher numbers. If floats are equal, their order after comparison is undefined. </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a60b92e4ec1dac6abca4fd6c09486c237_icgraph.png" border="0" usemap="#asim-operations_8c_a60b92e4ec1dac6abca4fd6c09486c237_icgraph" alt=""/></div>
<map name="asim-operations_8c_a60b92e4ec1dac6abca4fd6c09486c237_icgraph" id="asim-operations_8c_a60b92e4ec1dac6abca4fd6c09486c237_icgraph">
<area shape="rect" title="Comparator function for qsort." alt="" coords="184,13,367,41"/>
<area shape="rect" href="group__collgetters.html#ga5544a69d0f6abf525423479fc77d4805" title="Identify group numbers that currently has members and the number of genotypes currently allocated to ..." alt="" coords="5,5,136,49"/>
</map>
</div>

</div>
</div>
<a id="a877bbfdb03961d00c30f7e2fa55efe7c" name="a877bbfdb03961d00c30f7e2fa55efe7c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a877bbfdb03961d00c30f7e2fa55efe7c">&#9670;&nbsp;</a></span>cross_these_combinations()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="struct_group_num.html">GroupNum</a> cross_these_combinations </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="struct_sim_data.html">SimData</a> *&#160;</td>
          <td class="paramname"><em>d</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>n_combinations</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>firstParents</em>[n_combinations], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int&#160;</td>
          <td class="paramname"><em>secondParents</em>[n_combinations], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="struct_gen_options.html">GenOptions</a>&#160;</td>
          <td class="paramname"><em>g</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Performs the crosses of pairs of parents whose indexes are provided in an array. </p>
<p >The resulting genotypes are allocated to a new group.</p>
<p >Preferences in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> are applied to this cross. The family_size parameter in <a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> allows you to repeat each particular cross a certain number of times.</p>
<p >Previously had a parameter combinations[2][n_combinations] instead of firstParents and secondParents. This was changed to lower the boilerplate needs of calling this function: now there is no need to create a 2-wide int* to hold the two separate parent vectors if they already exist.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">d</td><td>pointer to the <a class="el" href="struct_sim_data.html" title="Composite type that is used to run crossing simulations.">SimData</a> object that includes genetic map data and allele data needed to simulate crossing. </td></tr>
    <tr><td class="paramname">n_combinations</td><td>the number of pairs of ids to cross/the length of <code>combinations</code> </td></tr>
    <tr><td class="paramname">firstParents</td><td>a vector of indexes of parents to be the first parent of each cross </td></tr>
    <tr><td class="paramname">secondParents</td><td>a vector of indexes of parents to be the second parent of each cross. firstParents[0] is crossed to secondParents[0], firstParents[1] is crossed to secondParents[1], and so forth. </td></tr>
    <tr><td class="paramname">g</td><td>options for the genotypes created. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="struct_gen_options.html" title="A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...">GenOptions</a> </dd></dl>
<dl class="section return"><dt>Returns</dt><dd>the group number of the group to which the produced offspring were allocated. </dd></dl>
<div id="dynsection-1" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-1-trigger" src="closed.png" alt="+"/> Here is the call graph for this function:</div>
<div id="dynsection-1-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-1-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_cgraph.png" border="0" usemap="#asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_cgraph" alt=""/></div>
<map name="asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_cgraph" id="asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_cgraph">
<area shape="rect" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="5,641,182,669"/>
<area shape="rect" href="group__calculators.html#ga75f4e1199f9ead752b64cd5fa3b5a1a7" title="Calculates the fitness metric/breeding value for each genotype in the AlleleMatrix,..." alt="" coords="267,209,366,237"/>
<area shape="rect" href="group__deletors.html#gaaf6542d59034f88f9e1ec48e5f053c4a" title="Deletes a DecimalMatrix and frees its memory." alt="" coords="488,329,594,357"/>
<area shape="rect" href="group__supporters.html#ga887027349503be24b1a847c9b833aaac" title="A function to tidy the internal storage of genotypes after addition or deletion of genotypes in the S..." alt="" coords="238,381,395,409"/>
<area shape="rect" href="group__deletors.html#gadd4a9c17e8c3e59fc00eee25635b110f" title="Delete the AlleleMatrix linked list from m onwards and frees its memory." alt="" coords="473,381,609,409"/>
<area shape="rect" href="group__loaders.html#gaf0399b852faa346adeab52d57fa9701b" title="Creator for an empty AlleleMatrix object of a given size." alt="" coords="230,537,403,565"/>
<area shape="rect" href="group__meiosis.html#gafc34eb6a9bf8128a2619ffd322db8689" title="Get the alleles of the outcome of crossing two genotypes." alt="" coords="261,485,372,513"/>
<area shape="rect" href="group__search.html#ga31eb9ab2cd17d7c107102922e2609aac" title="Get the alleles of a genotype by its index." alt="" coords="248,693,385,721"/>
<area shape="rect" href="group__search.html#ga8c2bc00d2b81a4c293a939f202250eb8" title="Get the id of a genotype by its index." alt="" coords="261,745,372,773"/>
<area shape="rect" href="group__supporters.html#gaa605100835fdec6b250fde0ad5a9cc8f" title="Function to identify the lookup index of a marker effect set identifier." alt="" coords="247,797,386,825"/>
<area shape="rect" href="group__supporters.html#gaf0bc58eab8b8e413c1b20140fab8d17f" title="Function to identify the next sequential integer that does not identify a group that currently has me..." alt="" coords="246,589,387,617"/>
<area shape="rect" href="group__savers.html#gaf147e82611c7d5d3cf3c5b8b27c4fb07" title="Prints all the genotype data saved in the linked list of AlleleMatrices starting with m to a file." alt="" coords="253,849,380,877"/>
<area shape="rect" href="group__savers.html#gabf1a9b82d2978a1563daf31c8814a783" title="Print the full known pedigree of each genotype in a single AlleleMatrix to a file." alt="" coords="251,901,382,929"/>
<area shape="rect" href="group__savers.html#ga4950114dda9c59fa276b72b13823d772" title="Print the provided breeding values of each provided name and id to a file, with the same format as a ..." alt="" coords="255,953,378,981"/>
<area shape="rect" href="group__savers.html#ga5b6c6c66af73fea9f0545b826d899a70" title="Prints a list of names as a tab&#45;separated row to a file." alt="" coords="247,1005,386,1033"/>
<area shape="rect" href="group__supporters.html#gafad0846098c4da7ca876f0a625f369e1" title="Fills the designated section of the .names array in an AlleleMatrix with the pattern prefixindex." alt="" coords="274,641,359,669"/>
<area shape="rect" href="group__maths.html#gab1f0bafd56d3496348ea3ccddfe9d937" title="Multiply two sets of a DecimalMatrix and vector, and add both products to the first column of a provi..." alt="" coords="462,261,619,305"/>
<area shape="rect" href="group__maths.html#ga029101c5b02d8ab22e74ce3c103f6114" title="Multiply a DecimalMatrix to a vector, and add that product to the first column of a provided DecimalM..." alt="" coords="456,5,625,49"/>
<area shape="rect" href="group__calculators.html#gae883ac8145dd7e45ada82cc878d405d7" title="Calculates the number of times at each marker that a particular allele appears for each genotype in a..." alt="" coords="463,73,618,117"/>
<area shape="rect" href="group__calculators.html#ga11a8e25ed696aaf04d27b2276e345fe6" title="Calculates the number of times at each marker that two particular alleles appear for each genotype in..." alt="" coords="466,141,616,185"/>
<area shape="rect" href="group__maths.html#ga26e7ca69aa8bd969e3a781331a0afe27" title="Generates a matrix of c columns, r rows with all 0." alt="" coords="464,209,617,237"/>
<area shape="rect" href="group__supporters.html#ga6072a152b624743eeafb30adaf9b7f17" title="Replace calls to malloc direct with this function, which errors and exits with status 2 if memory all..." alt="" coords="499,537,583,565"/>
<area shape="rect" href="group__supporters.html#ga3744bc6f3fcb52952d520029055e8af8" title="Comparator function for qsort." alt="" coords="451,485,630,513"/>
<area shape="rect" href="group__maths.html#gac0aed93b85c3ff464c1e29bd2093e066" title="Generates randomly a number from the Poisson distribution with parameter lambda, using the Knuth appr..." alt="" coords="509,433,572,461"/>
<area shape="rect" href="group__collgetters.html#gacadb6fd4167febae0fcd3f653ee8a534" title="Identify group numbers that currently have members." alt="" coords="472,589,609,617"/>
<area shape="rect" href="group__supporters.html#gadae2e9f7e1061cb20426ba2333b7629b" title="Comparator function for qsort." alt="" coords="678,589,848,617"/>
<area shape="rect" href="group__savers.html#ga2ad03e903efdd56805722af554328592" title="Recursively save the parents of a particular id to a file." alt="" coords="484,901,598,929"/>
<area shape="rect" href="group__search.html#gaaea63100f4c2639907cce72c29a413e8" title="Returns the name of the genotype with a given id." alt="" coords="707,875,819,903"/>
<area shape="rect" href="group__search.html#gaa814762960a964cd0aaf82be0c18b8cb" title="Saves the ids of the parents of a genotype with a particular id to the output array output." alt="" coords="701,927,825,955"/>
<area shape="rect" href="group__supporters.html#ga4e3d1fd7ae67dd133c4dbdec9a84d455" title="Returns the located index in an array of PedigreeIDs where the PedigreeID is target." alt="" coords="896,901,1097,929"/>
<area shape="rect" href="group__supporters.html#ga4467dd870605266ac95eff75bacffb1f" title="Count and return the number of digits in i." alt="" coords="479,641,602,669"/>
</map>
</div>
<div id="dynsection-2" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-2-trigger" src="closed.png" alt="+"/> Here is the caller graph for this function:</div>
<div id="dynsection-2-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-2-content" class="dyncontent" style="display:none;">
<div class="center"><img src="sim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_icgraph.png" border="0" usemap="#asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_icgraph" alt=""/></div>
<map name="asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_icgraph" id="asim-operations_8c_a877bbfdb03961d00c30f7e2fa55efe7c_icgraph">
<area shape="rect" title="Performs the crosses of pairs of parents whose indexes are provided in an array." alt="" coords="215,73,391,101"/>
<area shape="rect" href="group__crossers.html#gab173426b15ba3e135e1eea25035afd85" title="Perform crosses between all pairs of parents in the group from_group and allocates the resulting offs..." alt="" coords="8,5,164,49"/>
<area shape="rect" href="group__crossers.html#ga18a22fc990289359e790fa207613c121" title="Perform crosses between pairs of parents identified by name in a file and allocate the resulting offs..." alt="" coords="5,73,167,101"/>
<area shape="rect" href="group__crossers.html#ga0cc412ee0f607370ad3b34f4c39a15ff" title="Perform crosses between previously&#45;generated offspring of pairs of parents identified by name in a fi..." alt="" coords="10,125,162,169"/>
</map>
</div>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
