<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Features and Templates Guide</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Features and Templates Guide </div></div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md6">Setting Up</a><ul><li class="level2"><a href="#autotoc_md7">Load a genetic map and a set of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md8">Load a genetic map and several sets of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md9">Load a genetic map, a set of founder genotypes, and a set of additive trait effects</a></li>
<li class="level2"><a href="#autotoc_md10">Swap out the set of additive trait effects for another</a></li>
<li class="level2"><a href="#autotoc_md11">Input file formats:</a><ul><li class="level3"><a href="#autotoc_md12">Loading from other file formats</a></li>
</ul>
</li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md13">Crossing &amp; Other Ways to Generate New Genotypes: Plant-themed</a><ul><li class="level2"><a href="#autotoc_md14">Single seed descent</a></li>
<li class="level2"><a href="#autotoc_md15">Creating halfsib or fullsib families</a></li>
<li class="level2"><a href="#autotoc_md16">Updating marker effects</a></li>
<li class="level2"><a href="#autotoc_md17">Trials and locations</a></li>
<li class="level2"><a href="#autotoc_md18">Backcrossing</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md19">Crossing &amp; Other Ways to Generate New Genotypes: Animal-themed</a><ul><li class="level2"><a href="#autotoc_md20">Split offspring into male and female</a></li>
<li class="level2"><a href="#autotoc_md21">Introducing fresh blood to male and female kernels</a></li>
<li class="level2"><a href="#autotoc_md22">Random mating with caps on number of offspring per animal</a></li>
<li class="level2"><a href="#autotoc_md23">Mating all females to a good male</a></li>
<li class="level2"><a href="#autotoc_md24">Making specific chosen matings</a></li>
<li class="level2"><a href="#autotoc_md25">Tracking age and discarding individuals that are &#39;too old&#39;</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md26">Selection</a><ul><li class="level2"><a href="#autotoc_md27">Manually select on true breeding value</a></li>
<li class="level2"><a href="#autotoc_md28">Select on phenotype (simulated with a given heritability)</a></li>
<li class="level2"><a href="#autotoc_md29">Select on a qualitative trait</a></li>
<li class="level2"><a href="#autotoc_md30">Select on a qualitative trait, then a quantitative trait.</a></li>
<li class="level2"><a href="#autotoc_md31">Select on an index weighting two traits.</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p ><a class="anchor" id="md_Templates"></a></p>
<p >While the documentation pages of genomicSimulationC and its counterpart <a href="https://github.com/vllrs/genomicSimulation">genomicSimulation</a> do describe all the features the simulation tool offers, they don't do so in a way that is necessarily easy to browse or that's helpful for new users to understand the simulation tool.</p>
<p >This page provides a set of templates for implementing common breeding project actions in genomicSimulationC and its R counterpart.</p>
<h1><a class="anchor" id="autotoc_md6"></a>
Setting Up</h1>
<h2><a class="anchor" id="autotoc_md7"></a>
Load a genetic map and a set of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_struct" href="struct_sim_data.html">SimData</a>* d = <a class="code hl_function" href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">create_empty_simdata</a>();</div>
<div class="line"><span class="keywordtype">int</span> founders = <a class="code hl_function" href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">load_all_simdata</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL);</div>
<div class="ttc" id="agroup__loaders_html_ga54cd625669413c2ef171060e9844d814"><div class="ttname"><a href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">load_all_simdata</a></div><div class="ttdeci">int load_all_simdata(SimData *d, const char *data_file, const char *map_file, const char *effect_file)</div><div class="ttdoc">Populates a SimData combination from scratch with marker allele data, a genetic map,...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:3792</div></div>
<div class="ttc" id="agroup__loaders_html_gaf7350bdc690b7cb5592892858d2830a9"><div class="ttname"><a href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">create_empty_simdata</a></div><div class="ttdeci">SimData * create_empty_simdata()</div><div class="ttdoc">Creator for an empty SimData object on the heap.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:87</div></div>
<div class="ttc" id="astruct_sim_data_html"><div class="ttname"><a href="struct_sim_data.html">SimData</a></div><div class="ttdoc">Composite type that is used to run crossing simulations.</div><div class="ttdef"><b>Definition:</b> sim-operations.h:224</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">founders &lt;- load.data(allele.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Consider using the get_group_ family of functions (genomicSimulationC) and see.group.data function (genomicSimulation) to confirm data is correctly loaded.</p>
<h2><a class="anchor" id="autotoc_md8"></a>
Load a genetic map and several sets of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05</div>
<div class="line">m1  TT  TT  TT  TA  TT</div>
<div class="line">m3  TT  TT  TA  TA  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Genotype file 2:</p><ul>
<li>File location: genotype-file2.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G06</div>
<div class="line">m1  AT</div>
<div class="line">m2  AA</div>
<div class="line">m3  TT</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_struct" href="struct_sim_data.html">SimData</a>* d = <a class="code hl_function" href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">create_empty_simdata</a>();</div>
<div class="line"><span class="keywordtype">int</span> founders_a = <a class="code hl_function" href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">load_all_simdata</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL);</div>
<div class="line"><span class="keywordtype">int</span> founders_b = <a class="code hl_function" href="group__loaders.html#ga8bd85cb47ec8e0c4049f937ef40ae102">load_more_transposed_genes_to_simdata</a>(d, <span class="stringliteral">&quot;genotype-file2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__loaders_html_ga8bd85cb47ec8e0c4049f937ef40ae102"><div class="ttname"><a href="group__loaders.html#ga8bd85cb47ec8e0c4049f937ef40ae102">load_more_transposed_genes_to_simdata</a></div><div class="ttdeci">int load_more_transposed_genes_to_simdata(SimData *d, const char *filename)</div><div class="ttdoc">Appends genotype data from a file to an existing SimData.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:3293</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">founders_a &lt;- load.data(allele.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;)</div>
<div class="line">founders_b &lt;- load.more.genotypes(&quot;genotype-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md9"></a>
Load a genetic map, a set of founder genotypes, and a set of additive trait effects</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Effect file:</p><ul>
<li>File location: eff-file.txt</li>
<li>File contents: (line order does not matter) <div class="fragment"><div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_struct" href="struct_sim_data.html">SimData</a>* d = <a class="code hl_function" href="group__loaders.html#gaf7350bdc690b7cb5592892858d2830a9">create_empty_simdata</a>();</div>
<div class="line"><span class="keywordtype">int</span> founders = <a class="code hl_function" href="group__loaders.html#ga54cd625669413c2ef171060e9844d814">load_all_simdata</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, <span class="stringliteral">&quot;eff-file.txt&quot;</span>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">founders &lt;- load.data(allele.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;, effect.file=&quot;eff-file.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md10"></a>
Swap out the set of additive trait effects for another</h2>
<p >This assumes the simulation is set up, that is, one of <b>Load a genetic map and a set of founder genotypes</b>, <b>Load a genetic map and several sets of founder genotypes</b>, <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b> or equivalent has been carried out. If the simulation is not yet set up and you with to load trait effect values, see the section above, titled <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b>.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td>Effect file:<ul>
<li>File location: eff-file2.txt </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_function" href="group__loaders.html#ga9d6b231bd6ddde82af66a5d276a2c739">load_effects_to_simdata</a>(d, <span class="stringliteral">&quot;eff-file2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__loaders_html_ga9d6b231bd6ddde82af66a5d276a2c739"><div class="ttname"><a href="group__loaders.html#ga9d6b231bd6ddde82af66a5d276a2c739">load_effects_to_simdata</a></div><div class="ttdeci">void load_effects_to_simdata(SimData *d, const char *filename)</div><div class="ttdoc">Populates a SimData combination with effect values.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:3699</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">load.different.effects(&quot;eff-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md11"></a>
Input file formats:</h2>
<p >The file format expected for the genotype file is something like: </p><div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --><p> where G01, G02, ..., are names of the starter set of genotypes; m1, m2, ..., are the markers; and entries in the matrix are two-character pairs listing the two alleles that genotype has at that marker. Alleles can be any non-space character and the order of the two alleles is saved.</p>
<p >Cells may be space-separated or tab-separated. The order of rows and columns does not matter. The value in the first cell ("name" in this example) is ignored.</p>
<p >The C version of the package has additional functions for loading differently-formatted genotype files. This functionality is on track to be improved and shared between versions in some upcoming update.</p>
<p >The map file should be formatted as follows: </p><div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --><p >The first (header) line's values are not checked. After that, all rows must have three space-or-tab-separated values. The first should be the marker name, the second an integer representing the chromosome number, and the third a decimal representing the position of the SNP along the chromosome in centiMorgans (cM). The order of the rows in the file does not matter.</p>
<p >Only markers that appear both in the genotype file and the map file are used in simulation. The simulation tool will print as output the number of markers that were loaded and the number that were discarded.</p>
<p >Loading an effect file is optional for running the simulation. Its format should be: </p><div class="fragment"><div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --><p >The first column should be a marker name, corresponding to a name used in the map and genotype files. The second should be the allele (non-space character, eg A) this effect value corresponds to. The third should be a decimal representing the additive effect value of that allele for that marker.</p>
<p >A particular marker/allele combination not being included in the file is equivalent to that combination having an effect value of 0. If a particular marker/allele combination is included multiple times in a file, only the last occurence is saved. If a marker name in the file does not match any marker tracked by the simulation, that row will be ignored. The simulation will print out the number of marker/allele pairs for which effects are loaded: for the sample file above, that would be 5.</p>
<h3><a class="anchor" id="autotoc_md12"></a>
Loading from other file formats</h3>
<p >Updates to expand the range of allowed input formats are coming soon.</p>
<h1><a class="anchor" id="autotoc_md13"></a>
Crossing &amp; Other Ways to Generate New Genotypes: Plant-themed</h1>
<p >Templates in this section assume you have loaded a set of founders whose group id is saved in the variable <code>founders</code>.</p>
<h2><a class="anchor" id="autotoc_md14"></a>
Single seed descent</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>For six generations grow a single seed from each plant to maturity. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> f6 = <a class="code hl_function" href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91">self_n_times</a>(d, 6, founders, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__crossers_html_gadf5beb682ac15fa5a0176e480adc9d91"><div class="ttname"><a href="group__crossers.html#gadf5beb682ac15fa5a0176e480adc9d91">self_n_times</a></div><div class="ttdeci">int self_n_times(SimData *d, int n, int group, GenOptions g)</div><div class="ttdoc">Selfs each member of a group for a certain number of generations.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:5031</div></div>
<div class="ttc" id="asim-operations_8c_html_a5c6b92aede8f29f53476aa78d4e4133a"><div class="ttname"><a href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a></div><div class="ttdeci">const GenOptions BASIC_OPT</div><div class="ttdoc">Options parameter to run SimData functions in their bare-bones form.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:5</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f6 &lt;- self.n.times(founders, n=6)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md15"></a>
Creating halfsib or fullsib families</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Make 20 random crosses and collect families of 6 full siblings from the result of each cross. </td><td><div class="fragment"><div class="line"><a class="code hl_struct" href="struct_gen_options.html">GenOptions</a> opt = {.<a class="code hl_variable" href="struct_gen_options.html#a8b345355c2f0c8b0429380b3e8ef01b7">will_name_offspring</a>=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .offspring_name_prefix=NULL, .family_size=6,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, .will_allocate_ids=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, .will_save_pedigree_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .will_save_alleles_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><span class="keywordtype">int</span> crosses = <a class="code hl_function" href="group__crossers.html#gae66231fb141a76432d1211e2cf0baf25">cross_random_individuals</a>(d, founders, 20, 0, opt);</div>
<div class="line"><span class="keywordtype">int</span> families[20];</div>
<div class="line"><a class="code hl_function" href="group__groupmod.html#ga62aec37e088d85c1df20a53b6d65daa4">split_into_families</a>(d, crosses, families);</div>
<div class="ttc" id="agroup__crossers_html_gae66231fb141a76432d1211e2cf0baf25"><div class="ttname"><a href="group__crossers.html#gae66231fb141a76432d1211e2cf0baf25">cross_random_individuals</a></div><div class="ttdeci">int cross_random_individuals(SimData *d, int from_group, int n_crosses, int cap, GenOptions g)</div><div class="ttdoc">Performs random crosses among members of a group.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:4369</div></div>
<div class="ttc" id="agroup__groupmod_html_ga62aec37e088d85c1df20a53b6d65daa4"><div class="ttname"><a href="group__groupmod.html#ga62aec37e088d85c1df20a53b6d65daa4">split_into_families</a></div><div class="ttdeci">void split_into_families(SimData *d, int group_id, int *results)</div><div class="ttdoc">Split a group into a set of smaller groups, each of which contains the genotypes in the original grou...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:1333</div></div>
<div class="ttc" id="asim-operations_8h_html_aa8cecfc5c5c054d2875c03e77b7be15d"><div class="ttname"><a href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a></div><div class="ttdeci">#define TRUE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:13</div></div>
<div class="ttc" id="asim-operations_8h_html_aa93f0eb578d23995850d61f7d61c55c1"><div class="ttname"><a href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a></div><div class="ttdeci">#define FALSE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:14</div></div>
<div class="ttc" id="astruct_gen_options_html"><div class="ttname"><a href="struct_gen_options.html">GenOptions</a></div><div class="ttdoc">A type that contains choices of settings for SimData functions that create a new AlleleMatrix/generat...</div><div class="ttdef"><b>Definition:</b> sim-operations.h:120</div></div>
<div class="ttc" id="astruct_gen_options_html_a8b345355c2f0c8b0429380b3e8ef01b7"><div class="ttname"><a href="struct_gen_options.html#a8b345355c2f0c8b0429380b3e8ef01b7">GenOptions::will_name_offspring</a></div><div class="ttdeci">int will_name_offspring</div><div class="ttdoc">A boolean: whether generated offspring should be given names.</div><div class="ttdef"><b>Definition:</b> sim-operations.h:121</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">crosses &lt;- cross.randomly(founders, n.crosses=20, offspring=6)</div>
<div class="line">families &lt;- break.group.into.families(crosses)</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Make 10 random crosses with the second founder genotype, and collect halfsib families of 6 half-siblings from the result of each cross. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><span class="keywordtype">int</span> targetparent_group = <a class="code hl_function" href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c">split_from_group</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_struct" href="struct_gen_options.html">GenOptions</a> opt = {.<a class="code hl_variable" href="struct_gen_options.html#a8b345355c2f0c8b0429380b3e8ef01b7">will_name_offspring</a>=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .offspring_name_prefix=NULL, .family_size=6,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, .will_allocate_ids=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, .will_save_pedigree_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .will_save_alleles_to_file=<a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><span class="keywordtype">int</span> crosses = <a class="code hl_function" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a>(d, targetparent_group, founders, 10, 0, 0, opt);       </div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> families[10];</div>
<div class="line"><a class="code hl_function" href="group__groupmod.html#ga39c5cb49068858b095acea9385e017ec">split_into_halfsib_families</a>(d, crosses, 1, families);</div>
<div class="ttc" id="agroup__crossers_html_ga3be2cb18d1c722b3a766a7a5663a8f0c"><div class="ttname"><a href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a></div><div class="ttdeci">int cross_randomly_between(SimData *d, int group1, int group2, int n_crosses, int cap1, int cap2, GenOptions g)</div><div class="ttdoc">Performs random crosses where the first parent comes from one group and the second from another.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:4604</div></div>
<div class="ttc" id="agroup__groupmod_html_ga39c5cb49068858b095acea9385e017ec"><div class="ttname"><a href="group__groupmod.html#ga39c5cb49068858b095acea9385e017ec">split_into_halfsib_families</a></div><div class="ttdeci">void split_into_halfsib_families(SimData *d, int group_id, int parent, int *results)</div><div class="ttdoc">Split a group into a set of smaller groups, each containing the genotypes from the original group tha...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:1449</div></div>
<div class="ttc" id="agroup__groupmod_html_gaf2eda46d016c384b166ce65605952b7c"><div class="ttname"><a href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c">split_from_group</a></div><div class="ttdeci">int split_from_group(SimData *d, int n, int indexes_to_split[n])</div><div class="ttdoc">Take a list of indexes and allocate the genotypes at those indexes to a new group.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:1216</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1L))</div>
<div class="line"> </div>
<div class="line">crosses &lt;- cross.randomly.between(targetparent_group, founders, n.crosses=10, offspring=6)</div>
<div class="line"> </div>
<div class="line">families &lt;- break.group.into.halfsib.families(crosses)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md16"></a>
Updating marker effects</h2>
<p >Updating the marker effect estimates is a task that must be done outside of genomicSimulation. Export required data from the simulation, run the marker effect re-estimation, then re-import marker effects as in <b>Swap out the set of additive trait effects for another</b>.</p>
<h2><a class="anchor" id="autotoc_md17"></a>
Trials and locations</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to simulate the success of the founder population in different environments. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> location1 = <a class="code hl_function" href="group__crossers.html#gaa42e48ab06f9051e5ed6879409cbe2f4">make_clones</a>(d, founders, <a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="line"><span class="keywordtype">int</span> location2 = <a class="code hl_function" href="group__crossers.html#gaa42e48ab06f9051e5ed6879409cbe2f4">make_clones</a>(d, founders, <a class="code hl_define" href="sim-operations_8h.html#aa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__crossers_html_gaa42e48ab06f9051e5ed6879409cbe2f4"><div class="ttname"><a href="group__crossers.html#gaa42e48ab06f9051e5ed6879409cbe2f4">make_clones</a></div><div class="ttdeci">int make_clones(SimData *d, int group, int inherit_names, GenOptions g)</div><div class="ttdoc">Creates an identical copy of each member of a group.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:5469</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">location1 &lt;- make.clones(founders)</div>
<div class="line">location2 &lt;- make.clones(founders)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Once separate copies exist for trials at different locations, simulate selection with different intensities or heritabilities for each different location group. Eg <b>Select on phenotype (simulated with a given heritability)</b>.</p>
<h2><a class="anchor" id="autotoc_md18"></a>
Backcrossing</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Pull out a chosen founder and repeatedly (for 20 generations) cross back to it. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><span class="keywordtype">int</span> targetparent_group = <a class="code hl_function" href="group__groupmod.html#gaf2eda46d016c384b166ce65605952b7c">split_from_group</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> backcross_generations[20];</div>
<div class="line">backcross_generation[0] = founders;</div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; 20; ++i) {</div>
<div class="line">    backcross_generations[i] = <a class="code hl_function" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a>(d, targetparent_group, backcross_generations[i-1], 10, 0, 0, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="line">}   </div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1L))</div>
<div class="line"> </div>
<div class="line">backcross_generations &lt;- rep(0L, times=20);</div>
<div class="line">backcross_generations[1] &lt;- founders</div>
<div class="line">for (ii in 1:20) {</div>
<div class="line">    backcross_generations[ii] &lt;- cross.randomly.between(targetparent_group, backcross_generations[ii-1], n.crosses=10)</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >For marker-assisted backcrossing, select each loop on presence of the marker and/or the estimated breeding value. See <b>Select on a qualitative trait</b> and <b>Select on a qualitative trait, then a quantitative trait.</b></p>
<h1><a class="anchor" id="autotoc_md19"></a>
Crossing &amp; Other Ways to Generate New Genotypes: Animal-themed</h1>
<p >Templates in this section assume you have loaded two sets of founders whose group id are saved in the variables <code>cows</code> and <code>bulls</code>. See <b>Load a genetic map and several sets of founder genotypes</b>.</p>
<h2><a class="anchor" id="autotoc_md20"></a>
Split offspring into male and female</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you randomly cross your founders, then want to identify the male and female calves among the offspring. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> offspring = <a class="code hl_function" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a>(d, cows, bulls, 10, 0, 0, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> offspring_f = <a class="code hl_function" href="group__groupmod.html#gacd7e433cf6af229f0d2932b04468e9ca">split_randomly_into_two</a>(d, offspring);</div>
<div class="line"><span class="keywordtype">int</span> offspring_m = offspring;</div>
<div class="ttc" id="agroup__groupmod_html_gacd7e433cf6af229f0d2932b04468e9ca"><div class="ttname"><a href="group__groupmod.html#gacd7e433cf6af229f0d2932b04468e9ca">split_randomly_into_two</a></div><div class="ttdeci">int split_randomly_into_two(SimData *d, int group_id)</div><div class="ttdoc">Flip a coin for each member of the group to decide if it should be moved to the new group.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:1784</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- cross.randomly.between(cows, bulls, n.crosses=10)</div>
<div class="line"> </div>
<div class="line">temporary &lt;- break.group.randomly(offspring, into.n = 2)</div>
<div class="line">offspring_f &lt;- temporary[1]</div>
<div class="line">offspring_m &lt;- temporary[2]</div>
<div class="line">rm(temporary)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >split_randomly_into_two or break.group.randomly can be used to split a group into two sub-groups by flipping a coin on each group member. This may result in two groups that are not quite the same size. To split a group of eg. 10 offspring into two groups of exactly 5 members, the cousin functions of split_evenly_into_two or break.group.evenly can be used instead.</p>
<h2><a class="anchor" id="autotoc_md21"></a>
Introducing fresh blood to male and female kernels</h2>
<p >Suppose the <code>offspring_f</code> and <code>offspring_m</code> groups exist, as created in the previous section <b>Split offspring into male and female</b>.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to add your new calves to the breeding kernels from which their parents were chosen. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> cowGroups[2];</div>
<div class="line">cowGroups[0] = cows;</div>
<div class="line">cowGroups[1] = offspring_f;</div>
<div class="line"> </div>
<div class="line">cows = <a class="code hl_function" href="group__groupmod.html#gade6b01bfc6450638a3a341b76e28a97d">combine_groups</a>(d, 2, cowGroups);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> bullGroups[2];</div>
<div class="line">bullGroups[0] = bulls;</div>
<div class="line">bullGroups[1] = offspring_m;</div>
<div class="line"> </div>
<div class="line">bulls = <a class="code hl_function" href="group__groupmod.html#gade6b01bfc6450638a3a341b76e28a97d">combine_groups</a>(d, 2, bullGroups);</div>
<div class="ttc" id="agroup__groupmod_html_gade6b01bfc6450638a3a341b76e28a97d"><div class="ttname"><a href="group__groupmod.html#gade6b01bfc6450638a3a341b76e28a97d">combine_groups</a></div><div class="ttdeci">int combine_groups(SimData *d, int list_len, int group_ids[list_len])</div><div class="ttdoc">Combine a set of groups into one group.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:1156</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">cows &lt;- combine.groups(c(cows,offspring_f))</div>
<div class="line">bulls &lt;- combine.groups(c(bulls,offspring_m))</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md22"></a>
Random mating with caps on number of offspring per animal</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose each cow should only have one calf this generation. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> offspring = <a class="code hl_function" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a>(d, cows, bulls, 10, 1, 0, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- cross.randomly.between(cows, bulls, cap1=1, n.crosses=10)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md23"></a>
Mating all females to a good male</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose the best bull in the population is the only one that will father calves under the current breeding strategy. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> bestbull = <a class="code hl_function" href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">split_by_bv</a>(d, bulls, 1, <a class="code hl_define" href="sim-operations_8h.html#aa93f0eb578d23995850d61f7d61c55c1">FALSE</a>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> offspring = <a class="code hl_function" href="group__crossers.html#ga3be2cb18d1c722b3a766a7a5663a8f0c">cross_randomly_between</a>(d, cows, bestbull, 10, 1, 0, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__calculators_html_gae34db60faf62b898d0572ea6c039b96a"><div class="ttname"><a href="group__calculators.html#gae34db60faf62b898d0572ea6c039b96a">split_by_bv</a></div><div class="ttdeci">int split_by_bv(SimData *d, int group, int top_n, int lowIsBest)</div><div class="ttdoc">Takes the top_n individuals in the group with the best breeding values/fitnesses and puts them in a n...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:5887</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">bestbull &lt;- select.by.gebv(bulls, number=1)</div>
<div class="line">offspring &lt;- cross.randomly.between(cows, bestbull, cap1=1, n.crosses=10)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md24"></a>
Making specific chosen matings</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to cross Cow1 to Bull1, Cow2 to Bull2, and Daisy to Bull1 </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> namesNeeded[5];</div>
<div class="line"><span class="keywordtype">int</span> IDsNeeded[5];</div>
<div class="line">namesNeeded[0] = <span class="stringliteral">&quot;Cow1&quot;</span>;</div>
<div class="line">namesNeeded[1] = <span class="stringliteral">&quot;Cow2&quot;</span>;</div>
<div class="line">namesNeeded[2] = <span class="stringliteral">&quot;Daisy&quot;</span>;</div>
<div class="line">namesNeeded[3] = <span class="stringliteral">&quot;Bull1&quot;</span>;</div>
<div class="line">namesNeeded[4] = <span class="stringliteral">&quot;Bull2&quot;</span>;</div>
<div class="line"><a class="code hl_function" href="group__getters.html#ga84037d781b8850f9d90425ad16a0d059">get_ids_of_names</a>(d-&gt;<a class="code hl_variable" href="struct_sim_data.html#a0a3c4152d86786884939add509bfa325">m</a>, 5, namesNeeded, IDsNeeded);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][3];</div>
<div class="line">crossingPlan[0][0] = IDsNeeded[0];</div>
<div class="line">crossingPlan[0][1] = IDsNeeded[1];</div>
<div class="line">crossingPlan[0][2] = IDsNeeded[2];</div>
<div class="line">crossingPlan[1][0] = IDsNeeded[3];</div>
<div class="line">crossingPlan[1][1] = IDsNeeded[4];</div>
<div class="line">crossingPlan[1][2] = IDsNeeded[3];</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> offspring = <a class="code hl_function" href="group__crossers.html#gabb793988799d792dfe4007efdf988068">cross_these_combinations</a>(d, 3, crossingPlan, <a class="code hl_variable" href="sim-operations_8c.html#a5c6b92aede8f29f53476aa78d4e4133a">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__crossers_html_gabb793988799d792dfe4007efdf988068"><div class="ttname"><a href="group__crossers.html#gabb793988799d792dfe4007efdf988068">cross_these_combinations</a></div><div class="ttdeci">int cross_these_combinations(SimData *d, int n_combinations, int combinations[2][n_combinations], GenOptions g)</div><div class="ttdoc">Performs the crosses of pairs of parents whose ids are provided in an array.</div><div class="ttdef"><b>Definition:</b> sim-operations.c:4855</div></div>
<div class="ttc" id="agroup__getters_html_ga84037d781b8850f9d90425ad16a0d059"><div class="ttname"><a href="group__getters.html#ga84037d781b8850f9d90425ad16a0d059">get_ids_of_names</a></div><div class="ttdeci">void get_ids_of_names(AlleleMatrix *start, int n_names, char *names[n_names], unsigned int *output)</div><div class="ttdoc">Search for genotypes with certain names in a linked list of AlleleMatrix and save the ids of those na...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:921</div></div>
<div class="ttc" id="astruct_sim_data_html_a0a3c4152d86786884939add509bfa325"><div class="ttname"><a href="struct_sim_data.html#a0a3c4152d86786884939add509bfa325">SimData::m</a></div><div class="ttdeci">AlleleMatrix * m</div><div class="ttdoc">Pointer to an AlleleMatrix, which stores data and metadata of founders and simulated offspring.</div><div class="ttdef"><b>Definition:</b> sim-operations.h:230</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- cross.combinations(c(&quot;Cow1&quot;, &quot;Cow2&quot;, &quot;Daisy&quot;), c(&quot;Bull1&quot;, &quot;Bull2&quot;, &quot;Bull1&quot;))</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The other option is to create an input file of the following format: </p><div class="fragment"><div class="line">Cow1 Bull1</div>
<div class="line">Cow2 Bull2</div>
<div class="line">Daisy Bull1</div>
</div><!-- fragment --><p> and call <code>make_crosses_from_file</code> or <code>cross.combinations.file</code>.</p>
<h2><a class="anchor" id="autotoc_md25"></a>
Tracking age and discarding individuals that are 'too old'</h2>
<p >Updates to add an age/custom flag to each individual is coming soon.</p>
<h1><a class="anchor" id="autotoc_md26"></a>
Selection</h1>
<h2><a class="anchor" id="autotoc_md27"></a>
Manually select on true breeding value</h2>
<p >The template selection template. This does the exact same thing as split_by_bv / select.by.gebv, but using the 'custom selection method interface'. Hopefully this aids in understanding the concept.</p>
<h2><a class="anchor" id="autotoc_md28"></a>
Select on phenotype (simulated with a given heritability)</h2>
<h2><a class="anchor" id="autotoc_md29"></a>
Select on a qualitative trait</h2>
<h2><a class="anchor" id="autotoc_md30"></a>
Select on a qualitative trait, then a quantitative trait.</h2>
<h2><a class="anchor" id="autotoc_md31"></a>
Select on an index weighting two traits.</h2>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
