<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Features and Templates Guide</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.5.02</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Features and Templates Guide </div></div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md24">Setting Up</a><ul><li class="level2"><a href="#autotoc_md25">Input file formats:</a><ul><li class="level3"><a href="#autotoc_md26">Genetic map files</a></li>
<li class="level3"><a href="#autotoc_md27">Genotype matrix files</a></li>
<li class="level3"><a href="#autotoc_md28">Marker effect files</a></li>
<li class="level3"><a href="#autotoc_md29">Loading from other file formats</a></li>
</ul>
</li>
<li class="level2"><a href="#autotoc_md30">Set the random number generator</a></li>
<li class="level2"><a href="#autotoc_md31">Load a genetic map and a set of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md32">Load a genetic map and genotype file, specifying file format</a></li>
<li class="level2"><a href="#autotoc_md33">Load a genetic map and several sets of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md34">Load a genetic map, a set of founder genotypes, and a set of additive trait effects</a></li>
<li class="level2"><a href="#autotoc_md35">Load another set of additive trait effects</a></li>
<li class="level2"><a href="#autotoc_md36">Load another linkage map</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md37">Crossing &amp; Other Ways to Generate New Genotypes: Plant-themed</a><ul><li class="level2"><a href="#autotoc_md38">Single seed descent</a></li>
<li class="level2"><a href="#autotoc_md39">Creating halfsib or fullsib families</a></li>
<li class="level2"><a href="#autotoc_md40">Updating marker effects</a></li>
<li class="level2"><a href="#autotoc_md41">Trials and locations</a></li>
<li class="level2"><a href="#autotoc_md42">Backcrossing</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md43">Crossing &amp; Other Ways to Generate New Genotypes: Animal-themed</a><ul><li class="level2"><a href="#autotoc_md44">Split offspring into male and female</a></li>
<li class="level2"><a href="#autotoc_md45">Introducing fresh blood to male and female kernels</a></li>
<li class="level2"><a href="#autotoc_md46">Random mating with caps on number of offspring per animal</a></li>
<li class="level2"><a href="#autotoc_md47">Mating all females to a good male</a></li>
<li class="level2"><a href="#autotoc_md48">Making specific chosen matings</a></li>
<li class="level2"><a href="#autotoc_md49">Three-Way Crosses</a></li>
<li class="level2"><a href="#autotoc_md50">Tracking age and culling individuals that are too old</a></li>
<li class="level2"><a href="#autotoc_md51">Species with differing male and female recombination rates</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md52">Selection</a><ul><li class="level2"><a href="#autotoc_md53">Manually select on true breeding value</a></li>
<li class="level2"><a href="#autotoc_md54">Select on phenotype (simulated with a given heritability)</a></li>
<li class="level2"><a href="#autotoc_md55">Select on a qualitative trait</a></li>
<li class="level2"><a href="#autotoc_md56">Select on a qualitative trait, then a quantitative trait.</a></li>
<li class="level2"><a href="#autotoc_md57">Select on an index weighting two traits.</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p ><a class="anchor" id="md_Templates"></a></p>
<p >While the documentation pages of genomicSimulationC and its counterpart <a href="https://github.com/vllrs/genomicSimulation">genomicSimulation</a> do describe all the features the simulation tool offers, they don't do so in a way that is necessarily easy to browse or that's helpful for new users to understand the simulation tool.</p>
<p >This page provides a set of templates for implementing common breeding project actions in genomicSimulationC and its R counterpart.</p>
<h1><a class="anchor" id="autotoc_md24"></a>
Setting Up</h1>
<h2><a class="anchor" id="autotoc_md25"></a>
Input file formats:</h2>
<p >genomicSimulation has three types of input files: a genotype matrix, a genetic map, and a list of marker effects. If multiple files are provided at once in a call to <code>load_data_files</code> (C)/<code>load.data</code> (R), the map file will be loaded first, followed by the genotype matrix and marker effect file.</p>
<h3><a class="anchor" id="autotoc_md26"></a>
Genetic map files</h3>
<p >The simplest genetic map file is formatted as follows: </p><div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --><p> Other valid genetic map files might include: </p><div class="fragment"><div class="line">chr marker pos</div>
<div class="line">1A 1243509 173.2</div>
<div class="line">1A 2350898 462.2</div>
<div class="line">1B 4360992 32.009</div>
<div class="line">2A 1243556 243.5</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">gene 10 3.24</div>
<div class="line">othergene 10 8.3e-1</div>
<div class="line">etc 15 1.203e2</div>
</div><!-- fragment --><p >The header line is optional. If there is a header line, the three columns "marker" "chr" and "pos" may be rearranged into any ordering. If the header is not provided, the order is assumed to be marker name in the first column, followed by chromosome in the second column, followed by position in the third.</p>
<p >Regarding marker names: For all genomicSimulation features to work, genetic markers must have names. There is no issue with purely numeric names.</p>
<p >Regarding chromosomes/linkage group: any alphanumeric combination may be used to denote a chromosome/linkage group, for example '9' or '1A'. There is no limit on the number of unique chromosomes/linkage groups in the map.</p>
<p >Regarding marker position: this represents the position in centimorgans of each marker along its linkage group. (Distance of 1 cM = expected probability of 0.01 chromosomal crossovers in that range.)</p>
<p >The cells in the map file may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file, but therefore marker names/linkage group names cannot contain spaces, tabs, or commas.</p>
<p >The order in which genetic markers are presented in the file does not matter. If a marker is duplicated in the file, it will be recorded twice as two separate markers in the same position and with the same name. However, because loading genotypes depends on searching for markers by name, if two markers have the same name, genotypes may not be loaded correctly. Therefore it is suggested that all markers have unique names.</p>
<h3><a class="anchor" id="autotoc_md27"></a>
Genotype matrix files</h3>
<p >The simplest genotype matrix file is formatted as follows: </p><div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  CC  AA</div>
</div><!-- fragment --><p> where G01, G02, ..., are names of the founder genotypes; m1, m2, ..., are the genetic markers; and entries in the matrix are the alleles that the founder genotypes have at those markers.</p>
<p >Other valid genotype matrix files might include: </p><div class="fragment"><div class="line"> m100, m101, m102</div>
<div class="line">cand1,0,0,1</div>
<div class="line">cand2,1,2,2</div>
<div class="line">cand4,2,1,2</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">marker1 T/T T/T T/T T/A T/T A/T</div>
<div class="line">marker3 T/T T/T T/A T/A T/T T/T</div>
<div class="line">marker2 A/A A/A A/A A/A T/T A/A</div>
</div><!-- fragment --><p >The genotype matrix can be row-major or column-major (that is, the genetic markers may be rows, or columns). The program will determine the orientation by attempting to match row and column headers with names of markers tracked by the simulation (extracted from a genetic map file). If the simulation does not yet have a list of tracked markers (that is, no genetic map has been simultaneously or previously provided), then it defaults to assuming that rows represent genetic markers. <br  />
</p>
<p >The order in which genetic markers are presented in the file does not matter. Candidate genotypes, however, will be saved internally in the simulation in the order that they appear in the file. Candidate genotype names do not need to be unique. Candidate genotype names are optional, if candidates are stored as columns. genomicSimulation cannot read a gentoype matrix with no row headers, so if candidates are stored as rows, then they are required to have names.</p>
<p >Genetic markers' names are required.</p>
<p >When there are both row and column headers, the first/corner cell (the one that contained the value "name" in the first example above) can be deleted or can contain any text. Its value will be ignored.</p>
<p >The table cells in the genotype matrix file may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file.</p>
<p >There are several options for how the alleles at each marker can be presented. All allele pair cells in the same genotype matrix must be in the same format. The format of the allele pairs in the genotype matrix is automatically detected. There are four acceptable formats for allele pairs in the genotype matrix:</p>
<ol type="1">
<li>Any pair of characters (eg. "AA", "TA", "nW"). Each character is an allele. This is a format that specifies allele phase (ie. "AT" and "TA" are different genotypes).</li>
<li>Any pair of characters, separated by a forwards slash "/" character (eg. "A/A", "T/A", "n/W"). The two characters either side of the slash are the alleles. This is a format that specifies allele phase (ie. "AT" and "TA" are different genotypes).</li>
<li>Alternate allele counts ("0", "1", "2"). This is a format that does not specify allele phase, so phase of heterozygotes will be randomised when loaded. The counts represent the number of copies of the alternate allele (stored inside genomicSimulation as "A", while the reference allele is stored as "T"). "0" then corresponds to "TT", "2" to "AA", and "1" will be randomised as either "TA" or "AT". Corresponding marker effect files for calculating breeding values must use allele "A" consistently to represent the alternate allele, and "T" to represent the reference. (alternatively, <code>change_allele_symbol</code> (C)/<code>change.allele.symbol</code> (R) can be used after loading to change T and/or A to other symbols.)</li>
<li>(subset of) IUPAC encodings of DNA bases ("G", "A", "T", "C", "R", "Y", "M", "K", "S", "W", "N"). This is a format that does not specify allele phase, so phase of heterozygotes will be randomised when loaded. See below for meanings of the IUPAC encoding symbols that genomicSimulation can parse.</li>
</ol>
<table class="doxtable">
<tr>
<th>IUPAC symbol </th><th>genomicSimulation genotype </th></tr>
<tr>
</tr>
</table>
<p >GG </p>
<p >A</p>
<p >AA </p>
<p >T</p>
<p >TT </p>
<p >C</p>
<p >CC </p>
<p >R</p>
<p >GA or AG </p>
<p >Y</p>
<p >TC or CT </p>
<p >M</p>
<p >AC or CA </p>
<p >K</p>
<p >GT or TG </p>
<p >S</p>
<p >GC or CG </p>
<p >W</p>
<p >AT or TA </p>
<p >N</p>
<p >\0\0 (nulls represent unknown genotype) </p>
<p ><b>Note you might have a genotype matrix of only "AA", "AT", and "TT". This uses "alternate allele counts"-style encoding (like format 3) but presents it in a format that looks like pairs of alleles (format 1).</b> genomicSimulation expects allele pair encodings to include haplotype phase, (that is, to have four possible values for genotypes of two alleles, not three: eg. "AA", "AT", "TA", and "TT" instead of just "AA", "AT", "TT").</p>
<p >Two options for loading a dataset with non-phased "AA"/"AT"/"TT" allele pairs are:</p>
<ul>
<li>Use “haplotyping”/”haplotype phasing”/”haplotype inference” software to infer whether heterozygotes are "AT" or "TA", before loading into genomicSimulation.</li>
<li>Find-and-replace "TT" with "0", "AT" with "1", and "AA" with "2" before loading into genomicSimulation. genomicSimulation will then randomise the phase of each haplotype.</li>
</ul>
<h3><a class="anchor" id="autotoc_md28"></a>
Marker effect files</h3>
<p >Loading marker effect file(s) is optional for running the simulation. The simplest marker effect file is formatted as follows: </p><div class="fragment"><div class="line">marker allele eff</div>
<div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --><p> Other valid marker effect files might include: </p><div class="fragment"><div class="line">marker eff allele</div>
<div class="line">m1243509 0.1 A</div>
<div class="line">m1243509 -0.1 T</div>
<div class="line">m2350898 0.15 T</div>
<div class="line">m2350898 -0.1 A</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">specialgene G 1.0</div>
</div><!-- fragment --><p >A single marker effect file represents the allele effects for one trait. Multiple files like this can be loaded in order to calculate breeding values for multiple traits separately.</p>
<p >The header line is optional. If there is a header line, the three columns "marker" "allele" and "eff" may be rearranged into any ordering. If the header is not provided, the order is assumed to be marker name in the first column, followed by allele, followed by additive effect value.</p>
<p >The order of the rows in the file does not matter. If multiple rows exist in the file for the same marker name/allele combination, only the last row's additive effect value will be saved.</p>
<p >Regarding marker names: Only rows in this file whose marker name matches a marker name from a previously- or simultaneously-loaded map file or genotype matrix file will be loaded.</p>
<p >Regarding allele: This column should be the allele (non-space character, eg "A") that this effect value corresponds to.</p>
<p >Regarding allele effect: This column should be a decimal representing the additive effect value of the same-row allele for the same-row marker. It can be positive or negative or zero, and can be represented by an integer, a decimal, or scientific notation.</p>
<p >Cells may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file.</p>
<h3><a class="anchor" id="autotoc_md29"></a>
Loading from other file formats</h3>
<p >Updates to expand the range of allowed input formats are coming soon.</p>
<h2><a class="anchor" id="autotoc_md30"></a>
Set the random number generator</h2>
<p >In C, you should manually set the seed for the random number generator when you create the SimData object, before loading data. The R version borrows R's own random number generator so has no need for setting a seed, or for manually creating the simulation data structure before loading data.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* sd = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="ttc" id="agroup__shortnames_html_gaa9b2f128bb156d36d9334281b158c56a"><div class="ttname"><a href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a></div><div class="ttdeci">#define SimData</div><div class="ttdef"><b>Definition:</b> sim-operations.h:145</div></div>
<div class="ttc" id="agroup__shortnames_html_gacbedd673ae83f70154d663852acfea70"><div class="ttname"><a href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a></div><div class="ttdeci">#define create_empty_simdata</div><div class="ttdef"><b>Definition:</b> sim-operations.h:160</div></div>
</div><!-- fragment --> </td><td></td></tr>
</table>
<h2><a class="anchor" id="autotoc_md31"></a>
Load a genetic map and a set of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders = init.group;</div>
<div class="ttc" id="agroup__shortnames_html_ga1b342f7327286629391ccbad90ea19e4"><div class="ttname"><a href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a></div><div class="ttdeci">#define DETECT_FILE_FORMAT</div><div class="ttdef"><b>Definition:</b> sim-operations.h:166</div></div>
<div class="ttc" id="agroup__shortnames_html_ga2e10fe6c17f65cefc887a7b8868d9017"><div class="ttname"><a href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a></div><div class="ttdeci">#define load_data_files</div><div class="ttdef"><b>Definition:</b> sim-operations.h:165</div></div>
<div class="ttc" id="agroup__shortnames_html_gab9139b22223d33aa0a762d3ec00b3072"><div class="ttname"><a href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a></div><div class="ttdeci">#define MultiIDSet</div><div class="ttdef"><b>Definition:</b> sim-operations.h:134</div></div>
<div class="ttc" id="agroup__shortnames_html_gadb2ef9b75828b244e0b3e1cb3680d7ca"><div class="ttname"><a href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a></div><div class="ttdeci">#define GroupNum</div><div class="ttdef"><b>Definition:</b> sim-operations.h:126</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">founders &lt;- load.data(genotype.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The simulation will track genotypes across all markers present in the map file. Which markers are tracked by the simulation cannot currently be changed. Line order does not matter in any file.</p>
<p >Consider using the get_group_ family of functions (genomicSimulationC) and see.group.data function (genomicSimulation) to confirm data is correctly loaded.</p>
<h2><a class="anchor" id="autotoc_md32"></a>
Load a genetic map and genotype file, specifying file format</h2>
<p >By default, genomicSimulation attempts to detect the format of input files as it loads them. When a file loading command is run, genomicSimulation prints out its assumptions about the format of the file.</p>
<p >If the printed logs are incorrect about the layout of your map file or effect file, the most likely cause is a mis-spelling in the header row of the file. If the printed logs are incorrect about the size of the table in your map file or effect file, the issue may be that the numbers cannot be detected as numbers, or that there are issues with the column spacers/line breaks that cause there to be more than three entries to a row.</p>
<p >If the printed logs are incorrect about the layout or format of your genotype matrix file, you can manually specify the file format for genomicSimulation, as follows.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gae2dfa9fab7307781bef01124e32e7495">FileFormatSpec</a> mformat = <a class="code hl_define" href="group__shortnames.html#ga8bc4f8369ef14861b409317322046905">define_matrix_format_details</a>(<a class="code hl_define" href="sim-operations_8h.html#ad3981037d5b9c99a0c49de9f0dfc1e8c">GSC_TRUE</a>,<a class="code hl_define" href="sim-operations_8h.html#ad3981037d5b9c99a0c49de9f0dfc1e8c">GSC_TRUE</a>,<a class="code hl_enumvalue" href="group__structs.html#gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c">GSC_GENOTYPECELLSTYLE_PAIR</a>);</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, mformat);</div>
<div class="ttc" id="agroup__shortnames_html_ga8bc4f8369ef14861b409317322046905"><div class="ttname"><a href="group__shortnames.html#ga8bc4f8369ef14861b409317322046905">define_matrix_format_details</a></div><div class="ttdeci">#define define_matrix_format_details</div><div class="ttdef"><b>Definition:</b> sim-operations.h:168</div></div>
<div class="ttc" id="agroup__shortnames_html_gae2dfa9fab7307781bef01124e32e7495"><div class="ttname"><a href="group__shortnames.html#gae2dfa9fab7307781bef01124e32e7495">FileFormatSpec</a></div><div class="ttdeci">#define FileFormatSpec</div><div class="ttdef"><b>Definition:</b> sim-operations.h:167</div></div>
<div class="ttc" id="agroup__structs_html_gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c"><div class="ttname"><a href="group__structs.html#gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c">GSC_GENOTYPECELLSTYLE_PAIR</a></div><div class="ttdeci">@ GSC_GENOTYPECELLSTYLE_PAIR</div><div class="ttdef"><b>Definition:</b> sim-operations.h:835</div></div>
<div class="ttc" id="asim-operations_8h_html_ad3981037d5b9c99a0c49de9f0dfc1e8c"><div class="ttname"><a href="sim-operations_8h.html#ad3981037d5b9c99a0c49de9f0dfc1e8c">GSC_TRUE</a></div><div class="ttdeci">#define GSC_TRUE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:80</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">mformat &lt;- define.matrix.format.details(has.header=TRUE, markers.as.rows=TRUE, cell.style=&quot;Pair&quot;)</div>
<div class="line">founders &lt;- load.data(genotype.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;,format=mformat)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >It is not required that you manually specify all details of the file format. Whatever details of the file format are not manually specified will be automatically detected.</p>
<p >See the documentation for the function <code>define_matrix_format_details</code> (C)/<code>define.matrix.format.details</code> (R) for more information.</p>
<h2><a class="anchor" id="autotoc_md33"></a>
Load a genetic map and several sets of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">    G01 G02 G03 G04 G05</div>
<div class="line">m1  TT  TT  TT  TA  TT</div>
<div class="line">m3  TT  TT  TA  TA  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Genotype file 2:</p><ul>
<li>File location: genotype-file2.txt</li>
<li>File contents: <div class="fragment"><div class="line"> G06</div>
<div class="line">m1,AT</div>
<div class="line">m2,AA</div>
<div class="line">m3,TT</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders_a = init.group;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders_b = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(d, <span class="stringliteral">&quot;genotype-file2.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_ga23a7a92af469dfbba4905bb5c9c90a9b"><div class="ttname"><a href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a></div><div class="ttdeci">#define load_genotypefile</div><div class="ttdef"><b>Definition:</b> sim-operations.h:164</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">founders_a &lt;- load.data(genotype.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;)</div>
<div class="line">founders_b &lt;- load.genotypes(&quot;genotype-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md34"></a>
Load a genetic map, a set of founder genotypes, and a set of additive trait effects</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Effect file:</p><ul>
<li>File location: eff-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker allele eff</div>
<div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, <span class="stringliteral">&quot;eff-file.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders = init.group;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> eff1 = init.effectSet;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> map1 = init.map;</div>
<div class="ttc" id="agroup__shortnames_html_ga516a882a4bac611d495b8f454478d674"><div class="ttname"><a href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a></div><div class="ttdeci">#define EffectID</div><div class="ttdef"><b>Definition:</b> sim-operations.h:128</div></div>
<div class="ttc" id="agroup__shortnames_html_gad11f85a6aae488f7efaab5abecaf1ac7"><div class="ttname"><a href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a></div><div class="ttdeci">#define MapID</div><div class="ttdef"><b>Definition:</b> sim-operations.h:132</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line"># Note that the output is slightly different when an effect file is loaded.</div>
<div class="line">init &lt;- load.data(genotype.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;, effect.file=&quot;eff-file.txt&quot;)</div>
<div class="line">founders &lt;- init$groupNum</div>
<div class="line">eff1 &lt;- init$effectID</div>
<div class="line">map1 &lt;- init$mapID</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md35"></a>
Load another set of additive trait effects</h2>
<p >This assumes the simulation is set up, that is, one of <b>Load a genetic map and a set of founder genotypes</b>, <b>Load a genetic map and several sets of founder genotypes</b>, <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b> or equivalent has been carried out. If the simulation is not yet set up and you wish to load trait effect values, see the section above, titled <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b>.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td>Effect file:<ul>
<li>File location: eff-file2.txt </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> eff2 = <a class="code hl_define" href="group__shortnames.html#ga7b891ecd001d68f60b379b1bb14e43d7">load_effectfile</a>(d, <span class="stringliteral">&quot;eff-file2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__shortnames_html_ga7b891ecd001d68f60b379b1bb14e43d7"><div class="ttname"><a href="group__shortnames.html#ga7b891ecd001d68f60b379b1bb14e43d7">load_effectfile</a></div><div class="ttdeci">#define load_effectfile</div><div class="ttdef"><b>Definition:</b> sim-operations.h:163</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">eff2 &lt;- load.effects(&quot;eff-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md36"></a>
Load another linkage map</h2>
<p >Markers in this map that are not present in the primary linkage map will be ignored. (That is, the first genetic map loaded must contain all markers you wish to be tracked by the simulation. Later genetic maps can only provide different recombination frequencies or chromosome allocations).</p>
<table class="doxtable">
<tr>
<th>Task </th><th>Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td></td><td>Map file:<ul>
<li>File location: mapfile2.txt </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> map2 = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(d, <span class="stringliteral">&quot;mapfile2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__shortnames_html_gaf16559ebf0798beeaea167422915ece1"><div class="ttname"><a href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a></div><div class="ttdeci">#define load_mapfile</div><div class="ttdef"><b>Definition:</b> sim-operations.h:162</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">map2 &lt;- load.map(&quot;mapfile2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h1><a class="anchor" id="autotoc_md37"></a>
Crossing &amp; Other Ways to Generate New Genotypes: Plant-themed</h1>
<p >Templates in this section assume you have loaded a set of founders whose group number is saved in the variable <code>founders</code>.</p>
<h2><a class="anchor" id="autotoc_md38"></a>
Single seed descent</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>For six generations grow a single seed from each plant to maturity. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f6 = <a class="code hl_define" href="group__shortnames.html#gaea0e2102f5a29f876e34fb234e8d7d65">self_n_times</a>(d, 6, founders, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_ga3710da75e403e5a3da4e2deade0d50af"><div class="ttname"><a href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a></div><div class="ttdeci">#define NO_MAP</div><div class="ttdef"><b>Definition:</b> sim-operations.h:133</div></div>
<div class="ttc" id="agroup__shortnames_html_ga4af98a71b0736f81c32df8e5e4fd6a83"><div class="ttname"><a href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a></div><div class="ttdeci">#define BASIC_OPT</div><div class="ttdef"><b>Definition:</b> sim-operations.h:139</div></div>
<div class="ttc" id="agroup__shortnames_html_gaea0e2102f5a29f876e34fb234e8d7d65"><div class="ttname"><a href="group__shortnames.html#gaea0e2102f5a29f876e34fb234e8d7d65">self_n_times</a></div><div class="ttdeci">#define self_n_times</div><div class="ttdef"><b>Definition:</b> sim-operations.h:222</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f6 &lt;- self.n.times(founders, n=6)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Using NO_MAP for the recombination map parameter in any crossing function in the C version causes the tool to default to the first/primary recombination map. This is the same default as in the R version.</p>
<h2><a class="anchor" id="autotoc_md39"></a>
Creating halfsib or fullsib families</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Make 20 random crosses and collect families of 6 full siblings from the result of each cross. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .offspring_name_prefix=NULL, .family_size=6,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=NOT_AN_EFFECT_ID, .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> crosses = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, founders, 20, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> families[20];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga9b6583f62a092756b6a06e08536ed42b">split_into_families</a>(d, crosses, families);</div>
<div class="ttc" id="agroup__shortnames_html_ga87aa2de59b68e428e5b6f07f5955c15a"><div class="ttname"><a href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a></div><div class="ttdeci">#define make_random_crosses</div><div class="ttdef"><b>Definition:</b> sim-operations.h:219</div></div>
<div class="ttc" id="agroup__shortnames_html_ga9b6583f62a092756b6a06e08536ed42b"><div class="ttname"><a href="group__shortnames.html#ga9b6583f62a092756b6a06e08536ed42b">split_into_families</a></div><div class="ttdeci">#define split_into_families</div><div class="ttdef"><b>Definition:</b> sim-operations.h:207</div></div>
<div class="ttc" id="agroup__shortnames_html_gaa320efa0931830d69965beb2f5399acf"><div class="ttname"><a href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a></div><div class="ttdeci">#define GenOptions</div><div class="ttdef"><b>Definition:</b> sim-operations.h:138</div></div>
<div class="ttc" id="agroup__shortnames_html_gaa8cecfc5c5c054d2875c03e77b7be15d"><div class="ttname"><a href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a></div><div class="ttdeci">#define TRUE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:121</div></div>
<div class="ttc" id="agroup__shortnames_html_gaa93f0eb578d23995850d61f7d61c55c1"><div class="ttname"><a href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a></div><div class="ttdeci">#define FALSE</div><div class="ttdef"><b>Definition:</b> sim-operations.h:122</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">crosses &lt;- make.random.crosses(founders, n.crosses=20, offspring=6)</div>
<div class="line">families &lt;- break.group.into.families(crosses)</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Make 10 random crosses with the second founder genotype, and collect halfsib families of 6 half-siblings from the result of each cross. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> targetparent_group = <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .offspring_name_prefix=NULL, .family_size=6,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=NOT_AN_EFFECT_ID, .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> crosses = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, targetparent_group, founders, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> families[10];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga4fa671550fea7b5c723986703a3078ca">split_into_halfsib_families</a>(d, crosses, 1, families);</div>
<div class="ttc" id="agroup__shortnames_html_ga4fa671550fea7b5c723986703a3078ca"><div class="ttname"><a href="group__shortnames.html#ga4fa671550fea7b5c723986703a3078ca">split_into_halfsib_families</a></div><div class="ttdeci">#define split_into_halfsib_families</div><div class="ttdef"><b>Definition:</b> sim-operations.h:208</div></div>
<div class="ttc" id="agroup__shortnames_html_ga702855cf67ebd337941ec9c13e40dfe9"><div class="ttname"><a href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a></div><div class="ttdeci">#define make_group_from</div><div class="ttdef"><b>Definition:</b> sim-operations.h:203</div></div>
<div class="ttc" id="agroup__shortnames_html_ga9d7d8919aa31f1cfb54c902800a2d063"><div class="ttname"><a href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a></div><div class="ttdeci">#define make_random_crosses_between</div><div class="ttdef"><b>Definition:</b> sim-operations.h:220</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1L))</div>
<div class="line"> </div>
<div class="line">crosses &lt;- make.random.crosses.between(targetparent_group, founders, n.crosses=10, offspring=6)</div>
<div class="line"> </div>
<div class="line">families &lt;- break.group.into.halfsib.families(crosses)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md40"></a>
Updating marker effects</h2>
<p >Updating the marker effect estimates is a task that must be done outside of genomicSimulation. Export required data from the simulation, run the marker effect re-estimation, then re-import marker effects as in <b>Swap out the set of additive trait effects for another</b>.</p>
<h2><a class="anchor" id="autotoc_md41"></a>
Trials and locations</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to simulate the success of the founder population in different environments. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> location1 = <a class="code hl_define" href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a>(d, founders, <a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> location2 = <a class="code hl_define" href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a>(d, founders, <a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_gacd45e8b70bb44a78325f6947479db0ad"><div class="ttname"><a href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a></div><div class="ttdeci">#define make_clones</div><div class="ttdef"><b>Definition:</b> sim-operations.h:224</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">location1 &lt;- make.clones(founders)</div>
<div class="line">location2 &lt;- make.clones(founders)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Once separate copies exist for trials at different locations, simulate selection with different intensities or heritabilities for each different location group. Eg <b>Select on phenotype (simulated with a given heritability)</b>.</p>
<h2><a class="anchor" id="autotoc_md42"></a>
Backcrossing</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Pull out a chosen founder and repeatedly (for 20 generations) cross back to it. </td><td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> targetparent_group = <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> backcross_generations[20];</div>
<div class="line">backcross_generation[0] = founders;</div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; 20; ++i) {</div>
<div class="line">    backcross_generations[i] = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, targetparent_group, backcross_generations[i-1], 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line">}</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1L))</div>
<div class="line"> </div>
<div class="line">backcross_generations &lt;- rep(0L, times=20);</div>
<div class="line">backcross_generations[1] &lt;- founders</div>
<div class="line">for (ii in 1:20) {</div>
<div class="line">    backcross_generations[ii] &lt;- make.random.crosses.between(targetparent_group, backcross_generations[ii-1], n.crosses=10)</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >For marker-assisted backcrossing, select each loop on presence of the marker and/or the estimated breeding value. See <b>Select on a qualitative trait</b> and <b>Select on a qualitative trait, then a quantitative trait.</b></p>
<h1><a class="anchor" id="autotoc_md43"></a>
Crossing &amp; Other Ways to Generate New Genotypes: Animal-themed</h1>
<p >Templates in this section assume you have loaded two sets of founders whose group id are saved in the variables <code>cows</code> and <code>bulls</code>. See <b>Load a genetic map and several sets of founder genotypes</b>.</p>
<h2><a class="anchor" id="autotoc_md44"></a>
Split offspring into male and female</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you randomly cross your founders, then want to identify the male and female calves among the offspring. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_f = <a class="code hl_define" href="group__shortnames.html#ga846bceb22b2a254269d517342c60b189">split_randomly_into_two</a>(d, offspring);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_m = offspring;</div>
<div class="ttc" id="agroup__shortnames_html_ga846bceb22b2a254269d517342c60b189"><div class="ttname"><a href="group__shortnames.html#ga846bceb22b2a254269d517342c60b189">split_randomly_into_two</a></div><div class="ttdeci">#define split_randomly_into_two</div><div class="ttdef"><b>Definition:</b> sim-operations.h:212</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, n.crosses=10)</div>
<div class="line"> </div>
<div class="line">temporary &lt;- break.group.randomly(offspring, into.n = 2)</div>
<div class="line">offspring_f &lt;- temporary[1]</div>
<div class="line">offspring_m &lt;- temporary[2]</div>
<div class="line">rm(temporary)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >split_randomly_into_two or break.group.randomly can be used to split a group into two sub-groups by flipping a coin on each group member. This may result in two groups that are not quite the same size. To split a group of eg. 10 offspring into two groups of exactly 5 members, the cousin functions, split_evenly_into_two or break.group.evenly, can be used instead.</p>
<h2><a class="anchor" id="autotoc_md45"></a>
Introducing fresh blood to male and female kernels</h2>
<p >Suppose the <code>offspring_f</code> and <code>offspring_m</code> groups exist, as created in the previous section <b>Split offspring into male and female</b>.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to add your new calves to the breeding kernels from which their parents were chosen. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> cowGroups[2];</div>
<div class="line">cowGroups[0] = cows;</div>
<div class="line">cowGroups[1] = offspring_f;</div>
<div class="line"> </div>
<div class="line">cows = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, cowGroups);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> bullGroups[2];</div>
<div class="line">bullGroups[0] = bulls;</div>
<div class="line">bullGroups[1] = offspring_m;</div>
<div class="line"> </div>
<div class="line">bulls = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, bullGroups);</div>
<div class="ttc" id="agroup__shortnames_html_ga73e77b55777298bf0c2349ada77bc44b"><div class="ttname"><a href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a></div><div class="ttdeci">#define combine_groups</div><div class="ttdef"><b>Definition:</b> sim-operations.h:202</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">cows &lt;- combine.groups(c(cows,offspring_f))</div>
<div class="line">bulls &lt;- combine.groups(c(bulls,offspring_m))</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md46"></a>
Random mating with caps on number of offspring per animal</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose each cow should only have one calf this generation. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 1, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, cap1=1, n.crosses=10)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md47"></a>
Mating all females to a good male</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose the best bull in the population is the only one that will father calves under the current breeding strategy. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> bestbull = <a class="code hl_define" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a>(d, bulls, eff1, 1, <a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>); # where eff1 is an <a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> representing the marker effect set to use to calculate bvs</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bestbull, 10, 1, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_ga7359d7aff5260f43540819b922b4b8f9"><div class="ttname"><a href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a></div><div class="ttdeci">#define split_by_bv</div><div class="ttdef"><b>Definition:</b> sim-operations.h:230</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">bestbull &lt;- break.group.by.gebv(bulls, number=1, eff.set=eff1) # by default, this function uses the first effect set, so `eff.set=eff1` is optional</div>
<div class="line">offspring &lt;- make.random.crosses.between(cows, bestbull, cap1=1, n.crosses=10)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md48"></a>
Making specific chosen matings</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to cross Cow1 to Bull1, Cow2 to Bull2, and Daisy to Bull1 </td><td><div class="fragment"><div class="line">cow1_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Cow1&quot;</span>);</div>
<div class="line">cow2_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Cow2&quot;</span>);</div>
<div class="line">cow3_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Daisy&quot;</span>);</div>
<div class="line">bull1_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Bull1&quot;</span>);</div>
<div class="line">bull2_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Bull2&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][3];</div>
<div class="line">crossingPlan[0][0] = cow1_index;</div>
<div class="line">crossingPlan[0][1] = cow2_index;</div>
<div class="line">crossingPlan[0][2] = cow3_index;</div>
<div class="line">crossingPlan[1][0] = bull1_index;</div>
<div class="line">crossingPlan[1][1] = bull2_index;</div>
<div class="line">crossingPlan[1][2] = bull1_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 3, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__search_html_ga292d9c26b8d027c931a84fd09a32518f"><div class="ttname"><a href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a></div><div class="ttdeci">int gsc_get_index_of_name(const gsc_AlleleMatrix *start, const char *name)</div><div class="ttdoc">Search for a genotype with a particular name in a linked list of gsc_AlleleMatrix,...</div><div class="ttdef"><b>Definition:</b> sim-operations.c:2393</div></div>
<div class="ttc" id="agroup__shortnames_html_ga56e71147b1f8a71151ef81b4a828aa9b"><div class="ttname"><a href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a></div><div class="ttdeci">#define make_targeted_crosses</div><div class="ttdef"><b>Definition:</b> sim-operations.h:221</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.targeted.crosses(c(&quot;Cow1&quot;, &quot;Cow2&quot;, &quot;Daisy&quot;), c(&quot;Bull1&quot;, &quot;Bull2&quot;, &quot;Bull1&quot;))</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The other option is to create an input file of the following format: </p><div class="fragment"><div class="line">Cow1 Bull1</div>
<div class="line">Cow2 Bull2</div>
<div class="line">Daisy Bull1</div>
</div><!-- fragment --><p> and call <code>make_crosses_from_file</code> or <code>make.crosses.from.file</code>.</p>
<h2><a class="anchor" id="autotoc_md49"></a>
Three-Way Crosses</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to make a single Breed1/Breed2//Breed3 cross. That is, cross the F1 of a mating between Breed1 and Breed2 to Breed3. Assumes the genotypes for Breed1, Breed2, and Breed3 have those names. </td><td><div class="fragment"><div class="line">breed1_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed1&quot;</span>);</div>
<div class="line">breed2_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed2&quot;</span>);</div>
<div class="line">breed3_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed3&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][1];</div>
<div class="line">crossingPlan[0][0] = breed1_index;</div>
<div class="line">crossingPlan[1][0] = breed2_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1 = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"><span class="keywordtype">int</span>* f1_index = malloc(<span class="keyword">sizeof</span>(<span class="keywordtype">int</span>) * 1);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>(d, f1, 1, f1_index); <span class="comment">//we know this group has only one member</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Re-use crossingPlan</span></div>
<div class="line">crossingPlan[0][0] = breed3_index;</div>
<div class="line">crossingPlan[1][0] = f1_index[0];</div>
<div class="line">free(f1_index);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f3way = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_gaa7f7f1453a84746fa215fea171b80b97"><div class="ttname"><a href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a></div><div class="ttdeci">#define get_group_indexes</div><div class="ttdef"><b>Definition:</b> sim-operations.h:195</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f1 &lt;- make.targeted.crosses(c(&quot;Breed1&quot;), c(&quot;Breed2&quot;))</div>
<div class="line">f1_index &lt;- see.group.data(f1, &quot;X&quot;)</div>
<div class="line">f3way &lt;- make.targeted.crosses(c(&quot;Breed3&quot;), f1_index)</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want 25 offspring of a Breed1/Breed2//Breed3 cross. Assumes the genotypes for Breed1, Breed2, and Breed3 have those names. </td><td><div class="fragment"><div class="line">breed1_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed1&quot;</span>);</div>
<div class="line">breed2_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed2&quot;</span>);</div>
<div class="line">breed3_index = <a class="code hl_function" href="group__search.html#ga292d9c26b8d027c931a84fd09a32518f">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed3&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][1];</div>
<div class="line">crossingPlan[0][0] = breed1_index;</div>
<div class="line">crossingPlan[1][0] = breed2_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.family_size=5,</div>
<div class="line">        .will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .offspring_name_prefix=NULL,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1 = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"><span class="keywordtype">int</span>* f1_indexes = malloc(<span class="keyword">sizeof</span>(<span class="keywordtype">int</span>) * 5);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>(d, f1, 5, f1_indexes);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlanb[2][5];</div>
<div class="line">crossingPlanb[0][0] = breed3_index; crossingPlanb[0][1] = breed3_index;</div>
<div class="line">crossingPlanb[0][2] = breed3_index; crossingPlanb[0][3] = breed3_index;</div>
<div class="line">crossingPlanb[0][4] = breed3_index;</div>
<div class="line">crossingPlanb[1][0] = f1_indexes[0]; crossingPlanb[1][1] = f1_indexes[1];</div>
<div class="line">crossingPlanb[1][2] = f1_indexes[2]; crossingPlanb[1][3] = f1_indexes[3];</div>
<div class="line">crossingPlanb[1][4] = f1_indexes[4];</div>
<div class="line">free(f1_indexes);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f3way = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 5, crossingPlanb[0], crossingPlanb[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f1 &lt;- make.targeted.crosses(c(&quot;Breed1&quot;), c(&quot;Breed2&quot;), offspring=5)</div>
<div class="line"># This time, f1 is a group of multiple individuals</div>
<div class="line">f1_indexes &lt;- see.group.data(f1, &quot;X&quot;)</div>
<div class="line">f3way &lt;- make.targeted.crosses(rep(&quot;Breed3&quot;, times=length(f1_indexes)), f1_indexes, offspring=5)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md50"></a>
Tracking age and culling individuals that are too old</h2>
<p >genomicSimulation's custom labels can be used to track age (or some other known trait that can be represented with an integer for each individual).</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you only want to breed 3-year-old animals. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(1234567);</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> animals = init.group;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Create a new label to represent age, with default/at-birth value of 0.</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga5c7e03838336e41c1782e75c4f9029d6">LabelID</a> ageLabel = <a class="code hl_define" href="group__shortnames.html#gaa976db4aa919852354fc11c5c14647cf">create_new_label</a>(d, 0);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Founders are 3 years old at the beginning.</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga1c1680b1461267ccf7216e2b7ceb06dc">change_label_to</a>(d, animals, ageLabel, 3);</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> year = 0; year &lt; 10; ++year) {</div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> breedingGroup = <a class="code hl_define" href="group__shortnames.html#ga719bf50a9ff52b983e11186d11734e06">split_by_label_value</a>(d, animals, ageLabel, 3);</div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Do some breeding/selection steps as appropriate for the breeding program, eg:</span></div>
<div class="line">    GruopNum offspring = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, breedingGroup, 50, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line">    <span class="comment">// Offspring will have the default value for the label i.e. ageLabel = 0</span></div>
<div class="line">    </div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toCombine[3] = {animals, breedingGroup, offspring};</div>
<div class="line">    animals = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 3, toCombine);</div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Increase age of all by 1</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#ga53c6a459f13ae342d8e2a3ed88847a9a">change_label_by_amount</a>(d, animals, ageLabel, 1);</div>
<div class="line">}</div>
<div class="ttc" id="agroup__shortnames_html_ga1c1680b1461267ccf7216e2b7ceb06dc"><div class="ttname"><a href="group__shortnames.html#ga1c1680b1461267ccf7216e2b7ceb06dc">change_label_to</a></div><div class="ttdeci">#define change_label_to</div><div class="ttdef"><b>Definition:</b> sim-operations.h:154</div></div>
<div class="ttc" id="agroup__shortnames_html_ga53c6a459f13ae342d8e2a3ed88847a9a"><div class="ttname"><a href="group__shortnames.html#ga53c6a459f13ae342d8e2a3ed88847a9a">change_label_by_amount</a></div><div class="ttdeci">#define change_label_by_amount</div><div class="ttdef"><b>Definition:</b> sim-operations.h:155</div></div>
<div class="ttc" id="agroup__shortnames_html_ga5c7e03838336e41c1782e75c4f9029d6"><div class="ttname"><a href="group__shortnames.html#ga5c7e03838336e41c1782e75c4f9029d6">LabelID</a></div><div class="ttdeci">#define LabelID</div><div class="ttdef"><b>Definition:</b> sim-operations.h:130</div></div>
<div class="ttc" id="agroup__shortnames_html_ga719bf50a9ff52b983e11186d11734e06"><div class="ttname"><a href="group__shortnames.html#ga719bf50a9ff52b983e11186d11734e06">split_by_label_value</a></div><div class="ttdeci">#define split_by_label_value</div><div class="ttdef"><b>Definition:</b> sim-operations.h:204</div></div>
<div class="ttc" id="agroup__shortnames_html_gaa976db4aa919852354fc11c5c14647cf"><div class="ttname"><a href="group__shortnames.html#gaa976db4aa919852354fc11c5c14647cf">create_new_label</a></div><div class="ttdeci">#define create_new_label</div><div class="ttdef"><b>Definition:</b> sim-operations.h:152</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">animals &lt;- load.data(&quot;genotype-file.txt&quot;, &quot;map-file.txt&quot;)</div>
<div class="line"> </div>
<div class="line"># Create a new label to represent age, with default/at-birth value of 0.</div>
<div class="line">ageLabel &lt;- create.new.label(0L)</div>
<div class="line"> </div>
<div class="line"># Founders are 3 years old at the beginning.</div>
<div class="line">change.label.to.this(ageLabel, 3L, animals);</div>
<div class="line"> </div>
<div class="line">for (year in 0:10) {</div>
<div class="line">    breedingGroup &lt;- break.group.by.label.value(ageLabel, 3L, animals);</div>
<div class="line">    </div>
<div class="line">    # Do some breeding/selection steps as appropriate for the breeding program, eg:</div>
<div class="line">    offspring &lt;- make.random.crosses(breedingGroup, 50, 0)</div>
<div class="line">    # Offspring will have the default value for the label i.e. ageLabel = 0</div>
<div class="line">    </div>
<div class="line">    animals &lt;- combine.groups(c(animals, breedingGroup, offspring))</div>
<div class="line">    </div>
<div class="line">    # Increase age of all by 1</div>
<div class="line">    change.label.by.amount(ageLabel, 1, animals);</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to cull animals that are 12 years old or older. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toCull = <a class="code hl_define" href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a>(d, animals, ageLabel, 12, 1000000); </div>
<div class="line"><span class="comment">// some very large number for upper bound. User is responsible for the values</span></div>
<div class="line"><span class="comment">// of the custom label, so you can work out a number your script should never </span></div>
<div class="line"><span class="comment">// set the label above.</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">if</span> (toCull &gt; 0) { <span class="comment">// split_by_label_range did find some animals to cull</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, toCull);</div>
<div class="line">}</div>
<div class="ttc" id="agroup__shortnames_html_ga083d4709ce763182d5a00310b9ec459e"><div class="ttname"><a href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a></div><div class="ttdeci">#define split_by_label_range</div><div class="ttdef"><b>Definition:</b> sim-operations.h:205</div></div>
<div class="ttc" id="agroup__shortnames_html_gaed498b119b661a47824e76fece2b45df"><div class="ttname"><a href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a></div><div class="ttdeci">#define delete_group</div><div class="ttdef"><b>Definition:</b> sim-operations.h:248</div></div>
</div><!-- fragment --> or <div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toKeep = <a class="code hl_define" href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a>(d, animals, ageLabel, 0, 11); </div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, animals);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">toCull &lt;- break.group.by.label.range(ageLabel, 12, 1000000, animals)</div>
<div class="line"> </div>
<div class="line">if (toCull &gt; 0) { // split_by_label_range did find some animals to cull</div>
<div class="line">    delete.group(toCull) // delete from memory</div>
<div class="line">    rm(toCull) // remove `toCull` label from R environment.</div>
<div class="line">}</div>
</div><!-- fragment --> or <div class="fragment"><div class="line">toKeep &lt;- break.group.by.label.range(ageLabel, 0, 11, animals)</div>
<div class="line"> </div>
<div class="line">delete.group(animals)</div>
<div class="line">rm(animals)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md51"></a>
Species with differing male and female recombination rates</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you only want to simulate a cross in a species with differing male and female recombination maps. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(7654321);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> female_map = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(<span class="stringliteral">&quot;fmap.txt&quot;</span>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> male_map = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(<span class="stringliteral">&quot;mmap.txt&quot;</span>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> female_pop = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(<span class="stringliteral">&quot;fgenos.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> male_pop = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(<span class="stringliteral">&quot;mgenos.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, female_pop, male_pop, 1, 0, 0, female_map, male_map, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">female_map &lt;- load.map(&quot;fmap.txt&quot;)</div>
<div class="line">male_map &lt;- load.map(&quot;mmap.txt&quot;)</div>
<div class="line">female_pop &lt;- load.genotypes(&quot;fgenos.txt&quot;)</div>
<div class="line">male_pop &lt;- load.genotypes(&quot;mgenos.txt&quot;) </div>
<div class="line"> </div>
<div class="line">offspring &lt;- make.random.crosses.between(female_pop, male_pop, 1, map1=female_map, map2=male_map)</div>
</div><!-- fragment --> </td></tr>
</table>
<h1><a class="anchor" id="autotoc_md52"></a>
Selection</h1>
<p >Templates in this section assume you have a group <code>target</code> from which you want to select a subset of 'good' genotypes. Scripts are provided for selecting in different ways using the 'custom selection interface'. This interface is basically the existence the function <code>make.group</code>, allowing users to use any kind of scripting/data manipulation to choose the indexes of the genotypes they want to select.</p>
<p >Templates will be formatted as functions, so that selecting the 'good' genotypes out of <code>target</code> is simply a matter of calling that section's selection function.</p>
<p >Note the concept of the 'custom selection interface' is mainly appropriate to the R version of genomicSimulation. Due to C's lack of vector manipulation/sorting functionality compared to R, implementing these selection methods in C is less about scripting and more about knowing how to sort/loop through datasets efficiently in C. The writer of this page isn't going to bother to write up templates for all of these selection tasks in C. If the task is popular enough it will probably be written up as an easier-to-use library function instead of a template, anyway.</p>
<h2><a class="anchor" id="autotoc_md53"></a>
Manually select on true breeding value</h2>
<p >The template selection template. This does the exact same thing as <code>split_by_bv</code> / <code>break.group.by.gebv</code>, but using the 'custom selection method interface'. Hopefully this aids in understanding the concept.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to select the top 10 genotypes in <code>target</code> according to their true breeding value as calculated by the internal breeding value calculator. </td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> select_10_with_best_GEBV(<a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> group) {</div>
<div class="line">  <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> group_size = <a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>( d, group );</div>
<div class="line">  <span class="keywordtype">int</span> group_indexes[group_size];</div>
<div class="line">  memset(group_indexes, 0, <span class="keyword">sizeof</span>(<span class="keywordtype">int</span>)*group_size);</div>
<div class="line">  <a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>( d, group, group_size, group_indexes);</div>
<div class="line">  <span class="keywordtype">double</span> group_bvs[group_size];</div>
<div class="line">  memset(group_bvs, 0, <span class="keyword">sizeof</span>(<span class="keywordtype">double</span>)*group_size);</div>
<div class="line">  <a class="code hl_define" href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a>( d, group, group_size, group_bvs);</div>
<div class="line"> </div>
<div class="line"><span class="preprocessor">  # A pretty inefficient way of finding the top 10 scores and their corresponding</span></div>
<div class="line"><span class="preprocessor">  # indexes, but it will do for this example.</span></div>
<div class="line">  <span class="keywordtype">double</span> min_score = group_bvs[0];</div>
<div class="line">  <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; group_size; ++i) {</div>
<div class="line">    <span class="keywordflow">if</span> (group_bvs[i] &lt; min_score) {</div>
<div class="line">      min_score = group_bvs[i];</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line"> </div>
<div class="line">  <span class="keywordtype">int</span> top_individuals[10];</div>
<div class="line">  <span class="keywordtype">int</span> currentTopIndex = 0;</div>
<div class="line">  <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; 10; ++j) {</div>
<div class="line">    currentTopIndex = 0;</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; group_size; ++i) {</div>
<div class="line">      <span class="keywordflow">if</span> (group_bvs[i] &gt; group_bvs[currentTopIndex]) {</div>
<div class="line">        currentTopIndex = i;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">    top_individuals[j] = group_indexes[currentTopIndex];</div>
<div class="line">    group_bvs[currentTopIndex] = min_score;</div>
<div class="line">  }</div>
<div class="line"> </div>
<div class="line">  <span class="keywordflow">return</span> <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 10, top_individuals);</div>
<div class="line">}</div>
<div class="ttc" id="agroup__shortnames_html_ga0f4fcc00561afa1f820d7f11a02be1c7"><div class="ttname"><a href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a></div><div class="ttdeci">#define get_group_size</div><div class="ttdef"><b>Definition:</b> sim-operations.h:192</div></div>
<div class="ttc" id="agroup__shortnames_html_gaa68a8ebf286aa6192948ab56851b125f"><div class="ttname"><a href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a></div><div class="ttdeci">#define get_group_bvs</div><div class="ttdef"><b>Definition:</b> sim-operations.h:197</div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">select.10.best.GEBV &lt;- function(group) {</div>
<div class="line">  info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                     GEBV=see.group.data(group,&quot;BV&quot;))</div>
<div class="line"> </div>
<div class="line">  selected_group &lt;- make.group(info[order(info$GEBV, decreasing=TRUE),]$Index[1:10])</div>
<div class="line">  return( selected_group )</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md54"></a>
Select on phenotype (simulated with a given heritability)</h2>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to select the top 10 genotypes in <code>target</code> according to a phenotype simulated with a certain $h^2$ heritability value. </td><td><div class="fragment"><div class="line">select.top.10.phenotypes &lt;- function(group, heritability) {</div>
<div class="line">  info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                     GEBV=see.group.data(group,&quot;BV&quot;))</div>
<div class="line"> </div>
<div class="line">  # simulate phenotype = genotype + environmental variation</div>
<div class="line">  # using normally distributed Ve and heritability H^2 = (Ve + Vg)/Vg</div>
<div class="line">  Vg &lt;- var(info$GEBV)</div>
<div class="line">  Ve &lt;- Vg/heritability - Vg</div>
<div class="line">  info$Pheno &lt;- info$GEBV + rnorm(length(info$GEBV), mean=0, sd = sqrt(Ve))</div>
<div class="line"> </div>
<div class="line">  # Select those with the top phenotype</div>
<div class="line">  selected &lt;- make.group(info[order(info$Pheno, decreasing=TRUE),]$Index[1:10])</div>
<div class="line">  return( selected )</div>
<div class="line"> </div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md55"></a>
Select on a qualitative trait</h2>
<p >If you have a trait with only a small number of possible outcomes - perhaps a marker for which each individual can be homozygous, heterozygous, or carry no copies.</p>
<table class="doxtable">
<tr>
<th>Task </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>Suppose you want to select on a trait that has possible values Your effect file looks like: <div class="fragment"><div class="line">TARGETGENE T 1</div>
</div><!-- fragment --> </td><td><p class="starttd"></p>
<p class="endtd"></p>
</td></tr>
</table>
<h2><a class="anchor" id="autotoc_md56"></a>
Select on a qualitative trait, then a quantitative trait.</h2>
<p >Suppose you have an effect file for a quantitative trait, called <code>eff-large.txt</code>, as well as a one-line effect file for a qualitative trait, as described in the above section <b>Select on a qualitative trait</b>, named <code>eff-small.txt</code>.</p>
<h2><a class="anchor" id="autotoc_md57"></a>
Select on an index weighting two traits.</h2>
<p >(this list of examples will continue to be expanded...) </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
