<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.3"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>genomicSimulationC: Features and Templates Guide</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">genomicSimulationC<span id="projectnumber">&#160;0.2.6</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.3 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search",'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div><div class="header">
  <div class="headertitle"><div class="title">Features and Templates Guide </div></div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md26">Setting Up</a><ul><li class="level2"><a href="#autotoc_md27">Input file formats:</a><ul><li class="level3"><a href="#autotoc_md28">Genetic map files</a></li>
<li class="level3"><a href="#autotoc_md29">Genotype matrix files</a></li>
<li class="level3"><a href="#autotoc_md30">Marker effect files</a></li>
<li class="level3"><a href="#autotoc_md31">Loading from other file formats</a></li>
</ul>
</li>
<li class="level2"><a href="#autotoc_md32">Setting up the simulation and the random number generator</a></li>
<li class="level2"><a href="#autotoc_md33">Load a genetic map and a set of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md34">Load a genetic map and genotype file, specifying file format</a></li>
<li class="level2"><a href="#autotoc_md35">Load a genetic map and several sets of founder genotypes</a></li>
<li class="level2"><a href="#autotoc_md36">Load a genetic map, a set of founder genotypes, and a set of additive trait effects</a></li>
<li class="level2"><a href="#autotoc_md37">Load another set of additive trait effects</a></li>
<li class="level2"><a href="#autotoc_md38">Load another linkage map</a></li>
<li class="level2"><a href="#autotoc_md39">Simulating a breeding program</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md40">Crossing &amp; Generating New Genotypes: Plant-themed</a><ul><li class="level2"><a href="#autotoc_md41">Single seed descent and selfing</a></li>
<li class="level2"><a href="#autotoc_md42">Creating halfsib or fullsib families</a></li>
<li class="level2"><a href="#autotoc_md43">Updating marker effects</a></li>
<li class="level2"><a href="#autotoc_md44">Trials and locations</a></li>
<li class="level2"><a href="#autotoc_md45">Backcrossing</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md46">Crossing &amp; Generating New Genotypes: Animal-themed</a><ul><li class="level2"><a href="#autotoc_md47">Split offspring into male and female</a></li>
<li class="level2"><a href="#autotoc_md48">Introducing fresh blood to male and female kernels</a></li>
<li class="level2"><a href="#autotoc_md49">Tracking age and culling individuals that are too old</a></li>
<li class="level2"><a href="#autotoc_md50">Random mortality</a></li>
<li class="level2"><a href="#autotoc_md51">Culling or fitness-based animal loss</a></li>
<li class="level2"><a href="#autotoc_md52">Random mating with caps on number of offspring per animal</a></li>
<li class="level2"><a href="#autotoc_md53">Mating all females to a good male</a></li>
<li class="level2"><a href="#autotoc_md54">Random mating between the best performing individuals in a group</a></li>
<li class="level2"><a href="#autotoc_md55">Making specific chosen matings</a></li>
<li class="level2"><a href="#autotoc_md56">Three-Way Crosses</a></li>
<li class="level2"><a href="#autotoc_md57">Species with differing male and female recombination rates</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md58">Selection</a><ul><li class="level2"><a href="#autotoc_md59">Manually select on true breeding value</a></li>
<li class="level2"><a href="#autotoc_md60">Select on phenotype (simulated with a given heritability)</a></li>
<li class="level2"><a href="#autotoc_md61">Select on an index weighting two traits.</a></li>
<li class="level2"><a href="#autotoc_md62">Select on a qualitative trait</a></li>
<li class="level2"><a href="#autotoc_md63">Select on a qualitative trait, then a quantitative trait.</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p ><a class="anchor" id="md_Templates"></a></p>
<p >This page provides a set of templates for implementing common breeding project actions in genomicSimulationC and its R counterpart.</p>
<h1><a class="anchor" id="autotoc_md26"></a>
Setting Up</h1>
<h2><a class="anchor" id="autotoc_md27"></a>
Input file formats:</h2>
<p >genomicSimulation has three types of input files: a genotype matrix, a genetic map, and a list of marker effects. At minimum, a genotype matrix is required to run simulations, because genomicSimulation does not generate you founder genotypes.</p>
<p >The list of genetic markers tracked by the simulation is fixed after the simulation is set up (until you reset the simulation (R)/create another SimData object (C)). No other markers can be added and none can be removed by input files loaded after the simulation is set up. If a map file is the first file loaded, or the simulation is set up using the <code>load_data_files</code> (C)/<code>load.data</code> (R) function, then the markers listed in that genetic map file are the ones tracked by the simulation. If <code>load_genotypefile</code> (C)/<code>load.genotypes</code> (R) is the first function run, before loading a map, then the simulation tracks only the list of genetic markers from the genotype matrix, and the first genetic map in the simulation is an invented map with equally-spaced markers along a single chromosome.</p>
<p >The input file formats are described below.</p>
<h3><a class="anchor" id="autotoc_md28"></a>
Genetic map files</h3>
<p >The simplest genetic map file is formatted as follows: </p><div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --><p> Other valid genetic map files might include: </p><div class="fragment"><div class="line">chr marker pos</div>
<div class="line">1A 1243509 173.2</div>
<div class="line">1A 2350898 462.2</div>
<div class="line">1B 4360992 32.009</div>
<div class="line">2A 1243556 243.5</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">gene 10 3.24</div>
<div class="line">othergene 10 8.3e-1</div>
<div class="line">etc 15 1.203e2</div>
</div><!-- fragment --><p >The header line is optional. If there is a header line, the three columns "marker" "chr" and "pos" may be rearranged into any ordering. If the header is not provided, the order is assumed to be marker name in the first column, followed by chromosome in the second column, followed by position in the third.</p>
<p >Regarding marker names: For all genomicSimulation features to work, genetic markers must have names. There is no issue with purely numeric names.</p>
<p >Regarding chromosomes/linkage group: any alphanumeric combination may be used to denote a chromosome/linkage group, for example '9' or '1A'. There is no limit on the number of unique chromosomes/linkage groups in the map.</p>
<p >Regarding marker position: this represents the position in centimorgans of each marker along its linkage group. (Distance of 1 cM = expected probability of 0.01 chromosomal crossovers in that range.)</p>
<p >The cells in the map file may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file, but therefore marker names/linkage group names cannot contain spaces, tabs, or commas.</p>
<p >The order in which genetic markers are presented in the file does not matter. If a marker is duplicated in the file, it will be recorded twice as two separate markers in the same position and with the same name. However, because loading genotypes depends on searching for markers by name, if two markers have the same name, genotypes may not be loaded correctly. Therefore it is suggested that all markers have unique names.</p>
<h3><a class="anchor" id="autotoc_md29"></a>
Genotype matrix files</h3>
<p >The simplest genotype matrix file is formatted as follows: </p><div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  CC  AA</div>
</div><!-- fragment --><p> where G01, G02, ..., are names of the founder genotypes; m1, m2, ..., are the genetic markers; and entries in the matrix are the alleles that the founder genotypes have at those markers.</p>
<p >Other valid genotype matrix files might include: </p><div class="fragment"><div class="line"> m100, m101, m102</div>
<div class="line">cand1,0,0,1</div>
<div class="line">cand2,1,2,2</div>
<div class="line">cand4,2,1,2</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">marker1 T/T T/T T/T T/A T/T A/T</div>
<div class="line">marker3 T/T T/T T/A T/A T/T T/T</div>
<div class="line">marker2 A/A A/A A/A A/A T/T A/A</div>
</div><!-- fragment --><p >The genotype matrix can be row-major or column-major (that is, the genetic markers may be rows, or columns). The program will determine the orientation by attempting to match row and column headers with names of markers tracked by the simulation (extracted from a genetic map file). If the simulation does not yet have a list of tracked markers (that is, no genetic map has been simultaneously or previously provided), then it defaults to assuming that rows represent genetic markers. <br  />
</p>
<p >The order in which genetic markers are presented in the file does not matter. Candidate genotypes, however, will be saved internally in the simulation in the order that they appear in the file. Candidate genotype names do not need to be unique. Candidate genotype names are optional, if candidates are stored as columns. genomicSimulation cannot read a gentoype matrix with no row headers, so if candidates are stored as rows, then they are required to have names.</p>
<p >Genetic markers' names are required.</p>
<p >When there are both row and column headers, the first/corner cell (the one that contained the value "name" in the first example above) can be deleted or can contain any text. Its value will be ignored.</p>
<p >The table cells in the genotype matrix file may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file.</p>
<p >There are several options for how the alleles at each marker can be presented. All allele pair cells in the same genotype matrix must be in the same format. The format of the allele pairs in the genotype matrix is automatically detected. There are four acceptable formats for allele pairs in the genotype matrix:</p>
<ol type="1">
<li>Any pair of characters (eg. "AA", "TA", "nW"). Each character is an allele. This is a format that specifies allele phase (ie. "AT" and "TA" are different genotypes).</li>
<li>Any pair of characters, separated by a forwards slash "/" character (eg. "A/A", "T/A", "n/W"). The two characters either side of the slash are the alleles. This is a format that specifies allele phase (ie. "AT" and "TA" are different genotypes).</li>
<li>Alternate allele counts ("0", "1", "2"). This is a format that does not specify allele phase, so phase of heterozygotes will be randomised when loaded. The counts represent the number of copies of the alternate allele (stored inside genomicSimulation as "A", while the reference allele is stored as "T"). "0" then corresponds to "TT", "2" to "AA", and "1" will be randomised as either "TA" or "AT". Corresponding marker effect files for calculating breeding values must use allele "A" consistently to represent the alternate allele, and "T" to represent the reference. (alternatively, <code>change_allele_symbol</code> (C)/<code>change.allele.symbol</code> (R) can be used after loading to change T and/or A to other symbols.)</li>
<li>(subset of) IUPAC encodings of DNA bases ("G", "A", "T", "C", "R", "Y", "M", "K", "S", "W", "N"). This is a format that does not specify allele phase, so phase of heterozygotes will be randomised when loaded. See below for meanings of the IUPAC encoding symbols that genomicSimulation can parse.</li>
</ol>
<table class="doxtable">
<tr>
<th>IUPAC symbol </th><th>genomicSimulation genotype </th></tr>
<tr>
<td>G</td><td>GG </td></tr>
<tr>
<td>A</td><td>AA </td></tr>
<tr>
<td>T</td><td>TT </td></tr>
<tr>
<td>C</td><td>CC </td></tr>
<tr>
<td>R</td><td>GA or AG </td></tr>
<tr>
<td>Y</td><td>TC or CT </td></tr>
<tr>
<td>M</td><td>AC or CA </td></tr>
<tr>
<td>K</td><td>GT or TG </td></tr>
<tr>
<td>S</td><td>GC or CG </td></tr>
<tr>
<td>W</td><td>AT or TA </td></tr>
<tr>
<td>N</td><td>\0\0 (nulls represent unknown genotype) </td></tr>
</table>
<p ><b>Note you might have a genotype matrix of only "AA", "AT", and "TT". This uses "alternate allele counts"-style encoding (like format 3) but presents it in a format that looks like pairs of alleles (format 1).</b> genomicSimulation expects allele pair encodings to include haplotype phase, (that is, to have four possible values for genotypes of two alleles, not three: eg. "AA", "AT", "TA", and "TT" instead of just "AA", "AT", "TT").</p>
<p >Two options for loading a dataset with non-phased "AA"/"AT"/"TT" allele pairs are:</p>
<ul>
<li>Use “haplotyping”/”haplotype phasing”/”haplotype inference” software to infer whether heterozygotes are "AT" or "TA", before loading into genomicSimulation.</li>
<li>Find-and-replace "TT" with "0", "AT" with "1", and "AA" with "2" before loading into genomicSimulation. genomicSimulation will then randomise the phase of each haplotype.</li>
</ul>
<h3><a class="anchor" id="autotoc_md30"></a>
Marker effect files</h3>
<p >Loading marker effect file(s) is optional for running the simulation. The simplest marker effect file is formatted as follows: </p><div class="fragment"><div class="line">marker allele eff</div>
<div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --><p> Other valid marker effect files might include: </p><div class="fragment"><div class="line">marker eff allele</div>
<div class="line">m1243509 0.1 A</div>
<div class="line">m1243509 -0.1 T</div>
<div class="line">m2350898 0.15 T</div>
<div class="line">m2350898 -0.1 A</div>
</div><!-- fragment --><p> or </p><div class="fragment"><div class="line">specialgene G 1.0</div>
</div><!-- fragment --><p >A single marker effect file represents the allele effects for one trait. Multiple files like this can be loaded in order to calculate breeding values for multiple traits separately.</p>
<p >The header line is optional. If there is a header line, the three columns "marker" "allele" and "eff" may be rearranged into any ordering. If the header is not provided, the order is assumed to be marker name in the first column, followed by allele, followed by additive effect value.</p>
<p >The order of the rows in the file does not matter. If multiple rows exist in the file for the same marker name/allele combination, only the last row's additive effect value will be saved.</p>
<p >Regarding marker names: Only rows in this file whose marker name matches a marker name from a previously- or simultaneously-loaded map file or genotype matrix file will be loaded.</p>
<p >Regarding allele: This column should be the allele (non-space character, eg "A") that this effect value corresponds to.</p>
<p >Regarding allele effect: This column should be a decimal representing the additive effect value of the same-row allele for the same-row marker. It can be positive or negative or zero, and can be represented by an integer, a decimal, or scientific notation.</p>
<p >Cells may be separated by spaces, tabs, commas, or any combination thereof. Cell spacers do not need to be consistent across the file.</p>
<h3><a class="anchor" id="autotoc_md31"></a>
Loading from other file formats</h3>
<p >Updates to expand the range of allowed input formats are coming soon.</p>
<h2><a class="anchor" id="autotoc_md32"></a>
Setting up the simulation and the random number generator</h2>
<p >In C, you should manually set the seed for the random number generator when you create the SimData object, before loading data. The R version borrows R's own random number generator so has no need for setting a seed.</p>
<table class="doxtable">
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="ttc" id="agroup__shortnames_html_gaa9b2f128bb156d36d9334281b158c56a"><div class="ttname"><a href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a></div><div class="ttdeci">#define SimData</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00264">sim-operations.h:264</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gacbedd673ae83f70154d663852acfea70"><div class="ttname"><a href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a></div><div class="ttdeci">#define create_empty_simdata</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00279">sim-operations.h:279</a></div></div>
</div><!-- fragment --> </td><td>N/A </td></tr>
</table>
<p >In the remainder of this page, the variable <code>d</code> in genomicSimulationC scripts will always represent a pointer to the simulation object holding the current simulation's data.</p>
<p >In the R version, where only a single simulation data object exists at a time, the simulation is automatically set up at the first call to <code>load.data</code> or <code>load.genotypes</code> or <code>load.map</code>, and is reset whenever the functions <code>clear.simdata</code> or <code>load.data</code> are called. (Unlike the other loading functions, the function <code>load.data</code> can only be used as the first call to set up a simulation, as it deletes and clears any previous simulation data when it is run).</p>
<h2><a class="anchor" id="autotoc_md33"></a>
Load a genetic map and a set of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Example Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders = init.group;</div>
<div class="ttc" id="agroup__shortnames_html_ga1b342f7327286629391ccbad90ea19e4"><div class="ttname"><a href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a></div><div class="ttdeci">#define DETECT_FILE_FORMAT</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00285">sim-operations.h:285</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga2e10fe6c17f65cefc887a7b8868d9017"><div class="ttname"><a href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a></div><div class="ttdeci">#define load_data_files</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00284">sim-operations.h:284</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gab9139b22223d33aa0a762d3ec00b3072"><div class="ttname"><a href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a></div><div class="ttdeci">#define MultiIDSet</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00253">sim-operations.h:253</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gadb2ef9b75828b244e0b3e1cb3680d7ca"><div class="ttname"><a href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a></div><div class="ttdeci">#define GroupNum</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00245">sim-operations.h:245</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">init &lt;- load.data(&quot;genotype-file.txt&quot;, &quot;map-file.txt&quot;)</div>
<div class="line">founders = init$groupNum</div>
</div><!-- fragment --> </td></tr>
</table>
<p >As these functions run, they will print out debug information, including whether they detect header rows in the file, how many rows they read from the file, and how many rows had to be skipped for malformed/malformatted data. You can read this information to check that the input files are being read as expected.</p>
<p >The first list of genetic markers loaded into a simulation object, whether that list comes from a genetic map file or from the headers of a genotype matrix file, is the total list of all genetic markers that will be tracked by this simulation object. During the lifetime of the simulation, the list of markers that are being tracked cannot be changed or added to.</p>
<p >If, as in the table above, a genetic map and a genotype matrix are loaded in the same command (eg. <code>load_data_files</code> (C)/<code>load.data</code> (R)), then the genetic map takes precedence for setting the genetic markers that will be tracked. All markers that appear in the genetic map file will be tracked, including those that are missing from the genotype matrix (the founders from the genotype matrix will have blank/unknown alleles for those missing markers). Any markers that appear in the genotype matrix but not the genetic map will be ignored.</p>
<p >The line order of the genotype matrix and the genetic map does not need to match. genomicSimulation will match genetic markers using their names.</p>
<p >You can use the <code>get_group_</code> family of functions (C)/the <code>see.group.data</code> function (R) to confirm data is correctly loaded:</p>
<table class="doxtable">
<caption>Accessing simulation data to check information is correctly loaded</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>For users confident enough to use an iterator, the following method is recommended: <div class="fragment"><div class="line"><span class="comment">// Say we want to save some information about each founder to a logging file</span></div>
<div class="line">FILE* logfile = fopen(<span class="stringliteral">&quot;log.txt&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Set up an iterator that will step through all members of the &quot;founders&quot; group</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gab51c8cf307b7e6baaf9af222d1406ea3">BidirectionalIterator</a> it = <a class="code hl_define" href="group__shortnames.html#gaf6e8c60198060f13a3ade462c56b9d77">create_bidirectional_iter</a>(d, founders); <span class="comment">// &lt;-- Iterator setup</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga86d2d8f2c416cdbd37c3f99e5d80c4e3">GenoLocation</a> loc = <a class="code hl_define" href="group__shortnames.html#ga60fa7078ac84231afe8ec7cd9cf850ef">set_bidirectional_iter_to_start</a>(&amp;it);           <span class="comment">// &lt;--/</span></div>
<div class="line"><span class="keywordflow">while</span> (<a class="code hl_define" href="group__shortnames.html#ga74a123d16204f5ca9d85342f13d3ae9a">IS_VALID_LOCATION</a>(loc)) {                                   <span class="comment">// &lt;-- Iterator end condition</span></div>
<div class="line">    <span class="comment">// access some data on this founder. A range of `get_` functions are available</span></div>
<div class="line">    <span class="keywordtype">char</span>* this_founders_name = <a class="code hl_define" href="group__shortnames.html#ga555a72fbd7fef69ac4f057b1be64fb49">get_name</a>(loc);</div>
<div class="line">    <span class="keywordtype">char</span>* this_founders_alleles = <a class="code hl_define" href="group__shortnames.html#ga3b45650d84122a9a86af6759a5b36ab8">get_alleles</a>(loc);</div>
<div class="line">    <span class="comment">// save some of that data, perhaps</span></div>
<div class="line">    fprintf(logfile, <span class="stringliteral">&quot;%i %s\n&quot;</span>, <a class="code hl_define" href="group__shortnames.html#gaac44e9d83033fd5c5e4e449c0c23959b">get_id</a>(loc).<span class="keywordtype">id</span>, this_founders_name);</div>
<div class="line">    </div>
<div class="line">    loc = <a class="code hl_define" href="group__shortnames.html#ga78e7dac2c4c329adbd048de304ebb498">next_forwards</a>(&amp;it);                                      <span class="comment">// &lt;-- Iterator advances to next founder</span></div>
<div class="line">}</div>
<div class="line">fclose(logfile);</div>
<div class="ttc" id="agroup__shortnames_html_ga3b45650d84122a9a86af6759a5b36ab8"><div class="ttname"><a href="group__shortnames.html#ga3b45650d84122a9a86af6759a5b36ab8">get_alleles</a></div><div class="ttdeci">#define get_alleles</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00298">sim-operations.h:298</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga555a72fbd7fef69ac4f057b1be64fb49"><div class="ttname"><a href="group__shortnames.html#ga555a72fbd7fef69ac4f057b1be64fb49">get_name</a></div><div class="ttdeci">#define get_name</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00296">sim-operations.h:296</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga60fa7078ac84231afe8ec7cd9cf850ef"><div class="ttname"><a href="group__shortnames.html#ga60fa7078ac84231afe8ec7cd9cf850ef">set_bidirectional_iter_to_start</a></div><div class="ttdeci">#define set_bidirectional_iter_to_start</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00291">sim-operations.h:291</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga74a123d16204f5ca9d85342f13d3ae9a"><div class="ttname"><a href="group__shortnames.html#ga74a123d16204f5ca9d85342f13d3ae9a">IS_VALID_LOCATION</a></div><div class="ttdeci">#define IS_VALID_LOCATION</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00267">sim-operations.h:267</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga78e7dac2c4c329adbd048de304ebb498"><div class="ttname"><a href="group__shortnames.html#ga78e7dac2c4c329adbd048de304ebb498">next_forwards</a></div><div class="ttdeci">#define next_forwards</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00293">sim-operations.h:293</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga86d2d8f2c416cdbd37c3f99e5d80c4e3"><div class="ttname"><a href="group__shortnames.html#ga86d2d8f2c416cdbd37c3f99e5d80c4e3">GenoLocation</a></div><div class="ttdeci">#define GenoLocation</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00265">sim-operations.h:265</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaac44e9d83033fd5c5e4e449c0c23959b"><div class="ttname"><a href="group__shortnames.html#gaac44e9d83033fd5c5e4e449c0c23959b">get_id</a></div><div class="ttdeci">#define get_id</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00301">sim-operations.h:301</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gab51c8cf307b7e6baaf9af222d1406ea3"><div class="ttname"><a href="group__shortnames.html#gab51c8cf307b7e6baaf9af222d1406ea3">BidirectionalIterator</a></div><div class="ttdeci">#define BidirectionalIterator</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00268">sim-operations.h:268</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaf6e8c60198060f13a3ade462c56b9d77"><div class="ttname"><a href="group__shortnames.html#gaf6e8c60198060f13a3ade462c56b9d77">create_bidirectional_iter</a></div><div class="ttdeci">#define create_bidirectional_iter</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00289">sim-operations.h:289</a></div></div>
</div><!-- fragment --> Otherwise, the alternate and slightly slower method uses the <code>get_group_</code> family of functions, which are more similar to the R version's <code>see.group.data</code>: <div class="fragment"><div class="line"><span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> num_founders = <a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>(d, founders);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Get the names of every founder</span></div>
<div class="line"><span class="keywordtype">char</span>* founder_names[num_founders]; <span class="comment">// VLA, can replace with a known maximum length or a heap allocation if your C version does not support VLAs</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga2c52621edd21207fdb8660808976bdd3">get_group_names</a>(d, founders, num_founders, founder_names);</div>
<div class="line"><span class="comment">// names will now be stored in the founder_names array</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Get the allele sequence (a character vector of length 2*{num markers}) of every founder</span></div>
<div class="line"><span class="keywordtype">char</span>* founder_alleles[num_founders];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga39446372b6bcee2cdf6830715ad670a4">get_group_genes</a>(d, founders, num_founders, founder_alleles);</div>
<div class="line"><span class="comment">// genotypes will now be stored in the founder_alleles array. Pairs of alleles for the same genetic marker are consecutive. </span></div>
<div class="ttc" id="agroup__shortnames_html_ga0f4fcc00561afa1f820d7f11a02be1c7"><div class="ttname"><a href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a></div><div class="ttdeci">#define get_group_size</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00311">sim-operations.h:311</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga2c52621edd21207fdb8660808976bdd3"><div class="ttname"><a href="group__shortnames.html#ga2c52621edd21207fdb8660808976bdd3">get_group_names</a></div><div class="ttdeci">#define get_group_names</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00312">sim-operations.h:312</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga39446372b6bcee2cdf6830715ad670a4"><div class="ttname"><a href="group__shortnames.html#ga39446372b6bcee2cdf6830715ad670a4">get_group_genes</a></div><div class="ttdeci">#define get_group_genes</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00315">sim-operations.h:315</a></div></div>
</div><!-- fragment --> In the C version, you can also peek at internal data structures, like the list of stored genetic markers. This is not possible in the R version. <div class="fragment"><div class="line"><span class="comment">// (The order of the markers in the genotypes from `get_alleles` &amp; `get_group_genes` is:</span></div>
<div class="line"><span class="keywordtype">char</span>** marker_names_in_order = d-&gt;genome.marker_names;</div>
<div class="line"><span class="comment">// and there are</span></div>
<div class="line">unsined <span class="keywordtype">int</span> num_markers = d-&gt;genome.n_markers;</div>
<div class="line"><span class="comment">// of the markers.)</span></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line"># See the stored genotypes, as a (markers x founders) matrix</div>
<div class="line"># The markers may be reordered compared to the input genotype matrix file.</div>
<div class="line"># The markers will be ordered according to their order in the internal genetic map</div>
<div class="line">genomatrix &lt;- see.group.gene.data(founders)</div>
<div class="line"> </div>
<div class="line"># Other potential checks:</div>
<div class="line">founder.names &lt;- see.group.data(founders, &quot;N&quot;)</div>
<div class="line">founder.genotypes.as.strings &lt;- see.group.data(founders, &quot;G&quot;) # &lt;-- like concatenating the columns of &quot;genomatrix&quot;</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md34"></a>
Load a genetic map and genotype file, specifying file format</h2>
<p >By default, genomicSimulation attempts to detect the format of input files as it loads them. When a file loading command is run, genomicSimulation prints out its assumptions about the format of the file.</p>
<p >If the printed logs are incorrect about the layout of your map file or effect file, the most likely cause is a mis-spelling in the header row of the file. If the printed logs are incorrect about the size of the table in your map file or effect file, the issue may be that the numbers cannot be detected as numbers, or that there are issues with the column spacers/line breaks that cause there to be more than three entries to a row.</p>
<p >If the printed logs are incorrect about the layout or format of your genotype matrix file, you can manually specify the file format for genomicSimulation, as follows.</p>
<table class="doxtable">
<tr>
<th>Example Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gae2dfa9fab7307781bef01124e32e7495">FileFormatSpec</a> mformat = <a class="code hl_define" href="group__shortnames.html#ga8bc4f8369ef14861b409317322046905">define_matrix_format_details</a>(<a class="code hl_enumvalue" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8fa3587ddd94a959fe316d38be7023abb30">GSC_TRUE</a>,<a class="code hl_enumvalue" href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8fa3587ddd94a959fe316d38be7023abb30">GSC_TRUE</a>,<a class="code hl_enumvalue" href="group__structs.html#gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c">GSC_GENOTYPECELLSTYLE_PAIR</a>);</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, mformat);</div>
<div class="ttc" id="agroup__shortnames_html_ga8bc4f8369ef14861b409317322046905"><div class="ttname"><a href="group__shortnames.html#ga8bc4f8369ef14861b409317322046905">define_matrix_format_details</a></div><div class="ttdeci">#define define_matrix_format_details</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00287">sim-operations.h:287</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gae2dfa9fab7307781bef01124e32e7495"><div class="ttname"><a href="group__shortnames.html#gae2dfa9fab7307781bef01124e32e7495">FileFormatSpec</a></div><div class="ttdeci">#define FileFormatSpec</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00286">sim-operations.h:286</a></div></div>
<div class="ttc" id="agroup__structs_html_gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c"><div class="ttname"><a href="group__structs.html#gga169da3c2b32a6f9afd92e2e0801ad741a7ace3d1df4d9466ce2fdac1f1b2e678c">GSC_GENOTYPECELLSTYLE_PAIR</a></div><div class="ttdeci">@ GSC_GENOTYPECELLSTYLE_PAIR</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00929">sim-operations.h:929</a></div></div>
<div class="ttc" id="asim-operations_8h_html_a55485f3a8263f660077a88c6a7d05c8fa3587ddd94a959fe316d38be7023abb30"><div class="ttname"><a href="sim-operations_8h.html#a55485f3a8263f660077a88c6a7d05c8fa3587ddd94a959fe316d38be7023abb30">GSC_TRUE</a></div><div class="ttdeci">@ GSC_TRUE</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00095">sim-operations.h:95</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">mformat &lt;- define.matrix.format.details(has.header=TRUE, markers.as.rows=TRUE, cell.style=&quot;Pair&quot;)</div>
<div class="line">init &lt;- load.data(allele.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;,format=mformat)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >It is not required that you manually specify all details of the file format. Whatever details of the file format are not manually specified will be automatically detected.</p>
<p >See the documentation for the function <code>define_matrix_format_details</code> (C)/<code>define.matrix.format.details</code> (R) for more information.</p>
<h2><a class="anchor" id="autotoc_md35"></a>
Load a genetic map and several sets of founder genotypes</h2>
<table class="doxtable">
<tr>
<th>Example Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">    G01 G02 G03 G04 G05</div>
<div class="line">m1  TT  TT  TT  TA  TT</div>
<div class="line">m3  TT  TT  TA  TA  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Genotype file 2:</p><ul>
<li>File location: genotype-file2.txt</li>
<li>File contents: <div class="fragment"><div class="line"> G06</div>
<div class="line">m1,AT</div>
<div class="line">m2,AA</div>
<div class="line">m3,TT</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders_a = init.group;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders_b = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(d, <span class="stringliteral">&quot;genotype-file2.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_ga23a7a92af469dfbba4905bb5c9c90a9b"><div class="ttname"><a href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a></div><div class="ttdeci">#define load_genotypefile</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00283">sim-operations.h:283</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">init &lt;- load.data(&quot;genotype-file.txt&quot;, &quot;map-file.txt&quot;)</div>
<div class="line">founders_a &lt;- init$groupNum</div>
<div class="line">founders_b &lt;- load.genotypes(&quot;genotype-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md36"></a>
Load a genetic map, a set of founder genotypes, and a set of additive trait effects</h2>
<table class="doxtable">
<tr>
<th>Example Input Files </th><th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><p class="starttd">Genotype file/Allele file:</p><ul>
<li>File location: genotype-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">name    G01 G02 G03 G04 G05 G06</div>
<div class="line">m1  TT  TT  TT  TA  TT  AT</div>
<div class="line">m3  TT  TT  TA  TA  TT  TT</div>
<div class="line">m2  AA  AA  AA  AA  TT  AA</div>
</div><!-- fragment --></li>
</ul>
<p class="intertd">Map file:</p><ul>
<li>File location: map-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker chr pos</div>
<div class="line">m3 3 15</div>
<div class="line">m2 1 8.3</div>
<div class="line">m1 1 5.2</div>
</div><!-- fragment --></li>
</ul>
<p class="endtd">Effect file:</p><ul>
<li>File location: eff-file.txt</li>
<li>File contents: <div class="fragment"><div class="line">marker allele eff</div>
<div class="line">m1 A -0.8</div>
<div class="line">m2 A -0.1</div>
<div class="line">m3 A 0.1</div>
<div class="line">m1 T 0.9</div>
<div class="line">m3 T -0.1</div>
</div><!-- fragment --> </li>
</ul>
</td><td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, <span class="stringliteral">&quot;eff-file.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> founders = init.group;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> eff1 = init.effectSet;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> map1 = init.map;</div>
<div class="ttc" id="agroup__shortnames_html_ga516a882a4bac611d495b8f454478d674"><div class="ttname"><a href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a></div><div class="ttdeci">#define EffectID</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00247">sim-operations.h:247</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gad11f85a6aae488f7efaab5abecaf1ac7"><div class="ttname"><a href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a></div><div class="ttdeci">#define MapID</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00251">sim-operations.h:251</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">init &lt;- load.data(allele.file=&quot;genotype-file.txt&quot;, map.file=&quot;map-file.txt&quot;, effect.file=&quot;eff-file.txt&quot;)</div>
<div class="line">founders &lt;- init$groupNum</div>
<div class="line">eff1 &lt;- init$effectID</div>
<div class="line">map1 &lt;- init$mapID</div>
</div><!-- fragment --> </td></tr>
</table>
<p >This provides an introduction to three types of identifiers you will use in genomicSimulation. They are easier to distinguish in the C version, since they are defined as different types (<code>GroupNum</code> and <code>EffectID</code> and <code>MapID</code>). In the R environment, all genomicSimulation identifiers are integer values. An R user has to remember (or use variable names to help them remember) which one is which.</p><ul>
<li><code>GroupNum</code>: represents a group of candidates, eg your founder population or your F1 crosses.</li>
<li><code>EffectID</code>: represents a set of additive trait effects. You can have multiple of these sets of marker effects loaded to represent different traits. Each time you calculate breeding values, you choose which set of marker effects to use. By default it will use the first loaded set of marker effects.</li>
<li><code>MapID</code>: represents a particular arrangement of your simulation's tracked genetic markers in terms of chromosome positions and recombination frequencies. You can have multiple maps loaded at a time. Each time you generate offspring, you can choose which map to use to generate gametes from the parents. By default it will use the first loaded map.</li>
</ul>
<h2><a class="anchor" id="autotoc_md37"></a>
Load another set of additive trait effects</h2>
<p >This assumes the simulation is set up, that is, one of <b>Load a genetic map and a set of founder genotypes</b>, <b>Load a genetic map and several sets of founder genotypes</b>, <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b> or equivalent has been carried out. If the simulation is not yet set up and you wish to load trait effect values, see the section above, titled <b>Load a genetic map, a set of founder genotypes, and a set of additive trait effects</b>.</p>
<table class="doxtable">
<caption>To load an alternative set of marker effects stored in the file "eff-file2.txt":</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> eff2 = <a class="code hl_define" href="group__shortnames.html#ga7b891ecd001d68f60b379b1bb14e43d7">load_effectfile</a>(d, <span class="stringliteral">&quot;eff-file2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__shortnames_html_ga7b891ecd001d68f60b379b1bb14e43d7"><div class="ttname"><a href="group__shortnames.html#ga7b891ecd001d68f60b379b1bb14e43d7">load_effectfile</a></div><div class="ttdeci">#define load_effectfile</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00282">sim-operations.h:282</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">eff2 &lt;- load.effects(&quot;eff-file2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >If multiple sets of marker effects are loaded, you can choose which one to use each time you tell genomicSimulation to calculate breeding values, and so can calculate breeding values for multiple traits.</p>
<h2><a class="anchor" id="autotoc_md38"></a>
Load another linkage map</h2>
<p >Markers in this map that are not present in the primary linkage map will be ignored. (That is, the first genetic map loaded must contain all markers you wish to be tracked by the simulation. Later genetic maps can only provide different recombination frequencies or chromosome allocations).</p>
<table class="doxtable">
<caption>To load an alternative recombination map stored in the file "mapfile2.txt":</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> map2 = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(d, <span class="stringliteral">&quot;mapfile2.txt&quot;</span>);</div>
<div class="ttc" id="agroup__shortnames_html_gaf16559ebf0798beeaea167422915ece1"><div class="ttname"><a href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a></div><div class="ttdeci">#define load_mapfile</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00281">sim-operations.h:281</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">map2 &lt;- load.map(&quot;mapfile2.txt&quot;)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md39"></a>
Simulating a breeding program</h2>
<table class="doxtable">
<caption>To simulate recurrent truncation selection on a trait over 10 generations</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td>For the readability of the C simulation, we assume there exists a pair of helpful functions <code>double mean(int length, double* values)</code> and <code>double var(int length, double* values)</code> that can be called to calculate the mean and variance, respectively, of an array of doubles. <div class="fragment"><div class="line"><span class="comment">// Set up simulation object</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(time(NULL));</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, <span class="stringliteral">&quot;eff-file.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> parents = init.group;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> effset1 = init.effectSet;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> map1 = init.map;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Set up a file where we will output results</span></div>
<div class="line">FILE* results = fopen(<span class="stringliteral">&quot;simresults.txt&quot;</span>, <span class="stringliteral">&quot;w&quot;</span>);</div>
<div class="line">fprintf(results, <span class="stringliteral">&quot;Generation,MeanBV,VarBV\n&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">double</span> bvs[10]; <span class="comment">// assumes even in the founder population, there are only 10 candidates</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a>(d, parents, effset1, 10, bvs);</div>
<div class="line">fprintf(results, <span class="stringliteral">&quot;%i,%d,%d\n&quot;</span>, 0, mean(10, bvs), var(10, bvs));</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> gen = 1; gen &lt;= 10; ++gen) {</div>
<div class="line">    <span class="comment">// Generate 25 new genotypes by randomly crossing among last generation&#39;s parents</span></div>
<div class="line">    <span class="comment">// (BASIC_OPT is an alias for a default set of progeny generation options. </span></div>
<div class="line">    <span class="comment">// To use different progeny generation options, create a replacement GenOptions struct.</span></div>
<div class="line">    <span class="comment">// There are examples of this in later sections of this page.)</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1 = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, parents, 0, 25, map1, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>); </div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Now, perform truncation selection by taking the 10 candidates with </span></div>
<div class="line">    <span class="comment">// the best breeding values out of group &quot;f1&quot;, and put them in group &quot;f1selection&quot;.</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1selection = <a class="code hl_define" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a>(d, f1, effset1, 10, 0);</div>
<div class="line">    </div>
<div class="line">    <span class="comment">// (Delete last generation&#39;s parents and the F1s that were not selected, to free </span></div>
<div class="line">    <span class="comment">// some memory, since we will not use them again.)</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d,f1);</div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d,parents);</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Log some information about the current generation. Here, we log some statistics about</span></div>
<div class="line">    <span class="comment">// the breeding values measured in this generation.</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a>(d, f1selection, effset1, 10, bvs);</div>
<div class="line">    fprintf(results, <span class="stringliteral">&quot;%i,%d,%d\n&quot;</span>, gen, mean(10, bvs), var(10, bvs));</div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Define the current generation as parents of the next generation for the loop</span></div>
<div class="line">    parents = f1selection; <span class="comment">// Redefining the variable &quot;parents&quot;</span></div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line">fclose(results);</div>
<div class="ttc" id="agroup__shortnames_html_ga4af98a71b0736f81c32df8e5e4fd6a83"><div class="ttname"><a href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a></div><div class="ttdeci">#define BASIC_OPT</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00258">sim-operations.h:258</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga7359d7aff5260f43540819b922b4b8f9"><div class="ttname"><a href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a></div><div class="ttdeci">#define split_by_bv</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00348">sim-operations.h:348</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga87aa2de59b68e428e5b6f07f5955c15a"><div class="ttname"><a href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a></div><div class="ttdeci">#define make_random_crosses</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00338">sim-operations.h:338</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaa68a8ebf286aa6192948ab56851b125f"><div class="ttname"><a href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a></div><div class="ttdeci">#define get_group_bvs</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00316">sim-operations.h:316</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaed498b119b661a47824e76fece2b45df"><div class="ttname"><a href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a></div><div class="ttdeci">#define delete_group</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00361">sim-operations.h:361</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line"># Set up simulation</div>
<div class="line">init &lt;- load.data(&quot;genotype-file.txt&quot;, &quot;map-file.txt&quot;, &quot;eff-file.txt&quot;)</div>
<div class="line">parents &lt;- init$groupNum</div>
<div class="line">effset1 &lt;- init$effectID</div>
<div class="line">map1 &lt;- init$mapID</div>
<div class="line"> </div>
<div class="line"># Set up dataframe to save results</div>
<div class="line">founders.bv &lt;- see.group.data(parents, &quot;BV&quot;)</div>
<div class="line">results &lt;- data.frame(Generation=0, MeanBV=mean(founders.bv), VarBV=var(founders.bv))</div>
<div class="line"> </div>
<div class="line">for (gen in 1:10) {</div>
<div class="line">  # Generate 25 new genotypes by randomly crossing among last generation&#39;s parents.</div>
<div class="line">  # (&quot;map=map1&quot; is optional, because we have only loaded one map, and it will default to the first map that was loaded)</div>
<div class="line">  f1 &lt;- make.random.crosses(parents, n.crosses=25, map=map1)</div>
<div class="line">  </div>
<div class="line">  # Now, perform truncation selection by taking the 10 candidates with </div>
<div class="line">  # the best breeding values out of group &quot;f1&quot;, and put them in group &quot;f1.selection&quot;.</div>
<div class="line">  # (&quot;eff.set=effset1&quot; is optional, because we have only loaded one set of</div>
<div class="line">  # marker effects, and it will default to using the first loaded set of marker effects)</div>
<div class="line">  f1.selection &lt;- break.group.by.GEBV(f1, number=10, eff.set=effset1)</div>
<div class="line">  </div>
<div class="line">  # (Delete last generation&#39;s parents and the F1s that were not selected, to free </div>
<div class="line">  # some memory, since we will not use them again.)</div>
<div class="line">  delete.group(c(parents, f1))</div>
<div class="line"> </div>
<div class="line">  # Log some information about the current generation. Here, we log some statistics about</div>
<div class="line">  # the breeding values measured in this generation.</div>
<div class="line">  f1.selection.info &lt;- see.group.data(f1.selection, &quot;B&quot;)</div>
<div class="line">  results &lt;- rbind(results, c(gen, mean(f1.selection.info), var(f1.selection.info))</div>
<div class="line">  </div>
<div class="line">  # Define the current generation as parents of the next generation for the loop</div>
<div class="line">  parents &lt;- f1.selection</div>
<div class="line">  </div>
<div class="line">  rm(f1, f1.selection, f1.selection.info) # (optional tidying of R environment)</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The remainder of this document will provide sample scripts for simulating features that may appear in real breeding programs. The scripts for these breeding program components can be substituted or added into the structure above to develop a simulation of a more complex breeding program.</p>
<h1><a class="anchor" id="autotoc_md40"></a>
Crossing &amp; Generating New Genotypes: Plant-themed</h1>
<p >Templates in this section assume you have loaded a set of founders whose group number is saved in the variable <code>founders</code>.</p>
<h2><a class="anchor" id="autotoc_md41"></a>
Single seed descent and selfing</h2>
<table class="doxtable">
<caption>For six generations, to grow a single seed from each plant to maturity:</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f6 = <a class="code hl_define" href="group__shortnames.html#gaea0e2102f5a29f876e34fb234e8d7d65">self_n_times</a>(d, 6, founders, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_ga3710da75e403e5a3da4e2deade0d50af"><div class="ttname"><a href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a></div><div class="ttdeci">#define NO_MAP</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00252">sim-operations.h:252</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaea0e2102f5a29f876e34fb234e8d7d65"><div class="ttname"><a href="group__shortnames.html#gaea0e2102f5a29f876e34fb234e8d7d65">self_n_times</a></div><div class="ttdeci">#define self_n_times</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00341">sim-operations.h:341</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f6 &lt;- self.n.times(founders, n=6)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Using NO_MAP for the recombination map parameter in any crossing function in the C version causes the tool to default to the first/primary recombination map. This default behaviour is the same in the R version.</p>
<h2><a class="anchor" id="autotoc_md42"></a>
Creating halfsib or fullsib families</h2>
<table class="doxtable">
<caption>Make 20 random crosses and create families of 6 full siblings from each cross:</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, </div>
<div class="line">                  .offspring_name_prefix=NULL, </div>
<div class="line">                  .family_size=6,     <span class="comment">// &lt;---- Non-default value</span></div>
<div class="line">                  .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, </div>
<div class="line">                  .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">                  .filename_prefix=NULL, </div>
<div class="line">                  .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">                  .will_save_bvs_to_file=<a class="code hl_define" href="group__shortnames.html#ga27b7a0c46aa95c0f127549d8ae007b8d">NO_EFFECTSET</a>, </div>
<div class="line">                  .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">                  .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> crosses = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, founders, 20, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> families[20];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga9b6583f62a092756b6a06e08536ed42b">split_into_families</a>(d, crosses, families);</div>
<div class="ttc" id="agroup__shortnames_html_ga27b7a0c46aa95c0f127549d8ae007b8d"><div class="ttname"><a href="group__shortnames.html#ga27b7a0c46aa95c0f127549d8ae007b8d">NO_EFFECTSET</a></div><div class="ttdeci">#define NO_EFFECTSET</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00248">sim-operations.h:248</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga9b6583f62a092756b6a06e08536ed42b"><div class="ttname"><a href="group__shortnames.html#ga9b6583f62a092756b6a06e08536ed42b">split_into_families</a></div><div class="ttdeci">#define split_into_families</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00326">sim-operations.h:326</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaa320efa0931830d69965beb2f5399acf"><div class="ttname"><a href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a></div><div class="ttdeci">#define GenOptions</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00257">sim-operations.h:257</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaa8cecfc5c5c054d2875c03e77b7be15d"><div class="ttname"><a href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a></div><div class="ttdeci">#define TRUE</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00239">sim-operations.h:239</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaa93f0eb578d23995850d61f7d61c55c1"><div class="ttname"><a href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a></div><div class="ttdeci">#define FALSE</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00240">sim-operations.h:240</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">crosses &lt;- make.random.crosses(founders, n.crosses=20, offspring=6)</div>
<div class="line">families &lt;- break.group.into.families(crosses)</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<caption>Make 10 random crosses with the candidate stored at index 1, and create halfsib families of 6 half siblings from each cross:</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> targetparent_group = <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, </div>
<div class="line">                  .offspring_name_prefix=NULL, </div>
<div class="line">                  .family_size=6,     <span class="comment">// &lt;---- Non-default value</span></div>
<div class="line">                  .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, </div>
<div class="line">                  .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">                  .filename_prefix=NULL, </div>
<div class="line">                  .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">                  .will_save_bvs_to_file=<a class="code hl_define" href="group__shortnames.html#ga27b7a0c46aa95c0f127549d8ae007b8d">NO_EFFECTSET</a>, </div>
<div class="line">                  .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">                  .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> crosses = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, targetparent_group, founders, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> families[10];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga4fa671550fea7b5c723986703a3078ca">split_into_halfsib_families</a>(d, crosses, 2, families); <span class="comment">// &lt;-- 2 for &quot;group by second parent&quot;</span></div>
<div class="ttc" id="agroup__shortnames_html_ga4fa671550fea7b5c723986703a3078ca"><div class="ttname"><a href="group__shortnames.html#ga4fa671550fea7b5c723986703a3078ca">split_into_halfsib_families</a></div><div class="ttdeci">#define split_into_halfsib_families</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00327">sim-operations.h:327</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga702855cf67ebd337941ec9c13e40dfe9"><div class="ttname"><a href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a></div><div class="ttdeci">#define make_group_from</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00322">sim-operations.h:322</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga9d7d8919aa31f1cfb54c902800a2d063"><div class="ttname"><a href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a></div><div class="ttdeci">#define make_random_crosses_between</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00339">sim-operations.h:339</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1))</div>
<div class="line"> </div>
<div class="line">crosses &lt;- make.random.crosses.between(targetparent_group, founders, n.crosses=10, offspring=6)</div>
<div class="line"> </div>
<div class="line">families &lt;- break.group.into.halfsib.families(crosses, 2) # &lt;-- 2 for &quot;group by second parent&quot;</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md43"></a>
Updating marker effects</h2>
<p >Updating the marker effect estimates is a task that must be done outside of genomicSimulation. Export required data from the simulation, run the marker effect re-estimation, then re-import marker effects as in preceding section <b>Swap out the set of additive trait effects for another</b>.</p>
<h2><a class="anchor" id="autotoc_md44"></a>
Trials and locations</h2>
<p >Suppose you want to simulate the success of the founder population in different environments.</p>
<table class="doxtable">
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> location1 = <a class="code hl_define" href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a>(d, founders, <a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> location2 = <a class="code hl_define" href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a>(d, founders, <a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__shortnames_html_gacd45e8b70bb44a78325f6947479db0ad"><div class="ttname"><a href="group__shortnames.html#gacd45e8b70bb44a78325f6947479db0ad">make_clones</a></div><div class="ttdeci">#define make_clones</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00343">sim-operations.h:343</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">location1 &lt;- make.clones(founders)</div>
<div class="line">location2 &lt;- make.clones(founders)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Once separate copies exist for trials at different locations, simulate selection with different intensities or heritabilities for each different location group. Eg <b>Select on phenotype (simulated with a given heritability)</b>.</p>
<h2><a class="anchor" id="autotoc_md45"></a>
Backcrossing</h2>
<table class="doxtable">
<caption>Pull out a chosen founder and repeatedly (for 20 generations) cross back to it</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="keywordtype">int</span> targetparent = 1;</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> targetparent_group = <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 1, &amp;targetparent);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> backcross_generations[20];</div>
<div class="line">backcross_generation[0] = founders;</div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; 20; ++i) {</div>
<div class="line">    backcross_generations[i] = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, targetparent_group, backcross_generations[i-1], 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line">}</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">targetparent_group &lt;- make.group(c(1))</div>
<div class="line"> </div>
<div class="line">backcross_generations &lt;- rep(0, times=20) # initialise vector of correct size</div>
<div class="line">backcross_generations[1] &lt;- founders</div>
<div class="line">for (ii in 1:20) {</div>
<div class="line">    backcross_generations[ii] &lt;- make.random.crosses.between(targetparent_group, backcross_generations[ii-1], n.crosses=10)</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >For marker-assisted backcrossing, select each loop on presence of the marker and/or the estimated breeding value. See the sections <b>Select on a qualitative trait</b> and <b>Select on a qualitative trait, then a quantitative trait</b> further down this page.</p>
<h1><a class="anchor" id="autotoc_md46"></a>
Crossing &amp; Generating New Genotypes: Animal-themed</h1>
<p >Templates in this section assume you have loaded two sets of founders whose group numbers are saved in the variables <code>cows</code> and <code>bulls</code>. See preceding section <b>Load a genetic map and several sets of founder genotypes</b>.</p>
<h2><a class="anchor" id="autotoc_md47"></a>
Split offspring into male and female</h2>
<table class="doxtable">
<caption>Out of 10 offspring of random crosses, randomly define some as male and some as female</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_f = <a class="code hl_define" href="group__shortnames.html#ga846bceb22b2a254269d517342c60b189">split_randomly_into_two</a>(d, offspring);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_m = offspring;</div>
<div class="ttc" id="agroup__shortnames_html_ga846bceb22b2a254269d517342c60b189"><div class="ttname"><a href="group__shortnames.html#ga846bceb22b2a254269d517342c60b189">split_randomly_into_two</a></div><div class="ttdeci">#define split_randomly_into_two</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00331">sim-operations.h:331</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, n.crosses=10)</div>
<div class="line"> </div>
<div class="line">temporary &lt;- break.group.randomly(offspring, into.n = 2)</div>
<div class="line">offspring_f &lt;- temporary[1]</div>
<div class="line">offspring_m &lt;- temporary[2]</div>
<div class="line">rm(temporary)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >split_randomly_into_two or break.group.randomly can be used to split a group into two sub-groups by flipping a coin on each group member. This may result in two groups that are not quite the same size. To split a group of eg. 10 offspring into two groups of exactly 5 members, the cousin functions, <code>split_evenly_into_two</code> (C)/<code>break.group.evenly</code> (R), can be used instead:</p>
<table class="doxtable">
<caption>Out of 10 offspring of random crosses, get exactly 5 male calves and 5 females</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_f = <a class="code hl_define" href="group__shortnames.html#ga846bceb22b2a254269d517342c60b189">split_randomly_into_two</a>(d, offspring);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_m = offspring;</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, n.crosses=10)</div>
<div class="line"> </div>
<div class="line">temporary &lt;- break.group.evenly(offspring, into.n = 2)</div>
<div class="line">offspring_f &lt;- temporary[1]</div>
<div class="line">offspring_m &lt;- temporary[2]</div>
<div class="line">rm(temporary)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Or, for uneven proportions, use the functions</p><ul>
<li><code>split_by_probabilities</code> (C)/<code>break.group.with.probabilities</code> (R), corresponding to <code>split_randomly_into_n</code> (C)/<code>break.group.randomly</code> (R), to randomly allocate sex by performing a random draw/coin toss for each candidate to determine which group it is placed into, or</li>
<li><code>split_into_buckets</code> (C)/<code>break.group.into.buckets</code> (R), corresponding to <code>split_evenly_into_n</code> (C)/<code>break.group.evenly</code> (R), to randomly sort the list of candidates and allocate specific numbers of them to each group.</li>
</ul>
<table class="doxtable">
<caption>Out of 10 offspring of random crosses, get exactly 1 male calf and 9 females</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 0, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// split_into_buckets needs to be passed n-1 &quot;bucket sizes&quot; to make n groups</span></div>
<div class="line"><span class="comment">// so we tell it we want 9 females and the remainder (n=10-9=1) will be allocated male </span></div>
<div class="line"><span class="keywordtype">int</span> bucketsize = 9; </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> temporary[2];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga9d08b34cf3945aca7a0ce16dfcf4acc7">split_into_buckets</a>(d, offspring, 2, &amp;bucketsize, &amp;temporary);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_f = temporary[0];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_m = temporary[1];</div>
<div class="ttc" id="agroup__shortnames_html_ga9d08b34cf3945aca7a0ce16dfcf4acc7"><div class="ttname"><a href="group__shortnames.html#ga9d08b34cf3945aca7a0ce16dfcf4acc7">split_into_buckets</a></div><div class="ttdeci">#define split_into_buckets</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00330">sim-operations.h:330</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, n.crosses=10)</div>
<div class="line"> </div>
<div class="line"># break.group.into.buckets takes one less &quot;bucket size&quot; than the number of groups it will make</div>
<div class="line"># so we tell it we want 9 females and the remainder (n=10-9=1) will be allocated male.</div>
<div class="line">temporary &lt;- break.group.into.buckets(offspring, buckets=c(9))</div>
<div class="line">offspring_f &lt;- temporary[1]</div>
<div class="line">offspring_m &lt;- temporary[2]</div>
<div class="line">rm(temporary)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >As an alternative, you could create a custom label to track the animal's sex (eg. the label having value 1 for a female, 2 for a male). Sex could be initially set by passing a vector of 1s and 2s (random or otherwise) to <code>change_label_to_values</code> (C)/<code>change.label.to.values</code> (R). Afterwards, <code>split_by_label_value</code> (C)/<code>break.group.by.label.value</code> (R) can be used to separate the males and females as needed, and as needed <code>combine_groups</code> (C)/<code>combine.groups</code> (R) can merge the populations back together.</p>
<p >However, for most simulations of breeding programs (it is likely every generation would require separating out the males and females before crossing them), it would be more computationally efficient to maintain two separate "male" and "female" groups, rather than using a custom label and repeatedly merging and unmerging the groups.</p>
<h2><a class="anchor" id="autotoc_md48"></a>
Introducing fresh blood to male and female kernels</h2>
<p >Suppose the <code>offspring_f</code> and <code>offspring_m</code> groups exist, as created in the previous section <b>Split offspring into male and female</b>.</p>
<table class="doxtable">
<caption>Introduce new calves to the breeding kernels</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> cowGroups[2];</div>
<div class="line">cowGroups[0] = cows;</div>
<div class="line">cowGroups[1] = offspring_f;</div>
<div class="line"> </div>
<div class="line">cows = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, cowGroups);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> bullGroups[2];</div>
<div class="line">bullGroups[0] = bulls;</div>
<div class="line">bullGroups[1] = offspring_m;</div>
<div class="line"> </div>
<div class="line">bulls = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, bullGroups);</div>
<div class="ttc" id="agroup__shortnames_html_ga73e77b55777298bf0c2349ada77bc44b"><div class="ttname"><a href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a></div><div class="ttdeci">#define combine_groups</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00321">sim-operations.h:321</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">cows &lt;- combine.groups(c(cows,offspring_f))</div>
<div class="line">bulls &lt;- combine.groups(c(bulls,offspring_m))</div>
</div><!-- fragment --> </td></tr>
</table>
<p >It may be useful to use age tracking in these kernels. See the following section, <b>Tracking age and culling individuals that are too old</b>.</p>
<h2><a class="anchor" id="autotoc_md49"></a>
Tracking age and culling individuals that are too old</h2>
<p >genomicSimulation's custom labels can be used to track age (or some other known trait that can be represented with an integer for each individual).</p>
<p >Using custom labels involves three main operations:</p><ul>
<li>Creating the label and giving it a default value with <code>create_new_label</code> (C)/<code>create.new.label</code> (R)</li>
<li>Modifying the values of certain or all candidate genotypes, eg. to increase the age of the animals. Relevant functions include <code>change_label_by_amount</code> (C)/<code>change.label.by.amount</code> (R), <code>change.label.to</code> (C)/<code>change.label.to.this</code> (R).</li>
<li>Splitting a group based on the values of a custom label, so that you can carry out some crossing or selection operations on only a subset of the population (eg. those of a certain age). Relevant functions include <code>split_by_label_range</code> (C)/<code>break.group.by.label.range</code> (R), <code>split_by_label_value</code> (C)/<code>break.group.by.label.value</code> (R).</li>
</ul>
<table class="doxtable">
<caption>A population where only animals 3 years or older can breed</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(1234567);</div>
<div class="line"><span class="keyword">struct </span><a class="code hl_define" href="group__shortnames.html#gab9139b22223d33aa0a762d3ec00b3072">MultiIDSet</a> init = <a class="code hl_define" href="group__shortnames.html#ga2e10fe6c17f65cefc887a7b8868d9017">load_data_files</a>(d, <span class="stringliteral">&quot;genotype-file.txt&quot;</span>, <span class="stringliteral">&quot;map-file.txt&quot;</span>, NULL, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> animals = init.group;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Create a new label to represent age, with default/at-birth value of 0.</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga5c7e03838336e41c1782e75c4f9029d6">LabelID</a> ageLabel = <a class="code hl_define" href="group__shortnames.html#gaa976db4aa919852354fc11c5c14647cf">create_new_label</a>(d, 0);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Founders are 3 years old at the beginning.</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#ga1c1680b1461267ccf7216e2b7ceb06dc">change_label_to</a>(d, animals, ageLabel, 3);</div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> year = 0; year &lt; 10; ++year) {</div>
<div class="line">    <span class="comment">// Ages 3-20 can breed.</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> breedingGroup = <a class="code hl_define" href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a>(d, animals, ageLabel, 3, 20); </div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Do some breeding/selection steps as appropriate for the breeding program, eg:</span></div>
<div class="line">    GruopNum offspring = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, breedingGroup, 50, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line">    <span class="comment">// Offspring will have the default value for the label i.e. ageLabel = 0</span></div>
<div class="line">    </div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toCombine[3] = {animals, breedingGroup, offspring};</div>
<div class="line">    animals = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 3, toCombine);</div>
<div class="line">    </div>
<div class="line">    <span class="comment">// Increase age of all by 1</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#ga53c6a459f13ae342d8e2a3ed88847a9a">change_label_by_amount</a>(d, animals, ageLabel, 1);</div>
<div class="line">}</div>
<div class="ttc" id="agroup__shortnames_html_ga083d4709ce763182d5a00310b9ec459e"><div class="ttname"><a href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a></div><div class="ttdeci">#define split_by_label_range</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00324">sim-operations.h:324</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga1c1680b1461267ccf7216e2b7ceb06dc"><div class="ttname"><a href="group__shortnames.html#ga1c1680b1461267ccf7216e2b7ceb06dc">change_label_to</a></div><div class="ttdeci">#define change_label_to</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00273">sim-operations.h:273</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga53c6a459f13ae342d8e2a3ed88847a9a"><div class="ttname"><a href="group__shortnames.html#ga53c6a459f13ae342d8e2a3ed88847a9a">change_label_by_amount</a></div><div class="ttdeci">#define change_label_by_amount</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00274">sim-operations.h:274</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga5c7e03838336e41c1782e75c4f9029d6"><div class="ttname"><a href="group__shortnames.html#ga5c7e03838336e41c1782e75c4f9029d6">LabelID</a></div><div class="ttdeci">#define LabelID</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00249">sim-operations.h:249</a></div></div>
<div class="ttc" id="agroup__shortnames_html_gaa976db4aa919852354fc11c5c14647cf"><div class="ttname"><a href="group__shortnames.html#gaa976db4aa919852354fc11c5c14647cf">create_new_label</a></div><div class="ttdeci">#define create_new_label</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00271">sim-operations.h:271</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">init &lt;- load.data(&quot;genotype-file.txt&quot;, &quot;map-file.txt&quot;)</div>
<div class="line">animals &lt;- init$groupNum</div>
<div class="line"> </div>
<div class="line"># Create a new label to represent age, with default/at-birth value of 0.</div>
<div class="line">ageLabel &lt;- create.new.label(0L)</div>
<div class="line"> </div>
<div class="line"># Founders are 3 years old at the beginning.</div>
<div class="line">change.label.to.this(ageLabel, 3L, animals);</div>
<div class="line"> </div>
<div class="line">for (year in 0:10) {</div>
<div class="line">    // Ages 3-20 can breed.</div>
<div class="line">    breedingGroup &lt;- break.group.by.label.range(ageLabel, 3, 20, animals);</div>
<div class="line">    </div>
<div class="line">    # Do some breeding/selection steps as appropriate for the breeding program, eg:</div>
<div class="line">    offspring &lt;- make.random.crosses(breedingGroup, 50, 0)</div>
<div class="line">    # Offspring will have the default value for the label i.e. ageLabel = 0</div>
<div class="line">    </div>
<div class="line">    animals &lt;- combine.groups(c(animals, breedingGroup, offspring))</div>
<div class="line">    </div>
<div class="line">    # Increase age of all by 1</div>
<div class="line">    change.label.by.amount(ageLabel, 1, animals);</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<table class="doxtable">
<caption>Cull animals that are 12 years or older</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toCull = <a class="code hl_define" href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a>(d, animals, ageLabel, 12, 1000000); </div>
<div class="line"><span class="comment">// some very large number for upper bound. User is responsible for the values</span></div>
<div class="line"><span class="comment">// of the custom label, so you can work out a number your script should never </span></div>
<div class="line"><span class="comment">// set the label above.</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordflow">if</span> (toCull &gt; 0) { <span class="comment">// split_by_label_range did find some animals to cull</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, toCull);</div>
<div class="line">}</div>
</div><!-- fragment --> or <div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toKeep = <a class="code hl_define" href="group__shortnames.html#ga083d4709ce763182d5a00310b9ec459e">split_by_label_range</a>(d, animals, ageLabel, 0, 11); </div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, animals);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">toCull &lt;- break.group.by.label.range(ageLabel, 12, 1000000, animals)</div>
<div class="line"> </div>
<div class="line">if (toCull &gt; 0) { // split_by_label_range did find some animals to cull</div>
<div class="line">    delete.group(toCull) // delete from memory</div>
<div class="line">    rm(toCull) // remove `toCull` label from R environment.</div>
<div class="line">}</div>
</div><!-- fragment --> or <div class="fragment"><div class="line">toKeep &lt;- break.group.by.label.range(ageLabel, 0, 11, animals)</div>
<div class="line"> </div>
<div class="line">delete.group(animals)</div>
<div class="line">rm(animals)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md50"></a>
Random mortality</h2>
<p >In simulating a breeding program, there may often be stages where some of the population is lost in a way that can be best simulated as a random event.</p>
<table class="doxtable">
<caption>Some proportion of the animals will be lost this generation</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="keywordtype">double</span> mortality = 0.06; <span class="comment">// 6% chance of mortality</span></div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> temporary[2];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gabfc1ed9c84d045a8ea54c80645ecec95">split_by_probabilities</a>(d, animals, 1, &amp;mortality, &amp;temporary); <span class="comment">// &lt;-- 6% chance to be in first group, (100-6)% chance to survive.</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> lost_animals = temporary[0];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> animals_survived = temporary[1];</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, lost_animals);</div>
<div class="ttc" id="agroup__shortnames_html_gabfc1ed9c84d045a8ea54c80645ecec95"><div class="ttname"><a href="group__shortnames.html#gabfc1ed9c84d045a8ea54c80645ecec95">split_by_probabilities</a></div><div class="ttdeci">#define split_by_probabilities</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00333">sim-operations.h:333</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">mortality &lt;- 0.06  # 6% chance of mortality</div>
<div class="line"> </div>
<div class="line">temporary &lt;- break.group.with.probabilities(animals, mortality)  # &lt;-- 6% chance to be in first group, (100-6)% chance to survive.</div>
<div class="line">lost_animals &lt;- temporary[1]</div>
<div class="line">animals_survived &lt;- temporary[2] </div>
<div class="line">rm(temporary)</div>
<div class="line"> </div>
<div class="line">delete.group(lost_animals)   # &lt;-- to free the memory being used to store the candidates from &quot;lost_animals&quot;. The group &quot;lost_animals&quot; can no longer be used after this line</div>
<div class="line">rm(lost_animals)             # &lt;-- optional: to tidy up the R environment by removing the stale variable &quot;lost_animals&quot; so it does not confuse you later</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Because all offspring are generated independently in genomicSimulation, if you have a random culling/offspring loss step immediately after the command in which those genotypes were generated, you could get speedups in simulation by simply only generating the surviving offspring:</p>
<table class="doxtable">
<caption>Some proportion of offspring are immediately lost</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="comment">// Produce offspring</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, breedingGroup, 50000, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Immediate mortality</span></div>
<div class="line"><span class="keywordtype">double</span> mortality = 0.06; <span class="comment">// 6% chance of mortality</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> temporary[2];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gabfc1ed9c84d045a8ea54c80645ecec95">split_by_probabilities</a>(d, offspring, 1, &amp;mortality, &amp;temporary);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> lost_offspring = temporary[0];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring_survived = temporary[1];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, lost_offspring);</div>
</div><!-- fragment --> or, to combine those steps and potentially speed up simulation: <div class="fragment"><div class="line"><span class="comment">// 50000 offspring at 6% mortality rate. Calculate number of survivors:</span></div>
<div class="line"><span class="keywordtype">int</span> n_survivors = 0;</div>
<div class="line"><span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 0; i &lt; 50000; ++i) {</div>
<div class="line">    <span class="comment">// this line, run 50000 times, adds one survivor to the count if a random draw from a uniform distribution 0-1 is greater than 0.06</span></div>
<div class="line">    n_survivors += ((double)rand() / (double)RAND_MAX) &gt; 0.06;</div>
<div class="line">    <span class="comment">// If you have a proper C random number generation library included, use that instead.</span></div>
<div class="line">    <span class="comment">// (if you have a proper random number generation library, you can probably use its random binomial distribution </span></div>
<div class="line">    <span class="comment">// function, as show in R version opposite, instead of this loop)</span></div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, breedingGroup, n_survivors, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">// Produce offspring</div>
<div class="line">offspring &lt;- make.random.crosses(breedingGroup, 50000)</div>
<div class="line"> </div>
<div class="line">// Immediate mortality</div>
<div class="line">mortality &lt;- 0.06  # 6% chance of mortality</div>
<div class="line">temporary &lt;- break.group.with.probabilities(offspring, mortality)</div>
<div class="line">lost_offspring &lt;- temporary[1]</div>
<div class="line">offspring_survived &lt;- temporary[2] </div>
<div class="line">rm(temporary)</div>
<div class="line"> </div>
<div class="line">delete.group(lost_offspring)</div>
<div class="line">rm(lost_offspring)</div>
</div><!-- fragment --> or, to combine those steps and potentially speed up simulation: <div class="fragment"><div class="line">// 50000 offspring at 6% mortality rate. Calculate number of survivors:</div>
<div class="line">n_survivors &lt;- rbinom(1,50000,1-0.06)</div>
<div class="line">offspring &lt;- make.random.crosses(breedingGroup, n_survivors)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md51"></a>
Culling or fitness-based animal loss</h2>
<p >Alternatively, rather than a random mortality system, you could simulate loss of candidates according to their estimated breeding values. Perhaps those with the lowest breeding values for some fitness trait are the ones that are lost, or perhaps the ones with the highest for an economic value index are sold.</p>
<table class="doxtable">
<caption>The worst animals will be lost this generation</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="keywordtype">double</span> mortality = 0.06; <span class="comment">// 6% mortality proportion</span></div>
<div class="line"><span class="keywordtype">int</span> num_will_lose = round(mortality * <a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>(d, animals));</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> lost_animals = <a class="code hl_define" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a>(d, animals, fitness_effid, num_will_lose, 1);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, lost_animals);</div>
</div><!-- fragment --> If the situation is more complex than just deleting the animals with the lowest breeding values, the following equivalent template will be more useful: <div class="fragment"><div class="line"><span class="keywordtype">double</span> mortality = 0.06; <span class="comment">// 6% mortality proportion</span></div>
<div class="line"><span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> num_animals = <a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>(d, animals);</div>
<div class="line"><span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> num_will_lose = round(mortality * num_animals);</div>
<div class="line"><span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index_of_lost_animals[num_will_lose];</div>
<div class="line"> </div>
<div class="line"><span class="comment">// You would then access some group data and choose the indexes of the animals that</span></div>
<div class="line"><span class="comment">// will be lost however you would like. Here I pretend I have created a function </span></div>
<div class="line"><span class="comment">// &quot;void find_worst_animals(uint num_animals, uint* animal_indexes, uint* animal_bvs, uint find_how_many, uint* save_worst_animal_indexes_here);&quot;</span></div>
<div class="line"><span class="comment">// and I need to pass animal indexes and breeding values to that function.</span></div>
<div class="line"><span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> index_animals[num_animals];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>(d, animals, num_animals, index_animals);</div>
<div class="line"><span class="keywordtype">double</span> bv_animals[num_animals];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a>(d, animals, fitness_effid, num_animals, bv_animals);</div>
<div class="line">find_worst_animals(num_animals, index_animals, bv_animals, num_will_lose, index_of_lost_animals);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> lost_animals = make_group(d, num_will_lose, index_of_lost_animals);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaed498b119b661a47824e76fece2b45df">delete_group</a>(d, lost_animals);</div>
<div class="ttc" id="agroup__shortnames_html_gaa7f7f1453a84746fa215fea171b80b97"><div class="ttname"><a href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a></div><div class="ttdeci">#define get_group_indexes</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00314">sim-operations.h:314</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">mortality &lt;- 0.06</div>
<div class="line"> </div>
<div class="line">lost_animals &lt;- break.group.by.gebv(animals, low.score.best=TRUE, percentage=mortality*100, eff.set=fitness.effset)</div>
<div class="line">delete.group(lost_animals)</div>
<div class="line">rm(lost_animals)</div>
</div><!-- fragment --> If the situation is more complex than just deleting the animals with the lowest breeding values, the following equivalent template will be more useful: <div class="fragment"><div class="line">mortality &lt;- 0.06</div>
<div class="line"> </div>
<div class="line">info &lt;- data.frame(Index=see.group.data(animals,&quot;X&quot;),</div>
<div class="line">                   Fitness=see.group.data(animals,&quot;BV&quot;,eff.set=fitness.effset))</div>
<div class="line">     </div>
<div class="line"># Use the &quot;info&quot; table to choose which will be lost. We show a simple example here</div>
<div class="line"># of just taking the animals with the lowest breeding values, but more complex</div>
<div class="line"># choices could be made.</div>
<div class="line">num_will_lose &lt;- round(mortality * length(info$Index))</div>
<div class="line">index_of_lost_animals &lt;- info[order(info$GEBV, decreasing=FALSE),]$Index[1:num_will_lose]</div>
<div class="line"> </div>
<div class="line">lost_animals &lt;- make.group(index_of_lost_animals)</div>
<div class="line">delete.group(lost_animals)</div>
<div class="line">rm(lost_animals)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The important thing to remember when using a system like this is that any <code>delete_group</code> (C)/<code>delete.group</code> (R) call can potentially invalidate any position indexes calculated before it. There is a potential pitfall in a situation when you calculate scores for multiple groups, or multiple subgroups of a single group, and can become tempted to delete the unwanted groups as you go (see table below. The examples shown are for the R version of genomicSimulation, but the same pitfall exists in the C version).</p>
<table class="doxtable">
<caption>Pitfalls when deleting groups</caption>
<tr>
<th>INCORRECT SEQUENCE</th><th>CORRECT SEQUENCE </th></tr>
<tr>
<td>In this scenario, we are performing the same culling strategy as in the previous table. Some proportion of animals with the lowest breeding values will be lost/culled. However, in this case, we have different loss rates by sex, and a custom label that tracks the sex of each animal. <div class="fragment"><div class="line">female.mortality &lt;- 0.06</div>
<div class="line">male.mortality &lt;- 0.4</div>
<div class="line"> </div>
<div class="line">info &lt;- data.frame(Index=see.group.data(animals,&quot;X&quot;), # &lt;-- position indexes accessed here</div>
<div class="line">                   Fitness=see.group.data(animals,&quot;BV&quot;,eff.set=fitness.effset),</div>
<div class="line">                   Sex=see.group.data(animals,&quot;Label&quot;,label=label_sex))        </div>
<div class="line">info.females &lt;- info[Sex==1,]</div>
<div class="line">info.males &lt;- info[Sex==2,]</div>
<div class="line"> </div>
<div class="line">num_females_will_lose &lt;- round(female.mortality * length(info.females$Index))</div>
<div class="line">index_of_lost_females &lt;- info.females[order(info.females$GEBV, decreasing=FALSE),]$Index[1:num_females_will_lose]</div>
<div class="line">lost_females &lt;- make.group(index_of_lost_females)     # &lt;-- make.group uses position indexes</div>
<div class="line">delete.group(lost_females)                            # &lt;-- all previously calculated position indexes are now potentially invalid</div>
<div class="line">rm(lost_females)</div>
<div class="line"> </div>
<div class="line">num_males_will_lose &lt;- round(male.mortality * length(info.males$Index))</div>
<div class="line">index_of_lost_males &lt;- info.males[order(info.males$GEBV, decreasing=FALSE),]$Index[1:num_males_will_lose]</div>
<div class="line">                                                     # &lt;-- info tables contain invalid position indexes now</div>
<div class="line">lost_males &lt;- make.group(index_of_lost_males)        # &lt;-- !! make.group uses invalid position indexes, so the animals selected to be lost_males may not correspond to the ones we intended to select. </div>
<div class="line">delete.group(lost_males)</div>
<div class="line">rm(lost_males)</div>
</div><!-- fragment --> </td><td>To fix this, you could simply rearrange the script so that you only run <code>delete.group</code> once all interactions with the <code>info</code> tables are done: <div class="fragment"><div class="line">female.mortality &lt;- 0.06</div>
<div class="line">male.mortality &lt;- 0.4</div>
<div class="line"> </div>
<div class="line">info &lt;- data.frame(Index=see.group.data(animals,&quot;X&quot;), # &lt;-- position indexes accessed here</div>
<div class="line">                   Fitness=see.group.data(animals,&quot;BV&quot;,eff.set=fitness.effset),</div>
<div class="line">                   Sex=see.group.data(animals,&quot;Label&quot;,label=label_sex))        </div>
<div class="line">info.females &lt;- info[Sex==1,]</div>
<div class="line">info.males &lt;- info[Sex==2,]</div>
<div class="line"> </div>
<div class="line">num_females_will_lose &lt;- round(female.mortality * length(info.females$Index))</div>
<div class="line">index_of_lost_females &lt;- info.females[order(info.females$GEBV, decreasing=FALSE),]$Index[1:num_females_will_lose]</div>
<div class="line">lost_females &lt;- make.group(index_of_lost_females)     # &lt;-- make.group uses position indexes</div>
<div class="line"> </div>
<div class="line">num_males_will_lose &lt;- round(male.mortality * length(info.males$Index))</div>
<div class="line">index_of_lost_males &lt;- info.males[order(info.males$GEBV, decreasing=FALSE),]$Index[1:num_males_will_lose]</div>
<div class="line">lost_males &lt;- make.group(index_of_lost_males)         # &lt;-- make.group uses position indexes that are still valid</div>
<div class="line"> </div>
<div class="line"># As a helpful check, you may want to remove the tables containing position indexes</div>
<div class="line"># from the R environment so that there is no temptation to access them after groups are</div>
<div class="line"># deleted and the position indexes stored in them are no longer valid</div>
<div class="line">rm(info, info.females, info.males)</div>
<div class="line"> </div>
<div class="line">delete.group(c(lost_females, lost_males))             # &lt;-- position indexes from info tables now invalid (but the tables are already deleted)</div>
<div class="line">rm(lost_females,lost_males)</div>
</div><!-- fragment --> Be thoughtful about where in your script it is safe to place <code>delete.group</code> commands. </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md52"></a>
Random mating with caps on number of offspring per animal</h2>
<table class="doxtable">
<caption>Any cow should have at most one calf this generation</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bulls, 10, 1, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">offspring &lt;- make.random.crosses.between(cows, bulls, cap1=1, n.crosses=10)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md53"></a>
Mating all females to a good male</h2>
<table class="doxtable">
<caption>A breeding strategy where only the best bull in the population will father calves</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><span class="comment">// Assuming there exists a variable &quot;eff1&quot; which is an EffectID for the set of marker effects to use to calculate breeding values here</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> bestbull = <a class="code hl_define" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a>(d, bulls, eff1, 1, <a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, cows, bestbull, 10, 1, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// optional: put him back with the other bulls</span></div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> toCombine[2] = {bulls, bestbull};</div>
<div class="line">bulls = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, toCombine);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line"># by default, this function uses the first effect set loaded, so if &quot;eff1&quot; is </div>
<div class="line"># the first effect set loaded, `eff.set=eff1` is optional:</div>
<div class="line">bestbull &lt;- break.group.by.gebv(bulls, number=1, eff.set=eff1) </div>
<div class="line">offspring &lt;- make.random.crosses.between(cows, bestbull, cap1=1, n.crosses=10)</div>
<div class="line"> </div>
<div class="line"># optional: put him back with the other bulls</div>
<div class="line">bulls &lt;- combine.groups(c(bulls, bestbull))</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md54"></a>
Random mating between the best performing individuals in a group</h2>
<p >genomicSimulation used to have a function <code>make_n_crosses_from_top_m_percent</code>, which would identify a highest-performing slice of a group and make random crosses between them, but leave the high performers as members of their original group. Here are code snippets that replicate its functionality:</p>
<table class="doxtable">
<caption>A function to make n random crosses between the m percent of candidates in a group that have the highest breeding values</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> make_n_crosses_from_top_m_percent(<a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d, <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a>* group, <span class="keyword">const</span> <span class="keywordtype">int</span> n, <span class="keyword">const</span> <span class="keywordtype">float</span> m, <a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> effset) {</div>
<div class="line"><span class="preprocessor">    # Split off the best candidates </span></div>
<div class="line">    <span class="keywordtype">int</span> mcount = (int)(<a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>(d,*group) * m / 100); # close enough, you could round differently <span class="keywordflow">if</span> you preferred</div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> best_candidates = <a class="code hl_define" href="group__shortnames.html#ga7359d7aff5260f43540819b922b4b8f9">split_by_bv</a>(d, *group, effset, mcount, 0);</div>
<div class="line">    </div>
<div class="line"><span class="preprocessor">    # Make random crosses to get offspring, using default map</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, best_candidates, n, 0, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line">    </div>
<div class="line"><span class="preprocessor">    # Put the best candidates back into their original group</span></div>
<div class="line"><span class="preprocessor">    # Note: you cannot assume that the group number representing &quot;group&quot;</span></div>
<div class="line"><span class="preprocessor">    # will not change during a combine_groups operation</span></div>
<div class="line">    <a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> tmp[2] = {*group, best_candidates};</div>
<div class="line">    *group = <a class="code hl_define" href="group__shortnames.html#ga73e77b55777298bf0c2349ada77bc44b">combine_groups</a>(d, 2, tmp);</div>
<div class="line">    </div>
<div class="line">    <span class="keywordflow">return</span> offspring;</div>
<div class="line">}</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">make.n.crosses.from.top.m.percent &lt;- function(group, n, m, effset=1) {</div>
<div class="line">    # Split off the best candidates </div>
<div class="line">    best.candidates &lt;- break.group.by.gebv(group, percentage=m, eff.set=effset)</div>
<div class="line">    </div>
<div class="line">    # Make random crosses to get offspring </div>
<div class="line">    offspring &lt;- make.random.crosses(best.candidates, n.crosses=n)</div>
<div class="line">    </div>
<div class="line">    # Put the best candidates back into their original group</div>
<div class="line">    # Note: you cannot assume that the group number representing &quot;group&quot;</div>
<div class="line">    # will not change as a result of this line, thus the double arrow</div>
<div class="line">    # to make sure any changes to the group number representing &quot;group&quot;</div>
<div class="line">    # are also saved in this function&#39;s calling environment. </div>
<div class="line">    # (alternatively you could return &quot;group&quot; alongside the new &quot;offspring&quot; group)</div>
<div class="line">    group &lt;&lt;- combine.groups(c(group, best.candidates))</div>
<div class="line">    </div>
<div class="line">    return(offspring)</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md55"></a>
Making specific chosen matings</h2>
<table class="doxtable">
<caption>Suppose you want to cross Cow1 to Bull1, Cow2 to Bull2, and Daisy to Bull1</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">cow1_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Cow1&quot;</span>);</div>
<div class="line">cow2_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Cow2&quot;</span>);</div>
<div class="line">cow3_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Daisy&quot;</span>);</div>
<div class="line">bull1_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Bull1&quot;</span>);</div>
<div class="line">bull2_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Bull2&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][3];</div>
<div class="line">crossingPlan[0][0] = cow1_index;</div>
<div class="line">crossingPlan[0][1] = cow2_index;</div>
<div class="line">crossingPlan[0][2] = cow3_index;</div>
<div class="line">crossingPlan[1][0] = bull1_index;</div>
<div class="line">crossingPlan[1][1] = bull2_index;</div>
<div class="line">crossingPlan[1][2] = bull1_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 3, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="ttc" id="agroup__search_html_gad6c28080cf3e1f85fd78229aba53bf78"><div class="ttname"><a href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a></div><div class="ttdeci">unsigned int gsc_get_index_of_name(const gsc_AlleleMatrix *start, const char *name)</div><div class="ttdoc">Search for a genotype with a particular name in a linked list of gsc_AlleleMatrix,...</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8c_source.html#l02342">sim-operations.c:2342</a></div></div>
<div class="ttc" id="agroup__shortnames_html_ga56e71147b1f8a71151ef81b4a828aa9b"><div class="ttname"><a href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a></div><div class="ttdeci">#define make_targeted_crosses</div><div class="ttdef"><b>Definition:</b> <a href="sim-operations_8h_source.html#l00340">sim-operations.h:340</a></div></div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line"># The R version of the function (make.targeted.crosses) can be passed the indexes of these parent animals, like in the C version, </div>
<div class="line"># or it can use the names of the parent animals.</div>
<div class="line">offspring &lt;- make.targeted.crosses(c(&quot;Cow1&quot;, &quot;Cow2&quot;, &quot;Daisy&quot;), c(&quot;Bull1&quot;, &quot;Bull2&quot;, &quot;Bull1&quot;))</div>
</div><!-- fragment --> </td></tr>
</table>
<p >The other option is to create an input file of the following format: </p><div class="fragment"><div class="line">Cow1 Bull1</div>
<div class="line">Cow2 Bull2</div>
<div class="line">Daisy Bull1</div>
</div><!-- fragment --><p> and call <code>make_crosses_from_file</code> (C)/ <code>make.crosses.from.file</code> (R).</p>
<h2><a class="anchor" id="autotoc_md56"></a>
Three-Way Crosses</h2>
<p >Suppose you want to make a single Breed1/Breed2//Breed3 cross. That is, cross the F1 of a mating between Breed1 and Breed2 to Breed3. Assumes the parent you've chosen to represent Breed1 is named "Breed1Parent", and likewise for Breed2 and Breed3:</p>
<table class="doxtable">
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">breed1_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed1Parent&quot;</span>);</div>
<div class="line">breed2_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed2Parent&quot;</span>);</div>
<div class="line">breed3_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed3Parent&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][1];</div>
<div class="line">crossingPlan[0][0] = breed1_index;</div>
<div class="line">crossingPlan[1][0] = breed2_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1 = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
<div class="line"><span class="keywordtype">int</span> f1_index[1];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>(d, f1, 1, f1_index); <span class="comment">//we know this group has only one member</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Re-use crossingPlan</span></div>
<div class="line">crossingPlan[0][0] = breed3_index;</div>
<div class="line">crossingPlan[1][0] = f1_index[0];</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f3way = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f1 &lt;- make.targeted.crosses(c(&quot;Breed1Parent&quot;), c(&quot;Breed2Parent&quot;))</div>
<div class="line">f1_index &lt;- see.group.data(f1, &quot;X&quot;)</div>
<div class="line">f3way &lt;- make.targeted.crosses(c(&quot;Breed3Parent&quot;), f1_index)</div>
</div><!-- fragment --> </td></tr>
</table>
<p >And if you wanted more than one offspring from this three way cross:</p>
<table class="doxtable">
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">breed1_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed1Parent&quot;</span>);</div>
<div class="line">breed2_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed2Parent&quot;</span>);</div>
<div class="line">breed3_index = <a class="code hl_function" href="group__search.html#gad6c28080cf3e1f85fd78229aba53bf78">gsc_get_index_of_name</a>(d-&gt;m, <span class="stringliteral">&quot;Breed3Parent&quot;</span>);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlan[2][1];</div>
<div class="line">crossingPlan[0][0] = breed1_index;</div>
<div class="line">crossingPlan[1][0] = breed2_index;</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa320efa0931830d69965beb2f5399acf">GenOptions</a> opt = {.family_size=5,  <span class="comment">// &lt;-- changed from default</span></div>
<div class="line">        .will_name_offspring=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>, </div>
<div class="line">        .offspring_name_prefix=NULL,</div>
<div class="line">        .will_track_pedigree=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>, </div>
<div class="line">        .will_allocate_ids=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>,</div>
<div class="line">        .filename_prefix=NULL, </div>
<div class="line">        .will_save_pedigree_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_bvs_to_file=<a class="code hl_define" href="group__shortnames.html#ga27b7a0c46aa95c0f127549d8ae007b8d">NO_EFFECTSET</a>, </div>
<div class="line">        .will_save_alleles_to_file=<a class="code hl_define" href="group__shortnames.html#gaa93f0eb578d23995850d61f7d61c55c1">FALSE</a>,</div>
<div class="line">        .will_save_to_simdata=<a class="code hl_define" href="group__shortnames.html#gaa8cecfc5c5c054d2875c03e77b7be15d">TRUE</a>};</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f1 = <a class="code hl_define" href="group__shortnames.html#ga87aa2de59b68e428e5b6f07f5955c15a">make_random_crosses</a>(d, 1, crossingPlan[0], crossingPlan[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
<div class="line"><span class="keywordtype">int</span> f1_indexes[5];</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>(d, f1, 5, f1_indexes);</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> crossingPlanb[2][5];</div>
<div class="line">crossingPlanb[0][0] = breed3_index; crossingPlanb[0][1] = breed3_index;</div>
<div class="line">crossingPlanb[0][2] = breed3_index; crossingPlanb[0][3] = breed3_index;</div>
<div class="line">crossingPlanb[0][4] = breed3_index;</div>
<div class="line">crossingPlanb[1][0] = f1_indexes[0]; crossingPlanb[1][1] = f1_indexes[1];</div>
<div class="line">crossingPlanb[1][2] = f1_indexes[2]; crossingPlanb[1][3] = f1_indexes[3];</div>
<div class="line">crossingPlanb[1][4] = f1_indexes[4];</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> f3way = <a class="code hl_define" href="group__shortnames.html#ga56e71147b1f8a71151ef81b4a828aa9b">make_targeted_crosses</a>(d, 5, crossingPlanb[0], crossingPlanb[1], <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, <a class="code hl_define" href="group__shortnames.html#ga3710da75e403e5a3da4e2deade0d50af">NO_MAP</a>, opt);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">f1 &lt;- make.targeted.crosses(c(&quot;Breed1&quot;), c(&quot;Breed2&quot;), offspring=5)</div>
<div class="line"># This time, f1 is a group of multiple individuals</div>
<div class="line">f1_indexes &lt;- see.group.data(f1, &quot;X&quot;)</div>
<div class="line">f3way &lt;- make.targeted.crosses(rep(&quot;Breed3&quot;, times=length(f1_indexes)), f1_indexes, offspring=5)</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md57"></a>
Species with differing male and female recombination rates</h2>
<table class="doxtable">
<caption>Simulate crossing in a species with differing male and female recombination maps</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gaa9b2f128bb156d36d9334281b158c56a">SimData</a>* d = <a class="code hl_define" href="group__shortnames.html#gacbedd673ae83f70154d663852acfea70">create_empty_simdata</a>(7654321);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> female_map = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(<span class="stringliteral">&quot;fmap.txt&quot;</span>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gad11f85a6aae488f7efaab5abecaf1ac7">MapID</a> male_map = <a class="code hl_define" href="group__shortnames.html#gaf16559ebf0798beeaea167422915ece1">load_mapfile</a>(<span class="stringliteral">&quot;mmap.txt&quot;</span>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> female_pop = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(<span class="stringliteral">&quot;fgenos.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> male_pop = <a class="code hl_define" href="group__shortnames.html#ga23a7a92af469dfbba4905bb5c9c90a9b">load_genotypefile</a>(<span class="stringliteral">&quot;mgenos.txt&quot;</span>, <a class="code hl_define" href="group__shortnames.html#ga1b342f7327286629391ccbad90ea19e4">DETECT_FILE_FORMAT</a>);</div>
<div class="line"> </div>
<div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> offspring = <a class="code hl_define" href="group__shortnames.html#ga9d7d8919aa31f1cfb54c902800a2d063">make_random_crosses_between</a>(d, female_pop, male_pop, 1, 0, 0, female_map, male_map, <a class="code hl_define" href="group__shortnames.html#ga4af98a71b0736f81c32df8e5e4fd6a83">BASIC_OPT</a>);</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">female_map &lt;- load.map(&quot;fmap.txt&quot;)</div>
<div class="line">male_map &lt;- load.map(&quot;mmap.txt&quot;)</div>
<div class="line">female_pop &lt;- load.genotypes(&quot;fgenos.txt&quot;)</div>
<div class="line">male_pop &lt;- load.genotypes(&quot;mgenos.txt&quot;) </div>
<div class="line"> </div>
<div class="line">offspring &lt;- make.random.crosses.between(female_pop, male_pop, 1, map1=female_map, map2=male_map)</div>
</div><!-- fragment --> </td></tr>
</table>
<h1><a class="anchor" id="autotoc_md58"></a>
Selection</h1>
<p >Templates in this section assume you have a group <code>target</code> from which you want to select a subset of 'good' genotypes. Scripts are provided for selecting in different ways, showing some ways to extract information about the genotypes in the group, and showing how to select using this information using the function <code>make_group</code> (C)/<code>make.group</code> (R).</p>
<p >The C implementations in this section will be very naive/inefficient. The focus is on showing what kind of custom selection criteria can be implemented in genomicSimulation, not how best to sort/search through data efficiently in C. Sometimes, only R examples will be shown, because R's built-in table-manipulation functionalities make these tasks much simpler.</p>
<h2><a class="anchor" id="autotoc_md59"></a>
Manually select on true breeding value</h2>
<p >This is a selection function template. You can use the builtin functions <code>split_by_bv</code> (C)/ <code>break.group.by.gebv</code> (R) to carry out this exact task. It provides however a simple example of genomicSimulations "custom selection method interface", which involves accessing group data, choosing certain candidates based on that data, then calling <code>make_group</code> (C)/<code>make.group</code> (R) to move the selected candidates into a new group. This is the overall structure that the following sections will also use.</p>
<table class="doxtable">
<caption>Select the top 10 candidates in <code>target</code> according to their true breeding value as calculated by the internal breeding value calculator</caption>
<tr>
<th>genomicSimulationC (C) </th><th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line"><a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> select_10_with_best_GEBV(<a class="code hl_define" href="group__shortnames.html#gadb2ef9b75828b244e0b3e1cb3680d7ca">GroupNum</a> group, <a class="code hl_define" href="group__shortnames.html#ga516a882a4bac611d495b8f454478d674">EffectID</a> eff1) {</div>
<div class="line">  <span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> group_size = <a class="code hl_define" href="group__shortnames.html#ga0f4fcc00561afa1f820d7f11a02be1c7">get_group_size</a>( d, group );</div>
<div class="line">  </div>
<div class="line">  <span class="keywordtype">int</span> group_indexes[group_size]; <span class="comment">// VLA usage: swap for a heap array or a maximum group size if you don&#39;t like that</span></div>
<div class="line">  memset(group_indexes, 0, <span class="keyword">sizeof</span>(<span class="keywordtype">int</span>)*group_size);</div>
<div class="line">  <a class="code hl_define" href="group__shortnames.html#gaa7f7f1453a84746fa215fea171b80b97">get_group_indexes</a>( d, group, group_size, group_indexes);</div>
<div class="line">  </div>
<div class="line">  <span class="keywordtype">double</span> group_bvs[group_size];</div>
<div class="line">  memset(group_bvs, 0, <span class="keyword">sizeof</span>(<span class="keywordtype">double</span>)*group_size);</div>
<div class="line">  <a class="code hl_define" href="group__shortnames.html#gaa68a8ebf286aa6192948ab56851b125f">get_group_bvs</a>( d, group, eff1, group_size, group_bvs);</div>
<div class="line"> </div>
<div class="line"><span class="preprocessor">  # A pretty inefficient way of finding the top 10 scores and their corresponding</span></div>
<div class="line"><span class="preprocessor">  # indexes, but it will do for this example.</span></div>
<div class="line">  <span class="keywordtype">double</span> min_score = group_bvs[0];</div>
<div class="line">  <span class="keywordflow">for</span> (<span class="keywordtype">unsigned</span> <span class="keywordtype">int</span> i = 1; i &lt; group_size; ++i) {</div>
<div class="line">    <span class="keywordflow">if</span> (group_bvs[i] &lt; min_score) {</div>
<div class="line">      min_score = group_bvs[i];</div>
<div class="line">    }</div>
<div class="line">  }</div>
<div class="line"> </div>
<div class="line">  <span class="keywordtype">int</span> top_individuals[10];</div>
<div class="line">  <span class="keywordtype">int</span> currentTopIndex = 0;</div>
<div class="line">  <span class="keywordflow">for</span> (<span class="keywordtype">int</span> j = 0; j &lt; 10; ++j) {</div>
<div class="line">    currentTopIndex = 0;</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">int</span> i = 1; i &lt; group_size; ++i) {</div>
<div class="line">      <span class="keywordflow">if</span> (group_bvs[i] &gt; group_bvs[currentTopIndex]) {</div>
<div class="line">        currentTopIndex = i;</div>
<div class="line">      }</div>
<div class="line">    }</div>
<div class="line">    top_individuals[j] = group_indexes[currentTopIndex];</div>
<div class="line">    group_bvs[currentTopIndex] = min_score;</div>
<div class="line">  }</div>
<div class="line"> </div>
<div class="line">  <span class="keywordflow">return</span> <a class="code hl_define" href="group__shortnames.html#ga702855cf67ebd337941ec9c13e40dfe9">make_group_from</a>(d, 10, top_individuals);</div>
<div class="line">}</div>
</div><!-- fragment --> </td><td><div class="fragment"><div class="line">select.10.best.GEBV &lt;- function(group) {</div>
<div class="line">  info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                     GEBV=see.group.data(group,&quot;BV&quot;,eff.set=eff1))</div>
<div class="line"> </div>
<div class="line">  selected_group &lt;- make.group(info[order(info$GEBV, decreasing=TRUE),]$Index[1:10])</div>
<div class="line">  return( selected_group )</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md60"></a>
Select on phenotype (simulated with a given heritability)</h2>
<table class="doxtable">
<caption>Select the top 10 candidates in <code>target</code> according to a phenotype simulated with a certain $H^2$ broad-sense heritability value.</caption>
<tr>
<th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">select.top.10.phenotypes &lt;- function(group, heritability) {</div>
<div class="line">  info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                     GEBV=see.group.data(group,&quot;BV&quot;))</div>
<div class="line"> </div>
<div class="line">  # simulate phenotype = genotype + environmental variation</div>
<div class="line">  # using normally distributed Ve and heritability H^2 = Vg/(Vg + Ve)  </div>
<div class="line">  Vg &lt;- var(info$GEBV)</div>
<div class="line">  Ve &lt;- Vg/heritability - Vg</div>
<div class="line">  info$Pheno &lt;- info$GEBV + rnorm(length(info$GEBV), mean=0, sd = sqrt(Ve))</div>
<div class="line"> </div>
<div class="line">  # Select those with the top phenotype</div>
<div class="line">  selected &lt;- make.group(info[order(info$Pheno, decreasing=TRUE),]$Index[1:10])</div>
<div class="line">  return( selected )</div>
<div class="line"> </div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md61"></a>
Select on an index weighting two traits.</h2>
<table class="doxtable">
<caption>Select the top 10 candidates in <code>target</code> according to an index weighting two traits</caption>
<tr>
<th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">select.top.10.by.selectionindex &lt;- function(group, trait1.effset, trait2.effset, weighting=0.7) {</div>
<div class="line">  info &lt;- data.frame(PositionIndex=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                     BV1=see.group.data(group,&quot;BV&quot;, eff.set=trait1.effset),</div>
<div class="line">                     BV2=see.group.data(group,&quot;BV&quot;, eff.set=trait2.effset))</div>
<div class="line"> </div>
<div class="line">  # you could alter this formula as suits to calculate the selection index               </div>
<div class="line">  info$SelectionIndex &lt;- info$BV1 * weighting + info$BV2 * (1-weighting)</div>
<div class="line">  </div>
<div class="line">  selected &lt;- make.group(info[order(info$SelectionIndex, decreasing=TRUE),]$PositionIndex[1:10])</div>
<div class="line">  return( selected )</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md62"></a>
Select on a qualitative trait</h2>
<p >If you have a trait with only a small number of possible outcomes - perhaps a marker for which each individual can be homozygous, heterozygous, or carry no copies - and you want to select for individuals that carry that trait.</p>
<p >One option is to write a very small marker effect file, that looks something like: </p><div class="fragment"><div class="line">TARGETGENE T 1</div>
</div><!-- fragment --><p >You can load this effect file as in the earlier section <b>Load another set of additive trait effects</b>, and then use it with the function <code>split_by_bv</code> (C)/ <code>break.group.by.gebv</code> (R). Homozygous carriers will have a breeding value calculated as 2, heterozygous carriers will score 1, and non-carriers will score 0.</p>
<table class="doxtable">
<caption>Use a custom marker effect file to select on a qualitative trait</caption>
<tr>
<th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">qualtrait &lt;- load.effects(&quot;one_line_effectfile.txt&quot;)</div>
<div class="line"> </div>
<div class="line">select.top.10.qual.genotypes &lt;- function(group) {</div>
<div class="line">    return(break.group.by.GEBV(group, number=10, eff.set=qualtrait))</div>
<div class="line">}</div>
<div class="line"># or perhaps</div>
<div class="line">select.homozygous.carriers &lt;- function(group) {</div>
<div class="line">    info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                       GEBV=see.group.data(group,&quot;BV&quot;,eff.set=qualtrait))</div>
<div class="line">    </div>
<div class="line">    return(make.group(info$Index[info$GEBV == 2]))</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >Alternatively, you could read the genotype matrix manually:</p>
<table class="doxtable">
<caption>Access the genotype matrix to select on a qualitative trait</caption>
<tr>
<th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">select.homozygous.carriers &lt;- function(group, allele=&#39;T&#39;) {</div>
<div class="line">    counts &lt;- see.group.gene.data(group, count.allele=allele)</div>
<div class="line">    selected_names &lt;- colnames(counts)[counts[&quot;TARGETGENE&quot;,]==2]</div>
<div class="line">    rm(counts)</div>
<div class="line">    </div>
<div class="line">    info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                       Names=see.group.data(group,&quot;N&quot;))</div>
<div class="line">    return(make.group(info$Index[info$Names %in% selected_names]))</div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line"># With this method, you can also access phase information.</div>
<div class="line"># Say you wanted to select carriers of that &quot;T&quot; allele</div>
<div class="line"># who inherited the allele from their mother/first parent</div>
<div class="line">select.carriers.inheriting.from.parent1 &lt;- function(group) {</div>
<div class="line">    genotypes &lt;- see.group.gene.data(group)</div>
<div class="line">    selected_names &lt;- colnames(genotypes)[genotypes[&quot;TARGETGENE&quot;,] %in% c(&quot;TA&quot;, &quot;TT&quot;)]</div>
<div class="line">    rm(genotypes)</div>
<div class="line">    </div>
<div class="line">    info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                       Names=see.group.data(group,&quot;N&quot;))</div>
<div class="line">    return(make.group(info$Index[info$Names %in% selected_names]))</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<h2><a class="anchor" id="autotoc_md63"></a>
Select on a qualitative trait, then a quantitative trait.</h2>
<p >Suppose you have an effect file for a quantitative trait, called <code>eff-large.txt</code>, as well as a one-line effect file for a qualitative trait, as described in the above section <b>Select on a qualitative trait</b>, named <code>eff-small.txt</code>.</p>
<p >You would like to select primarily on the allele listed in <code>eff-small.txt</code>, and secondarily select on the quantitative trait. This means you would like your selection to be wholly homozygous for the allele from <code>eff-small.txt</code>, if possible, and if you have enough homozygous candidates in the population, the selection should be the homozygous candidates with the highest scores for the quantitative trait. If your population does not have enough homozygous individuals to fill up your selection, then you will take all the homozygous individuals and the best of the heterozygous individuals. And only if you do not have enough homozygotes and heterozygotes will you add the best of the non-carriers to your selection.</p>
<table class="doxtable">
<caption>Use two marker effect files, select first for carriers of a specific allele, then on a quantitative trait</caption>
<tr>
<th>genomicSimulation (R) </th></tr>
<tr>
<td><div class="fragment"><div class="line">quanttrait &lt;- load.effects(&quot;eff-large.txt&quot;)</div>
<div class="line">qualtrait &lt;- load.effects(&quot;eff-small.txt&quot;) # has one line and will return scores of 0, 1, or 2</div>
<div class="line"># Instead of the &quot;eff-small.txt&quot; file, you could also use a see.group.gene.data alternative method to calculate info$Copies, as discussed in the previous section.</div>
<div class="line"> </div>
<div class="line">select.n.best.carriers &lt;- function(population, n) {</div>
<div class="line">    info &lt;- data.frame(Index=see.group.data(group,&quot;X&quot;),</div>
<div class="line">                       BV=see.group.data(group,&quot;BV&quot;,eff.set=quanttrait),</div>
<div class="line">                       Copies=see.group.data(group,&quot;BV&quot;,eff.set=qualtrait))</div>
<div class="line"> </div>
<div class="line">    take.all.or.best.n &lt;- function(infosubset, n) {</div>
<div class="line">        if (n &lt;= 0) { return(c()) }</div>
<div class="line">        if (n &gt;= dim(infosubset)[1]) {</div>
<div class="line">            return(infosubset$Index)</div>
<div class="line">        } else { # rank and return best n</div>
<div class="line">            return(info[order(-info$BV),]$Index[1:n]</div>
<div class="line">        }</div>
<div class="line">    }</div>
<div class="line">    </div>
<div class="line">    will_be_selected &lt;- take.all.or.best.n(info[info$Copies == 2,], n)</div>
<div class="line">    will_be_selected &lt;- take.all.or.best.n(info[info$Copies == 1,], n-length(will_be_selected))</div>
<div class="line">    will_be_selected &lt;- take.all.or.best.n(info[info$Copies == 0,], n-length(will_be_selected))</div>
<div class="line"> </div>
<div class="line">    return(make.group(will_be_selected))</div>
<div class="line">}</div>
</div><!-- fragment --> </td></tr>
</table>
<p >(this list of examples will continue to be expanded...) </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.3
</small></address>
</body>
</html>
